##########################################################
# Summarization and Visualization of Dynamic
# Window Analysis Results using STAARpipelineSummary
# Xihao Li, Zilin Li
# Initiate date: 11/04/2021
# Current date: 03/10/2022
##########################################################
rm(list=ls())
gc()


Dynamic_Window_Results_Summary <- function(agds_dir,jobs_num,input_path,output_path,dynamic_window_results_name,
                                           obj_nullmodel,known_loci=NULL,
                                           method_cond=c("optimal","naive"),
                                           QC_label="annotation/filter",geno_missing_imputation=c("mean","minor"),variant_type=c("SNV","Indel","variant"),
                                           Annotation_dir="annotation/info/FunctionalAnnotation",Annotation_name_catalog,
                                           Use_annotation_weights=FALSE,Annotation_name=NULL,alpha=0.05){
  
  ## evaluate choices
  method_cond <- match.arg(method_cond)
  variant_type <- match.arg(variant_type)
  geno_missing_imputation <- match.arg(geno_missing_imputation)
  
  ##### SCANG_O
  SCANG_O_res <- c()
  SCANG_O_top1 <- c()
  SCANG_O_emthr <- c()
  
  ##### SCANG_S
  SCANG_S_res <- c()
  SCANG_S_top1 <- c()
  SCANG_S_emthr <- c()
  
  ##### SCANG_B
  SCANG_B_res <- c()
  SCANG_B_top1 <- c()
  SCANG_B_emthr <- c()
  
  for(i in 1:sum(jobs_num$scang_num))
  {
    print(i)
    results_scang <- get(load(paste0(input_path,dynamic_window_results_name,"_",i,".Rdata")))
    
    # SCANG-O
    SCANG_O_emthr <- rbind(SCANG_O_emthr,results_scang$SCANG_O_emthr)
    SCANG_O_res <- rbind(SCANG_O_res,results_scang$SCANG_O_sig)
    SCANG_O_top1 <- rbind(SCANG_O_top1,results_scang$SCANG_O_top1)
    
    # SCANG-S
    SCANG_S_emthr <- rbind(SCANG_S_emthr,results_scang$SCANG_S_emthr)
    SCANG_S_res <- rbind(SCANG_S_res,results_scang$SCANG_S_sig)
    SCANG_S_top1 <- rbind(SCANG_S_top1,results_scang$SCANG_S_top1)
    
    # SCANG-B
    SCANG_B_emthr <- rbind(SCANG_B_emthr,results_scang$SCANG_B_emthr)
    SCANG_B_res <- rbind(SCANG_B_res,results_scang$SCANG_B_sig)
    SCANG_B_top1 <- rbind(SCANG_B_top1,results_scang$SCANG_B_top1)
  }
  
  colnames(SCANG_O_res) <- colnames(SCANG_O_top1)
  colnames(SCANG_S_res) <- colnames(SCANG_S_top1)
  colnames(SCANG_B_res) <- colnames(SCANG_B_top1)
  
  ############ SCANG-O
  # empirical threshold
  emthr_genome <- apply(SCANG_O_emthr,2,max)
  
  thr <- quantile(emthr_genome,1-alpha)
  SCANG_O_res <- SCANG_O_res[SCANG_O_res[,1]>=thr,,drop=FALSE]
  
  if(length(SCANG_O_res)>=6)
  {
    for(i in 1:dim(SCANG_O_res)[1])
    {
      SCANG_O_res[i,5] <- mean(emthr_genome>SCANG_O_res[i,1])
    }
    SCANG_O_res <- SCANG_O_res[order(SCANG_O_res[,2]*1e10+SCANG_O_res[,3]),]
  }
  
  
  SCANG_O_top1 <- SCANG_O_top1[which.max(SCANG_O_top1[,1]),]
  SCANG_O_top1[5] <- mean(emthr_genome>SCANG_O_top1[1])
  
  thr_SCANG_O <- thr
  emthr_genome_SCANG_O <- emthr_genome
  
  ############ SCANG-S
  # empirical threshold
  emthr_genome <- apply(SCANG_S_emthr,2,max)
  
  thr <- quantile(emthr_genome,1-alpha)
  SCANG_S_res <- SCANG_S_res[SCANG_S_res[,1]>=thr,,drop=FALSE]
  
  if(length(SCANG_S_res)>=6)
  {
    for(i in 1:dim(SCANG_S_res)[1])
    {
      SCANG_S_res[i,5] <- mean(emthr_genome>SCANG_S_res[i,1])
    }
    SCANG_S_res <- SCANG_S_res[order(SCANG_S_res[,2]*1e10+SCANG_S_res[,3]),]
  }
  
  
  SCANG_S_top1 <- SCANG_S_top1[which.max(SCANG_S_top1[,1]),]
  SCANG_S_top1[5] <- mean(emthr_genome>SCANG_S_top1[1])
  
  thr_SCANG_S <- thr
  emthr_genome_SCANG_S <- emthr_genome
  
  
  ############ SCANG-B
  # empirical threshold
  emthr_genome <- apply(SCANG_B_emthr,2,max)
  
  thr <- quantile(emthr_genome,1-alpha)
  SCANG_B_res <- SCANG_B_res[SCANG_B_res[,1]>=thr,,drop=FALSE]
  
  if(length(SCANG_B_res)>=6)
  {
    for(i in 1:dim(SCANG_B_res)[1])
    {
      SCANG_B_res[i,5] <- mean(emthr_genome>SCANG_B_res[i,1])
    }
    SCANG_B_res <- SCANG_B_res[order(SCANG_B_res[,2]*1e10+SCANG_B_res[,3]),]
  }
  
  
  SCANG_B_top1 <- SCANG_B_top1[which.max(SCANG_B_top1[,1]),]
  SCANG_B_top1[5] <- mean(emthr_genome>SCANG_B_top1[1])
  
  thr_SCANG_B <- thr
  emthr_genome_SCANG_B <- emthr_genome
  
  SCANG_res <- list(SCANG_B_top1 = SCANG_B_top1, SCANG_B_emthr = emthr_genome_SCANG_B, SCANG_B_sig = SCANG_B_res,
                    SCANG_S_top1 = SCANG_S_top1, SCANG_S_emthr = emthr_genome_SCANG_S, SCANG_S_sig = SCANG_S_res,
                    SCANG_O_top1 = SCANG_O_top1, SCANG_O_emthr = emthr_genome_SCANG_O, SCANG_O_sig = SCANG_O_res,alpha=alpha)
  
  # Output path
  save(SCANG_res,file=paste0(output_path,"results_dynamic_window.Rdata"))
  
  # SCANG_S
  save(SCANG_S_res,file=paste0(output_path,"SCANG_S_res.Rdata"))
  write.csv(SCANG_S_res,paste0(output_path,"SCANG_S_res.csv"))
  # SCANG_O
  save(SCANG_O_res,file=paste0(output_path,"SCANG_O_res.Rdata"))
  write.csv(SCANG_O_res,paste0(output_path,"SCANG_O_res.csv"))
  # SCANG_B
  save(SCANG_B_res,file=paste0(output_path,"SCANG_B_res.Rdata"))
  write.csv(SCANG_B_res,paste0(output_path,"SCANG_B_res.csv"))
  
  ### TOP 1 Results
  # SCANG_S
  SCANG_S_top1 <- SCANG_res$SCANG_S_top1
  SCANG_S_top1[1] <- exp(-SCANG_S_top1[1])
  SCANG_S_top1 <- matrix(SCANG_S_top1,nrow=1)
  colnames(SCANG_S_top1) <- c("STAAR_S","chr","start_pos","end_pos","GWER","SNV_nos")
  
  save(SCANG_S_top1,file=paste0(output_path,"SCANG_S_top1.Rdata"))
  write.csv(SCANG_S_top1,paste0(output_path,"SCANG_S_top1.csv"))
  
  # SCANG_B
  SCANG_B_top1 <- SCANG_res$SCANG_B_top1
  SCANG_B_top1[1] <- exp(-SCANG_B_top1[1])
  SCANG_B_top1 <- matrix(SCANG_B_top1,nrow=1)
  colnames(SCANG_B_top1) <- c("STAAR_B","chr","start_pos","end_pos","GWER","SNV_nos")
  
  save(SCANG_B_top1,file=paste0(output_path,"SCANG_B_top1.Rdata"))
  write.csv(SCANG_B_top1,paste0(output_path,"SCANG_B_top1.csv"))
  
  # SCANG_O
  SCANG_O_top1 <- SCANG_res$SCANG_O_top1
  SCANG_O_top1[1] <- exp(-SCANG_O_top1[1])
  SCANG_O_top1 <- matrix(SCANG_O_top1,nrow=1)
  colnames(SCANG_O_top1) <- c("SCANG_STAAR_O","chr","start_pos","end_pos","GWER","SNV_nos")
  
  save(SCANG_O_top1,file=paste0(output_path,"SCANG_O_top1.Rdata"))
  write.csv(SCANG_O_top1,paste0(output_path,"SCANG_O_top1.csv"))
  
  
  ########################################################################
  #                    Conditional Analysis
  ########################################################################
  if(length(known_loci)!=0)
  {
    ################### SCANG-S
    SCANG_S_res <- SCANG_res$SCANG_S_sig
    
    if(class(SCANG_S_res)[1]!="matrix")
    {
      SCANG_S_res <- matrix(SCANG_S_res,nrow=1)
    }
    
    SCANG_S_res_cond <- c()
    if(length(SCANG_S_res)!=0)
    {
      for(kk in 1:dim(SCANG_S_res)[1])
      {
        chr <- as.numeric(SCANG_S_res[kk,2])
        start_loc <- as.numeric(SCANG_S_res[kk,3])
        end_loc <- as.numeric(SCANG_S_res[kk,4])
        
        gds.path <- agds_dir[chr]
        genofile <- seqOpen(gds.path)
        
        res_cond <- Sliding_Window_cond(chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,
                                        start_loc=start_loc,end_loc=end_loc,known_loci=known_loci,method_cond=method_cond,
                                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,
                                        Annotation_name_catalog=Annotation_name_catalog,Annotation_dir=Annotation_dir,
                                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)
        
        SCANG_S_res_cond <- rbind(SCANG_S_res_cond,res_cond)
        
        seqClose(genofile)
      }
    }
    
    save(SCANG_S_res_cond,file=paste0(output_path,"SCANG_S_res_cond.Rdata"))
    write.csv(SCANG_S_res_cond,paste0(output_path,"SCANG_S_res_cond.csv"))
    
    ### Uncond_Cond
    SCANG_S_res_uncond_cond <- c()
    if(length(SCANG_S_res)!=0)
    {
      ## Uncond
      SCANG_S_res[,1] <- exp(-SCANG_S_res[,1])
      colnames(SCANG_S_res) <- c("STAAR_S","chr","start_pos","end_pos","GWER","SNV_nos")
      SCANG_S_res <- SCANG_S_res[,c(2:4,6,5,1),drop=FALSE]
      
      ## SCANG_S_cond
      STAAR_S <- SCANG_S_res_cond[,c("STAAR-S(1,25)","STAAR-S(1,1)"),drop=FALSE]
      STAAR_S <- as.matrix(STAAR_S)
      STAAR_S_cond <- apply(STAAR_S,1,function(z) CCT(as.numeric(z)))
      
      STAAR_cond <- SCANG_S_res_cond[,c("ACAT-V(1,25)","Burden(1,1)","SKAT(1,25)","STAAR-O"),drop=FALSE]
      
      ## Uncond_Cond
      SCANG_S_res_uncond_cond <- cbind(SCANG_S_res,STAAR_S_cond)
      SCANG_S_res_uncond_cond <- cbind(SCANG_S_res_uncond_cond,STAAR_cond)
      colnames(SCANG_S_res_uncond_cond)[(dim(SCANG_S_res_uncond_cond)[2]-3):dim(SCANG_S_res_uncond_cond)[2]] <- c("ACAT_V_cond","Burden_cond","SKAT_cond","STAAR_O_cond")
    }
    
    save(SCANG_S_res_uncond_cond,file=paste0(output_path,"SCANG_S_res_uncond_cond.Rdata"))
    write.csv(SCANG_S_res_uncond_cond,paste0(output_path,"SCANG_S_res_uncond_cond.csv"))
    
    
    ######################## SCANG-O
    SCANG_O_res <- SCANG_res$SCANG_O_sig
    
    if(class(SCANG_O_res)[1]!="matrix")
    {
      SCANG_O_res <- matrix(SCANG_O_res,nrow=1)
    }
    
    SCANG_O_res_cond <- c()
    if(length(SCANG_O_res)!=0)
    {
      for(kk in 1:dim(SCANG_O_res)[1])
      {
        chr <- as.numeric(SCANG_O_res[kk,2])
        start_loc <- as.numeric(SCANG_O_res[kk,3])
        end_loc <- as.numeric(SCANG_O_res[kk,4])
        
        gds.path <- agds_dir[chr]
        genofile <- seqOpen(gds.path)
        
        res_cond <- Sliding_Window_cond(chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,
                                        start_loc=start_loc,end_loc=end_loc,known_loci=known_loci,method_cond=method_cond,
                                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,
                                        Annotation_name_catalog=Annotation_name_catalog,Annotation_dir=Annotation_dir,
                                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)
        
        SCANG_O_res_cond <- rbind(SCANG_O_res_cond,res_cond)
        
        seqClose(genofile)
      }
    }
    
    save(SCANG_O_res_cond,file=paste0(output_path,"SCANG_O_res_cond.Rdata"))
    write.csv(SCANG_O_res_cond,paste0(output_path,"SCANG_O_res_cond.csv"))
    
    ### Uncond_Cond
    SCANG_O_res_uncond_cond <- c()
    if(length(SCANG_O_res)!=0)
    {
      ## Uncond
      SCANG_O_res[,1] <- exp(-SCANG_O_res[,1])
      colnames(SCANG_O_res) <- c("SCANG_STAAR_O","chr","start_pos","end_pos","GWER","SNV_nos")
      SCANG_O_res <- SCANG_O_res[,c(2:4,6,5,1),drop=FALSE]
      
      ## SCANG_O_cond
      STAAR_O <- SCANG_O_res_cond[,c("STAAR-S(1,25)","STAAR-S(1,1)","STAAR-B(1,25)","STAAR-B(1,1)"),drop=FALSE]
      STAAR_O <- as.matrix(STAAR_O)
      SCANG_STAAR_O_cond <- apply(STAAR_O,1,function(z) CCT(as.numeric(z)))
      
      STAAR_cond <- SCANG_O_res_cond[,c("ACAT-V(1,25)","Burden(1,1)","SKAT(1,25)","STAAR-O"),drop=FALSE]
      
      SCANG_O_res_uncond_cond <- cbind(SCANG_O_res,SCANG_STAAR_O_cond)
      SCANG_O_res_uncond_cond <- cbind(SCANG_O_res_uncond_cond,STAAR_cond)
      colnames(SCANG_O_res_uncond_cond)[(dim(SCANG_O_res_uncond_cond)[2]-3):dim(SCANG_O_res_uncond_cond)[2]] <- c("ACAT_V_cond","Burden_cond","SKAT_cond","STAAR_O_cond")
    }
    
    save(SCANG_O_res_uncond_cond,file=paste0(output_path,"SCANG_O_res_uncond_cond.Rdata"))
    write.csv(SCANG_O_res_uncond_cond,paste0(output_path,"SCANG_O_res_uncond_cond.csv"))
    
    
    ######################## SCANG-B
    SCANG_B_res <- SCANG_res$SCANG_B_sig
    
    if(class(SCANG_B_res)[1]!="matrix")
    {
      SCANG_B_res <- matrix(SCANG_B_res,nrow=1)
    }
    
    SCANG_B_res_cond <- c()
    if(length(SCANG_B_res)!=0)
    {
      for(kk in 1:dim(SCANG_B_res)[1])
      {
        chr <- as.numeric(SCANG_B_res[kk,2])
        start_loc <- as.numeric(SCANG_B_res[kk,3])
        end_loc <- as.numeric(SCANG_B_res[kk,4])
        
        gds.path <- agds_dir[chr]
        genofile <- seqOpen(gds.path)
        
        res_cond <- Sliding_Window_cond(chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,
                                        start_loc=start_loc,end_loc=end_loc,known_loci=known_loci,method_cond=method_cond,
                                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,
                                        Annotation_name_catalog=Annotation_name_catalog,Annotation_dir=Annotation_dir,
                                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)
        
        SCANG_B_res_cond <- rbind(SCANG_B_res_cond,res_cond)
        
        seqClose(genofile)
      }
    }
    
    save(SCANG_B_res_cond,file=paste0(output_path,"SCANG_B_res_cond.Rdata"))
    write.csv(SCANG_B_res_cond,paste0(output_path,"SCANG_B_res_cond.csv"))
    
    ### Uncond_Cond
    SCANG_B_res_uncond_cond <- c()
    if(length(SCANG_B_res)!=0)
    {
      ## Uncond
      SCANG_B_res[,1] <- exp(-SCANG_B_res[,1])
      colnames(SCANG_B_res) <- c("STAAR_B","chr","start_pos","end_pos","GWER","SNV_nos")
      SCANG_B_res <- SCANG_B_res[,c(2:4,6,5,1),drop=FALSE]
      
      ## SCANG_B_cond
      STAAR_B <- SCANG_B_res_cond[,c("STAAR-B(1,25)","STAAR-B(1,1)"),drop=FALSE]
      STAAR_B <- as.matrix(STAAR_B)
      STAAR_B_cond <- apply(STAAR_B,1,function(z) CCT(as.numeric(z)))
      
      STAAR_cond <- SCANG_B_res_cond[,c("ACAT-V(1,25)","Burden(1,1)","SKAT(1,25)","STAAR-O"),drop=FALSE]
      
      SCANG_B_res_uncond_cond <- cbind(SCANG_B_res,STAAR_B_cond)
      SCANG_B_res_uncond_cond <- cbind(SCANG_B_res_uncond_cond,STAAR_cond)
      colnames(SCANG_B_res_uncond_cond)[(dim(SCANG_B_res_uncond_cond)[2]-3):dim(SCANG_B_res_uncond_cond)[2]] <- c("ACAT_V_cond","Burden_cond","SKAT_cond","STAAR_O_cond")
    }
    
    save(SCANG_B_res_uncond_cond,file=paste0(output_path,"SCANG_B_res_uncond_cond.Rdata"))
    write.csv(SCANG_B_res_uncond_cond,paste0(output_path,"SCANG_B_res_uncond_cond.csv"))
  }
}




## load required packages
library(gdsfmt)
library(SeqArray)
library(SeqVarTools)
library(STAAR)
library(STAARpipeline)
library(STAARpipelineSummary)

###########################################################
#           User Input
###########################################################
## Number of jobs for each chromosome
jobs_num <- get(load("/data/williamsjacr/UKB_WES_lipids/Data/agds/train_jobs_num.Rdata"))
## aGDS directory
agds_dir <- get(load("/data/williamsjacr/UKB_WES_lipids/Data/agds/train_agds_dir.Rdata"))
## Null model
obj_nullmodel_SCANG_STAAR <- get(load("/data/williamsjacr/UKB_WES_lipids/Data/nullmodels_staar/Train_Null_Model_SCANG_LDL.RData"))
## Known loci
known_loci <- NULL

## results path
input_path <- "/data/williamsjacr/UKB_WES_lipids/Data/Results/LDL/DynamicWindow/"
output_path <- input_path
## results name
dynamic_window_results_name <- "UKBB_WES_LDL_Dynamic_Train"

## QC_label
QC_label <- "annotation/info/QC_label"
## variant_type
variant_type <- "SNV"
## geno_missing_imputation
geno_missing_imputation <- "mean"
## method_cond
method_cond <- "optimal"
## alpha level
alpha <- 1

## Annotation_dir
Annotation_dir <- "annotation/info/FunctionalAnnotation/FunctionalAnnotation"
## Annotation channel
Annotation_name_catalog <- get(load("/data/williamsjacr/UKB_WES_lipids/Data/agds/train_Annotation_name_catalog.Rdata"))
## Use_annotation_weights
Use_annotation_weights <- TRUE
## Annotation name
Annotation_name <- c("CADD","LINSIGHT","FATHMM.XF","aPC.EpigeneticActive","aPC.EpigeneticRepressed","aPC.EpigeneticTranscription",
                     "aPC.Conservation","aPC.LocalDiversity","aPC.Mappability","aPC.TF","aPC.Protein")

###########################################################
#           Main Function 
###########################################################

# ##### SCANG_O
# SCANG_O_res <- c()
# SCANG_O_top1 <- c()
# SCANG_O_emthr <- c()
# 
# ##### SCANG_S
# SCANG_S_res <- c()
# SCANG_S_top1 <- c()
# SCANG_S_emthr <- c()
# 
# ##### SCANG_B
# SCANG_B_res <- c()
# SCANG_B_top1 <- c()
# SCANG_B_emthr <- c()
# 
# for(i in 1:sum(jobs_num$scang_num))
# {
#   print(i)
#   results_scang <- get(load(paste0(input_path,dynamic_window_results_name,"_",i,".Rdata")))
#   
#   # SCANG-O
#   SCANG_O_emthr <- rbind(SCANG_O_emthr,results_scang$SCANG_O_emthr)
#   SCANG_O_res <- rbind(SCANG_O_res,results_scang$SCANG_O_sig)
#   SCANG_O_top1 <- rbind(SCANG_O_top1,results_scang$SCANG_O_top1)
#   
#   # SCANG-S
#   SCANG_S_emthr <- rbind(SCANG_S_emthr,results_scang$SCANG_S_emthr)
#   SCANG_S_res <- rbind(SCANG_S_res,results_scang$SCANG_S_sig)
#   SCANG_S_top1 <- rbind(SCANG_S_top1,results_scang$SCANG_S_top1)
#   
#   # SCANG-B
#   SCANG_B_emthr <- rbind(SCANG_B_emthr,results_scang$SCANG_B_emthr)
#   SCANG_B_res <- rbind(SCANG_B_res,results_scang$SCANG_B_sig)
#   SCANG_B_top1 <- rbind(SCANG_B_top1,results_scang$SCANG_B_top1)
# }
# 
# colnames(SCANG_O_res) <- colnames(SCANG_O_top1)
# colnames(SCANG_S_res) <- colnames(SCANG_S_top1)
# colnames(SCANG_B_res) <- colnames(SCANG_B_top1)
# 
# 
# # empirical threshold
# emthr_genome <- apply(SCANG_O_emthr,2,max)
# 
# thr <- quantile(emthr_genome,1-alpha)
# SCANG_O_res <- SCANG_O_res[SCANG_O_res[,1]>=thr,,drop=FALSE]


Dynamic_Window_Results_Summary(agds_dir=agds_dir,jobs_num=jobs_num,
                               input_path=input_path,output_path=output_path,dynamic_window_results_name=dynamic_window_results_name,
                               obj_nullmodel=obj_nullmodel,known_loci=known_loci,
                               method_cond=method_cond,
                               QC_label=QC_label,geno_missing_imputation=geno_missing_imputation,variant_type=variant_type,
                               Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,
                               Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,
                               alpha=alpha)