{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"rm -r Results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"mkdir Results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"gsutil cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv Results/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "a <- read.csv(\"Results/all_phenotypes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"gsutil cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv Results/\")\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    system(paste0(\"gsutil cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\",trait,\"_BestPRS.csv Results/\"))\n",
    "    system(paste0(\"gsutil cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",trait,\"_Best_Validation_All.txt Results/\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'all_phenotypes.csv'</li><li>'BMI_Best_Validation_All.txt'</li><li>'BMI_BestPRS.csv'</li><li>'HDL_Best_Validation_All.txt'</li><li>'HDL_BestPRS.csv'</li><li>'Height_Best_Validation_All.txt'</li><li>'Height_BestPRS.csv'</li><li>'LDL_Best_Validation_All.txt'</li><li>'LDL_BestPRS.csv'</li><li>'logTG_Best_Validation_All.txt'</li><li>'logTG_BestPRS.csv'</li><li>'TC_Best_Validation_All.txt'</li><li>'TC_BestPRS.csv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'all\\_phenotypes.csv'\n",
       "\\item 'BMI\\_Best\\_Validation\\_All.txt'\n",
       "\\item 'BMI\\_BestPRS.csv'\n",
       "\\item 'HDL\\_Best\\_Validation\\_All.txt'\n",
       "\\item 'HDL\\_BestPRS.csv'\n",
       "\\item 'Height\\_Best\\_Validation\\_All.txt'\n",
       "\\item 'Height\\_BestPRS.csv'\n",
       "\\item 'LDL\\_Best\\_Validation\\_All.txt'\n",
       "\\item 'LDL\\_BestPRS.csv'\n",
       "\\item 'logTG\\_Best\\_Validation\\_All.txt'\n",
       "\\item 'logTG\\_BestPRS.csv'\n",
       "\\item 'TC\\_Best\\_Validation\\_All.txt'\n",
       "\\item 'TC\\_BestPRS.csv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'all_phenotypes.csv'\n",
       "2. 'BMI_Best_Validation_All.txt'\n",
       "3. 'BMI_BestPRS.csv'\n",
       "4. 'HDL_Best_Validation_All.txt'\n",
       "5. 'HDL_BestPRS.csv'\n",
       "6. 'Height_Best_Validation_All.txt'\n",
       "7. 'Height_BestPRS.csv'\n",
       "8. 'LDL_Best_Validation_All.txt'\n",
       "9. 'LDL_BestPRS.csv'\n",
       "10. 'logTG_Best_Validation_All.txt'\n",
       "11. 'logTG_BestPRS.csv'\n",
       "12. 'TC_Best_Validation_All.txt'\n",
       "13. 'TC_BestPRS.csv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"all_phenotypes.csv\"             \"BMI_Best_Validation_All.txt\"   \n",
       " [3] \"BMI_BestPRS.csv\"                \"HDL_Best_Validation_All.txt\"   \n",
       " [5] \"HDL_BestPRS.csv\"                \"Height_Best_Validation_All.txt\"\n",
       " [7] \"Height_BestPRS.csv\"             \"LDL_Best_Validation_All.txt\"   \n",
       " [9] \"LDL_BestPRS.csv\"                \"logTG_Best_Validation_All.txt\" \n",
       "[11] \"logTG_BestPRS.csv\"              \"TC_Best_Validation_All.txt\"    \n",
       "[13] \"TC_BestPRS.csv\"                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.files(\"Results/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/home/jupyter/packages’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"ggthemes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(IID)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(IID)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 6 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(IID)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 10 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(IID)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 6 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(IID)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 5 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(IID)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(Common_Bin, Rare_Bin)`\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2 rows containing missing values or values outside the scale range\n",
      "(`geom_segment()`).”\n"
     ]
    }
   ],
   "source": [
    "rm(list = ls())\n",
    "\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "\n",
    "theme_Publication <- function(base_size=12) {\n",
    "  library(grid)\n",
    "  library(ggthemes)\n",
    "  (theme_foundation(base_size=base_size, )\n",
    "    + theme(plot.title = element_text(face = \"bold\",\n",
    "                                      size = rel(1.1), hjust = 0.5),\n",
    "            text = element_text(),\n",
    "            panel.background = element_rect(colour = NA),\n",
    "            plot.background = element_rect(colour = NA),\n",
    "            panel.border = element_rect(colour = NA),\n",
    "            axis.title = element_text(face = \"bold\",size = 24),\n",
    "            strip.text.x = element_text(face = \"bold\",size = 24),\n",
    "            axis.title.y = element_text(angle=90,vjust =2),\n",
    "            axis.line = element_line(colour=\"black\",size=2),\n",
    "            axis.text.y = element_text(size = 20),\n",
    "            axis.text.x = element_text(size = 13),\n",
    "            axis.ticks = element_line(),\n",
    "            # panel.grid.major = element_line(colour=\"#f0f0f0\"),\n",
    "            # panel.grid.minor = element_line(colour=\"#f0f0f0\"),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor = element_blank(),\n",
    "            legend.key = element_rect(colour = NA),\n",
    "            legend.position = \"bottom\",\n",
    "            legend.text=element_text(size=20),\n",
    "            #legend.direction = \"horizontal\",\n",
    "            #legend.key.size= unit(0.2, \"cm\"),\n",
    "            #legend.margin = unit(0, \"cm\"),\n",
    "            legend.title = element_text(face=\"bold.italic\", size =24),\n",
    "            #legend.text = element_text(face =\"bold\"),\n",
    "            plot.margin=unit(c(10,5,5,5),\"mm\"),\n",
    "            strip.background=element_rect(colour=\"#f0f0f0\",fill=\"#f0f0f0\"),\n",
    "            strip.text = element_text(face=\"bold\")\n",
    "    ))\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "scale_fill_Publication <- function(...){\n",
    "  library(scales)\n",
    "  discrete_scale(\"fill\",\"Publication\",manual_pal(values = c(\"#386cb0\",\"#EF7E3D\",\"#ffd558\",\"#7fc97f\",\"#ef3b2c\",\"#662506\",\"#a6cee3\",\"#fb9a99\",\"#984ea3\",\"#ffff33\")), ...)\n",
    "  \n",
    "}\n",
    "\n",
    "scale_colour_Publication <- function(...){\n",
    "  library(scales)\n",
    "  discrete_scale(\"colour\",\"Publication\",manual_pal(values = c(\"#386cb0\",\"#EF7E3D\",\"#ffd558\",\"#7fc97f\",\"#ef3b2c\",\"#662506\",\"#a6cee3\",\"#fb9a99\",\"#984ea3\",\"#ffff33\")), ...)\n",
    "  \n",
    "}\n",
    "\n",
    "trait <- \"Height\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"HDL\",\"LDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "  RV_PRS <- read.csv(paste0(\"Results/\",trait,\"_BestPRS.csv\"))\n",
    "  CV_PRS <- read.delim(paste0(\"Results/\",trait,\"_Best_Validation_All.txt\"))\n",
    "  \n",
    "  CV_RV_PRS <- inner_join(RV_PRS,CV_PRS)\n",
    "  CV_RV_PRS_adjusted <- CV_RV_PRS\n",
    "  \n",
    "  for(i in c(\"RV_PRS\",\"prs\")){\n",
    "    tmp <- data.frame(y = CV_RV_PRS_adjusted[,i],CV_RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "    mod <- lm(y~.,data = tmp)\n",
    "    R <- mod$residuals\n",
    "    tmp <- data.frame(y = R^2,CV_RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "    mod <- lm(y~.,data = tmp)\n",
    "    y_hat <- predict(mod,tmp)\n",
    "    if(sum(y_hat < 0) > 0){\n",
    "      mod <- lm(y~1,data = tmp)\n",
    "      y_hat <- predict(mod,tmp)\n",
    "    }\n",
    "    if(sum(sqrt(y_hat)) == 0){\n",
    "      CV_RV_PRS_adjusted[,i] <- 0\n",
    "    }else{\n",
    "      CV_RV_PRS_adjusted[,i] <- R/sqrt(y_hat)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  \n",
    "  CV_RV_PRS_adjusted$Common_Bin <- 9\n",
    "  CV_RV_PRS_adjusted$Rare_Bin <- 9\n",
    "  \n",
    "  Common_quants <- quantile(CV_RV_PRS_adjusted$prs,c(0,.1,.2,.3,.4,.6,.7,.8,.9))\n",
    "  \n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted$Common_Bin[CV_RV_PRS_adjusted$prs < unname(Common_quants[i + 1]) & CV_RV_PRS_adjusted$prs >= unname(Common_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  ukb_pheno <- read.csv(\"Results/all_phenotypes.csv\")\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EUR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% ukb_pheno$IID[ukb_pheno$ancestry == \"EUR\"],]\n",
    "  CV_RV_PRS_adjusted_AMR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% ukb_pheno$IID[ukb_pheno$ancestry == \"AMR\"],]\n",
    "  CV_RV_PRS_adjusted_AFR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% ukb_pheno$IID[ukb_pheno$ancestry == \"AFR\"],]\n",
    "  CV_RV_PRS_adjusted_SAS <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% ukb_pheno$IID[ukb_pheno$ancestry == \"SAS\"],]\n",
    "  CV_RV_PRS_adjusted_EAS <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% ukb_pheno$IID[ukb_pheno$ancestry == \"EAS\"],]\n",
    "  CV_RV_PRS_adjusted_MID <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% ukb_pheno$IID[ukb_pheno$ancestry == \"MID\"],]\n",
    "  \n",
    "  Rare_quants <- quantile(CV_RV_PRS_adjusted_EUR$RV_PRS,c(0,.05,.2,.3,.4,.6,.7,.8,.95))\n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted_EUR$Rare_Bin[CV_RV_PRS_adjusted_EUR$RV_PRS < unname(Rare_quants[i + 1]) & CV_RV_PRS_adjusted_EUR$RV_PRS >= unname(Rare_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EUR$Rare_Bin[CV_RV_PRS_adjusted_EUR$Rare_Bin == 1] <- \"Below 5%\"\n",
    "  CV_RV_PRS_adjusted_EUR$Rare_Bin[CV_RV_PRS_adjusted_EUR$Rare_Bin %in% c(\"4\",\"5\",\"6\")] <- \"30% - 70%\"\n",
    "  CV_RV_PRS_adjusted_EUR$Rare_Bin[CV_RV_PRS_adjusted_EUR$Rare_Bin %in% c(\"9\")] <- \"Above 95%\"\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EUR <- CV_RV_PRS_adjusted_EUR[CV_RV_PRS_adjusted_EUR$Rare_Bin %in% c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"),]\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EUR$Rare_Bin <- factor(CV_RV_PRS_adjusted_EUR$Rare_Bin,levels = c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"))\n",
    "  \n",
    "  Rare_quants <- quantile(CV_RV_PRS_adjusted_AMR$RV_PRS,c(0,.05,.2,.3,.4,.6,.7,.8,.95))\n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted_AMR$Rare_Bin[CV_RV_PRS_adjusted_AMR$RV_PRS < unname(Rare_quants[i + 1]) & CV_RV_PRS_adjusted_AMR$RV_PRS >= unname(Rare_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AMR$Rare_Bin[CV_RV_PRS_adjusted_AMR$Rare_Bin == 1] <- \"Below 5%\"\n",
    "  CV_RV_PRS_adjusted_AMR$Rare_Bin[CV_RV_PRS_adjusted_AMR$Rare_Bin %in% c(\"4\",\"5\",\"6\")] <- \"30% - 70%\"\n",
    "  CV_RV_PRS_adjusted_AMR$Rare_Bin[CV_RV_PRS_adjusted_AMR$Rare_Bin %in% c(\"9\")] <- \"Above 95%\"\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AMR <- CV_RV_PRS_adjusted_AMR[CV_RV_PRS_adjusted_AMR$Rare_Bin %in% c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"),]\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AMR$Rare_Bin <- factor(CV_RV_PRS_adjusted_AMR$Rare_Bin,levels = c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"))\n",
    "  \n",
    "  Rare_quants <- quantile(CV_RV_PRS_adjusted_AFR$RV_PRS,c(0,.05,.2,.3,.4,.6,.7,.8,.95))\n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted_AFR$Rare_Bin[CV_RV_PRS_adjusted_AFR$RV_PRS < unname(Rare_quants[i + 1]) & CV_RV_PRS_adjusted_AFR$RV_PRS >= unname(Rare_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AFR$Rare_Bin[CV_RV_PRS_adjusted_AFR$Rare_Bin == 1] <- \"Below 5%\"\n",
    "  CV_RV_PRS_adjusted_AFR$Rare_Bin[CV_RV_PRS_adjusted_AFR$Rare_Bin %in% c(\"4\",\"5\",\"6\")] <- \"30% - 70%\"\n",
    "  CV_RV_PRS_adjusted_AFR$Rare_Bin[CV_RV_PRS_adjusted_AFR$Rare_Bin %in% c(\"9\")] <- \"Above 95%\"\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AFR <- CV_RV_PRS_adjusted_AFR[CV_RV_PRS_adjusted_AFR$Rare_Bin %in% c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"),]\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AFR$Rare_Bin <- factor(CV_RV_PRS_adjusted_AFR$Rare_Bin,levels = c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"))\n",
    "  \n",
    "  Rare_quants <- quantile(CV_RV_PRS_adjusted_SAS$RV_PRS,c(0,.05,.2,.3,.4,.6,.7,.8,.95))\n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted_SAS$Rare_Bin[CV_RV_PRS_adjusted_SAS$RV_PRS < unname(Rare_quants[i + 1]) & CV_RV_PRS_adjusted_SAS$RV_PRS >= unname(Rare_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  CV_RV_PRS_adjusted_SAS$Rare_Bin[CV_RV_PRS_adjusted_SAS$Rare_Bin == 1] <- \"Below 5%\"\n",
    "  CV_RV_PRS_adjusted_SAS$Rare_Bin[CV_RV_PRS_adjusted_SAS$Rare_Bin %in% c(\"4\",\"5\",\"6\")] <- \"30% - 70%\"\n",
    "  CV_RV_PRS_adjusted_SAS$Rare_Bin[CV_RV_PRS_adjusted_SAS$Rare_Bin %in% c(\"9\")] <- \"Above 95%\"\n",
    "  \n",
    "  CV_RV_PRS_adjusted_SAS <- CV_RV_PRS_adjusted_SAS[CV_RV_PRS_adjusted_SAS$Rare_Bin %in% c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"),]\n",
    "  \n",
    "  CV_RV_PRS_adjusted_SAS$Rare_Bin <- factor(CV_RV_PRS_adjusted_SAS$Rare_Bin,levels = c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"))\n",
    "  \n",
    "  \n",
    "  Rare_quants <- quantile(CV_RV_PRS_adjusted_EAS$RV_PRS,c(0,.05,.2,.3,.4,.6,.7,.8,.95))\n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted_EAS$Rare_Bin[CV_RV_PRS_adjusted_EAS$RV_PRS < unname(Rare_quants[i + 1]) & CV_RV_PRS_adjusted_EAS$RV_PRS >= unname(Rare_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EAS$Rare_Bin[CV_RV_PRS_adjusted_EAS$Rare_Bin == 1] <- \"Below 5%\"\n",
    "  CV_RV_PRS_adjusted_EAS$Rare_Bin[CV_RV_PRS_adjusted_EAS$Rare_Bin %in% c(\"4\",\"5\",\"6\")] <- \"30% - 70%\"\n",
    "  CV_RV_PRS_adjusted_EAS$Rare_Bin[CV_RV_PRS_adjusted_EAS$Rare_Bin %in% c(\"9\")] <- \"Above 95%\"\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EAS <- CV_RV_PRS_adjusted_EAS[CV_RV_PRS_adjusted_EAS$Rare_Bin %in% c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"),]\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EAS$Rare_Bin <- factor(CV_RV_PRS_adjusted_EAS$Rare_Bin,levels = c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"))\n",
    "  \n",
    "  \n",
    "  \n",
    "  Rare_quants <- quantile(CV_RV_PRS_adjusted_MID$RV_PRS,c(0,.05,.2,.3,.4,.6,.7,.8,.95))\n",
    "  for(i in 1:8){\n",
    "    CV_RV_PRS_adjusted_MID$Rare_Bin[CV_RV_PRS_adjusted_MID$RV_PRS < unname(Rare_quants[i + 1]) & CV_RV_PRS_adjusted_MID$RV_PRS >= unname(Rare_quants[i])] <- i\n",
    "  }\n",
    "  \n",
    "  CV_RV_PRS_adjusted_MID$Rare_Bin[CV_RV_PRS_adjusted_MID$Rare_Bin == 1] <- \"Below 5%\"\n",
    "  CV_RV_PRS_adjusted_MID$Rare_Bin[CV_RV_PRS_adjusted_MID$Rare_Bin %in% c(\"4\",\"5\",\"6\")] <- \"30% - 70%\"\n",
    "  CV_RV_PRS_adjusted_MID$Rare_Bin[CV_RV_PRS_adjusted_MID$Rare_Bin %in% c(\"9\")] <- \"Above 95%\"\n",
    "  \n",
    "  CV_RV_PRS_adjusted_MID <- CV_RV_PRS_adjusted_MID[CV_RV_PRS_adjusted_MID$Rare_Bin %in% c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"),]\n",
    "  \n",
    "  CV_RV_PRS_adjusted_MID$Rare_Bin <- factor(CV_RV_PRS_adjusted_MID$Rare_Bin,levels = c(\"Below 5%\",\"30% - 70%\",\"Above 95%\"))\n",
    "  \n",
    "  \n",
    "  \n",
    "  CV_RV_PRS_adjusted_EUR$Y <- scale(CV_RV_PRS_adjusted_EUR$Y)\n",
    "  CV_RV_PRS_adjusted_AMR$Y <- scale(CV_RV_PRS_adjusted_AMR$Y)\n",
    "  CV_RV_PRS_adjusted_AFR$Y <- scale(CV_RV_PRS_adjusted_AFR$Y)\n",
    "  CV_RV_PRS_adjusted_SAS$Y <- scale(CV_RV_PRS_adjusted_SAS$Y)\n",
    "  CV_RV_PRS_adjusted_EAS$Y <- scale(CV_RV_PRS_adjusted_EAS$Y)\n",
    "  CV_RV_PRS_adjusted_MID$Y <- scale(CV_RV_PRS_adjusted_MID$Y)\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EUR_se <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_EUR,function(x){sd(x)/sqrt(length(x))})\n",
    "  CV_RV_PRS_adjusted_EUR <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_EUR,mean)\n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_EUR) <- c(\"Common_Bin\",\"Rare_Bin\",\"Mean\")\n",
    "  colnames(CV_RV_PRS_adjusted_EUR_se) <- c(\"Common_Bin\",\"Rare_Bin\",\"SE\")\n",
    "  CV_RV_PRS_adjusted_EUR <- inner_join(CV_RV_PRS_adjusted_EUR,CV_RV_PRS_adjusted_EUR_se)\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AMR_se <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_AMR,function(x){sd(x)/sqrt(length(x))})\n",
    "  CV_RV_PRS_adjusted_AMR <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_AMR,mean)\n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_AMR) <- c(\"Common_Bin\",\"Rare_Bin\",\"Mean\")\n",
    "  colnames(CV_RV_PRS_adjusted_AMR_se) <- c(\"Common_Bin\",\"Rare_Bin\",\"SE\")\n",
    "  CV_RV_PRS_adjusted_AMR <- inner_join(CV_RV_PRS_adjusted_AMR,CV_RV_PRS_adjusted_AMR_se)\n",
    "  \n",
    "  CV_RV_PRS_adjusted_AFR_se <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_AFR,function(x){sd(x)/sqrt(length(x))})\n",
    "  CV_RV_PRS_adjusted_AFR <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_AFR,mean)\n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_AFR) <- c(\"Common_Bin\",\"Rare_Bin\",\"Mean\")\n",
    "  colnames(CV_RV_PRS_adjusted_AFR_se) <- c(\"Common_Bin\",\"Rare_Bin\",\"SE\")\n",
    "  CV_RV_PRS_adjusted_AFR <- inner_join(CV_RV_PRS_adjusted_AFR,CV_RV_PRS_adjusted_AFR_se)\n",
    "  \n",
    "  CV_RV_PRS_adjusted_SAS_se <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_SAS,function(x){sd(x)/sqrt(length(x))})\n",
    "  CV_RV_PRS_adjusted_SAS <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_SAS,mean)\n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_SAS) <- c(\"Common_Bin\",\"Rare_Bin\",\"Mean\")\n",
    "  colnames(CV_RV_PRS_adjusted_SAS_se) <- c(\"Common_Bin\",\"Rare_Bin\",\"SE\")\n",
    "  CV_RV_PRS_adjusted_SAS <- inner_join(CV_RV_PRS_adjusted_SAS,CV_RV_PRS_adjusted_SAS_se)\n",
    "  \n",
    "  CV_RV_PRS_adjusted_EAS_se <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_EAS,function(x){sd(x)/sqrt(length(x))})\n",
    "  CV_RV_PRS_adjusted_EAS <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_EAS,mean)\n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_EAS) <- c(\"Common_Bin\",\"Rare_Bin\",\"Mean\")\n",
    "  colnames(CV_RV_PRS_adjusted_EAS_se) <- c(\"Common_Bin\",\"Rare_Bin\",\"SE\")\n",
    "  CV_RV_PRS_adjusted_EAS <- inner_join(CV_RV_PRS_adjusted_EAS,CV_RV_PRS_adjusted_EAS_se)\n",
    "  \n",
    "  CV_RV_PRS_adjusted_MID_se <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_MID,function(x){sd(x)/sqrt(length(x))})\n",
    "  CV_RV_PRS_adjusted_MID <- aggregate(Y ~ Common_Bin + Rare_Bin,data = CV_RV_PRS_adjusted_MID,mean)\n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_MID) <- c(\"Common_Bin\",\"Rare_Bin\",\"Mean\")\n",
    "  colnames(CV_RV_PRS_adjusted_MID_se) <- c(\"Common_Bin\",\"Rare_Bin\",\"SE\")\n",
    "  CV_RV_PRS_adjusted_MID <- inner_join(CV_RV_PRS_adjusted_MID,CV_RV_PRS_adjusted_MID_se)\n",
    "  \n",
    "  \n",
    "  colnames(CV_RV_PRS_adjusted_EUR) <- c(\"Common_Bin\",\"RICE-RV Quantiles (Rare Variants)\",\"Mean\",\"SE\")\n",
    "  colnames(CV_RV_PRS_adjusted_AMR) <- c(\"Common_Bin\",\"RICE-RV Quantiles (Rare Variants)\",\"Mean\",\"SE\")\n",
    "  colnames(CV_RV_PRS_adjusted_AFR) <- c(\"Common_Bin\",\"RICE-RV Quantiles (Rare Variants)\",\"Mean\",\"SE\")\n",
    "  colnames(CV_RV_PRS_adjusted_SAS) <- c(\"Common_Bin\",\"RICE-RV Quantiles (Rare Variants)\",\"Mean\",\"SE\")\n",
    "  colnames(CV_RV_PRS_adjusted_EAS) <- c(\"Common_Bin\",\"RICE-RV Quantiles (Rare Variants)\",\"Mean\",\"SE\")\n",
    "  colnames(CV_RV_PRS_adjusted_MID) <- c(\"Common_Bin\",\"RICE-RV Quantiles (Rare Variants)\",\"Mean\",\"SE\")\n",
    "  \n",
    "  ymin <- round(min(c(CV_RV_PRS_adjusted_EUR$Mean - CV_RV_PRS_adjusted_EUR$SE,CV_RV_PRS_adjusted_AMR$Mean - CV_RV_PRS_adjusted_AMR$SE,CV_RV_PRS_adjusted_AFR$Mean - CV_RV_PRS_adjusted_AFR$SE,CV_RV_PRS_adjusted_SAS$Mean - CV_RV_PRS_adjusted_SAS$SE,CV_RV_PRS_adjusted_EAS$Mean - CV_RV_PRS_adjusted_EAS$SE,CV_RV_PRS_adjusted_MID$Mean - CV_RV_PRS_adjusted_MID$SE)) - 0.05,2)\n",
    "  ymax <- round(max(c(CV_RV_PRS_adjusted_EUR$Mean + CV_RV_PRS_adjusted_EUR$SE,CV_RV_PRS_adjusted_AMR$Mean + CV_RV_PRS_adjusted_AMR$SE,CV_RV_PRS_adjusted_AFR$Mean + CV_RV_PRS_adjusted_AFR$SE,CV_RV_PRS_adjusted_SAS$Mean + CV_RV_PRS_adjusted_SAS$SE,CV_RV_PRS_adjusted_EAS$Mean + CV_RV_PRS_adjusted_EAS$SE,CV_RV_PRS_adjusted_MID$Mean + CV_RV_PRS_adjusted_MID$SE)) + 0.05,2)\n",
    "  \n",
    "  plot1 <- ggplot(data=CV_RV_PRS_adjusted_EUR, aes(x=Common_Bin, y=Mean, color=`RICE-RV Quantiles (Rare Variants)`)) + geom_line() + geom_pointrange(aes(ymin=Mean-SE, ymax=Mean+SE)) + theme_Publication() + ylab(paste0(ifelse(trait == \"logTG\",\"log(TG)\",trait),\" Standardized\")) + ylim(c(ymin,ymax)) +\n",
    "    scale_x_continuous(breaks = c(1:9),labels = c(\"0-10%\",\"10-20%\",\"20-30%\",\"30-40%\",\"40-60%\",\"60-70%\",\"70-80%\",\"80-90%\",\"90-100%\")) + labs(x = \"RICE-CV Quantiles (Common Variants)\")\n",
    "  plot2 <- ggplot(data=CV_RV_PRS_adjusted_AMR, aes(x=Common_Bin, y=Mean, color=`RICE-RV Quantiles (Rare Variants)`)) + geom_line() + geom_pointrange(aes(ymin=Mean-SE, ymax=Mean+SE)) + theme_Publication() + ylab(paste0(ifelse(trait == \"logTG\",\"log(TG)\",trait),\" Standardized\")) + ylim(c(ymin,ymax)) +\n",
    "    scale_x_continuous(breaks = c(1:9),labels = c(\"0-10%\",\"10-20%\",\"20-30%\",\"30-40%\",\"40-60%\",\"60-70%\",\"70-80%\",\"80-90%\",\"90-100%\")) + labs(x = \"RICE-CV Quantiles (Common Variants)\")\n",
    "  plot3 <- ggplot(data=CV_RV_PRS_adjusted_AFR, aes(x=Common_Bin, y=Mean, color=`RICE-RV Quantiles (Rare Variants)`)) + geom_line() + geom_pointrange(aes(ymin=Mean-SE, ymax=Mean+SE)) + theme_Publication() + ylab(paste0(ifelse(trait == \"logTG\",\"log(TG)\",trait),\" Standardized\")) + ylim(c(ymin,ymax)) +\n",
    "    scale_x_continuous(breaks = c(1:9),labels = c(\"0-10%\",\"10-20%\",\"20-30%\",\"30-40%\",\"40-60%\",\"60-70%\",\"70-80%\",\"80-90%\",\"90-100%\")) + labs(x = \"RICE-CV Quantiles (Common Variants)\")\n",
    "  plot4 <- ggplot(data=CV_RV_PRS_adjusted_SAS, aes(x=Common_Bin, y=Mean, color=`RICE-RV Quantiles (Rare Variants)`)) + geom_line() + geom_pointrange(aes(ymin=Mean-SE, ymax=Mean+SE)) + theme_Publication() + ylab(paste0(ifelse(trait == \"logTG\",\"log(TG)\",trait),\" Standardized\")) + ylim(c(ymin,ymax)) +\n",
    "    scale_x_continuous(breaks = c(1:9),labels = c(\"0-10%\",\"10-20%\",\"20-30%\",\"30-40%\",\"40-60%\",\"60-70%\",\"70-80%\",\"80-90%\",\"90-100%\")) + labs(x = \"RICE-CV Quantiles (Common Variants)\")\n",
    "  plot5 <- ggplot(data=CV_RV_PRS_adjusted_EAS, aes(x=Common_Bin, y=Mean, color=`RICE-RV Quantiles (Rare Variants)`)) + geom_line() + geom_pointrange(aes(ymin=Mean-SE, ymax=Mean+SE)) + theme_Publication() + ylab(paste0(ifelse(trait == \"logTG\",\"log(TG)\",trait),\" Standardized\")) + ylim(c(ymin,ymax)) +\n",
    "    scale_x_continuous(breaks = c(1:9),labels = c(\"0-10%\",\"10-20%\",\"20-30%\",\"30-40%\",\"40-60%\",\"60-70%\",\"70-80%\",\"80-90%\",\"90-100%\")) + labs(x = \"RICE-CV Quantiles (Common Variants)\")\n",
    "  plot6 <- ggplot(data=CV_RV_PRS_adjusted_MID, aes(x=Common_Bin, y=Mean, color=`RICE-RV Quantiles (Rare Variants)`)) + geom_line() + geom_pointrange(aes(ymin=Mean-SE, ymax=Mean+SE)) + theme_Publication() + ylab(paste0(ifelse(trait == \"logTG\",\"log(TG)\",trait),\" Standardized\")) + ylim(c(ymin,ymax)) +\n",
    "    scale_x_continuous(breaks = c(1:9),labels = c(\"0-10%\",\"10-20%\",\"20-30%\",\"30-40%\",\"40-60%\",\"60-70%\",\"70-80%\",\"80-90%\",\"90-100%\")) + labs(x = \"RICE-CV Quantiles (Common Variants)\")\n",
    "  \n",
    "  \n",
    "  prow <- plot_grid(\n",
    "    plot1 + theme(legend.position=\"none\"),\n",
    "    plot2 + theme(legend.position=\"none\"),\n",
    "    plot3 + theme(legend.position=\"none\"),\n",
    "    plot4 + theme(legend.position=\"none\"),\n",
    "    plot5 + theme(legend.position=\"none\"),\n",
    "    plot6 + theme(legend.position=\"none\"),\n",
    "    align = 'vh',\n",
    "    labels = c(\"EUR\",\"AMR\",\"AFR\",\"SAS\",\"EAS\",\"MID\"),\n",
    "    hjust = -1,\n",
    "    ncol = 3, label_size = 20\n",
    "  )\n",
    "  \n",
    "  legend_b <- ggplotGrob(plot1)$grobs[[which(sapply(ggplotGrob(plot1)$grobs, function(x) x$name) == \"guide-box\")]]\n",
    "  \n",
    "  pdf(paste0(trait,\"_AoU_WES_Fig5.pdf\"), width=24, height=14.8331273177)\n",
    "  \n",
    "  print(plot_grid(prow, legend_b, ncol = 1, rel_heights = c(1, .1)))\n",
    "  \n",
    "  dev.off()\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"rm -r Results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"mkdir Results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"gsutil cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt Results/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "        system(paste0(\"gsutil cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned_QQPlots Results/\"))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'AFR_BMI_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AFR_HDL_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AFR_Height_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AFR_LDL_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AFR_logTG_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AFR_TC_GWAS_SumStats_Cleaned_QQPlots'</li><li>'All_Train.txt'</li><li>'AMR_BMI_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AMR_HDL_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AMR_Height_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AMR_LDL_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AMR_logTG_GWAS_SumStats_Cleaned_QQPlots'</li><li>'AMR_TC_GWAS_SumStats_Cleaned_QQPlots'</li><li>'EUR_BMI_GWAS_SumStats_Cleaned_QQPlots'</li><li>'EUR_HDL_GWAS_SumStats_Cleaned_QQPlots'</li><li>'EUR_Height_GWAS_SumStats_Cleaned_QQPlots'</li><li>'EUR_LDL_GWAS_SumStats_Cleaned_QQPlots'</li><li>'EUR_logTG_GWAS_SumStats_Cleaned_QQPlots'</li><li>'EUR_TC_GWAS_SumStats_Cleaned_QQPlots'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'AFR\\_BMI\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AFR\\_HDL\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AFR\\_Height\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AFR\\_LDL\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AFR\\_logTG\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AFR\\_TC\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'All\\_Train.txt'\n",
       "\\item 'AMR\\_BMI\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AMR\\_HDL\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AMR\\_Height\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AMR\\_LDL\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AMR\\_logTG\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'AMR\\_TC\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'EUR\\_BMI\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'EUR\\_HDL\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'EUR\\_Height\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'EUR\\_LDL\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'EUR\\_logTG\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\item 'EUR\\_TC\\_GWAS\\_SumStats\\_Cleaned\\_QQPlots'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'AFR_BMI_GWAS_SumStats_Cleaned_QQPlots'\n",
       "2. 'AFR_HDL_GWAS_SumStats_Cleaned_QQPlots'\n",
       "3. 'AFR_Height_GWAS_SumStats_Cleaned_QQPlots'\n",
       "4. 'AFR_LDL_GWAS_SumStats_Cleaned_QQPlots'\n",
       "5. 'AFR_logTG_GWAS_SumStats_Cleaned_QQPlots'\n",
       "6. 'AFR_TC_GWAS_SumStats_Cleaned_QQPlots'\n",
       "7. 'All_Train.txt'\n",
       "8. 'AMR_BMI_GWAS_SumStats_Cleaned_QQPlots'\n",
       "9. 'AMR_HDL_GWAS_SumStats_Cleaned_QQPlots'\n",
       "10. 'AMR_Height_GWAS_SumStats_Cleaned_QQPlots'\n",
       "11. 'AMR_LDL_GWAS_SumStats_Cleaned_QQPlots'\n",
       "12. 'AMR_logTG_GWAS_SumStats_Cleaned_QQPlots'\n",
       "13. 'AMR_TC_GWAS_SumStats_Cleaned_QQPlots'\n",
       "14. 'EUR_BMI_GWAS_SumStats_Cleaned_QQPlots'\n",
       "15. 'EUR_HDL_GWAS_SumStats_Cleaned_QQPlots'\n",
       "16. 'EUR_Height_GWAS_SumStats_Cleaned_QQPlots'\n",
       "17. 'EUR_LDL_GWAS_SumStats_Cleaned_QQPlots'\n",
       "18. 'EUR_logTG_GWAS_SumStats_Cleaned_QQPlots'\n",
       "19. 'EUR_TC_GWAS_SumStats_Cleaned_QQPlots'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"AFR_BMI_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       " [2] \"AFR_HDL_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       " [3] \"AFR_Height_GWAS_SumStats_Cleaned_QQPlots\"\n",
       " [4] \"AFR_LDL_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       " [5] \"AFR_logTG_GWAS_SumStats_Cleaned_QQPlots\" \n",
       " [6] \"AFR_TC_GWAS_SumStats_Cleaned_QQPlots\"    \n",
       " [7] \"All_Train.txt\"                           \n",
       " [8] \"AMR_BMI_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       " [9] \"AMR_HDL_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       "[10] \"AMR_Height_GWAS_SumStats_Cleaned_QQPlots\"\n",
       "[11] \"AMR_LDL_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       "[12] \"AMR_logTG_GWAS_SumStats_Cleaned_QQPlots\" \n",
       "[13] \"AMR_TC_GWAS_SumStats_Cleaned_QQPlots\"    \n",
       "[14] \"EUR_BMI_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       "[15] \"EUR_HDL_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       "[16] \"EUR_Height_GWAS_SumStats_Cleaned_QQPlots\"\n",
       "[17] \"EUR_LDL_GWAS_SumStats_Cleaned_QQPlots\"   \n",
       "[18] \"EUR_logTG_GWAS_SumStats_Cleaned_QQPlots\" \n",
       "[19] \"EUR_TC_GWAS_SumStats_Cleaned_QQPlots\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.files(\"Results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/home/jupyter/packages’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"plotrix\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/home/jupyter/packages’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"ggthemes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in max(lab2):\n",
      "“no non-missing arguments to max; returning -Inf”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n",
      "[1] \"okkkk2\"\n",
      "[1] \"okkkk3\"\n"
     ]
    }
   ],
   "source": [
    "rm(list = ls())\n",
    "\n",
    "pheno_train <- read.delim(\"Results/All_Train.txt\")\n",
    "\n",
    "theme_Publication <- function(base_size=12) {\n",
    "  library(grid)\n",
    "  library(ggthemes)\n",
    "  (theme_foundation(base_size=base_size, )\n",
    "    + theme(plot.title = element_text(face = \"bold\",\n",
    "                                      size = rel(1.1), hjust = 0.5),\n",
    "            text = element_text(),\n",
    "            panel.background = element_rect(colour = NA),\n",
    "            plot.background = element_rect(colour = NA),\n",
    "            panel.border = element_rect(colour = NA),\n",
    "            axis.title = element_text(face = \"bold\",size = 16),\n",
    "            axis.title.y = element_text(angle=90,vjust =2),\n",
    "            axis.title.x = element_blank(),\n",
    "            axis.text.x = element_blank(), \n",
    "            axis.line = element_line(colour=\"black\",size=2),\n",
    "            axis.ticks = element_line(),\n",
    "            title = element_text(face = \"bold\",size = 20),\n",
    "            # panel.grid.major = element_line(colour=\"#f0f0f0\"),\n",
    "            # panel.grid.minor = element_line(colour=\"#f0f0f0\"),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor = element_blank(),\n",
    "            legend.key = element_rect(colour = NA),\n",
    "            #legend.position = \"bottom\",\n",
    "            #legend.direction = \"horizontal\",\n",
    "            #legend.key.size= unit(0.2, \"cm\"),\n",
    "            #legend.margin = unit(0, \"cm\"),\n",
    "            legend.title = element_text(face=\"bold.italic\", size =18),\n",
    "            #legend.text = element_text(face =\"bold\"),\n",
    "            plot.margin=unit(c(10,5,5,5),\"mm\"),\n",
    "            strip.background=element_rect(colour=\"#f0f0f0\",fill=\"#f0f0f0\"),\n",
    "            strip.text = element_text(face=\"bold\")\n",
    "    ))\n",
    "  \n",
    "}\n",
    "\n",
    "library(plotrix)\n",
    "library(data.table)\n",
    "library(RColorBrewer)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "\n",
    "lambda_dat <- NULL\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "  dat <- fread(paste0(\"Results/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned_QQPlots\"))\n",
    "  dat <- as.data.frame(dat)\n",
    "  colnames(dat) <- c(\"CHR\",\"SNP\",\"BP\",\"A1_FREQ\",\"P\")\n",
    "  dat$MAF <- ifelse(dat$A1_FREQ <= 0.5, dat$A1_FREQ,1-dat$A1_FREQ)\n",
    "  dat <- dat[!is.na(dat$MAF),]\n",
    "  dat <- dat[!is.na(dat$P),]\n",
    "  dat <- dat[dat$MAF > 0.01,]\n",
    "  dat <- dat[dat$P != 0,]\n",
    "        \n",
    "if(anc == \"AFR\"){\n",
    "    anc_plot <- \"African\"\n",
    "}else if(anc == \"AMR\"){\n",
    "    anc_plot <- \"Latino\"\n",
    "}else{\n",
    "    anc_plot <- \"European\"\n",
    "}\n",
    "  \n",
    "  \n",
    "  x <- dat$P\n",
    "  z <- qnorm(x / 2)\n",
    "  lambda <- round(median(z^2) / qchisq(0.5,1), 3)\n",
    "  lambda_1000 <- round(1+1000*(lambda-1)/sum(!is.na(pheno_train[,trait])) ,3)\n",
    "  \n",
    "  lambda_dat <- rbind(lambda_dat,data.frame(Trait = trait,Ancestry_Group = anc_plot,lambda = lambda, lambda_1000 = lambda_1000, datasource = \"AoU WES\"))\n",
    "  \n",
    "  p.pwas <- 5E-08\n",
    "  \n",
    "  nCHR <- length(unique(dat$CHR))\n",
    "  dat$BPcum <- NA\n",
    "  s <- 0\n",
    "  nbp <- c()\n",
    "  for (i in unique(dat$CHR)){\n",
    "    nbp[i] <- max(dat[dat$CHR == i,]$BP)\n",
    "    dat$BPcum[dat$CHR == i] <- dat$BP[dat$CHR == i] + s\n",
    "    s <- s + nbp[i]\n",
    "  }\n",
    "  axis.set <- dat %>% \n",
    "    group_by(CHR) %>% \n",
    "    summarize(center = (max(BPcum) + min(BPcum)) / 2)\n",
    "  ylim <- abs(floor(log10(min(dat$P)))) + 2 \n",
    "  sig1 <- p.pwas\n",
    "  \n",
    "  sigline <- data.frame(sig=c(-log10(sig1)),val=c(paste0(\"P=\",signif(sig1,2))))\n",
    "  \n",
    "  p1 <- ggplot(dat, aes(x = BPcum, y = -log10(P), \n",
    "                        color = as.factor(CHR), size = -log10(P))) +\n",
    "    geom_point(alpha = 0.8, size=0.8) + \n",
    "    scale_x_continuous(label = axis.set$CHR, breaks = axis.set$center) +\n",
    "    scale_y_continuous(expand = c(0,0), limits = c(0, ylim)) +\n",
    "    scale_color_manual(values = rep(c(\"#08306b\", \"#4292c6\"), nCHR)) +\n",
    "    scale_size_continuous(range = c(0.5,3)) +\n",
    "    geom_hline(data = sigline, aes(yintercept = sig), color= \"red\", linetype=\"dashed\") +\n",
    "    guides(color = F) + \n",
    "    labs(x = NULL, \n",
    "         y = \"-log10(p)\", \n",
    "         linetype = \"\",\n",
    "         title = paste0(trait,\" (\",anc_plot,\")\"))+\n",
    "    theme_Publication()+\n",
    "    theme(\n",
    "      legend.position = \"top\",\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid.major.x = element_blank(),\n",
    "      panel.grid.minor.x = element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, size = 16, vjust = 0.5),\n",
    "      axis.text.y = element_text(size = 16),\n",
    "      plot.subtitle = element_text(size = 8)\n",
    "    )\n",
    "  \n",
    "  ggsave(paste0(trait,\"_\",anc,\"_AoU_WES_CV_Manhattan_Plot.png\"),p1,width = 15,height = 9.27070457355,dpi = 300)\n",
    "  \n",
    "  png(paste0(trait,\"_\",anc,\"_AoU_WES_CV_QQplot.png\"), width=10, height=10,units = \"in\",res = 300)\n",
    "  \n",
    "  qqplotdata <- function(logpvector){\n",
    "    o = sort(logpvector,decreasing=T)\n",
    "    e = -log10(ppoints(length(o)))       \n",
    "    qqdata <- data.frame(o,e)\n",
    "    qqdata$o <- round(qqdata$o,3)\n",
    "    qqdata$e <- round(qqdata$e,3)\n",
    "    keepU <- which(!duplicated(qqdata))\n",
    "    qqdata <- qqdata[keepU,]\n",
    "    \n",
    "    N <- length(logpvector) ## number of p-values\n",
    "    ## create the confidence intervals\n",
    "    qqdata$c975 <- NA\n",
    "    qqdata$c025 <- NA\n",
    "    \n",
    "    ## the jth order statistic from a\n",
    "    ## uniform(0,1) sample\n",
    "    ## has a beta(j,n-j+1) distribution\n",
    "    ## (Casella & Berger, 2002,\n",
    "    ## 2nd edition, pg 230, Duxbury)\n",
    "    \n",
    "    for(i in 1:length(keepU)){\n",
    "      j <- keepU[i]\n",
    "      qqdata$c975[i] <- -log10(qbeta(0.975,j,N-j+1))\n",
    "      qqdata$c025[i] <- -log10(qbeta(0.025,j,N-j+1))\n",
    "    }\n",
    "    return(qqdata)\n",
    "  }\n",
    "  yLine <- c(-log10(5E-8))\n",
    "  colLine <- c(\"red\")\n",
    "  dat$log10P <- -log10(dat$P)\n",
    "  gwas <- as.data.frame(dat)\n",
    "  # Determine frequency bins and create variable for binned QQ plot\n",
    "  \n",
    "  minMAF <- min(gwas$MAF)\n",
    "  \n",
    "  freqbins <- c(c(0.5,0.05,0.01,0)[which(c(0.5,0.05,0.01,0) > floor(minMAF*1000000)/1000000)],floor(minMAF*1000000)/1000000)\n",
    "  gwas$freqbin <- cut(gwas$MAF, freqbins,include.lowest=T)\n",
    "  freqtable <- table(gwas$freqbin)\n",
    "  freqtable <- freqtable[order(-as.numeric(gsub(\"[\\\\[\\\\(](.+),.+\",\"\\\\1\",names(freqtable))))]\n",
    "  freqtable <- freqtable[freqtable > 0]\n",
    "  \n",
    "  ## Generate QQ plot data by frequency bin\n",
    "  fbin <- character(0)\n",
    "  fN <- integer(0)\n",
    "  fx <- numeric(0)\n",
    "  fy <- numeric(0)\n",
    "  fcol <- character(0)\n",
    "  legendcol <- character(0)\n",
    "  conf <- list()\n",
    "  allcols <- brewer.pal(4,\"Set1\")\n",
    "  ycol <- \"log10P\"\n",
    "  for(f in 1:length(freqtable)){\n",
    "    fbin <- c(fbin,names(freqtable)[f])\n",
    "    fsnps <- which(gwas$freqbin ==names(freqtable)[f])\n",
    "    plotdata <- qqplotdata(gwas[[ycol]][fsnps])\n",
    "    fN <- c(fN,freqtable[f])\n",
    "    fx <- c(fx,plotdata$e)\n",
    "    fy <- c(fy,plotdata$o)\n",
    "    fcol <- c(fcol,rep(allcols[f],length(plotdata$o)))\n",
    "    conf[[f]] <- data.frame('x'=c(plotdata$e,rev(plotdata$e)),\n",
    "                            'y'=c(plotdata$c975,rev(plotdata$c025)))\n",
    "    legendcol <- c(legendcol,allcols[f])\n",
    "  }\n",
    "  legendtext <- paste0(\"MAF=\",fbin,\"; N SNPs=\",format(fN,big.mark=\",\",scientific=FALSE))\n",
    "  opt <-  list(break.top = 15,\n",
    "               top.size = 0.125)\n",
    "  \n",
    "  \n",
    "  xlim <- c(0,max(fx,na.rm=T))\n",
    "  ylim <- c(0,max(fy,na.rm=T))\n",
    "  maxY <- max(fy,na.rm=T)\n",
    "  print(\"okkkk2\")\n",
    "  par(mar=c(5.1,5.1,4.1,1.1))\n",
    "  print(\"okkkk3\")\n",
    "  \n",
    "  lab1 <- pretty(c(0,opt$break.top),n=ceiling(12 * (1-opt$top.size)))\n",
    "  lab1 <- c(lab1[lab1 < opt$break.top],opt$break.top)\n",
    "  lab2 <- pretty(c(opt$break.top,maxY),n=max(3,floor(12 * opt$top.size)))\n",
    "  lab2 <- lab2[lab2 > max(lab1)]\n",
    "  \n",
    "  # resulting range of top scale in bottom scale units\n",
    "  top.range = opt$break.top/(1 - opt$top.size) - opt$break.top\n",
    "  top.data = max(lab2)-opt$break.top\n",
    "  \n",
    "  # function to rescale the top part\n",
    "  rescale <- function(y) { opt$break.top+(y-opt$break.top)/(top.data/top.range)}\n",
    "  rescaled.y <- rescale(fy[fy>opt$break.top])\n",
    "  plot(0,0,\n",
    "       ylim=c(min(fy),opt$break.top*(1+opt$top.size)),xlim=xlim,axes=FALSE,\n",
    "       xlab=expression(plain(Expected)~~group(\"(\",-log[10]*italic(P),\")\")),\n",
    "       ylab=expression(plain(Observed)~~group(\"(\",-log[10]*italic(P),\")\")),\n",
    "       cex=1,cex.lab=1.5,cex.axis=1.5,bty=\"n\",col=\"transparent\",\n",
    "       main=opt$maintitle,pch=19)\n",
    "  \n",
    "  # Plot confidence intervals\t\n",
    "  for(p in 1:length(conf)){\n",
    "    polygon(conf[[p]]$'x',ifelse(conf[[p]]$'y'>opt$break.top,rescale(conf[[p]]$'y'),conf[[p]]$'y'),\n",
    "            col=grDevices::rgb(t(grDevices::col2rgb(allcols[p])),alpha=50,maxColorValue=255),\n",
    "            border = NA)\n",
    "  }\n",
    "  \n",
    "  # add points\n",
    "  points(fx[fy<opt$break.top],fy[fy<opt$break.top],cex=1,col=fcol[fy<opt$break.top],pch=19)\n",
    "  \n",
    "  # identify line & add axis break\n",
    "  lines(xlim,xlim,col=\"black\",lty = 2)\n",
    "  axis(1,cex.axis=1.5,cex.lab=1.5)\n",
    "  par(las=1)\n",
    "  axis(side=2,at=lab1,cex.axis=1.5,cex.lab=1.5)\n",
    "  par(las=0)\n",
    "  box()\n",
    "  par(las=0)\n",
    "  points(fx[fy>opt$break.top],rescaled.y,cex=1,col=fcol[fy>opt$break.top],pch=19)\n",
    "  par(las=1)\n",
    "  axis(side=2,at=rescale(lab2),labels=lab2,cex.axis=1.5,cex.lab=1.5)\n",
    "  axis.break(axis=2,breakpos=opt$break.top,style=\"zigzag\",brw=0.02)\n",
    "  axis.break(axis=4,breakpos=opt$break.top,style=\"zigzag\",brw=0.02)\n",
    "  lines(range(fx),c(opt$break.top,opt$break.top),col = \"grey\",lty = 6)\n",
    "  abline(h=ifelse(yLine<opt$break.top,\n",
    "                  yLine,\n",
    "                  rescale(yLine)),\n",
    "         col=colLine,lwd=1.5,lty=2)\n",
    "  legend(\"topleft\",legend=legendtext,col=legendcol,pch=15,bty=\"n\",cex=1.5)\n",
    "  text(4,1,expression(paste(lambda[1000],\" = \")),cex = 1.5)\n",
    "  text(4.65,1,paste(lambda_1000),cex = 1.5)\n",
    "  title(paste0(trait,\" (\",anc_plot,\")\"),cex.main = 1.5)\n",
    "  \n",
    "  # p2 <- recordPlot()\n",
    "  # \n",
    "  # title <- ggdraw() + \n",
    "  #   draw_label(\n",
    "  #     paste0(trait,\" for Europeans\"),\n",
    "  #     fontface = 'bold',\n",
    "  #     x = 0,\n",
    "  #     hjust = 0,\n",
    "  #     size = 30\n",
    "  #   ) +\n",
    "  #   theme(\n",
    "  #     # add margin on the left of the drawing canvas,\n",
    "  #     # so title is aligned with left edge of first plot\n",
    "  #     plot.margin = margin(0, 0, 0, 7)\n",
    "  #   )\n",
    "  \n",
    "  # plot_row <- plot_grid(p1, p2, align = \"h\", axis = \"bt\", rel_widths = c(6, 1))\n",
    "  # \n",
    "  # p3 <- plot_grid(\n",
    "  #   title, plot_row,\n",
    "  #   ncol = 1,\n",
    "  #   # rel_heights values control vertical title margins\n",
    "  #   rel_heights = c(0.1, 1)\n",
    "  # )\n",
    "  # \n",
    "  # cowplot::save_plot(filename = paste0(trait,\"_UKB_WES_CV_QQplot.pdf\"),plot = p3,nrow = 2,ncol = 2,base_width = c(24))\n",
    "  # \n",
    "  dev.off()\n",
    "}\n",
    "}\n",
    "write.csv(lambda_dat,file = \"AoU_WES_lambda.csv\",row.names = FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
