{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket = os.getenv(\"WORKSPACE_BUCKET\")\n",
    "\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: USER_NAME=williamsjacr\n"
     ]
    }
   ],
   "source": [
    "%env USER_NAME={USER_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dsub in /opt/conda/lib/python3.10/site-packages (0.4.11)\n",
      "Collecting dsub\n",
      "  Downloading dsub-0.4.13-py3-none-any.whl.metadata (33 kB)\n",
      "Requirement already satisfied: funcsigs==1.0.2 in /opt/conda/lib/python3.10/site-packages (from dsub) (1.0.2)\n",
      "Requirement already satisfied: google-api-core<=2.19.0,>=2.7.3 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.11.1)\n",
      "Requirement already satisfied: google-api-python-client<=2.131.0,>=2.47.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.127.0)\n",
      "Requirement already satisfied: google-auth-httplib2<=0.2.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.2.0)\n",
      "Requirement already satisfied: google-auth<=2.29.0,>=2.6.6 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.22.0)\n",
      "Requirement already satisfied: google-cloud-batch<=0.17.20 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.17.18)\n",
      "Requirement already satisfied: httplib2<=0.22.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.21.0)\n",
      "Requirement already satisfied: mock<=5.1.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (5.1.0)\n",
      "Requirement already satisfied: parameterized<=0.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.9.0)\n",
      "Requirement already satisfied: protobuf<=5.26.0,>=3.19.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (3.19.6)\n",
      "Requirement already satisfied: pyasn1-modules<=0.4.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.2.7)\n",
      "Requirement already satisfied: pyasn1<=0.6.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.4.8)\n",
      "Requirement already satisfied: python-dateutil<=2.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.8.2)\n",
      "Requirement already satisfied: pytz<=2024.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (2023.3)\n",
      "Requirement already satisfied: pyyaml<=6.0.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (6.0.1)\n",
      "Requirement already satisfied: rsa<=4.9 in /opt/conda/lib/python3.10/site-packages (from dsub) (4.9)\n",
      "Requirement already satisfied: tabulate<=0.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.9.0)\n",
      "Requirement already satisfied: tenacity<=8.2.3 in /opt/conda/lib/python3.10/site-packages (from dsub) (8.2.3)\n",
      "Requirement already satisfied: uritemplate<=4.1.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (3.0.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.dev0,>=1.56.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (1.60.0)\n",
      "Requirement already satisfied: requests<3.0.0.dev0,>=2.18.0 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (2.31.0)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (4.2.4)\n",
      "Requirement already satisfied: six>=1.9.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (1.16.0)\n",
      "Requirement already satisfied: urllib3<2.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (1.26.15)\n",
      "Requirement already satisfied: proto-plus<2.0.0dev,>=1.22.3 in /opt/conda/lib/python3.10/site-packages (from google-cloud-batch<=0.17.20->dsub) (1.22.3)\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in /opt/conda/lib/python3.10/site-packages (from httplib2<=0.22.0->dsub) (3.0.9)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.33.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-batch<=0.17.20->dsub) (1.48.1)\n",
      "Requirement already satisfied: grpcio-status<2.0.dev0,>=1.33.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-batch<=0.17.20->dsub) (1.48.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (3.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (3.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (2024.2.2)\n",
      "Downloading dsub-0.4.13-py3-none-any.whl (187 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m187.5/187.5 kB\u001b[0m \u001b[31m1.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25h\u001b[33mWARNING: Error parsing requirements for aiohttp: [Errno 2] No such file or directory: '/opt/conda/lib/python3.10/site-packages/aiohttp-3.8.5.dist-info/METADATA'\u001b[0m\u001b[33m\n",
      "\u001b[0mInstalling collected packages: dsub\n",
      "Successfully installed dsub-0.4.13\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.2\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install --upgrade dsub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /home/jupyter/aou_dsub.bash\n"
     ]
    }
   ],
   "source": [
    "%%writefile ~/aou_dsub.bash\n",
    "#!/bin/bash\n",
    "function aou_dsub () {\n",
    "\n",
    "  # Get a shorter username to leave more characters for the job name.\n",
    "  local DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "  # For AoU RWB projects network name is \"network\".\n",
    "  local AOU_NETWORK=network\n",
    "  local AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "  dsub \\\n",
    "      --provider google-cls-v2 \\\n",
    "      --user-project \"${GOOGLE_PROJECT}\"\\\n",
    "      --project \"${GOOGLE_PROJECT}\"\\\n",
    "      --image 'marketplace.gcr.io/google/ubuntu1804:latest' \\\n",
    "      --network \"${AOU_NETWORK}\" \\\n",
    "      --subnetwork \"${AOU_SUBNETWORK}\" \\\n",
    "      --service-account \"$(gcloud config get-value account)\" \\\n",
    "      --user \"${DSUB_USER_NAME}\" \\\n",
    "      --regions us-central1 \\\n",
    "      --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "      \"$@\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo source ~/aou_dsub.bash >> ~/.bashrc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                                      Last Update\n",
      "---------------  ------  ------------------------------------------  --------------\n",
      "single-rarev...       6  Started running \"user-command\"              08-06 18:47:34\n",
      "single-rarev...       5  Stopped running \"user-command\": exit st...  08-06 19:13:14\n",
      "single-rarev...       4  Stopped running \"user-command\": exit st...  08-06 19:17:42\n",
      "single-rarev...       3  Stopped running \"user-command\": exit st...  08-06 19:11:39\n",
      "single-rarev...       2  Stopped running \"user-command\": exit st...  08-06 19:10:33\n",
      "single-rarev...       1  Stopped running \"user-command\": exit st...  08-06 19:28:51\n",
      "single-rarev...       6  The operation was cancelled                 08-06 18:29:01\n",
      "single-rarev...       5  Stopped running \"user-command\": exit st...  08-06 18:09:51\n",
      "single-rarev...       4  Stopped running \"user-command\": exit st...  08-06 18:14:04\n",
      "single-rarev...       3  Stopped running \"user-command\": exit st...  08-06 18:08:20\n",
      "single-rarev...       2  Stopped running \"user-command\": exit st...  08-06 18:05:31\n",
      "single-rarev...       1  Stopped running \"user-command\": exit st...  08-06 18:27:10\n",
      "single-rarev...       6  The operation was cancelled                 08-06 17:26:07\n",
      "single-rarev...       5  Stopped running \"user-command\": exit st...  08-06 17:24:23\n",
      "single-rarev...       4  The operation was cancelled                 08-06 17:26:07\n",
      "single-rarev...       3  The operation was cancelled                 08-06 17:26:07\n",
      "single-rarev...       2  Stopped running \"user-command\": exit st...  08-06 17:22:59\n",
      "single-rarev...       1  The operation was cancelled                 08-06 17:26:07\n",
      "single-rarev...       6  Stopped running \"user-command\": exit st...  08-06 17:20:52\n",
      "single-rarev...       5  worker was terminated                       08-06 16:00:28\n",
      "single-rarev...       4  Stopped running \"user-command\": exit st...  08-06 16:31:37\n",
      "single-rarev...       3  worker was terminated                       08-06 16:08:54\n",
      "single-rarev...       2  worker was terminated                       08-06 16:00:43\n",
      "single-rarev...       1  worker was terminated                       08-06 16:02:26\n",
      "coding-analy...       1  Success                                     08-05 20:42:18\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 20:24:55\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 20:20:21\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 20:15:50\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 20:12:20\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 20:01:12\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 19:57:51\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 19:54:17\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 19:50:44\n",
      "coding-analy...       1  Stopped running \"user-command\": exit st...  08-05 19:47:13\n",
      "coding-analysis       3  Success                                     08-05 18:23:33\n",
      "coding-analysis       3  worker was terminated                       08-05 17:02:36\n",
      "coding-analysis       2  Success                                     08-05 17:36:46\n",
      "coding-analysis       3  worker was terminated                       08-05 16:37:48\n",
      "coding-analysis       4  Success                                     08-05 16:21:02\n",
      "coding-analysis       6  Success                                     08-05 16:44:42\n",
      "coding-analysis       8  Success                                     08-05 16:04:54\n",
      "coding-analysis       7  Success                                     08-05 16:47:22\n",
      "coding-analysis       6  worker was terminated                       08-05 15:39:07\n",
      "coding-analysis       5  Success                                     08-05 15:52:56\n",
      "coding-analysis       4  worker was terminated                       08-05 15:41:16\n",
      "coding-analysis       3  worker was terminated                       08-05 16:08:13\n",
      "coding-analysis       2  worker was terminated                       08-05 16:28:00\n",
      "coding-analysis       1  Success                                     08-05 15:53:40\n",
      "coding-analysis    1628  Stopped running \"user-command\": exit st...  08-04 23:35:54\n",
      "coding-analysis    1629  Stopped running \"user-command\": exit st...  08-04 23:30:15\n",
      "coding-analysis    1628  Stopped running \"user-command\": exit st...  08-04 23:11:43\n",
      "coding-analysis    1629  Stopped running \"user-command\": exit st...  08-04 23:07:23\n",
      "coding-analysis     112  Stopped running \"user-command\": exit st...  08-04 23:04:51\n",
      "coding-analysis     113  Stopped running \"user-command\": exit st...  08-04 23:01:55\n",
      "coding-analysis    1602  Stopped running \"user-command\": exit st...  08-04 22:41:28\n",
      "coding-analysis    1629  Stopped running \"user-command\": exit st...  08-04 22:44:22\n",
      "coding-analysis    1628  Stopped running \"user-command\": exit st...  08-04 22:47:22\n",
      "coding-analysis     112  Stopped running \"user-command\": exit st...  08-04 22:42:40\n",
      "coding-analysis     113  Stopped running \"user-command\": exit st...  08-04 22:39:55\n",
      "coding-analysis    1602  Stopped running \"user-command\": exit st...  08-04 22:26:58\n",
      "coding-analysis      86  Stopped running \"user-command\": exit st...  08-04 22:18:48\n",
      "coding-analysis    1629  Stopped running \"user-command\": exit st...  08-04 22:23:05\n",
      "coding-analysis    1848  Stopped running \"user-command\": exit st...  08-04 22:09:53\n",
      "coding-analysis    1628  Stopped running \"user-command\": exit st...  08-04 22:18:54\n",
      "coding-analysis    1602  Stopped running \"user-command\": exit st...  08-04 22:09:38\n",
      "coding-analysis     112  Stopped running \"user-command\": exit st...  08-04 22:18:33\n",
      "coding-analysis    1848  Stopped running \"user-command\": exit st...  08-04 22:00:20\n",
      "coding-analysis    2113  Success                                     08-04 22:09:23\n",
      "coding-analysis      86  Stopped running \"user-command\": exit st...  08-04 22:05:19\n",
      "coding-analysis     113  Stopped running \"user-command\": exit st...  08-04 22:17:39\n",
      "coding-analysis     332  Stopped running \"user-command\": exit st...  08-04 21:56:16\n",
      "coding-analysis    1705  Success                                     08-04 22:11:33\n",
      "coding-analysis    1848  Stopped running \"user-command\": exit st...  08-04 21:51:26\n",
      "coding-analysis    1602  Stopped running \"user-command\": exit st...  08-04 21:55:35\n",
      "coding-analysis     332  Stopped running \"user-command\": exit st...  08-04 21:47:22\n",
      "coding-analysis      86  Stopped running \"user-command\": exit st...  08-04 21:50:53\n",
      "coding-analysis    1523  Success                                     08-04 21:56:25\n",
      "coding-analysis    1848  Stopped running \"user-command\": exit st...  08-04 21:42:18\n",
      "coding-analysis     112  Stopped running \"user-command\": exit st...  08-04 21:53:26\n",
      "coding-analysis    2134  Success                                     08-04 21:47:38\n",
      "coding-analysis     113  Stopped running \"user-command\": exit st...  08-04 21:48:51\n",
      "coding-analysis     332  Stopped running \"user-command\": exit st...  08-04 21:36:25\n",
      "coding-analysis    1484  Success                                     08-04 21:53:36\n",
      "coding-analysis      86  Stopped running \"user-command\": exit st...  08-04 21:35:31\n",
      "coding-analysis    1356  Success                                     08-04 21:37:13\n",
      "coding-analysis       2  Success                                     08-04 21:44:01\n",
      "coding-analysis     332  Stopped running \"user-command\": exit st...  08-04 21:25:35\n",
      "coding-analysis    1342  Success                                     08-04 21:32:21\n",
      "coding-analysis     791  Success                                     08-04 21:36:54\n",
      "coding-analysis     758  Success                                     08-04 21:28:08\n",
      "coding-analysis    1333  Success                                     08-04 21:31:21\n",
      "coding-analysis     788  Success                                     08-04 21:38:35\n",
      "coding-analysis     786  Success                                     08-04 21:35:50\n",
      "coding-analysis    1361  Success                                     08-04 21:36:57\n",
      "coding-analysis    1358  Success                                     08-04 21:33:25\n",
      "coding-analysis    1389  Success                                     08-04 21:32:21\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     171  Success                                     08-04 21:39:33\n",
      "coding-analysis     778  Success                                     08-04 21:35:34\n",
      "coding-analysis     736  Success                                     08-04 21:31:55\n",
      "coding-analysis     233  Success                                     08-04 21:37:29\n",
      "coding-analysis      49  Success                                     08-04 21:38:26\n",
      "coding-analysis     229  Success                                     08-04 21:30:54\n",
      "coding-analysis     290  Success                                     08-04 22:07:05\n",
      "coding-analysis    2274  Success                                     08-04 21:49:28\n",
      "coding-analysis    2273  Success                                     08-04 21:58:39\n",
      "coding-analysis    2272  Success                                     08-04 21:57:38\n",
      "coding-analysis    2271  Success                                     08-04 21:42:18\n",
      "coding-analysis    2270  Success                                     08-04 21:42:01\n",
      "coding-analysis    2269  Success                                     08-04 21:38:26\n",
      "coding-analysis    2268  Success                                     08-04 21:39:55\n",
      "coding-analysis    2267  Success                                     08-04 21:40:58\n",
      "coding-analysis    2266  Success                                     08-04 21:52:58\n",
      "coding-analysis    2265  Success                                     08-04 21:53:13\n",
      "coding-analysis    2264  Success                                     08-04 21:53:57\n",
      "coding-analysis    2263  Success                                     08-04 21:53:35\n",
      "coding-analysis    2262  Success                                     08-04 21:42:18\n",
      "coding-analysis    2261  Success                                     08-04 21:39:02\n",
      "coding-analysis    2260  Success                                     08-04 21:52:54\n",
      "coding-analysis    2259  Success                                     08-04 21:58:45\n",
      "coding-analysis    2258  Success                                     08-04 21:36:57\n",
      "coding-analysis    2257  Success                                     08-04 21:43:16\n",
      "coding-analysis    2256  Success                                     08-04 22:01:31\n",
      "coding-analysis    2255  Success                                     08-04 21:55:56\n",
      "coding-analysis    2254  Success                                     08-04 21:44:01\n",
      "coding-analysis    2253  Success                                     08-04 21:45:11\n",
      "coding-analysis    2252  Success                                     08-04 21:39:33\n",
      "coding-analysis    2251  Success                                     08-04 21:58:57\n",
      "coding-analysis    2250  Success                                     08-04 21:54:29\n",
      "coding-analysis    2249  Success                                     08-04 21:56:12\n",
      "coding-analysis    2248  Success                                     08-04 21:44:54\n",
      "coding-analysis    2247  Success                                     08-04 21:59:08\n",
      "coding-analysis    2246  Success                                     08-04 22:01:02\n",
      "coding-analysis    2245  Success                                     08-04 21:43:27\n",
      "coding-analysis    2244  Success                                     08-04 21:45:24\n",
      "coding-analysis    2243  Success                                     08-04 21:54:13\n",
      "coding-analysis    2242  Success                                     08-04 21:44:08\n",
      "coding-analysis    2241  Success                                     08-04 21:42:01\n",
      "coding-analysis    2240  Success                                     08-04 21:45:16\n",
      "coding-analysis    2239  Success                                     08-04 21:57:19\n",
      "coding-analysis    2238  Success                                     08-04 21:58:08\n",
      "coding-analysis    2237  Success                                     08-04 21:46:54\n",
      "coding-analysis    2236  Success                                     08-04 22:02:29\n",
      "coding-analysis    2235  Success                                     08-04 21:58:33\n",
      "coding-analysis    2234  Success                                     08-04 21:59:25\n",
      "coding-analysis    2233  Success                                     08-04 21:57:19\n",
      "coding-analysis    2232  Success                                     08-04 21:55:22\n",
      "coding-analysis    2231  Success                                     08-04 21:48:50\n",
      "coding-analysis    2230  Success                                     08-04 21:46:12\n",
      "coding-analysis    2229  Success                                     08-04 21:56:37\n",
      "coding-analysis    2228  Success                                     08-04 21:44:03\n",
      "coding-analysis    2227  Success                                     08-04 22:03:13\n",
      "coding-analysis    2226  Success                                     08-04 21:58:11\n",
      "coding-analysis    2225  Success                                     08-04 21:48:24\n",
      "coding-analysis    2224  Success                                     08-04 21:49:47\n",
      "coding-analysis    2223  Success                                     08-04 21:51:44\n",
      "coding-analysis    2222  Success                                     08-04 21:55:16\n",
      "coding-analysis    2221  Success                                     08-04 21:48:31\n",
      "coding-analysis    2220  Success                                     08-04 21:43:00\n",
      "coding-analysis    2219  Success                                     08-04 21:44:01\n",
      "coding-analysis    2218  Success                                     08-04 21:43:19\n",
      "coding-analysis    2217  Success                                     08-04 21:52:41\n",
      "coding-analysis    2216  Success                                     08-04 22:01:21\n",
      "coding-analysis    2215  Success                                     08-04 21:45:45\n",
      "coding-analysis    2214  Success                                     08-04 21:58:04\n",
      "coding-analysis    2213  Success                                     08-04 21:47:21\n",
      "coding-analysis    2212  Success                                     08-04 21:55:14\n",
      "coding-analysis    2211  Success                                     08-04 21:44:01\n",
      "coding-analysis    2210  Success                                     08-04 22:00:22\n",
      "coding-analysis    2209  Success                                     08-04 22:01:56\n",
      "coding-analysis    2208  Success                                     08-04 21:57:57\n",
      "coding-analysis    2207  Success                                     08-04 21:58:23\n",
      "coding-analysis    2206  Success                                     08-04 21:41:58\n",
      "coding-analysis    2205  Success                                     08-04 21:44:55\n",
      "coding-analysis    2204  Success                                     08-04 21:53:27\n",
      "coding-analysis    2203  Success                                     08-04 21:44:20\n",
      "coding-analysis    2202  Success                                     08-04 21:54:15\n",
      "coding-analysis    2201  Success                                     08-04 21:59:06\n",
      "coding-analysis    2200  Success                                     08-04 21:44:46\n",
      "coding-analysis    2199  Success                                     08-04 21:59:55\n",
      "coding-analysis    2198  Success                                     08-04 21:55:47\n",
      "coding-analysis    2197  Success                                     08-04 21:46:11\n",
      "coding-analysis    2196  Success                                     08-04 22:02:23\n",
      "coding-analysis    2195  Success                                     08-04 21:46:50\n",
      "coding-analysis    2194  Success                                     08-04 21:44:58\n",
      "coding-analysis    2193  Success                                     08-04 21:46:40\n",
      "coding-analysis    2192  Success                                     08-04 21:46:39\n",
      "coding-analysis    2191  Success                                     08-04 21:46:23\n",
      "coding-analysis    2190  Success                                     08-04 21:46:13\n",
      "coding-analysis    2189  Success                                     08-04 22:01:58\n",
      "coding-analysis    2188  Success                                     08-04 21:44:16\n",
      "coding-analysis    2187  Success                                     08-04 21:56:12\n",
      "coding-analysis    2186  Success                                     08-04 21:54:00\n",
      "coding-analysis    2185  Success                                     08-04 21:46:04\n",
      "coding-analysis    2184  Success                                     08-04 21:44:01\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    2183  Success                                     08-04 21:42:57\n",
      "coding-analysis    2182  Success                                     08-04 22:03:43\n",
      "coding-analysis    2181  Success                                     08-04 21:42:58\n",
      "coding-analysis    2180  Success                                     08-04 22:00:17\n",
      "coding-analysis    2179  Success                                     08-04 21:57:49\n",
      "coding-analysis    2178  Success                                     08-04 21:54:38\n",
      "coding-analysis    2177  Success                                     08-04 21:51:00\n",
      "coding-analysis    2176  Success                                     08-04 21:44:57\n",
      "coding-analysis    2175  Success                                     08-04 21:47:30\n",
      "coding-analysis    2174  Success                                     08-04 21:47:32\n",
      "coding-analysis    2173  Success                                     08-04 21:48:17\n",
      "coding-analysis    2172  Success                                     08-04 21:38:26\n",
      "coding-analysis    2171  Success                                     08-04 21:57:02\n",
      "coding-analysis    2170  Success                                     08-04 21:56:57\n",
      "coding-analysis    2169  Success                                     08-04 21:56:21\n",
      "coding-analysis    2168  Success                                     08-04 21:59:02\n",
      "coding-analysis    2167  Success                                     08-04 21:43:18\n",
      "coding-analysis    2166  Success                                     08-04 21:59:02\n",
      "coding-analysis    2165  Success                                     08-04 21:59:56\n",
      "coding-analysis    2164  Success                                     08-04 22:00:03\n",
      "coding-analysis    2163  Success                                     08-04 22:02:55\n",
      "coding-analysis    2162  Success                                     08-04 21:57:19\n",
      "coding-analysis    2161  Success                                     08-04 21:58:22\n",
      "coding-analysis    2160  Success                                     08-04 21:39:33\n",
      "coding-analysis    2159  Success                                     08-04 21:56:57\n",
      "coding-analysis    2158  Success                                     08-04 22:00:39\n",
      "coding-analysis    2157  Success                                     08-04 21:46:18\n",
      "coding-analysis    2156  Success                                     08-04 21:43:28\n",
      "coding-analysis    2155  Success                                     08-04 22:00:34\n",
      "coding-analysis    2154  Success                                     08-04 22:00:13\n",
      "coding-analysis    2153  Success                                     08-04 21:57:46\n",
      "coding-analysis    2152  Success                                     08-04 21:57:19\n",
      "coding-analysis    2151  Success                                     08-04 21:46:16\n",
      "coding-analysis    2150  Success                                     08-04 21:46:41\n",
      "coding-analysis    2149  Success                                     08-04 21:43:19\n",
      "coding-analysis    2148  Success                                     08-04 21:50:18\n",
      "coding-analysis    2147  Success                                     08-04 21:58:35\n",
      "coding-analysis    2146  Success                                     08-04 21:45:50\n",
      "coding-analysis    2145  Success                                     08-04 21:45:29\n",
      "coding-analysis    2144  Success                                     08-04 22:01:30\n",
      "coding-analysis    2143  Success                                     08-04 21:53:58\n",
      "coding-analysis    2142  Success                                     08-04 21:42:18\n",
      "coding-analysis    2141  Success                                     08-04 22:00:18\n",
      "coding-analysis    2140  Success                                     08-04 21:58:52\n",
      "coding-analysis    2139  Success                                     08-04 22:01:50\n",
      "coding-analysis    2138  Success                                     08-04 21:55:33\n",
      "coding-analysis    2137  Success                                     08-04 21:57:51\n",
      "coding-analysis    2136  Success                                     08-04 21:44:06\n",
      "coding-analysis    2135  Success                                     08-04 22:01:28\n",
      "coding-analysis    2134  worker was terminated                       08-04 21:30:28\n",
      "coding-analysis    2133  Success                                     08-04 21:57:19\n",
      "coding-analysis    2132  Success                                     08-04 21:46:35\n",
      "coding-analysis    2131  Success                                     08-04 22:02:37\n",
      "coding-analysis    2130  Success                                     08-04 21:45:28\n",
      "coding-analysis    2129  Success                                     08-04 21:39:55\n",
      "coding-analysis    2128  Success                                     08-04 21:56:15\n",
      "coding-analysis    2127  Success                                     08-04 21:42:40\n",
      "coding-analysis    2126  Success                                     08-04 21:53:30\n",
      "coding-analysis    2125  Success                                     08-04 21:56:12\n",
      "coding-analysis    2124  Success                                     08-04 21:49:30\n",
      "coding-analysis    2123  Success                                     08-04 21:42:37\n",
      "coding-analysis    2122  Success                                     08-04 21:58:41\n",
      "coding-analysis    2121  Success                                     08-04 21:57:19\n",
      "coding-analysis    2120  Success                                     08-04 21:53:59\n",
      "coding-analysis    2119  Success                                     08-04 21:44:49\n",
      "coding-analysis    2118  Success                                     08-04 21:59:14\n",
      "coding-analysis    2117  Success                                     08-04 22:01:33\n",
      "coding-analysis    2116  Success                                     08-04 21:42:58\n",
      "coding-analysis    2115  Success                                     08-04 21:55:51\n",
      "coding-analysis    2114  Success                                     08-04 21:46:30\n",
      "coding-analysis    2113  worker was terminated                       08-04 21:51:00\n",
      "coding-analysis    2112  Success                                     08-04 21:41:39\n",
      "coding-analysis    2111  Success                                     08-04 21:56:13\n",
      "coding-analysis    2110  Success                                     08-04 21:58:35\n",
      "coding-analysis    2109  Success                                     08-04 21:56:12\n",
      "coding-analysis    2108  Success                                     08-04 21:41:58\n",
      "coding-analysis    2107  Success                                     08-04 21:55:03\n",
      "coding-analysis    2106  Success                                     08-04 21:44:01\n",
      "coding-analysis    2105  Success                                     08-04 21:39:55\n",
      "coding-analysis    2104  Success                                     08-04 21:56:23\n",
      "coding-analysis    2103  Success                                     08-04 21:56:59\n",
      "coding-analysis    2102  Success                                     08-04 21:59:24\n",
      "coding-analysis    2101  Success                                     08-04 21:59:23\n",
      "coding-analysis    2100  Success                                     08-04 21:40:40\n",
      "coding-analysis    2099  Success                                     08-04 21:59:59\n",
      "coding-analysis    2098  Success                                     08-04 21:46:03\n",
      "coding-analysis    2097  Success                                     08-04 21:55:11\n",
      "coding-analysis    2096  Success                                     08-04 21:43:25\n",
      "coding-analysis    2095  Success                                     08-04 21:45:11\n",
      "coding-analysis    2094  Success                                     08-04 21:59:04\n",
      "coding-analysis    2093  Success                                     08-04 21:42:02\n",
      "coding-analysis    2092  Success                                     08-04 21:58:25\n",
      "coding-analysis    2091  Success                                     08-04 22:01:07\n",
      "coding-analysis    2090  Success                                     08-04 21:59:53\n",
      "coding-analysis    2089  Success                                     08-04 22:02:55\n",
      "coding-analysis    2088  Success                                     08-04 21:56:50\n",
      "coding-analysis    2087  Success                                     08-04 21:58:44\n",
      "coding-analysis    2086  Success                                     08-04 21:59:44\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    2085  Success                                     08-04 21:55:26\n",
      "coding-analysis    2084  Success                                     08-04 21:57:54\n",
      "coding-analysis    2083  Success                                     08-04 21:43:28\n",
      "coding-analysis    2082  Success                                     08-04 21:44:06\n",
      "coding-analysis    2081  Success                                     08-04 21:58:55\n",
      "coding-analysis    2080  Success                                     08-04 21:45:00\n",
      "coding-analysis    2079  Success                                     08-04 21:58:58\n",
      "coding-analysis    2078  Success                                     08-04 21:47:24\n",
      "coding-analysis    2077  Success                                     08-04 21:41:58\n",
      "coding-analysis    2076  Success                                     08-04 21:44:24\n",
      "coding-analysis    2075  Success                                     08-04 21:52:11\n",
      "coding-analysis    2074  Success                                     08-04 21:59:43\n",
      "coding-analysis    2073  Success                                     08-04 21:45:27\n",
      "coding-analysis    2072  Success                                     08-04 21:55:05\n",
      "coding-analysis    2071  Success                                     08-04 22:00:24\n",
      "coding-analysis    2070  Success                                     08-04 21:56:49\n",
      "coding-analysis    2069  Success                                     08-04 22:02:37\n",
      "coding-analysis    2068  Success                                     08-04 21:49:01\n",
      "coding-analysis    2067  Success                                     08-04 21:31:41\n",
      "coding-analysis    2066  Success                                     08-04 21:44:34\n",
      "coding-analysis    2065  Success                                     08-04 21:56:12\n",
      "coding-analysis    2064  Success                                     08-04 21:58:21\n",
      "coding-analysis    2063  Success                                     08-04 21:57:30\n",
      "coding-analysis    2062  Success                                     08-04 21:43:16\n",
      "coding-analysis    2061  Success                                     08-04 21:58:47\n",
      "coding-analysis    2060  Success                                     08-04 21:57:42\n",
      "coding-analysis    2059  Success                                     08-04 21:52:21\n",
      "coding-analysis    2058  Success                                     08-04 21:47:15\n",
      "coding-analysis    2057  Success                                     08-04 21:46:46\n",
      "coding-analysis    2056  Success                                     08-04 21:54:07\n",
      "coding-analysis    2055  Success                                     08-04 21:55:43\n",
      "coding-analysis    2054  Success                                     08-04 21:37:29\n",
      "coding-analysis    2053  Success                                     08-04 21:52:08\n",
      "coding-analysis    2052  Success                                     08-04 21:43:43\n",
      "coding-analysis    2051  Success                                     08-04 21:42:18\n",
      "coding-analysis    2050  Success                                     08-04 21:56:57\n",
      "coding-analysis    2049  Success                                     08-04 21:56:56\n",
      "coding-analysis    2048  Success                                     08-04 21:43:01\n",
      "coding-analysis    2047  Success                                     08-04 21:59:12\n",
      "coding-analysis    2046  Success                                     08-04 21:44:11\n",
      "coding-analysis    2045  Success                                     08-04 22:01:04\n",
      "coding-analysis    2044  Success                                     08-04 21:41:58\n",
      "coding-analysis    2043  Success                                     08-04 21:58:51\n",
      "coding-analysis    2042  Success                                     08-04 22:01:18\n",
      "coding-analysis    2041  Success                                     08-04 21:53:15\n",
      "coding-analysis    2040  Success                                     08-04 21:58:51\n",
      "coding-analysis    2039  Success                                     08-04 21:56:55\n",
      "coding-analysis    2038  Success                                     08-04 21:59:48\n",
      "coding-analysis    2037  Success                                     08-04 21:55:31\n",
      "coding-analysis    2036  Success                                     08-04 21:56:12\n",
      "coding-analysis    2035  Success                                     08-04 21:29:03\n",
      "coding-analysis    2034  Success                                     08-04 22:00:12\n",
      "coding-analysis    2033  Success                                     08-04 21:44:44\n",
      "coding-analysis    2032  Success                                     08-04 21:58:07\n",
      "coding-analysis    2031  Success                                     08-04 21:57:19\n",
      "coding-analysis    2030  Success                                     08-04 21:58:12\n",
      "coding-analysis    2029  Success                                     08-04 21:59:17\n",
      "coding-analysis    2028  Success                                     08-04 22:01:02\n",
      "coding-analysis    2027  Success                                     08-04 21:44:58\n",
      "coding-analysis    2026  Success                                     08-04 21:43:43\n",
      "coding-analysis    2025  Success                                     08-04 21:44:01\n",
      "coding-analysis    2024  Success                                     08-04 21:53:00\n",
      "coding-analysis    2023  Success                                     08-04 21:56:58\n",
      "coding-analysis    2022  Success                                     08-04 21:43:46\n",
      "coding-analysis    2021  Success                                     08-04 21:47:21\n",
      "coding-analysis    2020  Success                                     08-04 21:53:04\n",
      "coding-analysis    2019  Success                                     08-04 21:54:25\n",
      "coding-analysis    2018  Success                                     08-04 21:47:59\n",
      "coding-analysis    2017  Success                                     08-04 21:43:43\n",
      "coding-analysis    2016  Success                                     08-04 21:44:29\n",
      "coding-analysis    2015  Success                                     08-04 22:00:12\n",
      "coding-analysis    2014  Success                                     08-04 21:26:55\n",
      "coding-analysis    2013  Success                                     08-04 21:46:36\n",
      "coding-analysis    2012  Success                                     08-04 21:56:31\n",
      "coding-analysis    2011  Success                                     08-04 21:42:02\n",
      "coding-analysis    2010  Success                                     08-04 21:55:56\n",
      "coding-analysis    2009  Success                                     08-04 21:47:24\n",
      "coding-analysis    2008  Success                                     08-04 22:31:28\n",
      "coding-analysis    2007  Success                                     08-04 22:08:55\n",
      "coding-analysis    2006  Success                                     08-04 21:41:39\n",
      "coding-analysis    2005  Success                                     08-04 21:45:01\n",
      "coding-analysis    2004  Success                                     08-04 21:43:43\n",
      "coding-analysis    2003  Success                                     08-04 22:00:44\n",
      "coding-analysis    2002  Success                                     08-04 21:53:57\n",
      "coding-analysis    2001  Success                                     08-04 21:45:53\n",
      "coding-analysis    2000  Success                                     08-04 21:42:19\n",
      "coding-analysis    1999  Success                                     08-04 22:00:21\n",
      "coding-analysis    1998  Success                                     08-04 21:55:33\n",
      "coding-analysis    1997  Success                                     08-04 21:45:39\n",
      "coding-analysis    1996  Success                                     08-04 21:41:23\n",
      "coding-analysis    1995  Success                                     08-04 22:00:08\n",
      "coding-analysis    1994  Success                                     08-04 21:48:18\n",
      "coding-analysis    1993  Success                                     08-04 21:47:06\n",
      "coding-analysis    1992  Success                                     08-04 21:43:16\n",
      "coding-analysis    1991  Success                                     08-04 21:56:27\n",
      "coding-analysis    1990  Success                                     08-04 21:59:36\n",
      "coding-analysis    1989  Success                                     08-04 22:00:46\n",
      "coding-analysis    1988  Success                                     08-04 21:42:37\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1987  Success                                     08-04 21:52:51\n",
      "coding-analysis    1986  Success                                     08-04 21:55:28\n",
      "coding-analysis    1985  Success                                     08-04 21:57:55\n",
      "coding-analysis    1984  Success                                     08-04 21:39:55\n",
      "coding-analysis    1983  Success                                     08-04 21:47:35\n",
      "coding-analysis    1982  Success                                     08-04 22:01:21\n",
      "coding-analysis    1981  Success                                     08-04 21:43:43\n",
      "coding-analysis    1980  Success                                     08-04 21:46:52\n",
      "coding-analysis    1979  Success                                     08-04 22:02:15\n",
      "coding-analysis    1978  Success                                     08-04 21:45:03\n",
      "coding-analysis    1977  Success                                     08-04 21:58:07\n",
      "coding-analysis    1976  Success                                     08-04 21:48:38\n",
      "coding-analysis    1975  Success                                     08-04 21:56:57\n",
      "coding-analysis    1974  Success                                     08-04 21:42:57\n",
      "coding-analysis    1973  Success                                     08-04 21:59:07\n",
      "coding-analysis    1972  Success                                     08-04 21:58:23\n",
      "coding-analysis    1971  Success                                     08-04 21:58:56\n",
      "coding-analysis    1970  Success                                     08-04 21:59:21\n",
      "coding-analysis    1969  Success                                     08-04 21:47:02\n",
      "coding-analysis    1968  Success                                     08-04 22:00:32\n",
      "coding-analysis    1967  Success                                     08-04 21:57:19\n",
      "coding-analysis    1966  Success                                     08-04 22:03:50\n",
      "coding-analysis    1965  Success                                     08-04 21:47:59\n",
      "coding-analysis    1964  Success                                     08-04 22:04:32\n",
      "coding-analysis    1963  Success                                     08-04 22:02:58\n",
      "coding-analysis    1962  Success                                     08-04 22:00:42\n",
      "coding-analysis    1961  Success                                     08-04 22:02:14\n",
      "coding-analysis    1960  Success                                     08-04 21:45:53\n",
      "coding-analysis    1959  Success                                     08-04 21:44:58\n",
      "coding-analysis    1958  Success                                     08-04 22:01:02\n",
      "coding-analysis    1957  Success                                     08-04 21:58:29\n",
      "coding-analysis    1956  Success                                     08-04 22:03:25\n",
      "coding-analysis    1955  Success                                     08-04 21:45:21\n",
      "coding-analysis    1954  Success                                     08-04 21:57:19\n",
      "coding-analysis    1953  Success                                     08-04 22:00:28\n",
      "coding-analysis    1952  Success                                     08-04 21:45:49\n",
      "coding-analysis    1951  Success                                     08-04 21:58:41\n",
      "coding-analysis    1950  Success                                     08-04 21:51:22\n",
      "coding-analysis    1949  Success                                     08-04 21:58:40\n",
      "coding-analysis    1948  Success                                     08-04 22:02:55\n",
      "coding-analysis    1947  Success                                     08-04 21:56:53\n",
      "coding-analysis    1946  Success                                     08-04 21:47:03\n",
      "coding-analysis    1945  Success                                     08-04 21:46:29\n",
      "coding-analysis    1944  Success                                     08-04 21:53:57\n",
      "coding-analysis    1943  Success                                     08-04 21:44:42\n",
      "coding-analysis    1942  Success                                     08-04 21:57:59\n",
      "coding-analysis    1941  Success                                     08-04 21:56:53\n",
      "coding-analysis    1940  Success                                     08-04 21:45:44\n",
      "coding-analysis    1939  Success                                     08-04 21:47:02\n",
      "coding-analysis    1938  Success                                     08-04 21:57:23\n",
      "coding-analysis    1937  Success                                     08-04 21:50:04\n",
      "coding-analysis    1936  Success                                     08-04 21:47:25\n",
      "coding-analysis    1935  Success                                     08-04 21:57:19\n",
      "coding-analysis    1934  Success                                     08-04 22:00:39\n",
      "coding-analysis    1933  Success                                     08-04 21:59:20\n",
      "coding-analysis    1932  Success                                     08-04 22:01:27\n",
      "coding-analysis    1931  Success                                     08-04 22:00:19\n",
      "coding-analysis    1930  Success                                     08-04 21:58:38\n",
      "coding-analysis    1929  Success                                     08-04 21:48:32\n",
      "coding-analysis    1928  Success                                     08-04 21:54:25\n",
      "coding-analysis    1927  Success                                     08-04 22:03:47\n",
      "coding-analysis    1926  Success                                     08-04 21:59:25\n",
      "coding-analysis    1925  Success                                     08-04 21:58:05\n",
      "coding-analysis    1924  Success                                     08-04 22:00:30\n",
      "coding-analysis    1923  Success                                     08-04 21:59:21\n",
      "coding-analysis    1922  Success                                     08-04 21:46:24\n",
      "coding-analysis    1921  Success                                     08-04 22:02:55\n",
      "coding-analysis    1920  Success                                     08-04 21:54:38\n",
      "coding-analysis    1919  Success                                     08-04 21:59:28\n",
      "coding-analysis    1918  Success                                     08-04 21:44:46\n",
      "coding-analysis    1917  Success                                     08-04 21:50:42\n",
      "coding-analysis    1916  Success                                     08-04 21:57:26\n",
      "coding-analysis    1915  Success                                     08-04 21:41:58\n",
      "coding-analysis    1914  Success                                     08-04 21:47:19\n",
      "coding-analysis    1913  Success                                     08-04 21:44:06\n",
      "coding-analysis    1912  Success                                     08-04 21:48:08\n",
      "coding-analysis    1911  Success                                     08-04 22:01:40\n",
      "coding-analysis    1910  Success                                     08-04 21:45:14\n",
      "coding-analysis    1909  Success                                     08-04 22:00:11\n",
      "coding-analysis    1908  Success                                     08-04 21:56:37\n",
      "coding-analysis    1907  Success                                     08-04 21:38:26\n",
      "coding-analysis    1906  Success                                     08-04 21:57:19\n",
      "coding-analysis    1905  Success                                     08-04 21:40:39\n",
      "coding-analysis    1904  Success                                     08-04 21:51:22\n",
      "coding-analysis    1903  Success                                     08-04 21:52:08\n",
      "coding-analysis    1902  Success                                     08-04 21:56:38\n",
      "coding-analysis    1901  Success                                     08-04 22:02:38\n",
      "coding-analysis    1900  Success                                     08-04 21:55:15\n",
      "coding-analysis    1899  Success                                     08-04 22:03:23\n",
      "coding-analysis    1898  Success                                     08-04 21:52:51\n",
      "coding-analysis    1897  Success                                     08-04 21:45:59\n",
      "coding-analysis    1896  Success                                     08-04 21:42:37\n",
      "coding-analysis    1895  Success                                     08-04 21:59:05\n",
      "coding-analysis    1894  Success                                     08-04 21:52:06\n",
      "coding-analysis    1893  Success                                     08-04 22:02:08\n",
      "coding-analysis    1892  Success                                     08-04 21:45:38\n",
      "coding-analysis    1891  Success                                     08-04 22:03:33\n",
      "coding-analysis    1890  Success                                     08-04 22:07:55\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1889  Success                                     08-04 22:01:02\n",
      "coding-analysis    1888  Success                                     08-04 21:41:58\n",
      "coding-analysis    1887  Success                                     08-04 22:02:00\n",
      "coding-analysis    1886  Success                                     08-04 21:39:10\n",
      "coding-analysis    1885  Success                                     08-04 22:01:25\n",
      "coding-analysis    1884  Success                                     08-04 22:01:04\n",
      "coding-analysis    1883  Success                                     08-04 21:44:58\n",
      "coding-analysis    1882  Success                                     08-04 21:59:50\n",
      "coding-analysis    1881  Success                                     08-04 21:46:02\n",
      "coding-analysis    1880  Success                                     08-04 21:48:15\n",
      "coding-analysis    1879  Success                                     08-04 22:00:14\n",
      "coding-analysis    1878  Success                                     08-04 21:44:58\n",
      "coding-analysis    1877  Success                                     08-04 22:02:58\n",
      "coding-analysis    1876  Success                                     08-04 22:04:44\n",
      "coding-analysis    1875  Success                                     08-04 21:54:06\n",
      "coding-analysis    1874  Success                                     08-04 21:57:22\n",
      "coding-analysis    1873  Success                                     08-04 22:03:38\n",
      "coding-analysis    1872  Success                                     08-04 22:02:10\n",
      "coding-analysis    1871  Success                                     08-04 21:41:58\n",
      "coding-analysis    1870  Success                                     08-04 21:46:33\n",
      "coding-analysis    1869  Success                                     08-04 21:57:43\n",
      "coding-analysis    1868  Success                                     08-04 22:04:21\n",
      "coding-analysis    1867  Success                                     08-04 21:50:38\n",
      "coding-analysis    1866  Success                                     08-04 21:56:13\n",
      "coding-analysis    1865  Success                                     08-04 22:01:41\n",
      "coding-analysis    1864  Success                                     08-04 21:54:35\n",
      "coding-analysis    1863  Success                                     08-04 22:05:12\n",
      "coding-analysis    1862  Success                                     08-04 22:08:59\n",
      "coding-analysis    1861  Success                                     08-04 22:02:38\n",
      "coding-analysis    1860  Success                                     08-04 21:50:03\n",
      "coding-analysis    1859  Success                                     08-04 22:11:23\n",
      "coding-analysis    1858  Success                                     08-04 22:03:34\n",
      "coding-analysis    1857  Success                                     08-04 21:58:38\n",
      "coding-analysis    1856  Success                                     08-04 21:58:47\n",
      "coding-analysis    1855  Success                                     08-04 22:05:02\n",
      "coding-analysis    1854  Success                                     08-04 22:08:52\n",
      "coding-analysis    1853  Success                                     08-04 21:50:02\n",
      "coding-analysis    1852  Success                                     08-04 21:58:17\n",
      "coding-analysis    1851  Success                                     08-04 21:50:58\n",
      "coding-analysis    1850  Success                                     08-04 22:04:45\n",
      "coding-analysis    1849  Success                                     08-04 22:08:33\n",
      "coding-analysis    1848  Stopped running \"user-command\": exit st...  08-04 21:31:44\n",
      "coding-analysis    1847  Success                                     08-04 21:51:22\n",
      "coding-analysis    1846  Success                                     08-04 21:52:49\n",
      "coding-analysis    1845  Success                                     08-04 22:06:58\n",
      "coding-analysis    1844  Success                                     08-04 22:00:23\n",
      "coding-analysis    1843  Success                                     08-04 22:02:32\n",
      "coding-analysis    1842  Success                                     08-04 21:50:14\n",
      "coding-analysis    1841  Success                                     08-04 22:09:56\n",
      "coding-analysis    1840  Success                                     08-04 21:45:03\n",
      "coding-analysis    1839  Success                                     08-04 22:09:03\n",
      "coding-analysis    1838  Success                                     08-04 22:05:13\n",
      "coding-analysis    1837  Success                                     08-04 22:07:02\n",
      "coding-analysis    1836  Success                                     08-04 22:06:09\n",
      "coding-analysis    1835  Success                                     08-04 21:57:19\n",
      "coding-analysis    1834  Success                                     08-04 21:56:12\n",
      "coding-analysis    1833  Success                                     08-04 21:47:16\n",
      "coding-analysis    1832  Success                                     08-04 22:08:44\n",
      "coding-analysis    1831  Success                                     08-04 22:07:57\n",
      "coding-analysis    1830  Success                                     08-04 21:49:53\n",
      "coding-analysis    1829  Success                                     08-04 22:07:42\n",
      "coding-analysis    1828  Success                                     08-04 22:12:13\n",
      "coding-analysis    1827  Success                                     08-04 22:03:39\n",
      "coding-analysis    1826  Success                                     08-04 21:48:04\n",
      "coding-analysis    1825  Success                                     08-04 21:45:59\n",
      "coding-analysis    1824  Success                                     08-04 21:57:45\n",
      "coding-analysis    1823  Success                                     08-04 22:05:38\n",
      "coding-analysis    1822  Success                                     08-04 21:50:16\n",
      "coding-analysis    1821  Success                                     08-04 21:58:11\n",
      "coding-analysis    1820  Success                                     08-04 22:04:25\n",
      "coding-analysis    1819  Success                                     08-04 21:51:49\n",
      "coding-analysis    1818  Success                                     08-04 22:08:56\n",
      "coding-analysis    1817  Success                                     08-04 22:03:12\n",
      "coding-analysis    1816  Success                                     08-04 22:00:26\n",
      "coding-analysis    1815  Success                                     08-04 21:51:44\n",
      "coding-analysis    1814  Success                                     08-04 21:54:17\n",
      "coding-analysis    1813  Success                                     08-04 21:49:30\n",
      "coding-analysis    1812  Success                                     08-04 22:07:25\n",
      "coding-analysis    1811  Success                                     08-04 22:03:02\n",
      "coding-analysis    1810  Success                                     08-04 22:05:32\n",
      "coding-analysis    1809  Success                                     08-04 22:05:40\n",
      "coding-analysis    1808  Success                                     08-04 22:01:16\n",
      "coding-analysis    1807  Success                                     08-04 22:02:07\n",
      "coding-analysis    1806  Success                                     08-04 22:03:04\n",
      "coding-analysis    1805  Success                                     08-04 21:50:36\n",
      "coding-analysis    1804  Success                                     08-04 21:42:18\n",
      "coding-analysis    1803  Success                                     08-04 21:49:04\n",
      "coding-analysis    1802  Success                                     08-04 22:10:04\n",
      "coding-analysis    1801  Success                                     08-04 22:07:36\n",
      "coding-analysis    1800  Success                                     08-04 22:08:14\n",
      "coding-analysis    1799  Success                                     08-04 21:52:06\n",
      "coding-analysis    1798  Success                                     08-04 22:15:58\n",
      "coding-analysis    1797  Success                                     08-04 22:03:48\n",
      "coding-analysis    1796  Success                                     08-04 21:49:43\n",
      "coding-analysis    1795  Success                                     08-04 22:06:54\n",
      "coding-analysis    1794  Success                                     08-04 21:55:12\n",
      "coding-analysis    1793  Success                                     08-04 22:03:32\n",
      "coding-analysis    1792  Success                                     08-04 21:46:41\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1791  Success                                     08-04 22:04:57\n",
      "coding-analysis    1790  Success                                     08-04 22:07:03\n",
      "coding-analysis    1789  Success                                     08-04 21:55:12\n",
      "coding-analysis    1788  Success                                     08-04 21:51:01\n",
      "coding-analysis    1787  Success                                     08-04 22:09:23\n",
      "coding-analysis    1786  Success                                     08-04 22:08:28\n",
      "coding-analysis    1785  Success                                     08-04 21:51:44\n",
      "coding-analysis    1784  Success                                     08-04 22:05:22\n",
      "coding-analysis    1783  Success                                     08-04 22:01:07\n",
      "coding-analysis    1782  Success                                     08-04 22:04:17\n",
      "coding-analysis    1781  Success                                     08-04 22:03:08\n",
      "coding-analysis    1780  Success                                     08-04 21:52:35\n",
      "coding-analysis    1779  Success                                     08-04 22:02:25\n",
      "coding-analysis    1778  Success                                     08-04 21:49:28\n",
      "coding-analysis    1777  Success                                     08-04 21:48:00\n",
      "coding-analysis    1776  Success                                     08-04 22:06:09\n",
      "coding-analysis    1775  Success                                     08-04 21:49:56\n",
      "coding-analysis    1774  Success                                     08-04 21:56:12\n",
      "coding-analysis    1773  Success                                     08-04 22:00:10\n",
      "coding-analysis    1772  Success                                     08-04 21:49:37\n",
      "coding-analysis    1771  Success                                     08-04 22:06:02\n",
      "coding-analysis    1770  Success                                     08-04 21:48:33\n",
      "coding-analysis    1769  Success                                     08-04 22:01:48\n",
      "coding-analysis    1768  Success                                     08-04 21:47:59\n",
      "coding-analysis    1767  Success                                     08-04 22:03:01\n",
      "coding-analysis    1766  Success                                     08-04 22:04:56\n",
      "coding-analysis    1765  Success                                     08-04 22:07:01\n",
      "coding-analysis    1764  Success                                     08-04 22:12:57\n",
      "coding-analysis    1763  Success                                     08-04 21:58:09\n",
      "coding-analysis    1762  Success                                     08-04 21:53:09\n",
      "coding-analysis    1761  Success                                     08-04 22:05:40\n",
      "coding-analysis    1760  Success                                     08-04 21:51:22\n",
      "coding-analysis    1759  Success                                     08-04 21:51:25\n",
      "coding-analysis    1758  Success                                     08-04 21:45:28\n",
      "coding-analysis    1757  Success                                     08-04 22:07:11\n",
      "coding-analysis    1756  Success                                     08-04 22:04:22\n",
      "coding-analysis    1755  Success                                     08-04 21:43:16\n",
      "coding-analysis    1754  Success                                     08-04 22:02:40\n",
      "coding-analysis    1753  Success                                     08-04 22:05:30\n",
      "coding-analysis    1752  Success                                     08-04 22:08:05\n",
      "coding-analysis    1751  Success                                     08-04 21:48:43\n",
      "coding-analysis    1750  Success                                     08-04 22:08:07\n",
      "coding-analysis    1749  Success                                     08-04 22:06:37\n",
      "coding-analysis    1748  Success                                     08-04 22:02:34\n",
      "coding-analysis    1747  Success                                     08-04 21:50:47\n",
      "coding-analysis    1746  Success                                     08-04 22:09:00\n",
      "coding-analysis    1745  Success                                     08-04 21:58:16\n",
      "coding-analysis    1744  Success                                     08-04 21:45:27\n",
      "coding-analysis    1743  Success                                     08-04 21:59:28\n",
      "coding-analysis    1742  Success                                     08-04 21:44:59\n",
      "coding-analysis    1741  Success                                     08-04 21:48:43\n",
      "coding-analysis    1740  Success                                     08-04 22:09:21\n",
      "coding-analysis    1739  Success                                     08-04 22:01:23\n",
      "coding-analysis    1738  Success                                     08-04 22:02:17\n",
      "coding-analysis    1737  Success                                     08-04 21:50:40\n",
      "coding-analysis    1736  Success                                     08-04 21:45:25\n",
      "coding-analysis    1735  Success                                     08-04 21:57:45\n",
      "coding-analysis    1734  Success                                     08-04 22:04:48\n",
      "coding-analysis    1733  Success                                     08-04 22:01:39\n",
      "coding-analysis    1732  Success                                     08-04 22:06:25\n",
      "coding-analysis    1731  Success                                     08-04 21:46:18\n",
      "coding-analysis    1730  Success                                     08-04 22:05:55\n",
      "coding-analysis    1729  Success                                     08-04 22:01:40\n",
      "coding-analysis    1728  Success                                     08-04 21:51:44\n",
      "coding-analysis    1727  Success                                     08-04 22:02:39\n",
      "coding-analysis    1726  Success                                     08-04 21:51:22\n",
      "coding-analysis    1725  Success                                     08-04 21:48:06\n",
      "coding-analysis    1724  Success                                     08-04 22:13:25\n",
      "coding-analysis    1723  Success                                     08-04 21:47:41\n",
      "coding-analysis    1722  Success                                     08-04 22:05:54\n",
      "coding-analysis    1721  Success                                     08-04 22:00:30\n",
      "coding-analysis    1720  Success                                     08-04 21:57:47\n",
      "coding-analysis    1719  Success                                     08-04 21:47:28\n",
      "coding-analysis    1718  Success                                     08-04 22:06:29\n",
      "coding-analysis    1717  Success                                     08-04 22:00:00\n",
      "coding-analysis    1716  Success                                     08-04 21:58:23\n",
      "coding-analysis    1715  Success                                     08-04 21:47:33\n",
      "coding-analysis    1714  Success                                     08-04 22:05:06\n",
      "coding-analysis    1713  Success                                     08-04 21:49:57\n",
      "coding-analysis    1712  Success                                     08-04 22:00:38\n",
      "coding-analysis    1711  Success                                     08-04 22:04:35\n",
      "coding-analysis    1710  Success                                     08-04 22:02:55\n",
      "coding-analysis    1709  Success                                     08-04 22:08:50\n",
      "coding-analysis    1708  Success                                     08-04 21:53:08\n",
      "coding-analysis    1707  Success                                     08-04 22:04:54\n",
      "coding-analysis    1706  Success                                     08-04 22:00:08\n",
      "coding-analysis    1705  worker was terminated                       08-04 21:42:37\n",
      "coding-analysis    1704  Success                                     08-04 21:33:49\n",
      "coding-analysis    1703  Success                                     08-04 22:02:22\n",
      "coding-analysis    1702  Success                                     08-04 22:05:55\n",
      "coding-analysis    1701  Success                                     08-04 21:58:52\n",
      "coding-analysis    1700  Success                                     08-04 21:47:41\n",
      "coding-analysis    1699  Success                                     08-04 22:01:02\n",
      "coding-analysis    1698  Success                                     08-04 21:50:21\n",
      "coding-analysis    1697  Success                                     08-04 21:52:20\n",
      "coding-analysis    1696  Success                                     08-04 22:05:33\n",
      "coding-analysis    1695  Success                                     08-04 22:06:46\n",
      "coding-analysis    1694  Success                                     08-04 21:51:22\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1693  Success                                     08-04 22:03:31\n",
      "coding-analysis    1692  Success                                     08-04 22:06:57\n",
      "coding-analysis    1691  Success                                     08-04 21:53:03\n",
      "coding-analysis    1690  Success                                     08-04 22:07:27\n",
      "coding-analysis    1689  Success                                     08-04 21:53:13\n",
      "coding-analysis    1688  Success                                     08-04 21:46:14\n",
      "coding-analysis    1687  Success                                     08-04 22:12:50\n",
      "coding-analysis    1686  Success                                     08-04 21:50:40\n",
      "coding-analysis    1685  Success                                     08-04 22:05:10\n",
      "coding-analysis    1684  Success                                     08-04 21:52:06\n",
      "coding-analysis    1683  Success                                     08-04 22:00:05\n",
      "coding-analysis    1682  Success                                     08-04 22:14:57\n",
      "coding-analysis    1681  Success                                     08-04 22:05:18\n",
      "coding-analysis    1680  Success                                     08-04 22:07:27\n",
      "coding-analysis    1679  Success                                     08-04 21:57:36\n",
      "coding-analysis    1678  Success                                     08-04 22:08:33\n",
      "coding-analysis    1677  Success                                     08-04 22:02:55\n",
      "coding-analysis    1676  Success                                     08-04 21:49:56\n",
      "coding-analysis    1675  Success                                     08-04 22:03:09\n",
      "coding-analysis    1674  Success                                     08-04 22:00:28\n",
      "coding-analysis    1673  Success                                     08-04 22:06:26\n",
      "coding-analysis    1672  Success                                     08-04 22:01:05\n",
      "coding-analysis    1671  Success                                     08-04 21:54:27\n",
      "coding-analysis    1670  Success                                     08-04 21:57:19\n",
      "coding-analysis    1669  Success                                     08-04 22:04:58\n",
      "coding-analysis    1668  Success                                     08-04 22:01:08\n",
      "coding-analysis    1667  Success                                     08-04 22:06:03\n",
      "coding-analysis    1666  Success                                     08-04 22:08:26\n",
      "coding-analysis    1665  Success                                     08-04 21:54:33\n",
      "coding-analysis    1664  Success                                     08-04 22:02:55\n",
      "coding-analysis    1663  Success                                     08-04 21:50:19\n",
      "coding-analysis    1662  Success                                     08-04 21:55:30\n",
      "coding-analysis    1661  Success                                     08-04 22:00:36\n",
      "coding-analysis    1660  Success                                     08-04 22:10:07\n",
      "coding-analysis    1659  Success                                     08-04 22:11:44\n",
      "coding-analysis    1658  Success                                     08-04 21:53:22\n",
      "coding-analysis    1657  Success                                     08-04 21:49:03\n",
      "coding-analysis    1656  Success                                     08-04 21:49:39\n",
      "coding-analysis    1655  Success                                     08-04 21:49:06\n",
      "coding-analysis    1654  Success                                     08-04 22:01:22\n",
      "coding-analysis    1653  Success                                     08-04 21:49:28\n",
      "coding-analysis    1652  Success                                     08-04 21:47:21\n",
      "coding-analysis    1651  Success                                     08-04 21:55:18\n",
      "coding-analysis    1650  Success                                     08-04 21:55:00\n",
      "coding-analysis    1649  Success                                     08-04 22:02:55\n",
      "coding-analysis    1648  Success                                     08-04 22:00:25\n",
      "coding-analysis    1647  Success                                     08-04 21:49:07\n",
      "coding-analysis    1646  Success                                     08-04 22:06:07\n",
      "coding-analysis    1645  Success                                     08-04 22:02:31\n",
      "coding-analysis    1644  Success                                     08-04 22:01:59\n",
      "coding-analysis    1643  Success                                     08-04 21:49:28\n",
      "coding-analysis    1642  Success                                     08-04 22:04:47\n",
      "coding-analysis    1641  Success                                     08-04 22:00:15\n",
      "coding-analysis    1640  Success                                     08-04 21:51:22\n",
      "coding-analysis    1639  Success                                     08-04 22:00:39\n",
      "coding-analysis    1638  Success                                     08-04 21:43:01\n",
      "coding-analysis    1637  Success                                     08-04 21:55:08\n",
      "coding-analysis    1636  Success                                     08-04 21:50:19\n",
      "coding-analysis    1635  Success                                     08-04 21:41:39\n",
      "coding-analysis    1634  Success                                     08-04 22:02:33\n",
      "coding-analysis    1633  Success                                     08-04 22:03:49\n",
      "coding-analysis    1632  Success                                     08-04 21:55:32\n",
      "coding-analysis    1631  Success                                     08-04 21:52:06\n",
      "coding-analysis    1630  Success                                     08-04 21:54:03\n",
      "coding-analysis    1629  Stopped running \"user-command\": exit st...  08-04 22:02:27\n",
      "coding-analysis    1628  Stopped running \"user-command\": exit st...  08-04 21:56:48\n",
      "coding-analysis    1627  Success                                     08-04 21:48:24\n",
      "coding-analysis    1626  Success                                     08-04 22:04:14\n",
      "coding-analysis    1625  Success                                     08-04 22:00:42\n",
      "coding-analysis    1624  Success                                     08-04 22:03:29\n",
      "coding-analysis    1623  Success                                     08-04 22:06:34\n",
      "coding-analysis    1622  Success                                     08-04 22:00:13\n",
      "coding-analysis    1621  Success                                     08-04 22:04:24\n",
      "coding-analysis    1620  Success                                     08-04 21:35:32\n",
      "coding-analysis    1619  Success                                     08-04 21:48:04\n",
      "coding-analysis    1618  Success                                     08-04 22:00:23\n",
      "coding-analysis    1617  Success                                     08-04 22:00:25\n",
      "coding-analysis    1616  Success                                     08-04 21:31:14\n",
      "coding-analysis    1615  Success                                     08-04 21:34:32\n",
      "coding-analysis    1614  Success                                     08-04 22:06:17\n",
      "coding-analysis    1613  Success                                     08-04 21:57:39\n",
      "coding-analysis    1612  Success                                     08-04 21:33:21\n",
      "coding-analysis    1611  Success                                     08-04 21:34:32\n",
      "coding-analysis    1610  Success                                     08-04 21:50:08\n",
      "coding-analysis    1609  Success                                     08-04 21:57:51\n",
      "coding-analysis    1608  Success                                     08-04 21:32:48\n",
      "coding-analysis    1607  Success                                     08-04 21:32:50\n",
      "coding-analysis    1606  Success                                     08-04 21:56:20\n",
      "coding-analysis    1605  Success                                     08-04 21:34:51\n",
      "coding-analysis    1604  Success                                     08-04 22:11:40\n",
      "coding-analysis    1603  Success                                     08-04 21:41:39\n",
      "coding-analysis    1602  Stopped running \"user-command\": exit st...  08-04 21:39:04\n",
      "coding-analysis    1601  Success                                     08-04 21:35:09\n",
      "coding-analysis    1600  Success                                     08-04 22:07:50\n",
      "coding-analysis    1599  Success                                     08-04 21:38:07\n",
      "coding-analysis    1598  Success                                     08-04 21:34:32\n",
      "coding-analysis    1597  Success                                     08-04 21:38:26\n",
      "coding-analysis    1596  Success                                     08-04 22:05:00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1595  Success                                     08-04 21:37:29\n",
      "coding-analysis    1594  Success                                     08-04 21:51:44\n",
      "coding-analysis    1593  Success                                     08-04 21:33:59\n",
      "coding-analysis    1592  Success                                     08-04 21:33:58\n",
      "coding-analysis    1591  Success                                     08-04 21:34:51\n",
      "coding-analysis    1590  Success                                     08-04 21:43:21\n",
      "coding-analysis    1589  Success                                     08-04 21:41:41\n",
      "coding-analysis    1588  Success                                     08-04 21:35:31\n",
      "coding-analysis    1587  Success                                     08-04 21:42:37\n",
      "coding-analysis    1586  Success                                     08-04 21:38:26\n",
      "coding-analysis    1585  Success                                     08-04 21:37:29\n",
      "coding-analysis    1584  Success                                     08-04 21:39:01\n",
      "coding-analysis    1583  Success                                     08-04 21:35:31\n",
      "coding-analysis    1582  Success                                     08-04 21:36:25\n",
      "coding-analysis    1581  Success                                     08-04 21:22:53\n",
      "coding-analysis    1580  Success                                     08-04 21:39:38\n",
      "coding-analysis    1579  Success                                     08-04 21:39:05\n",
      "coding-analysis    1578  Success                                     08-04 21:35:32\n",
      "coding-analysis    1577  Success                                     08-04 21:36:07\n",
      "coding-analysis    1576  Success                                     08-04 21:42:18\n",
      "coding-analysis    1575  Success                                     08-04 21:33:19\n",
      "coding-analysis    1574  Success                                     08-04 21:37:13\n",
      "coding-analysis    1573  Success                                     08-04 21:35:09\n",
      "coding-analysis    1572  Success                                     08-04 21:42:01\n",
      "coding-analysis    1571  Success                                     08-04 21:38:56\n",
      "coding-analysis    1570  Success                                     08-04 21:34:51\n",
      "coding-analysis    1569  Success                                     08-04 21:34:51\n",
      "coding-analysis    1568  Success                                     08-04 21:34:34\n",
      "coding-analysis    1567  Success                                     08-04 21:38:55\n",
      "coding-analysis    1566  Success                                     08-04 21:38:07\n",
      "coding-analysis    1565  Success                                     08-04 21:37:48\n",
      "coding-analysis    1564  Success                                     08-04 21:37:29\n",
      "coding-analysis    1563  Success                                     08-04 21:35:48\n",
      "coding-analysis    1562  Success                                     08-04 21:39:56\n",
      "coding-analysis    1561  Success                                     08-04 21:38:07\n",
      "coding-analysis    1560  Success                                     08-04 21:37:29\n",
      "coding-analysis    1559  Success                                     08-04 21:41:58\n",
      "coding-analysis    1558  Success                                     08-04 21:42:57\n",
      "coding-analysis    1557  Success                                     08-04 21:33:11\n",
      "coding-analysis    1556  Success                                     08-04 21:38:54\n",
      "coding-analysis    1555  Success                                     08-04 21:42:37\n",
      "coding-analysis    1554  Success                                     08-04 21:32:21\n",
      "coding-analysis    1553  Success                                     08-04 21:40:18\n",
      "coding-analysis    1552  Success                                     08-04 21:36:45\n",
      "coding-analysis    1551  Success                                     08-04 21:36:55\n",
      "coding-analysis    1550  Success                                     08-04 21:34:05\n",
      "coding-analysis    1549  Success                                     08-04 21:34:51\n",
      "coding-analysis    1548  Success                                     08-04 21:40:36\n",
      "coding-analysis    1547  Success                                     08-04 21:34:07\n",
      "coding-analysis    1546  Success                                     08-04 21:32:53\n",
      "coding-analysis    1545  Success                                     08-04 21:35:09\n",
      "coding-analysis    1544  Success                                     08-04 21:38:07\n",
      "coding-analysis    1543  Success                                     08-04 21:31:14\n",
      "coding-analysis    1542  Success                                     08-04 21:35:49\n",
      "coding-analysis    1541  Success                                     08-04 21:39:33\n",
      "coding-analysis    1540  Success                                     08-04 21:46:56\n",
      "coding-analysis    1539  Success                                     08-04 21:38:56\n",
      "coding-analysis    1538  Success                                     08-04 21:38:54\n",
      "coding-analysis    1537  Success                                     08-04 21:36:45\n",
      "coding-analysis    1536  Success                                     08-04 21:36:45\n",
      "coding-analysis    1535  Success                                     08-04 21:34:16\n",
      "coding-analysis    1534  Success                                     08-04 21:33:12\n",
      "coding-analysis    1533  Success                                     08-04 21:41:58\n",
      "coding-analysis    1532  Success                                     08-04 21:39:12\n",
      "coding-analysis    1531  Success                                     08-04 21:35:13\n",
      "coding-analysis    1530  Success                                     08-04 21:35:09\n",
      "coding-analysis    1529  Success                                     08-04 21:34:51\n",
      "coding-analysis    1528  Success                                     08-04 21:36:45\n",
      "coding-analysis    1527  Success                                     08-04 21:32:39\n",
      "coding-analysis    1526  Success                                     08-04 21:34:51\n",
      "coding-analysis    1525  Success                                     08-04 21:42:18\n",
      "coding-analysis    1524  Success                                     08-04 21:52:37\n",
      "coding-analysis    1523  worker was terminated                       08-04 21:31:36\n",
      "coding-analysis    1522  Success                                     08-04 21:38:26\n",
      "coding-analysis    1521  Success                                     08-04 21:35:09\n",
      "coding-analysis    1520  Success                                     08-04 21:36:25\n",
      "coding-analysis    1519  Success                                     08-04 21:33:33\n",
      "coding-analysis    1518  Success                                     08-04 21:36:25\n",
      "coding-analysis    1517  Success                                     08-04 21:35:48\n",
      "coding-analysis    1516  Success                                     08-04 21:22:02\n",
      "coding-analysis    1515  Success                                     08-04 21:29:15\n",
      "coding-analysis    1514  Success                                     08-04 21:26:05\n",
      "coding-analysis    1513  Success                                     08-04 21:24:44\n",
      "coding-analysis    1512  Success                                     08-04 21:27:16\n",
      "coding-analysis    1511  Success                                     08-04 21:30:46\n",
      "coding-analysis    1510  Success                                     08-04 21:27:14\n",
      "coding-analysis    1509  Success                                     08-04 21:27:33\n",
      "coding-analysis    1508  Success                                     08-04 21:28:15\n",
      "coding-analysis    1507  Success                                     08-04 21:22:34\n",
      "coding-analysis    1506  Success                                     08-04 21:27:17\n",
      "coding-analysis    1505  Success                                     08-04 21:25:20\n",
      "coding-analysis    1504  Success                                     08-04 21:23:07\n",
      "coding-analysis    1503  Success                                     08-04 21:22:35\n",
      "coding-analysis    1502  Success                                     08-04 21:20:52\n",
      "coding-analysis    1501  Success                                     08-04 21:28:21\n",
      "coding-analysis    1500  Success                                     08-04 21:24:27\n",
      "coding-analysis    1499  Success                                     08-04 21:25:39\n",
      "coding-analysis    1498  Success                                     08-04 21:26:22\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1497  Success                                     08-04 21:28:42\n",
      "coding-analysis    1496  Success                                     08-04 21:28:47\n",
      "coding-analysis    1495  Success                                     08-04 21:23:40\n",
      "coding-analysis    1494  Success                                     08-04 21:27:32\n",
      "coding-analysis    1493  Success                                     08-04 21:25:02\n",
      "coding-analysis    1492  Success                                     08-04 21:27:59\n",
      "coding-analysis    1491  Success                                     08-04 21:26:29\n",
      "coding-analysis    1490  Success                                     08-04 21:32:21\n",
      "coding-analysis    1489  Success                                     08-04 21:30:39\n",
      "coding-analysis    1488  Success                                     08-04 21:30:22\n",
      "coding-analysis    1487  Success                                     08-04 21:28:39\n",
      "coding-analysis    1486  Success                                     08-04 21:30:59\n",
      "coding-analysis    1485  Success                                     08-04 21:29:29\n",
      "coding-analysis    1484  worker was terminated                       08-04 21:22:02\n",
      "coding-analysis    1483  Success                                     08-04 21:30:08\n",
      "coding-analysis    1482  Success                                     08-04 21:27:53\n",
      "coding-analysis    1481  Success                                     08-04 21:31:21\n",
      "coding-analysis    1480  Success                                     08-04 21:27:29\n",
      "coding-analysis    1479  Success                                     08-04 21:26:51\n",
      "coding-analysis    1478  Success                                     08-04 21:33:30\n",
      "coding-analysis    1477  Success                                     08-04 21:26:55\n",
      "coding-analysis    1476  Success                                     08-04 21:26:59\n",
      "coding-analysis    1475  Success                                     08-04 21:32:21\n",
      "coding-analysis    1474  Success                                     08-04 21:30:05\n",
      "coding-analysis    1473  Success                                     08-04 21:29:02\n",
      "coding-analysis    1472  Success                                     08-04 21:27:45\n",
      "coding-analysis    1471  Success                                     08-04 21:26:52\n",
      "coding-analysis    1470  Success                                     08-04 21:27:34\n",
      "coding-analysis    1469  Success                                     08-04 21:35:16\n",
      "coding-analysis    1468  Success                                     08-04 21:31:14\n",
      "coding-analysis    1467  Success                                     08-04 21:31:36\n",
      "coding-analysis    1466  Success                                     08-04 21:27:35\n",
      "coding-analysis    1465  Success                                     08-04 21:30:58\n",
      "coding-analysis    1464  Success                                     08-04 21:32:34\n",
      "coding-analysis    1463  Success                                     08-04 21:18:02\n",
      "coding-analysis    1462  Success                                     08-04 21:26:02\n",
      "coding-analysis    1461  Success                                     08-04 21:25:20\n",
      "coding-analysis    1460  Success                                     08-04 21:27:12\n",
      "coding-analysis    1459  Success                                     08-04 21:31:20\n",
      "coding-analysis    1458  Success                                     08-04 21:29:55\n",
      "coding-analysis    1457  Success                                     08-04 21:26:34\n",
      "coding-analysis    1456  Success                                     08-04 21:30:00\n",
      "coding-analysis    1455  Success                                     08-04 21:33:13\n",
      "coding-analysis    1454  Success                                     08-04 21:26:51\n",
      "coding-analysis    1453  Success                                     08-04 21:29:07\n",
      "coding-analysis    1452  Success                                     08-04 21:25:43\n",
      "coding-analysis    1451  Success                                     08-04 21:30:10\n",
      "coding-analysis    1450  Success                                     08-04 21:26:03\n",
      "coding-analysis    1449  Success                                     08-04 21:29:01\n",
      "coding-analysis    1448  Success                                     08-04 21:26:29\n",
      "coding-analysis    1447  Success                                     08-04 21:26:52\n",
      "coding-analysis    1446  Success                                     08-04 21:27:08\n",
      "coding-analysis    1445  Success                                     08-04 21:26:03\n",
      "coding-analysis    1444  Success                                     08-04 21:26:03\n",
      "coding-analysis    1443  Success                                     08-04 21:26:20\n",
      "coding-analysis    1442  Success                                     08-04 21:28:15\n",
      "coding-analysis    1441  Success                                     08-04 21:29:33\n",
      "coding-analysis    1440  Success                                     08-04 21:27:41\n",
      "coding-analysis    1439  Success                                     08-04 21:34:51\n",
      "coding-analysis    1438  Success                                     08-04 21:29:16\n",
      "coding-analysis    1437  Success                                     08-04 21:27:32\n",
      "coding-analysis    1436  Success                                     08-04 21:28:53\n",
      "coding-analysis    1435  Success                                     08-04 21:30:26\n",
      "coding-analysis    1434  Success                                     08-04 21:15:57\n",
      "coding-analysis    1433  Success                                     08-04 21:35:48\n",
      "coding-analysis    1432  Success                                     08-04 21:28:10\n",
      "coding-analysis    1431  Success                                     08-04 21:32:50\n",
      "coding-analysis    1430  Success                                     08-04 21:31:36\n",
      "coding-analysis    1429  Success                                     08-04 21:26:23\n",
      "coding-analysis    1428  Success                                     08-04 21:28:19\n",
      "coding-analysis    1427  Success                                     08-04 21:27:49\n",
      "coding-analysis    1426  Success                                     08-04 21:27:53\n",
      "coding-analysis    1425  Success                                     08-04 21:27:42\n",
      "coding-analysis    1424  Success                                     08-04 21:27:16\n",
      "coding-analysis    1423  Success                                     08-04 21:31:36\n",
      "coding-analysis    1422  Success                                     08-04 21:26:00\n",
      "coding-analysis    1421  Success                                     08-04 21:29:09\n",
      "coding-analysis    1420  Success                                     08-04 21:25:20\n",
      "coding-analysis    1419  Success                                     08-04 21:31:15\n",
      "coding-analysis    1418  Success                                     08-04 21:26:34\n",
      "coding-analysis    1417  Success                                     08-04 21:19:46\n",
      "coding-analysis    1416  Success                                     08-04 21:29:59\n",
      "coding-analysis    1415  Success                                     08-04 21:27:40\n",
      "coding-analysis    1414  Success                                     08-04 21:26:51\n",
      "coding-analysis    1413  Success                                     08-04 21:26:23\n",
      "coding-analysis    1412  Success                                     08-04 21:26:49\n",
      "coding-analysis    1411  Success                                     08-04 21:27:48\n",
      "coding-analysis    1410  Success                                     08-04 21:27:41\n",
      "coding-analysis    1409  Success                                     08-04 21:28:39\n",
      "coding-analysis    1408  Success                                     08-04 21:28:44\n",
      "coding-analysis    1407  Success                                     08-04 21:25:20\n",
      "coding-analysis    1406  Success                                     08-04 21:27:05\n",
      "coding-analysis    1405  Success                                     08-04 21:28:41\n",
      "coding-analysis    1404  Success                                     08-04 21:26:52\n",
      "coding-analysis    1403  Success                                     08-04 21:29:50\n",
      "coding-analysis    1402  Success                                     08-04 21:30:13\n",
      "coding-analysis    1401  Success                                     08-04 21:27:02\n",
      "coding-analysis    1400  Success                                     08-04 21:28:13\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1399  Success                                     08-04 21:30:15\n",
      "coding-analysis    1398  Success                                     08-04 21:29:36\n",
      "coding-analysis    1397  Success                                     08-04 21:30:43\n",
      "coding-analysis    1396  Success                                     08-04 21:27:46\n",
      "coding-analysis    1395  Success                                     08-04 21:26:55\n",
      "coding-analysis    1394  Success                                     08-04 21:24:44\n",
      "coding-analysis    1393  Success                                     08-04 21:18:13\n",
      "coding-analysis    1392  Success                                     08-04 21:27:29\n",
      "coding-analysis    1391  Success                                     08-04 21:27:28\n",
      "coding-analysis    1390  Success                                     08-04 21:25:38\n",
      "coding-analysis    1389  worker was terminated                       08-04 21:14:39\n",
      "coding-analysis    1388  Success                                     08-04 21:27:32\n",
      "coding-analysis    1387  Success                                     08-04 21:29:45\n",
      "coding-analysis    1386  Success                                     08-04 21:35:48\n",
      "coding-analysis    1385  Success                                     08-04 21:27:48\n",
      "coding-analysis    1384  Success                                     08-04 21:29:19\n",
      "coding-analysis    1383  Success                                     08-04 21:28:39\n",
      "coding-analysis    1382  Success                                     08-04 21:32:21\n",
      "coding-analysis    1381  Success                                     08-04 21:28:04\n",
      "coding-analysis    1380  Success                                     08-04 21:27:53\n",
      "coding-analysis    1379  Success                                     08-04 21:26:02\n",
      "coding-analysis    1378  Success                                     08-04 21:27:29\n",
      "coding-analysis    1377  Success                                     08-04 21:26:52\n",
      "coding-analysis    1376  Success                                     08-04 21:26:02\n",
      "coding-analysis    1375  Success                                     08-04 21:29:30\n",
      "coding-analysis    1374  Success                                     08-04 21:27:44\n",
      "coding-analysis    1373  Success                                     08-04 21:27:46\n",
      "coding-analysis    1372  Success                                     08-04 21:31:41\n",
      "coding-analysis    1371  Success                                     08-04 21:27:41\n",
      "coding-analysis    1370  Success                                     08-04 21:26:02\n",
      "coding-analysis    1369  Success                                     08-04 21:24:44\n",
      "coding-analysis    1368  Success                                     08-04 21:29:26\n",
      "coding-analysis    1367  Success                                     08-04 21:28:39\n",
      "coding-analysis    1366  Success                                     08-04 21:23:07\n",
      "coding-analysis    1365  Success                                     08-04 21:25:36\n",
      "coding-analysis    1364  Success                                     08-04 21:30:09\n",
      "coding-analysis    1363  Success                                     08-04 21:27:10\n",
      "coding-analysis    1362  Success                                     08-04 21:30:12\n",
      "coding-analysis    1361  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis    1360  Success                                     08-04 21:26:32\n",
      "coding-analysis    1359  Success                                     08-04 21:27:36\n",
      "coding-analysis    1358  worker was terminated                       08-04 21:14:39\n",
      "coding-analysis    1357  Success                                     08-04 21:28:04\n",
      "coding-analysis    1356  worker was terminated                       08-04 21:17:53\n",
      "coding-analysis    1355  Success                                     08-04 21:31:14\n",
      "coding-analysis    1354  Success                                     08-04 21:31:43\n",
      "coding-analysis    1353  Success                                     08-04 21:28:12\n",
      "coding-analysis    1352  Success                                     08-04 21:25:03\n",
      "coding-analysis    1351  Success                                     08-04 21:26:00\n",
      "coding-analysis    1350  Success                                     08-04 21:24:45\n",
      "coding-analysis    1349  Success                                     08-04 21:27:02\n",
      "coding-analysis    1348  Success                                     08-04 21:26:49\n",
      "coding-analysis    1347  Success                                     08-04 21:31:50\n",
      "coding-analysis    1346  Success                                     08-04 21:27:38\n",
      "coding-analysis    1345  Success                                     08-04 21:28:05\n",
      "coding-analysis    1344  Success                                     08-04 21:27:01\n",
      "coding-analysis    1343  Success                                     08-04 21:27:15\n",
      "coding-analysis    1342  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis    1341  Success                                     08-04 21:30:28\n",
      "coding-analysis    1340  Success                                     08-04 21:17:05\n",
      "coding-analysis    1339  Success                                     08-04 21:26:56\n",
      "coding-analysis    1338  Success                                     08-04 21:30:39\n",
      "coding-analysis    1337  Success                                     08-04 21:32:30\n",
      "coding-analysis    1336  Success                                     08-04 21:26:52\n",
      "coding-analysis    1335  Success                                     08-04 21:26:29\n",
      "coding-analysis    1334  Success                                     08-04 21:26:52\n",
      "coding-analysis    1333  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis    1332  Success                                     08-04 21:25:36\n",
      "coding-analysis    1331  Success                                     08-04 21:32:21\n",
      "coding-analysis    1330  Success                                     08-04 21:28:12\n",
      "coding-analysis    1329  Success                                     08-04 21:26:03\n",
      "coding-analysis    1328  Success                                     08-04 21:28:21\n",
      "coding-analysis    1327  Success                                     08-04 21:25:42\n",
      "coding-analysis    1326  Success                                     08-04 21:27:08\n",
      "coding-analysis    1325  Success                                     08-04 21:14:28\n",
      "coding-analysis    1324  Success                                     08-04 21:26:22\n",
      "coding-analysis    1323  Success                                     08-04 21:26:53\n",
      "coding-analysis    1322  Success                                     08-04 21:26:02\n",
      "coding-analysis    1321  Success                                     08-04 21:26:23\n",
      "coding-analysis    1320  Success                                     08-04 21:26:52\n",
      "coding-analysis    1319  Success                                     08-04 21:26:52\n",
      "coding-analysis    1318  Success                                     08-04 21:27:02\n",
      "coding-analysis    1317  Success                                     08-04 21:26:52\n",
      "coding-analysis    1316  Success                                     08-04 21:26:59\n",
      "coding-analysis    1315  Success                                     08-04 21:28:39\n",
      "coding-analysis    1314  Success                                     08-04 21:27:01\n",
      "coding-analysis    1313  Success                                     08-04 21:26:32\n",
      "coding-analysis    1312  Success                                     08-04 21:27:23\n",
      "coding-analysis    1311  Success                                     08-04 21:31:56\n",
      "coding-analysis    1310  Success                                     08-04 21:27:57\n",
      "coding-analysis    1309  Success                                     08-04 21:14:05\n",
      "coding-analysis    1308  Success                                     08-04 21:26:31\n",
      "coding-analysis    1307  Success                                     08-04 21:26:52\n",
      "coding-analysis    1306  Success                                     08-04 21:28:03\n",
      "coding-analysis    1305  Success                                     08-04 21:27:36\n",
      "coding-analysis    1304  Success                                     08-04 21:25:35\n",
      "coding-analysis    1303  Success                                     08-04 21:32:21\n",
      "coding-analysis    1302  Success                                     08-04 21:26:57\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1301  Success                                     08-04 21:26:23\n",
      "coding-analysis    1300  Success                                     08-04 21:30:38\n",
      "coding-analysis    1299  Success                                     08-04 21:26:30\n",
      "coding-analysis    1298  Success                                     08-04 21:27:03\n",
      "coding-analysis    1297  Success                                     08-04 21:31:38\n",
      "coding-analysis    1296  Success                                     08-04 21:24:45\n",
      "coding-analysis    1295  Success                                     08-04 21:21:08\n",
      "coding-analysis    1294  Success                                     08-04 21:26:34\n",
      "coding-analysis    1293  Success                                     08-04 21:30:22\n",
      "coding-analysis    1292  Success                                     08-04 21:27:13\n",
      "coding-analysis    1291  Success                                     08-04 21:27:20\n",
      "coding-analysis    1290  Success                                     08-04 21:27:49\n",
      "coding-analysis    1289  Success                                     08-04 21:28:12\n",
      "coding-analysis    1288  Success                                     08-04 21:31:38\n",
      "coding-analysis    1287  Success                                     08-04 21:29:34\n",
      "coding-analysis    1286  Success                                     08-04 21:28:10\n",
      "coding-analysis    1285  Success                                     08-04 21:27:07\n",
      "coding-analysis    1284  Success                                     08-04 21:30:33\n",
      "coding-analysis    1283  Success                                     08-04 21:31:58\n",
      "coding-analysis    1282  Success                                     08-04 21:26:51\n",
      "coding-analysis    1281  Success                                     08-04 21:27:35\n",
      "coding-analysis    1280  Success                                     08-04 21:26:54\n",
      "coding-analysis    1279  Success                                     08-04 21:27:36\n",
      "coding-analysis    1278  Success                                     08-04 21:28:17\n",
      "coding-analysis    1277  Success                                     08-04 21:16:42\n",
      "coding-analysis    1276  Success                                     08-04 21:26:52\n",
      "coding-analysis    1275  Success                                     08-04 21:37:13\n",
      "coding-analysis    1274  Success                                     08-04 21:26:20\n",
      "coding-analysis    1273  Success                                     08-04 21:30:25\n",
      "coding-analysis    1272  Success                                     08-04 21:27:06\n",
      "coding-analysis    1271  Success                                     08-04 21:26:54\n",
      "coding-analysis    1270  Success                                     08-04 21:25:36\n",
      "coding-analysis    1269  Success                                     08-04 21:30:07\n",
      "coding-analysis    1268  Success                                     08-04 21:26:58\n",
      "coding-analysis    1267  Success                                     08-04 21:26:00\n",
      "coding-analysis    1266  Success                                     08-04 21:30:35\n",
      "coding-analysis    1265  Success                                     08-04 21:27:36\n",
      "coding-analysis    1264  Success                                     08-04 21:25:44\n",
      "coding-analysis    1263  Success                                     08-04 21:26:34\n",
      "coding-analysis    1262  Success                                     08-04 21:30:05\n",
      "coding-analysis    1261  Success                                     08-04 21:26:31\n",
      "coding-analysis    1260  Success                                     08-04 21:30:48\n",
      "coding-analysis    1259  Success                                     08-04 21:25:36\n",
      "coding-analysis    1258  Success                                     08-04 21:26:55\n",
      "coding-analysis    1257  Success                                     08-04 21:25:39\n",
      "coding-analysis    1256  Success                                     08-04 21:15:35\n",
      "coding-analysis    1255  Success                                     08-04 21:26:55\n",
      "coding-analysis    1254  Success                                     08-04 21:26:20\n",
      "coding-analysis    1253  Success                                     08-04 21:26:30\n",
      "coding-analysis    1252  Success                                     08-04 21:26:50\n",
      "coding-analysis    1251  Success                                     08-04 21:30:06\n",
      "coding-analysis    1250  Success                                     08-04 22:03:45\n",
      "coding-analysis    1249  Success                                     08-04 21:45:38\n",
      "coding-analysis    1248  Success                                     08-04 21:25:36\n",
      "coding-analysis    1247  Success                                     08-04 21:26:23\n",
      "coding-analysis    1246  Success                                     08-04 21:31:18\n",
      "coding-analysis    1245  Success                                     08-04 21:27:46\n",
      "coding-analysis    1244  Success                                     08-04 21:27:49\n",
      "coding-analysis    1243  Success                                     08-04 21:25:35\n",
      "coding-analysis    1242  Success                                     08-04 21:26:22\n",
      "coding-analysis    1241  Success                                     08-04 21:27:39\n",
      "coding-analysis    1240  Success                                     08-04 21:25:38\n",
      "coding-analysis    1239  Success                                     08-04 21:28:52\n",
      "coding-analysis    1238  Success                                     08-04 21:24:44\n",
      "coding-analysis    1237  Success                                     08-04 21:26:03\n",
      "coding-analysis    1236  Success                                     08-04 21:27:44\n",
      "coding-analysis    1235  Success                                     08-04 21:26:03\n",
      "coding-analysis    1234  Success                                     08-04 21:31:57\n",
      "coding-analysis    1233  Success                                     08-04 21:26:22\n",
      "coding-analysis    1232  Success                                     08-04 21:27:00\n",
      "coding-analysis    1231  Success                                     08-04 21:27:34\n",
      "coding-analysis    1230  Success                                     08-04 21:26:03\n",
      "coding-analysis    1229  Success                                     08-04 21:25:36\n",
      "coding-analysis    1228  Success                                     08-04 21:26:49\n",
      "coding-analysis    1227  Success                                     08-04 21:31:14\n",
      "coding-analysis    1226  Success                                     08-04 21:24:13\n",
      "coding-analysis    1225  Success                                     08-04 21:31:36\n",
      "coding-analysis    1224  Success                                     08-04 21:30:02\n",
      "coding-analysis    1223  Success                                     08-04 21:28:04\n",
      "coding-analysis    1222  Success                                     08-04 21:26:23\n",
      "coding-analysis    1221  Success                                     08-04 21:27:41\n",
      "coding-analysis    1220  Success                                     08-04 21:29:45\n",
      "coding-analysis    1219  Success                                     08-04 21:28:14\n",
      "coding-analysis    1218  Success                                     08-04 21:27:27\n",
      "coding-analysis    1217  Success                                     08-04 21:27:59\n",
      "coding-analysis    1216  Success                                     08-04 21:35:31\n",
      "coding-analysis    1215  Success                                     08-04 21:32:38\n",
      "coding-analysis    1214  Success                                     08-04 21:26:56\n",
      "coding-analysis    1213  Success                                     08-04 21:31:14\n",
      "coding-analysis    1212  Success                                     08-04 21:27:25\n",
      "coding-analysis    1211  Success                                     08-04 21:29:24\n",
      "coding-analysis    1210  Success                                     08-04 21:28:39\n",
      "coding-analysis    1209  Success                                     08-04 21:31:00\n",
      "coding-analysis    1208  Success                                     08-04 21:27:35\n",
      "coding-analysis    1207  Success                                     08-04 21:27:21\n",
      "coding-analysis    1206  Success                                     08-04 21:33:34\n",
      "coding-analysis    1205  Success                                     08-04 21:31:36\n",
      "coding-analysis    1204  Success                                     08-04 21:28:39\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1203  Success                                     08-04 21:29:42\n",
      "coding-analysis    1202  Success                                     08-04 21:19:39\n",
      "coding-analysis    1201  Success                                     08-04 21:28:39\n",
      "coding-analysis    1200  Success                                     08-04 21:30:52\n",
      "coding-analysis    1199  Success                                     08-04 21:31:14\n",
      "coding-analysis    1198  Success                                     08-04 21:26:51\n",
      "coding-analysis    1197  Success                                     08-04 21:29:08\n",
      "coding-analysis    1196  Success                                     08-04 21:27:15\n",
      "coding-analysis    1195  Success                                     08-04 21:29:32\n",
      "coding-analysis    1194  Success                                     08-04 21:27:36\n",
      "coding-analysis    1193  Success                                     08-04 21:30:08\n",
      "coding-analysis    1192  Success                                     08-04 21:29:08\n",
      "coding-analysis    1191  Success                                     08-04 21:29:19\n",
      "coding-analysis    1190  Success                                     08-04 21:27:59\n",
      "coding-analysis    1189  Success                                     08-04 21:27:03\n",
      "coding-analysis    1188  Success                                     08-04 21:27:36\n",
      "coding-analysis    1187  Success                                     08-04 21:27:34\n",
      "coding-analysis    1186  Success                                     08-04 21:27:36\n",
      "coding-analysis    1185  Success                                     08-04 21:32:21\n",
      "coding-analysis    1184  Success                                     08-04 21:26:22\n",
      "coding-analysis    1183  Success                                     08-04 21:29:36\n",
      "coding-analysis    1182  Success                                     08-04 21:27:41\n",
      "coding-analysis    1181  Success                                     08-04 21:29:54\n",
      "coding-analysis    1180  Success                                     08-04 21:29:29\n",
      "coding-analysis    1179  Success                                     08-04 21:28:56\n",
      "coding-analysis    1178  Success                                     08-04 21:30:46\n",
      "coding-analysis    1177  Success                                     08-04 21:27:31\n",
      "coding-analysis    1176  Success                                     08-04 21:27:57\n",
      "coding-analysis    1175  Success                                     08-04 21:28:50\n",
      "coding-analysis    1174  Success                                     08-04 21:29:32\n",
      "coding-analysis    1173  Success                                     08-04 21:28:15\n",
      "coding-analysis    1172  Success                                     08-04 21:33:18\n",
      "coding-analysis    1171  Success                                     08-04 21:28:05\n",
      "coding-analysis    1170  Success                                     08-04 21:27:36\n",
      "coding-analysis    1169  Success                                     08-04 21:30:08\n",
      "coding-analysis    1168  Success                                     08-04 21:31:55\n",
      "coding-analysis    1167  Success                                     08-04 21:28:39\n",
      "coding-analysis    1166  Success                                     08-04 21:28:12\n",
      "coding-analysis    1165  Success                                     08-04 21:30:13\n",
      "coding-analysis    1164  Success                                     08-04 21:29:51\n",
      "coding-analysis    1163  Success                                     08-04 21:27:26\n",
      "coding-analysis    1162  Success                                     08-04 21:31:50\n",
      "coding-analysis    1161  Success                                     08-04 21:33:53\n",
      "coding-analysis    1160  Success                                     08-04 21:27:46\n",
      "coding-analysis    1159  Success                                     08-04 21:28:58\n",
      "coding-analysis    1158  Success                                     08-04 21:33:04\n",
      "coding-analysis    1157  Success                                     08-04 21:32:36\n",
      "coding-analysis    1156  Success                                     08-04 21:31:39\n",
      "coding-analysis    1155  Success                                     08-04 21:27:28\n",
      "coding-analysis    1154  Success                                     08-04 21:28:39\n",
      "coding-analysis    1153  Success                                     08-04 21:27:17\n",
      "coding-analysis    1152  Success                                     08-04 21:31:21\n",
      "coding-analysis    1151  Success                                     08-04 21:27:26\n",
      "coding-analysis    1150  Success                                     08-04 21:28:54\n",
      "coding-analysis    1149  Success                                     08-04 21:27:41\n",
      "coding-analysis    1148  Success                                     08-04 21:31:17\n",
      "coding-analysis    1147  Success                                     08-04 21:28:19\n",
      "coding-analysis    1146  Success                                     08-04 21:29:47\n",
      "coding-analysis    1145  Success                                     08-04 21:31:36\n",
      "coding-analysis    1144  Success                                     08-04 21:27:40\n",
      "coding-analysis    1143  Success                                     08-04 21:28:13\n",
      "coding-analysis    1142  Success                                     08-04 21:30:48\n",
      "coding-analysis    1141  Success                                     08-04 21:27:21\n",
      "coding-analysis    1140  Success                                     08-04 21:27:54\n",
      "coding-analysis    1139  Success                                     08-04 21:32:21\n",
      "coding-analysis    1138  Success                                     08-04 21:31:37\n",
      "coding-analysis    1137  Success                                     08-04 21:21:48\n",
      "coding-analysis    1136  Success                                     08-04 21:26:26\n",
      "coding-analysis    1135  Success                                     08-04 21:25:35\n",
      "coding-analysis    1134  Success                                     08-04 21:27:52\n",
      "coding-analysis    1133  Success                                     08-04 21:29:30\n",
      "coding-analysis    1132  Success                                     08-04 21:26:03\n",
      "coding-analysis    1131  Success                                     08-04 21:26:22\n",
      "coding-analysis    1130  Success                                     08-04 21:27:03\n",
      "coding-analysis    1129  Success                                     08-04 21:26:26\n",
      "coding-analysis    1128  Success                                     08-04 21:21:47\n",
      "coding-analysis    1127  Success                                     08-04 21:27:34\n",
      "coding-analysis    1126  Success                                     08-04 21:25:20\n",
      "coding-analysis    1125  Success                                     08-04 21:25:18\n",
      "coding-analysis    1124  Success                                     08-04 21:21:47\n",
      "coding-analysis    1123  Success                                     08-04 21:19:55\n",
      "coding-analysis    1122  Success                                     08-04 21:32:24\n",
      "coding-analysis    1121  Success                                     08-04 21:24:44\n",
      "coding-analysis    1120  Success                                     08-04 21:25:20\n",
      "coding-analysis    1119  Success                                     08-04 21:25:43\n",
      "coding-analysis    1118  Success                                     08-04 21:25:20\n",
      "coding-analysis    1117  Success                                     08-04 21:26:20\n",
      "coding-analysis    1116  Success                                     08-04 21:26:22\n",
      "coding-analysis    1115  Success                                     08-04 21:25:02\n",
      "coding-analysis    1114  Success                                     08-04 21:24:11\n",
      "coding-analysis    1113  Success                                     08-04 21:27:04\n",
      "coding-analysis    1112  Success                                     08-04 21:24:27\n",
      "coding-analysis    1111  Success                                     08-04 21:28:10\n",
      "coding-analysis    1110  Success                                     08-04 21:27:32\n",
      "coding-analysis    1109  Success                                     08-04 21:26:30\n",
      "coding-analysis    1108  Success                                     08-04 21:31:36\n",
      "coding-analysis    1107  Success                                     08-04 21:28:12\n",
      "coding-analysis    1106  Success                                     08-04 21:26:32\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1105  Success                                     08-04 21:27:44\n",
      "coding-analysis    1104  Success                                     08-04 21:27:21\n",
      "coding-analysis    1103  Success                                     08-04 21:28:21\n",
      "coding-analysis    1102  Success                                     08-04 21:26:51\n",
      "coding-analysis    1101  Success                                     08-04 21:28:07\n",
      "coding-analysis    1100  Success                                     08-04 21:27:31\n",
      "coding-analysis    1099  Success                                     08-04 21:29:32\n",
      "coding-analysis    1098  Success                                     08-04 21:30:35\n",
      "coding-analysis    1097  Success                                     08-04 21:27:45\n",
      "coding-analysis    1096  Success                                     08-04 21:32:29\n",
      "coding-analysis    1095  Success                                     08-04 21:27:37\n",
      "coding-analysis    1094  Success                                     08-04 21:31:36\n",
      "coding-analysis    1093  Success                                     08-04 21:31:52\n",
      "coding-analysis    1092  Success                                     08-04 21:31:14\n",
      "coding-analysis    1091  Success                                     08-04 21:27:45\n",
      "coding-analysis    1090  Success                                     08-04 21:34:15\n",
      "coding-analysis    1089  Success                                     08-04 21:27:18\n",
      "coding-analysis    1088  Success                                     08-04 21:28:12\n",
      "coding-analysis    1087  Success                                     08-04 21:26:54\n",
      "coding-analysis    1086  Success                                     08-04 21:33:30\n",
      "coding-analysis    1085  Success                                     08-04 21:29:05\n",
      "coding-analysis    1084  Success                                     08-04 21:16:42\n",
      "coding-analysis    1083  Success                                     08-04 21:25:41\n",
      "coding-analysis    1082  Success                                     08-04 21:25:36\n",
      "coding-analysis    1081  Success                                     08-04 21:26:34\n",
      "coding-analysis    1080  Success                                     08-04 21:26:51\n",
      "coding-analysis    1079  Success                                     08-04 21:26:02\n",
      "coding-analysis    1078  Success                                     08-04 21:27:24\n",
      "coding-analysis    1077  Success                                     08-04 21:29:37\n",
      "coding-analysis    1076  Success                                     08-04 21:29:55\n",
      "coding-analysis    1075  Success                                     08-04 21:29:29\n",
      "coding-analysis    1074  Success                                     08-04 21:32:22\n",
      "coding-analysis    1073  Success                                     08-04 21:30:08\n",
      "coding-analysis    1072  Success                                     08-04 21:30:01\n",
      "coding-analysis    1071  Success                                     08-04 21:29:04\n",
      "coding-analysis    1070  Success                                     08-04 21:27:23\n",
      "coding-analysis    1069  Success                                     08-04 21:26:55\n",
      "coding-analysis    1068  Success                                     08-04 21:27:16\n",
      "coding-analysis    1067  Success                                     08-04 21:24:29\n",
      "coding-analysis    1066  Success                                     08-04 21:27:35\n",
      "coding-analysis    1065  Success                                     08-04 21:24:13\n",
      "coding-analysis    1064  Success                                     08-04 21:26:26\n",
      "coding-analysis    1063  Success                                     08-04 21:27:20\n",
      "coding-analysis    1062  Success                                     08-04 21:26:30\n",
      "coding-analysis    1061  Success                                     08-04 21:26:59\n",
      "coding-analysis    1060  Success                                     08-04 21:33:37\n",
      "coding-analysis    1059  Success                                     08-04 21:28:43\n",
      "coding-analysis    1058  Success                                     08-04 21:26:20\n",
      "coding-analysis    1057  Success                                     08-04 21:27:55\n",
      "coding-analysis    1056  Success                                     08-04 21:27:02\n",
      "coding-analysis    1055  Success                                     08-04 21:15:56\n",
      "coding-analysis    1054  Success                                     08-04 21:29:41\n",
      "coding-analysis    1053  Success                                     08-04 21:27:29\n",
      "coding-analysis    1052  Success                                     08-04 21:29:53\n",
      "coding-analysis    1051  Success                                     08-04 21:27:36\n",
      "coding-analysis    1050  Success                                     08-04 21:28:45\n",
      "coding-analysis    1049  Success                                     08-04 21:26:29\n",
      "coding-analysis    1048  Success                                     08-04 21:31:36\n",
      "coding-analysis    1047  Success                                     08-04 21:29:18\n",
      "coding-analysis    1046  Success                                     08-04 21:24:45\n",
      "coding-analysis    1045  Success                                     08-04 21:26:27\n",
      "coding-analysis    1044  Success                                     08-04 21:27:39\n",
      "coding-analysis    1043  Success                                     08-04 21:27:13\n",
      "coding-analysis    1042  Success                                     08-04 21:28:55\n",
      "coding-analysis    1041  Success                                     08-04 21:27:37\n",
      "coding-analysis    1040  Success                                     08-04 21:29:30\n",
      "coding-analysis    1039  Success                                     08-04 21:27:51\n",
      "coding-analysis    1038  Success                                     08-04 21:19:25\n",
      "coding-analysis    1037  Success                                     08-04 21:27:40\n",
      "coding-analysis    1036  Success                                     08-04 21:26:49\n",
      "coding-analysis    1035  Success                                     08-04 21:25:41\n",
      "coding-analysis    1034  Success                                     08-04 21:26:52\n",
      "coding-analysis    1033  Success                                     08-04 21:26:00\n",
      "coding-analysis    1032  Success                                     08-04 21:28:22\n",
      "coding-analysis    1031  Success                                     08-04 21:31:52\n",
      "coding-analysis    1030  Success                                     08-04 21:27:07\n",
      "coding-analysis    1029  Success                                     08-04 21:33:41\n",
      "coding-analysis    1028  Success                                     08-04 21:26:00\n",
      "coding-analysis    1027  Success                                     08-04 21:28:06\n",
      "coding-analysis    1026  Success                                     08-04 21:31:45\n",
      "coding-analysis    1025  Success                                     08-04 21:26:55\n",
      "coding-analysis    1024  Success                                     08-04 21:27:05\n",
      "coding-analysis    1023  Success                                     08-04 21:29:28\n",
      "coding-analysis    1022  Success                                     08-04 21:26:57\n",
      "coding-analysis    1021  Success                                     08-04 21:26:56\n",
      "coding-analysis    1020  Success                                     08-04 21:27:38\n",
      "coding-analysis    1019  Success                                     08-04 21:25:35\n",
      "coding-analysis    1018  Success                                     08-04 21:30:03\n",
      "coding-analysis    1017  Success                                     08-04 21:25:35\n",
      "coding-analysis    1016  Success                                     08-04 21:26:27\n",
      "coding-analysis    1015  Success                                     08-04 21:26:03\n",
      "coding-analysis    1014  Success                                     08-04 21:17:05\n",
      "coding-analysis    1013  Success                                     08-04 21:31:18\n",
      "coding-analysis    1012  Success                                     08-04 21:26:26\n",
      "coding-analysis    1011  Success                                     08-04 21:24:13\n",
      "coding-analysis    1010  Success                                     08-04 21:26:27\n",
      "coding-analysis    1009  Success                                     08-04 21:27:23\n",
      "coding-analysis    1008  Success                                     08-04 21:25:20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis    1007  Success                                     08-04 21:31:14\n",
      "coding-analysis    1006  Success                                     08-04 21:29:59\n",
      "coding-analysis    1005  Success                                     08-04 21:26:49\n",
      "coding-analysis    1004  Success                                     08-04 21:26:55\n",
      "coding-analysis    1003  Success                                     08-04 21:27:55\n",
      "coding-analysis    1002  Success                                     08-04 21:28:23\n",
      "coding-analysis    1001  Success                                     08-04 21:27:55\n",
      "coding-analysis    1000  Success                                     08-04 21:28:39\n",
      "coding-analysis     999  Success                                     08-04 21:27:12\n",
      "coding-analysis     998  Success                                     08-04 21:26:23\n",
      "coding-analysis     997  Success                                     08-04 21:27:38\n",
      "coding-analysis     996  Success                                     08-04 21:28:48\n",
      "coding-analysis     995  Success                                     08-04 21:28:57\n",
      "coding-analysis     994  Success                                     08-04 21:30:25\n",
      "coding-analysis     993  Success                                     08-04 21:26:54\n",
      "coding-analysis     992  Success                                     08-04 21:30:11\n",
      "coding-analysis     991  Success                                     08-04 21:26:02\n",
      "coding-analysis     990  Success                                     08-04 21:23:55\n",
      "coding-analysis     989  Success                                     08-04 21:24:27\n",
      "coding-analysis     988  Success                                     08-04 21:27:54\n",
      "coding-analysis     987  Success                                     08-04 21:22:19\n",
      "coding-analysis     986  Success                                     08-04 21:25:18\n",
      "coding-analysis     985  Success                                     08-04 21:29:41\n",
      "coding-analysis     984  Success                                     08-04 21:25:20\n",
      "coding-analysis     983  Success                                     08-04 21:29:20\n",
      "coding-analysis     982  Success                                     08-04 21:27:21\n",
      "coding-analysis     981  Success                                     08-04 21:26:30\n",
      "coding-analysis     980  Success                                     08-04 21:28:06\n",
      "coding-analysis     979  Success                                     08-04 21:26:49\n",
      "coding-analysis     978  Success                                     08-04 21:24:44\n",
      "coding-analysis     977  Success                                     08-04 21:25:38\n",
      "coding-analysis     976  Success                                     08-04 21:27:27\n",
      "coding-analysis     975  Success                                     08-04 21:27:24\n",
      "coding-analysis     974  Success                                     08-04 21:26:54\n",
      "coding-analysis     973  Success                                     08-04 21:28:01\n",
      "coding-analysis     972  Success                                     08-04 21:26:51\n",
      "coding-analysis     971  Success                                     08-04 21:26:51\n",
      "coding-analysis     970  Success                                     08-04 21:30:33\n",
      "coding-analysis     969  Success                                     08-04 21:29:33\n",
      "coding-analysis     968  Success                                     08-04 21:31:14\n",
      "coding-analysis     967  Success                                     08-04 21:27:13\n",
      "coding-analysis     966  Success                                     08-04 21:27:37\n",
      "coding-analysis     965  Success                                     08-04 21:28:06\n",
      "coding-analysis     964  Success                                     08-04 21:25:18\n",
      "coding-analysis     963  Success                                     08-04 21:26:05\n",
      "coding-analysis     962  Success                                     08-04 21:30:26\n",
      "coding-analysis     961  Success                                     08-04 21:16:42\n",
      "coding-analysis     960  Success                                     08-04 21:26:26\n",
      "coding-analysis     959  Success                                     08-04 21:26:31\n",
      "coding-analysis     958  Success                                     08-04 21:26:25\n",
      "coding-analysis     957  Success                                     08-04 21:24:29\n",
      "coding-analysis     956  Success                                     08-04 21:25:41\n",
      "coding-analysis     955  Success                                     08-04 21:30:59\n",
      "coding-analysis     954  Success                                     08-04 21:24:45\n",
      "coding-analysis     953  Success                                     08-04 21:30:12\n",
      "coding-analysis     952  Success                                     08-04 21:27:31\n",
      "coding-analysis     951  Success                                     08-04 21:26:23\n",
      "coding-analysis     950  Success                                     08-04 21:25:44\n",
      "coding-analysis     949  Success                                     08-04 21:27:07\n",
      "coding-analysis     948  Success                                     08-04 21:26:49\n",
      "coding-analysis     947  Success                                     08-04 21:26:00\n",
      "coding-analysis     946  Success                                     08-04 21:14:16\n",
      "coding-analysis     945  Success                                     08-04 21:25:36\n",
      "coding-analysis     944  Success                                     08-04 21:25:36\n",
      "coding-analysis     943  Success                                     08-04 21:25:43\n",
      "coding-analysis     942  Success                                     08-04 21:29:29\n",
      "coding-analysis     941  Success                                     08-04 21:26:49\n",
      "coding-analysis     940  Success                                     08-04 21:25:35\n",
      "coding-analysis     939  Success                                     08-04 21:27:05\n",
      "coding-analysis     938  Success                                     08-04 21:26:00\n",
      "coding-analysis     937  Success                                     08-04 21:26:31\n",
      "coding-analysis     936  Success                                     08-04 21:28:02\n",
      "coding-analysis     935  Success                                     08-04 21:26:29\n",
      "coding-analysis     934  Success                                     08-04 21:26:54\n",
      "coding-analysis     933  Success                                     08-04 21:27:57\n",
      "coding-analysis     932  Success                                     08-04 21:27:43\n",
      "coding-analysis     931  Success                                     08-04 21:26:55\n",
      "coding-analysis     930  Success                                     08-04 21:13:55\n",
      "coding-analysis     929  Success                                     08-04 21:30:04\n",
      "coding-analysis     928  Success                                     08-04 21:25:38\n",
      "coding-analysis     927  Success                                     08-04 21:27:22\n",
      "coding-analysis     926  Success                                     08-04 21:30:04\n",
      "coding-analysis     925  Success                                     08-04 21:25:36\n",
      "coding-analysis     924  Success                                     08-04 21:31:37\n",
      "coding-analysis     923  Success                                     08-04 21:30:33\n",
      "coding-analysis     922  Success                                     08-04 21:26:49\n",
      "coding-analysis     921  Success                                     08-04 21:28:22\n",
      "coding-analysis     920  Success                                     08-04 21:26:20\n",
      "coding-analysis     919  Success                                     08-04 21:25:36\n",
      "coding-analysis     918  Success                                     08-04 21:29:44\n",
      "coding-analysis     917  Success                                     08-04 21:24:44\n",
      "coding-analysis     916  Success                                     08-04 21:20:02\n",
      "coding-analysis     915  Success                                     08-04 21:26:03\n",
      "coding-analysis     914  Success                                     08-04 21:25:02\n",
      "coding-analysis     913  Success                                     08-04 21:29:48\n",
      "coding-analysis     912  Success                                     08-04 21:25:43\n",
      "coding-analysis     911  Success                                     08-04 21:28:17\n",
      "coding-analysis     910  Success                                     08-04 21:31:39\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     909  Success                                     08-04 21:27:17\n",
      "coding-analysis     908  Success                                     08-04 21:27:10\n",
      "coding-analysis     907  Success                                     08-04 21:27:58\n",
      "coding-analysis     906  Success                                     08-04 21:26:52\n",
      "coding-analysis     905  Success                                     08-04 21:24:44\n",
      "coding-analysis     904  Success                                     08-04 21:31:36\n",
      "coding-analysis     903  Success                                     08-04 21:29:18\n",
      "coding-analysis     902  Success                                     08-04 21:27:17\n",
      "coding-analysis     901  Success                                     08-04 21:26:52\n",
      "coding-analysis     900  Success                                     08-04 21:26:49\n",
      "coding-analysis     899  Success                                     08-04 21:30:36\n",
      "coding-analysis     898  Success                                     08-04 21:17:11\n",
      "coding-analysis     897  Success                                     08-04 21:26:57\n",
      "coding-analysis     896  Success                                     08-04 21:30:45\n",
      "coding-analysis     895  Success                                     08-04 21:29:48\n",
      "coding-analysis     894  Success                                     08-04 21:25:21\n",
      "coding-analysis     893  Success                                     08-04 21:27:05\n",
      "coding-analysis     892  Success                                     08-04 21:25:36\n",
      "coding-analysis     891  Success                                     08-04 21:25:41\n",
      "coding-analysis     890  Success                                     08-04 21:26:26\n",
      "coding-analysis     889  Success                                     08-04 21:29:53\n",
      "coding-analysis     888  Success                                     08-04 21:25:38\n",
      "coding-analysis     887  Success                                     08-04 21:26:22\n",
      "coding-analysis     886  Success                                     08-04 21:30:06\n",
      "coding-analysis     885  Success                                     08-04 21:25:03\n",
      "coding-analysis     884  Success                                     08-04 21:26:23\n",
      "coding-analysis     883  Success                                     08-04 21:28:56\n",
      "coding-analysis     882  Success                                     08-04 21:29:48\n",
      "coding-analysis     881  Success                                     08-04 21:26:31\n",
      "coding-analysis     880  Success                                     08-04 21:23:40\n",
      "coding-analysis     879  Success                                     08-04 21:26:56\n",
      "coding-analysis     878  Success                                     08-04 21:24:27\n",
      "coding-analysis     877  Success                                     08-04 21:12:59\n",
      "coding-analysis     876  Success                                     08-04 21:29:17\n",
      "coding-analysis     875  Success                                     08-04 21:26:00\n",
      "coding-analysis     874  Success                                     08-04 21:25:45\n",
      "coding-analysis     873  Success                                     08-04 21:26:56\n",
      "coding-analysis     872  Success                                     08-04 21:28:51\n",
      "coding-analysis     871  Success                                     08-04 22:00:45\n",
      "coding-analysis     870  Success                                     08-04 21:44:44\n",
      "coding-analysis     869  Success                                     08-04 21:25:02\n",
      "coding-analysis     868  Success                                     08-04 21:26:51\n",
      "coding-analysis     867  Success                                     08-04 21:26:49\n",
      "coding-analysis     866  Success                                     08-04 21:26:22\n",
      "coding-analysis     865  Success                                     08-04 21:27:35\n",
      "coding-analysis     864  Success                                     08-04 21:28:02\n",
      "coding-analysis     863  Success                                     08-04 21:28:50\n",
      "coding-analysis     862  Success                                     08-04 21:27:34\n",
      "coding-analysis     861  Success                                     08-04 21:28:39\n",
      "coding-analysis     860  Success                                     08-04 21:26:02\n",
      "coding-analysis     859  Success                                     08-04 21:23:55\n",
      "coding-analysis     858  Success                                     08-04 21:26:22\n",
      "coding-analysis     857  Success                                     08-04 21:27:08\n",
      "coding-analysis     856  Success                                     08-04 21:25:18\n",
      "coding-analysis     855  Success                                     08-04 21:31:14\n",
      "coding-analysis     854  Success                                     08-04 21:25:35\n",
      "coding-analysis     853  Success                                     08-04 21:26:49\n",
      "coding-analysis     852  Success                                     08-04 21:26:51\n",
      "coding-analysis     851  Success                                     08-04 21:27:03\n",
      "coding-analysis     850  Success                                     08-04 21:25:18\n",
      "coding-analysis     849  Success                                     08-04 21:26:22\n",
      "coding-analysis     848  Success                                     08-04 21:25:35\n",
      "coding-analysis     847  Success                                     08-04 21:26:52\n",
      "coding-analysis     846  Success                                     08-04 21:26:52\n",
      "coding-analysis     845  Success                                     08-04 21:29:16\n",
      "coding-analysis     844  Success                                     08-04 21:28:39\n",
      "coding-analysis     843  Success                                     08-04 21:26:49\n",
      "coding-analysis     842  Success                                     08-04 21:26:26\n",
      "coding-analysis     841  Success                                     08-04 21:29:33\n",
      "coding-analysis     840  Success                                     08-04 21:25:44\n",
      "coding-analysis     839  Success                                     08-04 21:29:34\n",
      "coding-analysis     838  Success                                     08-04 21:31:36\n",
      "coding-analysis     837  Success                                     08-04 21:26:23\n",
      "coding-analysis     836  Success                                     08-04 21:27:24\n",
      "coding-analysis     835  Success                                     08-04 21:26:20\n",
      "coding-analysis     834  Success                                     08-04 21:27:13\n",
      "coding-analysis     833  Success                                     08-04 21:27:49\n",
      "coding-analysis     832  Success                                     08-04 21:29:24\n",
      "coding-analysis     831  Success                                     08-04 21:31:41\n",
      "coding-analysis     830  Success                                     08-04 21:27:34\n",
      "coding-analysis     829  Success                                     08-04 21:28:39\n",
      "coding-analysis     828  Success                                     08-04 21:27:30\n",
      "coding-analysis     827  Success                                     08-04 21:28:42\n",
      "coding-analysis     826  Success                                     08-04 21:27:13\n",
      "coding-analysis     825  Success                                     08-04 21:28:01\n",
      "coding-analysis     824  Success                                     08-04 21:27:56\n",
      "coding-analysis     823  Success                                     08-04 21:18:51\n",
      "coding-analysis     822  Success                                     08-04 21:28:21\n",
      "coding-analysis     821  Success                                     08-04 21:32:51\n",
      "coding-analysis     820  Success                                     08-04 21:31:14\n",
      "coding-analysis     819  Success                                     08-04 21:26:20\n",
      "coding-analysis     818  Success                                     08-04 21:27:29\n",
      "coding-analysis     817  Success                                     08-04 21:31:36\n",
      "coding-analysis     816  Success                                     08-04 21:28:18\n",
      "coding-analysis     815  Success                                     08-04 21:26:52\n",
      "coding-analysis     814  Success                                     08-04 21:30:40\n",
      "coding-analysis     813  Success                                     08-04 21:28:03\n",
      "coding-analysis     812  Success                                     08-04 21:29:06\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     811  Success                                     08-04 21:27:13\n",
      "coding-analysis     810  Success                                     08-04 21:26:25\n",
      "coding-analysis     809  Success                                     08-04 21:26:52\n",
      "coding-analysis     808  Success                                     08-04 21:27:16\n",
      "coding-analysis     807  Success                                     08-04 21:26:26\n",
      "coding-analysis     806  Success                                     08-04 21:28:52\n",
      "coding-analysis     805  Success                                     08-04 21:25:35\n",
      "coding-analysis     804  Success                                     08-04 21:29:12\n",
      "coding-analysis     803  Success                                     08-04 21:29:20\n",
      "coding-analysis     802  Success                                     08-04 21:32:37\n",
      "coding-analysis     801  Success                                     08-04 21:28:11\n",
      "coding-analysis     800  Success                                     08-04 21:27:44\n",
      "coding-analysis     799  Success                                     08-04 21:27:20\n",
      "coding-analysis     798  Success                                     08-04 21:27:13\n",
      "coding-analysis     797  Success                                     08-04 21:29:01\n",
      "coding-analysis     796  Success                                     08-04 21:27:59\n",
      "coding-analysis     795  Success                                     08-04 21:28:42\n",
      "coding-analysis     794  Success                                     08-04 21:27:34\n",
      "coding-analysis     793  Success                                     08-04 21:28:39\n",
      "coding-analysis     792  Success                                     08-04 21:27:01\n",
      "coding-analysis     791  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis     790  Success                                     08-04 21:28:43\n",
      "coding-analysis     789  Success                                     08-04 21:27:56\n",
      "coding-analysis     788  worker was terminated                       08-04 21:14:39\n",
      "coding-analysis     787  Success                                     08-04 21:28:21\n",
      "coding-analysis     786  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis     785  Success                                     08-04 21:27:10\n",
      "coding-analysis     784  Success                                     08-04 21:27:15\n",
      "coding-analysis     783  Success                                     08-04 21:27:55\n",
      "coding-analysis     782  Success                                     08-04 21:29:22\n",
      "coding-analysis     781  Success                                     08-04 21:31:14\n",
      "coding-analysis     780  Success                                     08-04 21:28:39\n",
      "coding-analysis     779  Success                                     08-04 21:28:40\n",
      "coding-analysis     778  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis     777  Success                                     08-04 21:28:08\n",
      "coding-analysis     776  Success                                     08-04 21:27:12\n",
      "coding-analysis     775  Success                                     08-04 21:31:37\n",
      "coding-analysis     774  Success                                     08-04 21:26:49\n",
      "coding-analysis     773  Success                                     08-04 21:28:24\n",
      "coding-analysis     772  Success                                     08-04 21:27:38\n",
      "coding-analysis     771  Success                                     08-04 21:26:20\n",
      "coding-analysis     770  Success                                     08-04 21:25:36\n",
      "coding-analysis     769  Success                                     08-04 21:26:51\n",
      "coding-analysis     768  Success                                     08-04 21:31:18\n",
      "coding-analysis     767  Success                                     08-04 21:28:39\n",
      "coding-analysis     766  Success                                     08-04 21:31:56\n",
      "coding-analysis     765  Success                                     08-04 21:27:39\n",
      "coding-analysis     764  Success                                     08-04 21:29:52\n",
      "coding-analysis     763  Success                                     08-04 21:29:39\n",
      "coding-analysis     762  Success                                     08-04 21:26:52\n",
      "coding-analysis     761  Success                                     08-04 21:26:52\n",
      "coding-analysis     760  Success                                     08-04 21:26:51\n",
      "coding-analysis     759  Success                                     08-04 21:27:08\n",
      "coding-analysis     758  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis     757  Success                                     08-04 21:25:02\n",
      "coding-analysis     756  Success                                     08-04 21:25:36\n",
      "coding-analysis     755  Success                                     08-04 21:23:55\n",
      "coding-analysis     754  Success                                     08-04 21:24:11\n",
      "coding-analysis     753  Success                                     08-04 21:25:02\n",
      "coding-analysis     752  Success                                     08-04 21:26:22\n",
      "coding-analysis     751  Success                                     08-04 21:23:07\n",
      "coding-analysis     750  Success                                     08-04 21:25:03\n",
      "coding-analysis     749  Success                                     08-04 21:21:00\n",
      "coding-analysis     748  Success                                     08-04 21:23:55\n",
      "coding-analysis     747  Success                                     08-04 21:28:39\n",
      "coding-analysis     746  Success                                     08-04 21:21:26\n",
      "coding-analysis     745  Success                                     08-04 21:24:11\n",
      "coding-analysis     744  Success                                     08-04 21:19:46\n",
      "coding-analysis     743  Success                                     08-04 21:31:14\n",
      "coding-analysis     742  Success                                     08-04 21:26:49\n",
      "coding-analysis     741  Success                                     08-04 21:25:02\n",
      "coding-analysis     740  Success                                     08-04 21:25:02\n",
      "coding-analysis     739  Success                                     08-04 21:25:02\n",
      "coding-analysis     738  Success                                     08-04 21:27:34\n",
      "coding-analysis     737  Success                                     08-04 21:26:22\n",
      "coding-analysis     736  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis     735  Success                                     08-04 21:24:27\n",
      "coding-analysis     734  Success                                     08-04 21:23:42\n",
      "coding-analysis     733  Success                                     08-04 21:23:07\n",
      "coding-analysis     732  Success                                     08-04 21:26:25\n",
      "coding-analysis     731  Success                                     08-04 21:29:35\n",
      "coding-analysis     730  Success                                     08-04 21:28:43\n",
      "coding-analysis     729  Success                                     08-04 21:30:50\n",
      "coding-analysis     728  Success                                     08-04 21:30:48\n",
      "coding-analysis     727  Success                                     08-04 21:26:25\n",
      "coding-analysis     726  Success                                     08-04 21:29:58\n",
      "coding-analysis     725  Success                                     08-04 21:26:24\n",
      "coding-analysis     724  Success                                     08-04 21:26:51\n",
      "coding-analysis     723  Success                                     08-04 21:26:54\n",
      "coding-analysis     722  Success                                     08-04 21:25:45\n",
      "coding-analysis     721  Success                                     08-04 21:25:20\n",
      "coding-analysis     720  Success                                     08-04 21:28:19\n",
      "coding-analysis     719  Success                                     08-04 21:26:23\n",
      "coding-analysis     718  Success                                     08-04 21:26:52\n",
      "coding-analysis     717  Success                                     08-04 21:28:20\n",
      "coding-analysis     716  Success                                     08-04 21:27:11\n",
      "coding-analysis     715  Success                                     08-04 21:27:41\n",
      "coding-analysis     714  Success                                     08-04 21:28:07\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     713  Success                                     08-04 21:26:25\n",
      "coding-analysis     712  Success                                     08-04 21:30:48\n",
      "coding-analysis     711  Success                                     08-04 21:32:22\n",
      "coding-analysis     710  Success                                     08-04 21:26:23\n",
      "coding-analysis     709  Success                                     08-04 21:26:25\n",
      "coding-analysis     708  Success                                     08-04 21:29:58\n",
      "coding-analysis     707  Success                                     08-04 21:27:28\n",
      "coding-analysis     706  Success                                     08-04 21:27:11\n",
      "coding-analysis     705  Success                                     08-04 21:17:05\n",
      "coding-analysis     704  Success                                     08-04 21:25:20\n",
      "coding-analysis     703  Success                                     08-04 21:24:28\n",
      "coding-analysis     702  Success                                     08-04 21:25:20\n",
      "coding-analysis     701  Success                                     08-04 21:26:20\n",
      "coding-analysis     700  Success                                     08-04 21:28:44\n",
      "coding-analysis     699  Success                                     08-04 21:27:00\n",
      "coding-analysis     698  Success                                     08-04 21:32:55\n",
      "coding-analysis     697  Success                                     08-04 21:29:03\n",
      "coding-analysis     696  Success                                     08-04 21:26:30\n",
      "coding-analysis     695  Success                                     08-04 21:31:14\n",
      "coding-analysis     694  Success                                     08-04 21:29:57\n",
      "coding-analysis     693  Success                                     08-04 21:29:30\n",
      "coding-analysis     692  Success                                     08-04 21:25:03\n",
      "coding-analysis     691  Success                                     08-04 21:25:02\n",
      "coding-analysis     690  Success                                     08-04 21:24:45\n",
      "coding-analysis     689  Success                                     08-04 21:27:11\n",
      "coding-analysis     688  Success                                     08-04 21:26:49\n",
      "coding-analysis     687  Success                                     08-04 21:26:00\n",
      "coding-analysis     686  Success                                     08-04 21:25:18\n",
      "coding-analysis     685  Success                                     08-04 21:24:44\n",
      "coding-analysis     684  Success                                     08-04 21:26:03\n",
      "coding-analysis     683  Success                                     08-04 21:26:03\n",
      "coding-analysis     682  Success                                     08-04 21:26:56\n",
      "coding-analysis     681  Success                                     08-04 21:28:00\n",
      "coding-analysis     680  Success                                     08-04 21:26:49\n",
      "coding-analysis     679  Success                                     08-04 21:26:02\n",
      "coding-analysis     678  Success                                     08-04 21:30:47\n",
      "coding-analysis     677  Success                                     08-04 21:26:49\n",
      "coding-analysis     676  Success                                     08-04 21:16:22\n",
      "coding-analysis     675  Success                                     08-04 21:29:49\n",
      "coding-analysis     674  Success                                     08-04 21:31:36\n",
      "coding-analysis     673  Success                                     08-04 21:28:16\n",
      "coding-analysis     672  Success                                     08-04 21:26:20\n",
      "coding-analysis     671  Success                                     08-04 21:28:39\n",
      "coding-analysis     670  Success                                     08-04 21:28:23\n",
      "coding-analysis     669  Success                                     08-04 21:26:22\n",
      "coding-analysis     668  Success                                     08-04 21:25:35\n",
      "coding-analysis     667  Success                                     08-04 21:24:11\n",
      "coding-analysis     666  Success                                     08-04 21:26:20\n",
      "coding-analysis     665  Success                                     08-04 21:27:26\n",
      "coding-analysis     664  Success                                     08-04 21:24:44\n",
      "coding-analysis     663  Success                                     08-04 21:32:31\n",
      "coding-analysis     662  Success                                     08-04 21:26:00\n",
      "coding-analysis     661  Success                                     08-04 21:30:12\n",
      "coding-analysis     660  Success                                     08-04 21:30:53\n",
      "coding-analysis     659  Success                                     08-04 21:18:40\n",
      "coding-analysis     658  Success                                     08-04 21:27:55\n",
      "coding-analysis     657  Success                                     08-04 21:26:04\n",
      "coding-analysis     656  Success                                     08-04 21:25:35\n",
      "coding-analysis     655  Success                                     08-04 21:26:03\n",
      "coding-analysis     654  Success                                     08-04 21:26:20\n",
      "coding-analysis     653  Success                                     08-04 21:26:29\n",
      "coding-analysis     652  Success                                     08-04 21:26:51\n",
      "coding-analysis     651  Success                                     08-04 21:30:33\n",
      "coding-analysis     650  Success                                     08-04 21:28:44\n",
      "coding-analysis     649  Success                                     08-04 21:26:25\n",
      "coding-analysis     648  Success                                     08-04 21:30:16\n",
      "coding-analysis     647  Success                                     08-04 21:30:45\n",
      "coding-analysis     646  Success                                     08-04 21:25:18\n",
      "coding-analysis     645  Success                                     08-04 21:27:56\n",
      "coding-analysis     644  Success                                     08-04 21:24:45\n",
      "coding-analysis     643  Success                                     08-04 21:29:18\n",
      "coding-analysis     642  Success                                     08-04 21:30:06\n",
      "coding-analysis     641  Success                                     08-04 21:26:22\n",
      "coding-analysis     640  Success                                     08-04 21:25:03\n",
      "coding-analysis     639  Success                                     08-04 21:25:38\n",
      "coding-analysis     638  Success                                     08-04 21:26:20\n",
      "coding-analysis     637  Success                                     08-04 21:25:03\n",
      "coding-analysis     636  Success                                     08-04 21:24:11\n",
      "coding-analysis     635  Success                                     08-04 21:15:15\n",
      "coding-analysis     634  Success                                     08-04 21:31:14\n",
      "coding-analysis     633  Success                                     08-04 21:25:18\n",
      "coding-analysis     632  Success                                     08-04 21:23:55\n",
      "coding-analysis     631  Success                                     08-04 21:26:52\n",
      "coding-analysis     630  Success                                     08-04 21:25:35\n",
      "coding-analysis     629  Success                                     08-04 21:29:14\n",
      "coding-analysis     628  Success                                     08-04 21:28:43\n",
      "coding-analysis     627  Success                                     08-04 21:26:00\n",
      "coding-analysis     626  Success                                     08-04 21:25:20\n",
      "coding-analysis     625  Success                                     08-04 21:29:29\n",
      "coding-analysis     624  Success                                     08-04 21:26:23\n",
      "coding-analysis     623  Success                                     08-04 21:28:03\n",
      "coding-analysis     622  Success                                     08-04 21:26:23\n",
      "coding-analysis     621  Success                                     08-04 21:25:20\n",
      "coding-analysis     620  Success                                     08-04 21:26:20\n",
      "coding-analysis     619  Success                                     08-04 21:26:02\n",
      "coding-analysis     618  Success                                     08-04 21:26:25\n",
      "coding-analysis     617  Success                                     08-04 21:31:37\n",
      "coding-analysis     616  Success                                     08-04 21:29:51\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     615  Success                                     08-04 21:27:12\n",
      "coding-analysis     614  Success                                     08-04 21:26:52\n",
      "coding-analysis     613  Success                                     08-04 21:29:24\n",
      "coding-analysis     612  Success                                     08-04 21:25:20\n",
      "coding-analysis     611  Success                                     08-04 21:25:43\n",
      "coding-analysis     610  Success                                     08-04 21:26:54\n",
      "coding-analysis     609  Success                                     08-04 21:28:07\n",
      "coding-analysis     608  Success                                     08-04 21:21:27\n",
      "coding-analysis     607  Success                                     08-04 21:25:02\n",
      "coding-analysis     606  Success                                     08-04 21:26:31\n",
      "coding-analysis     605  Success                                     08-04 21:25:35\n",
      "coding-analysis     604  Success                                     08-04 21:25:36\n",
      "coding-analysis     603  Success                                     08-04 21:26:51\n",
      "coding-analysis     602  Success                                     08-04 21:29:19\n",
      "coding-analysis     601  Success                                     08-04 21:26:59\n",
      "coding-analysis     600  Success                                     08-04 21:25:36\n",
      "coding-analysis     599  Success                                     08-04 21:25:18\n",
      "coding-analysis     598  Success                                     08-04 21:27:08\n",
      "coding-analysis     597  Success                                     08-04 21:29:09\n",
      "coding-analysis     596  Success                                     08-04 21:26:22\n",
      "coding-analysis     595  Success                                     08-04 21:29:49\n",
      "coding-analysis     594  Success                                     08-04 21:27:38\n",
      "coding-analysis     593  Success                                     08-04 21:27:49\n",
      "coding-analysis     592  Success                                     08-04 21:23:55\n",
      "coding-analysis     591  Success                                     08-04 21:29:54\n",
      "coding-analysis     590  Success                                     08-04 21:26:04\n",
      "coding-analysis     589  Success                                     08-04 21:26:20\n",
      "coding-analysis     588  Success                                     08-04 21:29:47\n",
      "coding-analysis     587  Success                                     08-04 21:26:03\n",
      "coding-analysis     586  Success                                     08-04 21:27:28\n",
      "coding-analysis     585  Success                                     08-04 21:29:48\n",
      "coding-analysis     584  Success                                     08-04 21:26:00\n",
      "coding-analysis     583  Success                                     08-04 21:28:40\n",
      "coding-analysis     582  Success                                     08-04 21:15:49\n",
      "coding-analysis     581  Success                                     08-04 21:26:22\n",
      "coding-analysis     580  Success                                     08-04 21:25:36\n",
      "coding-analysis     579  Success                                     08-04 21:29:53\n",
      "coding-analysis     578  Success                                     08-04 21:28:08\n",
      "coding-analysis     577  Success                                     08-04 21:26:00\n",
      "coding-analysis     576  Success                                     08-04 21:26:23\n",
      "coding-analysis     575  Success                                     08-04 21:28:25\n",
      "coding-analysis     574  Success                                     08-04 21:25:02\n",
      "coding-analysis     573  Success                                     08-04 21:30:13\n",
      "coding-analysis     572  Success                                     08-04 21:27:16\n",
      "coding-analysis     571  Success                                     08-04 21:27:35\n",
      "coding-analysis     570  Success                                     08-04 21:26:23\n",
      "coding-analysis     569  Success                                     08-04 21:26:20\n",
      "coding-analysis     568  Success                                     08-04 21:24:45\n",
      "coding-analysis     567  Success                                     08-04 21:13:26\n",
      "coding-analysis     566  Success                                     08-04 21:24:11\n",
      "coding-analysis     565  Success                                     08-04 21:26:22\n",
      "coding-analysis     564  Success                                     08-04 21:24:44\n",
      "coding-analysis     563  Success                                     08-04 21:24:45\n",
      "coding-analysis     562  Success                                     08-04 21:25:35\n",
      "coding-analysis     561  Success                                     08-04 21:29:35\n",
      "coding-analysis     560  Success                                     08-04 21:30:06\n",
      "coding-analysis     559  Success                                     08-04 21:25:42\n",
      "coding-analysis     558  Success                                     08-04 21:26:20\n",
      "coding-analysis     557  Success                                     08-04 21:31:36\n",
      "coding-analysis     556  Success                                     08-04 21:25:03\n",
      "coding-analysis     555  Success                                     08-04 21:26:02\n",
      "coding-analysis     554  Success                                     08-04 21:25:03\n",
      "coding-analysis     553  Success                                     08-04 21:26:25\n",
      "coding-analysis     552  Success                                     08-04 21:26:54\n",
      "coding-analysis     551  Success                                     08-04 21:13:26\n",
      "coding-analysis     550  Success                                     08-04 21:26:02\n",
      "coding-analysis     549  Success                                     08-04 21:29:05\n",
      "coding-analysis     548  Success                                     08-04 21:30:21\n",
      "coding-analysis     547  Success                                     08-04 21:29:31\n",
      "coding-analysis     546  Success                                     08-04 21:24:45\n",
      "coding-analysis     545  Success                                     08-04 21:30:27\n",
      "coding-analysis     544  Success                                     08-04 21:29:31\n",
      "coding-analysis     543  Success                                     08-04 21:26:00\n",
      "coding-analysis     542  Success                                     08-04 21:32:38\n",
      "coding-analysis     541  Success                                     08-04 21:28:39\n",
      "coding-analysis     540  Success                                     08-04 21:28:08\n",
      "coding-analysis     539  Success                                     08-04 21:29:10\n",
      "coding-analysis     538  Success                                     08-04 21:24:27\n",
      "coding-analysis     537  Success                                     08-04 21:22:34\n",
      "coding-analysis     536  Success                                     08-04 21:25:36\n",
      "coding-analysis     535  Success                                     08-04 21:25:36\n",
      "coding-analysis     534  Success                                     08-04 21:29:28\n",
      "coding-analysis     533  Success                                     08-04 21:27:07\n",
      "coding-analysis     532  Success                                     08-04 21:26:51\n",
      "coding-analysis     531  Success                                     08-04 21:25:42\n",
      "coding-analysis     530  Success                                     08-04 21:26:20\n",
      "coding-analysis     529  Success                                     08-04 21:27:49\n",
      "coding-analysis     528  Success                                     08-04 21:27:15\n",
      "coding-analysis     527  Success                                     08-04 21:29:18\n",
      "coding-analysis     526  Success                                     08-04 21:29:23\n",
      "coding-analysis     525  Success                                     08-04 21:25:20\n",
      "coding-analysis     524  Success                                     08-04 21:26:00\n",
      "coding-analysis     523  Success                                     08-04 21:30:01\n",
      "coding-analysis     522  Success                                     08-04 21:25:38\n",
      "coding-analysis     521  Success                                     08-04 21:29:28\n",
      "coding-analysis     520  Success                                     08-04 21:27:06\n",
      "coding-analysis     519  Success                                     08-04 21:15:35\n",
      "coding-analysis     518  Success                                     08-04 21:26:00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     517  Success                                     08-04 21:27:16\n",
      "coding-analysis     516  Success                                     08-04 21:25:36\n",
      "coding-analysis     515  Success                                     08-04 21:24:27\n",
      "coding-analysis     514  Success                                     08-04 21:29:28\n",
      "coding-analysis     513  Success                                     08-04 21:28:24\n",
      "coding-analysis     512  Success                                     08-04 21:23:25\n",
      "coding-analysis     511  Success                                     08-04 21:25:36\n",
      "coding-analysis     510  Success                                     08-04 21:26:26\n",
      "coding-analysis     509  Success                                     08-04 21:28:01\n",
      "coding-analysis     508  Success                                     08-04 21:26:03\n",
      "coding-analysis     507  Success                                     08-04 21:28:58\n",
      "coding-analysis     506  Success                                     08-04 21:27:55\n",
      "coding-analysis     505  Success                                     08-04 21:25:03\n",
      "coding-analysis     504  Success                                     08-04 21:24:11\n",
      "coding-analysis     503  Success                                     08-04 21:28:59\n",
      "coding-analysis     502  Success                                     08-04 21:25:36\n",
      "coding-analysis     501  Success                                     08-04 21:25:18\n",
      "coding-analysis     500  Success                                     08-04 21:26:26\n",
      "coding-analysis     499  Success                                     08-04 21:25:02\n",
      "coding-analysis     498  Success                                     08-04 21:12:42\n",
      "coding-analysis     497  Success                                     08-04 21:24:45\n",
      "coding-analysis     496  Success                                     08-04 21:28:12\n",
      "coding-analysis     495  Success                                     08-04 21:28:39\n",
      "coding-analysis     494  Success                                     08-04 21:25:43\n",
      "coding-analysis     493  Success                                     08-04 21:25:20\n",
      "coding-analysis     492  Success                                     08-04 21:46:31\n",
      "coding-analysis     491  Success                                     08-04 21:44:42\n",
      "coding-analysis     490  Success                                     08-04 21:24:44\n",
      "coding-analysis     489  Success                                     08-04 21:25:41\n",
      "coding-analysis     488  Success                                     08-04 21:24:27\n",
      "coding-analysis     487  Success                                     08-04 21:25:02\n",
      "coding-analysis     486  Success                                     08-04 21:25:20\n",
      "coding-analysis     485  Success                                     08-04 21:24:27\n",
      "coding-analysis     484  Success                                     08-04 21:25:36\n",
      "coding-analysis     483  Success                                     08-04 21:29:49\n",
      "coding-analysis     482  Success                                     08-04 21:25:35\n",
      "coding-analysis     481  Success                                     08-04 21:26:23\n",
      "coding-analysis     480  Success                                     08-04 21:23:55\n",
      "coding-analysis     479  Success                                     08-04 21:28:39\n",
      "coding-analysis     478  Success                                     08-04 21:26:22\n",
      "coding-analysis     477  Success                                     08-04 21:27:34\n",
      "coding-analysis     476  Success                                     08-04 21:30:00\n",
      "coding-analysis     475  Success                                     08-04 21:28:22\n",
      "coding-analysis     474  Success                                     08-04 21:31:36\n",
      "coding-analysis     473  Success                                     08-04 21:26:02\n",
      "coding-analysis     472  Success                                     08-04 21:23:07\n",
      "coding-analysis     471  Success                                     08-04 21:28:23\n",
      "coding-analysis     470  Success                                     08-04 21:28:49\n",
      "coding-analysis     469  Success                                     08-04 21:25:36\n",
      "coding-analysis     468  Success                                     08-04 21:24:11\n",
      "coding-analysis     467  Success                                     08-04 21:26:23\n",
      "coding-analysis     466  Success                                     08-04 21:28:39\n",
      "coding-analysis     465  Success                                     08-04 21:33:01\n",
      "coding-analysis     464  Success                                     08-04 21:24:27\n",
      "coding-analysis     463  Success                                     08-04 21:29:32\n",
      "coding-analysis     462  Success                                     08-04 21:28:45\n",
      "coding-analysis     461  Success                                     08-04 21:26:02\n",
      "coding-analysis     460  Success                                     08-04 21:26:02\n",
      "coding-analysis     459  Success                                     08-04 21:27:41\n",
      "coding-analysis     458  Success                                     08-04 21:24:45\n",
      "coding-analysis     457  Success                                     08-04 21:28:02\n",
      "coding-analysis     456  Success                                     08-04 21:28:49\n",
      "coding-analysis     455  Success                                     08-04 21:28:39\n",
      "coding-analysis     454  Success                                     08-04 21:25:20\n",
      "coding-analysis     453  Success                                     08-04 21:28:09\n",
      "coding-analysis     452  Success                                     08-04 21:27:44\n",
      "coding-analysis     451  Success                                     08-04 21:29:35\n",
      "coding-analysis     450  Success                                     08-04 21:27:08\n",
      "coding-analysis     449  Success                                     08-04 21:29:05\n",
      "coding-analysis     448  Success                                     08-04 21:27:33\n",
      "coding-analysis     447  Success                                     08-04 21:26:00\n",
      "coding-analysis     446  Success                                     08-04 21:30:29\n",
      "coding-analysis     445  Success                                     08-04 21:32:22\n",
      "coding-analysis     444  Success                                     08-04 21:19:02\n",
      "coding-analysis     443  Success                                     08-04 21:32:23\n",
      "coding-analysis     442  Success                                     08-04 21:27:56\n",
      "coding-analysis     441  Success                                     08-04 21:30:31\n",
      "coding-analysis     440  Success                                     08-04 21:26:57\n",
      "coding-analysis     439  Success                                     08-04 21:27:05\n",
      "coding-analysis     438  Success                                     08-04 21:26:00\n",
      "coding-analysis     437  Success                                     08-04 21:28:00\n",
      "coding-analysis     436  Success                                     08-04 21:26:02\n",
      "coding-analysis     435  Success                                     08-04 21:36:45\n",
      "coding-analysis     434  Success                                     08-04 21:32:21\n",
      "coding-analysis     433  Success                                     08-04 21:27:46\n",
      "coding-analysis     432  Success                                     08-04 21:27:46\n",
      "coding-analysis     431  Success                                     08-04 21:26:24\n",
      "coding-analysis     430  Success                                     08-04 21:26:54\n",
      "coding-analysis     429  Success                                     08-04 21:28:58\n",
      "coding-analysis     428  Success                                     08-04 21:26:20\n",
      "coding-analysis     427  Success                                     08-04 21:27:35\n",
      "coding-analysis     426  Success                                     08-04 21:25:20\n",
      "coding-analysis     425  Success                                     08-04 21:28:46\n",
      "coding-analysis     424  Success                                     08-04 21:28:15\n",
      "coding-analysis     423  Success                                     08-04 21:32:21\n",
      "coding-analysis     422  Success                                     08-04 21:26:55\n",
      "coding-analysis     421  Success                                     08-04 21:28:49\n",
      "coding-analysis     420  Success                                     08-04 21:30:22\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     419  Success                                     08-04 21:27:51\n",
      "coding-analysis     418  Success                                     08-04 21:27:36\n",
      "coding-analysis     417  Success                                     08-04 21:28:39\n",
      "coding-analysis     416  Success                                     08-04 21:27:25\n",
      "coding-analysis     415  Success                                     08-04 21:28:21\n",
      "coding-analysis     414  Success                                     08-04 21:31:14\n",
      "coding-analysis     413  Success                                     08-04 21:27:10\n",
      "coding-analysis     412  Success                                     08-04 21:27:08\n",
      "coding-analysis     411  Success                                     08-04 21:31:59\n",
      "coding-analysis     410  Success                                     08-04 21:26:55\n",
      "coding-analysis     409  Success                                     08-04 21:28:41\n",
      "coding-analysis     408  Success                                     08-04 21:26:59\n",
      "coding-analysis     407  Success                                     08-04 21:33:09\n",
      "coding-analysis     406  Success                                     08-04 21:29:07\n",
      "coding-analysis     405  Success                                     08-04 21:27:13\n",
      "coding-analysis     404  Success                                     08-04 21:27:52\n",
      "coding-analysis     403  Success                                     08-04 21:33:22\n",
      "coding-analysis     402  Success                                     08-04 21:30:20\n",
      "coding-analysis     401  Success                                     08-04 21:27:48\n",
      "coding-analysis     400  Success                                     08-04 21:32:29\n",
      "coding-analysis     399  Success                                     08-04 21:28:51\n",
      "coding-analysis     398  Success                                     08-04 21:28:23\n",
      "coding-analysis     397  Success                                     08-04 21:29:23\n",
      "coding-analysis     396  Success                                     08-04 21:31:14\n",
      "coding-analysis     395  Success                                     08-04 21:26:49\n",
      "coding-analysis     394  Success                                     08-04 21:26:31\n",
      "coding-analysis     393  Success                                     08-04 21:27:14\n",
      "coding-analysis     392  Success                                     08-04 21:30:55\n",
      "coding-analysis     391  Success                                     08-04 21:29:18\n",
      "coding-analysis     390  Success                                     08-04 21:29:41\n",
      "coding-analysis     389  Success                                     08-04 21:28:06\n",
      "coding-analysis     388  Success                                     08-04 21:29:39\n",
      "coding-analysis     387  Success                                     08-04 21:31:36\n",
      "coding-analysis     386  Success                                     08-04 21:27:41\n",
      "coding-analysis     385  Success                                     08-04 21:32:04\n",
      "coding-analysis     384  Success                                     08-04 21:26:56\n",
      "coding-analysis     383  Success                                     08-04 21:26:03\n",
      "coding-analysis     382  Success                                     08-04 21:30:30\n",
      "coding-analysis     381  Success                                     08-04 21:26:25\n",
      "coding-analysis     380  Success                                     08-04 21:26:20\n",
      "coding-analysis     379  Success                                     08-04 21:25:18\n",
      "coding-analysis     378  Success                                     08-04 21:38:28\n",
      "coding-analysis     377  Success                                     08-04 21:32:21\n",
      "coding-analysis     376  Success                                     08-04 21:28:18\n",
      "coding-analysis     375  Success                                     08-04 21:34:32\n",
      "coding-analysis     374  Success                                     08-04 21:32:33\n",
      "coding-analysis     373  Success                                     08-04 21:38:26\n",
      "coding-analysis     372  Success                                     08-04 21:28:59\n",
      "coding-analysis     371  Success                                     08-04 21:32:21\n",
      "coding-analysis     370  Success                                     08-04 21:23:23\n",
      "coding-analysis     369  Success                                     08-04 21:35:14\n",
      "coding-analysis     368  Success                                     08-04 21:31:36\n",
      "coding-analysis     367  Success                                     08-04 21:29:56\n",
      "coding-analysis     366  Success                                     08-04 21:26:03\n",
      "coding-analysis     365  Success                                     08-04 21:23:07\n",
      "coding-analysis     364  Success                                     08-04 21:35:32\n",
      "coding-analysis     363  Success                                     08-04 21:29:01\n",
      "coding-analysis     362  Success                                     08-04 21:29:39\n",
      "coding-analysis     361  Success                                     08-04 21:31:17\n",
      "coding-analysis     360  Success                                     08-04 21:31:15\n",
      "coding-analysis     359  Success                                     08-04 21:31:14\n",
      "coding-analysis     358  Success                                     08-04 21:27:37\n",
      "coding-analysis     357  Success                                     08-04 21:31:15\n",
      "coding-analysis     356  Success                                     08-04 21:35:32\n",
      "coding-analysis     355  Success                                     08-04 21:27:21\n",
      "coding-analysis     354  Success                                     08-04 21:29:13\n",
      "coding-analysis     353  Success                                     08-04 21:35:09\n",
      "coding-analysis     352  Success                                     08-04 21:33:55\n",
      "coding-analysis     351  Success                                     08-04 21:32:47\n",
      "coding-analysis     350  Success                                     08-04 21:38:55\n",
      "coding-analysis     349  Success                                     08-04 21:32:21\n",
      "coding-analysis     348  Success                                     08-04 21:32:41\n",
      "coding-analysis     347  Success                                     08-04 21:37:41\n",
      "coding-analysis     346  Success                                     08-04 21:33:04\n",
      "coding-analysis     345  Success                                     08-04 21:35:32\n",
      "coding-analysis     344  Success                                     08-04 21:34:34\n",
      "coding-analysis     343  Success                                     08-04 21:33:25\n",
      "coding-analysis     342  Success                                     08-04 21:33:23\n",
      "coding-analysis     341  Success                                     08-04 21:38:07\n",
      "coding-analysis     340  Success                                     08-04 21:32:27\n",
      "coding-analysis     339  Success                                     08-04 21:38:07\n",
      "coding-analysis     338  Success                                     08-04 21:36:04\n",
      "coding-analysis     337  Success                                     08-04 21:38:26\n",
      "coding-analysis     336  Success                                     08-04 21:35:09\n",
      "coding-analysis     335  Success                                     08-04 21:39:33\n",
      "coding-analysis     334  Success                                     08-04 21:32:21\n",
      "coding-analysis     333  Success                                     08-04 21:32:21\n",
      "coding-analysis     332  Stopped running \"user-command\": exit st...  08-04 21:16:08\n",
      "coding-analysis     331  Success                                     08-04 21:38:26\n",
      "coding-analysis     330  Success                                     08-04 21:34:32\n",
      "coding-analysis     329  Success                                     08-04 21:34:33\n",
      "coding-analysis     328  Success                                     08-04 21:37:45\n",
      "coding-analysis     327  Success                                     08-04 21:35:48\n",
      "coding-analysis     326  Success                                     08-04 21:18:01\n",
      "coding-analysis     325  Success                                     08-04 21:32:21\n",
      "coding-analysis     324  Success                                     08-04 21:30:42\n",
      "coding-analysis     323  Success                                     08-04 21:33:08\n",
      "coding-analysis     322  Success                                     08-04 21:34:01\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     321  Success                                     08-04 21:34:13\n",
      "coding-analysis     320  Success                                     08-04 21:36:45\n",
      "coding-analysis     319  Success                                     08-04 21:37:12\n",
      "coding-analysis     318  Success                                     08-04 21:37:29\n",
      "coding-analysis     317  Success                                     08-04 21:35:50\n",
      "coding-analysis     316  Success                                     08-04 21:42:01\n",
      "coding-analysis     315  Success                                     08-04 21:32:56\n",
      "coding-analysis     314  Success                                     08-04 21:31:14\n",
      "coding-analysis     313  Success                                     08-04 21:28:39\n",
      "coding-analysis     312  Success                                     08-04 21:30:50\n",
      "coding-analysis     311  Success                                     08-04 21:30:43\n",
      "coding-analysis     310  Success                                     08-04 21:32:55\n",
      "coding-analysis     309  Success                                     08-04 21:27:08\n",
      "coding-analysis     308  Success                                     08-04 21:37:13\n",
      "coding-analysis     307  Success                                     08-04 21:33:30\n",
      "coding-analysis     306  Success                                     08-04 21:31:14\n",
      "coding-analysis     305  Success                                     08-04 21:35:31\n",
      "coding-analysis     304  Success                                     08-04 21:31:14\n",
      "coding-analysis     303  Success                                     08-04 21:40:58\n",
      "coding-analysis     302  Success                                     08-04 21:38:26\n",
      "coding-analysis     301  Success                                     08-04 21:41:23\n",
      "coding-analysis     300  Success                                     08-04 21:32:01\n",
      "coding-analysis     299  Success                                     08-04 21:35:32\n",
      "coding-analysis     298  Success                                     08-04 21:34:14\n",
      "coding-analysis     297  Success                                     08-04 21:17:55\n",
      "coding-analysis     296  Success                                     08-04 21:38:26\n",
      "coding-analysis     295  Success                                     08-04 21:41:39\n",
      "coding-analysis     294  Success                                     08-04 21:38:26\n",
      "coding-analysis     293  Success                                     08-04 21:34:35\n",
      "coding-analysis     292  Success                                     08-04 21:30:32\n",
      "coding-analysis     291  Success                                     08-04 21:30:44\n",
      "coding-analysis     290  worker was terminated                       08-04 21:11:13\n",
      "coding-analysis     289  Success                                     08-04 21:31:38\n",
      "coding-analysis     288  Success                                     08-04 21:27:54\n",
      "coding-analysis     287  Success                                     08-04 21:30:58\n",
      "coding-analysis     286  Success                                     08-04 21:34:51\n",
      "coding-analysis     285  Success                                     08-04 21:32:21\n",
      "coding-analysis     284  Success                                     08-04 21:36:25\n",
      "coding-analysis     283  Success                                     08-04 21:33:32\n",
      "coding-analysis     282  Success                                     08-04 21:47:59\n",
      "coding-analysis     281  Success                                     08-04 21:33:44\n",
      "coding-analysis     280  Success                                     08-04 21:25:03\n",
      "coding-analysis     279  Success                                     08-04 21:42:02\n",
      "coding-analysis     278  Success                                     08-04 21:37:38\n",
      "coding-analysis     277  Success                                     08-04 21:31:54\n",
      "coding-analysis     276  Success                                     08-04 21:36:45\n",
      "coding-analysis     275  Success                                     08-04 21:37:29\n",
      "coding-analysis     274  Success                                     08-04 21:36:25\n",
      "coding-analysis     273  Success                                     08-04 21:36:25\n",
      "coding-analysis     272  Success                                     08-04 21:34:53\n",
      "coding-analysis     271  Success                                     08-04 21:39:06\n",
      "coding-analysis     270  Success                                     08-04 21:39:58\n",
      "coding-analysis     269  Success                                     08-04 21:33:19\n",
      "coding-analysis     268  Success                                     08-04 21:38:33\n",
      "coding-analysis     267  Success                                     08-04 21:30:46\n",
      "coding-analysis     266  Success                                     08-04 21:35:48\n",
      "coding-analysis     265  Success                                     08-04 21:36:25\n",
      "coding-analysis     264  Success                                     08-04 21:34:33\n",
      "coding-analysis     263  Success                                     08-04 21:41:58\n",
      "coding-analysis     262  Success                                     08-04 21:30:52\n",
      "coding-analysis     261  Success                                     08-04 21:36:08\n",
      "coding-analysis     260  Success                                     08-04 21:31:54\n",
      "coding-analysis     259  Success                                     08-04 21:32:28\n",
      "coding-analysis     258  Success                                     08-04 21:30:36\n",
      "coding-analysis     257  Success                                     08-04 21:29:43\n",
      "coding-analysis     256  Success                                     08-04 21:16:42\n",
      "coding-analysis     255  Success                                     08-04 21:42:37\n",
      "coding-analysis     254  Success                                     08-04 21:38:59\n",
      "coding-analysis     253  Success                                     08-04 21:30:02\n",
      "coding-analysis     252  Success                                     08-04 21:30:29\n",
      "coding-analysis     251  Success                                     08-04 21:33:23\n",
      "coding-analysis     250  Success                                     08-04 21:37:45\n",
      "coding-analysis     249  Success                                     08-04 21:37:29\n",
      "coding-analysis     248  Success                                     08-04 21:32:29\n",
      "coding-analysis     247  Success                                     08-04 21:31:37\n",
      "coding-analysis     246  Success                                     08-04 21:34:32\n",
      "coding-analysis     245  Success                                     08-04 21:39:01\n",
      "coding-analysis     244  Success                                     08-04 21:35:09\n",
      "coding-analysis     243  Success                                     08-04 21:34:51\n",
      "coding-analysis     242  Success                                     08-04 21:31:36\n",
      "coding-analysis     241  Success                                     08-04 21:32:22\n",
      "coding-analysis     240  Success                                     08-04 21:32:37\n",
      "coding-analysis     239  Success                                     08-04 21:28:05\n",
      "coding-analysis     238  Success                                     08-04 21:42:03\n",
      "coding-analysis     237  Success                                     08-04 21:35:13\n",
      "coding-analysis     236  Success                                     08-04 21:32:06\n",
      "coding-analysis     235  Success                                     08-04 21:34:35\n",
      "coding-analysis     234  Success                                     08-04 21:33:32\n",
      "coding-analysis     233  worker was terminated                       08-04 21:13:26\n",
      "coding-analysis     232  Success                                     08-04 21:31:36\n",
      "coding-analysis     231  Success                                     08-04 21:36:45\n",
      "coding-analysis     230  Success                                     08-04 21:36:45\n",
      "coding-analysis     229  worker was terminated                       08-04 21:12:14\n",
      "coding-analysis     228  Success                                     08-04 21:28:59\n",
      "coding-analysis     227  Success                                     08-04 21:32:57\n",
      "coding-analysis     226  Success                                     08-04 21:30:07\n",
      "coding-analysis     225  Success                                     08-04 21:29:39\n",
      "coding-analysis     224  Success                                     08-04 21:32:05\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     223  Success                                     08-04 21:34:06\n",
      "coding-analysis     222  Success                                     08-04 21:40:01\n",
      "coding-analysis     221  Success                                     08-04 21:33:22\n",
      "coding-analysis     220  Success                                     08-04 21:37:29\n",
      "coding-analysis     219  Success                                     08-04 21:34:51\n",
      "coding-analysis     218  Success                                     08-04 21:33:01\n",
      "coding-analysis     217  Success                                     08-04 21:37:50\n",
      "coding-analysis     216  Success                                     08-04 21:39:33\n",
      "coding-analysis     215  Success                                     08-04 21:27:56\n",
      "coding-analysis     214  Success                                     08-04 21:32:58\n",
      "coding-analysis     213  Success                                     08-04 21:32:02\n",
      "coding-analysis     212  Success                                     08-04 21:38:26\n",
      "coding-analysis     211  Success                                     08-04 21:36:47\n",
      "coding-analysis     210  Success                                     08-04 21:33:25\n",
      "coding-analysis     209  Success                                     08-04 21:33:20\n",
      "coding-analysis     208  Success                                     08-04 21:36:47\n",
      "coding-analysis     207  Success                                     08-04 21:33:03\n",
      "coding-analysis     206  Success                                     08-04 21:33:00\n",
      "coding-analysis     205  Success                                     08-04 21:35:31\n",
      "coding-analysis     204  Success                                     08-04 21:46:34\n",
      "coding-analysis     203  Success                                     08-04 21:18:38\n",
      "coding-analysis     202  Success                                     08-04 21:38:28\n",
      "coding-analysis     201  Success                                     08-04 21:37:12\n",
      "coding-analysis     200  Success                                     08-04 21:34:33\n",
      "coding-analysis     199  Success                                     08-04 21:36:45\n",
      "coding-analysis     198  Success                                     08-04 21:32:23\n",
      "coding-analysis     197  Success                                     08-04 21:34:08\n",
      "coding-analysis     196  Success                                     08-04 21:29:36\n",
      "coding-analysis     195  Success                                     08-04 21:31:36\n",
      "coding-analysis     194  Success                                     08-04 21:32:34\n",
      "coding-analysis     193  Success                                     08-04 21:33:33\n",
      "coding-analysis     192  Success                                     08-04 21:31:14\n",
      "coding-analysis     191  Success                                     08-04 21:40:58\n",
      "coding-analysis     190  Success                                     08-04 21:33:24\n",
      "coding-analysis     189  Success                                     08-04 21:30:47\n",
      "coding-analysis     188  Success                                     08-04 21:13:26\n",
      "coding-analysis     187  Success                                     08-04 21:31:43\n",
      "coding-analysis     186  Success                                     08-04 21:29:47\n",
      "coding-analysis     185  Success                                     08-04 21:34:09\n",
      "coding-analysis     184  Success                                     08-04 21:32:48\n",
      "coding-analysis     183  Success                                     08-04 21:30:36\n",
      "coding-analysis     182  Success                                     08-04 21:30:46\n",
      "coding-analysis     181  Success                                     08-04 21:32:42\n",
      "coding-analysis     180  Success                                     08-04 21:30:58\n",
      "coding-analysis     179  Success                                     08-04 21:32:06\n",
      "coding-analysis     178  Success                                     08-04 21:34:32\n",
      "coding-analysis     177  Success                                     08-04 21:34:14\n",
      "coding-analysis     176  Success                                     08-04 21:38:54\n",
      "coding-analysis     175  Success                                     08-04 21:33:41\n",
      "coding-analysis     174  Success                                     08-04 21:39:59\n",
      "coding-analysis     173  Success                                     08-04 21:33:15\n",
      "coding-analysis     172  Success                                     08-04 21:15:32\n",
      "coding-analysis     171  worker was terminated                       08-04 21:14:40\n",
      "coding-analysis     170  Success                                     08-04 21:32:49\n",
      "coding-analysis     169  Success                                     08-04 21:40:58\n",
      "coding-analysis     168  Success                                     08-04 21:35:31\n",
      "coding-analysis     167  Success                                     08-04 21:30:45\n",
      "coding-analysis     166  Success                                     08-04 21:35:48\n",
      "coding-analysis     165  Success                                     08-04 21:32:53\n",
      "coding-analysis     164  Success                                     08-04 21:36:52\n",
      "coding-analysis     163  Success                                     08-04 21:47:22\n",
      "coding-analysis     162  Success                                     08-04 21:30:59\n",
      "coding-analysis     161  Success                                     08-04 21:32:02\n",
      "coding-analysis     160  Success                                     08-04 21:31:36\n",
      "coding-analysis     159  Success                                     08-04 21:32:21\n",
      "coding-analysis     158  Success                                     08-04 21:23:55\n",
      "coding-analysis     157  Success                                     08-04 21:31:37\n",
      "coding-analysis     156  Success                                     08-04 21:29:49\n",
      "coding-analysis     155  Success                                     08-04 21:31:57\n",
      "coding-analysis     154  Success                                     08-04 21:33:09\n",
      "coding-analysis     153  Success                                     08-04 21:40:41\n",
      "coding-analysis     152  Success                                     08-04 21:31:36\n",
      "coding-analysis     151  Success                                     08-04 21:31:14\n",
      "coding-analysis     150  Success                                     08-04 21:35:31\n",
      "coding-analysis     149  Success                                     08-04 21:35:09\n",
      "coding-analysis     148  Success                                     08-04 21:31:49\n",
      "coding-analysis     147  Success                                     08-04 21:38:07\n",
      "coding-analysis     146  Success                                     08-04 21:33:10\n",
      "coding-analysis     145  Success                                     08-04 21:37:13\n",
      "coding-analysis     144  Success                                     08-04 21:39:19\n",
      "coding-analysis     143  Success                                     08-04 21:39:08\n",
      "coding-analysis     142  Success                                     08-04 21:32:21\n",
      "coding-analysis     141  Success                                     08-04 21:44:41\n",
      "coding-analysis     140  Success                                     08-04 21:18:30\n",
      "coding-analysis     139  Success                                     08-04 21:37:29\n",
      "coding-analysis     138  Success                                     08-04 21:37:13\n",
      "coding-analysis     137  Success                                     08-04 21:32:04\n",
      "coding-analysis     136  Success                                     08-04 21:33:47\n",
      "coding-analysis     135  Success                                     08-04 21:35:48\n",
      "coding-analysis     134  Success                                     08-04 21:29:14\n",
      "coding-analysis     133  Success                                     08-04 21:36:45\n",
      "coding-analysis     132  Success                                     08-04 21:34:33\n",
      "coding-analysis     131  Success                                     08-04 21:30:12\n",
      "coding-analysis     130  Success                                     08-04 21:36:45\n",
      "coding-analysis     129  Success                                     08-04 21:35:15\n",
      "coding-analysis     128  Success                                     08-04 21:36:45\n",
      "coding-analysis     127  Success                                     08-04 21:30:27\n",
      "coding-analysis     126  Success                                     08-04 21:31:14\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis     125  Success                                     08-04 21:34:32\n",
      "coding-analysis     124  Success                                     08-04 21:33:26\n",
      "coding-analysis     123  Success                                     08-04 21:33:47\n",
      "coding-analysis     122  Success                                     08-04 21:28:09\n",
      "coding-analysis     121  Success                                     08-04 21:30:22\n",
      "coding-analysis     120  Success                                     08-04 21:31:22\n",
      "coding-analysis     119  Success                                     08-04 21:13:26\n",
      "coding-analysis     118  Success                                     08-04 21:32:21\n",
      "coding-analysis     117  Success                                     08-04 21:30:20\n",
      "coding-analysis     116  Success                                     08-04 21:36:45\n",
      "coding-analysis     115  Success                                     08-04 21:38:07\n",
      "coding-analysis     114  Success                                     08-04 21:30:41\n",
      "coding-analysis     113  Stopped running \"user-command\": exit st...  08-04 21:28:04\n",
      "coding-analysis     112  Stopped running \"user-command\": exit st...  08-04 21:31:14\n",
      "coding-analysis     111  Success                                     08-04 21:33:48\n",
      "coding-analysis     110  Success                                     08-04 21:30:34\n",
      "coding-analysis     109  Success                                     08-04 21:31:22\n",
      "coding-analysis     108  Success                                     08-04 21:33:48\n",
      "coding-analysis     107  Success                                     08-04 21:35:09\n",
      "coding-analysis     106  Success                                     08-04 21:29:43\n",
      "coding-analysis     105  Success                                     08-04 21:36:45\n",
      "coding-analysis     104  Success                                     08-04 21:34:12\n",
      "coding-analysis     103  Success                                     08-04 21:31:44\n",
      "coding-analysis     102  Success                                     08-04 21:36:45\n",
      "coding-analysis     101  Success                                     08-04 21:33:00\n",
      "coding-analysis     100  Success                                     08-04 21:29:57\n",
      "coding-analysis      99  Success                                     08-04 21:33:01\n",
      "coding-analysis      98  Success                                     08-04 21:29:36\n",
      "coding-analysis      97  Success                                     08-04 21:34:51\n",
      "coding-analysis      96  Success                                     08-04 21:29:00\n",
      "coding-analysis      95  Success                                     08-04 21:31:36\n",
      "coding-analysis      94  Success                                     08-04 21:32:42\n",
      "coding-analysis      93  Success                                     08-04 21:25:42\n",
      "coding-analysis      92  Success                                     08-04 21:31:14\n",
      "coding-analysis      91  Success                                     08-04 21:32:22\n",
      "coding-analysis      90  Success                                     08-04 21:32:43\n",
      "coding-analysis      89  Success                                     08-04 21:27:53\n",
      "coding-analysis      88  Success                                     08-04 21:34:10\n",
      "coding-analysis      87  Success                                     08-04 21:35:17\n",
      "coding-analysis      86  Stopped running \"user-command\": exit st...  08-04 21:21:26\n",
      "coding-analysis      85  Success                                     08-04 21:31:36\n",
      "coding-analysis      84  Success                                     08-04 21:32:21\n",
      "coding-analysis      83  Success                                     08-04 21:31:36\n",
      "coding-analysis      82  Success                                     08-04 21:31:14\n",
      "coding-analysis      81  Success                                     08-04 21:32:59\n",
      "coding-analysis      80  Success                                     08-04 21:33:13\n",
      "coding-analysis      79  Success                                     08-04 21:32:23\n",
      "coding-analysis      78  Success                                     08-04 21:34:51\n",
      "coding-analysis      77  Success                                     08-04 21:36:04\n",
      "coding-analysis      76  Success                                     08-04 21:31:46\n",
      "coding-analysis      75  Success                                     08-04 21:33:09\n",
      "coding-analysis      74  Success                                     08-04 21:36:06\n",
      "coding-analysis      73  Success                                     08-04 21:39:55\n",
      "coding-analysis      72  Success                                     08-04 21:32:36\n",
      "coding-analysis      71  Success                                     08-04 21:36:27\n",
      "coding-analysis      70  Success                                     08-04 21:32:31\n",
      "coding-analysis      69  Success                                     08-04 21:34:12\n",
      "coding-analysis      68  Success                                     08-04 21:31:36\n",
      "coding-analysis      67  Success                                     08-04 21:34:51\n",
      "coding-analysis      66  Success                                     08-04 21:42:21\n",
      "coding-analysis      65  Success                                     08-04 21:21:28\n",
      "coding-analysis      64  Success                                     08-04 21:33:35\n",
      "coding-analysis      63  Success                                     08-04 21:37:14\n",
      "coding-analysis      62  Success                                     08-04 21:32:36\n",
      "coding-analysis      61  Success                                     08-04 21:32:35\n",
      "coding-analysis      60  Success                                     08-04 21:36:25\n",
      "coding-analysis      59  Success                                     08-04 21:33:48\n",
      "coding-analysis      58  Success                                     08-04 21:37:13\n",
      "coding-analysis      57  Success                                     08-04 21:30:43\n",
      "coding-analysis      56  Success                                     08-04 21:40:39\n",
      "coding-analysis      55  Success                                     08-04 21:44:24\n",
      "coding-analysis      54  Success                                     08-04 21:36:28\n",
      "coding-analysis      53  Success                                     08-04 21:33:44\n",
      "coding-analysis      52  Success                                     08-04 21:32:21\n",
      "coding-analysis      51  Success                                     08-04 21:36:45\n",
      "coding-analysis      50  Success                                     08-04 21:36:09\n",
      "coding-analysis      49  worker was terminated                       08-04 21:12:57\n",
      "coding-analysis      48  Success                                     08-04 21:39:55\n",
      "coding-analysis      47  Success                                     08-04 21:28:39\n",
      "coding-analysis      46  Success                                     08-04 21:36:04\n",
      "coding-analysis      45  Success                                     08-04 21:36:25\n",
      "coding-analysis      44  Success                                     08-04 21:34:32\n",
      "coding-analysis      43  Success                                     08-04 21:32:49\n",
      "coding-analysis      42  Success                                     08-04 21:34:32\n",
      "coding-analysis      41  Success                                     08-04 21:30:29\n",
      "coding-analysis      40  Success                                     08-04 21:32:21\n",
      "coding-analysis      39  Success                                     08-04 21:40:36\n",
      "coding-analysis      38  Success                                     08-04 21:35:31\n",
      "coding-analysis      37  Success                                     08-04 21:39:05\n",
      "coding-analysis      36  Success                                     08-04 21:33:08\n",
      "coding-analysis      35  Success                                     08-04 21:34:32\n",
      "coding-analysis      34  Success                                     08-04 21:31:50\n",
      "coding-analysis      33  Success                                     08-04 21:32:35\n",
      "coding-analysis      32  Success                                     08-04 21:36:50\n",
      "coding-analysis      31  Success                                     08-04 21:39:33\n",
      "coding-analysis      30  Success                                     08-04 21:39:55\n",
      "coding-analysis      29  Success                                     08-04 21:31:37\n",
      "coding-analysis      28  Success                                     08-04 21:36:28\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis      27  Success                                     08-04 21:32:24\n",
      "coding-analysis      26  Success                                     08-04 21:31:37\n",
      "coding-analysis      25  Success                                     08-04 21:31:36\n",
      "coding-analysis      24  Success                                     08-04 21:38:54\n",
      "coding-analysis      23  Success                                     08-04 21:31:36\n",
      "coding-analysis      22  Success                                     08-04 21:34:32\n",
      "coding-analysis      21  Success                                     08-04 21:36:25\n",
      "coding-analysis      20  Success                                     08-04 21:40:36\n",
      "coding-analysis      19  Success                                     08-04 21:32:21\n",
      "coding-analysis      18  Success                                     08-04 21:30:20\n",
      "coding-analysis      17  Success                                     08-04 21:34:51\n",
      "coding-analysis      16  Success                                     08-04 21:31:36\n",
      "coding-analysis      15  Success                                     08-04 21:33:31\n",
      "coding-analysis      14  Success                                     08-04 21:34:05\n",
      "coding-analysis      13  Success                                     08-04 21:33:02\n",
      "coding-analysis      12  Success                                     08-04 21:30:41\n",
      "coding-analysis      11  Success                                     08-04 21:33:20\n",
      "coding-analysis      10  Success                                     08-04 21:32:32\n",
      "coding-analysis       9  Success                                     08-04 21:35:09\n",
      "coding-analysis       8  Success                                     08-04 21:50:14\n",
      "coding-analysis       7  Success                                     08-04 21:32:21\n",
      "coding-analysis       6  Success                                     08-04 21:42:57\n",
      "coding-analysis       5  Success                                     08-04 21:30:43\n",
      "coding-analysis       4  Success                                     08-04 21:30:21\n",
      "coding-analysis       3  Success                                     08-04 21:30:55\n",
      "coding-analysis       2  worker was terminated                       08-04 21:16:42\n",
      "coding-analysis       1  Success                                     08-04 21:32:06\n",
      "coding-analysis       6  Success                                     08-04 20:58:15\n",
      "coding-analysis       5  Success                                     08-04 21:04:21\n",
      "coding-analysis       4  Success                                     08-04 20:58:14\n",
      "coding-analysis       3  Success                                     08-04 20:58:19\n",
      "coding-analysis       2  Success                                     08-04 20:58:03\n",
      "coding-analysis       1  Success                                     08-04 21:05:07\n",
      "coding-analysis       6  Success                                     08-04 20:18:36\n",
      "coding-analysis       5  Success                                     08-04 20:22:22\n",
      "coding-analysis       4  Success                                     08-04 20:18:37\n",
      "coding-analysis       3  Success                                     08-04 20:18:23\n",
      "coding-analysis       2  Success                                     08-04 20:18:55\n",
      "coding-analysis       1  Success                                     08-04 20:23:15\n",
      "coding-analysis       6  Success                                     08-04 19:37:17\n",
      "coding-analysis       5  Success                                     08-04 19:40:43\n",
      "coding-analysis       4  Success                                     08-04 19:36:38\n",
      "coding-analysis       3  Success                                     08-04 19:37:18\n",
      "coding-analysis       2  Success                                     08-04 19:36:28\n",
      "coding-analysis       1  Success                                     08-04 19:40:36\n",
      "coding-analysis       6  Success                                     08-04 19:12:41\n",
      "coding-analysis       5  Success                                     08-04 19:17:07\n",
      "coding-analysis       4  Success                                     08-04 19:13:05\n",
      "coding-analysis       3  Success                                     08-04 19:12:47\n",
      "coding-analysis       2  Success                                     08-04 19:12:15\n",
      "coding-analysis       1  Success                                     08-04 19:17:00\n",
      "coding-analysis       8  Stopped running \"user-command\": exit st...  08-04 17:56:37\n",
      "coding-analysis      48  Stopped running \"user-command\": exit st...  08-04 17:56:50\n",
      "coding-analysis       9  Stopped running \"user-command\": exit st...  08-04 17:44:45\n",
      "coding-analysis      49  Stopped running \"user-command\": exit st...  08-04 17:42:56\n",
      "coding-analysis      60  Success                                     08-04 17:38:59\n",
      "coding-analysis      59  Success                                     08-04 17:39:19\n",
      "coding-analysis      58  Success                                     08-04 17:40:28\n",
      "coding-analysis      57  Success                                     08-04 17:38:23\n",
      "coding-analysis      56  Success                                     08-04 17:39:33\n",
      "coding-analysis      55  Success                                     08-04 17:37:08\n",
      "coding-analysis      54  Success                                     08-04 17:40:35\n",
      "coding-analysis      53  Success                                     08-04 17:37:36\n",
      "coding-analysis      52  Success                                     08-04 17:38:12\n",
      "coding-analysis      51  Success                                     08-04 17:38:55\n",
      "coding-analysis      50  Success                                     08-04 17:42:45\n",
      "coding-analysis      49  Stopped running \"user-command\": exit st...  08-04 17:33:54\n",
      "coding-analysis      48  Stopped running \"user-command\": exit st...  08-04 17:40:07\n",
      "coding-analysis      47  Success                                     08-04 17:43:41\n",
      "coding-analysis      46  Success                                     08-04 17:44:24\n",
      "coding-analysis      45  Success                                     08-04 17:41:01\n",
      "coding-analysis      44  Success                                     08-04 17:42:52\n",
      "coding-analysis      43  Success                                     08-04 17:42:56\n",
      "coding-analysis      42  Success                                     08-04 17:43:06\n",
      "coding-analysis      41  Success                                     08-04 17:40:50\n",
      "coding-analysis      40  Success                                     08-04 17:37:47\n",
      "coding-analysis      39  Success                                     08-04 17:38:58\n",
      "coding-analysis      38  Success                                     08-04 17:40:05\n",
      "coding-analysis      37  Success                                     08-04 17:38:15\n",
      "coding-analysis      36  Success                                     08-04 17:39:08\n",
      "coding-analysis      35  Success                                     08-04 17:37:29\n",
      "coding-analysis      34  Success                                     08-04 17:37:30\n",
      "coding-analysis      33  Success                                     08-04 17:38:17\n",
      "coding-analysis      32  Success                                     08-04 17:38:48\n",
      "coding-analysis      31  Success                                     08-04 17:38:04\n",
      "coding-analysis      30  Success                                     08-04 17:38:04\n",
      "coding-analysis      29  Success                                     08-04 17:39:58\n",
      "coding-analysis      28  Success                                     08-04 17:40:28\n",
      "coding-analysis      27  Success                                     08-04 17:37:58\n",
      "coding-analysis      26  Success                                     08-04 17:39:55\n",
      "coding-analysis      25  Success                                     08-04 17:37:15\n",
      "coding-analysis      24  Success                                     08-04 17:37:43\n",
      "coding-analysis      23  Success                                     08-04 17:38:17\n",
      "coding-analysis      22  Success                                     08-04 17:39:03\n",
      "coding-analysis      21  Success                                     08-04 17:37:36\n",
      "coding-analysis      20  Success                                     08-04 17:38:21\n",
      "coding-analysis      19  Success                                     08-04 17:38:55\n",
      "coding-analysis      18  Success                                     08-04 17:39:16\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis      17  Success                                     08-04 17:38:21\n",
      "coding-analysis      16  Success                                     08-04 17:39:15\n",
      "coding-analysis      15  Success                                     08-04 17:37:58\n",
      "coding-analysis      14  Success                                     08-04 17:37:22\n",
      "coding-analysis      13  Success                                     08-04 17:40:33\n",
      "coding-analysis      12  Success                                     08-04 17:39:30\n",
      "coding-analysis      11  Success                                     08-04 17:38:13\n",
      "coding-analysis      10  Success                                     08-04 17:42:55\n",
      "coding-analysis       9  Stopped running \"user-command\": exit st...  08-04 17:35:43\n",
      "coding-analysis       8  Stopped running \"user-command\": exit st...  08-04 17:40:10\n",
      "coding-analysis       7  Success                                     08-04 17:43:39\n",
      "coding-analysis       6  Success                                     08-04 17:45:30\n",
      "coding-analysis       5  Success                                     08-04 17:41:05\n",
      "coding-analysis       4  Success                                     08-04 17:42:04\n",
      "coding-analysis       3  Success                                     08-04 17:42:03\n",
      "coding-analysis       2  Success                                     08-04 17:43:24\n",
      "coding-analysis       1  Success                                     08-04 17:45:22\n",
      "coding-analysis       7  Stopped running \"user-command\": exit st...  08-04 16:53:55\n",
      "coding-analysis       2  Stopped running \"user-command\": exit st...  08-04 16:46:57\n",
      "coding-analysis       4  Stopped running \"user-command\": exit st...  08-04 16:50:17\n",
      "coding-analysis      47  Stopped running \"user-command\": exit st...  08-04 16:48:54\n",
      "coding-analysis      44  Stopped running \"user-command\": exit st...  08-04 16:49:07\n",
      "coding-analysis      43  Stopped running \"user-command\": exit st...  08-04 16:44:44\n",
      "coding-analysis      48  Stopped running \"user-command\": exit st...  08-04 16:46:07\n",
      "coding-analysis       8  Stopped running \"user-command\": exit st...  08-04 16:45:45\n",
      "coding-analysis      41  Stopped running \"user-command\": exit st...  08-04 16:48:11\n",
      "coding-analysis       1  Stopped running \"user-command\": exit st...  08-04 16:43:22\n",
      "coding-analysis      42  Stopped running \"user-command\": exit st...  08-04 16:42:13\n",
      "coding-analysis      38  Stopped running \"user-command\": exit st...  08-04 16:41:04\n",
      "coding-analysis       3  Stopped running \"user-command\": exit st...  08-04 16:39:40\n",
      "coding-analysis      58  Stopped running \"user-command\": exit st...  08-04 16:43:29\n",
      "coding-analysis      28  Stopped running \"user-command\": exit st...  08-04 16:39:31\n",
      "coding-analysis      18  Stopped running \"user-command\": exit st...  08-04 16:40:16\n",
      "coding-analysis      19  Stopped running \"user-command\": exit st...  08-04 16:33:34\n",
      "coding-analysis      49  Stopped running \"user-command\": exit st...  08-04 16:34:48\n",
      "coding-analysis      46  Stopped running \"user-command\": exit st...  08-04 16:31:53\n",
      "coding-analysis       9  Stopped running \"user-command\": exit st...  08-04 16:32:45\n",
      "coding-analysis      59  Stopped running \"user-command\": exit st...  08-04 16:32:00\n",
      "coding-analysis      39  Stopped running \"user-command\": exit st...  08-04 16:31:12\n",
      "coding-analysis      29  Stopped running \"user-command\": exit st...  08-04 16:31:11\n",
      "coding-analysis       6  Stopped running \"user-command\": exit st...  08-04 16:30:52\n",
      "coding-analysis      60  Success                                     08-04 16:30:03\n",
      "coding-analysis      59  Stopped running \"user-command\": exit st...  08-04 16:23:23\n",
      "coding-analysis      58  Stopped running \"user-command\": exit st...  08-04 16:27:33\n",
      "coding-analysis      57  Success                                     08-04 16:29:54\n",
      "coding-analysis      56  Success                                     08-04 16:30:55\n",
      "coding-analysis      55  Success                                     08-04 16:31:11\n",
      "coding-analysis      54  Success                                     08-04 16:29:06\n",
      "coding-analysis      53  Success                                     08-04 16:32:29\n",
      "coding-analysis      52  Success                                     08-04 16:30:01\n",
      "coding-analysis      51  Success                                     08-04 16:29:07\n",
      "coding-analysis      50  Success                                     08-04 16:34:35\n",
      "coding-analysis      49  Stopped running \"user-command\": exit st...  08-04 16:24:04\n",
      "coding-analysis      48  Stopped running \"user-command\": exit st...  08-04 16:30:52\n",
      "coding-analysis      47  Stopped running \"user-command\": exit st...  08-04 16:32:06\n",
      "coding-analysis      46  Stopped running \"user-command\": exit st...  08-04 16:23:44\n",
      "coding-analysis      45  Success                                     08-04 16:37:10\n",
      "coding-analysis      44  Stopped running \"user-command\": exit st...  08-04 16:32:07\n",
      "coding-analysis      43  Stopped running \"user-command\": exit st...  08-04 16:31:36\n",
      "coding-analysis      42  Stopped running \"user-command\": exit st...  08-04 16:28:42\n",
      "coding-analysis      41  Stopped running \"user-command\": exit st...  08-04 16:29:27\n",
      "coding-analysis      40  Success                                     08-04 16:29:27\n",
      "coding-analysis      39  Stopped running \"user-command\": exit st...  08-04 16:23:02\n",
      "coding-analysis      38  Stopped running \"user-command\": exit st...  08-04 16:27:53\n",
      "coding-analysis      37  Success                                     08-04 16:29:11\n",
      "coding-analysis      36  Success                                     08-04 16:34:40\n",
      "coding-analysis      35  Success                                     08-04 16:31:53\n",
      "coding-analysis      34  Success                                     08-04 16:29:47\n",
      "coding-analysis      33  Success                                     08-04 16:29:08\n",
      "coding-analysis      32  Success                                     08-04 16:29:47\n",
      "coding-analysis      31  Success                                     08-04 16:29:27\n",
      "coding-analysis      30  Success                                     08-04 16:29:31\n",
      "coding-analysis      29  Stopped running \"user-command\": exit st...  08-04 16:23:06\n",
      "coding-analysis      28  Stopped running \"user-command\": exit st...  08-04 16:27:14\n",
      "coding-analysis      27  Success                                     08-04 16:33:08\n",
      "coding-analysis      26  Success                                     08-04 16:30:56\n",
      "coding-analysis      25  Success                                     08-04 16:29:06\n",
      "coding-analysis      24  Success                                     08-04 16:28:48\n",
      "coding-analysis      23  Success                                     08-04 16:31:53\n",
      "coding-analysis      22  Success                                     08-04 16:29:47\n",
      "coding-analysis      21  Success                                     08-04 16:32:45\n",
      "coding-analysis      20  Success                                     08-04 16:29:06\n",
      "coding-analysis      19  Stopped running \"user-command\": exit st...  08-04 16:25:49\n",
      "coding-analysis      18  Stopped running \"user-command\": exit st...  08-04 16:27:09\n",
      "coding-analysis      17  Success                                     08-04 16:28:36\n",
      "coding-analysis      16  Success                                     08-04 16:35:11\n",
      "coding-analysis      15  Success                                     08-04 16:27:57\n",
      "coding-analysis      14  Success                                     08-04 16:28:29\n",
      "coding-analysis      13  Success                                     08-04 16:29:27\n",
      "coding-analysis      12  Success                                     08-04 16:29:27\n",
      "coding-analysis      11  Success                                     08-04 16:33:17\n",
      "coding-analysis      10  Success                                     08-04 16:40:21\n",
      "coding-analysis       9  Stopped running \"user-command\": exit st...  08-04 16:23:44\n",
      "coding-analysis       8  Stopped running \"user-command\": exit st...  08-04 16:30:32\n",
      "coding-analysis       7  Stopped running \"user-command\": exit st...  08-04 16:36:42\n",
      "coding-analysis       6  Stopped running \"user-command\": exit st...  08-04 16:22:40\n",
      "coding-analysis       5  Success                                     08-04 16:31:19\n",
      "coding-analysis       4  Stopped running \"user-command\": exit st...  08-04 16:32:26\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coding-analysis       3  Stopped running \"user-command\": exit st...  08-04 16:27:34\n",
      "coding-analysis       2  Stopped running \"user-command\": exit st...  08-04 16:32:46\n",
      "coding-analysis       1  Stopped running \"user-command\": exit st...  08-04 16:29:10\n",
      "coding-analysis       1  Success                                     08-02 16:49:24\n",
      "coding-analysis       6  Success                                     08-02 16:42:51\n",
      "coding-analysis       5  Success                                     08-02 16:43:17\n",
      "coding-analysis       4  Success                                     08-02 16:43:03\n",
      "coding-analysis       3  Success                                     08-02 16:42:56\n",
      "coding-analysis       2  Success                                     08-02 16:43:02\n",
      "coding-analysis       1  worker was terminated                       08-02 16:42:23\n",
      "coding-analysis       6  Success                                     08-02 16:12:04\n",
      "coding-analysis       5  Success                                     08-02 16:12:02\n",
      "coding-analysis       4  Success                                     08-02 16:11:48\n",
      "coding-analysis       3  Success                                     08-02 16:12:00\n",
      "coding-analysis       2  Success                                     08-02 16:12:07\n",
      "coding-analysis       1  Success                                     08-02 16:12:13\n",
      "coding-analysis       6  Success                                     08-02 15:51:59\n",
      "coding-analysis       5  Success                                     08-02 15:52:09\n",
      "coding-analysis       4  Success                                     08-02 15:51:23\n",
      "coding-analysis       3  Success                                     08-02 15:57:08\n",
      "coding-analysis       2  Success                                     08-02 15:51:59\n",
      "coding-analysis       1  Success                                     08-02 15:51:37\n",
      "coding-analysis       6  Success                                     08-02 15:34:52\n",
      "coding-analysis       5  Success                                     08-02 15:34:42\n",
      "coding-analysis       4  Success                                     08-02 15:35:02\n",
      "coding-analysis       3  Success                                     08-02 15:34:58\n",
      "coding-analysis       2  Success                                     08-02 15:35:10\n",
      "coding-analysis       1  Success                                     08-02 15:34:10\n",
      "coding-analysis       6  Success                                     08-02 15:18:22\n",
      "coding-analysis       5  Success                                     08-02 15:18:02\n",
      "coding-analysis       4  Success                                     08-02 15:18:23\n",
      "coding-analysis       3  Success                                     08-02 15:18:47\n",
      "coding-analysis       2  Success                                     08-02 15:18:02\n",
      "coding-analysis       1  Success                                     08-02 15:26:51\n",
      "noncoding-an...       6  Success                                     08-02 14:28:26\n",
      "noncoding-an...       5  Success                                     08-02 14:35:16\n",
      "noncoding-an...       4  Success                                     08-02 14:28:42\n",
      "noncoding-an...       3  Success                                     08-02 14:29:28\n",
      "noncoding-an...       2  Success                                     08-02 14:28:26\n",
      "noncoding-an...       1  Success                                     08-02 14:37:06\n",
      "ncrna-analysis        6  Success                                     08-02 14:27:00\n",
      "ncrna-analysis        5  Success                                     08-02 14:24:56\n",
      "ncrna-analysis        4  Success                                     08-02 14:24:33\n",
      "ncrna-analysis        3  Success                                     08-02 14:24:57\n",
      "ncrna-analysis        2  Success                                     08-02 14:24:48\n",
      "ncrna-analysis        1  Success                                     08-02 14:25:09\n",
      "noncoding-an...       6  Success                                     08-02 13:36:20\n",
      "noncoding-an...       5  Success                                     08-02 13:37:31\n",
      "noncoding-an...       4  Success                                     08-02 13:36:44\n",
      "noncoding-an...       3  Success                                     08-02 13:36:46\n",
      "noncoding-an...       2  Success                                     08-02 13:38:21\n",
      "noncoding-an...       1  Success                                     08-02 13:38:01\n",
      "noncoding-an...       6  Stopped running \"user-command\": exit st...  08-02 13:25:55\n",
      "noncoding-an...       5  Stopped running \"user-command\": exit st...  08-02 13:26:43\n",
      "noncoding-an...       4  Stopped running \"user-command\": exit st...  08-02 13:26:18\n",
      "noncoding-an...       3  Stopped running \"user-command\": exit st...  08-02 13:25:29\n",
      "noncoding-an...       2  Stopped running \"user-command\": exit st...  08-02 13:26:21\n",
      "noncoding-an...       1  Stopped running \"user-command\": exit st...  08-02 13:25:37\n",
      "ncrna-analysis        6  Success                                     08-02 13:24:10\n",
      "ncrna-analysis        5  Success                                     08-02 13:26:27\n",
      "ncrna-analysis        4  Success                                     08-02 13:21:56\n",
      "ncrna-analysis        3  Success                                     08-02 13:22:09\n",
      "ncrna-analysis        2  Success                                     08-02 13:22:49\n",
      "ncrna-analysis        1  Success                                     08-02 13:22:03\n",
      "ncrna-analysis        6  Stopped running \"user-command\": exit st...  08-02 13:02:37\n",
      "ncrna-analysis        5  Stopped running \"user-command\": exit st...  08-02 13:03:20\n",
      "ncrna-analysis        4  Stopped running \"user-command\": exit st...  08-02 13:02:41\n",
      "ncrna-analysis        3  Stopped running \"user-command\": exit st...  08-02 13:03:43\n",
      "ncrna-analysis        2  Stopped running \"user-command\": exit st...  08-02 13:03:08\n",
      "ncrna-analysis        1  Stopped running \"user-command\": exit st...  08-02 13:03:32\n",
      "nullmodel             1  Success                                     08-01 21:52:06\n",
      "nullmodel             1  Stopped running \"user-command\": exit st...  08-01 13:14:11\n",
      "nullmodel             1  Stopped running \"user-command\": exit st...  08-01 12:53:22\n",
      "onecommonprs          6  Success                                     08-01 12:16:58\n",
      "onecommonprs          5  Success                                     08-01 06:01:06\n",
      "onecommonprs          4  Success                                     08-01 06:50:37\n",
      "onecommonprs          3  Success                                     08-01 07:02:13\n",
      "onecommonprs          2  Success                                     08-01 05:21:51\n",
      "onecommonprs          1  Success                                     08-01 11:45:18\n",
      "joint-prs-re...       6  Success                                     07-31 19:21:26\n",
      "joint-prs-re...       5  Success                                     07-31 19:21:18\n",
      "joint-prs-re...       4  Success                                     07-31 19:21:59\n",
      "joint-prs-re...       3  Success                                     07-31 19:21:17\n",
      "joint-prs-re...       2  Success                                     07-31 19:21:26\n",
      "joint-prs-re...       1  Success                                     07-31 19:21:46\n",
      "jointprs-score        1  Success                                     07-31 17:53:41\n",
      "onecommonprs          6  Stopped running \"logging\": exit status ...  07-30 02:48:25\n",
      "onecommonprs          5  Success                                     07-29 19:26:20\n",
      "onecommonprs          4  Stopped running \"logging\": exit status ...  07-29 22:02:35\n",
      "onecommonprs          3  Success                                     07-29 19:42:07\n",
      "onecommonprs          2  Success                                     07-29 20:11:36\n",
      "onecommonprs          1  Stopped running \"logging\": exit status ...  07-30 06:31:11\n",
      "onecommonprs          6  Stopped running \"user-command\": exit st...  07-29 10:21:15\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-29 04:00:36\n",
      "onecommonprs          4  Stopped running \"user-command\": exit st...  07-29 06:12:02\n",
      "onecommonprs          3  Stopped running \"user-command\": exit st...  07-29 04:02:07\n",
      "onecommonprs          2  Stopped running \"user-command\": exit st...  07-29 03:37:06\n",
      "onecommonprs          1  The operation was cancelled                 07-29 12:07:32\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "onecommonprs          6  The operation was cancelled                 07-28 20:15:30\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-28 19:35:23\n",
      "onecommonprs          4  The operation was cancelled                 07-28 20:15:30\n",
      "onecommonprs          3  The operation was cancelled                 07-28 20:15:30\n",
      "onecommonprs          2  The operation was cancelled                 07-28 20:15:30\n",
      "onecommonprs          1  The operation was cancelled                 07-28 20:15:31\n",
      "onecommonprs          6  Stopped running \"user-command\": exit st...  07-28 02:50:10\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-27 19:25:31\n",
      "onecommonprs          4  Stopped running \"user-command\": exit st...  07-27 22:22:23\n",
      "onecommonprs          3  Stopped running \"user-command\": exit st...  07-27 19:08:08\n",
      "onecommonprs          2  Stopped running \"user-command\": exit st...  07-27 19:20:38\n",
      "onecommonprs          1  Stopped running \"user-command\": exit st...  07-28 04:57:50\n",
      "onecommonprs          6  The operation was cancelled                 07-27 11:49:07\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-27 04:56:48\n",
      "onecommonprs          4  Stopped running \"user-command\": exit st...  07-27 07:37:46\n",
      "onecommonprs          3  Stopped running \"user-command\": exit st...  07-27 05:31:04\n",
      "onecommonprs          2  Stopped running \"user-command\": exit st...  07-27 04:51:30\n",
      "onecommonprs          1  The operation was cancelled                 07-27 11:49:07\n",
      "onecommonprs          6  Stopped running \"user-command\": exit st...  07-26 21:13:10\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-26 21:13:05\n",
      "onecommonprs          4  Stopped running \"user-command\": exit st...  07-26 21:12:47\n",
      "onecommonprs          3  Stopped running \"user-command\": exit st...  07-26 21:13:23\n",
      "onecommonprs          2  Stopped running \"user-command\": exit st...  07-26 21:12:41\n",
      "onecommonprs          1  Stopped running \"user-command\": exit st...  07-26 21:12:10\n",
      "onecommonprs          6  Stopped running \"user-command\": exit st...  07-26 20:54:45\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-26 20:54:35\n",
      "onecommonprs          4  Stopped running \"user-command\": exit st...  07-26 20:54:26\n",
      "onecommonprs          3  Stopped running \"user-command\": exit st...  07-26 20:54:36\n",
      "onecommonprs          2  Stopped running \"user-command\": exit st...  07-26 20:54:25\n",
      "onecommonprs          1  Stopped running \"user-command\": exit st...  07-26 20:54:08\n",
      "onecommonprs          6  Stopped running \"user-command\": exit st...  07-26 20:41:50\n",
      "onecommonprs          5  Stopped running \"user-command\": exit st...  07-26 20:41:55\n",
      "onecommonprs          4  Stopped running \"user-command\": exit st...  07-26 20:41:43\n",
      "onecommonprs          3  Stopped running \"user-command\": exit st...  07-26 20:41:51\n",
      "onecommonprs          2  Stopped running \"user-command\": exit st...  07-26 20:41:49\n",
      "onecommonprs          1  Stopped running \"user-command\": exit st...  07-26 20:41:38\n",
      "prosper-results       6  Success                                     07-26 16:48:41\n",
      "prosper-results       5  Success                                     07-26 16:49:06\n",
      "prosper-results       4  Success                                     07-26 16:48:44\n",
      "prosper-results       3  Success                                     07-26 16:48:52\n",
      "prosper-results       2  Success                                     07-26 16:48:59\n",
      "prosper-results       1  Success                                     07-26 16:49:10\n",
      "joint-prs-re...       6  Success                                     07-26 16:17:57\n",
      "joint-prs-re...       5  Success                                     07-26 16:17:54\n",
      "joint-prs-re...       4  Success                                     07-26 16:17:41\n",
      "joint-prs-re...       3  Success                                     07-26 16:18:09\n",
      "joint-prs-re...       2  Success                                     07-26 16:17:56\n",
      "joint-prs-re...       1  Success                                     07-26 16:17:52\n",
      "joint-prs-re...       6  Stopped running \"user-command\": exit st...  07-26 15:56:58\n",
      "joint-prs-re...       5  Stopped running \"user-command\": exit st...  07-26 15:56:48\n",
      "joint-prs-re...       4  Stopped running \"user-command\": exit st...  07-26 15:56:56\n",
      "joint-prs-re...       3  Stopped running \"user-command\": exit st...  07-26 15:57:35\n",
      "joint-prs-re...       2  Stopped running \"user-command\": exit st...  07-26 15:57:07\n",
      "joint-prs-re...       1  Stopped running \"user-command\": exit st...  07-26 15:57:23\n",
      "ct-sleb-sl            6  Success                                     07-27 06:07:20\n",
      "ct-sleb-sl            5  Success                                     07-26 23:49:08\n",
      "ct-sleb-sl            4  Success                                     07-27 00:39:37\n",
      "ct-sleb-sl            3  Success                                     07-26 23:18:53\n",
      "ct-sleb-sl            2  Success                                     07-26 23:10:27\n",
      "ct-sleb-sl            1  Success                                     07-27 06:34:10\n",
      "prosper-scor...       6  Success                                     07-26 16:42:28\n",
      "prosper-scor...       5  Success                                     07-26 16:30:33\n",
      "prosper-scor...       4  Success                                     07-26 16:33:17\n",
      "prosper-scor...       3  Success                                     07-26 16:33:39\n",
      "prosper-scor...       2  Success                                     07-26 16:32:08\n",
      "prosper-scor...       1  Success                                     07-26 16:51:45\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-26 13:40:21\n",
      "ct-sleb-sl            5  The operation was cancelled                 07-26 13:54:35\n",
      "ct-sleb-sl            4  The operation was cancelled                 07-26 13:54:35\n",
      "ct-sleb-sl            3  The operation was cancelled                 07-26 13:54:35\n",
      "ct-sleb-sl            2  The operation was cancelled                 07-26 13:54:35\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-26 13:50:08\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-26 12:34:10\n",
      "ct-sleb-sl            5  The operation was cancelled                 07-26 12:38:48\n",
      "ct-sleb-sl            4  The operation was cancelled                 07-26 12:38:48\n",
      "ct-sleb-sl            3  The operation was cancelled                 07-26 12:38:48\n",
      "ct-sleb-sl            2  The operation was cancelled                 07-26 12:38:48\n",
      "ct-sleb-sl            1  The operation was cancelled                 07-26 12:38:48\n",
      "prosper-scor...       6  Stopped running \"user-command\": exit st...  07-26 13:09:22\n",
      "prosper-scor...       5  Stopped running \"user-command\": exit st...  07-26 13:08:42\n",
      "prosper-scor...       4  Stopped running \"user-command\": exit st...  07-26 13:08:35\n",
      "prosper-scor...       3  Stopped running \"user-command\": exit st...  07-26 13:08:15\n",
      "prosper-scor...       2  Stopped running \"user-command\": exit st...  07-26 13:12:13\n",
      "prosper-scor...       1  Stopped running \"user-command\": exit st...  07-26 13:08:25\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-26 12:19:00\n",
      "ct-sleb-sl            5  The operation was cancelled                 07-26 12:27:46\n",
      "ct-sleb-sl            4  The operation was cancelled                 07-26 12:27:46\n",
      "ct-sleb-sl            3  The operation was cancelled                 07-26 12:27:46\n",
      "ct-sleb-sl            2  The operation was cancelled                 07-26 12:27:46\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-26 12:14:36\n",
      "prosper-score         1  Success                                     07-26 05:22:25\n",
      "jointprs-score        1  Success                                     07-26 04:06:29\n",
      "jointprs-score        1  Stopped running \"user-command\": exit st...  07-25 21:58:00\n",
      "prosper-sl            6  Success                                     07-25 22:24:08\n",
      "prosper-sl            5  Success                                     07-25 22:26:14\n",
      "prosper-sl            4  Success                                     07-25 22:23:25\n",
      "prosper-sl            3  Success                                     07-25 22:20:23\n",
      "prosper-sl            2  Success                                     07-25 22:25:03\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prosper-sl            1  Success                                     07-25 22:21:18\n",
      "jointprs-score        1  Stopped running \"user-command\": exit st...  07-25 21:07:50\n",
      "prosper-score         1  Stopped running \"user-command\": exit st...  07-25 17:43:04\n",
      "prosper-score         1  Stopped running \"user-command\": exit st...  07-25 01:33:54\n",
      "prosper-score         1  Stopped running \"localization\": exit st...  07-24 19:25:13\n",
      "prosper-score         1  Stopped running \"localization\": exit st...  07-24 19:16:17\n",
      "prosper-score         1  The operation was cancelled                 07-24 19:12:07\n",
      "jointprs-score        1  Stopped running \"user-command\": exit st...  07-24 19:51:55\n",
      "jointprs-score        1  The operation was cancelled                 07-24 19:06:17\n",
      "jointprs-score        1  The operation was cancelled                 07-24 18:59:17\n",
      "prosper-sl            6  Success                                     07-24 16:45:59\n",
      "prosper-sl            5  Success                                     07-24 16:47:00\n",
      "prosper-sl            4  Success                                     07-24 16:47:04\n",
      "prosper-sl            3  Success                                     07-24 16:45:24\n",
      "prosper-sl            2  Success                                     07-24 16:46:54\n",
      "prosper-sl            1  Success                                     07-24 16:46:59\n",
      "prosper-sl            6  Stopped running \"user-command\": exit st...  07-24 14:36:01\n",
      "prosper-sl            5  Stopped running \"user-command\": exit st...  07-24 14:36:53\n",
      "prosper-sl            4  Stopped running \"user-command\": exit st...  07-24 14:37:01\n",
      "prosper-sl            3  Stopped running \"user-command\": exit st...  07-24 14:36:08\n",
      "prosper-sl            2  Stopped running \"user-command\": exit st...  07-24 14:35:59\n",
      "prosper-sl            1  Stopped running \"user-command\": exit st...  07-24 14:36:19\n",
      "prosper-prosper       6  Stopped running \"localization\": exit st...  07-24 13:37:08\n",
      "prosper-prosper       5  Stopped running \"localization\": exit st...  07-24 13:37:16\n",
      "prosper-prosper       4  Stopped running \"localization\": exit st...  07-24 13:36:42\n",
      "prosper-prosper       3  Stopped running \"localization\": exit st...  07-24 13:36:56\n",
      "prosper-prosper       2  Stopped running \"localization\": exit st...  07-24 13:36:47\n",
      "prosper-prosper       1  Stopped running \"localization\": exit st...  07-24 13:37:51\n",
      "prosper-prosper       6  Success                                     07-23 17:07:20\n",
      "prosper-prosper       5  Success                                     07-23 17:09:53\n",
      "prosper-prosper       4  Success                                     07-23 18:32:59\n",
      "prosper-prosper       3  Success                                     07-23 18:03:57\n",
      "prosper-prosper       2  Success                                     07-23 17:43:58\n",
      "prosper-prosper       1  Success                                     07-23 17:04:05\n",
      "mussel-ldpred2        6  The operation was cancelled                 07-25 20:33:22\n",
      "mussel-ldpred2        5  The operation was cancelled                 07-25 20:33:22\n",
      "mussel-ldpred2        4  The operation was cancelled                 07-25 20:33:22\n",
      "mussel-ldpred2        3  The operation was cancelled                 07-25 20:33:22\n",
      "mussel-ldpred2        2  The operation was cancelled                 07-25 20:33:22\n",
      "mussel-ldpred2        1  The operation was cancelled                 07-25 20:33:22\n",
      "prosper-lass...       6  Success                                     07-22 20:09:39\n",
      "prosper-lass...       5  Success                                     07-22 20:06:39\n",
      "prosper-lass...       4  Success                                     07-22 20:08:25\n",
      "prosper-lass...       3  Success                                     07-22 20:06:17\n",
      "prosper-lass...       2  Success                                     07-22 20:06:57\n",
      "prosper-lass...       1  Success                                     07-22 20:06:37\n",
      "tune-validat...       1  Success                                     07-22 18:00:29\n",
      "mussel-ldpred2        6  The operation was cancelled                 07-22 17:58:47\n",
      "mussel-ldpred2        5  The operation was cancelled                 07-22 17:58:47\n",
      "mussel-ldpred2        4  The operation was cancelled                 07-22 17:58:47\n",
      "mussel-ldpred2        3  The operation was cancelled                 07-22 17:58:47\n",
      "mussel-ldpred2        2  The operation was cancelled                 07-22 17:58:47\n",
      "mussel-ldpred2        1  The operation was cancelled                 07-22 17:58:47\n",
      "mussel-ldpred2        6  Stopped running \"localization\": exit st...  07-22 16:24:24\n",
      "mussel-ldpred2        5  Stopped running \"localization\": exit st...  07-22 16:24:22\n",
      "mussel-ldpred2        4  Stopped running \"localization\": exit st...  07-22 16:24:09\n",
      "mussel-ldpred2        3  Stopped running \"localization\": exit st...  07-22 16:24:15\n",
      "mussel-ldpred2        2  Stopped running \"localization\": exit st...  07-22 16:24:13\n",
      "mussel-ldpred2        1  Stopped running \"localization\": exit st...  07-22 16:24:31\n",
      "prosper-lass...       6  Stopped running \"user-command\": exit st...  07-22 13:37:59\n",
      "prosper-lass...       5  Stopped running \"user-command\": exit st...  07-22 13:36:54\n",
      "prosper-lass...       4  Stopped running \"user-command\": exit st...  07-22 13:39:09\n",
      "prosper-lass...       3  Stopped running \"user-command\": exit st...  07-22 13:40:15\n",
      "prosper-lass...       2  Stopped running \"user-command\": exit st...  07-22 13:38:41\n",
      "prosper-lass...       1  Stopped running \"user-command\": exit st...  07-22 13:39:36\n",
      "ct-sleb-dimc...      12  Success                                     07-25 10:41:16\n",
      "ct-sleb-dimc...      11  Success                                     07-25 03:44:58\n",
      "ct-sleb-dimc...      10  Success                                     07-24 19:24:01\n",
      "ct-sleb-dimc...       9  Success                                     07-25 06:55:21\n",
      "ct-sleb-dimc...       8  Success                                     07-25 04:10:06\n",
      "ct-sleb-dimc...       7  Success                                     07-25 17:03:01\n",
      "ct-sleb-dimc...       6  Success                                     07-26 03:59:09\n",
      "ct-sleb-dimc...       5  Success                                     07-25 09:18:17\n",
      "ct-sleb-dimc...       4  Success                                     07-25 08:11:38\n",
      "ct-sleb-dimc...       3  Success                                     07-25 12:02:03\n",
      "ct-sleb-dimc...       2  Success                                     07-25 14:00:39\n",
      "ct-sleb-dimc...       1  Success                                     07-25 08:06:10\n",
      "mussel-sumst...      18  Success                                     07-22 12:31:31\n",
      "mussel-sumst...      17  Success                                     07-22 12:31:30\n",
      "mussel-sumst...      16  Success                                     07-22 12:31:30\n",
      "mussel-sumst...      15  Success                                     07-22 12:31:25\n",
      "mussel-sumst...      14  Success                                     07-22 12:31:07\n",
      "mussel-sumst...      13  Success                                     07-22 12:31:27\n",
      "mussel-sumst...      12  Success                                     07-22 12:32:24\n",
      "mussel-sumst...      11  Success                                     07-22 12:31:23\n",
      "mussel-sumst...      10  Success                                     07-22 12:31:41\n",
      "mussel-sumst...       9  Success                                     07-22 12:31:56\n",
      "mussel-sumst...       8  Success                                     07-22 12:31:22\n",
      "mussel-sumst...       7  Success                                     07-22 12:31:38\n",
      "mussel-sumst...       6  Success                                     07-22 12:31:45\n",
      "mussel-sumst...       5  Success                                     07-22 12:32:08\n",
      "mussel-sumst...       4  Success                                     07-22 12:31:38\n",
      "mussel-sumst...       3  Success                                     07-22 12:31:25\n",
      "mussel-sumst...       2  Success                                     07-22 12:31:40\n",
      "mussel-sumst...       1  Success                                     07-22 12:32:00\n",
      "mussel-sumst...      18  Stopped running \"localization\": exit st...  07-21 12:47:17\n",
      "mussel-sumst...      17  Stopped running \"localization\": exit st...  07-21 12:46:53\n",
      "mussel-sumst...      16  Stopped running \"localization\": exit st...  07-21 12:46:58\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mussel-sumst...      15  Stopped running \"localization\": exit st...  07-21 12:46:55\n",
      "mussel-sumst...      14  Stopped running \"localization\": exit st...  07-21 12:46:54\n",
      "mussel-sumst...      13  Stopped running \"localization\": exit st...  07-21 12:47:13\n",
      "mussel-sumst...      12  Stopped running \"localization\": exit st...  07-21 12:46:59\n",
      "mussel-sumst...      11  Stopped running \"localization\": exit st...  07-21 12:47:00\n",
      "mussel-sumst...      10  Stopped running \"localization\": exit st...  07-21 12:46:56\n",
      "mussel-sumst...       9  Stopped running \"localization\": exit st...  07-21 12:46:54\n",
      "mussel-sumst...       8  Stopped running \"localization\": exit st...  07-21 12:46:57\n",
      "mussel-sumst...       7  Stopped running \"localization\": exit st...  07-21 12:47:03\n",
      "mussel-sumst...       6  Stopped running \"localization\": exit st...  07-21 12:47:07\n",
      "mussel-sumst...       5  Stopped running \"localization\": exit st...  07-21 12:47:19\n",
      "mussel-sumst...       4  Stopped running \"localization\": exit st...  07-21 12:47:04\n",
      "mussel-sumst...       3  Stopped running \"localization\": exit st...  07-21 12:47:09\n",
      "mussel-sumst...       2  Stopped running \"localization\": exit st...  07-21 12:47:19\n",
      "mussel-sumst...       1  Stopped running \"localization\": exit st...  07-21 12:47:47\n",
      "prosper-sums...      18  Success                                     07-21 12:47:43\n",
      "prosper-sums...      17  Success                                     07-21 12:47:31\n",
      "prosper-sums...      16  Success                                     07-21 12:47:04\n",
      "prosper-sums...      15  Success                                     07-21 12:47:20\n",
      "prosper-sums...      14  Success                                     07-21 12:47:16\n",
      "prosper-sums...      13  Success                                     07-21 12:47:00\n",
      "prosper-sums...      12  Success                                     07-21 12:47:06\n",
      "prosper-sums...      11  Success                                     07-21 12:47:13\n",
      "prosper-sums...      10  Success                                     07-21 12:47:14\n",
      "prosper-sums...       9  Success                                     07-21 12:47:20\n",
      "prosper-sums...       8  Success                                     07-21 12:47:11\n",
      "prosper-sums...       7  Success                                     07-21 12:47:05\n",
      "prosper-sums...       6  Success                                     07-21 12:47:42\n",
      "prosper-sums...       5  Success                                     07-21 12:47:36\n",
      "prosper-sums...       4  Success                                     07-21 12:47:09\n",
      "prosper-sums...       3  Success                                     07-21 12:47:45\n",
      "prosper-sums...       2  Success                                     07-21 12:47:11\n",
      "prosper-sums...       1  Success                                     07-21 12:47:19\n",
      "jointprs-mus...       1  Success                                     07-21 04:06:35\n",
      "jointprs-mus...       1  User specified operation timeout reached    07-20 12:01:25\n",
      "jointprs-mus...       1  Stopped running \"user-command\": exit st...  07-16 01:50:44\n",
      "jointprs-mus...       1  Stopped running \"user-command\": exit st...  07-15 15:23:15\n",
      "jointprs-mus...       1  Success                                     07-14 15:37:45\n",
      "jointprs-mus...       1  Success                                     07-14 04:43:10\n",
      "jointprs-mus...       1  Stopped running \"user-command\": exit st...  07-13 15:29:31\n",
      "jointprs-mus...       1  Stopped running \"user-command\": exit st...  07-13 04:14:53\n",
      "jointprs-mus...       1  The assigned worker has failed to compl...  07-12 22:15:33\n",
      "jointprs-mus...       1  Stopped running \"user-command\": exit st...  07-12 17:46:54\n",
      "jointprs-mus...       3  The operation was cancelled                 07-12 13:35:53\n",
      "jointprs-mus...       2  The operation was cancelled                 07-12 13:35:53\n",
      "jointprs-mus...       1  The operation was cancelled                 07-12 13:35:53\n",
      "mussel-ldpred2        6  Stopped running \"user-command\": exit st...  07-10 00:33:19\n",
      "mussel-ldpred2        5  Stopped running \"user-command\": exit st...  07-10 00:30:49\n",
      "mussel-ldpred2        4  Stopped running \"user-command\": exit st...  07-10 00:30:23\n",
      "mussel-ldpred2        3  Stopped running \"user-command\": exit st...  07-10 00:30:57\n",
      "mussel-ldpred2        2  Stopped running \"user-command\": exit st...  07-10 00:30:25\n",
      "mussel-ldpred2        1  Stopped running \"user-command\": exit st...  07-10 00:29:08\n",
      "ct-sleb-sl            6  Success                                     07-10 17:03:43\n",
      "ct-sleb-sl            5  Success                                     07-10 09:40:49\n",
      "ct-sleb-sl            4  Success                                     07-10 10:16:41\n",
      "ct-sleb-sl            3  Success                                     07-10 10:29:19\n",
      "ct-sleb-sl            2  Success                                     07-10 11:44:41\n",
      "ct-sleb-sl            1  Success                                     07-10 15:34:11\n",
      "mussel-ldpred2        6  Stopped running \"user-command\": exit st...  07-09 22:01:41\n",
      "mussel-ldpred2        5  Stopped running \"user-command\": exit st...  07-09 22:03:39\n",
      "mussel-ldpred2        4  Stopped running \"user-command\": exit st...  07-09 22:02:13\n",
      "mussel-ldpred2        3  Stopped running \"user-command\": exit st...  07-09 22:02:02\n",
      "mussel-ldpred2        2  Stopped running \"user-command\": exit st...  07-09 22:02:01\n",
      "mussel-ldpred2        1  Stopped running \"user-command\": exit st...  07-09 22:02:19\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-10 05:38:04\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-10 01:18:34\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-09 23:21:34\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-10 00:24:30\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-09 23:30:21\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-10 05:21:05\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-09 15:36:09\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-09 06:41:07\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-09 08:59:33\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-09 07:53:49\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-09 08:13:06\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-09 14:32:22\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-09 03:53:00\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-08 23:03:58\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-08 23:18:03\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-08 22:56:27\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-08 21:12:48\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-09 04:17:18\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-08 08:00:33\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-07 23:48:39\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-08 01:43:36\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-08 02:01:33\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-07 22:55:55\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-08 07:04:29\n",
      "prosper-lass...       6  The operation was cancelled                 07-10 21:02:39\n",
      "prosper-lass...       5  The operation was cancelled                 07-10 21:02:39\n",
      "prosper-lass...       4  The operation was cancelled                 07-10 21:02:39\n",
      "prosper-lass...       3  The operation was cancelled                 07-10 21:02:39\n",
      "prosper-lass...       2  The operation was cancelled                 07-10 21:02:39\n",
      "prosper-lass...       1  The operation was cancelled                 07-10 21:02:39\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-07 04:46:23\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-07 00:08:20\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-06 23:13:05\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-06 23:56:30\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-06 21:43:35\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-07 06:16:08\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-06 10:01:49\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-06 03:04:01\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-06 04:26:19\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-06 04:35:13\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-06 02:10:33\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-06 09:58:41\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-05 17:53:53\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-05 17:49:13\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-05 17:50:09\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-05 17:49:21\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-05 17:49:14\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-05 17:53:19\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-05 17:38:35\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-05 17:36:55\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-05 17:36:33\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-05 17:35:19\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-05 17:35:24\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-05 17:39:01\n",
      "ct-sleb-sl            6  Stopped running \"user-command\": exit st...  07-05 17:14:37\n",
      "ct-sleb-sl            5  Stopped running \"user-command\": exit st...  07-05 17:14:34\n",
      "ct-sleb-sl            4  Stopped running \"user-command\": exit st...  07-05 17:14:26\n",
      "ct-sleb-sl            3  Stopped running \"user-command\": exit st...  07-05 17:14:56\n",
      "ct-sleb-sl            2  Stopped running \"user-command\": exit st...  07-05 17:14:21\n",
      "ct-sleb-sl            1  Stopped running \"user-command\": exit st...  07-05 17:14:27\n",
      "ct-sleb-sl            6  Stopped running \"localization\": exit st...  07-05 16:23:27\n",
      "ct-sleb-sl            5  Stopped running \"localization\": exit st...  07-05 16:23:43\n",
      "ct-sleb-sl            4  Stopped running \"localization\": exit st...  07-05 16:23:45\n",
      "ct-sleb-sl            3  Stopped running \"localization\": exit st...  07-05 16:23:57\n",
      "ct-sleb-sl            2  Stopped running \"localization\": exit st...  07-05 16:23:44\n",
      "ct-sleb-sl            1  Stopped running \"localization\": exit st...  07-05 16:24:30\n",
      "prosper-lass...       6  User specified operation timeout reached    07-06 13:12:53\n",
      "prosper-lass...       5  User specified operation timeout reached    07-06 13:12:54\n",
      "prosper-lass...       4  User specified operation timeout reached    07-06 13:12:54\n",
      "prosper-lass...       3  User specified operation timeout reached    07-06 13:12:53\n",
      "prosper-lass...       2  User specified operation timeout reached    07-06 13:12:53\n",
      "prosper-lass...       1  User specified operation timeout reached    07-06 13:12:53\n",
      "mussel-ldpred2        6  Stopped running \"user-command\": exit st...  07-02 13:37:00\n",
      "mussel-ldpred2        5  Stopped running \"user-command\": exit st...  07-02 13:41:11\n",
      "mussel-ldpred2        4  Stopped running \"user-command\": exit st...  07-02 13:37:16\n",
      "mussel-ldpred2        3  Stopped running \"user-command\": exit st...  07-02 13:34:33\n",
      "mussel-ldpred2        2  Stopped running \"user-command\": exit st...  07-02 13:35:04\n",
      "mussel-ldpred2        1  Stopped running \"user-command\": exit st...  07-02 13:36:43\n",
      "prosper-lass...       6  The operation was cancelled                 07-02 14:54:18\n",
      "prosper-lass...       5  The operation was cancelled                 07-02 14:54:18\n",
      "prosper-lass...       4  The operation was cancelled                 07-02 14:54:18\n",
      "prosper-lass...       3  The operation was cancelled                 07-02 14:54:18\n",
      "prosper-lass...       2  The operation was cancelled                 07-02 14:54:18\n",
      "prosper-lass...       1  The operation was cancelled                 07-02 14:54:18\n",
      "ct-sleb-dimc...      12  Success                                     07-04 19:34:04\n",
      "ct-sleb-dimc...      11  Success                                     07-03 21:55:08\n",
      "ct-sleb-dimc...      10  Success                                     07-03 23:30:58\n",
      "ct-sleb-dimc...       9  Success                                     07-04 08:36:45\n",
      "ct-sleb-dimc...       8  Success                                     07-03 22:39:49\n",
      "ct-sleb-dimc...       7  Success                                     07-04 22:13:04\n",
      "ct-sleb-dimc...       6  Success                                     07-04 17:10:06\n",
      "ct-sleb-dimc...       5  Success                                     07-04 14:13:42\n",
      "ct-sleb-dimc...       4  Success                                     07-04 04:28:31\n",
      "ct-sleb-dimc...       3  Success                                     07-04 20:13:14\n",
      "ct-sleb-dimc...       2  Success                                     07-04 06:51:57\n",
      "ct-sleb-dimc...       1  Success                                     07-04 19:42:13\n",
      "prosper-lass...       6  Success                                     07-01 16:34:21\n",
      "prosper-lass...       5  Success                                     07-01 16:44:33\n",
      "prosper-lass...       4  Success                                     07-01 16:36:24\n",
      "prosper-lass...       3  Success                                     07-01 16:40:48\n",
      "prosper-lass...       2  Success                                     07-01 16:33:36\n",
      "prosper-lass...       1  Success                                     07-01 16:33:14\n",
      "prosper-lass...       6  Success                                     07-01 15:36:17\n",
      "prosper-lass...       5  Success                                     07-01 15:31:23\n",
      "prosper-lass...       4  Success                                     07-01 15:41:50\n",
      "prosper-lass...       3  Success                                     07-01 15:31:13\n",
      "prosper-lass...       2  Success                                     07-01 15:32:25\n",
      "prosper-lass...       1  Success                                     07-01 15:35:01\n",
      "ct-sleb-dimct        12  Stopped running \"user-command\": exit st...  07-01 16:11:09\n",
      "ct-sleb-dimct        11  Stopped running \"user-command\": exit st...  07-01 15:59:57\n",
      "ct-sleb-dimct        10  Stopped running \"user-command\": exit st...  07-01 16:15:31\n",
      "ct-sleb-dimct         9  Stopped running \"user-command\": exit st...  07-01 15:59:45\n",
      "ct-sleb-dimct         8  Stopped running \"user-command\": exit st...  07-01 16:27:28\n",
      "ct-sleb-dimct         7  Stopped running \"user-command\": exit st...  07-01 16:18:10\n",
      "ct-sleb-dimct         6  Stopped running \"user-command\": exit st...  07-01 16:06:01\n",
      "ct-sleb-dimct         5  Stopped running \"user-command\": exit st...  07-01 16:02:28\n",
      "ct-sleb-dimct         4  Stopped running \"user-command\": exit st...  07-01 16:26:30\n",
      "ct-sleb-dimct         3  Stopped running \"user-command\": exit st...  07-01 16:10:29\n",
      "ct-sleb-dimct         2  Stopped running \"user-command\": exit st...  07-01 16:26:02\n",
      "ct-sleb-dimct         1  Stopped running \"user-command\": exit st...  07-01 16:22:57\n",
      "prosper-lass...       6  Stopped running \"user-command\": exit st...  06-29 23:46:31\n",
      "prosper-lass...       5  Stopped running \"user-command\": exit st...  06-29 23:45:29\n",
      "prosper-lass...       4  Stopped running \"user-command\": exit st...  06-29 23:50:06\n",
      "prosper-lass...       3  Stopped running \"user-command\": exit st...  06-29 23:46:03\n",
      "prosper-lass...       2  Stopped running \"user-command\": exit st...  06-29 23:50:07\n",
      "prosper-lass...       1  Stopped running \"user-command\": exit st...  06-29 23:47:02\n",
      "prosper-lass...       6  Stopped running \"user-command\": exit st...  06-29 02:05:48\n",
      "prosper-lass...       5  Stopped running \"user-command\": exit st...  06-29 02:07:41\n",
      "prosper-lass...       4  Stopped running \"user-command\": exit st...  06-29 02:03:54\n",
      "prosper-lass...       3  Stopped running \"user-command\": exit st...  06-29 02:03:02\n",
      "prosper-lass...       2  Stopped running \"user-command\": exit st...  06-29 02:07:26\n",
      "prosper-lass...       1  Stopped running \"user-command\": exit st...  06-29 02:03:03\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tune-validat...       1  Success                                     06-29 12:49:20\n",
      "prosper-lass...       6  Stopped running \"localization\": exit st...  06-29 01:18:39\n",
      "prosper-lass...       5  Stopped running \"localization\": exit st...  06-29 01:22:42\n",
      "prosper-lass...       4  Stopped running \"localization\": exit st...  06-29 01:34:45\n",
      "prosper-lass...       3  Stopped running \"localization\": exit st...  06-29 01:25:30\n",
      "prosper-lass...       2  Stopped running \"localization\": exit st...  06-29 01:18:10\n",
      "prosper-lass...       1  Stopped running \"localization\": exit st...  06-29 01:25:39\n",
      "prosper-lass...       6  Stopped running \"localization\": exit st...  06-29 01:00:29\n",
      "prosper-lass...       5  Stopped running \"localization\": exit st...  06-29 01:00:30\n",
      "prosper-lass...       4  Stopped running \"localization\": exit st...  06-29 01:00:34\n",
      "prosper-lass...       3  Stopped running \"localization\": exit st...  06-29 01:00:35\n",
      "prosper-lass...       2  Stopped running \"localization\": exit st...  06-29 01:00:41\n",
      "prosper-lass...       1  Stopped running \"localization\": exit st...  06-29 01:00:31\n",
      "mussel-sumst...      18  Success                                     06-29 01:00:20\n",
      "mussel-sumst...      17  Success                                     06-29 01:00:11\n",
      "mussel-sumst...      16  Success                                     06-29 01:00:25\n",
      "mussel-sumst...      15  Success                                     06-29 01:00:18\n",
      "mussel-sumst...      14  Success                                     06-29 01:00:29\n",
      "mussel-sumst...      13  Success                                     06-29 01:00:13\n",
      "mussel-sumst...      12  Success                                     06-29 01:00:47\n",
      "mussel-sumst...      11  Success                                     06-29 01:00:22\n",
      "mussel-sumst...      10  Success                                     06-29 01:00:15\n",
      "mussel-sumst...       9  Success                                     06-29 01:00:35\n",
      "mussel-sumst...       8  Success                                     06-29 01:00:32\n",
      "mussel-sumst...       7  Success                                     06-29 01:00:18\n",
      "mussel-sumst...       6  Success                                     06-29 01:00:23\n",
      "mussel-sumst...       5  Success                                     06-29 01:00:44\n",
      "mussel-sumst...       4  Success                                     06-29 01:00:24\n",
      "mussel-sumst...       3  Success                                     06-29 01:00:27\n",
      "mussel-sumst...       2  Success                                     06-29 01:00:28\n",
      "mussel-sumst...       1  Success                                     06-29 01:00:17\n",
      "ct-sleb-dimct        12  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct        11  Success                                     07-01 09:50:42\n",
      "ct-sleb-dimct        10  Success                                     06-30 22:32:20\n",
      "ct-sleb-dimct         9  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct         8  Success                                     07-01 09:05:15\n",
      "ct-sleb-dimct         7  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct         6  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct         5  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct         4  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct         3  Success                                     07-01 05:09:32\n",
      "ct-sleb-dimct         2  The operation was cancelled                 07-01 13:37:31\n",
      "ct-sleb-dimct         1  The operation was cancelled                 07-01 13:37:31\n",
      "mussel-sumst...      18  Success                                     06-29 00:51:06\n",
      "mussel-sumst...      17  Success                                     06-29 00:51:10\n",
      "mussel-sumst...      16  Success                                     06-29 00:50:59\n",
      "mussel-sumst...      15  Success                                     06-29 00:51:08\n",
      "mussel-sumst...      14  Success                                     06-29 00:51:11\n",
      "mussel-sumst...      13  Success                                     06-29 00:51:07\n",
      "mussel-sumst...      12  Success                                     06-29 00:51:26\n",
      "mussel-sumst...      11  Success                                     06-29 00:51:04\n",
      "mussel-sumst...      10  Success                                     06-29 00:50:59\n",
      "mussel-sumst...       9  Success                                     06-29 00:51:20\n",
      "mussel-sumst...       8  Success                                     06-29 00:51:12\n",
      "mussel-sumst...       7  Success                                     06-29 00:51:07\n",
      "mussel-sumst...       6  Success                                     06-29 00:51:01\n",
      "mussel-sumst...       5  Success                                     06-29 00:51:04\n",
      "mussel-sumst...       4  Success                                     06-29 00:51:38\n",
      "mussel-sumst...       3  Success                                     06-29 00:51:04\n",
      "mussel-sumst...       2  Success                                     06-29 00:51:00\n",
      "mussel-sumst...       1  Success                                     06-29 00:51:01\n",
      "prosper-sums...      18  Success                                     06-29 00:50:26\n",
      "prosper-sums...      17  Success                                     06-29 00:50:28\n",
      "prosper-sums...      16  Success                                     06-29 00:50:32\n",
      "prosper-sums...      15  Success                                     06-29 00:50:42\n",
      "prosper-sums...      14  Success                                     06-29 00:50:19\n",
      "prosper-sums...      13  Success                                     06-29 00:51:05\n",
      "prosper-sums...      12  Success                                     06-29 00:50:30\n",
      "prosper-sums...      11  Success                                     06-29 00:50:31\n",
      "prosper-sums...      10  Success                                     06-29 00:50:22\n",
      "prosper-sums...       9  Success                                     06-29 00:50:35\n",
      "prosper-sums...       8  Success                                     06-29 00:50:27\n",
      "prosper-sums...       7  Success                                     06-29 00:50:31\n",
      "prosper-sums...       6  Success                                     06-29 00:50:59\n",
      "prosper-sums...       5  Success                                     06-29 00:50:35\n",
      "prosper-sums...       4  Success                                     06-29 00:50:33\n",
      "prosper-sums...       3  Success                                     06-29 00:51:04\n",
      "prosper-sums...       2  Success                                     06-29 00:50:46\n",
      "prosper-sums...       1  Success                                     06-29 00:50:29\n",
      "add-rsids-su...      18  Success                                     06-28 14:12:37\n",
      "add-rsids-su...      17  Success                                     06-28 14:10:06\n",
      "add-rsids-su...      16  Success                                     06-28 14:14:22\n",
      "add-rsids-su...      15  Success                                     06-28 14:12:41\n",
      "add-rsids-su...      14  Success                                     06-28 14:13:10\n",
      "add-rsids-su...      13  Success                                     06-28 14:13:00\n",
      "add-rsids-su...      12  Success                                     06-28 14:13:46\n",
      "add-rsids-su...      11  Success                                     06-28 14:18:05\n",
      "add-rsids-su...      10  Success                                     06-28 14:12:13\n",
      "add-rsids-su...       9  Success                                     06-28 14:14:16\n",
      "add-rsids-su...       8  Success                                     06-28 14:12:12\n",
      "add-rsids-su...       7  Success                                     06-28 14:09:06\n",
      "add-rsids-su...       6  Success                                     06-28 14:19:53\n",
      "add-rsids-su...       5  Success                                     06-28 14:12:12\n",
      "add-rsids-su...       4  Success                                     06-28 14:15:28\n",
      "add-rsids-su...       3  Success                                     06-28 14:10:50\n",
      "add-rsids-su...       2  Success                                     06-28 14:14:11\n",
      "add-rsids-su...       1  Success                                     06-28 14:11:50\n",
      "add-rsids-su...      18  Stopped running \"localization\": exit st...  06-28 13:29:58\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "add-rsids-su...      17  Stopped running \"localization\": exit st...  06-28 13:30:08\n",
      "add-rsids-su...      16  Stopped running \"localization\": exit st...  06-28 13:31:31\n",
      "add-rsids-su...      15  Stopped running \"localization\": exit st...  06-28 13:31:35\n",
      "add-rsids-su...      14  Stopped running \"localization\": exit st...  06-28 13:30:01\n",
      "add-rsids-su...      13  Stopped running \"localization\": exit st...  06-28 13:29:58\n",
      "add-rsids-su...      12  Stopped running \"localization\": exit st...  06-28 13:29:44\n",
      "add-rsids-su...      11  Stopped running \"localization\": exit st...  06-28 13:30:05\n",
      "add-rsids-su...      10  Stopped running \"localization\": exit st...  06-28 13:29:48\n",
      "add-rsids-su...       9  Stopped running \"localization\": exit st...  06-28 13:29:51\n",
      "add-rsids-su...       8  Stopped running \"localization\": exit st...  06-28 13:31:33\n",
      "add-rsids-su...       7  Stopped running \"localization\": exit st...  06-28 13:29:52\n",
      "add-rsids-su...       6  Stopped running \"localization\": exit st...  06-28 13:29:39\n",
      "add-rsids-su...       5  Stopped running \"localization\": exit st...  06-28 13:29:35\n",
      "add-rsids-su...       4  Stopped running \"localization\": exit st...  06-28 13:29:58\n",
      "add-rsids-su...       3  Stopped running \"localization\": exit st...  06-28 13:31:31\n",
      "add-rsids-su...       2  Stopped running \"localization\": exit st...  06-28 13:31:31\n",
      "add-rsids-su...       1  Stopped running \"localization\": exit st...  06-28 13:31:07\n",
      "gwas-sumstat...      18  Success                                     06-28 13:18:49\n",
      "gwas-sumstat...      17  Success                                     06-28 13:19:27\n",
      "gwas-sumstat...      16  Success                                     06-28 13:22:32\n",
      "gwas-sumstat...      15  Success                                     06-28 13:19:52\n",
      "gwas-sumstat...      14  Success                                     06-28 13:20:00\n",
      "gwas-sumstat...      13  Success                                     06-28 13:22:37\n",
      "gwas-sumstat...      12  Success                                     06-28 13:19:44\n",
      "gwas-sumstat...      11  Success                                     06-28 13:17:56\n",
      "gwas-sumstat...      10  Success                                     06-28 13:15:51\n",
      "gwas-sumstat...       9  Success                                     06-28 13:20:56\n",
      "gwas-sumstat...       8  Success                                     06-28 13:19:23\n",
      "gwas-sumstat...       7  Success                                     06-28 13:23:57\n",
      "gwas-sumstat...       6  Success                                     06-28 13:20:20\n",
      "gwas-sumstat...       5  Success                                     06-28 13:22:08\n",
      "gwas-sumstat...       4  Success                                     06-28 13:17:51\n",
      "gwas-sumstat...       3  Success                                     06-28 13:25:17\n",
      "gwas-sumstat...       2  Success                                     06-28 13:19:29\n",
      "gwas-sumstat...       1  Success                                     06-28 13:19:24\n",
      "gwas-sumstat...      18  Success                                     06-28 12:39:24\n",
      "gwas-sumstat...      17  Success                                     06-28 12:31:12\n",
      "gwas-sumstat...      16  Success                                     06-28 12:32:07\n",
      "gwas-sumstat...      15  Success                                     06-28 12:32:37\n",
      "gwas-sumstat...      14  Success                                     06-28 12:35:27\n",
      "gwas-sumstat...      13  Success                                     06-28 12:35:53\n",
      "gwas-sumstat...      12  Success                                     06-28 12:34:14\n",
      "gwas-sumstat...      11  Success                                     06-28 12:29:47\n",
      "gwas-sumstat...      10  Success                                     06-28 12:31:49\n",
      "gwas-sumstat...       9  Success                                     06-28 12:32:02\n",
      "gwas-sumstat...       8  Success                                     06-28 12:31:48\n",
      "gwas-sumstat...       7  Success                                     06-28 12:33:57\n",
      "gwas-sumstat...       6  Success                                     06-28 12:30:29\n",
      "gwas-sumstat...       5  Success                                     06-28 12:34:26\n",
      "gwas-sumstat...       4  Success                                     06-28 12:33:20\n",
      "gwas-sumstat...       3  Success                                     06-28 12:36:48\n",
      "gwas-sumstat...       2  Success                                     06-28 12:32:54\n",
      "gwas-sumstat...       1  The operation was cancelled                 06-28 12:40:49\n",
      "tune-validat...       1  Success                                     06-28 13:41:21\n",
      "gwas-sumstat...      18  Success                                     06-28 11:34:15\n",
      "gwas-sumstat...      17  Success                                     06-28 11:33:20\n",
      "gwas-sumstat...      16  Success                                     06-28 11:32:35\n",
      "gwas-sumstat...      15  Success                                     06-28 11:32:58\n",
      "gwas-sumstat...      14  Success                                     06-28 11:34:14\n",
      "gwas-sumstat...      13  Success                                     06-28 11:35:00\n",
      "gwas-sumstat...      12  Success                                     06-28 11:39:36\n",
      "gwas-sumstat...      11  Success                                     06-28 11:30:52\n",
      "gwas-sumstat...      10  Success                                     06-28 11:33:32\n",
      "gwas-sumstat...       9  Success                                     06-28 11:33:33\n",
      "gwas-sumstat...       8  Success                                     06-28 11:36:48\n",
      "gwas-sumstat...       7  Success                                     06-28 11:36:42\n",
      "gwas-sumstat...       6  Success                                     06-28 11:35:17\n",
      "gwas-sumstat...       5  Success                                     06-28 11:36:17\n",
      "gwas-sumstat...       4  Success                                     06-28 11:33:57\n",
      "gwas-sumstat...       3  Success                                     06-28 11:34:55\n",
      "gwas-sumstat...       2  Success                                     06-28 11:34:51\n",
      "gwas-sumstat...       1  Success                                     06-28 11:35:18\n",
      "tune-validat...       1  Success                                     06-27 16:33:24\n",
      "tune-validat...       1  Stopped running \"user-command\": exit st...  06-27 16:26:41\n",
      "tune-validat...       1  The operation was cancelled                 06-27 16:17:56\n",
      "tune-validat...       1  Stopped running \"user-command\": exit st...  06-27 16:18:27\n",
      "tune-validat...       1  Stopped running \"user-command\": exit st...  06-27 18:38:48\n",
      "tune-validat...       3  Success                                     06-28 02:15:12\n",
      "tune-validat...       2  Success                                     06-28 02:39:25\n",
      "tune-validat...       1  Success                                     06-28 02:38:59\n",
      "gwas-sumstats        18  Success                                     06-28 03:00:13\n",
      "gwas-sumstats        17  Success                                     06-27 22:46:17\n",
      "gwas-sumstats        16  Success                                     06-27 23:43:55\n",
      "gwas-sumstats        15  Success                                     06-27 22:46:18\n",
      "gwas-sumstats        14  Success                                     06-27 23:16:05\n",
      "gwas-sumstats        13  Success                                     06-28 03:26:32\n",
      "gwas-sumstats        12  Success                                     06-27 22:11:28\n",
      "gwas-sumstats        11  Success                                     06-27 22:56:21\n",
      "gwas-sumstats        10  Success                                     06-27 23:12:19\n",
      "gwas-sumstats         9  Success                                     06-27 22:11:04\n",
      "gwas-sumstats         8  Success                                     06-27 22:24:56\n",
      "gwas-sumstats         7  Success                                     06-27 22:45:40\n",
      "gwas-sumstats         6  Success                                     06-27 22:35:59\n",
      "gwas-sumstats         5  Success                                     06-27 22:40:54\n",
      "gwas-sumstats         4  Success                                     06-27 22:12:02\n",
      "gwas-sumstats         3  Success                                     06-27 22:06:28\n",
      "gwas-sumstats         2  Success                                     06-27 22:18:14\n",
      "gwas-sumstats         1  Success                                     06-27 22:36:33\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train-tune-v...       1  Success                                     06-27 15:37:44\n",
      "train-tune-v...       1  Stopped running \"user-command\": exit st...  06-27 15:30:31\n",
      "all-phenotypes        1  Success                                     06-27 15:26:30\n",
      "all-phenotypes        1  Success                                     06-27 15:20:38\n",
      "all-phenotypes        1  Success                                     06-27 15:14:34\n",
      "all-phenotypes        1  Stopped running \"user-command\": exit st...  06-27 15:04:32\n",
      "ct-sleb-dimct        12  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct        11  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct        10  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         9  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         8  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         7  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         6  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         5  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         4  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         3  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         2  The operation was cancelled                 06-27 14:37:42\n",
      "ct-sleb-dimct         1  The operation was cancelled                 06-27 14:37:42\n",
      "tune-validat...       1  The operation was cancelled                 06-27 14:37:42\n",
      "tune-validat...       3  The operation was cancelled                 06-27 12:20:35\n",
      "tune-validat...       2  The operation was cancelled                 06-27 12:20:35\n",
      "tune-validat...       1  The operation was cancelled                 06-27 12:20:35\n",
      "tune-validat...       3  Success                                     06-27 03:49:00\n",
      "tune-validat...       2  Success                                     06-27 04:35:25\n",
      "tune-validat...       1  Success                                     06-27 05:22:59\n",
      "tune-validat...       3  Stopped running \"localization\": exit st...  06-27 01:41:43\n",
      "tune-validat...       2  Stopped running \"localization\": exit st...  06-27 01:41:36\n",
      "tune-validat...       1  Stopped running \"localization\": exit st...  06-27 01:25:28\n",
      "prosper-sums...      18  Success                                     06-26 15:22:08\n",
      "prosper-sums...      17  Success                                     06-26 15:13:46\n",
      "prosper-sums...      16  Success                                     06-26 15:11:41\n",
      "prosper-sums...      15  Success                                     06-26 15:11:40\n",
      "prosper-sums...      14  Success                                     06-26 15:11:42\n",
      "prosper-sums...      13  Success                                     06-26 15:11:56\n",
      "prosper-sums...      12  Success                                     06-26 15:22:00\n",
      "prosper-sums...      11  Success                                     06-26 15:11:34\n",
      "prosper-sums...      10  Success                                     06-26 15:22:20\n",
      "prosper-sums...       9  Success                                     06-26 15:22:37\n",
      "prosper-sums...       8  Success                                     06-26 15:11:36\n",
      "prosper-sums...       7  Success                                     06-26 15:11:33\n",
      "prosper-sums...       6  Success                                     06-26 15:22:14\n",
      "prosper-sums...       5  Success                                     06-26 15:11:39\n",
      "prosper-sums...       4  Success                                     06-26 15:11:34\n",
      "prosper-sums...       3  Success                                     06-26 15:22:08\n",
      "prosper-sums...       2  Success                                     06-26 15:13:48\n",
      "prosper-sums...       1  Success                                     06-26 15:11:35\n",
      "prosper-sums...      18  Success                                     06-26 14:47:12\n",
      "prosper-sums...      17  Success                                     06-26 14:47:14\n",
      "prosper-sums...      16  Success                                     06-26 14:47:11\n",
      "prosper-sums...      15  Success                                     06-26 14:47:13\n",
      "prosper-sums...      14  Success                                     06-26 14:47:08\n",
      "prosper-sums...      13  Success                                     06-26 14:47:24\n",
      "prosper-sums...      12  Success                                     06-26 14:47:29\n",
      "prosper-sums...      11  Success                                     06-26 14:47:55\n",
      "prosper-sums...      10  Success                                     06-26 14:47:27\n",
      "prosper-sums...       9  Success                                     06-26 14:47:03\n",
      "prosper-sums...       8  Success                                     06-26 14:47:13\n",
      "prosper-sums...       7  Success                                     06-26 14:47:07\n",
      "prosper-sums...       6  Success                                     06-26 14:48:24\n",
      "prosper-sums...       5  Success                                     06-26 14:47:18\n",
      "prosper-sums...       4  Success                                     06-26 14:46:59\n",
      "prosper-sums...       3  Success                                     06-26 14:47:03\n",
      "prosper-sums...       2  Success                                     06-26 14:47:22\n",
      "prosper-sums...       1  Success                                     06-26 14:47:12\n",
      "add-rsids-su...      18  Success                                     06-26 14:01:31\n",
      "add-rsids-su...      17  Success                                     06-26 14:01:59\n",
      "add-rsids-su...      16  Success                                     06-26 14:01:50\n",
      "add-rsids-su...      15  Success                                     06-26 14:02:18\n",
      "add-rsids-su...      14  Success                                     06-26 14:01:00\n",
      "add-rsids-su...      13  Success                                     06-26 14:03:31\n",
      "add-rsids-su...      12  Success                                     06-26 14:02:24\n",
      "add-rsids-su...      11  Success                                     06-26 14:03:12\n",
      "add-rsids-su...      10  Success                                     06-26 14:03:09\n",
      "add-rsids-su...       9  Success                                     06-26 14:01:12\n",
      "add-rsids-su...       8  Success                                     06-26 14:00:29\n",
      "add-rsids-su...       7  Success                                     06-26 14:01:33\n",
      "add-rsids-su...       6  Success                                     06-26 14:01:11\n",
      "add-rsids-su...       5  Success                                     06-26 14:01:07\n",
      "add-rsids-su...       4  Success                                     06-26 14:00:51\n",
      "add-rsids-su...       3  Success                                     06-26 14:00:24\n",
      "add-rsids-su...       2  Success                                     06-26 14:00:29\n",
      "add-rsids-su...       1  Success                                     06-26 14:00:26\n",
      "prosper-sums...      18  Success                                     06-26 12:23:18\n",
      "prosper-sums...      17  Success                                     06-26 12:23:19\n",
      "prosper-sums...      16  Success                                     06-26 12:23:08\n",
      "prosper-sums...      15  Success                                     06-26 12:22:53\n",
      "prosper-sums...      14  Success                                     06-26 12:22:46\n",
      "prosper-sums...      13  Success                                     06-26 12:23:17\n",
      "prosper-sums...      12  Success                                     06-26 12:22:55\n",
      "prosper-sums...      11  Success                                     06-26 12:22:52\n",
      "prosper-sums...      10  Success                                     06-26 12:22:57\n",
      "prosper-sums...       9  Success                                     06-26 12:22:58\n",
      "prosper-sums...       8  Success                                     06-26 12:22:59\n",
      "prosper-sums...       7  Success                                     06-26 12:23:04\n",
      "prosper-sums...       6  Success                                     06-26 12:22:53\n",
      "prosper-sums...       5  Success                                     06-26 12:22:54\n",
      "prosper-sums...       4  Success                                     06-26 12:22:46\n",
      "prosper-sums...       3  Success                                     06-26 12:22:48\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prosper-sums...       2  Success                                     06-26 12:23:17\n",
      "prosper-sums...       1  Success                                     06-26 12:22:42\n",
      "prosper-sums...      18  Stopped running \"user-command\": exit st...  06-25 21:24:18\n",
      "prosper-sums...      17  Stopped running \"user-command\": exit st...  06-25 21:24:44\n",
      "prosper-sums...      16  Stopped running \"user-command\": exit st...  06-25 21:24:21\n",
      "prosper-sums...      15  Stopped running \"user-command\": exit st...  06-25 21:23:55\n",
      "prosper-sums...      14  Stopped running \"user-command\": exit st...  06-25 21:24:18\n",
      "prosper-sums...      13  Stopped running \"user-command\": exit st...  06-25 21:24:38\n",
      "prosper-sums...      12  Stopped running \"user-command\": exit st...  06-25 21:24:04\n",
      "prosper-sums...      11  Stopped running \"user-command\": exit st...  06-25 21:24:08\n",
      "prosper-sums...      10  Stopped running \"user-command\": exit st...  06-25 21:24:26\n",
      "prosper-sums...       9  Stopped running \"user-command\": exit st...  06-25 21:23:54\n",
      "prosper-sums...       8  Stopped running \"user-command\": exit st...  06-25 21:24:03\n",
      "prosper-sums...       7  Stopped running \"user-command\": exit st...  06-25 21:24:13\n",
      "prosper-sums...       6  Stopped running \"user-command\": exit st...  06-25 21:24:22\n",
      "prosper-sums...       5  Stopped running \"user-command\": exit st...  06-25 21:24:07\n",
      "prosper-sums...       4  Stopped running \"user-command\": exit st...  06-25 21:23:59\n",
      "prosper-sums...       3  Stopped running \"user-command\": exit st...  06-25 21:24:05\n",
      "prosper-sums...       2  Stopped running \"user-command\": exit st...  06-25 21:24:22\n",
      "prosper-sums...       1  Stopped running \"user-command\": exit st...  06-25 21:24:15\n",
      "prosper-sums...      18  Stopped running \"user-command\": exit st...  06-25 20:00:41\n",
      "prosper-sums...      17  Stopped running \"user-command\": exit st...  06-25 20:00:49\n",
      "prosper-sums...      16  Stopped running \"user-command\": exit st...  06-25 20:00:53\n",
      "prosper-sums...      15  Stopped running \"user-command\": exit st...  06-25 20:00:41\n",
      "prosper-sums...      14  Stopped running \"user-command\": exit st...  06-25 20:00:48\n",
      "prosper-sums...      13  Stopped running \"user-command\": exit st...  06-25 20:00:52\n",
      "prosper-sums...      12  Stopped running \"user-command\": exit st...  06-25 20:00:43\n",
      "prosper-sums...      11  Stopped running \"user-command\": exit st...  06-25 20:01:04\n",
      "prosper-sums...      10  Stopped running \"user-command\": exit st...  06-25 20:00:48\n",
      "prosper-sums...       9  Stopped running \"user-command\": exit st...  06-25 20:00:48\n",
      "prosper-sums...       8  Stopped running \"user-command\": exit st...  06-25 20:00:46\n",
      "prosper-sums...       7  Stopped running \"user-command\": exit st...  06-25 20:00:54\n",
      "prosper-sums...       6  Stopped running \"user-command\": exit st...  06-25 20:00:47\n",
      "prosper-sums...       5  Stopped running \"user-command\": exit st...  06-25 20:00:52\n",
      "prosper-sums...       4  Stopped running \"user-command\": exit st...  06-25 20:00:34\n",
      "prosper-sums...       3  Stopped running \"user-command\": exit st...  06-25 20:00:44\n",
      "prosper-sums...       2  Stopped running \"user-command\": exit st...  06-25 20:00:46\n",
      "prosper-sums...       1  Stopped running \"user-command\": exit st...  06-25 20:00:47\n",
      "prosper-sums...      18  Stopped running \"user-command\": exit st...  06-25 19:08:17\n",
      "prosper-sums...      17  Stopped running \"user-command\": exit st...  06-25 19:08:21\n",
      "prosper-sums...      16  Stopped running \"user-command\": exit st...  06-25 19:08:36\n",
      "prosper-sums...      15  Stopped running \"user-command\": exit st...  06-25 19:08:22\n",
      "prosper-sums...      14  Stopped running \"user-command\": exit st...  06-25 19:08:20\n",
      "prosper-sums...      13  Stopped running \"user-command\": exit st...  06-25 19:08:14\n",
      "prosper-sums...      12  Stopped running \"user-command\": exit st...  06-25 19:08:20\n",
      "prosper-sums...      11  Stopped running \"user-command\": exit st...  06-25 19:08:22\n",
      "prosper-sums...      10  Stopped running \"user-command\": exit st...  06-25 19:08:19\n",
      "prosper-sums...       9  Stopped running \"user-command\": exit st...  06-25 19:08:09\n",
      "prosper-sums...       8  Stopped running \"user-command\": exit st...  06-25 19:08:31\n",
      "prosper-sums...       7  Stopped running \"user-command\": exit st...  06-25 19:08:07\n",
      "prosper-sums...       6  Stopped running \"user-command\": exit st...  06-25 19:08:30\n",
      "prosper-sums...       5  Stopped running \"user-command\": exit st...  06-25 19:08:13\n",
      "prosper-sums...       4  Stopped running \"user-command\": exit st...  06-25 19:08:32\n",
      "prosper-sums...       3  Stopped running \"user-command\": exit st...  06-25 19:08:21\n",
      "prosper-sums...       2  Stopped running \"user-command\": exit st...  06-25 19:08:13\n",
      "prosper-sums...       1  Stopped running \"user-command\": exit st...  06-25 19:08:27\n",
      "prosper-sums...      18  pulling image: docker pull: retry budge...  06-25 19:03:09\n",
      "prosper-sums...      17  pulling image: docker pull: retry budge...  06-25 18:59:46\n",
      "prosper-sums...      16  pulling image: docker pull: retry budge...  06-25 18:59:28\n",
      "prosper-sums...      15  pulling image: docker pull: retry budge...  06-25 19:01:57\n",
      "prosper-sums...      14  pulling image: docker pull: retry budge...  06-25 19:00:41\n",
      "prosper-sums...      13  pulling image: docker pull: retry budge...  06-25 19:00:53\n",
      "prosper-sums...      12  pulling image: docker pull: retry budge...  06-25 18:59:54\n",
      "prosper-sums...      11  pulling image: docker pull: retry budge...  06-25 19:00:16\n",
      "prosper-sums...      10  pulling image: docker pull: retry budge...  06-25 19:00:45\n",
      "prosper-sums...       9  pulling image: docker pull: retry budge...  06-25 19:01:57\n",
      "prosper-sums...       8  pulling image: docker pull: retry budge...  06-25 18:58:30\n",
      "prosper-sums...       7  pulling image: docker pull: retry budge...  06-25 19:00:51\n",
      "prosper-sums...       6  pulling image: docker pull: retry budge...  06-25 19:02:39\n",
      "prosper-sums...       5  pulling image: docker pull: retry budge...  06-25 19:02:08\n",
      "prosper-sums...       4  pulling image: docker pull: retry budge...  06-25 19:00:25\n",
      "prosper-sums...       3  pulling image: docker pull: retry budge...  06-25 19:00:39\n",
      "prosper-sums...       2  pulling image: docker pull: retry budge...  06-25 19:02:07\n",
      "prosper-sums...       1  pulling image: docker pull: retry budge...  06-25 19:02:46\n",
      "prosper-sums...      18  pulling image: docker pull: retry budge...  06-25 16:03:26\n",
      "prosper-sums...      17  pulling image: docker pull: retry budge...  06-25 16:03:52\n",
      "prosper-sums...      16  pulling image: docker pull: retry budge...  06-25 16:03:44\n",
      "prosper-sums...      15  pulling image: docker pull: retry budge...  06-25 16:03:53\n",
      "prosper-sums...      14  pulling image: docker pull: retry budge...  06-25 16:02:36\n",
      "prosper-sums...      13  pulling image: docker pull: retry budge...  06-25 16:03:15\n",
      "prosper-sums...      12  pulling image: docker pull: retry budge...  06-25 16:02:42\n",
      "prosper-sums...      11  pulling image: docker pull: retry budge...  06-25 16:02:05\n",
      "prosper-sums...      10  pulling image: docker pull: retry budge...  06-25 16:02:08\n",
      "prosper-sums...       9  pulling image: docker pull: retry budge...  06-25 16:03:09\n",
      "prosper-sums...       8  pulling image: docker pull: retry budge...  06-25 16:03:10\n",
      "prosper-sums...       7  pulling image: docker pull: retry budge...  06-25 16:02:29\n",
      "prosper-sums...       6  pulling image: docker pull: retry budge...  06-25 16:01:29\n",
      "prosper-sums...       5  pulling image: docker pull: retry budge...  06-25 16:03:18\n",
      "prosper-sums...       4  pulling image: docker pull: retry budge...  06-25 16:01:12\n",
      "prosper-sums...       3  pulling image: docker pull: retry budge...  06-25 16:02:06\n",
      "prosper-sums...       2  pulling image: docker pull: retry budge...  06-25 16:01:55\n",
      "prosper-sums...       1  pulling image: docker pull: retry budge...  06-25 16:00:35\n",
      "prosper-sums...      18  pulling image: docker pull: retry budge...  06-25 15:19:54\n",
      "prosper-sums...      17  pulling image: docker pull: retry budge...  06-25 15:22:25\n",
      "prosper-sums...      16  pulling image: docker pull: retry budge...  06-25 15:19:22\n",
      "prosper-sums...      15  pulling image: docker pull: retry budge...  06-25 15:20:18\n",
      "prosper-sums...      14  pulling image: docker pull: retry budge...  06-25 15:21:50\n",
      "prosper-sums...      13  pulling image: docker pull: retry budge...  06-25 15:21:55\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prosper-sums...      12  pulling image: docker pull: retry budge...  06-25 15:23:04\n",
      "prosper-sums...      11  pulling image: docker pull: retry budge...  06-25 15:21:33\n",
      "prosper-sums...      10  pulling image: docker pull: retry budge...  06-25 15:20:38\n",
      "prosper-sums...       9  pulling image: docker pull: retry budge...  06-25 15:20:27\n",
      "prosper-sums...       8  pulling image: docker pull: retry budge...  06-25 15:20:21\n",
      "prosper-sums...       7  pulling image: docker pull: retry budge...  06-25 15:21:06\n",
      "prosper-sums...       6  pulling image: docker pull: retry budge...  06-25 15:21:59\n",
      "prosper-sums...       5  pulling image: docker pull: retry budge...  06-25 15:21:16\n",
      "prosper-sums...       4  pulling image: docker pull: retry budge...  06-25 15:20:07\n",
      "prosper-sums...       3  pulling image: docker pull: retry budge...  06-25 15:20:04\n",
      "prosper-sums...       2  pulling image: docker pull: retry budge...  06-25 15:21:33\n",
      "prosper-sums...       1  pulling image: docker pull: retry budge...  06-25 15:20:44\n",
      "ct-sleb-dimct        12  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct        11  Stopped running \"user-command\": exit st...  06-27 08:26:44\n",
      "ct-sleb-dimct        10  Stopped running \"user-command\": exit st...  06-27 11:05:06\n",
      "ct-sleb-dimct         9  Stopped running \"user-command\": exit st...  06-27 10:35:12\n",
      "ct-sleb-dimct         8  Stopped running \"user-command\": exit st...  06-27 02:11:05\n",
      "ct-sleb-dimct         7  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct         6  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct         5  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct         4  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct         3  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct         2  The operation was cancelled                 06-27 12:29:56\n",
      "ct-sleb-dimct         1  The operation was cancelled                 06-27 12:29:56\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "dstat \\\n",
    "    --provider google-cls-v2 \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --location us-central1 \\\n",
    "    --users \"${USER_NAME}\" \\\n",
    "    --jobs '*' \\\n",
    "    --status '*' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Delete running jobs:\n",
      "  user:\n",
      "    {'williamsjacr'}\n",
      "\n",
      "  job-id:\n",
      "    ['*']\n",
      "\n",
      "Found 0 tasks to delete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 jobs deleted\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ddel \\\n",
    "    --provider google-cls-v2 \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --location us-central1 \\\n",
    "    --jobs '*'\\\n",
    "    --users 'williamsjacr'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile Common_Variants.R\n",
    "rm(list = ls())\n",
    "\n",
    "chr <- as.numeric(commandArgs(TRUE)[1])\n",
    "print(chr)\n",
    "\n",
    "BED_file <- commandArgs(TRUE)[2]\n",
    "BED_file <- gsub(\".bed\",\"\",BED_file)\n",
    "print(BED_file)\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[5]\n",
    "print(Ancestry_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "\n",
    "AFR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AFR\"])\n",
    "AMR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AMR\"])\n",
    "EUR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"EUR\"])\n",
    "\n",
    "\n",
    "write.table(AFR_Keep,\"AFR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(AMR_Keep,\"AMR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(EUR_Keep,\"EUR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AFR_Keep.txt --freq --out chr\",chr,\"_freq_AFR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AMR_Keep.txt --freq --out chr\",chr,\"_freq_AMR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep EUR_Keep.txt --freq --out chr\",chr,\"_freq_EUR\"),intern = TRUE)\n",
    "\n",
    "a <- read.csv(paste0(\"chr\",chr,\"_freq_AFR.frq\"),sep = \"\")\n",
    "b <- read.csv(paste0(\"chr\",chr,\"_freq_AMR.frq\"),sep = \"\")\n",
    "c <- read.csv(paste0(\"chr\",chr,\"_freq_EUR.frq\"),sep = \"\")\n",
    "\n",
    "exclude_list <- a$SNP[is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF) | a$MAF <= 0.01 | b$MAF <= 0.01 | c$MAF <= 0.01]\n",
    "write.table(exclude_list,\"exclude_list.txt\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AFR_Keep.txt --exclude exclude_list.txt --make-bed --out \",OUTPUT_PATH,\"/AFR_chr\",chr,\"_corrected\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AMR_Keep.txt --exclude exclude_list.txt --make-bed --out \",OUTPUT_PATH,\"/AMR_chr\",chr,\"_corrected\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep EUR_Keep.txt --exclude exclude_list.txt --make-bed --out \",OUTPUT_PATH,\"/EUR_chr\",chr,\"_corrected\"),intern = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for (chrom in 1:22) {\n",
    "  tasks <- rbind(tasks, data.frame(\n",
    "    '--env CHR'=chrom,\n",
    "    '--input BED_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/acaf_threshold_v7.1/plink_bed/acaf_threshold.chr\",chrom,\".bed\"),\n",
    "    '--input BIM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/acaf_threshold_v7.1/plink_bed/acaf_threshold.chr\",chrom,\".bim\"),\n",
    "    '--input FAM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/acaf_threshold_v7.1/plink_bed/acaf_threshold.chr\",chrom,\".fam\"),\n",
    "    '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Common_Variants.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "}\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile Common_Variants.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${CHR} ${BED_File} ${BIM_File} ${FAM_File} ${Ancestry_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Variants_BED_Sum.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Variants_BED_Sum.R\n",
    "rm(list = ls())\n",
    "\n",
    "chr <- as.numeric(commandArgs(TRUE)[1])\n",
    "print(chr)\n",
    "\n",
    "BED_file <- commandArgs(TRUE)[2]\n",
    "BED_file <- gsub(\".bed\",\"\",BED_file,fixed = TRUE)\n",
    "print(BED_file)\n",
    "\n",
    "CT_BIM_FILE <- commandArgs(TRUE)[3]\n",
    "print(CT_BIM_FILE)\n",
    "\n",
    "PROSPER_BIM_FILE <- commandArgs(TRUE)[4]\n",
    "print(PROSPER_BIM_FILE)\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[5]\n",
    "print(Ancestry_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "\n",
    "AFR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AFR\"])\n",
    "AMR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AMR\"])\n",
    "EUR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"EUR\"])\n",
    "\n",
    "\n",
    "write.table(AFR_Keep,\"AFR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(AMR_Keep,\"AMR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(EUR_Keep,\"EUR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AFR_Keep.txt --freq --out chr\",chr,\"_freq_AFR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AMR_Keep.txt --freq --out chr\",chr,\"_freq_AMR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep EUR_Keep.txt --freq --out chr\",chr,\"_freq_EUR\"),intern = TRUE)\n",
    "\n",
    "a <- read.csv(paste0(\"chr\",chr,\"_freq_AFR.frq\"),sep = \"\")\n",
    "b <- read.csv(paste0(\"chr\",chr,\"_freq_AMR.frq\"),sep = \"\")\n",
    "c <- read.csv(paste0(\"chr\",chr,\"_freq_EUR.frq\"),sep = \"\")\n",
    "\n",
    "exclude_list <- a$SNP[is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF) | a$MAF <= 0.01 | b$MAF <= 0.01 | c$MAF <= 0.01]\n",
    "rarevariant_list <- a$SNP[is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF)]\n",
    "\n",
    "Number_Common_Variants <- nrow(a) - length(exclude_list)\n",
    "Number_Rare_Variants <- nrow(a) - length(rarevariant_list) - Number_Common_Variants\n",
    "\n",
    "CT_BIM <- read.delim(CT_BIM_FILE,header = FALSE)\n",
    "CT_SNPs <- nrow(CT_BIM)\n",
    "\n",
    "PROSPER_BIM <- read.delim(PROSPER_BIM_FILE,header = FALSE)\n",
    "PROSPER_SNPs <- nrow(PROSPER_BIM)\n",
    "\n",
    "tmp <- data.frame(Chr = chr,\n",
    "                  Number_Common_Variants = Number_Common_Variants,\n",
    "                  Number_Rare_Variants = Number_Rare_Variants,\n",
    "                 CT_SNPs = CT_SNPs,\n",
    "                 PROSPER_SNPs = PROSPER_SNPs)\n",
    "write.csv(tmp,row.names = FALSE,file = paste0(OUTPUT_PATH,\"/\",chr,\"_NumSNPs.csv\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for (chrom in 1:22) {\n",
    "  tasks <- rbind(tasks, data.frame(\n",
    "    '--env CHR'=chrom,\n",
    "    '--input BED_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/plink_bed/exome.chr\",chrom,\".bed\"),\n",
    "    '--input BIM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/plink_bed/exome.chr\",chrom,\".bim\"),\n",
    "    '--input FAM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/plink_bed/exome.chr\",chrom,\".fam\"),\n",
    "    '--input CT_BIM_FILE'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input PROSPER_BIM_FILE'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",  \n",
    "    '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Common_Variants_BED_Sum.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Number_SNPs\",\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "}\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Variants_BED_Sum.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Variants_BED_Sum.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${CHR} ${BED_File} ${CT_BIM_FILE} ${PROSPER_BIM_FILE} ${Ancestry_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Common_Variants_BED_Sum.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.3 KiB/  2.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Common_Variants_BED_Sum.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: common-var--williamsjacr--240901-131943-66\n",
      "  job-name: common-variants-bed-sum\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11196432129391680803\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7824575572013587726\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9758527684499917372\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15865405136015907616\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13738081057261705017\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1215447884564362733\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12706738842639517649\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9128748637006943225\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17641149042402703339\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9298657386333898969\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8368203038466850967\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6101676909241046982\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10799503408380772836\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9032221348798239212\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7251358705368745639\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9939768856006987805\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13772409251482507036\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17914230607144014707\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13585512460409370635\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10923239924929804856\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13427428528047602329\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13504958931632819405\n",
      "Launched job-id: common-var--williamsjacr--240901-131943-66\n",
      "22 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--240901-131943-66' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--240901-131943-66' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 400 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 32 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Common_Variants_BED_Sum.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "common-varia...      22  Success   09-01 14:09:08\r\n",
      "common-varia...      21  Success   09-01 13:33:01\r\n",
      "common-varia...      20  Success   09-01 14:19:11\r\n",
      "common-varia...      19  Success   09-01 16:18:57\r\n",
      "common-varia...      18  Success   09-01 14:00:15\r\n",
      "common-varia...      17  Success   09-01 15:48:39\r\n",
      "common-varia...      16  Success   09-01 15:15:55\r\n",
      "common-varia...      15  Success   09-01 14:50:13\r\n",
      "common-varia...      14  Success   09-01 14:45:19\r\n",
      "common-varia...      13  Success   09-01 14:04:41\r\n",
      "common-varia...      12  Success   09-01 15:38:07\r\n",
      "common-varia...      11  Success   09-01 15:49:05\r\n",
      "common-varia...      10  Success   09-01 14:59:51\r\n",
      "common-varia...       9  Success   09-01 15:02:14\r\n",
      "common-varia...       8  Success   09-01 14:53:29\r\n",
      "common-varia...       7  Success   09-01 15:14:49\r\n",
      "common-varia...       6  Success   09-01 15:30:24\r\n",
      "common-varia...       5  Success   09-01 15:21:54\r\n",
      "common-varia...       4  Success   09-01 15:10:32\r\n",
      "common-varia...       3  Success   09-01 16:03:44\r\n",
      "common-varia...       2  Success   09-01 16:28:35\r\n",
      "common-varia...       1  Success   09-01 17:43:54\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--240901-131943-66' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile merge_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[1]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "merge_list_train <- paste0(INPUT_PATH,\"/chr\",2:22,\"_corrected\")\n",
    "write.table(merge_list_train,file = \"merge_list.txt\",col.names = F,row.names = F,quote=F)\n",
    "system(paste0(\"plink --bfile \",INPUT_PATH,\"/chr\",1,\"_corrected --merge-list merge_list.txt --make-bed --out \",OUTPUT_PATH,\"/all_chr\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/merge_bed.R\",\n",
    "  '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "  # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile merge_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting all_phenotypes.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile all_phenotypes.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Related_File <- commandArgs(TRUE)[1]\n",
    "print(Related_File)\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[2]\n",
    "print(Ancestry_File)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "flagged_relatedness <- read.delim(Related_File)\n",
    "colnames(flagged_relatedness) <- \"IID\"\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "aou_BMI_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_BMI_pheno.tsv\"))\n",
    "aou_LDL_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_LDL_pheno.tsv\"))\n",
    "aou_HDL_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_HDL_pheno.tsv\"))\n",
    "aou_TG_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_TG_pheno.tsv\"))\n",
    "colnames(aou_TG_pheno)[3] <- \"logTG\"\n",
    "aou_TC_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_TC_pheno.tsv\"))\n",
    "aou_Height_pheno <- read.csv(paste0(INPUT_PATH,\"/aou_Height_pheno.tsv\"),sep =\"\")\n",
    "\n",
    "aou_BMI_cov <- read.delim(paste0(INPUT_PATH,\"/aou_BMI_cov.tsv\"))\n",
    "aou_LDL_cov <- read.csv(paste0(INPUT_PATH,\"/aou_LDL_cov.tsv\"),sep =\"\")\n",
    "aou_HDL_cov <- read.csv(paste0(INPUT_PATH,\"/aou_HDL_cov.tsv\"),sep =\"\")\n",
    "aou_TG_cov <- read.csv(paste0(INPUT_PATH,\"/aou_TG_cov.tsv\"),sep =\"\")\n",
    "aou_TC_cov <- read.csv(paste0(INPUT_PATH,\"/aou_TC_cov.tsv\"),sep =\"\")\n",
    "aou_Height_cov <- read.csv(paste0(INPUT_PATH,\"/aou_Height_cov.tsv\"),sep =\"\")\n",
    "\n",
    "library(dplyr)\n",
    "unique_ids_pheno <- unique(c(aou_BMI_pheno$IID,aou_LDL_pheno$IID,aou_HDL_pheno$IID,aou_TG_pheno$IID,aou_TC_pheno$IID,aou_Height_pheno$IID))\n",
    "unique_ids_cov <- unique(c(aou_BMI_cov$IID,aou_LDL_cov$IID,aou_HDL_cov$IID,aou_TG_cov$IID,aou_TC_cov$IID,aou_Height_cov$IID))\n",
    "\n",
    "\n",
    "all_phenotypes <- data.frame(IID = unique(c(unique_ids_pheno,unique_ids_cov)))\n",
    "all_phenotypes <- all_phenotypes[!(all_phenotypes$IID %in% flagged_relatedness$IID),,drop = FALSE]\n",
    "\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_BMI_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_LDL_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_HDL_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TG_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TC_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_Height_pheno)\n",
    "\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_BMI_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_LDL_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_HDL_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TG_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TC_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_Height_cov)\n",
    "\n",
    "all_phenotypes <- left_join(all_phenotypes,ancestries)\n",
    "\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$age),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$female),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC1),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC2),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC3),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC4),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC5),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC6),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC7),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC8),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC9),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC10),]\n",
    "\n",
    "write.csv(all_phenotypes,file = paste0(OUTPUT_PATH,\"/all_phenotypes.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/all_phenotypes.R\",\n",
    "  '--input Related_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/relatedness/relatedness_flagged_samples.tsv\",\n",
    "  '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "  '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/Tony_Files\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/\",\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting all_phenotypes.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile all_phenotypes.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Related_File} ${Ancestry_File} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://all_phenotypes.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  3.4 KiB/  3.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/3.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp all_phenotypes.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: all-phenot--williamsjacr--240627-152247-11\n",
      "  job-name: all-phenotypes\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3165015638967736981\n",
      "Launched job-id: all-phenot--williamsjacr--240627-152247-11\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'all-phenot--williamsjacr--240627-152247-11' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'all-phenot--williamsjacr--240627-152247-11' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script all_phenotypes.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name          Task  Status    Last Update\r\n",
      "--------------  ------  --------  --------------\r\n",
      "all-phenotypes       1  Success   06-27 15:20:38\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'all-phenot--williamsjacr--240627-151710-43' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Train_Tune_Validation.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Train_Tune_Validation.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "all_phenotype_File <- commandArgs(TRUE)[1]\n",
    "print(all_phenotype_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotype_File)\n",
    "\n",
    "i <- (1:length(all_phenotypes$IID))[all_phenotypes$ancestry %in% c(\"EUR\",\"AFR\",\"AMR\")]\n",
    "train_number <- round(length(all_phenotypes$IID)*0.7) + 1\n",
    "train <- sample(i, train_number)\n",
    "\n",
    "i <- (1:length(all_phenotypes$IID))[!((1:length(all_phenotypes$IID)) %in% train)]\n",
    "i_EUR <- i[all_phenotypes$ancestry[i] == \"EUR\"]\n",
    "i_AFR <- i[all_phenotypes$ancestry[i] == \"AFR\"]\n",
    "i_SAS <- i[all_phenotypes$ancestry[i] == \"SAS\"]\n",
    "i_EAS <- i[all_phenotypes$ancestry[i] == \"EAS\"]\n",
    "i_MID <- i[all_phenotypes$ancestry[i] == \"MID\"]\n",
    "i_AMR <- i[all_phenotypes$ancestry[i] == \"AMR\"]\n",
    "\n",
    "\n",
    "tune <- c(sample(i_EUR,round(length(i_EUR)/2)),\n",
    "          sample(i_AFR,round(length(i_AFR)/2)),\n",
    "          sample(i_SAS,round(length(i_SAS)/2)),\n",
    "          sample(i_EAS,round(length(i_EAS)/2)),\n",
    "          sample(i_MID,round(length(i_MID)/2)),\n",
    "          sample(i_AMR,round(length(i_AMR)/2)))\n",
    "\n",
    "validation <- i[!(i %in% tune)]\n",
    "\n",
    "\n",
    "train <- all_phenotypes$IID[train]\n",
    "tune <- all_phenotypes$IID[tune]\n",
    "validation <- all_phenotypes$IID[validation]\n",
    "\n",
    "\n",
    "write.table(train,paste0(OUTPUT_PATH,\"/train.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "write.table(tune,paste0(OUTPUT_PATH,\"/tune.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "write.table(validation,paste0(OUTPUT_PATH,\"/validation.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "reference <- train[sample(1:length(train),3000,replace = FALSE)]\n",
    "\n",
    "write.table(reference,paste0(OUTPUT_PATH,\"/reference.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "colnames(all_phenotypes)[10] <- \"sex\"\n",
    "\n",
    "phenotype_train <- all_phenotypes[all_phenotypes$IID %in% train,]\n",
    "phenotype_tune <- all_phenotypes[all_phenotypes$IID %in% tune,]\n",
    "phenotype_validation <- all_phenotypes[all_phenotypes$IID %in% validation,]\n",
    "\n",
    "phenotype_train$FID <- 0\n",
    "phenotype_tune$FID <- 0\n",
    "phenotype_validation$FID <- 0\n",
    "\n",
    "phenotype_train <- phenotype_train[,c(\"IID\",\"FID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "phenotype_train$age2 <- phenotype_train$age^2\n",
    "phenotype_tune <- phenotype_tune[,c(\"IID\",\"FID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "phenotype_tune$age2 <- phenotype_tune$age^2\n",
    "phenotype_validation <- phenotype_validation[,c(\"IID\",\"FID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "phenotype_validation$age2 <- phenotype_validation$age^2\n",
    "\n",
    "colnames(phenotype_train) <- c(\"IID\",colnames(phenotype_train)[-1])\n",
    "colnames(phenotype_tune) <- c(\"IID\",colnames(phenotype_tune)[-1])\n",
    "colnames(phenotype_validation) <- c(\"IID\",colnames(phenotype_validation)[-1])\n",
    "\n",
    "phenotype_train[c(9,11:21)] <- lapply(phenotype_train[c(9,11:21)], function(x){c(scale(x))})\n",
    "phenotype_tune[c(9,11:21)] <- lapply(phenotype_tune[c(9,11:21)], function(x){c(scale(x))})\n",
    "phenotype_validation[c(9,11:21)] <- lapply(phenotype_validation[c(9,11:21)], function(x){c(scale(x))})\n",
    "\n",
    "write.table(phenotype_train,file = paste0(OUTPUT_PATH,\"/All_Train.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "write.table(phenotype_tune,file = paste0(OUTPUT_PATH,\"/All_Tune.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "write.table(phenotype_validation,file = paste0(OUTPUT_PATH,\"/All_Validation.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Train_Tune_Validation.R\",\n",
    "  '--input all_phenotype_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/\",\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Train_Tune_Validation.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Train_Tune_Validation.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${all_phenotype_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Train_Tune_Validation.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  3.4 KiB/  3.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/3.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Train_Tune_Validation.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: train-tune--williamsjacr--240627-153410-54\n",
      "  job-name: train-tune-validation\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9958687970653510792\n",
      "Launched job-id: train-tune--williamsjacr--240627-153410-54\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'train-tune--williamsjacr--240627-153410-54' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'train-tune--williamsjacr--240627-153410-54' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 60 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Train_Tune_Validation.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "train-tune-v...       1  Success   06-27 15:37:44\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'train-tune--williamsjacr--240627-153410-54' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing GWAS_SumStats.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats.R\n",
    "rm(list = ls())\n",
    "\n",
    "BED_file <- commandArgs(TRUE)[1]\n",
    "BED_file <- gsub(\".bed\",\"\",BED_file)\n",
    "print(BED_file)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[3]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "anc <- commandArgs(TRUE)[4]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[5]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "tmp <- read.delim(all_train_file)\n",
    "write.table(tmp[tmp$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == anc],],file = paste0(\"All_Train_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "\n",
    "system(paste0(\"plink2 --bfile \",BED_file,\" --pheno All_Train_\",anc,\".txt --pheno-name \",trait,\" --linear --covar All_Train_\",anc,\".txt --covar-name age, age2, sex, PC1, PC2, PC3, PC4, PC5, PC6, PC7, PC8, PC9, PC10 --vif 999 --out \",OUTPUT_PATH,\"/\",trait,\"_sumstats_\",anc))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "  for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "      '--input BED_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "      '--input BIM_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "      '--input FAM_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "      '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "      '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "      '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats.R\",\n",
    "      '--env anc'=anc,\n",
    "      '--env trait'=trait,\n",
    "      '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "      check.names = FALSE\n",
    "    ))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing GWAS_SumStats.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${BED_File} ${all_phenotypes_file} ${all_train_file} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  905.0 B/  905.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/905.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--240627-154651-71\n",
      "  job-name: gwas-sumstats\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8204918055333401999\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17136902588676638775\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8833417735648720266\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13338584417530022384\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11984449143327305597\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14282491897173240579\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8331165620674023108\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2532056940772854370\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2792657320549818446\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14452553402639685255\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11008324446314991533\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9467252047295087971\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15944107210577122977\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3955676320343102319\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2048023195318102093\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15446919442826121498\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5107762276113667612\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13177588171934138516\n",
      "Launched job-id: gwas-sumst--williamsjacr--240627-154651-71\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240627-154651-71' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240627-154651-71' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 60 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name         Task  Status                          Last Update\r\n",
      "-------------  ------  ------------------------------  --------------\r\n",
      "gwas-sumstats      18  Started running \"localization\"  06-27 15:49:27\r\n",
      "gwas-sumstats      17  Started running \"localization\"  06-27 15:49:24\r\n",
      "gwas-sumstats      16  Started running \"localization\"  06-27 15:49:39\r\n",
      "gwas-sumstats      15  Started running \"localization\"  06-27 15:49:25\r\n",
      "gwas-sumstats      14  Started running \"localization\"  06-27 15:49:57\r\n",
      "gwas-sumstats      13  Started running \"localization\"  06-27 15:49:43\r\n",
      "gwas-sumstats      12  Started running \"localization\"  06-27 15:49:27\r\n",
      "gwas-sumstats      11  Started running \"localization\"  06-27 15:49:41\r\n",
      "gwas-sumstats      10  Started running \"localization\"  06-27 15:49:40\r\n",
      "gwas-sumstats       9  Started running \"localization\"  06-27 15:49:33\r\n",
      "gwas-sumstats       8  Started running \"localization\"  06-27 15:49:26\r\n",
      "gwas-sumstats       7  Started running \"localization\"  06-27 15:49:43\r\n",
      "gwas-sumstats       6  Started running \"localization\"  06-27 15:49:23\r\n",
      "gwas-sumstats       5  Started running \"localization\"  06-27 15:49:32\r\n",
      "gwas-sumstats       4  Started running \"localization\"  06-27 15:49:24\r\n",
      "gwas-sumstats       3  Started running \"localization\"  06-27 15:49:25\r\n",
      "gwas-sumstats       2  Started running \"localization\"  06-27 15:49:41\r\n",
      "gwas-sumstats       1  Started running \"localization\"  06-27 15:49:25\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240627-154651-71' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting tune_validation_anc_bed.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_anc_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "all_File <- commandArgs(TRUE)[3]\n",
    "print(all_File)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[4]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "anc <- commandArgs(TRUE)[5]\n",
    "print(anc)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Tune$IID[Tune$IID %in% All$IID[All$ancestry == anc]])\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune_\",anc),intern = TRUE)\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Valid$IID[Valid$IID %in% All$IID[All$ancestry == anc]])\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_valid_\",anc),intern = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(anc in c(\"AFR\",\"EUR\",\"AMR\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--env anc'=anc,\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input all_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/tune_validation_anc_bed.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting tune_validation_anc_bed.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_anc_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${all_File} ${BED_Full_File} ${anc} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://tune_validation_anc_bed.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.0 KiB/  1.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.0 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp tune_validation_anc_bed.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: tune-valid--williamsjacr--240627-160659-03\n",
      "  job-name: tune-validation-anc-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11205462176011724146\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13842708501326363877\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3863052655684454429\n",
      "Launched job-id: tune-valid--williamsjacr--240627-160659-03\n",
      "3 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240627-160659-03' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240627-160659-03' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 80 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script tune_validation_anc_bed.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                                      Last Update\r\n",
      "---------------  ------  ------------------------------------------  --------------\r\n",
      "tune-validat...       3  Pulling \"gcr.io/google.com/cloudsdktool...  06-27 16:07:39\r\n",
      "tune-validat...       2  Pulling \"gcr.io/google.com/cloudsdktool...  06-27 16:07:40\r\n",
      "tune-validat...       1  Pulling \"gcr.io/google.com/cloudsdktool...  06-27 16:07:37\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240627-160659-03' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting JointPRS_MUSSEL_PROSPER_BED_Files.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_MUSSEL_PROSPER_BED_Files.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "library(dplyr)\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "all_File <- commandArgs(TRUE)[3]\n",
    "print(all_File)\n",
    "\n",
    "snp_info_rsids <- commandArgs(TRUE)[4]\n",
    "print(snp_info_rsids)\n",
    "\n",
    "snp_info_extractlist <- commandArgs(TRUE)[5]\n",
    "print(snp_info_extractlist)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[6]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_chr_bim <- read.delim(paste0(BED_Full_File,\".bim\"), header=FALSE)\n",
    "all_chr_bim <- as.data.frame(all_chr_bim)\n",
    "colnames(all_chr_bim) <- c(\"CHR\",\"SNP\",\"IDK\",\"BP\",\"A1\",\"A2\")\n",
    "\n",
    "all_chr_bim$snp <- all_chr_bim$SNP\n",
    "all_chr_bim$rs_ID <- all_chr_bim$SNP\n",
    "\n",
    "all_chr_bim$SNP <- toupper(all_chr_bim$SNP)\n",
    "all_chr_bim$SNP <- gsub(\" \",\"\",all_chr_bim$SNP)\n",
    "\n",
    "SNP_GRCh37_38_match_update <- readRDS(snp_info_rsids)\n",
    "\n",
    "system(paste0(\"rm \",snp_info_rsids))\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele1_38,\":\",SNP_GRCh37_38_match_update$allele2_38)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele2_38,\":\",SNP_GRCh37_38_match_update$allele1_38)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- toupper(SNP_GRCh37_38_match_update$unique_id1)\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id1)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- toupper(SNP_GRCh37_38_match_update$unique_id2)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id2)\n",
    "\n",
    "all_chr_bim <- left_join(all_chr_bim,SNP_GRCh37_38_match_update[,c(\"unique_id1\",\"rsid\")],by = c(\"SNP\"=\"unique_id1\"))\n",
    "all_chr_bim$rs_ID[!is.na(all_chr_bim$rsid)] <- all_chr_bim$rsid[!is.na(all_chr_bim$rsid)]\n",
    "    \n",
    "all_chr_bim <- subset(all_chr_bim,select = -c(rsid))\n",
    "    \n",
    "all_chr_bim <- left_join(all_chr_bim,SNP_GRCh37_38_match_update[,c(\"unique_id2\",\"rsid\")],by = c(\"SNP\"=\"unique_id2\"))\n",
    "all_chr_bim$rs_ID[!is.na(all_chr_bim$rsid)] <- all_chr_bim$rsid[!is.na(all_chr_bim$rsid)]\n",
    "\n",
    "snp_info_extractlist <- read.csv(snp_info_extractlist)\n",
    "extract_list <- all_chr_bim$snp[all_chr_bim$rs_ID %in% snp_info_extractlist$SNP]\n",
    "write.table(extract_list,\"extract_list.txt\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "# Extract Needed SNPs in Full File\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --extract extract_list.txt --make-bed --out \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER\"),intern = TRUE)\n",
    "\n",
    "system(paste0(\"rm \",BED_Full_File,\".bed\"))\n",
    "system(paste0(\"rm \",BED_Full_File,\".bim\"))\n",
    "system(paste0(\"rm \",BED_Full_File,\".fam\"))\n",
    "\n",
    "write.table(all_chr_bim[,c(\"snp\",\"rs_ID\")],file = \"snp_rename.txt\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "# Rename SNPs to rsids\n",
    "system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --update-name snp_rename.txt 2 1 --make-bed --out \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER\"))\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Tune$IID)\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "## Full Tune\n",
    "system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune_JointPRS_MUSSEL_PROSPER\"),intern = TRUE)\n",
    "\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "# Anc Specific Tune\n",
    "for(anc in c(\"AFR\",\"EUR\",\"AMR\")){\n",
    "    Keep <- data.frame(FID = 0,IID = Tune$IID[Tune$IID %in% All$IID[All$ancestry == anc]])\n",
    "    write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "    system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune_JointPRS_MUSSEL_PROSPER_\",anc),intern = TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Valid$IID)\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "## Full Valid\n",
    "system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_valid_JointPRS_MUSSEL_PROSPER\"),intern = TRUE)\n",
    "\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "# Anc Specific Tune\n",
    "for(anc in c(\"AFR\",\"EUR\",\"AMR\")){\n",
    "    Keep <- data.frame(FID = 0,IID = Valid$IID[Valid$IID %in% All$IID[All$ancestry == anc]])\n",
    "    write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "    system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_Valid_JointPRS_MUSSEL_PROSPER_\",anc),intern = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input all_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input snp_info_rsids'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\",\n",
    "    '--input snp_info_extractlist'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_Info_JointPRS_PROSPER_MUSSEL.csv\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/JointPRS_MUSSEL_PROSPER_BED_Files.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\",\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting JointPRS_MUSSEL_PROSPER_BED_Files.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_MUSSEL_PROSPER_BED_Files.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${all_File} ${snp_info_rsids} ${snp_info_extractlist} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://JointPRS_MUSSEL_PROSPER_BED_Files.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  4.5 KiB/  4.5 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/4.5 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp JointPRS_MUSSEL_PROSPER_BED_Files.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: jointprs-m--williamsjacr--240720-133736-74\n",
      "  job-name: jointprs-mussel-prosper-bed-files\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6898823270404457997\n",
      "Launched job-id: jointprs-m--williamsjacr--240720-133736-74\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-m--williamsjacr--240720-133736-74' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-m--williamsjacr--240720-133736-74' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"168h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script JointPRS_MUSSEL_PROSPER_BED_Files.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                          Last Update\r\n",
      "---------------  ------  ------------------------------  --------------\r\n",
      "jointprs-mus...       1  Started running \"user-command\"  07-20 16:15:24\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-m--williamsjacr--240720-133736-74' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_Info_JointPRS_PROSPER_MUSSEL.csv\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr\"\r\n",
      "[1] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\"\r\n",
      " [1] \"PLINK v1.90b7.2 64-bit (11 Dec 2023)           www.cog-genomics.org/plink/1.9/\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [2] \"(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [3] \"Logging to /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.log.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      " [4] \"Options in effect:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \r\n",
      " [5] \"  --bfile /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \r\n",
      " [6] \"  --extract extract_list.txt\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      " [7] \"  --make-bed\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      " [8] \"  --out /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \r\n",
      " [9] \"\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \r\n",
      "[10] \"64311 MB RAM detected; reserving 32155 MB for main workspace.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \r\n",
      "[11] \"10486613 variants loaded from .bim file.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \r\n",
      "[12] \"245394 people (0 males, 0 females, 245394 ambiguous) loaded from .fam.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[13] \"Ambiguous sex IDs written to\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[14] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.nosex\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[15] \".\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[16] \"--extract: 1572405 variants remaining.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[17] \"Using 1 thread (no multithreaded calculations invoked).\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[18] \"Before main variant filters, 245394 founders and 0 nonfounders present.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[19] \"Calculating allele frequencies... 0%\\b\\b1%\\b\\b2%\\b\\b3%\\b\\b4%\\b\\b5%\\b\\b6%\\b\\b7%\\b\\b8%\\b\\b9%\\b\\b10%\\b\\b\\b11%\\b\\b\\b12%\\b\\b\\b13%\\b\\b\\b14%\\b\\b\\b15%\\b\\b\\b16%\\b\\b\\b17%\\b\\b\\b18%\\b\\b\\b19%\\b\\b\\b20%\\b\\b\\b21%\\b\\b\\b22%\\b\\b\\b23%\\b\\b\\b24%\\b\\b\\b25%\\b\\b\\b26%\\b\\b\\b27%\\b\\b\\b28%\\b\\b\\b29%\\b\\b\\b30%\\b\\b\\b31%\\b\\b\\b32%\\b\\b\\b33%\\b\\b\\b34%\\b\\b\\b35%\\b\\b\\b36%\\b\\b\\b37%\\b\\b\\b38%\\b\\b\\b39%\\b\\b\\b40%\\b\\b\\b41%\\b\\b\\b42%\\b\\b\\b43%\\b\\b\\b44%\\b\\b\\b45%\\b\\b\\b46%\\b\\b\\b47%\\b\\b\\b48%\\b\\b\\b49%\\b\\b\\b50%\\b\\b\\b51%\\b\\b\\b52%\\b\\b\\b53%\\b\\b\\b54%\\b\\b\\b55%\\b\\b\\b56%\\b\\b\\b57%\\b\\b\\b58%\\b\\b\\b59%\\b\\b\\b60%\\b\\b\\b61%\\b\\b\\b62%\\b\\b\\b63%\\b\\b\\b64%\\b\\b\\b65%\\b\\b\\b66%\\b\\b\\b67%\\b\\b\\b68%\\b\\b\\b69%\\b\\b\\b70%\\b\\b\\b71%\\b\\b\\b72%\\b\\b\\b73%\\b\\b\\b74%\\b\\b\\b75%\\b\\b\\b76%\\b\\b\\b77%\\b\\b\\b78%\\b\\b\\b79%\\b\\b\\b80%\\b\\b\\b81%\\b\\b\\b82%\\b\\b\\b83%\\b\\b\\b84%\\b\\b\\b85%\\b\\b\\b86%\\b\\b\\b87%\\b\\b\\b88%\\b\\b\\b89%\\b\\b\\b90%\\b\\b\\b91%\\b\\b\\b92%\\b\\b\\b93%\\b\\b\\b94%\\b\\b\\b95%\\b\\b\\b96%\\b\\b\\b97%\\b\\b\\b98%\\b\\b\\b99%\\b\\b\\b\\b done.\"\r\n",
      "[20] \"Total genotyping rate is 0.993161.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \r\n",
      "[21] \"1572405 variants and 245394 people pass filters and QC.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[22] \"Note: No phenotypes present.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[23] \"--make-bed to\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \r\n",
      "[24] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \r\n",
      "[25] \"+\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[26] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \r\n",
      "[27] \"+\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[28] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \r\n",
      "[29] \"... 0%\\b\\b1%\\b\\b2%\\b\\b3%\\b\\b4%\\b\\b5%\\b\\b6%\\b\\b7%\\b\\b8%\\b\\b9%\\b\\b10%\\b\\b\\b11%\\b\\b\\b12%\\b\\b\\b13%\\b\\b\\b14%\\b\\b\\b15%\\b\\b\\b16%\\b\\b\\b17%\\b\\b\\b18%\\b\\b\\b19%\\b\\b\\b20%\\b\\b\\b21%\\b\\b\\b22%\\b\\b\\b23%\\b\\b\\b24%\\b\\b\\b25%\\b\\b\\b26%\\b\\b\\b27%\\b\\b\\b28%\\b\\b\\b29%\\b\\b\\b30%\\b\\b\\b31%\\b\\b\\b32%\\b\\b\\b33%\\b\\b\\b34%\\b\\b\\b35%\\b\\b\\b36%\\b\\b\\b37%\\b\\b\\b38%\\b\\b\\b39%\\b\\b\\b40%\\b\\b\\b41%\\b\\b\\b42%\\b\\b\\b43%\\b\\b\\b44%\\b\\b\\b45%\\b\\b\\b46%\\b\\b\\b47%\\b\\b\\b48%\\b\\b\\b49%\\b\\b\\b50%\\b\\b\\b51%\\b\\b\\b52%\\b\\b\\b53%\\b\\b\\b54%\\b\\b\\b55%\\b\\b\\b56%\\b\\b\\b57%\\b\\b\\b58%\\b\\b\\b59%\\b\\b\\b60%\\b\\b\\b61%\\b\\b\\b62%\\b\\b\\b63%\\b\\b\\b64%\\b\\b\\b65%\\b\\b\\b66%\\b\\b\\b67%\\b\\b\\b68%\\b\\b\\b69%\\b\\b\\b70%\\b\\b\\b71%\\b\\b\\b72%\\b\\b\\b73%\\b\\b\\b74%\\b\\b\\b75%\\b\\b\\b76%\\b\\b\\b77%\\b\\b\\b78%\\b\\b\\b79%\\b\\b\\b80%\\b\\b\\b81%\\b\\b\\b82%\\b\\b\\b83%\\b\\b\\b84%\\b\\b\\b85%\\b\\b\\b86%\\b\\b\\b87%\\b\\b\\b88%\\b\\b\\b89%\\b\\b\\b90%\\b\\b\\b91%\\b\\b\\b92%\\b\\b\\b93%\\b\\b\\b94%\\b\\b\\b95%\\b\\b\\b96%\\b\\b\\b97%\\b\\b\\b98%\\b\\b\\b99%\\b\\b\\bdone.\"                                 \r\n",
      "PLINK v1.90b7.2 64-bit (11 Dec 2023)           www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.log.\r\n",
      "Options in effect:\r\n",
      "  --bfile /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\r\n",
      "  --make-bed\r\n",
      "  --out /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\r\n",
      "  --update-name snp_rename.txt 2 1\r\n",
      "\r\n",
      "64311 MB RAM detected; reserving 32155 MB for main workspace.\r\n",
      "1572405 variants loaded from .bim file.\r\n",
      "245394 people (0 males, 0 females, 245394 ambiguous) loaded from .fam.\r\n",
      "Ambiguous sex IDs written to\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.nosex\r\n",
      ".\r\n",
      "--update-name: 1572405 values updated, 8914211 variant IDs not present.\r\n",
      "Using 1 thread (no multithreaded calculations invoked).\r\n",
      "Before main variant filters, 245394 founders and 0 nonfounders present.\r\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\r\n",
      "Total genotyping rate is 0.993161.\r\n",
      "1572405 variants and 245394 people pass filters and QC.\r\n",
      "Note: No phenotypes present.\r\n",
      "--make-bed to\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\r\n",
      "+\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\r\n",
      "+\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\r\n",
      "... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/jointprs-m--williamsjacr--240720-133736-74.1-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_Info_JointPRS_PROSPER_MUSSEL.csv\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr\"\r\n",
      "[1] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\"\r\n",
      " [1] \"PLINK v1.90b7.2 64-bit (11 Dec 2023)           www.cog-genomics.org/plink/1.9/\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [2] \"(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [3] \"Logging to /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.log.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      " [4] \"Options in effect:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \r\n",
      " [5] \"  --bfile /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \r\n",
      " [6] \"  --extract extract_list.txt\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      " [7] \"  --make-bed\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      " [8] \"  --out /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \r\n",
      " [9] \"\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \r\n",
      "[10] \"64311 MB RAM detected; reserving 32155 MB for main workspace.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \r\n",
      "[11] \"10486613 variants loaded from .bim file.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \r\n",
      "[12] \"245394 people (0 males, 0 females, 245394 ambiguous) loaded from .fam.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[13] \"Ambiguous sex IDs written to\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[14] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.nosex\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[15] \".\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[16] \"--extract: 1572405 variants remaining.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[17] \"Using 1 thread (no multithreaded calculations invoked).\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[18] \"Before main variant filters, 245394 founders and 0 nonfounders present.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[19] \"Calculating allele frequencies... 0%\\b\\b1%\\b\\b2%\\b\\b3%\\b\\b4%\\b\\b5%\\b\\b6%\\b\\b7%\\b\\b8%\\b\\b9%\\b\\b10%\\b\\b\\b11%\\b\\b\\b12%\\b\\b\\b13%\\b\\b\\b14%\\b\\b\\b15%\\b\\b\\b16%\\b\\b\\b17%\\b\\b\\b18%\\b\\b\\b19%\\b\\b\\b20%\\b\\b\\b21%\\b\\b\\b22%\\b\\b\\b23%\\b\\b\\b24%\\b\\b\\b25%\\b\\b\\b26%\\b\\b\\b27%\\b\\b\\b28%\\b\\b\\b29%\\b\\b\\b30%\\b\\b\\b31%\\b\\b\\b32%\\b\\b\\b33%\\b\\b\\b34%\\b\\b\\b35%\\b\\b\\b36%\\b\\b\\b37%\\b\\b\\b38%\\b\\b\\b39%\\b\\b\\b40%\\b\\b\\b41%\\b\\b\\b42%\\b\\b\\b43%\\b\\b\\b44%\\b\\b\\b45%\\b\\b\\b46%\\b\\b\\b47%\\b\\b\\b48%\\b\\b\\b49%\\b\\b\\b50%\\b\\b\\b51%\\b\\b\\b52%\\b\\b\\b53%\\b\\b\\b54%\\b\\b\\b55%\\b\\b\\b56%\\b\\b\\b57%\\b\\b\\b58%\\b\\b\\b59%\\b\\b\\b60%\\b\\b\\b61%\\b\\b\\b62%\\b\\b\\b63%\\b\\b\\b64%\\b\\b\\b65%\\b\\b\\b66%\\b\\b\\b67%\\b\\b\\b68%\\b\\b\\b69%\\b\\b\\b70%\\b\\b\\b71%\\b\\b\\b72%\\b\\b\\b73%\\b\\b\\b74%\\b\\b\\b75%\\b\\b\\b76%\\b\\b\\b77%\\b\\b\\b78%\\b\\b\\b79%\\b\\b\\b80%\\b\\b\\b81%\\b\\b\\b82%\\b\\b\\b83%\\b\\b\\b84%\\b\\b\\b85%\\b\\b\\b86%\\b\\b\\b87%\\b\\b\\b88%\\b\\b\\b89%\\b\\b\\b90%\\b\\b\\b91%\\b\\b\\b92%\\b\\b\\b93%\\b\\b\\b94%\\b\\b\\b95%\\b\\b\\b96%\\b\\b\\b97%\\b\\b\\b98%\\b\\b\\b99%\\b\\b\\b\\b done.\"\r\n",
      "[20] \"Total genotyping rate is 0.993161.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \r\n",
      "[21] \"1572405 variants and 245394 people pass filters and QC.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[22] \"Note: No phenotypes present.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[23] \"--make-bed to\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \r\n",
      "[24] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \r\n",
      "[25] \"+\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[26] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \r\n",
      "[27] \"+\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[28] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \r\n",
      "[29] \"... 0%\\b\\b1%\\b\\b2%\\b\\b3%\\b\\b4%\\b\\b5%\\b\\b6%\\b\\b7%\\b\\b8%\\b\\b9%\\b\\b10%\\b\\b\\b11%\\b\\b\\b12%\\b\\b\\b13%\\b\\b\\b14%\\b\\b\\b15%\\b\\b\\b16%\\b\\b\\b17%\\b\\b\\b18%\\b\\b\\b19%\\b\\b\\b20%\\b\\b\\b21%\\b\\b\\b22%\\b\\b\\b23%\\b\\b\\b24%\\b\\b\\b25%\\b\\b\\b26%\\b\\b\\b27%\\b\\b\\b28%\\b\\b\\b29%\\b\\b\\b30%\\b\\b\\b31%\\b\\b\\b32%\\b\\b\\b33%\\b\\b\\b34%\\b\\b\\b35%\\b\\b\\b36%\\b\\b\\b37%\\b\\b\\b38%\\b\\b\\b39%\\b\\b\\b40%\\b\\b\\b41%\\b\\b\\b42%\\b\\b\\b43%\\b\\b\\b44%\\b\\b\\b45%\\b\\b\\b46%\\b\\b\\b47%\\b\\b\\b48%\\b\\b\\b49%\\b\\b\\b50%\\b\\b\\b51%\\b\\b\\b52%\\b\\b\\b53%\\b\\b\\b54%\\b\\b\\b55%\\b\\b\\b56%\\b\\b\\b57%\\b\\b\\b58%\\b\\b\\b59%\\b\\b\\b60%\\b\\b\\b61%\\b\\b\\b62%\\b\\b\\b63%\\b\\b\\b64%\\b\\b\\b65%\\b\\b\\b66%\\b\\b\\b67%\\b\\b\\b68%\\b\\b\\b69%\\b\\b\\b70%\\b\\b\\b71%\\b\\b\\b72%\\b\\b\\b73%\\b\\b\\b74%\\b\\b\\b75%\\b\\b\\b76%\\b\\b\\b77%\\b\\b\\b78%\\b\\b\\b79%\\b\\b\\b80%\\b\\b\\b81%\\b\\b\\b82%\\b\\b\\b83%\\b\\b\\b84%\\b\\b\\b85%\\b\\b\\b86%\\b\\b\\b87%\\b\\b\\b88%\\b\\b\\b89%\\b\\b\\b90%\\b\\b\\b91%\\b\\b\\b92%\\b\\b\\b93%\\b\\b\\b94%\\b\\b\\b95%\\b\\b\\b96%\\b\\b\\b97%\\b\\b\\b98%\\b\\b\\b99%\\b\\b\\bdone.\"                                 \r\n",
      "PLINK v1.90b7.2 64-bit (11 Dec 2023)           www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.log.\r\n",
      "Options in effect:\r\n",
      "  --bfile /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\r\n",
      "  --make-bed\r\n",
      "  --out /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\r\n",
      "  --update-name snp_rename.txt 2 1\r\n",
      "\r\n",
      "64311 MB RAM detected; reserving 32155 MB for main workspace.\r\n",
      "1572405 variants loaded from .bim file.\r\n",
      "245394 people (0 males, 0 females, 245394 ambiguous) loaded from .fam.\r\n",
      "Ambiguous sex IDs written to\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.nosex\r\n",
      ".\r\n",
      "--update-name: 1572405 values updated, 8914211 variant IDs not present.\r\n",
      "Using 1 thread (no multithreaded calculations invoked).\r\n",
      "Before main variant filters, 245394 founders and 0 nonfounders present.\r\n",
      "Calculating allele frequencies... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\b\b done.\r\n",
      "Total genotyping rate is 0.993161.\r\n",
      "1572405 variants and 245394 people pass filters and QC.\r\n",
      "Note: No phenotypes present.\r\n",
      "--make-bed to\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\r\n",
      "+\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\r\n",
      "+\r\n",
      "/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\r\n",
      "... 0%\b\b1%\b\b2%\b\b3%\b\b4%\b\b5%\b\b6%\b\b7%\b\b8%\b\b9%\b\b10%\b\b\b11%\b\b\b12%\b\b\b13%\b\b\b14%\b\b\b15%\b\b\b16%\b\b\b17%\b\b\b18%\b\b\b19%\b\b\b20%\b\b\b21%\b\b\b22%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b26%\b\b\b27%\b\b\b28%\b\b\b29%\b\b\b30%\b\b\b31%\b\b\b32%\b\b\b33%\b\b\b34%\b\b\b35%\b\b\b36%\b\b\b37%\b\b\b38%\b\b\b39%\b\b\b40%\b\b\b41%\b\b\b42%\b\b\b43%\b\b\b44%\b\b\b45%\b\b\b46%\b\b\b47%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b51%\b\b\b52%\b\b\b53%\b\b\b54%\b\b\b55%\b\b\b56%\b\b\b57%\b\b\b58%\b\b\b59%\b\b\b60%\b\b\b61%\b\b\b62%\b\b\b63%\b\b\b64%\b\b\b65%\b\b\b66%\b\b\b67%\b\b\b68%\b\b\b69%\b\b\b70%\b\b\b71%\b\b\b72%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b76%\b\b\b77%\b\b\b78%\b\b\b79%\b\b\b80%\b\b\b81%\b\b\b82%\b\b\b83%\b\b\b84%\b\b\b85%\b\b\b86%\b\b\b87%\b\b\b88%\b\b\b89%\b\b\b90%\b\b\b91%\b\b\b92%\b\b\b93%\b\b\b94%\b\b\b95%\b\b\b96%\b\b\b97%\b\b\b98%\b\b\b99%\b\b\bdone.\r\n",
      " [1] \"PLINK v1.90b7.2 64-bit (11 Dec 2023)           www.cog-genomics.org/plink/1.9/\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [2] \"(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [3] \"Logging to /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.log.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      " [4] \"Options in effect:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \r\n",
      " [5] \"  --bfile /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \r\n",
      " [6] \"  --keep Keep.txt\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      " [7] \"  --make-bed\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      " [8] \"  --out /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \r\n",
      " [9] \"\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \r\n",
      "[10] \"64311 MB RAM detected; reserving 32155 MB for main workspace.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \r\n",
      "[11] \"1572405 variants loaded from .bim file.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[12] \"245394 people (0 males, 0 females, 245394 ambiguous) loaded from .fam.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[13] \"Ambiguous sex IDs written to\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[14] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.nosex\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[15] \".\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[16] \"--keep: 33738 people remaining.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \r\n",
      "[17] \"Using 1 thread (no multithreaded calculations invoked).\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[18] \"Before main variant filters, 33738 founders and 0 nonfounders present.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[19] \"Calculating allele frequencies... 0%\\b\\b1%\\b\\b2%\\b\\b3%\\b\\b4%\\b\\b5%\\b\\b6%\\b\\b7%\\b\\b8%\\b\\b9%\\b\\b10%\\b\\b\\b11%\\b\\b\\b12%\\b\\b\\b13%\\b\\b\\b14%\\b\\b\\b15%\\b\\b\\b16%\\b\\b\\b17%\\b\\b\\b18%\\b\\b\\b19%\\b\\b\\b20%\\b\\b\\b21%\\b\\b\\b22%\\b\\b\\b23%\\b\\b\\b24%\\b\\b\\b25%\\b\\b\\b26%\\b\\b\\b27%\\b\\b\\b28%\\b\\b\\b29%\\b\\b\\b30%\\b\\b\\b31%\\b\\b\\b32%\\b\\b\\b33%\\b\\b\\b34%\\b\\b\\b35%\\b\\b\\b36%\\b\\b\\b37%\\b\\b\\b38%\\b\\b\\b39%\\b\\b\\b40%\\b\\b\\b41%\\b\\b\\b42%\\b\\b\\b43%\\b\\b\\b44%\\b\\b\\b45%\\b\\b\\b46%\\b\\b\\b47%\\b\\b\\b48%\\b\\b\\b49%\\b\\b\\b50%\\b\\b\\b51%\\b\\b\\b52%\\b\\b\\b53%\\b\\b\\b54%\\b\\b\\b55%\\b\\b\\b56%\\b\\b\\b57%\\b\\b\\b58%\\b\\b\\b59%\\b\\b\\b60%\\b\\b\\b61%\\b\\b\\b62%\\b\\b\\b63%\\b\\b\\b64%\\b\\b\\b65%\\b\\b\\b66%\\b\\b\\b67%\\b\\b\\b68%\\b\\b\\b69%\\b\\b\\b70%\\b\\b\\b71%\\b\\b\\b72%\\b\\b\\b73%\\b\\b\\b74%\\b\\b\\b75%\\b\\b\\b76%\\b\\b\\b77%\\b\\b\\b78%\\b\\b\\b79%\\b\\b\\b80%\\b\\b\\b81%\\b\\b\\b82%\\b\\b\\b83%\\b\\b\\b84%\\b\\b\\b85%\\b\\b\\b86%\\b\\b\\b87%\\b\\b\\b88%\\b\\b\\b89%\\b\\b\\b90%\\b\\b\\b91%\\b\\b\\b92%\\b\\b\\b93%\\b\\b\\b94%\\b\\b\\b95%\\b\\b\\b96%\\b\\b\\b97%\\b\\b\\b98%\\b\\b\\b99%\\b\\b\\b\\b done.\"\r\n",
      "[20] \"Total genotyping rate in remaining samples is 0.993159.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \r\n",
      "[21] \"1572405 variants and 33738 people pass filters and QC.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "[22] \"Note: No phenotypes present.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \r\n",
      "[23] \"--make-bed to\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \r\n",
      "[24] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.bed\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \r\n",
      "[25] \"+\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[26] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.bim\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \r\n",
      "[27] \"+\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \r\n",
      "[28] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.fam\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \r\n",
      "[29] \"... 0%\\b\\b1%\\b\\b2%\\b\\b3%\\b\\b4%\\b\\b5%\\b\\b6%\\b\\b7%\\b\\b8%\\b\\b9%\\b\\b10%\\b\\b\\b11%\\b\\b\\b12%\\b\\b\\b13%\\b\\b\\b14%\\b\\b\\b15%\\b\\b\\b16%\\b\\b\\b17%\\b\\b\\b18%\\b\\b\\b19%\\b\\b\\b20%\\b\\b\\b21%\\b\\b\\b22%\\b\\b\\b23%\\b\\b\\b24%\\b\\b\\b25%\\b\\b\\b26%\\b\\b\\b27%\\b\\b\\b28%\\b\\b\\b29%\\b\\b\\b30%\\b\\b\\b31%\\b\\b\\b32%\\b\\b\\b33%\\b\\b\\b34%\\b\\b\\b35%\\b\\b\\b36%\\b\\b\\b37%\\b\\b\\b38%\\b\\b\\b39%\\b\\b\\b40%\\b\\b\\b41%\\b\\b\\b42%\\b\\b\\b43%\\b\\b\\b44%\\b\\b\\b45%\\b\\b\\b46%\\b\\b\\b47%\\b\\b\\b48%\\b\\b\\b49%\\b\\b\\b50%\\b\\b\\b51%\\b\\b\\b52%\\b\\b\\b53%\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \r\n",
      "attr(,\"status\")\r\n",
      "[1] 6\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/jointprs-m--williamsjacr--240715-165205-70.1-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_bed.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[3]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Tune$IID)\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune\"),intern = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Valid$IID)\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_valid\"),intern = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/tune_validation_bed.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_bed.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://tune_validation_bed.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  823.0 B/  823.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/823.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp tune_validation_bed.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: tune-valid--williamsjacr--240629-011433-61\n",
      "  job-name: tune-validation-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15401220709288402463\n",
      "Launched job-id: tune-valid--williamsjacr--240629-011433-61\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240629-011433-61' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240629-011433-61' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 40 \\\n",
    "  --min-ram 80 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script tune_validation_bed.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "tune-validat...       1  Pending   06-29 01:14:33\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240629-011433-61' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_phenotypes.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_phenotypes.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "all_File <- commandArgs(TRUE)[3]\n",
    "print(all_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "covar_tmp <- data.frame(FID = Tune$FID,IID = Tune$IID,age = Tune$age,age2 = Tune$age2,sex = Tune$sex,\n",
    "                        PC1 = Tune$PC1,PC2 = Tune$PC2,PC3 = Tune$PC3,PC4 = Tune$PC4,PC5 = Tune$PC5,\n",
    "                        PC6 = Tune$PC6,PC7 = Tune$PC7,PC8 = Tune$PC8,PC9 = Tune$PC9,PC10 = Tune$PC10)\n",
    "\n",
    "write.table(covar_tmp,file = paste0(OUTPUT_PATH,\"/full_tune_covar.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    covar_tmp_anc <- covar_tmp[covar_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "    write.table(covar_tmp_anc,file = paste0(OUTPUT_PATH,\"/full_tune_covar_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "}\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    fam_tmp <- data.frame(FID = Tune$FID,IID = Tune$IID,Y = Tune[,trait])\n",
    "    write.table(fam_tmp,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_tune_y.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "    for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "        fam_tmp_anc <- fam_tmp[fam_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "        write.table(fam_tmp_anc,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_tune_y_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "covar_tmp <- data.frame(FID = Valid$FID,IID = Valid$IID,age = Valid$age,age2 = Valid$age2,sex = Valid$sex,\n",
    "                        PC1 = Valid$PC1,PC2 = Valid$PC2,PC3 = Valid$PC3,PC4 = Valid$PC4,PC5 = Valid$PC5,\n",
    "                        PC6 = Valid$PC6,PC7 = Valid$PC7,PC8 = Valid$PC8,PC9 = Valid$PC9,PC10 = Valid$PC10)\n",
    "\n",
    "write.table(covar_tmp,file = paste0(OUTPUT_PATH,\"/full_valid_covar.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    covar_tmp_anc <- covar_tmp[covar_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "    write.table(covar_tmp_anc,file = paste0(OUTPUT_PATH,\"/full_valid_covar_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "}\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    fam_tmp <- data.frame(FID = Valid$FID,IID = Valid$IID,Y = Valid[,trait])\n",
    "    write.table(fam_tmp,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_valid_y.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "    for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "        fam_tmp_anc <- fam_tmp[fam_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "        write.table(fam_tmp_anc,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_valid_y_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input all_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/tune_validation_phenotypes.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes\",\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_phenotypes.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_phenotypes.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${all_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://tune_validation_phenotypes.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.8 KiB/  2.8 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.8 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp tune_validation_phenotypes.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: tune-valid--williamsjacr--240722-175637-83\n",
      "  job-name: tune-validation-phenotypes\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16970955615058126312\n",
      "Launched job-id: tune-valid--williamsjacr--240722-175637-83\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240722-175637-83' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240722-175637-83' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 40 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 8 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script tune_validation_phenotypes.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "tune-validat...       1  Pending   07-22 17:56:37\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240722-175637-83' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean.R\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "\n",
    "anc <- commandArgs(TRUE)[2]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[3]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "sumstats <- read.delim(sumstats, header=FALSE, comment.char=\"#\")\n",
    "colnames(sumstats) <- c(\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"PROVISIONAL_REF\",\"A1\",\"OMITTED\",\"A1_FREQ\",\"TEST\",\"OBS_CT\",\"BETA\",\"SE\",\"T_STAT\",\"P\",\"ERRCODE\")\n",
    "sumstats <- sumstats[sumstats$TEST == \"ADD\",]\n",
    "sumstats <- sumstats[,c(\"CHROM\",\"ID\",\"POS\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\")\n",
    "write.table(sumstats, file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing GWAS_SumStats_Clean.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"EUR\"\n",
    "trait <- \"HDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AMR\",\"AFR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"Height\",\"TC\",\"logTG\")){\n",
    "     tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",trait,\"_sumstats_\",anc,\".\",trait,\".glm.linear\"),\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats_Clean.R\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats_Clean.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  746.0 B/  746.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/746.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats_Clean.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--240628-124107-15\n",
      "  job-name: gwas-sumstats-clean\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11902671578390765630\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14232277539497694331\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9176029849610361544\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11783176353215920730\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9411797687353540377\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17709795272143767001\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12861438726488854482\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11276180165803765558\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17092567833679330304\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11932236714567507877\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9876759812377941978\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12363945166688179911\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6647186722262533557\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12389200928970243072\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4860045537164910537\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11910010614503323706\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3202666522087632644\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14821564890709452870\n",
      "Launched job-id: gwas-sumst--williamsjacr--240628-124107-15\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240628-124107-15' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240628-124107-15' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 60 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats_Clean.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "gwas-sumstat...      18  Success   06-28 13:18:49\r\n",
      "gwas-sumstat...      17  Success   06-28 13:19:27\r\n",
      "gwas-sumstat...      16  Success   06-28 13:22:32\r\n",
      "gwas-sumstat...      15  Success   06-28 13:19:52\r\n",
      "gwas-sumstat...      14  Success   06-28 13:20:00\r\n",
      "gwas-sumstat...      13  Success   06-28 13:22:37\r\n",
      "gwas-sumstat...      12  Success   06-28 13:19:44\r\n",
      "gwas-sumstat...      11  Success   06-28 13:17:56\r\n",
      "gwas-sumstat...      10  Success   06-28 13:15:51\r\n",
      "gwas-sumstat...       9  Success   06-28 13:20:56\r\n",
      "gwas-sumstat...       8  Success   06-28 13:19:23\r\n",
      "gwas-sumstat...       7  Success   06-28 13:23:57\r\n",
      "gwas-sumstat...       6  Success   06-28 13:20:20\r\n",
      "gwas-sumstat...       5  Success   06-28 13:22:08\r\n",
      "gwas-sumstat...       4  Success   06-28 13:17:51\r\n",
      "gwas-sumstat...       3  Success   06-28 13:25:17\r\n",
      "gwas-sumstat...       2  Success   06-28 13:19:29\r\n",
      "gwas-sumstat...       1  Success   06-28 13:19:24\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240628-124107-15' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean_QQPlots.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean_QQPlots.R\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "\n",
    "anc <- commandArgs(TRUE)[2]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[3]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "sumstats <- read.delim(sumstats, header=FALSE, comment.char=\"#\")\n",
    "colnames(sumstats) <- c(\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"PROVISIONAL_REF\",\"A1\",\"OMITTED\",\"A1_FREQ\",\"TEST\",\"OBS_CT\",\"BETA\",\"SE\",\"T_STAT\",\"P\",\"ERRCODE\")\n",
    "sumstats <- sumstats[sumstats$TEST == \"ADD\",]\n",
    "sumstats <- sumstats[,c(\"CHROM\",\"ID\",\"POS\",\"A1_FREQ\",\"P\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"A1_FREQ\",\"P\")\n",
    "write.table(sumstats, file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned_QQPlots\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean_QQPlots.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean_QQPlots.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"EUR\"\n",
    "trait <- \"HDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AMR\",\"AFR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"Height\",\"TC\",\"logTG\")){\n",
    "     tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",trait,\"_sumstats_\",anc,\".\",trait,\".glm.linear\"),\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats_Clean_QQPlots.R\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats_Clean_QQPlots.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  698.0 B/  698.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/698.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats_Clean_QQPlots.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--240813-154726-79\n",
      "  job-name: gwas-sumstats-clean-qqplots\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5519588274004611166\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7476960220488265516\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6735816343148071838\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13156461183141985096\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5264466287709118151\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11897381666247388351\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17490078748917350948\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12347047526151723993\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18172776244062677380\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8614277087355113269\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10496336135676992825\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12995564038882800843\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15335774630137363136\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4457741117537690394\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7561312576099810793\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9187935349901062374\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2690730917378102129\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2110307076805893207\n",
      "Launched job-id: gwas-sumst--williamsjacr--240813-154726-79\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240813-154726-79' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240813-154726-79' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 60 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats_Clean_QQPlots.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "gwas-sumstat...      18  Success   08-13 16:26:08\r\n",
      "gwas-sumstat...      17  Success   08-13 16:25:11\r\n",
      "gwas-sumstat...      16  Success   08-13 16:25:27\r\n",
      "gwas-sumstat...      15  Success   08-13 16:23:44\r\n",
      "gwas-sumstat...      14  Success   08-13 16:25:01\r\n",
      "gwas-sumstat...      13  Success   08-13 16:22:28\r\n",
      "gwas-sumstat...      12  Success   08-13 16:22:16\r\n",
      "gwas-sumstat...      11  Success   08-13 16:21:35\r\n",
      "gwas-sumstat...      10  Success   08-13 16:25:12\r\n",
      "gwas-sumstat...       9  Success   08-13 16:29:40\r\n",
      "gwas-sumstat...       8  Success   08-13 16:36:59\r\n",
      "gwas-sumstat...       7  Success   08-13 16:22:48\r\n",
      "gwas-sumstat...       6  Success   08-13 16:25:23\r\n",
      "gwas-sumstat...       5  Success   08-13 16:24:32\r\n",
      "gwas-sumstat...       4  Success   08-13 16:22:15\r\n",
      "gwas-sumstat...       3  Success   08-13 16:23:24\r\n",
      "gwas-sumstat...       2  Success   08-13 16:23:29\r\n",
      "gwas-sumstat...       1  Success   08-13 16:25:01\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--240813-154726-79' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/data/script/GWAS_SumStats_Clean_QQPlots.sh: line 6:     7 Killed                  Rscript ${R_Script} ${sumstats} ${anc} ${trait} ${OUTPUT_PATH}\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/gwas-sumst--williamsjacr--240813-144353-97.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Add_RSIDs_Sumstats.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Add_RSIDs_Sumstats.R\n",
    "rm(list = ls())\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "print(sumstats)\n",
    "\n",
    "snp_info <- commandArgs(TRUE)[2]\n",
    "print(snp_info)\n",
    "\n",
    "anc <- commandArgs(TRUE)[3]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[4]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "SNP_GRCh37_38_match_update <- readRDS(snp_info)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele1_38,\":\",SNP_GRCh37_38_match_update$allele2_38)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele2_38,\":\",SNP_GRCh37_38_match_update$allele1_38)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- toupper(SNP_GRCh37_38_match_update$unique_id1)\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id1)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- toupper(SNP_GRCh37_38_match_update$unique_id2)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id2)\n",
    "\n",
    "\n",
    "sumstats <- fread(sumstats)\n",
    "    \n",
    "sumstats <- as.data.frame(sumstats)\n",
    "sumstats <- sumstats[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "    \n",
    "sumstats$SNP <- toupper(sumstats$SNP)\n",
    "sumstats$SNP <- gsub(\" \",\"\",sumstats$SNP)\n",
    "    \n",
    "sumstats <- left_join(sumstats,SNP_GRCh37_38_match_update[,c(\"unique_id1\",\"rsid\")],by = c(\"SNP\"=\"unique_id1\"))\n",
    "sumstats$rs_ID[!is.na(sumstats$rsid)] <- sumstats$rsid[!is.na(sumstats$rsid)]\n",
    "    \n",
    "sumstats <- subset(sumstats,select = -c(rsid))\n",
    "    \n",
    "sumstats <- left_join(sumstats,SNP_GRCh37_38_match_update[,c(\"unique_id2\",\"rsid\")],by = c(\"SNP\"=\"unique_id2\"))\n",
    "sumstats$rs_ID[!is.na(sumstats$rsid)] <- sumstats$rsid[!is.na(sumstats$rsid)]\n",
    "    \n",
    "sumstats <- sumstats[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "    \n",
    "sumstats$SNP <- sumstats$rs_ID\n",
    "    \n",
    "doubles <- names(which(table(sumstats$rs_ID) > 1))\n",
    "    \n",
    "sumstats <- sumstats[!(sumstats$SNP %in% doubles),]\n",
    "    \n",
    "write.csv(sumstats,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_RSIDs.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Add_RSIDs_Sumstats.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Add_RSIDs_Sumstats.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${snp_info} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input snp_info'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Add_RSIDs_Sumstats.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Add_RSIDs_Sumstats.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.3 KiB/  2.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Add_RSIDs_Sumstats.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: add-rsids---williamsjacr--240628-133112-38\n",
      "  job-name: add-rsids-sumstats\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/596445960202423148\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10680619813942882207\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11523552278899343275\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14699910471220127595\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14751023867850149884\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10842220169385773957\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2424441718384203511\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4459686452747780496\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2575789168063597997\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2438685205286195760\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4567010862885207510\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17495988694034571192\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17978034972007395152\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10961078861505061882\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9896934920230142708\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18119545141173856538\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4509579815050431866\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1738477489566109627\n",
      "Launched job-id: add-rsids---williamsjacr--240628-133112-38\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'add-rsids---williamsjacr--240628-133112-38' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'add-rsids---williamsjacr--240628-133112-38' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 80 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Add_RSIDs_Sumstats.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                          Last Update\r\n",
      "---------------  ------  ------------------------------  --------------\r\n",
      "add-rsids-su...      18  Started running \"user-command\"  06-28 13:34:13\r\n",
      "add-rsids-su...      17  Started running \"user-command\"  06-28 13:34:04\r\n",
      "add-rsids-su...      16  Started running \"user-command\"  06-28 13:34:16\r\n",
      "add-rsids-su...      15  Started running \"user-command\"  06-28 13:34:12\r\n",
      "add-rsids-su...      14  Started running \"user-command\"  06-28 13:34:25\r\n",
      "add-rsids-su...      13  Started running \"user-command\"  06-28 13:34:22\r\n",
      "add-rsids-su...      12  Started running \"user-command\"  06-28 13:34:05\r\n",
      "add-rsids-su...      11  Started running \"user-command\"  06-28 13:34:05\r\n",
      "add-rsids-su...      10  Started running \"user-command\"  06-28 13:34:09\r\n",
      "add-rsids-su...       9  Started running \"user-command\"  06-28 13:34:22\r\n",
      "add-rsids-su...       8  Started running \"user-command\"  06-28 13:34:09\r\n",
      "add-rsids-su...       7  Started running \"user-command\"  06-28 13:34:04\r\n",
      "add-rsids-su...       6  Started running \"user-command\"  06-28 13:34:21\r\n",
      "add-rsids-su...       5  Started running \"user-command\"  06-28 13:34:08\r\n",
      "add-rsids-su...       4  Started running \"user-command\"  06-28 13:34:14\r\n",
      "add-rsids-su...       3  Started running \"user-command\"  06-28 13:34:02\r\n",
      "add-rsids-su...       2  Started running \"user-command\"  06-28 13:34:28\r\n",
      "add-rsids-su...       1  Started running \"user-command\"  06-28 13:34:03\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'add-rsids---williamsjacr--240628-133112-38' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_SumStats_Format.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_SumStats_Format.R\n",
    "rm(list = ls())\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "print(sumstats)\n",
    "\n",
    "ref_bim <- commandArgs(TRUE)[2]\n",
    "print(ref_bim)\n",
    "\n",
    "anc <- commandArgs(TRUE)[3]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[4]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "sumstats <- fread(sumstats)\n",
    "    \n",
    "sumstats <- as.data.frame(sumstats)\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "\n",
    "sumstats <- sumstats[,c(\"rs_ID\",\"CHR\",\"ALT\",\"REF\",\"BETA\",\"SE\",\"OBS_CT\")]\n",
    "colnames(sumstats) <- c(\"rsid\",\"chr\",\"a1\",\"a0\",\"beta\",\"beta_se\",\"n_eff\")\n",
    "\n",
    "sumstats$n_eff <- as.numeric(sumstats$n_eff)\n",
    "\n",
    "ref_bim <- fread(ref_bim)\n",
    "ref_bim <- as.data.frame(ref_bim)\n",
    "\n",
    "sumstats <- sumstats[sumstats$rsid %in% ref_bim[,2],]\n",
    "\n",
    "ref_bim <- ref_bim[ref_bim[,2] %in% sumstats$rsid,]\n",
    "\n",
    "order_vec <- match(sumstats$rsid,ref_bim[,2])\n",
    "\n",
    "ref_bim <- ref_bim[order_vec,]\n",
    "\n",
    "a <- which((sumstats$a0 != ref_bim[,6]) & (sumstats$a1 != ref_bim[,6]))\n",
    "\n",
    "b <- which((sumstats$a0 != ref_bim[,5]) & (sumstats$a1 != ref_bim[,5]))\n",
    "\n",
    "sumstats$a0[a] <- ref_bim$V5[a]\n",
    "sumstats$a1[a] <- ref_bim$V6[a]\n",
    "\n",
    "b <- which((sumstats$a0 != ref_bim[,5]) & (sumstats$a1 != ref_bim[,5]))\n",
    "\n",
    "if(length(b) > 0){\n",
    "  sumstats$a0[b] <- ref_bim$V5[b]\n",
    "  sumstats$a1[b] <- ref_bim$V6[b]\n",
    "}\n",
    "\n",
    "print(summary(sumstats$n_eff))\n",
    "\n",
    "sumstats <- sumstats[sumstats$n_eff > (median(sumstats$n_eff)*.9),]\n",
    "\n",
    "write.csv(sumstats,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_SumStats_Format.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_SumStats_Format.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${ref_bim} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_RSIDs.csv\"),\n",
    "    '--input ref_bim'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER/ref_bim.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_SumStats_Format.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_SumStats_Format.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.4 KiB/  1.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_SumStats_Format.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-su--williamsjacr--240721-124238-39\n",
      "  job-name: prosper-sumstats-format\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5943329446903232607\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10164156545079215933\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6012435391604352405\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5775686596901906735\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3353824575108734335\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17537601535375617419\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6819809204638678646\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16083125786427904745\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14826923869042126249\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15580849815880190842\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1930944712024022465\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10753895111553124555\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4617382583028289194\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2406795728784846537\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7103877516864775440\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17738116763106711491\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10816255693716591330\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13688461453926655159\n",
      "Launched job-id: prosper-su--williamsjacr--240721-124238-39\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-su--williamsjacr--240721-124238-39' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-su--williamsjacr--240721-124238-39' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 8 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_SumStats_Format.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "prosper-sums...      18  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      17  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      16  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      15  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      14  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      13  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      12  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      11  Pending   07-21 12:42:39\r\n",
      "prosper-sums...      10  Pending   07-21 12:42:39\r\n",
      "prosper-sums...       9  Pending   07-21 12:42:39\r\n",
      "prosper-sums...       8  Pending   07-21 12:42:39\r\n",
      "prosper-sums...       7  Pending   07-21 12:42:38\r\n",
      "prosper-sums...       6  Pending   07-21 12:42:38\r\n",
      "prosper-sums...       5  Pending   07-21 12:42:38\r\n",
      "prosper-sums...       4  Pending   07-21 12:42:38\r\n",
      "prosper-sums...       3  Pending   07-21 12:42:38\r\n",
      "prosper-sums...       2  Pending   07-21 12:42:38\r\n",
      "prosper-sums...       1  Pending   07-21 12:42:38\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-su--williamsjacr--240721-124238-39' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting MUSSEL_SumStats_Format.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile MUSSEL_SumStats_Format.R\n",
    "rm(list = ls())\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "print(sumstats)\n",
    "\n",
    "ref_bim <- commandArgs(TRUE)[2]\n",
    "print(ref_bim)\n",
    "\n",
    "anc <- commandArgs(TRUE)[3]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[4]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "sumstats <- fread(sumstats)\n",
    "    \n",
    "sumstats <- as.data.frame(sumstats)\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "\n",
    "sumstats <- sumstats[,c(\"rs_ID\",\"CHR\",\"ALT\",\"REF\",\"BETA\",\"SE\",\"OBS_CT\")]\n",
    "colnames(sumstats) <- c(\"rsid\",\"chr\",\"a1\",\"a0\",\"beta\",\"beta_se\",\"n_eff\")\n",
    "\n",
    "sumstats$n_eff <- as.numeric(sumstats$n_eff)\n",
    "\n",
    "ref_bim <- fread(ref_bim)\n",
    "ref_bim <- as.data.frame(ref_bim)\n",
    "\n",
    "sumstats <- sumstats[sumstats$rsid %in% ref_bim[,2],]\n",
    "\n",
    "ref_bim <- ref_bim[ref_bim[,2] %in% sumstats$rsid,]\n",
    "\n",
    "order_vec <- match(sumstats$rsid,ref_bim[,2])\n",
    "\n",
    "ref_bim <- ref_bim[order_vec,]\n",
    "\n",
    "a <- which((sumstats$a0 != ref_bim[,6]) & (sumstats$a1 != ref_bim[,6]))\n",
    "\n",
    "b <- which((sumstats$a0 != ref_bim[,5]) & (sumstats$a1 != ref_bim[,5]))\n",
    "\n",
    "sumstats$a0[a] <- ref_bim$V5[a]\n",
    "sumstats$a1[a] <- ref_bim$V6[a]\n",
    "\n",
    "b <- which((sumstats$a0 != ref_bim[,5]) & (sumstats$a1 != ref_bim[,5]))\n",
    "\n",
    "if(length(b) > 0){\n",
    "  sumstats$a0[b] <- ref_bim$V5[b]\n",
    "  sumstats$a1[b] <- ref_bim$V6[b]\n",
    "}\n",
    "\n",
    "print(summary(sumstats$n_eff))\n",
    "\n",
    "sumstats <- sumstats[sumstats$n_eff > (median(sumstats$n_eff)*.9),]\n",
    "\n",
    "write.csv(sumstats,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting MUSSEL_SumStats_Format.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile MUSSEL_SumStats_Format.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${ref_bim} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_RSIDs.csv\"),\n",
    "    '--input ref_bim'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/MUSSEL/ref_bim.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL_SumStats_Format.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://MUSSEL_SumStats_Format.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.4 KiB/  1.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp MUSSEL_SumStats_Format.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: mussel-sum--williamsjacr--240722-122648-79\n",
      "  job-name: mussel-sumstats-format\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18368647893015607889\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11161361928737729861\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3139711350381230008\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6110467123046620153\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1005811247559210268\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10423943641961380668\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15924590537914389518\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3826222295720045467\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11238638565207718403\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11463398504302800953\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1442354105770765068\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10090876728245039343\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12314980190451875002\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17945156052511371940\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3345148964388653604\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10822960612426980708\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13232446326907583331\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14149867169014191577\n",
      "Launched job-id: mussel-sum--williamsjacr--240722-122648-79\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-sum--williamsjacr--240722-122648-79' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-sum--williamsjacr--240722-122648-79' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 8 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script MUSSEL_SumStats_Format.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "mussel-sumst...      18  Pending   07-22 12:26:50\r\n",
      "mussel-sumst...      17  Pending   07-22 12:26:50\r\n",
      "mussel-sumst...      16  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...      15  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...      14  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...      13  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...      12  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...      11  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...      10  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       9  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       8  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       7  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       6  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       5  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       4  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       3  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       2  Pending   07-22 12:26:49\r\n",
      "mussel-sumst...       1  Pending   07-22 12:26:48\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-sum--williamsjacr--240722-122648-79' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-21 12:46:42 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/mussel-sum--williamsjacr--240721-124323-60.1.log\r\n",
      "2024-07-21 12:46:42 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\r\n",
      "2024-07-21 12:46:42 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts\r\n",
      "2024-07-21 12:46:42 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL\r\n",
      "2024-07-21 12:46:42 INFO: mkdir -m 777 -p /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/\r\n",
      "2024-07-21 12:46:43 INFO: Localizing sumstats\r\n",
      "2024-07-21 12:46:43 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/EUR_BMI_GWAS_SumStats_RSIDs.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/EUR_BMI_GWAS_SumStats_RSIDs.csv\r\n",
      "2024-07-21 12:47:06 INFO: Localizing ref_bim\r\n",
      "2024-07-21 12:47:06 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-07-21 12:47:09 WARNING: Sleeping 10s before the next attempt of failed gsutil command\r\n",
      "2024-07-21 12:47:09 WARNING: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "2024-07-21 12:47:19 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-07-21 12:47:20 WARNING: Sleeping 10s before the next attempt of failed gsutil command\r\n",
      "2024-07-21 12:47:20 WARNING: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "2024-07-21 12:47:30 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-07-21 12:47:32 WARNING: Sleeping 10s before the next attempt of failed gsutil command\r\n",
      "2024-07-21 12:47:32 WARNING: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "2024-07-21 12:47:42 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-07-21 12:47:43 ERROR: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/ref_bim.txt\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/mussel-sum--williamsjacr--240721-124323-60.1.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting CT_SLEB_dimCT_prseb.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_dimCT_prseb.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(devtools)\n",
    "library(CTSLEB)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "BED_Tune_File <- commandArgs(TRUE)[1]\n",
    "BED_Tune_File <- gsub(\".bed\",\"\",BED_Tune_File)\n",
    "print(BED_Tune_File)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[2]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "#EUR\n",
    "BED_Ref_pop1_File <- commandArgs(TRUE)[3]\n",
    "BED_Ref_pop1_File <- gsub(\".bed\",\"\",BED_Ref_pop1_File)\n",
    "print(BED_Ref_pop1_File)\n",
    "\n",
    "#ANC\n",
    "BED_Ref_pop2_File <- commandArgs(TRUE)[4]\n",
    "BED_Ref_pop2_File <- gsub(\".bed\",\"\",BED_Ref_pop2_File)\n",
    "print(BED_Ref_pop2_File)\n",
    "\n",
    "pop1_sumstats <- commandArgs(TRUE)[5]\n",
    "print(pop1_sumstats)\n",
    "\n",
    "pop2_sumstats <- commandArgs(TRUE)[6]\n",
    "print(pop2_sumstats)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[7]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[8]\n",
    "print(all_tune_file)\n",
    "\n",
    "anc <- commandArgs(TRUE)[9]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[10]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[11]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "system(\"mkdir CTSLEB_TestCode\")\n",
    "\n",
    "sum_pop1 <- fread(pop1_sumstats)\n",
    "    \n",
    "sum_pop1 <- as.data.frame(sum_pop1)\n",
    "sum_pop1 <- sum_pop1[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(sum_pop1) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "sum_pop1 <- sum_pop1[,c(\"CHR\",\"SNP\",\"BP\",\"A1\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "\n",
    "sum_pop1$P <- as.numeric(sum_pop1$P)\n",
    "sum_pop1 <- sum_pop1[which(sum_pop1$P < 0.1),]\n",
    "\n",
    "sum_pop2 <- fread(pop2_sumstats)\n",
    "    \n",
    "sum_pop2 <- as.data.frame(sum_pop2)\n",
    "sum_pop2 <- sum_pop2[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(sum_pop2) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "sum_pop2 <- sum_pop2[,c(\"CHR\",\"SNP\",\"BP\",\"A1\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "\n",
    "sum_pop2$P <- as.numeric(sum_pop2$P)\n",
    "sum_pop2 <- sum_pop2[which(sum_pop2$P < 0.1),]\n",
    "\n",
    "print(nrow(sum_pop1));print(nrow(sum_pop2))\n",
    "\n",
    "PRS_farm <- SetParamsFarm(plink19_exec = \"plink\",plink2_exec = \"plink2\",mem = 80490,pthres = c(5E-08,5E-07,5E-06,5E-05,5E-04,5E-03,0.05,0.1))\n",
    "\n",
    "prs_mat <- dimCT(results_dir = \"CTSLEB_TestCode/\",sum_target = sum_pop2,sum_ref = sum_pop1,\n",
    "                 ref_plink = BED_Ref_pop1_File,\n",
    "                 target_plink = BED_Ref_pop2_File,\n",
    "                 test_target_plink = BED_Tune_File,\n",
    "                 out_prefix = \"dimCT\",\n",
    "                 params_farm = PRS_farm)\n",
    "\n",
    "# save(prs_mat,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_prs_mat_dimCT.RData\"))\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,prs_mat,by = \"IID\")\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "\n",
    "n.total.prs <- length(pthres)^2*length(r2_vec)*length(wc_base_vec)\n",
    "prs_r2_vec_test <- rep(0,n.total.prs)\n",
    "indices <- (ncol(pheno_tuning) - n.total.prs + 1):ncol(pheno_tuning)\n",
    "\n",
    "for(p_ind in 1:n.total.prs){\n",
    "  prs <- pheno_tuning[!is.na(pheno_tuning[,trait]),indices[p_ind]]\n",
    "  model <- lm(model.null$residual~prs)\n",
    "  prs_r2_vec_test[p_ind] <- summary(model)$r.square\n",
    "}\n",
    "\n",
    "max_ind <- which.max(prs_r2_vec_test)\n",
    "\n",
    "print(colnames(prs_mat)[indices[max_ind]])\n",
    "\n",
    "best_snps <- colnames(prs_mat)[indices[max_ind]]\n",
    "#calculate eb effect using EB coefficients\n",
    "prs_mat_eb <- CalculateEBEffectSize(bfile = BED_Full_File,\n",
    "                                   snp_ind = best_snps,\n",
    "                                   plink_list = plink_list,\n",
    "                                   out_prefix = \"EB\",\n",
    "                                   results_dir = \"CTSLEB_TestCode/\",\n",
    "                                   params_farm = PRS_farm)\n",
    "\n",
    "\n",
    "\n",
    "score_full <- matrix(NA,ncol = ncol(prs_mat_eb),nrow = nrow(unique_infor_post))\n",
    "names_score_full <- NULL\n",
    "count <- 1\n",
    "for(i in 1:length(pthres)){\n",
    "  for(j in 1:length(pthres)){\n",
    "    for(k in 1:24){\n",
    "      tmp <- matrix(0,nrow = nrow(unique_infor_post),ncol = 1)\n",
    "      tmp[which(unique_infor_post$P < pthres[j] | unique_infor_post$P_ref < pthres[i]),1] <- plink_list_eb$scores_eb[which(unique_infor_post$P < pthres[j] | unique_infor_post$P_ref < pthres[i]),k + 2]\n",
    "      names_score_full <- c(names_score_full,paste0(colnames(plink_list_eb$scores_eb)[k + 2],\"_p_other_\",pthres[i],\"_p_tar_\",pthres[j]))\n",
    "      score_full[,count] <- tmp[,1]\n",
    "      count <- count + 1\n",
    "    }\n",
    "  }\n",
    "}\n",
    "  \n",
    "final <- data.frame(SNP = unique_infor_post$SNP,A1 = unique_infor_post$A1,score_full)\n",
    "colnames(final) <- c(\"SNP\",\"A1\",names_score_full)\n",
    "\n",
    "\n",
    "save(final,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_betas_eb.RData\"))\n",
    "save(prs_mat_eb,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_prs_mat_eb.RData\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting CT_SLEB_dimCT_prseb.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_dimCT_prseb.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${BED_Tune_File} ${BED_Full_File} ${BED_Ref_pop1_File} ${BED_Ref_pop2_File} ${pop1_sumstats} ${pop2_sumstats} ${all_phenotypes_file} ${all_tune_file} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input BED_Tune_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles/all_tune_\",anc,\".bed\"),\n",
    "    '--input BIM_Tune_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles/all_tune_\",anc,\".bim\"),\n",
    "    '--input FAM_Tune_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles/all_tune_\",anc,\".fam\"),\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input BED_Ref_pop1_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",\"EUR\",\".bed\"),\n",
    "    '--input BIM_Ref_pop1_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",\"EUR\",\".bim\"),\n",
    "    '--input FAM_Ref_pop1_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",\"EUR\",\".fam\"),\n",
    "    '--input BED_Ref_pop2_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",anc,\".bed\"),\n",
    "    '--input BIM_Ref_pop2_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",anc,\".bim\"),\n",
    "    '--input FAM_Ref_pop2_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",anc,\".fam\"),\n",
    "    '--input pop1_sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/EUR_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input pop2_sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/CT_SLEB_dimCT_prseb.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://CT_SLEB_dimCT_prseb.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  4.4 KiB/  4.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/4.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp CT_SLEB_dimCT_prseb.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: ct-sleb-di--williamsjacr--240722-123250-39\n",
      "  job-name: ct-sleb-dimct-prseb\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3266674414708638102\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17516157729654335075\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15003487564096970620\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10431250898719543418\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17480979131394875063\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12974867493337158376\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9396346449314849525\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7046307428555817315\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/368328631847936472\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17480892385270171289\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15563485788652614311\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14388007805605636527\n",
      "Launched job-id: ct-sleb-di--williamsjacr--240722-123250-39\n",
      "12 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-di--williamsjacr--240722-123250-39' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-di--williamsjacr--240722-123250-39' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script CT_SLEB_dimCT_prseb.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "ct-sleb-dimc...      12  Pending   07-22 12:32:51\r\n",
      "ct-sleb-dimc...      11  Pending   07-22 12:32:51\r\n",
      "ct-sleb-dimc...      10  Pending   07-22 12:32:51\r\n",
      "ct-sleb-dimc...       9  Pending   07-22 12:32:51\r\n",
      "ct-sleb-dimc...       8  Pending   07-22 12:32:51\r\n",
      "ct-sleb-dimc...       7  Pending   07-22 12:32:51\r\n",
      "ct-sleb-dimc...       6  Pending   07-22 12:32:50\r\n",
      "ct-sleb-dimc...       5  Pending   07-22 12:32:50\r\n",
      "ct-sleb-dimc...       4  Pending   07-22 12:32:50\r\n",
      "ct-sleb-dimc...       3  Pending   07-22 12:32:50\r\n",
      "ct-sleb-dimc...       2  Pending   07-22 12:32:50\r\n",
      "ct-sleb-dimc...       1  Pending   07-22 12:32:50\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-di--williamsjacr--240722-123250-39' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_LASSOSUM.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_LASSOSUM.sh\n",
    "#!/bin/bash\n",
    "\n",
    "path_plink='plink2'\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "tuning_bed_EUR=${tuning_bed_EUR%??}\n",
    "tuning_bed_AFR=${tuning_bed_AFR%??}\n",
    "tuning_bed_AMR=${tuning_bed_AMR%??}\n",
    "\n",
    "valid_bed_EUR=${valid_bed_EUR%??}\n",
    "valid_bed_AFR=${valid_bed_AFR%??}\n",
    "valid_bed_AMR=${valid_bed_AMR%??}\n",
    "\n",
    "\n",
    "Rscript ${PROSPER_Scripts}/scripts/lassosum2.R \\\n",
    "--PATH_package ${PROSPER_Scripts} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--PATH_plink ${path_plink} \\\n",
    "--FILE_sst ${EUR_Sumstats},${AFR_Sumstats},${AMR_Sumstats} \\\n",
    "--pop EUR,AFR,AMR \\\n",
    "--chrom 1-22 \\\n",
    "--bfile_tuning ${tuning_bed_EUR},${tuning_bed_AFR},${tuning_bed_AMR} \\\n",
    "--pheno_tuning ${tuning_y_EUR},${tuning_y_AFR},${tuning_y_AMR} \\\n",
    "--covar_tuning ${tuning_covar_EUR},${tuning_covar_AFR},${tuning_covar_AMR} \\\n",
    "--bfile_testing ${valid_bed_EUR},${valid_bed_AFR},${valid_bed_AMR} \\\n",
    "--pheno_testing ${valid_y_EUR},${valid_y_AFR},${valid_y_AMR} \\\n",
    "--covar_testing ${valid_covar_EUR},${valid_covar_AFR},${valid_covar_AMR} \\\n",
    "--testing TRUE \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input EUR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/EUR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AFR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AFR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AMR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AMR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input-recursive PROSPER_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER\",\n",
    "        '--input tuning_bed_EUR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_EUR.*\",\n",
    "        '--input tuning_bed_AMR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AMR.*\",\n",
    "        '--input tuning_bed_AFR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AFR.*\",\n",
    "        '--input valid_bed_EUR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_Valid_JointPRS_MUSSEL_PROSPER_EUR.*\",\n",
    "        '--input valid_bed_AMR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_Valid_JointPRS_MUSSEL_PROSPER_AMR.*\",\n",
    "        '--input valid_bed_AFR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_Valid_JointPRS_MUSSEL_PROSPER_AFR.*\",\n",
    "        '--input tuning_y_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y_EUR.txt\"),\n",
    "        '--input tuning_y_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y_AMR.txt\"),\n",
    "        '--input tuning_y_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y_AFR.txt\"),\n",
    "        '--input valid_y_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y_EUR.txt\"),\n",
    "        '--input valid_y_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y_AMR.txt\"),\n",
    "        '--input valid_y_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y_AFR.txt\"),\n",
    "        '--input tuning_covar_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar_EUR.txt\"),\n",
    "        '--input tuning_covar_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar_AMR.txt\"),\n",
    "        '--input tuning_covar_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar_AFR.txt\"),\n",
    "        '--input valid_covar_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar_EUR.txt\"),\n",
    "        '--input valid_covar_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar_AMR.txt\"),\n",
    "        '--input valid_covar_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar_AFR.txt\"),\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-la--williamsjacr--240722-180538-32\n",
      "  job-name: prosper-lassosum\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4638120276011749943\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8663400788070535812\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6956666000463151820\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4609012853128535744\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10967767430252248676\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5771711663660864596\n",
      "Launched job-id: prosper-la--williamsjacr--240722-180538-32\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-la--williamsjacr--240722-180538-32' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-la--williamsjacr--240722-180538-32' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 60 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_LASSOSUM.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "prosper-lass...       6  Pending   07-22 18:05:38\r\n",
      "prosper-lass...       5  Pending   07-22 18:05:38\r\n",
      "prosper-lass...       4  Pending   07-22 18:05:38\r\n",
      "prosper-lass...       3  Pending   07-22 18:05:38\r\n",
      "prosper-lass...       2  Pending   07-22 18:05:38\r\n",
      "prosper-lass...       1  Pending   07-22 18:05:38\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-la--williamsjacr--240722-180538-32' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "*************************\r\n",
      "*** LASSOSUM2 FOR EUR ***\r\n",
      "*************************\r\n",
      "\r\n",
      "** Step 1. Preprocessing data **\r\n",
      "\r\n",
      "** Step 2. Fitting models by chromosome **\r\n",
      "\r\n",
      "** Step 3. Tuning **\r\n",
      "Loaded 13 covariates\r\n",
      "Calculating lassosum2 scores for tuning samples\r\n",
      "lassosum2 optimal tuning parameters are saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/EUR/optimal_param.txt \r\n",
      "lassosum2 model is saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/EUR/lassosum2_prs_file.txt \r\n",
      "\r\n",
      "** Step 4. Testing **\r\n",
      "Loaded 13 covariates\r\n",
      "0.4559 variance in phenotype explained by covariates in testing samples \r\n",
      "** !COMPLETED! R2 is saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/EUR/R2.txt \r\n",
      "\r\n",
      "*************************\r\n",
      "*** LASSOSUM2 FOR AFR ***\r\n",
      "*************************\r\n",
      "\r\n",
      "** Step 1. Preprocessing data **\r\n",
      "\r\n",
      "** Step 2. Fitting models by chromosome **\r\n",
      "\r\n",
      "** Step 3. Tuning **\r\n",
      "Loaded 13 covariates\r\n",
      "Calculating lassosum2 scores for tuning samples\r\n",
      "lassosum2 optimal tuning parameters are saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/AFR/optimal_param.txt \r\n",
      "lassosum2 model is saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/AFR/lassosum2_prs_file.txt \r\n",
      "\r\n",
      "** Step 4. Testing **\r\n",
      "Loaded 13 covariates\r\n",
      "0.4235 variance in phenotype explained by covariates in testing samples \r\n",
      "** !COMPLETED! R2 is saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/AFR/R2.txt \r\n",
      "\r\n",
      "*************************\r\n",
      "*** LASSOSUM2 FOR AMR ***\r\n",
      "*************************\r\n",
      "\r\n",
      "** Step 1. Preprocessing data **\r\n",
      "\r\n",
      "** Step 2. Fitting models by chromosome **\r\n",
      "\r\n",
      "** Step 3. Tuning **\r\n",
      "Loaded 13 covariates\r\n",
      "Calculating lassosum2 scores for tuning samples\r\n",
      "lassosum2 optimal tuning parameters are saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/AMR/optimal_param.txt \r\n",
      "lassosum2 model is saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/AMR/lassosum2_prs_file.txt \r\n",
      "\r\n",
      "** Step 4. Testing **\r\n",
      "Loaded 13 covariates\r\n",
      "0.5006 variance in phenotype explained by covariates in testing samples \r\n",
      "** !COMPLETED! R2 is saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/AMR/R2.txt \r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-la--williamsjacr--240722-180538-32.6-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting MUSSEL_LDPred2.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile MUSSEL_LDPred2.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "tuning_bed_EUR=${tuning_bed_EUR%??}\n",
    "tuning_bed_AFR=${tuning_bed_AFR%??}\n",
    "tuning_bed_AMR=${tuning_bed_AMR%??}\n",
    "\n",
    "Rscript ${MUSSEL_Scripts}/R/LDpred2.R \\\n",
    "--PATH_package ${MUSSEL_Scripts} \\\n",
    "--PATH_ref ${MUSSEL_LD} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--FILE_sst ${EUR_Sumstats},${AFR_Sumstats},${AMR_Sumstats} \\\n",
    "--pop EUR,AFR,AMR \\\n",
    "--bfile_tuning ${tuning_bed_EUR},${tuning_bed_AFR},${tuning_bed_AMR} \\\n",
    "--verbose 2 \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input EUR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/EUR_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),\n",
    "        '--input AFR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/AFR_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),\n",
    "        '--input AMR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/AMR_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),\n",
    "        '--input-recursive MUSSEL_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/MUSSEL\",\n",
    "        '--input-recursive MUSSEL_LD'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/MUSSEL_LD\",\n",
    "        '--input tuning_bed_EUR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_EUR.*\",\n",
    "        '--input tuning_bed_AMR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AMR.*\",\n",
    "        '--input tuning_bed_AFR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AFR.*\",\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: mussel-ldp--williamsjacr--240722-180603-76\n",
      "  job-name: mussel-ldpred2\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1876996133311691452\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2094648451624967140\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13690489926848506048\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5508878527126587761\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16970031399968614553\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5105032732639826203\n",
      "Launched job-id: mussel-ldp--williamsjacr--240722-180603-76\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-ldp--williamsjacr--240722-180603-76' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-ldp--williamsjacr--240722-180603-76' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script MUSSEL_LDPred2.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Delete running jobs:\n",
      "  user:\n",
      "    {'williamsjacr'}\n",
      "\n",
      "  job-id:\n",
      "    ['mussel-ldp--williamsjacr--240722-180603-76']\n",
      "\n",
      "Found 6 tasks to delete.\n",
      "1 job deleted (6 tasks)\n"
     ]
    }
   ],
   "source": [
    "!ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-ldp--williamsjacr--240722-180603-76' --users 'williamsjacr'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "*************************\r\n",
      "**** LDpred2 on EUR ****\r\n",
      "*************************\r\n",
      "\r\n",
      "** Step 1: data preparation **\r\n",
      "[1] \"Complete pre-processing GWAS summary data.\"\r\n",
      "1.. [1] \"Complete summarizing LD matrix info for CHR 1\"\r\n",
      "2.. [1] \"Complete summarizing LD matrix info for CHR 2\"\r\n",
      "3.. [1] \"Complete summarizing LD matrix info for CHR 3\"\r\n",
      "4.. [1] \"Complete summarizing LD matrix info for CHR 4\"\r\n",
      "5.. [1] \"Complete summarizing LD matrix info for CHR 5\"\r\n",
      "6.. [1] \"Complete summarizing LD matrix info for CHR 6\"\r\n",
      "7.. [1] \"Complete summarizing LD matrix info for CHR 7\"\r\n",
      "8.. [1] \"Complete summarizing LD matrix info for CHR 8\"\r\n",
      "9.. [1] \"Complete summarizing LD matrix info for CHR 9\"\r\n",
      "10.. [1] \"Complete summarizing LD matrix info for CHR 10\"\r\n",
      "11.. [1] \"Complete summarizing LD matrix info for CHR 11\"\r\n",
      "12.. [1] \"Complete summarizing LD matrix info for CHR 12\"\r\n",
      "13.. [1] \"Complete summarizing LD matrix info for CHR 13\"\r\n",
      "14.. [1] \"Complete summarizing LD matrix info for CHR 14\"\r\n",
      "15.. [1] \"Complete summarizing LD matrix info for CHR 15\"\r\n",
      "16.. [1] \"Complete summarizing LD matrix info for CHR 16\"\r\n",
      "17.. [1] \"Complete summarizing LD matrix info for CHR 17\"\r\n",
      "18.. [1] \"Complete summarizing LD matrix info for CHR 18\"\r\n",
      "19.. [1] \"Complete summarizing LD matrix info for CHR 19\"\r\n",
      "20.. [1] \"Complete summarizing LD matrix info for CHR 20\"\r\n",
      "21.. [1] \"Complete summarizing LD matrix info for CHR 21\"\r\n",
      "22.. [1] \"Complete summarizing LD matrix info for CHR 22\"\r\n",
      "\r\n",
      "** Step 2: run LDpred2 under all tuning parameter settings **\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/mussel-ldp--williamsjacr--240722-180603-76.1-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning message:\r\n",
      "NAs introduced by coercion \r\n",
      "Warning message:\r\n",
      "NAs introduced by coercion \r\n",
      "1,545,398 variants to be matched.\r\n",
      "63,059 ambiguous SNPs have been removed.\r\n",
      "1,482,339 variants have been matched; 0 were flipped and 1,080,289 were reversed.\r\n",
      "Loading required package: rngtools\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/mussel-ldp--williamsjacr--240722-180603-76.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Prosper.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Prosper.sh\n",
    "\n",
    "path_plink='plink2'\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${PROSPER_Scripts}/scripts/PROSPER.R \\\n",
    "--PATH_package ${PROSPER_Scripts} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--FILE_sst ${EUR_Sumstats},${AFR_Sumstats},${AMR_Sumstats} \\\n",
    "--pop EUR,AFR,AMR \\\n",
    "--lassosum_param ${EUR_Optim_Param},${AFR_Optim_Param},${AMR_Optim_Param} \\\n",
    "--chrom 1-22 \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input EUR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/EUR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AFR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AFR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AMR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AMR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input EUR_Optim_Param'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/EUR/optimal_param.txt\"),\n",
    "        '--input AFR_Optim_Param'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/AFR/optimal_param.txt\"),\n",
    "        '--input AMR_Optim_Param'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/AMR/optimal_param.txt\"),\n",
    "        '--input-recursive PROSPER_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER\",\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-pr--williamsjacr--240723-155606-10\n",
      "  job-name: prosper-prosper\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2200656647477152566\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10608213276173193313\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11366293823247759560\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3874828222003813483\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18258188515994055652\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5478996094867208468\n",
      "Launched job-id: prosper-pr--williamsjacr--240723-155606-10\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-pr--williamsjacr--240723-155606-10' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-pr--williamsjacr--240723-155606-10' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 60 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Prosper.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                          Last Update\r\n",
      "---------------  ------  ------------------------------  --------------\r\n",
      "prosper-prosper       6  Started running \"user-command\"  07-23 16:07:08\r\n",
      "prosper-prosper       5  Started running \"user-command\"  07-23 16:07:16\r\n",
      "prosper-prosper       4  Started running \"user-command\"  07-23 16:07:28\r\n",
      "prosper-prosper       3  Started running \"user-command\"  07-23 16:08:49\r\n",
      "prosper-prosper       2  Started running \"user-command\"  07-23 16:07:10\r\n",
      "prosper-prosper       1  Started running \"user-command\"  07-23 16:07:17\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-pr--williamsjacr--240723-155606-10' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "** Step 1. Preprocessing data **\r\n",
      "\r\n",
      "** Step 2. Fitting models by chromosome **\r\n",
      "PRSs in all tuning parameter settings are saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/BMI/before_ensemble/score_file.txt \r\n",
      "Their corresponding tuning parameter settings are saved in /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/BMI/before_ensemble/score_param.txt \r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-pr--williamsjacr--240723-155606-10.1-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_SL.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_SL.sh\n",
    "\n",
    "path_plink='plink2'\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "tuning_bed=${tuning_bed%??}\n",
    "\n",
    "valid_bed=${valid_bed%??}\n",
    "\n",
    "Rscript ${PROSPER_Scripts}/scripts/tuning_testing.R \\\n",
    "--PATH_plink ${path_plink} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--PATH_input ${INPUT_PATH} \\\n",
    "--prefix 'full_results' \\\n",
    "--SL_library 'SL.glmnet,SL.ridge,SL.lm' \\\n",
    "--testing TRUE \\\n",
    "--linear_score TRUE \\\n",
    "--bfile_tuning ${tuning_bed} \\\n",
    "--pheno_tuning ${tuning_y} \\\n",
    "--covar_tuning ${tuning_covar} \\\n",
    "--bfile_testing ${valid_bed} \\\n",
    "--pheno_testing ${valid_y} \\\n",
    "--covar_testing ${valid_covar} \\\n",
    "--cleanup F \\\n",
    "--verbose 2 \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input-recursive PROSPER_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER\",\n",
    "        '--input-recursive INPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "        '--input tuning_bed'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.*\",\n",
    "        '--input valid_bed'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_valid_JointPRS_MUSSEL_PROSPER.*\",\n",
    "        '--input tuning_y'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y.txt\"),\n",
    "        '--input valid_y'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y.txt\"),\n",
    "        '--input tuning_covar'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar.txt\"),\n",
    "        '--input valid_covar'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar.txt\"),\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-sl--williamsjacr--240725-203220-36\n",
      "  job-name: prosper-sl\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3198059155794176789\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9150731821324572012\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18392933430793886820\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17610843321921514412\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3790050121890991472\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10535344993978084634\n",
      "Launched job-id: prosper-sl--williamsjacr--240725-203220-36\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sl--williamsjacr--240725-203220-36' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sl--williamsjacr--240725-203220-36' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 60 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_SL.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name      Task  Status    Last Update\r\n",
      "----------  ------  --------  --------------\r\n",
      "prosper-sl       6  Pending   07-25 20:32:20\r\n",
      "prosper-sl       5  Pending   07-25 20:32:20\r\n",
      "prosper-sl       4  Pending   07-25 20:32:20\r\n",
      "prosper-sl       3  Pending   07-25 20:32:20\r\n",
      "prosper-sl       2  Pending   07-25 20:32:20\r\n",
      "prosper-sl       1  Pending   07-25 20:32:20\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sl--williamsjacr--240725-203220-36' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 1031 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n",
      "Warning: 1031 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sl--williamsjacr--240724-145841-65.6-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting JointPRS_Score.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_Score.R\n",
    "rm(list = ls())\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[1]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[2]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    print(trait)\n",
    "    a <- read.csv(paste0(INPUT_PATH,\"/JointPRS_META_Score_\",trait,\".csv\"))\n",
    "    colnames(a) <- c(\"CHR\", \"rsID\", \"BP\", \"A1\", \"A2\", \"Post_EffectSize\")\n",
    "    a <- data.frame(SNP = a$rsID,A1 = a$A1,BETA = a$Post_EffectSize)\n",
    "    write.table(a,file = \"Final_PRS_Coefficients\",col.names = T,row.names = F,quote=F)\n",
    "    \n",
    "    system(paste0(\"plink2 --threads 2 --score Final_PRS_Coefficients cols=+scoresums,-scoreavgs header no-mean-imputation  --bfile \",BED_Full_File,\" --out \",OUTPUT_PATH,\"/PRS_META_\",trait))\n",
    "} \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting JointPRS_Score.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_Score.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${INPUT_PATH} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/Scores\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/JointPRS_Score.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://JointPRS_Score.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  812.0 B/  812.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/812.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp JointPRS_Score.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: jointprs-s--williamsjacr--240731-121008-31\n",
      "  job-name: jointprs-score\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12804961307672585039\n",
      "Launched job-id: jointprs-s--williamsjacr--240731-121008-31\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-s--williamsjacr--240731-121008-31' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-s--williamsjacr--240731-121008-31' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script JointPRS_Score.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name          Task  Status    Last Update\r\n",
      "--------------  ------  --------  --------------\r\n",
      "jointprs-score       1  Pending   07-31 12:10:08\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-s--williamsjacr--240731-121008-31' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: object 'anc' not found\r\n",
      "Execution halted\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/jointprs-s--williamsjacr--240725-211311-32.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_Score.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score.R\n",
    "rm(list = ls())\n",
    "\n",
    "Score_BMI <- commandArgs(TRUE)[1]\n",
    "print(Score_BMI)\n",
    "\n",
    "Score_LDL <- commandArgs(TRUE)[2]\n",
    "print(Score_LDL)\n",
    "\n",
    "Score_HDL <- commandArgs(TRUE)[3]\n",
    "print(Score_HDL)\n",
    "\n",
    "Score_logTG <- commandArgs(TRUE)[4]\n",
    "print(Score_logTG)\n",
    "\n",
    "Score_TC <- commandArgs(TRUE)[5]\n",
    "print(Score_TC)\n",
    "\n",
    "Score_Height <- commandArgs(TRUE)[6]\n",
    "print(Score_Height)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[7]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[8]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    \n",
    "    if(trait == \"BMI\"){\n",
    "        Score_File <- Score_BMI\n",
    "    }else if(trait == \"LDL\"){\n",
    "        Score_File <- Score_LDL\n",
    "    }else if(trait == \"HDL\"){\n",
    "        Score_File <- Score_HDL\n",
    "    }else if(trait == \"logTG\"){\n",
    "        Score_File <- Score_logTG\n",
    "    }else if(trait == \"TC\"){\n",
    "        Score_File <- Score_TC\n",
    "    }else {\n",
    "        Score_File <- Score_Height\n",
    "    }\n",
    "    \n",
    "    b <- read.delim(Score_File)\n",
    "    b <- data.frame(SNP = b$rsid,A1 = b$a1,BETA = b$weight)\n",
    "    write.table(b,file = \"Final_PRS_Coefficients\",col.names = T,row.names = F,quote=F)\n",
    "    \n",
    "    system(paste0(\"plink2 --threads 2 --score Final_PRS_Coefficients cols=+scoresums,-scoreavgs header no-mean-imputation  --bfile \",BED_Full_File,\" --out \",OUTPUT_PATH,\"/PRS_\",trait))\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_Score.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Score_BMI} ${Score_LDL} ${Score_HDL} ${Score_logTG} ${Score_TC} ${Score_Height} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Score_BMI'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/BMI/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_LDL'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/LDL/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_HDL'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/HDL/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_logTG'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/logTG/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_TC'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/TC/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_Height'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Score.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_Score.R [Content-Type=application/octet-stream]...\n",
      "\\ [1/1 files][  1.3 KiB/  1.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_Score.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-sc--williamsjacr--240725-225501-88\n",
      "  job-name: prosper-score\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7038750270459992951\n",
      "Launched job-id: prosper-sc--williamsjacr--240725-225501-88\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--240725-225501-88' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--240725-225501-88' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Score.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name         Task  Status    Last Update\r\n",
      "-------------  ------  --------  --------------\r\n",
      "prosper-score       1  Pending   07-25 22:55:02\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--240725-225501-88' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 1029 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n",
      "Warning: 1038 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n",
      "Warning: 1040 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n",
      "Warning: 1038 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n",
      "Warning: 1038 --score file entries were skipped due to mismatching allele\r\n",
      "codes.\r\n",
      "(Add the 'list-variants' modifier to see which variants were actually used for\r\n",
      "scoring.)\r\n",
      "Error in data.frame(SNP = b$rsid, A1 = b$a1, BETA = b$weight) : \r\n",
      "  arguments imply differing number of rows: 1456203, 0\r\n",
      "Execution halted\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240725-121247-15.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_Score_All.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score_All.R\n",
    "rm(list = ls())\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Score_Trait <- commandArgs(TRUE)[2]\n",
    "print(Score_Trait)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[3]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Score_File <- Score_Trait\n",
    "    \n",
    "b <- read.delim(Score_File)\n",
    "\n",
    "system(paste0(\"plink2 --threads 2 --score \",Score_Trait,\" cols=+scoresums,-scoreavgs header no-mean-imputation --score-col-nums 4-\",ncol(b),\" --bfile \",BED_Full_File,\" --out \",OUTPUT_PATH,\"/PRS_All_\",trait))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting PROSPER_Score_All.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score_All.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Score_Trait} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "    '--env trait'=trait,\n",
    "    '--input Score_Trait'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/before_ensemble/score_file.txt\"),\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Score_All.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   \n",
    "}\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_Score_All.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  548.0 B/  548.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/548.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_Score_All.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-sc--williamsjacr--240726-131839-85\n",
      "  job-name: prosper-score-all\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16919446399436233190\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16216689967082005827\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/519434646358723671\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14020884406289596863\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/855727823133521907\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16841389287832901535\n",
      "Launched job-id: prosper-sc--williamsjacr--240726-131839-85\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--240726-131839-85' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--240726-131839-85' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Score_All.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "prosper-scor...       6  Pending   07-26 13:18:40\r\n",
      "prosper-scor...       5  Pending   07-26 13:18:40\r\n",
      "prosper-scor...       4  Pending   07-26 13:18:40\r\n",
      "prosper-scor...       3  Pending   07-26 13:18:40\r\n",
      "prosper-scor...       2  Pending   07-26 13:18:40\r\n",
      "prosper-scor...       1  Pending   07-26 13:18:40\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--240726-131839-85' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-07-26 12:27:37 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\r\n",
      "2024-07-26 12:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\r\n",
      "2024-07-26 12:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\r\n",
      "2024-07-26 12:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/before_ensemble\r\n",
      "2024-07-26 12:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts\r\n",
      "2024-07-26 12:27:37 INFO: mkdir -m 777 -p /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/\r\n",
      "2024-07-26 12:27:38 INFO: Localizing BIM_Full_File\r\n",
      "2024-07-26 12:27:38 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\r\n",
      "2024-07-26 12:27:40 INFO: Localizing Score_Trait\r\n",
      "2024-07-26 12:27:40 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/before_ensemble/score_file.txt /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/before_ensemble/score_file.txt\r\n",
      "2024-07-26 12:28:04 INFO: Localizing BED_Full_File\r\n",
      "2024-07-26 12:28:04 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\r\n",
      "2024-07-26 12:28:39 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:29:41 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:30:44 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:31:46 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:32:49 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:33:52 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:34:54 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:35:57 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:36:59 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:38:02 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:39:04 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:40:06 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:41:09 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:42:11 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:43:13 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:44:16 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:45:18 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:46:20 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:47:22 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:48:25 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:49:27 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:50:29 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:51:32 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:52:34 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:53:36 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:54:38 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:55:40 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:56:43 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:57:45 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:58:47 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 12:59:49 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:00:52 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:01:54 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:02:56 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:03:59 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:05:01 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:06:03 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:07:05 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:08:08 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:08:49 INFO: Localizing R_Script\r\n",
      "2024-07-26 13:08:49 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Score_All.R /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Score_All.R\r\n",
      "2024-07-26 13:08:51 INFO: Localizing FAM_Full_File\r\n",
      "2024-07-26 13:08:51 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\r\n",
      "2024-07-26 13:09:10 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/stdout gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6-stdout.log\r\n",
      "2024-07-26 13:09:10 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log\r\n",
      "2024-07-26 13:09:10 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/stderr gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6-stderr.log\r\n",
      "/mnt/data/script/PROSPER_Score_All.sh: line 6: trait: unbound variable\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/prosper-sc--williamsjacr--240726-122424-49.6.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting CT_SLEB_SL.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_SL.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "\n",
    "AFR_PRS <- commandArgs(TRUE)[1]\n",
    "print(AFR_PRS)\n",
    "\n",
    "AMR_PRS <- commandArgs(TRUE)[2]\n",
    "print(AMR_PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[3]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[4]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[5]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[6]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[7]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[8]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "\n",
    "load(AFR_PRS)\n",
    "colnames(prs_mat_eb)[3:ncol(prs_mat_eb)] <- paste0(\"AFR_\",colnames(prs_mat_eb)[3:ncol(prs_mat_eb)])\n",
    "\n",
    "PRS_Mat_AFR <- prs_mat_eb\n",
    "rm(prs_mat_eb)\n",
    "\n",
    "load(AMR_PRS)\n",
    "colnames(prs_mat_eb)[3:ncol(prs_mat_eb)] <- paste0(\"AMR_\",colnames(prs_mat_eb)[3:ncol(prs_mat_eb)])\n",
    "\n",
    "PRS_Mat_AMR <- prs_mat_eb\n",
    "rm(prs_mat_eb)\n",
    "\n",
    "PRS_Mat_All <- inner_join(PRS_Mat_AFR,PRS_Mat_AMR)\n",
    "\n",
    "rm(PRS_Mat_AFR);rm(PRS_Mat_AMR)\n",
    "\n",
    "\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "prs_train_all <- pheno_training[!is.na(pheno_training[,trait]),-c(1:22)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "prs_tune_all <- pheno_tuning[!is.na(pheno_tuning[,trait]),-c(1:22)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "prs_validation_all <- pheno_valid[,-c(1:22)]\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "\n",
    "drop <- caret::findLinearCombos(prs_tune_all)$remove\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))\n",
    "    \n",
    "\n",
    "mtx <- cor(prs_tune_all)\n",
    "drop <- findCorrelation(mtx,cutoff=0.98)\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))    \n",
    "\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "\n",
    "\n",
    "SL.library <- c(\n",
    "  \"SL.glmnet\",\n",
    "  \"SL.ridge\"\n",
    ")\n",
    "sl <- SuperLearner(Y = y_tune, X = prs_tune_all, family = gaussian(),\n",
    "                   # For a real analysis we would use V = 10.\n",
    "                   # V = 3,\n",
    "                   SL.library = SL.library,cvControl = list(V = 20))\n",
    "cvsl <- CV.SuperLearner(Y = y_tune, X = prs_tune_all, family = gaussian(),\n",
    "                        # For a real analysis we would use V = 10.\n",
    "                        # V = 3,\n",
    "                        SL.library = SL.library,cvControl = list(V = 20))\n",
    "\n",
    "best_algorithm <- summary(cvsl)$Table$Algorithm[which.min(summary(cvsl)$Table$Ave)]\n",
    "\n",
    "a <- predict(sl, prs_validation_all, onlySL = FALSE)\n",
    "\n",
    "prs_best_validation_sl <- a$pred\n",
    "\n",
    "prs_best_validation <- data.frame(IID = pheno_valid$IID,prs = prs_best_validation_sl)\n",
    "\n",
    "\n",
    "a <- predict(sl, prs_train_all, onlySL = FALSE)\n",
    "\n",
    "prs_best_train_sl <- a$pred\n",
    "\n",
    "prs_best_train <- data.frame(IID = pheno_training$IID[!is.na(pheno_training[,trait])],prs = prs_best_train_sl)\n",
    "\n",
    "\n",
    "a <- predict(sl, prs_tune_all, onlySL = FALSE)\n",
    "\n",
    "prs_best_tune_sl <- a$pred\n",
    "\n",
    "prs_best_tune <- data.frame(IID = pheno_tuning$IID[!is.na(pheno_tuning[,trait])],prs = prs_best_tune_sl)\n",
    "\n",
    "pheno_training <- left_join(pheno_training,prs_best_train)\n",
    "pheno_valid <- left_join(pheno_valid,prs_best_validation)\n",
    "pheno_tuning <- left_join(pheno_tuning,prs_best_tune)\n",
    "\n",
    "\n",
    "beta_lasso <- data.frame(Coef = row.names(coef(sl$fitLibrary$SL.glmnet_All$object)),Beta = as.numeric(coef(sl$fitLibrary$SL.glmnet_All$object)))\n",
    "beta_ridge <- data.frame(Coef = row.names(sl$fitLibrary$SL.ridge_All$bestCoef),Beta = as.numeric(sl$fitLibrary$SL.ridge_All$bestCoef[,1]))\n",
    "    \n",
    "beta_sl <- data.frame(Coef = row.names(coef(sl$fitLibrary$SL.glmnet_All$object)), Beta = 0)\n",
    "beta_sl$Beta <- as.numeric(sl$coef[names(sl$coef) == \"SL.glmnet_All\"]) * beta_lasso$Beta + \n",
    "      as.numeric(sl$coef[names(sl$coef) == \"SL.ridge_All\"]) * beta_ridge$Beta \n",
    "  \n",
    "write.csv(beta_sl,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Final_Coefficients.csv\"),row.names = FALSE)\n",
    "\n",
    "\n",
    "prs_best_train <- data.frame(IID = pheno_training$IID,prs = pheno_training[,\"prs\"])\n",
    "\n",
    "prs_best_tune <- data.frame(IID = pheno_tuning$IID,prs = pheno_tuning[,\"prs\"])\n",
    "\n",
    "prs_best_validation <- data.frame(IID = pheno_valid$IID,prs = pheno_valid[,\"prs\"])\n",
    "\n",
    "write.table(prs_best_train,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Train_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(prs_best_tune,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Tune_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(prs_best_validation,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Validation_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- pheno_valid\n",
    "pheno_validation_adjusted <- pheno_valid\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"prs\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"prs\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"prs\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR[,\"prs\"] <- scale(pheno_validation_raw_EUR[,\"prs\"])\n",
    "pheno_validation_raw_SAS[,\"prs\"] <- scale(pheno_validation_raw_SAS[,\"prs\"])\n",
    "pheno_validation_raw_AMR[,\"prs\"] <- scale(pheno_validation_raw_AMR[,\"prs\"])\n",
    "pheno_validation_raw_AFR[,\"prs\"] <- scale(pheno_validation_raw_AFR[,\"prs\"])\n",
    "pheno_validation_raw_EAS[,\"prs\"] <- scale(pheno_validation_raw_EAS[,\"prs\"])\n",
    "pheno_validation_raw_MID[,\"prs\"] <- scale(pheno_validation_raw_MID[,\"prs\"])\n",
    "\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))[2]\n",
    "se_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))$coefficients[2,2]\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))[2]\n",
    "se_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))$coefficients[2,2]\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))[2]\n",
    "se_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))$coefficients[2,2]\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))[2]\n",
    "se_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))$coefficients[2,2]\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))[2]\n",
    "se_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))$coefficients[2,2]\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))[2]\n",
    "se_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))$coefficients[2,2]\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "se_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))$coefficients[2,2]\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "se_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "se_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))$coefficients[2,2]\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "se_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))$coefficients[2,2]\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "se_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))[2]\n",
    "se_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))$coefficients[2,2]\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         se_raw = c(se_validation_raw_EUR,se_validation_raw_SAS,se_validation_raw_AMR,se_validation_raw_AFR,se_validation_raw_EAS,se_validation_raw_MID), \n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         se_adjusted = c(se_validation_adjusted_EUR,se_validation_adjusted_SAS,se_validation_adjusted_AMR,se_validation_adjusted_AFR,se_validation_adjusted_EAS,se_validation_adjusted_MID))\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting CT_SLEB_SL.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_SL.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${AFR_PRS} ${AMR_PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input AFR_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AFR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input AMR_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AMR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/CT_SLEB_SL.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/Results/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://CT_SLEB_SL.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 12.9 KiB/ 12.9 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/12.9 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp CT_SLEB_SL.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: ct-sleb-sl--williamsjacr--240726-135410-37\n",
      "  job-name: ct-sleb-sl\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12577038072026570578\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10747239263645617330\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8395246878476471389\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14450160083624827693\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14140011373402086818\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7320237539570158463\n",
      "Launched job-id: ct-sleb-sl--williamsjacr--240726-135410-37\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-sl--williamsjacr--240726-135410-37' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-sl--williamsjacr--240726-135410-37' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script CT_SLEB_SL.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name      Task  Status                          Last Update\r\n",
      "----------  ------  ------------------------------  --------------\r\n",
      "ct-sleb-sl       6  Started running \"user-command\"  07-26 13:57:47\r\n",
      "ct-sleb-sl       5  Started running \"user-command\"  07-26 13:57:57\r\n",
      "ct-sleb-sl       4  Started running \"user-command\"  07-26 13:57:32\r\n",
      "ct-sleb-sl       3  Started running \"user-command\"  07-26 13:57:36\r\n",
      "ct-sleb-sl       2  Started running \"user-command\"  07-26 13:58:03\r\n",
      "ct-sleb-sl       1  Started running \"user-command\"  07-26 13:57:37\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-sl--williamsjacr--240726-135410-37' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: ggplot2\r\n",
      "Loading required package: lattice\r\n",
      "Loading required package: nnls\r\n",
      "Loading required package: gam\r\n",
      "Loading required package: splines\r\n",
      "Loading required package: foreach\r\n",
      "Loaded gam 1.22-3\r\n",
      "\r\n",
      "Super Learner\r\n",
      "Version: 2.0-29\r\n",
      "Package created on 2024-02-06\r\n",
      "\r\n",
      "\r\n",
      "Attaching package: ‘dplyr’\r\n",
      "\r\n",
      "The following objects are masked from ‘package:stats’:\r\n",
      "\r\n",
      "    filter, lag\r\n",
      "\r\n",
      "The following objects are masked from ‘package:base’:\r\n",
      "\r\n",
      "    intersect, setdiff, setequal, union\r\n",
      "\r\n",
      "\r\n",
      "Attaching package: ‘boot’\r\n",
      "\r\n",
      "The following object is masked from ‘package:lattice’:\r\n",
      "\r\n",
      "    melanoma\r\n",
      "\r\n",
      "Joining with `by = join_by(`#FID`, IID)`\r\n",
      "Loading required namespace: glmnet\r\n",
      "/mnt/data/script/CT_SLEB_SL.sh: line 6:     7 Killed                  Rscript ${R_Script} ${AFR_PRS} ${AMR_PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/ct-sleb-sl--williamsjacr--240726-123906-81.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Joint_PRS_Results.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Joint_PRS_Results.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "\n",
    "PRS <- commandArgs(TRUE)[1]\n",
    "print(PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[3]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[4]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[5]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[6]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "PRS_Mat_All <- read.delim(PRS)\n",
    "colnames(PRS_Mat_All) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"prs\")\n",
    "PRS_Mat_All <- PRS_Mat_All[,c(\"IID\",\"prs\")]\n",
    "\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- pheno_valid\n",
    "pheno_validation_adjusted <- pheno_valid\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"prs\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"prs\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"prs\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR[,\"prs\"] <- scale(pheno_validation_raw_EUR[,\"prs\"])\n",
    "pheno_validation_raw_SAS[,\"prs\"] <- scale(pheno_validation_raw_SAS[,\"prs\"])\n",
    "pheno_validation_raw_AMR[,\"prs\"] <- scale(pheno_validation_raw_AMR[,\"prs\"])\n",
    "pheno_validation_raw_AFR[,\"prs\"] <- scale(pheno_validation_raw_AFR[,\"prs\"])\n",
    "pheno_validation_raw_EAS[,\"prs\"] <- scale(pheno_validation_raw_EAS[,\"prs\"])\n",
    "pheno_validation_raw_MID[,\"prs\"] <- scale(pheno_validation_raw_MID[,\"prs\"])\n",
    "\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))[2]\n",
    "se_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))$coefficients[2,2]\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))[2]\n",
    "se_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))$coefficients[2,2]\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))[2]\n",
    "se_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))$coefficients[2,2]\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))[2]\n",
    "se_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))$coefficients[2,2]\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))[2]\n",
    "se_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))$coefficients[2,2]\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))[2]\n",
    "se_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))$coefficients[2,2]\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "se_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))$coefficients[2,2]\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "se_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "se_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))$coefficients[2,2]\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "se_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))$coefficients[2,2]\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "se_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))[2]\n",
    "se_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))$coefficients[2,2]\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         se_raw = c(se_validation_raw_EUR,se_validation_raw_SAS,se_validation_raw_AMR,se_validation_raw_AFR,se_validation_raw_EAS,se_validation_raw_MID), \n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         se_adjusted = c(se_validation_adjusted_EUR,se_validation_adjusted_SAS,se_validation_adjusted_AMR,se_validation_adjusted_AFR,se_validation_adjusted_EAS,se_validation_adjusted_MID))\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Joint_PRS_Results.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Joint_PRS_Results.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs/PRS_META_\",trait,\".sscore\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Joint_PRS_Results.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/Results/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Joint_PRS_Results.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  9.4 KiB/  9.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/9.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Joint_PRS_Results.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: joint-prs---williamsjacr--240731-191801-94\n",
      "  job-name: joint-prs-results\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3712880415493156591\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13846036390793409358\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7413637980004994758\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2494075548292990445\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4063203274783889299\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17990280119822650816\n",
      "Launched job-id: joint-prs---williamsjacr--240731-191801-94\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'joint-prs---williamsjacr--240731-191801-94' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'joint-prs---williamsjacr--240731-191801-94' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Joint_PRS_Results.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "joint-prs-re...       6  Pending   07-31 19:18:02\r\n",
      "joint-prs-re...       5  Pending   07-31 19:18:02\r\n",
      "joint-prs-re...       4  Pending   07-31 19:18:02\r\n",
      "joint-prs-re...       3  Pending   07-31 19:18:02\r\n",
      "joint-prs-re...       2  Pending   07-31 19:18:02\r\n",
      "joint-prs-re...       1  Pending   07-31 19:18:02\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'joint-prs---williamsjacr--240731-191801-94' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: ggplot2\r\n",
      "Loading required package: lattice\r\n",
      "Loading required package: nnls\r\n",
      "Loading required package: gam\r\n",
      "Loading required package: splines\r\n",
      "Loading required package: foreach\r\n",
      "Loaded gam 1.22-3\r\n",
      "\r\n",
      "Super Learner\r\n",
      "Version: 2.0-29\r\n",
      "Package created on 2024-02-06\r\n",
      "\r\n",
      "\r\n",
      "Attaching package: ‘dplyr’\r\n",
      "\r\n",
      "The following objects are masked from ‘package:stats’:\r\n",
      "\r\n",
      "    filter, lag\r\n",
      "\r\n",
      "The following objects are masked from ‘package:base’:\r\n",
      "\r\n",
      "    intersect, setdiff, setequal, union\r\n",
      "\r\n",
      "\r\n",
      "Attaching package: ‘boot’\r\n",
      "\r\n",
      "The following object is masked from ‘package:lattice’:\r\n",
      "\r\n",
      "    melanoma\r\n",
      "\r\n",
      "Error in array(x, c(length(x), 1L), if (!is.null(names(x))) list(names(x),  : \r\n",
      "  'data' must be of a vector type, was 'NULL'\r\n",
      "Calls: scale ... scale.default -> as.matrix -> as.matrix.default -> array\r\n",
      "Execution halted\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/joint-prs---williamsjacr--240726-155335-82.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting OneCommonPRS.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Results.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "\n",
    "PRS <- commandArgs(TRUE)[1]\n",
    "print(PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[3]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[4]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[5]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[6]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "PRS_Mat_All <- read.delim(PRS)\n",
    "colnames(PRS_Mat_All) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"prs\")\n",
    "PRS_Mat_All <- PRS_Mat_All[,c(\"IID\",\"prs\")]\n",
    "\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- pheno_valid\n",
    "pheno_validation_adjusted <- pheno_valid\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"prs\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"prs\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"prs\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR[,\"prs\"] <- scale(pheno_validation_raw_EUR[,\"prs\"])\n",
    "pheno_validation_raw_SAS[,\"prs\"] <- scale(pheno_validation_raw_SAS[,\"prs\"])\n",
    "pheno_validation_raw_AMR[,\"prs\"] <- scale(pheno_validation_raw_AMR[,\"prs\"])\n",
    "pheno_validation_raw_AFR[,\"prs\"] <- scale(pheno_validation_raw_AFR[,\"prs\"])\n",
    "pheno_validation_raw_EAS[,\"prs\"] <- scale(pheno_validation_raw_EAS[,\"prs\"])\n",
    "pheno_validation_raw_MID[,\"prs\"] <- scale(pheno_validation_raw_MID[,\"prs\"])\n",
    "\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))[2]\n",
    "se_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))$coefficients[2,2]\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))[2]\n",
    "se_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))$coefficients[2,2]\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))[2]\n",
    "se_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))$coefficients[2,2]\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))[2]\n",
    "se_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))$coefficients[2,2]\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))[2]\n",
    "se_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))$coefficients[2,2]\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))[2]\n",
    "se_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))$coefficients[2,2]\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "se_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))$coefficients[2,2]\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "se_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "se_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))$coefficients[2,2]\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "se_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))$coefficients[2,2]\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "se_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))[2]\n",
    "se_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))$coefficients[2,2]\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         se_raw = c(se_validation_raw_EUR,se_validation_raw_SAS,se_validation_raw_AMR,se_validation_raw_AFR,se_validation_raw_EAS,se_validation_raw_MID), \n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         se_adjusted = c(se_validation_adjusted_EUR,se_validation_adjusted_SAS,se_validation_adjusted_AMR,se_validation_adjusted_AFR,se_validation_adjusted_EAS,se_validation_adjusted_MID))\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Results.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Results.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/PRS_\",trait,\".sscore\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Results.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Results/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CommandException: No URLs matched: PROSPER_Results.R\n",
      "CommandException: 1 file/object could not be transferred.\n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_Results.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-re--williamsjacr--240726-164526-12\n",
      "  job-name: prosper-results\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14514243183779343518\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11259223612414206690\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12115918917554650631\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6019796038400644284\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10214161872501419410\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8329389636366739508\n",
      "Launched job-id: prosper-re--williamsjacr--240726-164526-12\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-re--williamsjacr--240726-164526-12' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-re--williamsjacr--240726-164526-12' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Results.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "prosper-results       6  Pending   07-26 16:45:26\r\n",
      "prosper-results       5  Pending   07-26 16:45:26\r\n",
      "prosper-results       4  Pending   07-26 16:45:26\r\n",
      "prosper-results       3  Pending   07-26 16:45:26\r\n",
      "prosper-results       2  Pending   07-26 16:45:26\r\n",
      "prosper-results       1  Pending   07-26 16:45:26\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-re--williamsjacr--240726-164526-12' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing OneCommonPRS.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile OneCommonPRS.R\n",
    "rm(list = ls())\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "\n",
    "PROSPER_PRSs <- commandArgs(TRUE)[1]\n",
    "print(PROSPER_PRSs)\n",
    "\n",
    "JointPRS_PRS <- commandArgs(TRUE)[2]\n",
    "print(JointPRS_PRS)\n",
    "\n",
    "CTSLEB_AFR <- commandArgs(TRUE)[3]\n",
    "print(CTSLEB_AFR)\n",
    "\n",
    "CTSLEB_AMR <- commandArgs(TRUE)[4]\n",
    "print(CTSLEB_AMR)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[5]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[6]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[7]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[8]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[9]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[10]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "PROSPER_PRSs <- read.delim(PROSPER_PRSs)\n",
    "colnames(PROSPER_PRSs) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_PRSs) - 4)))\n",
    "PROSPER_PRSs <- PROSPER_PRSs[,c(\"IID\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_PRSs) - 4)))]\n",
    "\n",
    "JointPRS_PRS <- read.delim(JointPRS_PRS)\n",
    "colnames(JointPRS_PRS) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"JointPRS_PRS\")\n",
    "JointPRS_PRS <- JointPRS_PRS[,c(\"IID\",\"JointPRS_PRS\")]\n",
    "\n",
    "load(CTSLEB_AFR)\n",
    "CTSLEB_PRS_AFR <- prs_mat_eb;rm(prs_mat_eb)\n",
    "colnames(CTSLEB_PRS_AFR) <- c(\"FID\",\"IID\",paste0(\"AFR_\",colnames(CTSLEB_PRS_AFR)[3:ncol(CTSLEB_PRS_AFR)]))\n",
    "colnames(CTSLEB_PRS_AFR) <- gsub(\"-\",\"_\",colnames(CTSLEB_PRS_AFR))\n",
    "CTSLEB_PRS_AFR <- CTSLEB_PRS_AFR[,c(\"IID\",colnames(CTSLEB_PRS_AFR)[3:ncol(CTSLEB_PRS_AFR)])]\n",
    "\n",
    "load(CTSLEB_AMR)\n",
    "CTSLEB_PRS_AMR <- prs_mat_eb;rm(prs_mat_eb)\n",
    "colnames(CTSLEB_PRS_AMR) <- c(\"FID\",\"IID\",paste0(\"AMR_\",colnames(CTSLEB_PRS_AMR)[3:ncol(CTSLEB_PRS_AMR)]))\n",
    "colnames(CTSLEB_PRS_AMR) <- gsub(\"-\",\"_\",colnames(CTSLEB_PRS_AMR))\n",
    "CTSLEB_PRS_AMR <- CTSLEB_PRS_AMR[,c(\"IID\",colnames(CTSLEB_PRS_AMR)[3:ncol(CTSLEB_PRS_AMR)])]\n",
    "\n",
    "PRS_All <- inner_join(PROSPER_PRSs,JointPRS_PRS)\n",
    "PRS_All <- inner_join(PRS_All,CTSLEB_PRS_AFR)\n",
    "PRS_All <- inner_join(PRS_All,CTSLEB_PRS_AMR)\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_All[PRS_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "prs_train_all <- pheno_training[!is.na(pheno_training[,trait]),-c(1:22)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_All[PRS_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "prs_tune_all <- pheno_tuning[!is.na(pheno_tuning[,trait]),-c(1:22)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_All[PRS_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "prs_validation_all <- pheno_valid[,-c(1:22)]\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "\n",
    "drop <- caret::findLinearCombos(prs_tune_all)$remove\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))\n",
    "\n",
    "\n",
    "mtx <- cor(prs_tune_all)\n",
    "drop <- findCorrelation(mtx,cutoff=0.98)\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))    \n",
    "\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "## SL\n",
    "\n",
    "SL.library <- c(\n",
    "  \"SL.glmnet\",\n",
    "  \"SL.ridge\",\n",
    "  \"SL.glm\",\n",
    "  \"SL.mean\"\n",
    ")\n",
    "sl <- SuperLearner(Y = y_tune, X = prs_tune_all, family = gaussian(),\n",
    "                   # For a real analysis we would use V = 10.\n",
    "                   # V = 3,\n",
    "                   SL.library = SL.library,cvControl = list(V = 20))\n",
    "cvsl <- CV.SuperLearner(Y = y_tune, X = prs_tune_all, family = gaussian(),\n",
    "                        # For a real analysis we would use V = 10.\n",
    "                        # V = 3,\n",
    "                        SL.library = SL.library,cvControl = list(V = 20))\n",
    "\n",
    "best_algorithm <- summary(cvsl)$Table$Algorithm[which.min(summary(cvsl)$Table$Ave)]\n",
    "\n",
    "a <- predict(sl, prs_validation_all, onlySL = FALSE)\n",
    "\n",
    "prs_best_validation_sl <- a$pred\n",
    "prs_best_validation_glmnet <- a$library.predict[,1]\n",
    "prs_best_validation_ridge <- a$library.predict[,2]\n",
    "prs_best_validation_glm <- a$library.predict[,3]\n",
    "\n",
    "if(best_algorithm == \"SL.glmnet_All\"){\n",
    "  #final\n",
    "  prs_best_validation <- prs_best_validation_glmnet\n",
    "}else if(best_algorithm == \"SL.ridge_All\"){\n",
    "  #final\n",
    "  prs_best_validation <- prs_best_validation_ridge\n",
    "}else if(best_algorithm == \"SL.glm_All\"){\n",
    "  #final\n",
    "  prs_best_validation <- prs_best_validation_glm\n",
    "}else{\n",
    "  #final\n",
    "  prs_best_validation <- prs_best_validation_sl\n",
    "}\n",
    "\n",
    "prs_best_validation <- data.frame(IID = pheno_valid$IID,prs = prs_best_validation)\n",
    "\n",
    "\n",
    "a <- predict(sl, prs_train_all, onlySL = FALSE)\n",
    "\n",
    "prs_best_train_sl <- a$pred\n",
    "prs_best_train_glmnet <- a$library.predict[,1]\n",
    "prs_best_train_ridge <- a$library.predict[,2]\n",
    "prs_best_train_glm <- a$library.predict[,3]\n",
    "\n",
    "if(best_algorithm == \"SL.glmnet_All\"){\n",
    "  #final\n",
    "  prs_best_train <- prs_best_train_glmnet\n",
    "}else if(best_algorithm == \"SL.ridge_All\"){\n",
    "  #final\n",
    "  prs_best_train <- prs_best_train_ridge\n",
    "}else if(best_algorithm == \"SL.glm_All\"){\n",
    "  #final\n",
    "  prs_best_train <- prs_best_train_glm\n",
    "}else{\n",
    "  #final\n",
    "  prs_best_train <- prs_best_train_sl\n",
    "}\n",
    "prs_best_train <- data.frame(IID = pheno_training$IID[!is.na(pheno_training[,trait])],prs = prs_best_train)\n",
    "\n",
    "\n",
    "a <- predict(sl, prs_tune_all, onlySL = FALSE)\n",
    "\n",
    "prs_best_tune_sl <- a$pred\n",
    "prs_best_tune_glmnet <- a$library.predict[,1]\n",
    "prs_best_tune_ridge <- a$library.predict[,2]\n",
    "prs_best_tune_glm <- a$library.predict[,3]\n",
    "\n",
    "if(best_algorithm == \"SL.glmnet_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_glmnet\n",
    "}else if(best_algorithm == \"SL.ridge_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_ridge\n",
    "}else if(best_algorithm == \"SL.glm_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_glm\n",
    "}else{\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_sl\n",
    "}\n",
    "prs_best_tune <- data.frame(IID = pheno_tuning$IID[!is.na(pheno_tuning[,trait])],prs = prs_best_tune)\n",
    "\n",
    "pheno_training <- left_join(pheno_training,prs_best_train)\n",
    "pheno_valid <- left_join(pheno_valid,prs_best_validation)\n",
    "pheno_tuning <- left_join(pheno_tuning,prs_best_tune)\n",
    "\n",
    "prs_columns <- c(which(str_detect(colnames(pheno_tuning),\"PROSPER_PRS_\")),which(str_detect(colnames(pheno_tuning),\"JointPRS_PRS\")),which(str_detect(colnames(pheno_tuning),\"AFR_\")),which(str_detect(colnames(pheno_tuning),\"AMR_\")),which(str_detect(colnames(pheno_tuning),\"prs\")))\n",
    "\n",
    "r2_tune <- vector()\n",
    "for(i in 1:length(prs_columns)){\n",
    "  r2_tune[i] <- summary(lm(as.formula(paste0(\"y_tune ~\",colnames(pheno_tuning)[prs_columns[i]])),data = pheno_tuning))$r.squared\n",
    "}\n",
    "\n",
    "if(colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]] == \"prs\"){\n",
    "  if(best_algorithm == \"SL.glmnet_All\"){\n",
    "    beta_sl <- data.frame(Coef = row.names(coef(sl$fitLibrary$SL.glmnet_All$object)),Beta = as.numeric(coef(sl$fitLibrary$SL.glmnet_All$object)))\n",
    "  }else if(best_algorithm == \"SL.ridge_All\"){\n",
    "    beta_sl <- data.frame(Coef = row.names(sl$fitLibrary$SL.ridge_All$bestCoef),Beta = as.numeric(sl$fitLibrary$SL.ridge_All$bestCoef[,1]))\n",
    "  }else if(best_algorithm == \"SL.glm_All\"){\n",
    "    beta_sl <- data.frame(Coef = names(coef(sl$fitLibrary$SL.glm_All$object)),Beta = as.numeric(coef(sl$fitLibrary$SL.glm_All$object)))\n",
    "  }else{\n",
    "    beta_lasso <- data.frame(Coef = row.names(coef(sl$fitLibrary$SL.glmnet_All$object)),Beta = as.numeric(coef(sl$fitLibrary$SL.glmnet_All$object)))\n",
    "    beta_ridge <- data.frame(Coef = row.names(sl$fitLibrary$SL.ridge_All$bestCoef),Beta = as.numeric(sl$fitLibrary$SL.ridge_All$bestCoef[,1]))\n",
    "    beta_glm <- data.frame(Coef = names(coef(sl$fitLibrary$SL.glm_All$object)),Beta = as.numeric(coef(sl$fitLibrary$SL.glm_All$object)))\n",
    "    beta_mean <- data.frame(Coef = names(coef(sl$fitLibrary$SL.glm_All$object)), Beta = 0)\n",
    "    beta_mean$Beta[1] <- sl$fitLibrary$SL.mean_All$object\n",
    "    \n",
    "    beta_sl <- data.frame(Coef = names(coef(sl$fitLibrary$SL.glm_All$object)), Beta = 0)\n",
    "    beta_sl$Beta <- as.numeric(sl$coef[names(sl$coef) == \"SL.glmnet_All\"]) * beta_lasso$Beta + \n",
    "      as.numeric(sl$coef[names(sl$coef) == \"SL.ridge_All\"]) * beta_ridge$Beta + \n",
    "      as.numeric(sl$coef[names(sl$coef) == \"SL.glm_All\"]) * beta_glm$Beta + \n",
    "      as.numeric(sl$coef[names(sl$coef) == \"SL.mean_All\"]) * beta_mean$Beta \n",
    "  }\n",
    "  write.csv(beta_sl,file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)\n",
    "  \n",
    "}else{\n",
    "  beta_coefs <- data.frame(Coef = colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]],Beta = 1)\n",
    "  write.csv(beta_coefs,file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "prs_best_train <- data.frame(IID = pheno_training$IID,prs = pheno_training[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "\n",
    "prs_best_tune <- data.frame(IID = pheno_tuning$IID,prs = pheno_tuning[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "\n",
    "prs_best_validation <- data.frame(IID = pheno_valid$IID,prs = pheno_valid[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "\n",
    "write.table(prs_best_train,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Train_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(prs_best_tune,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Tune_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(prs_best_validation,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Validation_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "\n",
    "pheno_valid_raw <- pheno_valid\n",
    "pheno_valid_adjusted <- pheno_valid\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_valid_adjusted[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]],pheno_valid_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_valid_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_valid_adjusted[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- 0\n",
    "}else{\n",
    "  pheno_valid_adjusted[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_valid_raw_EUR <- pheno_valid_raw[pheno_valid_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_valid_raw_SAS <- pheno_valid_raw[pheno_valid_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_valid_raw_AMR <- pheno_valid_raw[pheno_valid_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_valid_raw_AFR <- pheno_valid_raw[pheno_valid_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_valid_raw_EAS <- pheno_valid_raw[pheno_valid_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_valid_raw_MID <- pheno_valid_raw[pheno_valid_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "pheno_valid_adjusted_EUR <- pheno_valid_adjusted[pheno_valid_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_valid_adjusted_SAS <- pheno_valid_adjusted[pheno_valid_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_valid_adjusted_AMR <- pheno_valid_adjusted[pheno_valid_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_valid_adjusted_AFR <- pheno_valid_adjusted[pheno_valid_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_valid_adjusted_EAS <- pheno_valid_adjusted[pheno_valid_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_valid_adjusted_MID <- pheno_valid_adjusted[pheno_valid_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "pheno_valid_raw_EUR$y_validation <- scale(pheno_valid_raw_EUR$y_validation)\n",
    "pheno_valid_raw_SAS$y_validation <- scale(pheno_valid_raw_SAS$y_validation)\n",
    "pheno_valid_raw_AMR$y_validation <- scale(pheno_valid_raw_AMR$y_validation)\n",
    "pheno_valid_raw_AFR$y_validation <- scale(pheno_valid_raw_AFR$y_validation)\n",
    "pheno_valid_raw_EAS$y_validation <- scale(pheno_valid_raw_EAS$y_validation)\n",
    "pheno_valid_raw_MID$y_validation <- scale(pheno_valid_raw_MID$y_validation)\n",
    "\n",
    "pheno_valid_adjusted_EUR$y_validation <- scale(pheno_valid_adjusted_EUR$y_validation)\n",
    "pheno_valid_adjusted_SAS$y_validation <- scale(pheno_valid_adjusted_SAS$y_validation)\n",
    "pheno_valid_adjusted_AMR$y_validation <- scale(pheno_valid_adjusted_AMR$y_validation)\n",
    "pheno_valid_adjusted_AFR$y_validation <- scale(pheno_valid_adjusted_AFR$y_validation)\n",
    "pheno_valid_adjusted_EAS$y_validation <- scale(pheno_valid_adjusted_EAS$y_validation)\n",
    "pheno_valid_adjusted_MID$y_validation <- scale(pheno_valid_adjusted_MID$y_validation)\n",
    "\n",
    "pheno_valid_raw_EUR[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- scale(pheno_valid_raw_EUR[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "pheno_valid_raw_SAS[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- scale(pheno_valid_raw_SAS[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "pheno_valid_raw_AMR[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- scale(pheno_valid_raw_AMR[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "pheno_valid_raw_AFR[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- scale(pheno_valid_raw_AFR[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "pheno_valid_raw_EAS[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- scale(pheno_valid_raw_EAS[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "pheno_valid_raw_MID[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]] <- scale(pheno_valid_raw_MID[,colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]]])\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_EUR))[2]\n",
    "se_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_EUR))$coefficients[2,2]\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_SAS))[2]\n",
    "se_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_SAS))$coefficients[2,2]\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_AMR))[2]\n",
    "se_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_AMR))$coefficients[2,2]\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_AFR))[2]\n",
    "se_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_AFR))$coefficients[2,2]\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_EAS))[2]\n",
    "se_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_EAS))$coefficients[2,2]\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_MID))[2]\n",
    "se_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_raw_MID))$coefficients[2,2]\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_EUR))[2]\n",
    "se_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_EUR))$coefficients[2,2]\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_SAS))[2]\n",
    "se_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_SAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_AMR))[2]\n",
    "se_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_AMR))$coefficients[2,2]\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_AFR))[2]\n",
    "se_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_AFR))$coefficients[2,2]\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_EAS))[2]\n",
    "se_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_EAS))$coefficients[2,2]\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_MID))[2]\n",
    "se_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",colnames(pheno_tuning)[prs_columns[which.max(r2_tune)]])),data = pheno_valid_adjusted_MID))$coefficients[2,2]\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                                                    beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                                                    se_raw = c(se_validation_raw_EUR,se_validation_raw_SAS,se_validation_raw_AMR,se_validation_raw_AFR,se_validation_raw_EAS,se_validation_raw_MID), \n",
    "                                                    beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                                                    se_adjusted = c(se_validation_adjusted_EUR,se_validation_adjusted_SAS,se_validation_adjusted_AMR,se_validation_adjusted_AFR,se_validation_adjusted_EAS,se_validation_adjusted_MID))\n",
    "                         \n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing OneCommonPRS.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile OneCommonPRS.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${PROSPER_PRSs} ${JointPRS_PRS} ${CTSLEB_AFR} ${CTSLEB_AMR} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input PROSPER_PRSs'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/PRS_All_\",trait,\".sscore\"),\n",
    "    '--input JointPRS_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs/PRS_META_\",trait,\".sscore\"),\n",
    "    '--input CTSLEB_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AFR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input CTSLEB_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AMR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/OneCommonPRS.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://OneCommonPRS.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 18.7 KiB/ 18.7 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/18.7 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp OneCommonPRS.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: onecommonp--williamsjacr--240731-191832-07\n",
      "  job-name: onecommonprs\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5269959543922159773\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4104167026320805035\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17363237223867866755\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14958567778078369968\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6334697042897950361\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12828164815721406103\n",
      "Launched job-id: onecommonp--williamsjacr--240731-191832-07\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'onecommonp--williamsjacr--240731-191832-07' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'onecommonp--williamsjacr--240731-191832-07' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script OneCommonPRS.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name        Task  Status    Last Update\r\n",
      "------------  ------  --------  --------------\r\n",
      "onecommonprs       6  Pending   07-31 19:18:32\r\n",
      "onecommonprs       5  Pending   07-31 19:18:32\r\n",
      "onecommonprs       4  Pending   07-31 19:18:32\r\n",
      "onecommonprs       3  Pending   07-31 19:18:32\r\n",
      "onecommonprs       2  Pending   07-31 19:18:32\r\n",
      "onecommonprs       1  Pending   07-31 19:18:32\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'onecommonp--williamsjacr--240731-191832-07' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting NullModel.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile NullModel.R\n",
    "rm(list = ls())\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[1]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[2]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[3]\n",
    "print(all_validation_file)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(dplyr)\n",
    "library(STAAR)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(Matrix)\n",
    "library(SCANG)\n",
    "library(STAARpipeline)\n",
    "\n",
    "trait <- \"BMI\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"TC\",\"LDL\",\"HDL\",\"logTG\",\"Height\")){\n",
    "  pheno_train <- read.delim(all_train_file)\n",
    "  \n",
    "  common_prs <- read.delim(paste0(INPUT_PATH,\"/\",trait,\"_Best_Train_All.txt\"))\n",
    "  \n",
    "  pheno_train <- inner_join(pheno_train,common_prs,by = \"IID\")\n",
    "  \n",
    "  obj.STAAR.UKB <- fit_nullmodel(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10 + prs\")), data = pheno_train,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "  save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Train_Null_Model.RData\"))\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  pheno_tune <- read.delim(all_tune_file)\n",
    "  \n",
    "  common_prs <- read.delim(paste0(INPUT_PATH,\"/\",trait,\"_Best_Tune_All.txt\"))\n",
    "  \n",
    "  pheno_tune <- inner_join(pheno_tune,common_prs,by = \"IID\")\n",
    "  \n",
    "  obj.STAAR.UKB <- fit_nullmodel(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10 + prs\")), data = pheno_tune,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "  save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Tune_Null_Model.RData\"))\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  pheno_validation <- read.delim(all_validation_file)\n",
    "  \n",
    "  common_prs <- read.delim(paste0(INPUT_PATH,\"/\",trait,\"_Best_Validation_All.txt\"))\n",
    "  \n",
    "  pheno_validation <- inner_join(pheno_validation,common_prs,by = \"IID\")\n",
    "  \n",
    "  obj.STAAR.UKB <- fit_nullmodel(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10 + prs\")), data = pheno_validation,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "  save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Validation_Null_Model.RData\"))\n",
    "  \n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/NullModel.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting NullModel.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile NullModel.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${all_train_file} ${all_tune_file} ${all_validation_file} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://NullModel.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.0 KiB/  2.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.0 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp NullModel.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: nullmodel--williamsjacr--240801-214711-99\n",
      "  job-name: nullmodel\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4654998437295274832\n",
      "Launched job-id: nullmodel--williamsjacr--240801-214711-99\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240801-214711-99' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240801-214711-99' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script NullModel.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name      Task  Status    Last Update\r\n",
      "----------  ------  --------  --------------\r\n",
      "nullmodel        1  Success   08-01 21:52:06\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240801-214711-99' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error in library(SeqArray) : there is no package called ‘SeqArray’\r\n",
      "Execution halted\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/nullmodel--williamsjacr--240801-131108-38.1-stderr.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /home/jupyter/aou_dsub.bash\n"
     ]
    }
   ],
   "source": [
    "%%writefile ~/aou_dsub.bash\n",
    "#!/bin/bash\n",
    "function aou_dsub () {\n",
    "\n",
    "  # Get a shorter username to leave more characters for the job name.\n",
    "  local DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "  # For AoU RWB projects network name is \"network\".\n",
    "  local AOU_NETWORK=network\n",
    "  local AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "  dsub \\\n",
    "      --provider google-cls-v2 \\\n",
    "      --user-project \"${GOOGLE_PROJECT}\"\\\n",
    "      --project \"${GOOGLE_PROJECT}\"\\\n",
    "      --image 'marketplace.gcr.io/google/ubuntu1804:latest' \\\n",
    "      --network \"${AOU_NETWORK}\" \\\n",
    "      --subnetwork \"${AOU_SUBNETWORK}\" \\\n",
    "      --service-account \"$(gcloud config get-value account)\" \\\n",
    "      --user \"${DSUB_USER_NAME}\" \\\n",
    "      --regions us-central1 \\\n",
    "      --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "      \"$@\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo source ~/aou_dsub.bash >> ~/.bashrc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "code_folding": [
     81,
     271
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ncRNA_Analysis.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile ncRNA_Analysis.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "trait <- commandArgs(TRUE)[2]\n",
    "print(trait)\n",
    "\n",
    "Train_Null_Model <- commandArgs(TRUE)[3]\n",
    "print(Train_Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[4]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[5]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "\n",
    "obj_nullmodel <- get(load(Train_Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output file name\n",
    "output_file_name <- paste0(OUTPUT_PATH,\"/\",trait,\"_AoU_WGS_ncRNA_Train\")\n",
    "\n",
    "## gene number in job\n",
    "gene_num_in_array <- 100 \n",
    "group.num.allchr <- ceiling(table(ncRNA_gene[,1])/gene_num_in_array)\n",
    "\n",
    "chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if (chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "ncRNA_gene_chr <- ncRNA_gene[ncRNA_gene[,1]==chr,]\n",
    "ncRNA_longmasks <- c(\"KCNQ1OT1\",\"AC006548.28\",\"RP3-394A18.1\",\"RP3-323A16.1\",\"RP5-1039K5.19\",\"SNHG14\",\"LL22NC03-86G7.1\")\n",
    "ncRNA_gene_chr <- ncRNA_gene_chr[!ncRNA_gene_chr[,2] %in% ncRNA_longmasks,]\n",
    "sub_seq_num <- dim(ncRNA_gene_chr)[1]\n",
    "\n",
    "if(groupid < group.num){\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "###########################################################\n",
    "#           Main Function \n",
    "###########################################################\n",
    "\n",
    "ncRNA_nolongmask <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                             rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                             QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                             Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                             Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                             SPA_p_filter=TRUE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ## ncRNA SNVs\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  is.in <- ((GENCODE.Category==\"ncRNA_exonic\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\"))&(SNVlist)\n",
    "  \n",
    "  variant.id.ncRNA <- variant.id[is.in]\n",
    "  \n",
    "  rm(GENCODE.Category)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.ncRNA,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.ncRNA)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[;]\")\n",
    "  Gene <- as.character(sapply(GENCODE.Info.split,function(z) gsub(\"\\\\(.*\\\\)\",\"\",z[1])))\n",
    "  \n",
    "  Gene_list_1 <- as.character(sapply(strsplit(Gene,','),'[',1))\n",
    "  Gene_list_2 <- as.character(sapply(strsplit(Gene,','),'[',2))\n",
    "  Gene_list_3 <- as.character(sapply(strsplit(Gene,','),'[',3))\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  variant.id.ncRNA <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- union(which(Gene_list_1==gene_name),which(Gene_list_2==gene_name))\n",
    "  is.in <- union(is.in,which(Gene_list_3==gene_name))\n",
    "  \n",
    "  variant.is.in <- variant.id.ncRNA[is.in]\n",
    "  \n",
    "  if(length(variant.is.in) > 10000){\n",
    "    return(NULL)\n",
    "  }\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"ncRNA\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "agds.path <- GDS_File\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "gene_centric_ncRNA_dnanexus <- function(ncRNA_gene_chr,gene_name,chr,genofile,obj_nullmodel,rare_maf_cutoff,\n",
    "                                        QC_label,variant_type,geno_missing_imputation,\n",
    "                                        Annotation_dir,Annotation_name_catalog,\n",
    "                                        Use_annotation_weights,Annotation_name,silent){\n",
    "  results <- try(ncRNA_nolongmask(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                  QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                  Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                  Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=silent),silent=TRUE)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "results_ncRNA <- c()\n",
    "for(kk in sub_seq_id){\n",
    "  print(kk)\n",
    "  gene_name <- ncRNA_gene_chr[kk,2]\n",
    "  results <- gene_centric_ncRNA_dnanexus(ncRNA_gene_chr=ncRNA_gene_chr,gene_name = gene_name,chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=max.maf,\n",
    "                                         QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                         Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=TRUE)\n",
    "  \n",
    "  results_ncRNA <- append(results_ncRNA,results)\n",
    "}\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "rm(list=setdiff(ls(), c(\"results_ncRNA\", \"output_file_name\",\"arrayid\",\"agds.path\",\"trait\"))); gc()\n",
    "save(results_ncRNA, file = paste0(output_file_name,\"_\",arrayid,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ncRNA_Analysis.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile ncRNA_Analysis.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${trait} ${Train_Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"Height\",\"TC\")){\n",
    "    for(arrayid in 222){\n",
    "        if(arrayid < 19){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 35){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 46){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 56){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 69){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 79){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 89){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 100){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 108){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 117){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 128){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 147){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 156){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 166){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 178){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 191){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 198){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 208){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 214){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 218){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--env trait'=trait,\n",
    "            '--input Train_Null_Model'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/acaf_threshold_v7/acaf_threshold.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/ncRNA_Analysis.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentric_ncRNA\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "}  \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://ncRNA_Analysis.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 10.5 KiB/ 10.5 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/10.5 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp ncRNA_Analysis.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: ncrna-anal--williamsjacr--240802-141249-95\n",
      "  job-name: ncrna-analysis\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15945330792918160714\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12474725655731960867\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10526454129034710544\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2471146011125752468\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16532595984810075093\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6231998838750485626\n",
      "Launched job-id: ncrna-anal--williamsjacr--240802-141249-95\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ncrna-anal--williamsjacr--240802-141249-95' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ncrna-anal--williamsjacr--240802-141249-95' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 32 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script ncRNA_Analysis.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name          Task  Status    Last Update\r\n",
      "--------------  ------  --------  --------------\r\n",
      "ncrna-analysis       6  Success   08-02 14:27:00\r\n",
      "ncrna-analysis       5  Success   08-02 14:24:56\r\n",
      "ncrna-analysis       4  Success   08-02 14:24:33\r\n",
      "ncrna-analysis       3  Success   08-02 14:24:57\r\n",
      "ncrna-analysis       2  Success   08-02 14:24:48\r\n",
      "ncrna-analysis       1  Success   08-02 14:25:09\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ncrna-anal--williamsjacr--240802-141249-95' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         used (Mb) gc trigger (Mb) max used (Mb)\r\n",
      "Ncells 294492 15.8     692571   37   480244 25.7\r\n",
      "Vcells 510864  3.9    8388608   64  1926841 14.8\r\n",
      "[1] \"Height\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/Height_Train_Null_Model.RData\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/acaf_threshold_v7/acaf_threshold.chr22.gds\"\r\n",
      "[1] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentric_ncRNA\"\r\n",
      "[1] 401\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 402\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 111\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 403\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 404\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 13\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 405\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 406\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 407\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 193\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 408\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 3\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 409\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 410\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 2\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 411\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 412\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 413\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 414\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 415\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 416\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 51\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 417\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 418\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 419\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 420\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 30\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 421\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 82\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 422\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 423\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 424\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 425\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 426\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 427\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 428\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 429\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 430\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 431\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 432\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 433\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 135\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 434\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 435\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 174\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 436\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 437\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 438\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 439\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 440\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 441\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 442\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 443\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 8\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 444\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 445\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 446\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 447\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 448\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 449\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 2\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 450\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 451\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 453\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 454\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 455\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 456\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 457\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 458\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 459\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 460\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 461\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 462\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 463\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 464\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 465\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 466\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 467\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 468\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 469\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 4\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 470\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 471\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 472\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 473\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 474\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 475\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 476\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 477\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 478\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 18,254\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 155,611\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells  9901586 528.9   16402042 876.0 12330846 658.6\r\n",
      "Vcells 19966829 152.4   66531959 507.6 93247597 711.5\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/ncrna-anal--williamsjacr--240802-141249-95.5-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "code_folding": [
     85,
     184,
     1082,
     1213,
     1344,
     1474,
     1604
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Noncoding_Analysis.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Noncoding_Analysis.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "trait <- commandArgs(TRUE)[2]\n",
    "print(trait)\n",
    "\n",
    "Train_Null_Model <- commandArgs(TRUE)[3]\n",
    "print(Train_Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[4]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[5]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "\n",
    "obj_nullmodel <- get(load(Train_Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\t\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output file name\n",
    "output_file_name <- paste0(OUTPUT_PATH,\"/\",trait,\"_AoU_WGS_Noncoding_Train\")\n",
    "\n",
    "###########################################################\n",
    "#           Main Function \n",
    "###########################################################\n",
    "\n",
    "\n",
    "## gene number in job\n",
    "gene_num_in_array <- 50 \n",
    "group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "\n",
    "chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if(chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "noncoding_longmasks <- c(\"ARID4B\",\"RERE\",\"CRIM1\",\"EML4\",\"EML6\",\"GLI2\",\"ITGB6\",\"FNDC3B\",\"AFAP1\",\"TRIO\",\n",
    "                         \"NR3C1\",\"SGK1\",\"PRDM1\",\"TNS3\",\"GNA12\",\"HOXA2\",\"PLEC\",\"ASAP1\",\"CSGALNACT1\",\"RAPGEF1\",\n",
    "                         \"KLF6\",\"DNAJB12\",\"ABLIM1\",\"ETV6\",\"IGF1R\",\"RMI2\",\"CMIP\",\"HOXB2\",\"SEPT9\",\"SPECC1\",\n",
    "                         \"CIRBP\",\"ADNP\",\"CDC42EP3\",\"EPAS1\",\"MGAT5\")\n",
    "genes_info_chr <- genes_info_chr[!genes_info_chr[,1] %in% noncoding_longmasks,]\n",
    "sub_seq_num <- dim(genes_info_chr)[1]\n",
    "\n",
    "if(groupid < group.num){\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "Gene_Centric_Noncoding_preload <- function(chr,gene_name,category=c(\"all_categories\",\"downstream\",\"upstream\",\"UTR\",\"promoter_CAGE\",\"promoter_DHS\",\"enhancer_CAGE\",\"enhancer_DHS\"),\n",
    "                                           genofile,obj_nullmodel,\n",
    "                                           dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                                           dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                                           dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                                           dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                                           rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                           QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                           Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                           Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                           SPA_p_filter=TRUE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  if(category==\"all_categories\"){\n",
    "    results <- noncoding_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                 dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                 dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                 dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                 dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                 rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                 Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                 Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                 SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"downstream\"){\n",
    "    results <- downstream(chr,gene_name,genofile,obj_nullmodel,\n",
    "                          rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                          QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                          Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                          Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                          SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"upstream\"){\n",
    "    results <- upstream(chr,gene_name,genofile,obj_nullmodel,\n",
    "                        rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                        Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                        SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"UTR\"){\n",
    "    results <- UTR(chr,gene_name,genofile,obj_nullmodel,\n",
    "                   rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                   QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                   Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                   Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                   SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"promoter_CAGE\"){\n",
    "    results <- promoter_CAGE_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                     dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                     rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                     QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                     Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                     Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                     SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"promoter_DHS\"){\n",
    "    results <- promoter_DHS_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                    dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                    rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                    QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                    Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                    Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                    SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"enhancer_CAGE\"){\n",
    "    results <- enhancer_CAGE_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                     dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                     rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                     QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                     Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                     Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                     SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"enhancer_DHS\"){\n",
    "    results <- enhancer_DHS_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                    dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                    rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                    QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                    Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                    Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                    SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "noncoding_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                              dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                              dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                              dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                              dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                              rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                              QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                              Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                              Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                              SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  #####################################\n",
    "  #   Gene Info\n",
    "  ## get SNV id\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ########################################\n",
    "  #   Downstream\n",
    "  ## downstream SNVs\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  is.in <- (GENCODE.Category==\"downstream\")&(SNVlist)\n",
    "  variant.id.downstream <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.downstream,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.downstream)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[,]\")\n",
    "  variant_gene_num <- sapply(GENCODE.Info.split,function(z) length(z))\n",
    "  \n",
    "  variant.id.SNV <- seqGetData(genofile, \"variant.id\")\n",
    "  variant.id.SNV <- rep(variant.id.SNV,variant_gene_num)\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  rm(variant_gene_num)\n",
    "  gc()\n",
    "  \n",
    "  Gene <- as.character(unlist(GENCODE.Info.split))\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(Gene==gene_name)\n",
    "  variant.is.in <- variant.id.SNV[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_downstream <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"downstream\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_downstream <- rbind(results_downstream,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_downstream)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_downstream) <- colnames(results_downstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_downstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_downstream)[(dim(results_downstream)[2]-1):dim(results_downstream)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_downstream) <- colnames(results_downstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_downstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_downstream)[dim(results_downstream)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ########################################\n",
    "  #   Upstream\n",
    "  \n",
    "  is.in <- (GENCODE.Category==\"upstream\")&(SNVlist)\n",
    "  variant.id.upstream <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.upstream,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.upstream)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[,]\")\n",
    "  variant_gene_num <- sapply(GENCODE.Info.split,function(z) length(z))\n",
    "  \n",
    "  variant.id.SNV <- seqGetData(genofile, \"variant.id\")\n",
    "  variant.id.SNV <- rep(variant.id.SNV,variant_gene_num)\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  rm(variant_gene_num)\n",
    "  gc()\n",
    "  \n",
    "  Gene <- as.character(unlist(GENCODE.Info.split))\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(Gene==gene_name)\n",
    "  variant.is.in <- variant.id.SNV[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_upstream <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"upstream\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_upstream <- rbind(results_upstream,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_upstream)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_upstream) <- colnames(results_upstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_upstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_upstream)[(dim(results_upstream)[2]-1):dim(results_upstream)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_upstream) <- colnames(results_upstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_upstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_upstream)[dim(results_upstream)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ########################################################\n",
    "  #                UTR\n",
    "  \n",
    "  is.in <- ((GENCODE.Category==\"UTR3\")|(GENCODE.Category==\"UTR5\")|(GENCODE.Category==\"UTR5;UTR3\"))&(SNVlist)\n",
    "  variant.id.UTR <- variant.id[is.in]\n",
    "  \n",
    "  rm(GENCODE.Category)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.UTR,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.UTR)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[(]\")\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  # Gene <- as.character(sapply(GENCODE.Info.split,function(z) z[seq(1,length(z),2)]))\n",
    "  Gene <- as.character(sapply(GENCODE.Info.split,function(z) z[1]))\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  variant.id.SNV <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(Gene==gene_name)\n",
    "  variant.is.in <- variant.id.SNV[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_UTR <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"UTR\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_UTR <- rbind(results_UTR,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_UTR)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_UTR) <- colnames(results_UTR, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_UTR)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_UTR)[(dim(results_UTR)[2]-1):dim(results_UTR)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_UTR) <- colnames(results_UTR, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_UTR)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_UTR)[dim(results_UTR)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  #############################################\n",
    "  #   Promoter-CAGE\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_promoter_CAGE <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_promoter_CAGE <- rbind(results_promoter_CAGE,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_promoter_CAGE)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_promoter_CAGE) <- colnames(results_promoter_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_CAGE)[(dim(results_promoter_CAGE)[2]-1):dim(results_promoter_CAGE)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_promoter_CAGE) <- colnames(results_promoter_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_CAGE)[dim(results_promoter_CAGE)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ##################################################\n",
    "  #       Promoter-DHS\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_promoter_DHS <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_promoter_DHS <- rbind(results_promoter_DHS ,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_promoter_DHS)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_promoter_DHS) <- colnames(results_promoter_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_DHS)[(dim(results_promoter_DHS)[2]-1):dim(results_promoter_DHS)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_promoter_DHS) <- colnames(results_promoter_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_DHS)[dim(results_promoter_DHS)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ###########################################\n",
    "  #        Enhancer-CAGE\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_enhancer_CAGE <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_enhancer_CAGE <- rbind(results_enhancer_CAGE,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_enhancer_CAGE)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_enhancer_CAGE) <- colnames(results_enhancer_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_CAGE)[(dim(results_enhancer_CAGE)[2]-1):dim(results_enhancer_CAGE)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_enhancer_CAGE) <- colnames(results_enhancer_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_CAGE)[dim(results_enhancer_CAGE)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ##################################################\n",
    "  #       Enhancer-DHS\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_enhancer_DHS <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_enhancer_DHS <- rbind(results_enhancer_DHS,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_enhancer_DHS)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_enhancer_DHS) <- colnames(results_enhancer_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_DHS)[(dim(results_enhancer_DHS)[2]-1):dim(results_enhancer_DHS)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_enhancer_DHS) <- colnames(results_enhancer_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_DHS)[dim(results_enhancer_DHS)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ############################################\n",
    "  #           results\n",
    "  \n",
    "  results_noncoding <- list(upstream=results_upstream,downstream=results_downstream,UTR=results_UTR,\n",
    "                            promoter_CAGE=results_promoter_CAGE,promoter_DHS=results_promoter_DHS,\n",
    "                            enhancer_CAGE=results_enhancer_CAGE,enhancer_DHS=results_enhancer_DHS)\n",
    "  \n",
    "  return(results_noncoding)\n",
    "  \n",
    "}\n",
    "\n",
    "promoter_CAGE_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                  dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                                  rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                  QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                  Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                  Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                  SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "promoter_DHS_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                 dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                                 rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                 QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                 Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                 SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "enhancer_CAGE_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                  dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                                  rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                  QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                  Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                  Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                  SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "enhancer_DHS_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                 dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                                 rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                 QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                 Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                 SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "gene_centric_noncoding_dnanexus <- function(genes_info_chr,gene_name,chr,genofile,obj_nullmodel,\n",
    "                                            dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                                            dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                                            dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                                            dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                                            rare_maf_cutoff,QC_label,variant_type,geno_missing_imputation,\n",
    "                                            Annotation_dir,Annotation_name_catalog,\n",
    "                                            Use_annotation_weights,Annotation_name,silent){\n",
    "  results <- try(Gene_Centric_Noncoding_preload(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,\n",
    "                                                dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                                dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                                dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                                dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                                rare_maf_cutoff=rare_maf_cutoff,QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                                Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=silent),silent=TRUE)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "#########################################################\n",
    "#             Promoter_CAGE\n",
    "#########################################################\n",
    "\n",
    "agds.path <- GDS_File\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "promGobj <- promoters(genes(txdb), upstream = 3000, downstream = 3000)\n",
    "\n",
    "#Subsetting Promoters that within +/-3kb of TSS and have CAGE signals\n",
    "CAGEAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"CAGE\")]))\n",
    "CAGEBvt <- CAGEAnno!=\"\"\n",
    "CAGEidx <- which(CAGEBvt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[CAGEidx])\n",
    "seqSetFilter(genofile,promGobj,intersect=TRUE)\n",
    "CAGEpromgene <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "CAGEGene <- unlist(lapply(strsplit(CAGEpromgene,\"\\\\(|\\\\,|;|-\"),`[[`,1))\n",
    "##obtain variants info\n",
    "CAGEvchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "CAGEvpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "CAGEvref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "CAGEvalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfPromCAGEVarGene <- data.frame(CAGEvchr,CAGEvpos,CAGEvref,CAGEvalt,CAGEGene)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.PromCAGE <- variant.id[SNVlist]\n",
    "\n",
    "dfPromCAGEVarGene.SNV <- dfPromCAGEVarGene[SNVlist,]\n",
    "dfPromCAGEVarGene.SNV$CAGEvpos <- as.character(dfPromCAGEVarGene.SNV$CAGEvpos)\n",
    "dfPromCAGEVarGene.SNV$CAGEvref <- as.character(dfPromCAGEVarGene.SNV$CAGEvref)\n",
    "dfPromCAGEVarGene.SNV$CAGEvalt <- as.character(dfPromCAGEVarGene.SNV$CAGEvalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfPromCAGEVarGene)\n",
    "gc()\n",
    "\n",
    "#########################################################\n",
    "#             Promoter_DHS\n",
    "#########################################################\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "promGobj <- promoters(genes(txdb), upstream = 3000, downstream = 3000)\n",
    "\n",
    "# Subsetting Promoters that within +/-3kb of TSS and have rOCRs signals\n",
    "rOCRsAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"DHS\")]))\n",
    "rOCRsBvt <- rOCRsAnno!=\"\"\n",
    "rOCRsidx <- which(rOCRsBvt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[rOCRsidx])\n",
    "\n",
    "seqSetFilter(genofile,promGobj,intersect=TRUE)\n",
    "rOCRspromgene <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "rOCRsGene <- unlist(lapply(strsplit(rOCRspromgene,\"\\\\(|\\\\,|;|-\"),`[[`,1))\n",
    "## obtain variants info\n",
    "rOCRsvchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "rOCRsvpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "rOCRsvref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "rOCRsvalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfPromrOCRsVarGene <- data.frame(rOCRsvchr,rOCRsvpos,rOCRsvref,rOCRsvalt,rOCRsGene)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.PromrOCRs <- variant.id[SNVlist]\n",
    "\n",
    "dfPromrOCRsVarGene.SNV <- dfPromrOCRsVarGene[SNVlist,]\n",
    "dfPromrOCRsVarGene.SNV$rOCRsvpos <- as.character(dfPromrOCRsVarGene.SNV$rOCRsvpos)\n",
    "dfPromrOCRsVarGene.SNV$rOCRsvref <- as.character(dfPromrOCRsVarGene.SNV$rOCRsvref)\n",
    "dfPromrOCRsVarGene.SNV$rOCRsvalt <- as.character(dfPromrOCRsVarGene.SNV$rOCRsvalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfPromrOCRsVarGene)\n",
    "gc()\n",
    "\n",
    "#########################################################\n",
    "#             Enhancer_CAGE\n",
    "#########################################################\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "\n",
    "#Now extract the GeneHancer with CAGE Signal Overlay\n",
    "genehancerAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "genehancer <- genehancerAnno!=\"\"\n",
    "\n",
    "CAGEAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"CAGE\")]))\n",
    "CAGE <- CAGEAnno!=\"\"\n",
    "CAGEGeneHancervt <- CAGEAnno!=\"\"&genehancerAnno!=\"\"\n",
    "CAGEGeneHanceridx <- which(CAGEGeneHancervt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[CAGEGeneHanceridx])\n",
    "\n",
    "# variants that covered by whole GeneHancer without CAGE overlap.\n",
    "genehancerSet <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "enhancerGene <- unlist(lapply(strsplit(genehancerSet,\"=\"),`[[`,4))\n",
    "enhancer2GENE <- unlist(lapply(strsplit(enhancerGene,\";\"),`[[`,1))\n",
    "enhancervchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "enhancervpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "enhancervref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "enhancervalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfHancerCAGEVarGene <- data.frame(enhancervchr,enhancervpos,enhancervref,enhancervalt,enhancer2GENE)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.HancerCAGE <- variant.id[SNVlist]\n",
    "\n",
    "dfHancerCAGEVarGene.SNV <- dfHancerCAGEVarGene[SNVlist,]\n",
    "dfHancerCAGEVarGene.SNV$enhancervpos <- as.character(dfHancerCAGEVarGene.SNV$enhancervpos)\n",
    "dfHancerCAGEVarGene.SNV$enhancervref <- as.character(dfHancerCAGEVarGene.SNV$enhancervref)\n",
    "dfHancerCAGEVarGene.SNV$enhancervalt <- as.character(dfHancerCAGEVarGene.SNV$enhancervalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfHancerCAGEVarGene)\n",
    "gc()\n",
    "\n",
    "#########################################################\n",
    "#             Enhancer_DHS\n",
    "#########################################################\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "\n",
    "#Now extract the GeneHancer with rOCRs Signal Overlay\n",
    "genehancerAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "genehancer <- genehancerAnno!=\"\"\n",
    "\n",
    "rOCRsAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"DHS\")]))\n",
    "rOCRs <- rOCRsAnno!=\"\"\n",
    "rOCRsGeneHancervt <- rOCRsAnno!=\"\"&genehancerAnno!=\"\"\n",
    "rOCRsGeneHanceridx <- which(rOCRsGeneHancervt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[rOCRsGeneHanceridx])\n",
    "# variants that covered by whole GeneHancer without rOCRs overlap.\n",
    "\n",
    "genehancerSet <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "enhancerGene <- unlist(lapply(strsplit(genehancerSet,\"=\"),`[[`,4))\n",
    "enhancer2GENE <- unlist(lapply(strsplit(enhancerGene,\";\"),`[[`,1))\n",
    "enhancervchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "enhancervpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "enhancervref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "enhancervalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfHancerrOCRsVarGene <- data.frame(enhancervchr,enhancervpos,enhancervref,enhancervalt,enhancer2GENE)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.HancerrOCRs <- variant.id[SNVlist]\n",
    "\n",
    "dfHancerrOCRsVarGene.SNV <- dfHancerrOCRsVarGene[SNVlist,]\n",
    "dfHancerrOCRsVarGene.SNV$enhancervpos <- as.character(dfHancerrOCRsVarGene.SNV$enhancervpos)\n",
    "dfHancerrOCRsVarGene.SNV$enhancervref <- as.character(dfHancerrOCRsVarGene.SNV$enhancervref)\n",
    "dfHancerrOCRsVarGene.SNV$enhancervalt <- as.character(dfHancerrOCRsVarGene.SNV$enhancervalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfHancerrOCRsVarGene)\n",
    "gc()\n",
    "\n",
    "results_noncoding <- list()\n",
    "count <- 1\n",
    "for(kk in sub_seq_id){\n",
    "  print(kk)\n",
    "  gene_name <- genes_info_chr[kk,1]\n",
    "  results <- gene_centric_noncoding_dnanexus(genes_info_chr=genes_info_chr,gene_name = gene_name,chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,\n",
    "                                             dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                             dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                             dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                             dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                             rare_maf_cutoff=0.01,QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                             Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                             Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=TRUE)\n",
    "  \n",
    "  results_noncoding[[count]] <-results\n",
    "  count <- count + 1\n",
    "}\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "rm(list=setdiff(ls(), c(\"results_noncoding\", \"output_file_name\",\"arrayid\",\"agds.path\",\"trait\"))); gc()\n",
    "save(results_noncoding, file = paste0(output_file_name,\"_\",arrayid,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Noncoding_Analysis.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Noncoding_Analysis.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${trait} ${Train_Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"Height\",\"TC\")){\n",
    "    for(arrayid in 379){\n",
    "        if(arrayid < 41){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 66){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 87){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 102){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 120){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 159){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 173){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 189){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 204){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 230){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 250){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 257){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 269){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 281){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 298){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 321){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 327){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 355){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 366){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 371){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--env trait'=trait,\n",
    "            '--input Train_Null_Model'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/acaf_threshold_v7/acaf_threshold.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Noncoding_Analysis.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricNoncoding\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "}  \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Noncoding_Analysis.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 75.8 KiB/ 75.8 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/75.8 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Noncoding_Analysis.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/bin/dsub\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/jupyter/.local/lib/python3.10/site-packages/dsub/commands/dsub.py\", line 1150, in main\n",
      "    dsub_main(prog, argv)\n",
      "  File \"/home/jupyter/.local/lib/python3.10/site-packages/dsub/commands/dsub.py\", line 1135, in dsub_main\n",
      "    launched_job = run_main(args)\n",
      "  File \"/home/jupyter/.local/lib/python3.10/site-packages/dsub/commands/dsub.py\", line 1212, in run_main\n",
      "    return run(\n",
      "  File \"/home/jupyter/.local/lib/python3.10/site-packages/dsub/commands/dsub.py\", line 1284, in run\n",
      "    raise ValueError('Requesting retries requires requesting wait')\n",
      "ValueError: Requesting retries requires requesting wait\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\nsource ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\\n\\naou_dsub \\\\\\n  --image willja16/r_with_plink \\\\\\n  --disk-size 100 \\\\\\n  --boot-disk-size 25 \\\\\\n  --min-ram 32 \\\\\\n  --preemptible \\\\\\n  --retries 1 \\\\\\n  --timeout \"96h\" \\\\\\n  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\\\\n  --script Noncoding_Analysis.sh \\\\\\n  --tasks score_task.txt\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[110], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m--out score_batch\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43msource ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43maou_dsub \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --image willja16/r_with_plink \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --disk-size 100 \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --boot-disk-size 25 \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --min-ram 32 \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --preemptible \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --retries 1 \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --timeout \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m96h\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --logging \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m$\u001b[39;49m\u001b[38;5;132;43;01m{WORKSPACE_BUCKET}\u001b[39;49;00m\u001b[38;5;124;43m/data/logging\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --script Noncoding_Analysis.sh \u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m  --tasks score_task.txt\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2478\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2476\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2477\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2478\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2480\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2481\u001b[0m \u001b[38;5;66;03m# when using magics with decodator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2482\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2483\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/IPython/core/magics/script.py:154\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    153\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 154\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/IPython/core/magics/script.py:314\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    310\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    313\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 314\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\nsource ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\\n\\naou_dsub \\\\\\n  --image willja16/r_with_plink \\\\\\n  --disk-size 100 \\\\\\n  --boot-disk-size 25 \\\\\\n  --min-ram 32 \\\\\\n  --preemptible \\\\\\n  --retries 1 \\\\\\n  --timeout \"96h\" \\\\\\n  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\\\\n  --script Noncoding_Analysis.sh \\\\\\n  --tasks score_task.txt\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 32 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Noncoding_Analysis.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                          Last Update\r\n",
      "---------------  ------  ------------------------------  --------------\r\n",
      "noncoding-an...       6  Success                         08-02 14:28:26\r\n",
      "noncoding-an...       5  Success                         08-02 14:35:16\r\n",
      "noncoding-an...       4  Success                         08-02 14:28:42\r\n",
      "noncoding-an...       3  Success                         08-02 14:29:28\r\n",
      "noncoding-an...       2  Success                         08-02 14:28:26\r\n",
      "noncoding-an...       1  Started running \"user-command\"  08-02 14:18:37\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'noncoding---williamsjacr--240802-141312-91' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         used (Mb) gc trigger (Mb) max used (Mb)\r\n",
      "Ncells 294492 15.8     692571   37   480244 25.7\r\n",
      "Vcells 510864  3.9    8388608   64  1926841 14.8\r\n",
      "[1] \"TC\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/TC_Train_Null_Model.RData\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\"\r\n",
      "[1] \"/mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/acaf_threshold_v7/acaf_threshold.chr22.gds\"\r\n",
      "[1] \"/mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricNoncoding\"\r\n",
      "# of selected variants: 19,697\r\n",
      "# of selected variants: 6,975\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10008001 534.5   16402044 876.0 12330846 658.6\r\n",
      "Vcells 24815785 189.4   42645776 325.4 35469499 270.7\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10008168 534.5   16402044 876.0 12330846 658.6\r\n",
      "Vcells 24860376 189.7   42645776 325.4 35469499 270.7\r\n",
      "# of selected variants: 201,885\r\n",
      "# of selected variants: 28,970\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10047002 536.6   16402044 876.0 12330846 658.6\r\n",
      "Vcells 26851653 204.9   42645776 325.4 42645776 325.4\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10047014 536.6   16402044 876.0 12330846 658.6\r\n",
      "Vcells 27027792 206.3   42645776 325.4 42645776 325.4\r\n",
      "# of selected variants: 12,056\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10053339 537.0   16402044 876.0 12330846 658.6\r\n",
      "Vcells 29933358 228.4   51254931 391.1 42645776 325.4\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10053351 537.0   16402044 876.0 12330846 658.6\r\n",
      "Vcells 29965503 228.7   51254931 391.1 42645776 325.4\r\n",
      "# of selected variants: 83,340\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10056185 537.1   16402044 876.0 12330846 658.6\r\n",
      "Vcells 31584140 241.0   51254931 391.1 42645776 325.4\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells 10056197 537.1   16402044 876.0 12330846 658.6\r\n",
      "Vcells 32108089 245.0   51254931 391.1 42645776 325.4\r\n",
      "[1] 401\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 39\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 21\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 20\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 83\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 51\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 143\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 402\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 7\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 8\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 65\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 4\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 104\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 403\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 7\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 12\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 16\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 32\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 22\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 46\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 404\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 13\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 6\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 17\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 34\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 405\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 14\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 20\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 19\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 20\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 59\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 21\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 50\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 406\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 18\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 23\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 8\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 66\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 19\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 99\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 407\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 29\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 8\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 24\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 58\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 71\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 295\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 408\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 7\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 20\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 56\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 64\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 12\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 67\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 409\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 12\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 43\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 21\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 16\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 54\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 410\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 19\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 94\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 12\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 14\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 100\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 411\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 14\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 109\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 8\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 59\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 2\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 14\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 412\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 26\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 34\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 6\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 52\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 6\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 54\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 413\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 28\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 31\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 65\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 29\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 70\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 414\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 10\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 6\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 13\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 415\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 8\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 26\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 44\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 107\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 174\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 416\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 5\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 40\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 54\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 417\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 7\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 46\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 83\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 418\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 3\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 13\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 20\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 16\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 47\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 419\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 3\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 6\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 32\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 26\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 99\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 420\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 23\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 10\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 2\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 2\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 10\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 26\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 421\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 21\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 4\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 422\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 47\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 87\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 40\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 79\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 423\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 34\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 21\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 32\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 17\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 30\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 424\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 16\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 17\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 71\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 23\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 64\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 26\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 97\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 425\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 13\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 33\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 0\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 39\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 120\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 426\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 21\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 3\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 7\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 50\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 1\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "[1] 427\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,628\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 5\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 9,544\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 3\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 15,320\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 70\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 6\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 10\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 11\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "# of selected samples: 66,458\r\n",
      "# of selected variants: 48\r\n",
      "# of selected samples: 245,394\r\n",
      "# of selected variants: 1,069,452\r\n",
      "           used  (Mb) gc trigger  (Mb) max used  (Mb)\r\n",
      "Ncells  9947233 531.3   16402044 876.0 15740782 840.7\r\n",
      "Vcells 20109995 153.5   85756017 654.3 95118218 725.7\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/noncoding---williamsjacr--240802-141312-91.6-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "code_folding": [
     33,
     785,
     975,
     1168,
     1644,
     1834,
     2022,
     2229,
     2439,
     3408
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "trait <- commandArgs(TRUE)[2]\n",
    "print(trait)\n",
    "\n",
    "Train_Null_Model <- commandArgs(TRUE)[3]\n",
    "print(Train_Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[4]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[5]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "\n",
    "coding <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                   rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                   QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                   Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                   Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                   SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  rm(position)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ################################################\n",
    "  #           Coding\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.coding <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.EXONIC.Category==\"nonsynonymous SNV\")|(GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.coding]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ################################################\n",
    "  #                  plof_ds\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof_ds <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof_ds <- rbind(results_plof_ds,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof_ds))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[(dim(results_plof_ds)[2]-1):dim(results_plof_ds)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[dim(results_plof_ds)[2]] <- c(\"STAAR-B\")\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #####################################################\n",
    "  #                      plof\n",
    "  #####################################################\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof <- rbind(results_plof,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[(dim(results_plof)[2]-1):dim(results_plof)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[dim(results_plof)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #############################################\n",
    "  #             synonymous\n",
    "  #############################################\n",
    "  lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.synonymous]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.synonymous,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_synonymous <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"synonymous\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_synonymous <- rbind(results_synonymous,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_synonymous))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[(dim(results_synonymous)[2]-1):dim(results_synonymous)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[dim(results_synonymous)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #        missense\n",
    "  #################################################\n",
    "  lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.missense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.missense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #         disruptive missense\n",
    "  #################################################\n",
    "  lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.dmissense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.dmissense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results <- cbind(results,matrix(1,1,6))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }else{\n",
    "          results <- cbind(results,matrix(1,1,2))\n",
    "          colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }\n",
    "      }else\n",
    "      {\n",
    "        results_missense <- c()\n",
    "        results_ds <- results\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,6))\n",
    "          names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT(1,25)\",\"SKAT(1,1)\",\"Burden(1,25)\",\"Burden(1,1)\",\"ACAT-V(1,25)\",\"ACAT-V(1,1)\")]\n",
    "          apc_num <- (length(results_m)-19)/6\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "          results_m[\"STAAR-O\"] <- CCT(as.numeric(results_m[6:length(results_m)][p_seq]))\n",
    "          results_m[\"STAAR-S(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "          results_m[\"STAAR-S(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "          results_m[\"STAAR-B(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "          results_m[\"STAAR-B(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "          results_m[\"STAAR-A(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "          results_m[\"STAAR-A(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }else\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,2))\n",
    "          names(results_m)[(length(results_m)-1):length(results_m)] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-1):length(results_m)] <- results[2,c(\"Burden(1,25)\",\"Burden(1,1)\")]\n",
    "          \n",
    "          ## check whether the p-values is NA. If so, set NA equals 1.\n",
    "          if(is.na(results_m[(length(results_m)-1)]))\n",
    "          {\n",
    "            results_m[(length(results_m)-1)] <- 1\n",
    "          }\n",
    "          \n",
    "          if(is.na(results_m[length(results_m)]))\n",
    "          {\n",
    "            results_m[length(results_m)] <- 1\n",
    "          }\n",
    "          \n",
    "          apc_num <- (length(results_m)-10)/2\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),(length(results_m)-6):(length(results_m)-5))\n",
    "          \n",
    "          ## calculate STAAR-B\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][p_seq])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,25)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num,(length(results_m)-6))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,1)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),(length(results_m)-5))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "    results_ds <- c()\n",
    "  }\n",
    "  \n",
    "  results_coding <- list(plof=results_plof,plof_ds=results_plof_ds,missense=results,disruptive_missense=results_ds,synonymous=results_synonymous)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results_coding)\n",
    "}\n",
    "\n",
    "plof <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                 rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                 QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                 Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                 SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "plof_ds <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                    rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                    QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                    Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                    Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                    SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof_ds\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else\n",
    "    {\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "missense <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                     rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                     QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                     Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                     Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                     SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  ############################################################\n",
    "  #                      missense\n",
    "  ############################################################\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.missense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  ############################################################\n",
    "  #                      disruptive missense\n",
    "  ############################################################\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.dmissense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results <- cbind(results,matrix(1,1,6))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "        }else{\n",
    "          results <- cbind(results,matrix(1,1,2))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "        }\n",
    "      }else\n",
    "      {\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,6))\n",
    "          names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT(1,25)\",\"SKAT(1,1)\",\"Burden(1,25)\",\"Burden(1,1)\",\"ACAT-V(1,25)\",\"ACAT-V(1,1)\")]\n",
    "          apc_num <- (length(results_m)-19)/6\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "          results_m[\"STAAR-O\"] <- CCT(as.numeric(results_m[6:length(results_m)][p_seq]))\n",
    "          results_m[\"STAAR-S(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "          results_m[\"STAAR-S(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "          results_m[\"STAAR-B(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "          results_m[\"STAAR-B(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "          results_m[\"STAAR-A(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "          results_m[\"STAAR-A(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }else\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,2))\n",
    "          names(results_m)[(length(results_m)-1):length(results_m)] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-1):length(results_m)] <- results[2,c(\"Burden(1,25)\",\"Burden(1,1)\")]\n",
    "          \n",
    "          ## check whether the p-values is NA. If so, set NA equals 1.\n",
    "          if(is.na(results_m[(length(results_m)-1)]))\n",
    "          {\n",
    "            results_m[(length(results_m)-1)] <- 1\n",
    "          }\n",
    "          \n",
    "          if(is.na(results_m[length(results_m)]))\n",
    "          {\n",
    "            results_m[length(results_m)] <- 1\n",
    "          }\n",
    "          \n",
    "          apc_num <- (length(results_m)-10)/2\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),(length(results_m)-6):(length(results_m)-5))\n",
    "          \n",
    "          ## calculate STAAR-B\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][p_seq])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,25)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num,(length(results_m)-6))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,1)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),(length(results_m)-5))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "          \n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "disruptive_missense <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                                rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## disruptive_missense\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.ds <- ((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene <- variant.id.gene[lof.in.ds]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "synonymous <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                       rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                       QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                       Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                       Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                       SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ### synonymous\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.synonymous]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"synonymous\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "ptv <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel\n",
    "  }\n",
    "  \n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "ptv_ds <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                   rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                   QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                   Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                   Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                   SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof_ds\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else\n",
    "    {\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "coding_incl_ptv <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                            rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                            QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                            Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                            Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                            SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  rm(position)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ################################################\n",
    "  #           Coding\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.coding <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.EXONIC.Category==\"nonsynonymous SNV\")|(GENCODE.EXONIC.Category==\"synonymous SNV\")|(GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.coding]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ################################################\n",
    "  #                  plof_ds\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof_ds <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof_ds <- rbind(results_plof_ds,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof_ds))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[(dim(results_plof_ds)[2]-1):dim(results_plof_ds)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[dim(results_plof_ds)[2]] <- c(\"STAAR-B\")\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ################################################\n",
    "  #                  ptv_ds\n",
    "  ################################################\n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_ptv_ds <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_ptv_ds <- rbind(results_ptv_ds,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_ptv_ds))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_ptv_ds) <- colnames(results_ptv_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv_ds)[(dim(results_ptv_ds)[2]-1):dim(results_ptv_ds)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_ptv_ds) <- colnames(results_ptv_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv_ds)[dim(results_ptv_ds)[2]] <- c(\"STAAR-B\")\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #####################################################\n",
    "  #                      plof\n",
    "  #####################################################\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof <- rbind(results_plof,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[(dim(results_plof)[2]-1):dim(results_plof)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[dim(results_plof)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #####################################################\n",
    "  #                      ptv\n",
    "  #####################################################\n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel\n",
    "  }\n",
    "  \n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_ptv <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_ptv <- rbind(results_ptv,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_ptv))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_ptv) <- colnames(results_ptv, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv)[(dim(results_ptv)[2]-1):dim(results_ptv)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_ptv) <- colnames(results_ptv, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv)[dim(results_ptv)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #############################################\n",
    "  #             synonymous\n",
    "  #############################################\n",
    "  lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.synonymous]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.synonymous,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_synonymous <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"synonymous\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_synonymous <- rbind(results_synonymous,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_synonymous))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[(dim(results_synonymous)[2]-1):dim(results_synonymous)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[dim(results_synonymous)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #        missense\n",
    "  #################################################\n",
    "  lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.missense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.missense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #         disruptive missense\n",
    "  #################################################\n",
    "  lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.dmissense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.dmissense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results <- cbind(results,matrix(1,1,6))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }else{\n",
    "          results <- cbind(results,matrix(1,1,2))\n",
    "          colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }\n",
    "      }else\n",
    "      {\n",
    "        results_missense <- c()\n",
    "        results_ds <- results\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,6))\n",
    "          names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT(1,25)\",\"SKAT(1,1)\",\"Burden(1,25)\",\"Burden(1,1)\",\"ACAT-V(1,25)\",\"ACAT-V(1,1)\")]\n",
    "          apc_num <- (length(results_m)-19)/6\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "          results_m[\"STAAR-O\"] <- CCT(as.numeric(results_m[6:length(results_m)][p_seq]))\n",
    "          results_m[\"STAAR-S(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "          results_m[\"STAAR-S(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "          results_m[\"STAAR-B(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "          results_m[\"STAAR-B(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "          results_m[\"STAAR-A(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "          results_m[\"STAAR-A(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }else\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,2))\n",
    "          names(results_m)[(length(results_m)-1):length(results_m)] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-1):length(results_m)] <- results[2,c(\"Burden(1,25)\",\"Burden(1,1)\")]\n",
    "          \n",
    "          ## check whether the p-values is NA. If so, set NA equals 1.\n",
    "          if(is.na(results_m[(length(results_m)-1)]))\n",
    "          {\n",
    "            results_m[(length(results_m)-1)] <- 1\n",
    "          }\n",
    "          \n",
    "          if(is.na(results_m[length(results_m)]))\n",
    "          {\n",
    "            results_m[length(results_m)] <- 1\n",
    "          }\n",
    "          \n",
    "          apc_num <- (length(results_m)-10)/2\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),(length(results_m)-6):(length(results_m)-5))\n",
    "          \n",
    "          ## calculate STAAR-B\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][p_seq])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,25)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num,(length(results_m)-6))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,1)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),(length(results_m)-5))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "    results_ds <- c()\n",
    "  }\n",
    "  \n",
    "  results_coding <- list(plof=results_plof,plof_ds=results_plof_ds,missense=results,disruptive_missense=results_ds,synonymous=results_synonymous,ptv=results_ptv,ptv_ds=results_ptv_ds)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results_coding)\n",
    "}\n",
    "\n",
    "Gene_Centric_Coding <- function(chr,gene_name,category=c(\"all_categories\",\"plof\",\"plof_ds\",\"missense\",\"disruptive_missense\",\"synonymous\",\"ptv\",\"ptv_ds\",\"all_categories_incl_ptv\"),\n",
    "                                genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                SPA_p_filter=TRUE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  genes <- genes_info[genes_info[,2]==chr,]\n",
    "  \n",
    "  if(category==\"all_categories\")\n",
    "  {\n",
    "    results <- coding(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                      rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                      QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                      Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                      Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                      SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"plof\")\n",
    "  {\n",
    "    results <- plof(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                    rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                    QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                    Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                    Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                    SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"plof_ds\")\n",
    "  {\n",
    "    results <- plof_ds(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                       rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                       QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                       Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                       Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                       SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"missense\")\n",
    "  {\n",
    "    results <- missense(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                        rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                        Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                        SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"disruptive_missense\")\n",
    "  {\n",
    "    results <- disruptive_missense(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                                   rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                   QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                   Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                   Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                   SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"synonymous\")\n",
    "  {\n",
    "    results <- synonymous(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                          rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                          QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                          Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                          Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                          SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(category==\"ptv\")\n",
    "  {\n",
    "    results <- ptv(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                   rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                   QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                   Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                   Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                   SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(category==\"ptv_ds\")\n",
    "  {\n",
    "    results <- ptv_ds(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                      rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                      QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                      Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                      Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                      SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(category==\"all_categories_incl_ptv\")\n",
    "  {\n",
    "    results <- coding_incl_ptv(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                               rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                               QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                               Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                               Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                               SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "obj_nullmodel <- get(load(Train_Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output file name\n",
    "output_file_name <- paste0(OUTPUT_PATH,\"/\",trait,\"_AoU_WGS_Coding_Train\")\n",
    "\n",
    "## Chr\n",
    "gene_num_in_array <- 50 \n",
    "group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "sum(group.num.allchr)\n",
    "\n",
    "chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if (chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "coding_longmasks <- c(\"TTN\",\"PCDHA2\",\"PCDHA3\",\"PCDHGA1\",\"PCDHGA2\",\"PCDHGA3\",\"PCDHGB1\",\"PCDHGA4\",\"PCDHGB2\",\"PCDHGB3\")\n",
    "genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "genes_info_chr <- genes_info_chr[!genes_info_chr[,1] %in% coding_longmasks,]\n",
    "sub_seq_num <- dim(genes_info_chr)[1]\n",
    "\n",
    "if(groupid < group.num){ \n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "agds.path <- GDS_File\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "gene_centric_coding_dnanexus <- function(genes_info_chr,gene_name,chr,genofile,obj_nullmodel,rare_maf_cutoff,\n",
    "                                         QC_label,variant_type,geno_missing_imputation,\n",
    "                                         Annotation_dir,Annotation_name_catalog,\n",
    "                                         Use_annotation_weights,Annotation_name,silent){\n",
    "  results <- try(Gene_Centric_Coding(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                     QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                     Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                     Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=silent),silent=TRUE)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "results_coding <- list()\n",
    "count <- 1\n",
    "for(kk in sub_seq_id){\n",
    "  print(kk)\n",
    "  gene_name <- genes_info_chr[kk,1]\n",
    "  results <- gene_centric_coding_dnanexus(genes_info_chr=genes_info_chr,gene_name = gene_name,chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=0.01,\n",
    "                                          QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                          Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                          Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=TRUE)\n",
    "  \n",
    "  results_coding[[count]] <- results\n",
    "  count <- count + 1\n",
    "}\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "rm(list=setdiff(ls(), c(\"results_coding\", \"output_file_name\",\"arrayid\",\"agds.path\",\"trait\"))); gc()\n",
    "save(results_coding, file = paste0(output_file_name,\"_\",arrayid,\".Rdata\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${trait} ${Train_Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"Height\")){\n",
    "    for(arrayid in c(86,112,113,332)){\n",
    "        if(arrayid < 41){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 66){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 87){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 102){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 120){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 159){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 173){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 189){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 204){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 230){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 250){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 257){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 269){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 281){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 298){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 321){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 327){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 355){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 366){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 371){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--env trait'=trait,\n",
    "            '--input Train_Null_Model'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/exome.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Coding_Analysis.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "}  \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Coding_Analysis.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][129.1 KiB/129.1 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/129.1 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Coding_Analysis.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: coding-ana--williamsjacr--240805-152344-37\n",
      "  job-name: coding-analysis\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17061318534041850195\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/840692259032739568\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9318378643191275524\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15924022795533132610\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3860031643969366395\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18348738718336721892\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10678717588606309768\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9692262135783927507\n",
      "Launched job-id: coding-ana--williamsjacr--240805-152344-37\n",
      "8 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-152344-37' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-152344-37' --users 'williamsjacr'\n",
      "Waiting for job to complete...\n",
      "Monitoring for failed tasks to retry...\n",
      "*** This dsub process must continue running to retry failed tasks.\n",
      "  coding-ana--williamsjacr--240805-152344-37.6 (attempt 1) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2964234674434847443\n",
      "  coding-ana--williamsjacr--240805-152344-37.3 (attempt 1) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11529758471240069288\n",
      "  coding-ana--williamsjacr--240805-152344-37.2 (attempt 1) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8349154917546125030\n",
      "  coding-ana--williamsjacr--240805-152344-37.3 (attempt 2) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7701904607760559155\n",
      "  coding-ana--williamsjacr--240805-152344-37.3 (attempt 3) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7912129358242913412\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Coding_Analysis.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status                 Last Update\r\n",
      "---------------  ------  ---------------------  --------------\r\n",
      "coding-analysis       3  Success                08-05 18:23:33\r\n",
      "coding-analysis       3  worker was terminated  08-05 17:02:36\r\n",
      "coding-analysis       2  Success                08-05 17:36:46\r\n",
      "coding-analysis       3  worker was terminated  08-05 16:37:48\r\n",
      "coding-analysis       4  Success                08-05 16:21:02\r\n",
      "coding-analysis       6  Success                08-05 16:44:42\r\n",
      "coding-analysis       8  Success                08-05 16:04:54\r\n",
      "coding-analysis       7  Success                08-05 16:47:22\r\n",
      "coding-analysis       6  worker was terminated  08-05 15:39:07\r\n",
      "coding-analysis       5  Success                08-05 15:52:56\r\n",
      "coding-analysis       4  worker was terminated  08-05 15:41:16\r\n",
      "coding-analysis       3  worker was terminated  08-05 16:08:13\r\n",
      "coding-analysis       2  worker was terminated  08-05 16:28:00\r\n",
      "coding-analysis       1  Success                08-05 15:53:40\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-152344-37' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis_Summary.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis_Summary.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "library(data.table)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[1]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "gene_centric_coding_jobs_num <- 379\n",
    "\n",
    "trait <- \"BMI\"\n",
    "i <- 1\n",
    "\n",
    "for(trait in c(\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\")){\n",
    "  gene_centric_results_name <- paste0(trait,\"_AoU_WGS_Coding_Train\")\n",
    "  \n",
    "  results_coding_genome <- NULL\n",
    "  for (i in 1:gene_centric_coding_jobs_num){\n",
    "    results <- get(load(paste0(INPUT_PATH,\"/\",gene_centric_results_name,\"_\",i,\".Rdata\")))\n",
    "    results_coding_genome <- c(results_coding_genome, results)\n",
    "  }\n",
    "  \n",
    "  combine_jake <- function(x){\n",
    "    a <- data.frame(Gene = c(unlist(x$plof[,c(\"Gene name\")]),unlist(x$plof_ds[,c(\"Gene name\")]),unlist(x$missense[,c(\"Gene name\")]),unlist(x$disruptive_missense[,c(\"Gene name\")]),unlist(x$synonymous[,c(\"Gene name\")])),\n",
    "                    Chr = c(unlist(x$plof[,c(\"Chr\")]),unlist(x$plof_ds[,c(\"Chr\")]),unlist(x$missense[,c(\"Chr\")]),unlist(x$disruptive_missense[,c(\"Chr\")]),unlist(x$synonymous[,c(\"Chr\")])),\n",
    "                    Category = c(unlist(x$plof[,c(\"Category\")]),unlist(x$plof_ds[,c(\"Category\")]),unlist(x$missense[,c(\"Category\")]),unlist(x$disruptive_missense[,c(\"Category\")]),unlist(x$synonymous[,c(\"Category\")])),\n",
    "                    Number_SNV = c(unlist(x$plof[,c(\"#SNV\")]),unlist(x$plof_ds[,c(\"#SNV\")]),unlist(x$missense[,c(\"#SNV\")]),unlist(x$disruptive_missense[,c(\"#SNV\")]),unlist(x$synonymous[,c(\"#SNV\")])),\n",
    "                    Burden_1_1 = c(unlist(x$plof[,c(\"Burden(1,1)\")]),unlist(x$plof_ds[,c(\"Burden(1,1)\")]),unlist(x$missense[,c(\"Burden(1,1)\")]),unlist(x$disruptive_missense[,c(\"Burden(1,1)\")]),unlist(x$synonymous[,c(\"Burden(1,1)\")])),\n",
    "                    STAARB = c(unlist(x$plof[,c(\"STAAR-B(1,1)\")]),unlist(x$plof_ds[,c(\"STAAR-B(1,1)\")]),unlist(x$missense[,c(\"STAAR-B(1,1)\")]),unlist(x$disruptive_missense[,c(\"STAAR-B(1,1)\")]),unlist(x$synonymous[,c(\"STAAR-B(1,1)\")])))\n",
    "    return(a)\n",
    "  }\n",
    "  \n",
    "  results_coding_genome <- lapply(results_coding_genome, combine_jake)\n",
    "  results_coding_genome <- rbindlist(results_coding_genome)\n",
    "  results_coding_genome$Number_SNV <- as.numeric(results_coding_genome$Number_SNV)\n",
    "  results_coding_genome <- results_coding_genome[results_coding_genome$Number_SNV < 2000,]\n",
    "  \n",
    "  write.csv(results_coding_genome,file = paste0(OUTPUT_PATH,\"/\",trait,\"_coding_sig.csv\"),row.names = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis_Summary.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis_Summary.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "            '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Coding_Analysis_Summary.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\",\n",
    "            check.names = FALSE\n",
    "        ))\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Coding_Analysis_Summary.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.3 KiB/  2.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Coding_Analysis_Summary.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: coding-ana--williamsjacr--240805-203638-20\n",
      "  job-name: coding-analysis-summary\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18328105492588917735\n",
      "Launched job-id: coding-ana--williamsjacr--240805-203638-20\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-203638-20' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-203638-20' --users 'williamsjacr'\n",
      "Waiting for job to complete...\n",
      "Monitoring for failed tasks to retry...\n",
      "*** This dsub process must continue running to retry failed tasks.\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Coding_Analysis_Summary.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-194340-74' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "code_folding": [
     19
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Single_RareVariant_PRS.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Single_RareVariant_PRS.R\n",
    "rm(list = ls())\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "Gene_Centric_Coding_G_Star <- function(chr,gene_name,category=c(\"plof\",\"plof_ds\",\"missense\",\"disruptive_missense\",\"synonymous\"),\n",
    "                                       genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                       QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                       Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,silent=FALSE){\n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  genes <- genes_info[genes_info[,2]==chr,]  \n",
    "\n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category  <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")  \n",
    "  \n",
    "  if(category == \"plof\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"plof_ds\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"missense\"){\n",
    "    lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.missense]\n",
    "  }else if(category == \"disruptive_missense\"){\n",
    "    lof.in.ds <- ((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.ds]\n",
    "  }else{\n",
    "    lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.synonymous]\n",
    "  }\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  genotype <- Geno\n",
    "  \n",
    "  if(dim(genotype)[2] == 1){\n",
    "    return(matrix(0,nrow = dim(genotype)[1],ncol = 1))\n",
    "  }\n",
    "  \n",
    "  if(!is.null(attr(class(genotype), \"package\")) && attr(class(genotype), \"package\") == \"Matrix\"){\n",
    "    genotype <- as.matrix(genotype)\n",
    "  }\n",
    "  genotype <- matrix_flip(genotype)\n",
    "  MAF <- genotype$MAF\n",
    "  RV_label <- as.vector((MAF<rare_maf_cutoff)&(MAF>0))\n",
    "  Geno_rare <- genotype$Geno[,RV_label]\n",
    "  G <- Geno_rare\n",
    "  rm(Geno_rare)\n",
    "  gc()\n",
    "  \n",
    "  if(is.null(dim(G))){\n",
    "    G <- matrix(G,ncol = 1)\n",
    "  }\n",
    "  \n",
    "  C <- G%*%matrix(1,nrow=ncol(G),ncol = 1)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(C)\n",
    "}\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Train_Pvals <- commandArgs(TRUE)[2]\n",
    "print(Train_Pvals)\n",
    "\n",
    "Train_NullModel <- commandArgs(TRUE)[3]\n",
    "print(Train_NullModel)\n",
    "\n",
    "Tune_NullModel <- commandArgs(TRUE)[4]\n",
    "print(Tune_NullModel)\n",
    "\n",
    "Validation_NullModel <- commandArgs(TRUE)[5]\n",
    "print(Validation_NullModel)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[6]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[8]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[9]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[10]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[11]\n",
    "print(all_validation_file)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[12]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Train_PVals_All <- read.csv(Train_Pvals)\n",
    "Train_PVals_All <- Train_PVals_All[Train_PVals_All$STAARB <= 1e-03,]\n",
    "\n",
    "## Null Model\n",
    "obj_nullmodel_train <- get(load(Train_NullModel))\n",
    "obj_nullmodel_tune <- get(load(Tune_NullModel))\n",
    "obj_nullmodel_validation <- get(load(Validation_NullModel))\n",
    "\n",
    "obj_nullmodel <- obj_nullmodel_train\n",
    "obj_nullmodel$id_include <- c(obj_nullmodel_train$id_include,obj_nullmodel_tune$id_include,obj_nullmodel_validation$id_include)\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "\n",
    "G_star_gene_centric_coding <- list()\n",
    "\n",
    "for(i in 1:nrow(Train_PVals_All)){\n",
    "  ## Chr\n",
    "  chr <- Train_PVals_All$Chr[i]\n",
    "  ## Gene name\n",
    "  gene_name <- Train_PVals_All$Gene[i]\n",
    "  ## Coding mask\n",
    "  category <- Train_PVals_All$Category[i]\n",
    "  \n",
    "  ### gds file\n",
    "  gds.path <- paste0(INPUT_PATH,\"/exome.chr\",chr,\".gds\")\n",
    "  genofile <- seqOpen(gds.path)\n",
    "  \n",
    "  G_star_gene_centric_coding[[i]] <- Gene_Centric_Coding_G_Star(chr=chr,gene_name=gene_name,category=category ,\n",
    "                                                                genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                                                QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,silent=FALSE) \n",
    "  seqClose(genofile) \n",
    "} \n",
    "\n",
    "G_star_gene_centric_coding <- do.call(cbind,G_star_gene_centric_coding)\n",
    "\n",
    "col_remove <- apply(G_star_gene_centric_coding,2,function(x){sum(x != 0)}) > 10 & colSums(G_star_gene_centric_coding) > 10 \n",
    "G_star_gene_centric_coding <- G_star_gene_centric_coding[,col_remove,drop = FALSE]\n",
    "\n",
    "Train_PVals_All <- Train_PVals_All[col_remove,]\n",
    "\n",
    "print(str(G_star_gene_centric_coding))\n",
    "\n",
    "ids_gstar <- obj_nullmodel$id_include\n",
    "\n",
    "G_star_gene_centric_coding_train <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_train$id_include,]\n",
    "G_star_gene_centric_coding_tune <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_tune$id_include,]\n",
    "G_star_gene_centric_coding_vad <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_validation$id_include,]\n",
    "\n",
    "rm(G_star_gene_centric_coding)\n",
    "\n",
    "X_train <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_train$id_include],G_star_gene_centric_coding_train)\n",
    "pheno_train <- read.delim(all_train_file)\n",
    "pheno_train <- inner_join(pheno_train,X_train)\n",
    "print(str(pheno_train))\n",
    "X_train <- as.matrix(pheno_train[,22:ncol(pheno_train),drop = FALSE])\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_train)\n",
    "y_train <- model.null$residual\n",
    "\n",
    "X_tune <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_tune$id_include],G_star_gene_centric_coding_tune)\n",
    "pheno_tune <- read.delim(all_tune_file)\n",
    "pheno_tune <- inner_join(pheno_tune,X_tune)\n",
    "X_tune <- as.matrix(pheno_tune[,22:ncol(pheno_tune),drop = FALSE])\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tune)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "X_valid <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_validation$id_include],G_star_gene_centric_coding_vad)\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "pheno_valid <- inner_join(pheno_valid,X_valid)\n",
    "X_valid <- as.matrix(pheno_valid[,22:ncol(pheno_valid),drop = FALSE])\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_valid <- model.null$residual\n",
    "\n",
    "lasso_train <- glmnet(X_train,y_train,family = \"gaussian\",alpha = 1)\n",
    "ridge_train <- glmnet(X_train,y_train,family = \"gaussian\",alpha = 0)\n",
    "lm_train <- lm.fit(cbind(1,X_train),y_train)\n",
    "lm_train$coefficients[is.na(lm_train$coefficients)] <- 0\n",
    "\n",
    "beta_matrix <- cbind(as.matrix(lasso_train$beta),as.matrix(ridge_train$beta),lm_train$coefficients[-1])\n",
    "beta_matrix <- as.data.frame(beta_matrix)\n",
    "colnames(beta_matrix) <- c(paste0(\"lasso_prs\",1:ncol(as.matrix(lasso_train$beta))),paste0(\"ridge_prs\",1:ncol(as.matrix(ridge_train$beta))),paste0(\"lm_prs\",1))\n",
    "\n",
    "beta_matrix <- cbind(Train_PVals_All[,c(1:4)],beta_matrix)\n",
    "\n",
    "\n",
    "lasso_prs_tune <- predict(lasso_train,X_tune)\n",
    "ridge_prs_tune <- predict(ridge_train,X_tune)\n",
    "lm_prs_tune <- as.numeric(cbind(1,X_tune)%*%matrix(lm_train$coefficients,ncol = 1))\n",
    "\n",
    "lasso_prs_vad <- predict(lasso_train,X_valid)\n",
    "ridge_prs_vad <- predict(ridge_train,X_valid)\n",
    "lm_prs_vad <- as.numeric(cbind(1,X_valid)%*%matrix(lm_train$coefficients,ncol = 1))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "lasso_tune_dat <- data.frame(y = y_tune,lasso_prs_tune)\n",
    "colnames(lasso_tune_dat) <- c(\"y\",paste0(\"lasso_prs\",1:(ncol(lasso_tune_dat) - 1)))\n",
    "lasso_valid_dat <- data.frame(y = y_valid,lasso_prs_vad)\n",
    "colnames(lasso_valid_dat) <- c(\"y\",paste0(\"lasso_prs\",1:(ncol(lasso_valid_dat) - 1)))\n",
    "\n",
    "\n",
    "ridge_tune_dat <- data.frame(y = y_tune,ridge_prs_tune)\n",
    "colnames(ridge_tune_dat) <- c(\"y\",paste0(\"ridge_prs\",1:(ncol(ridge_tune_dat) - 1)))\n",
    "ridge_valid_dat <- data.frame(y = y_valid,ridge_prs_vad)\n",
    "colnames(ridge_valid_dat) <- c(\"y\",paste0(\"ridge_prs\",1:(ncol(ridge_valid_dat) - 1)))\n",
    "\n",
    "lm_tune_dat <- data.frame(y = y_tune,lm_prs_tune)\n",
    "colnames(lm_tune_dat) <- c(\"y\",paste0(\"lm_prs\",1:(ncol(lm_tune_dat) - 1)))\n",
    "lm_valid_dat <- data.frame(y = y_valid,lm_prs_vad)\n",
    "colnames(lm_valid_dat) <- c(\"y\",paste0(\"lm_prs\",1:(ncol(lm_valid_dat) - 1)))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "all_prs_tune <- cbind(lasso_prs_tune,ridge_prs_tune,lm_prs_tune)\n",
    "colnames(all_prs_tune) <- c(paste0(\"lasso_prs\",1:ncol(lasso_prs_tune)),paste0(\"ridge_prs\",1:ncol(ridge_prs_tune)),\"lm_prs1\")\n",
    "all_prs_valid <- cbind(lasso_prs_vad,ridge_prs_vad,lm_prs_vad)\n",
    "colnames(all_prs_valid) <- c(paste0(\"lasso_prs\",1:ncol(lasso_prs_vad)),paste0(\"ridge_prs\",1:ncol(ridge_prs_vad)),\"lm_prs1\")\n",
    "\n",
    "all_tune <- data.frame(y = y_tune,all_prs_tune)\n",
    "all_valid <- data.frame(y = y_valid,all_prs_valid)\n",
    "\n",
    "best_r2 <- vector()\n",
    "count <- 1\n",
    "for(i in colnames(all_prs_tune)){\n",
    "  tmp <- data.frame(y = all_tune$y,x = all_prs_tune[,i])\n",
    "  best_r2[count] <- summary(lm(y~x,data = tmp))$r.squared\n",
    "  count <- count + 1\n",
    "}\n",
    "\n",
    "best_thresh <- colnames(all_prs_tune)[which.max(best_r2)]\n",
    "r2_bestoverall_tune <- best_r2[which.max(best_r2)]\n",
    "\n",
    "\n",
    "all_prs_tune <- as.data.frame(all_prs_tune)\n",
    "all_prs_valid <- as.data.frame(all_prs_valid)\n",
    "\n",
    "mtx <- cor(all_prs_tune)\n",
    "drop <- names(all_prs_tune)[apply(mtx,2,function(x){sum(is.na(x))}) == (nrow(mtx) - 1)]\n",
    "\n",
    "all_prs_tune <- dplyr::select(all_prs_tune, -c(drop))\n",
    "all_prs_valid <- dplyr::select(all_prs_valid, -c(drop))\n",
    "\n",
    "mtx <- cor(all_prs_tune)\n",
    "drop <- findCorrelation(mtx,cutoff=0.98)\n",
    "drop <- names(all_prs_tune)[drop]\n",
    "\n",
    "all_prs_tune <- dplyr::select(all_prs_tune, -c(drop))\n",
    "all_prs_valid <- dplyr::select(all_prs_valid, -c(drop))\n",
    "\n",
    "drop <- findLinearCombos(all_prs_tune)$remove\n",
    "drop <- names(data.frame(all_prs_tune))[drop]\n",
    "\n",
    "all_prs_tune <- dplyr::select(all_prs_tune, -c(drop))\n",
    "all_prs_valid <- dplyr::select(all_prs_valid, -c(drop))\n",
    "\n",
    "\n",
    "\n",
    "SL.library <- c(\n",
    "  \"SL.glmnet\",\n",
    "  \"SL.ridge\",\n",
    "  \"SL.glm\",\n",
    "  \"SL.mean\"\n",
    ")\n",
    "\n",
    "full_superlearner <- SuperLearner(Y = y_tune, X = all_prs_tune, family = gaussian(),\n",
    "                                  # For a real analysis we would use V = 10.\n",
    "                                  # V = 3,\n",
    "                                  SL.library = SL.library,cvControl = list(V = 20))\n",
    "cvsl <- CV.SuperLearner(Y = y_tune, X = all_prs_tune, family = gaussian(),\n",
    "                        # For a real analysis we would use V = 10.\n",
    "                        # V = 3,\n",
    "                        SL.library = SL.library,cvControl = list(V = 20))\n",
    "\n",
    "best_algorithm <- summary(cvsl)$Table$Algorithm[which.min(summary(cvsl)$Table$Ave)]\n",
    "\n",
    "a_tune <- predict(full_superlearner, all_prs_tune, onlySL = FALSE)\n",
    "a_vad <- predict(full_superlearner, all_prs_valid, onlySL = FALSE)\n",
    "\n",
    "prs_best_tune_sl <- a_tune$pred\n",
    "prs_best_tune_glmnet <- a_tune$library.predict[,1]\n",
    "prs_best_tune_ridge <- a_tune$library.predict[,2]\n",
    "prs_best_tune_glm <- a_tune$library.predict[,3]\n",
    "prs_best_tune_mean <- a_tune$library.predict[,4]\n",
    "\n",
    "prs_best_vad_sl <- a_vad$pred\n",
    "prs_best_vad_glmnet <- a_vad$library.predict[,1]\n",
    "prs_best_vad_ridge <- a_vad$library.predict[,2]\n",
    "prs_best_vad_glm <- a_vad$library.predict[,3]\n",
    "prs_best_vad_mean <- a_vad$library.predict[,4]\n",
    "\n",
    "if(best_algorithm == \"SL.glmnet_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_glmnet\n",
    "  prs_best_vad <- prs_best_vad_glmnet\n",
    "}else if(best_algorithm == \"SL.ridge_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_ridge\n",
    "  prs_best_vad <- prs_best_vad_ridge\n",
    "}else if(best_algorithm == \"SL.glm_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_glm\n",
    "  prs_best_vad <- prs_best_vad_glm\n",
    "}else if(best_algorithm == \"SL.mean_All\"){\n",
    "  #final\n",
    "  prs_best_tune <- prs_best_tune_mean\n",
    "  prs_best_vad <- prs_best_vad_mean\n",
    "} else {\n",
    "  prs_best_tune <- prs_best_tune_sl\n",
    "  prs_best_vad <- prs_best_vad_sl\n",
    "}\n",
    "\n",
    "tune_dat_sl_R2 <- data.frame(y = y_tune,x = prs_best_tune)\n",
    "valid_dat_sl_R2 <- data.frame(y = y_valid,x = prs_best_vad)\n",
    "\n",
    "r2_sl_tune <- summary(lm(y~x,data = tune_dat_sl_R2))$r.squared\n",
    "\n",
    "if(r2_sl_tune < r2_bestoverall_tune){\n",
    "  prs_best_tune_sl <- all_tune[,best_thresh]\n",
    "  prs_best_vad_sl <- all_valid[,best_thresh] \n",
    "  \n",
    "  tune_dat_sl_R2 <- data.frame(y = y_tune,x = prs_best_tune_sl)\n",
    "  valid_dat_sl_R2 <- data.frame(y = y_valid,x = prs_best_vad_sl)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "if(r2_sl_tune < r2_bestoverall_tune){\n",
    "  beta_matrix$Beta_Final <- beta_matrix[,best_thresh]\n",
    "  write.csv(beta_matrix[,c(\"Gene\",\"Chr\",\"Category\",\"Number_SNV\",\"Beta_Final\")],file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)\n",
    "}else{\n",
    "  if(best_algorithm == \"SL.glmnet_All\"){\n",
    "    beta_full_superlearner <- data.frame(Coef = row.names(coef(full_superlearner$fitLibrary$SL.glmnet_All$object)),Beta = as.numeric(coef(full_superlearner$fitLibrary$SL.glmnet_All$object)))\n",
    "  }else if(best_algorithm == \"SL.ridge_All\"){\n",
    "    beta_full_superlearner <- data.frame(Coef = row.names(full_superlearner$fitLibrary$SL.ridge_All$bestCoef),Beta = as.numeric(full_superlearner$fitLibrary$SL.ridge_All$bestCoef[,1]))\n",
    "  }else if(best_algorithm == \"SL.glm_All\"){\n",
    "    beta_full_superlearner <- data.frame(Coef = names(coef(full_superlearner$fitLibrary$SL.glm_All$object)),Beta = as.numeric(coef(full_superlearner$fitLibrary$SL.glm_All$object)))\n",
    "  }else{\n",
    "    beta_lasso <- data.frame(Coef = row.names(coef(full_superlearner$fitLibrary$SL.glmnet_All$object)),Beta = as.numeric(coef(full_superlearner$fitLibrary$SL.glmnet_All$object)))\n",
    "    beta_ridge <- data.frame(Coef = row.names(full_superlearner$fitLibrary$SL.ridge_All$bestCoef),Beta = as.numeric(full_superlearner$fitLibrary$SL.ridge_All$bestCoef[,1]))\n",
    "    beta_glm <- data.frame(Coef = names(coef(full_superlearner$fitLibrary$SL.glm_All$object)),Beta = as.numeric(coef(full_superlearner$fitLibrary$SL.glm_All$object)))\n",
    "    beta_mean <- data.frame(Coef = names(coef(full_superlearner$fitLibrary$SL.glm_All$object)), Beta = 0)\n",
    "    beta_mean$Beta[1] <- full_superlearner$fitLibrary$SL.mean_All$object\n",
    "    \n",
    "    beta_full_superlearner <- data.frame(Coef = names(coef(full_superlearner$fitLibrary$SL.glm_All$object)), Beta = 0)\n",
    "    beta_full_superlearner$Beta <- as.numeric(full_superlearner$coef[names(full_superlearner$coef) == \"SL.glmnet_All\"]) * beta_lasso$Beta + \n",
    "      as.numeric(full_superlearner$coef[names(full_superlearner$coef) == \"SL.ridge_All\"]) * beta_ridge$Beta + \n",
    "      as.numeric(full_superlearner$coef[names(full_superlearner$coef) == \"SL.glm_All\"]) * beta_glm$Beta + \n",
    "      as.numeric(full_superlearner$coef[names(full_superlearner$coef) == \"SL.mean_All\"]) * beta_mean$Beta \n",
    "  }\n",
    "  \n",
    "  beta_full_superlearner <- beta_full_superlearner[-1,]\n",
    "  \n",
    "  beta_matrix$Beta_Final <- 0\n",
    "  for(i in 1:nrow(beta_full_superlearner)){\n",
    "    beta_matrix$Beta_Final <- beta_matrix[,beta_full_superlearner$Coef[i]]*beta_full_superlearner$Beta[i]\n",
    "  }\n",
    "  write.csv(beta_matrix[,c(\"Gene\",\"Chr\",\"Category\",\"Number_SNV\",\"Beta_Final\")],file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "print(str(all_phenotypes))\n",
    "\n",
    "RV_PRS <- data.frame(IID = pheno_valid$IID,Y = valid_dat_sl_R2[,1],RV_PRS = valid_dat_sl_R2[,2],pheno_valid[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "\n",
    "print(str(RV_PRS))\n",
    "\n",
    "write.csv(RV_PRS,file = paste0(OUTPUT_PATH,\"/\",trait,\"_BestPRS.csv\"),row.names = FALSE)\n",
    "\n",
    "RV_PRS_raw <- RV_PRS\n",
    "RV_PRS_adjusted <- RV_PRS\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = RV_PRS_adjusted[,\"RV_PRS\"],RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "print(summary(mod))\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "print(summary(mod))\n",
    "\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(y_hat < 0) > 0){\n",
    "  mod <- lm(y~1,data = tmp)\n",
    "  y_hat <- predict(mod,tmp)\n",
    "}\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  RV_PRS_adjusted[,\"RV_PRS\"] <- 0\n",
    "}else{\n",
    "  RV_PRS_adjusted[,\"RV_PRS\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "RV_PRS_raw_EUR <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "RV_PRS_raw_SAS <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "RV_PRS_raw_AMR <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "RV_PRS_raw_AFR <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "RV_PRS_raw_EAS <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "RV_PRS_raw_MID <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "str(RV_PRS_raw_EUR)\n",
    "str(RV_PRS_raw_SAS)\n",
    "str(RV_PRS_raw_AMR)\n",
    "str(RV_PRS_raw_AFR)\n",
    "str(RV_PRS_raw_EAS)\n",
    "str(RV_PRS_raw_MID)\n",
    "\n",
    "RV_PRS_adjusted_EUR <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "RV_PRS_adjusted_SAS <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "RV_PRS_adjusted_AMR <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "RV_PRS_adjusted_AFR <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "RV_PRS_adjusted_EAS <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "RV_PRS_adjusted_MID <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "str(RV_PRS_adjusted_EUR)\n",
    "str(RV_PRS_adjusted_SAS)\n",
    "str(RV_PRS_adjusted_AMR)\n",
    "str(RV_PRS_adjusted_AFR)\n",
    "str(RV_PRS_adjusted_EAS)\n",
    "str(RV_PRS_adjusted_MID)\n",
    "\n",
    "RV_PRS_raw_EUR$Y <- scale(RV_PRS_raw_EUR$Y)\n",
    "RV_PRS_raw_SAS$Y <- scale(RV_PRS_raw_SAS$Y)\n",
    "RV_PRS_raw_AMR$Y <- scale(RV_PRS_raw_AMR$Y)\n",
    "RV_PRS_raw_AFR$Y <- scale(RV_PRS_raw_AFR$Y)\n",
    "RV_PRS_raw_EAS$Y <- scale(RV_PRS_raw_EAS$Y)\n",
    "RV_PRS_raw_MID$Y <- scale(RV_PRS_raw_MID$Y)\n",
    "\n",
    "RV_PRS_adjusted_EUR$Y <- scale(RV_PRS_adjusted_EUR$Y)\n",
    "RV_PRS_adjusted_SAS$Y <- scale(RV_PRS_adjusted_SAS$Y)\n",
    "RV_PRS_adjusted_AMR$Y <- scale(RV_PRS_adjusted_AMR$Y)\n",
    "RV_PRS_adjusted_AFR$Y <- scale(RV_PRS_adjusted_AFR$Y)\n",
    "RV_PRS_adjusted_EAS$Y <- scale(RV_PRS_adjusted_EAS$Y)\n",
    "RV_PRS_adjusted_MID$Y <- scale(RV_PRS_adjusted_MID$Y)\n",
    "\n",
    "RV_PRS_raw_EUR$RV_PRS <- scale(RV_PRS_raw_EUR$RV_PRS)\n",
    "RV_PRS_raw_SAS$RV_PRS <- scale(RV_PRS_raw_SAS$RV_PRS)\n",
    "RV_PRS_raw_AMR$RV_PRS <- scale(RV_PRS_raw_AMR$RV_PRS)\n",
    "RV_PRS_raw_AFR$RV_PRS <- scale(RV_PRS_raw_AFR$RV_PRS)\n",
    "RV_PRS_raw_EAS$RV_PRS <- scale(RV_PRS_raw_EAS$RV_PRS)\n",
    "RV_PRS_raw_MID$RV_PRS <- scale(RV_PRS_raw_MID$RV_PRS)\n",
    "\n",
    "best_beta_raw_RV_EUR <- coef(lm(Y~RV_PRS,data = RV_PRS_raw_EUR))[2]\n",
    "se_beta_raw_RV_EUR <- summary(lm(Y~RV_PRS,data = RV_PRS_raw_EUR))$coefficients[2,2]\n",
    "best_beta_raw_RV_SAS <- coef(lm(Y~RV_PRS,data = RV_PRS_raw_SAS))[2]\n",
    "se_beta_raw_RV_SAS <- summary(lm(Y~RV_PRS,data = RV_PRS_raw_SAS))$coefficients[2,2]\n",
    "best_beta_raw_RV_AMR <- coef(lm(Y~RV_PRS,data = RV_PRS_raw_AMR))[2]\n",
    "se_beta_raw_RV_AMR <- summary(lm(Y~RV_PRS,data = RV_PRS_raw_AMR))$coefficients[2,2]\n",
    "best_beta_raw_RV_AFR <- coef(lm(Y~RV_PRS,data = RV_PRS_raw_AFR))[2]\n",
    "se_beta_raw_RV_AFR <- summary(lm(Y~RV_PRS,data = RV_PRS_raw_AFR))$coefficients[2,2]\n",
    "best_beta_raw_RV_EAS <- coef(lm(Y~RV_PRS,data = RV_PRS_raw_EAS))[2]\n",
    "se_beta_raw_RV_EAS <- summary(lm(Y~RV_PRS,data = RV_PRS_raw_EAS))$coefficients[2,2]\n",
    "best_beta_raw_RV_MID <- coef(lm(Y~RV_PRS,data = RV_PRS_raw_MID))[2]\n",
    "se_beta_raw_RV_MID <- summary(lm(Y~RV_PRS,data = RV_PRS_raw_MID))$coefficients[2,2]\n",
    "\n",
    "best_beta_adjusted_RV_EUR <- coef(lm(Y~RV_PRS,data = RV_PRS_adjusted_EUR))[2]\n",
    "se_beta_adjusted_RV_EUR <- summary(lm(Y~RV_PRS,data = RV_PRS_adjusted_EUR))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_SAS <- coef(lm(Y~RV_PRS,data = RV_PRS_adjusted_SAS))[2]\n",
    "se_beta_adjusted_RV_SAS <- summary(lm(Y~RV_PRS,data = RV_PRS_adjusted_SAS))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_AMR <- coef(lm(Y~RV_PRS,data = RV_PRS_adjusted_AMR))[2]\n",
    "se_beta_adjusted_RV_AMR <- summary(lm(Y~RV_PRS,data = RV_PRS_adjusted_AMR))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_AFR <- coef(lm(Y~RV_PRS,data = RV_PRS_adjusted_AFR))[2]\n",
    "se_beta_adjusted_RV_AFR <- summary(lm(Y~RV_PRS,data = RV_PRS_adjusted_AFR))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_EAS <- coef(lm(Y~RV_PRS,data = RV_PRS_adjusted_EAS))[2]\n",
    "se_beta_adjusted_RV_EAS <- summary(lm(Y~RV_PRS,data = RV_PRS_adjusted_EAS))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_MID <- coef(lm(Y~RV_PRS,data = RV_PRS_adjusted_MID))[2]\n",
    "se_beta_adjusted_RV_MID <- summary(lm(Y~RV_PRS,data = RV_PRS_adjusted_MID))$coefficients[2,2]\n",
    "\n",
    "RV_PRS_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                             beta_raw = c(best_beta_raw_RV_EUR,best_beta_raw_RV_SAS,best_beta_raw_RV_AMR,best_beta_raw_RV_AFR,best_beta_raw_RV_EAS,best_beta_raw_RV_MID), \n",
    "                             se_raw = c(se_beta_raw_RV_EUR,se_beta_raw_RV_SAS,se_beta_raw_RV_AMR,se_beta_raw_RV_AFR,se_beta_raw_RV_EAS,se_beta_raw_RV_MID), \n",
    "                             beta_adjusted = c(best_beta_adjusted_RV_EUR,best_beta_adjusted_RV_SAS,best_beta_adjusted_RV_AMR,best_beta_adjusted_RV_AFR,best_beta_adjusted_RV_EAS,best_beta_adjusted_RV_MID), \n",
    "                             se_adjusted = c(se_beta_adjusted_RV_EUR,se_beta_adjusted_RV_SAS,se_beta_adjusted_RV_AMR,se_beta_adjusted_RV_AFR,se_beta_adjusted_RV_EAS,se_beta_adjusted_RV_MID))\n",
    "\n",
    "write.csv(RV_PRS_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Single_RareVariant_PRS.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Single_RareVariant_PRS.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Train_Pvals} ${Train_NullModel} ${Tune_NullModel} ${Validation_NullModel} ${Annotation_name_catalog} ${INPUT_PATH} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "            '--env trait'=trait,\n",
    "            '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Single_RareVariant_PRS.R\",\n",
    "            '--input Train_Pvals'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding/\",trait,\"_coding_sig.csv\"),\n",
    "            '--input Train_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Tune_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Tune_Null_Model.RData\"),\n",
    "            '--input Validation_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Validation_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "            '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "            '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "            '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS\",\n",
    "            check.names = FALSE\n",
    "        ))   \n",
    "}\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Single_RareVariant_PRS.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 23.6 KiB/ 23.6 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/23.6 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Single_RareVariant_PRS.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: single-rar--williamsjacr--240807-002443-67\n",
      "  job-name: single-rarevariant-prs\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5211521007745152674\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3582161128721624158\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12226037870621958711\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3508336469533073098\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10313404690612761132\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14000302688532583800\n",
      "Launched job-id: single-rar--williamsjacr--240807-002443-67\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'single-rar--williamsjacr--240807-002443-67' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'single-rar--williamsjacr--240807-002443-67' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Single_RareVariant_PRS.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "single-rarev...       6  Pending   08-07 00:24:44\r\n",
      "single-rarev...       5  Pending   08-07 00:24:44\r\n",
      "single-rarev...       4  Pending   08-07 00:24:44\r\n",
      "single-rarev...       3  Pending   08-07 00:24:43\r\n",
      "single-rarev...       2  Pending   08-07 00:24:43\r\n",
      "single-rarev...       1  Pending   08-07 00:24:43\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'single-rar--williamsjacr--240807-002443-67' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing RareVariantCoefficients.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile RareVariantCoefficients.R\n",
    "rm(list = ls())\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "Gene_Centric_Coding_G_Star <- function(chr,gene_name,category=c(\"plof\",\"plof_ds\",\"missense\",\"disruptive_missense\",\"synonymous\"),\n",
    "                                       genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                       QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                       Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,silent=FALSE){\n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  genes <- genes_info[genes_info[,2]==chr,]  \n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category  <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")  \n",
    "  \n",
    "  if(category == \"plof\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"plof_ds\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"missense\"){\n",
    "    lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.missense]\n",
    "  }else if(category == \"disruptive_missense\"){\n",
    "    lof.in.ds <- ((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.ds]\n",
    "  }else{\n",
    "    lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.synonymous]\n",
    "  }\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  genotype <- Geno\n",
    "  \n",
    "  if(dim(genotype)[2] == 1){\n",
    "    return(matrix(0,nrow = dim(genotype)[1],ncol = 1))\n",
    "  }\n",
    "  \n",
    "  if(!is.null(attr(class(genotype), \"package\")) && attr(class(genotype), \"package\") == \"Matrix\"){\n",
    "    genotype <- as.matrix(genotype)\n",
    "  }\n",
    "  genotype <- matrix_flip(genotype)\n",
    "  MAF <- genotype$MAF\n",
    "  RV_label <- as.vector((MAF<rare_maf_cutoff)&(MAF>0))\n",
    "  Geno_rare <- genotype$Geno[,RV_label]\n",
    "  G <- Geno_rare\n",
    "  rm(Geno_rare)\n",
    "  gc()\n",
    "  \n",
    "  if(is.null(dim(G))){\n",
    "    G <- matrix(G,ncol = 1)\n",
    "  }\n",
    "  \n",
    "  C <- G%*%matrix(1,nrow=ncol(G),ncol = 1)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(C)\n",
    "}\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Train_Pvals <- commandArgs(TRUE)[2]\n",
    "print(Train_Pvals)\n",
    "\n",
    "Train_NullModel <- commandArgs(TRUE)[3]\n",
    "print(Train_NullModel)\n",
    "\n",
    "Tune_NullModel <- commandArgs(TRUE)[4]\n",
    "print(Tune_NullModel)\n",
    "\n",
    "Validation_NullModel <- commandArgs(TRUE)[5]\n",
    "print(Validation_NullModel)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[6]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "RV_PRS <- commandArgs(TRUE)[8]\n",
    "print(RV_PRS)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[9]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Train_PVals_All <- read.csv(Train_Pvals)\n",
    "Train_PVals_All <- Train_PVals_All[Train_PVals_All$STAARB <= 1e-03,]\n",
    "\n",
    "## Null Model\n",
    "obj_nullmodel_train <- get(load(Train_NullModel))\n",
    "obj_nullmodel_tune <- get(load(Tune_NullModel))\n",
    "obj_nullmodel_validation <- get(load(Validation_NullModel))\n",
    "\n",
    "obj_nullmodel <- obj_nullmodel_train\n",
    "obj_nullmodel$id_include <- c(obj_nullmodel_train$id_include,obj_nullmodel_tune$id_include,obj_nullmodel_validation$id_include)\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "\n",
    "G_star_gene_centric_coding <- list()\n",
    "\n",
    "for(i in 1:nrow(Train_PVals_All)){\n",
    "  ## Chr\n",
    "  chr <- Train_PVals_All$Chr[i]\n",
    "  ## Gene name\n",
    "  gene_name <- Train_PVals_All$Gene[i]\n",
    "  ## Coding mask\n",
    "  category <- Train_PVals_All$Category[i]\n",
    "  \n",
    "  ### gds file\n",
    "  gds.path <- paste0(INPUT_PATH,\"/exome.chr\",chr,\".gds\")\n",
    "  genofile <- seqOpen(gds.path)\n",
    "  \n",
    "  G_star_gene_centric_coding[[i]] <- Gene_Centric_Coding_G_Star(chr=chr,gene_name=gene_name,category=category ,\n",
    "                                                                genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                                                QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,silent=FALSE) \n",
    "  seqClose(genofile) \n",
    "} \n",
    "\n",
    "G_star_gene_centric_coding <- do.call(cbind,G_star_gene_centric_coding)\n",
    "\n",
    "col_remove <- apply(G_star_gene_centric_coding,2,function(x){sum(x != 0)}) > 10 & colSums(G_star_gene_centric_coding) > 10 \n",
    "G_star_gene_centric_coding <- G_star_gene_centric_coding[,col_remove,drop = FALSE]\n",
    "\n",
    "Train_PVals_All <- Train_PVals_All[col_remove,]\n",
    "\n",
    "print(str(G_star_gene_centric_coding))\n",
    "\n",
    "ids_gstar <- obj_nullmodel$id_include\n",
    "\n",
    "G_star_gene_centric_coding_train <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_train$id_include,]\n",
    "G_star_gene_centric_coding_tune <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_tune$id_include,]\n",
    "G_star_gene_centric_coding_vad <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_validation$id_include,]\n",
    "\n",
    "rm(G_star_gene_centric_coding)\n",
    "\n",
    "\n",
    "\n",
    "X_valid <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_validation$id_include],G_star_gene_centric_coding_vad)\n",
    "RV_PRS <- read.csv(RV_PRS)\n",
    "tmp <- inner_join(RV_PRS[,c(\"IID\",\"RV_PRS\")],X_valid)\n",
    "tmp <- subset(tmp,select = -c(IID))\n",
    "Train_PVals_All$Beta <- coef(lm(RV_PRS~.,tmp))[-1]\n",
    "Train_PVals_All$Beta[is.na(Train_PVals_All$Beta)] <- 0\n",
    "Train_PVals_All$Beta[abs(Train_PVals_All$Beta) < 1e-10] <- 0\n",
    "\n",
    "write.csv(Train_PVals_All,file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing RareVariantCoefficients.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile RareVariantCoefficients.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Train_Pvals} ${Train_NullModel} ${Tune_NullModel} ${Validation_NullModel} ${Annotation_name_catalog} ${INPUT_PATH} ${RV_PRS} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "            '--env trait'=trait,\n",
    "            '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/RareVariantCoefficients.R\",\n",
    "            '--input Train_Pvals'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding/\",trait,\"_coding_sig.csv\"),\n",
    "            '--input Train_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Tune_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Tune_Null_Model.RData\"),\n",
    "            '--input Validation_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Validation_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input RV_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\",trait,\"_BestPRS.csv\"),\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS\",\n",
    "            check.names = FALSE\n",
    "        ))   \n",
    "}\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://RareVariantCoefficients.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  8.4 KiB/  8.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/8.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp RareVariantCoefficients.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: rarevarian--williamsjacr--240902-144749-64\n",
      "  job-name: rarevariantcoefficients\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/313460965501063702\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1070285458177013171\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8852094039577346100\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3294359260856517488\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1907302209952564823\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12671696551866663672\n",
      "Launched job-id: rarevarian--williamsjacr--240902-144749-64\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'rarevarian--williamsjacr--240902-144749-64' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'rarevarian--williamsjacr--240902-144749-64' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script RareVariantCoefficients.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "rarevariantc...       6  Success   09-02 16:33:40\r\n",
      "rarevariantc...       5  Success   09-02 15:24:48\r\n",
      "rarevariantc...       4  Success   09-02 15:32:14\r\n",
      "rarevariantc...       3  Success   09-02 15:28:02\r\n",
      "rarevariantc...       2  Success   09-02 15:27:59\r\n",
      "rarevariantc...       1  Success   09-02 15:41:49\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'rarevarian--williamsjacr--240902-144749-64' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-09-02 14:27:36 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/rarevarian--williamsjacr--240902-142432-20.1.log\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7\r\n",
      "2024-09-02 14:27:37 INFO: mkdir -m 777 -p /mnt/data/output/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\r\n",
      "2024-09-02 14:27:37 INFO: Localizing INPUT_PATH\r\n",
      "2024-09-02 14:27:37 INFO: gsutil -u terra-vpc-sc-804f445b -mq rsync -r gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1 /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/\r\n",
      "2024-09-02 14:28:41 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/rarevarian--williamsjacr--240902-142432-20.1.log\r\n",
      "2024-09-02 14:31:46 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/rarevarian--williamsjacr--240902-142432-20.1.log\r\n",
      "2024-09-02 14:40:19 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/rarevarian--williamsjacr--240902-142432-20.1.log\r\n",
      "2024-09-02 14:42:01 INFO: Localizing R_Script\r\n",
      "2024-09-02 14:42:01 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/RareVariantCoefficients.R /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/RareVariantCoefficients.R\r\n",
      "2024-09-02 14:42:16 INFO: Localizing Train_NullModel\r\n",
      "2024-09-02 14:42:16 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/BMI_Train_Null_Model.RData /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/BMI_Train_Null_Model.RData\r\n",
      "2024-09-02 14:42:23 INFO: Localizing Validation_NullModel\r\n",
      "2024-09-02 14:42:23 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/BMI_Validation_Null_Model.RData /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/BMI_Validation_Null_Model.RData\r\n",
      "2024-09-02 14:42:26 INFO: Localizing RV_PRS\r\n",
      "2024-09-02 14:42:26 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-09-02 14:42:33 WARNING: Sleeping 10s before the next attempt of failed gsutil command\r\n",
      "2024-09-02 14:42:33 WARNING: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "2024-09-02 14:42:43 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-09-02 14:42:45 WARNING: Sleeping 10s before the next attempt of failed gsutil command\r\n",
      "2024-09-02 14:42:45 WARNING: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "2024-09-02 14:42:55 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-09-02 14:42:56 WARNING: Sleeping 10s before the next attempt of failed gsutil command\r\n",
      "2024-09-02 14:42:56 WARNING: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "2024-09-02 14:43:06 INFO: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: No URLs matched: gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "CommandException: 1 file/object could not be transferred.\r\n",
      "2024-09-02 14:43:08 ERROR: gsutil  -u terra-vpc-sc-804f445b -mq cp gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Continuous/RareVariant_PRS/BMI_BestPRS.csv\r\n",
      "2024-09-02 14:43:15 INFO: gsutil -h Content-Type:text/plain -u terra-vpc-sc-804f445b -mq cp /tmp/continuous_logging_action/output gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/rarevarian--williamsjacr--240902-142432-20.1.log\r\n"
     ]
    }
   ],
   "source": [
    "!gsutil cat gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/data/logging/rarevarian--williamsjacr--240902-142432-20.1.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Plus_Rare.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Plus_Rare.R\n",
    "rm(list = ls())\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "RareVariantPRS <- commandArgs(TRUE)[2]\n",
    "print(RareVariantPRS)\n",
    "\n",
    "CommonVariantPRS <- commandArgs(TRUE)[3]\n",
    "print(CommonVariantPRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[4]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "RV_PRS <- read.csv(RareVariantPRS)\n",
    "CV_PRS <- read.delim(CommonVariantPRS)\n",
    "\n",
    "CV_RV_PRS <- inner_join(RV_PRS,CV_PRS)\n",
    "\n",
    "CV_RV_PRS_raw <- CV_RV_PRS\n",
    "CV_RV_PRS_adjusted <- CV_RV_PRS\n",
    "\n",
    "for(i in c(\"RV_PRS\",\"prs\")){\n",
    "  tmp <- data.frame(y = CV_RV_PRS_adjusted[,i],CV_RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "  mod <- lm(y~.,data = tmp)\n",
    "  R <- mod$residuals\n",
    "  tmp <- data.frame(y = R^2,CV_RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "  mod <- lm(y~.,data = tmp)\n",
    "  y_hat <- predict(mod,tmp)\n",
    "  if(sum(y_hat < 0) > 0){\n",
    "    mod <- lm(y~1,data = tmp)\n",
    "    y_hat <- predict(mod,tmp)\n",
    "  }\n",
    "  if(sum(sqrt(y_hat)) == 0){\n",
    "    CV_RV_PRS_adjusted[,i] <- 0\n",
    "  }else{\n",
    "    CV_RV_PRS_adjusted[,i] <- R/sqrt(y_hat)\n",
    "  }\n",
    "}\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "CV_RV_PRS_raw_EUR <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "CV_RV_PRS_raw_SAS <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "CV_RV_PRS_raw_AMR <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "CV_RV_PRS_raw_AFR <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "CV_RV_PRS_raw_EAS <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "CV_RV_PRS_raw_MID <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "CV_RV_PRS_adjusted_EUR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "CV_RV_PRS_adjusted_SAS <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "CV_RV_PRS_adjusted_AMR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "CV_RV_PRS_adjusted_AFR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "CV_RV_PRS_adjusted_EAS <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "CV_RV_PRS_adjusted_MID <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "CV_RV_PRS_raw_EUR$Y <- scale(CV_RV_PRS_raw_EUR$Y)\n",
    "CV_RV_PRS_raw_SAS$Y <- scale(CV_RV_PRS_raw_SAS$Y)\n",
    "CV_RV_PRS_raw_AMR$Y <- scale(CV_RV_PRS_raw_AMR$Y)\n",
    "CV_RV_PRS_raw_AFR$Y <- scale(CV_RV_PRS_raw_AFR$Y)\n",
    "CV_RV_PRS_raw_EAS$Y <- scale(CV_RV_PRS_raw_EAS$Y)\n",
    "CV_RV_PRS_raw_MID$Y <- scale(CV_RV_PRS_raw_MID$Y)\n",
    "\n",
    "CV_RV_PRS_adjusted_EUR$Y <- scale(CV_RV_PRS_adjusted_EUR$Y)\n",
    "CV_RV_PRS_adjusted_SAS$Y <- scale(CV_RV_PRS_adjusted_SAS$Y)\n",
    "CV_RV_PRS_adjusted_AMR$Y <- scale(CV_RV_PRS_adjusted_AMR$Y)\n",
    "CV_RV_PRS_adjusted_AFR$Y <- scale(CV_RV_PRS_adjusted_AFR$Y)\n",
    "CV_RV_PRS_adjusted_EAS$Y <- scale(CV_RV_PRS_adjusted_EAS$Y)\n",
    "CV_RV_PRS_adjusted_MID$Y <- scale(CV_RV_PRS_adjusted_MID$Y)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$RV_PRS <- scale(CV_RV_PRS_raw_EUR$RV_PRS)\n",
    "CV_RV_PRS_raw_SAS$RV_PRS <- scale(CV_RV_PRS_raw_SAS$RV_PRS)\n",
    "CV_RV_PRS_raw_AMR$RV_PRS <- scale(CV_RV_PRS_raw_AMR$RV_PRS)\n",
    "CV_RV_PRS_raw_AFR$RV_PRS <- scale(CV_RV_PRS_raw_AFR$RV_PRS)\n",
    "CV_RV_PRS_raw_EAS$RV_PRS <- scale(CV_RV_PRS_raw_EAS$RV_PRS)\n",
    "CV_RV_PRS_raw_MID$RV_PRS <- scale(CV_RV_PRS_raw_MID$RV_PRS)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$prs <- scale(CV_RV_PRS_raw_EUR$prs)\n",
    "CV_RV_PRS_raw_SAS$prs <- scale(CV_RV_PRS_raw_SAS$prs)\n",
    "CV_RV_PRS_raw_AMR$prs <- scale(CV_RV_PRS_raw_AMR$prs)\n",
    "CV_RV_PRS_raw_AFR$prs <- scale(CV_RV_PRS_raw_AFR$prs)\n",
    "CV_RV_PRS_raw_EAS$prs <- scale(CV_RV_PRS_raw_EAS$prs)\n",
    "CV_RV_PRS_raw_MID$prs <- scale(CV_RV_PRS_raw_MID$prs)\n",
    "\n",
    "\n",
    "best_beta_raw_CV_EUR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EUR))[2]\n",
    "se_beta_raw_CV_EUR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EUR))$coefficients[2,2]\n",
    "best_beta_raw_RV_EUR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EUR))[3]\n",
    "se_beta_raw_RV_EUR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EUR))$coefficients[3,2]\n",
    "\n",
    "best_beta_raw_CV_SAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_SAS))[2]\n",
    "se_beta_raw_CV_SAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_SAS))$coefficients[2,2]\n",
    "best_beta_raw_RV_SAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_SAS))[3]\n",
    "se_beta_raw_RV_SAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_SAS))$coefficients[3,2]\n",
    "\n",
    "best_beta_raw_CV_AMR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AMR))[2]\n",
    "se_beta_raw_CV_AMR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AMR))$coefficients[2,2]\n",
    "best_beta_raw_RV_AMR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AMR))[3]\n",
    "se_beta_raw_RV_AMR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AMR))$coefficients[3,2]\n",
    "\n",
    "best_beta_raw_CV_AFR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AFR))[2]\n",
    "se_beta_raw_CV_AFR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AFR))$coefficients[2,2]\n",
    "best_beta_raw_RV_AFR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AFR))[3]\n",
    "se_beta_raw_RV_AFR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_AFR))$coefficients[3,2]\n",
    "\n",
    "best_beta_raw_CV_EAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EAS))[2]\n",
    "se_beta_raw_CV_EAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EAS))$coefficients[2,2]\n",
    "best_beta_raw_RV_EAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EAS))[3]\n",
    "se_beta_raw_RV_EAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_EAS))$coefficients[3,2]\n",
    "\n",
    "best_beta_raw_CV_MID <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_MID))[2]\n",
    "se_beta_raw_CV_MID <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_MID))$coefficients[2,2]\n",
    "best_beta_raw_RV_MID <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_MID))[3]\n",
    "se_beta_raw_RV_MID <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_raw_MID))$coefficients[3,2]\n",
    "\n",
    "\n",
    "\n",
    "best_beta_adjusted_CV_EUR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EUR))[2]\n",
    "se_beta_adjusted_CV_EUR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EUR))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_EUR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EUR))[3]\n",
    "se_beta_adjusted_RV_EUR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EUR))$coefficients[3,2]\n",
    "\n",
    "best_beta_adjusted_CV_SAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_SAS))[2]\n",
    "se_beta_adjusted_CV_SAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_SAS))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_SAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_SAS))[3]\n",
    "se_beta_adjusted_RV_SAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_SAS))$coefficients[3,2]\n",
    "\n",
    "best_beta_adjusted_CV_AMR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AMR))[2]\n",
    "se_beta_adjusted_CV_AMR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AMR))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_AMR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AMR))[3]\n",
    "se_beta_adjusted_RV_AMR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AMR))$coefficients[3,2]\n",
    "\n",
    "best_beta_adjusted_CV_AFR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AFR))[2]\n",
    "se_beta_adjusted_CV_AFR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AFR))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_AFR <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AFR))[3]\n",
    "se_beta_adjusted_RV_AFR <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_AFR))$coefficients[3,2]\n",
    "\n",
    "best_beta_adjusted_CV_EAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EAS))[2]\n",
    "se_beta_adjusted_CV_EAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EAS))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_EAS <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EAS))[3]\n",
    "se_beta_adjusted_RV_EAS <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_EAS))$coefficients[3,2]\n",
    "\n",
    "best_beta_adjusted_CV_MID <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_MID))[2]\n",
    "se_beta_adjusted_CV_MID <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_MID))$coefficients[2,2]\n",
    "best_beta_adjusted_RV_MID <- coef(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_MID))[3]\n",
    "se_beta_adjusted_RV_MID <- summary(lm(Y~prs + RV_PRS,data = CV_RV_PRS_adjusted_MID))$coefficients[3,2]\n",
    "\n",
    "\n",
    "CV_PRS_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                             beta_raw = c(best_beta_raw_CV_EUR,best_beta_raw_CV_SAS,best_beta_raw_CV_AMR,best_beta_raw_CV_AFR,best_beta_raw_CV_EAS,best_beta_raw_CV_MID), \n",
    "                             se_raw = c(se_beta_raw_CV_EUR,se_beta_raw_CV_SAS,se_beta_raw_CV_AMR,se_beta_raw_CV_AFR,se_beta_raw_CV_EAS,se_beta_raw_CV_MID), \n",
    "                             beta_adjusted = c(best_beta_adjusted_CV_EUR,best_beta_adjusted_CV_SAS,best_beta_adjusted_CV_AMR,best_beta_adjusted_CV_AFR,best_beta_adjusted_CV_EAS,best_beta_adjusted_CV_MID), \n",
    "                             se_adjusted = c(se_beta_adjusted_CV_EUR,se_beta_adjusted_CV_SAS,se_beta_adjusted_CV_AMR,se_beta_adjusted_CV_AFR,se_beta_adjusted_CV_EAS,se_beta_adjusted_CV_MID),\n",
    "                             Method = \"CV\")\n",
    "\n",
    "RV_PRS_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                             beta_raw = c(best_beta_raw_RV_EUR,best_beta_raw_RV_SAS,best_beta_raw_RV_AMR,best_beta_raw_RV_AFR,best_beta_raw_RV_EAS,best_beta_raw_RV_MID), \n",
    "                             se_raw = c(se_beta_raw_RV_EUR,se_beta_raw_RV_SAS,se_beta_raw_RV_AMR,se_beta_raw_RV_AFR,se_beta_raw_RV_EAS,se_beta_raw_RV_MID), \n",
    "                             beta_adjusted = c(best_beta_adjusted_RV_EUR,best_beta_adjusted_RV_SAS,best_beta_adjusted_RV_AMR,best_beta_adjusted_RV_AFR,best_beta_adjusted_RV_EAS,best_beta_adjusted_RV_MID), \n",
    "                             se_adjusted = c(se_beta_adjusted_RV_EUR,se_beta_adjusted_RV_SAS,se_beta_adjusted_RV_AMR,se_beta_adjusted_RV_AFR,se_beta_adjusted_RV_EAS,se_beta_adjusted_RV_MID),\n",
    "                             Method = \"RV\")\n",
    "\n",
    "\n",
    "write.csv(rbind(CV_PRS_Results,RV_PRS_Results),file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Common_Plus_Rare.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Plus_Rare.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${RareVariantPRS} ${CommonVariantPRS} ${all_phenotypes_file} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "            '--env trait'=trait,\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Common_Plus_Rare.R\",\n",
    "            '--input RareVariantPRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\",trait,\"_BestPRS.csv\"),\n",
    "            '--input CommonVariantPRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",trait,\"_Best_Validation_All.txt\"),\n",
    "            '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/Common_Plus_Rare/\",\n",
    "            check.names = FALSE\n",
    "        ))   \n",
    "}\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Common_Plus_Rare.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 10.0 KiB/ 10.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/10.0 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Common_Plus_Rare.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: common-plu--williamsjacr--240807-165037-00\n",
      "  job-name: common-plus-rare\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11963387640502094454\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7050916832642191274\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16627059054397606154\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17531345581312317542\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6893272649865612905\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15003385858120078740\n",
      "Launched job-id: common-plu--williamsjacr--240807-165037-00\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-plu--williamsjacr--240807-165037-00' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-plu--williamsjacr--240807-165037-00' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Common_Plus_Rare.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "common-plus-...       6  Success   08-07 16:54:03\r\n",
      "common-plus-...       5  Success   08-07 16:54:02\r\n",
      "common-plus-...       4  Success   08-07 16:54:08\r\n",
      "common-plus-...       3  Success   08-07 16:54:07\r\n",
      "common-plus-...       2  Success   08-07 16:54:14\r\n",
      "common-plus-...       1  Success   08-07 16:53:58\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-plu--williamsjacr--240807-165037-00' --users 'williamsjacr' --status '*'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
