{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket = os.getenv(\"WORKSPACE_BUCKET\")\n",
    "\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: USER_NAME=williamsjacr\n"
     ]
    }
   ],
   "source": [
    "%env USER_NAME={USER_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dsub in /opt/conda/lib/python3.10/site-packages (0.5.0)\n",
      "Requirement already satisfied: funcsigs==1.0.2 in /opt/conda/lib/python3.10/site-packages (from dsub) (1.0.2)\n",
      "Requirement already satisfied: google-api-core<=2.19.0,>=2.7.3 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.19.0)\n",
      "Requirement already satisfied: google-api-python-client<=2.131.0,>=2.47.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.131.0)\n",
      "Requirement already satisfied: google-auth-httplib2<=0.2.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.2.0)\n",
      "Requirement already satisfied: google-auth<=2.29.0,>=2.6.6 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.29.0)\n",
      "Requirement already satisfied: google-cloud-batch<=0.17.20 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.17.20)\n",
      "Requirement already satisfied: httplib2<=0.22.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.21.0)\n",
      "Requirement already satisfied: mock<=5.1.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (5.1.0)\n",
      "Requirement already satisfied: parameterized<=0.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.9.0)\n",
      "Requirement already satisfied: protobuf<=5.26.0,>=3.19.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (3.20.2)\n",
      "Requirement already satisfied: pyasn1-modules<=0.4.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.4.0)\n",
      "Requirement already satisfied: pyasn1<=0.6.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.6.0)\n",
      "Requirement already satisfied: python-dateutil<=2.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.8.2)\n",
      "Requirement already satisfied: pytz<=2024.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (2023.3)\n",
      "Requirement already satisfied: pyyaml<=6.0.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (6.0.1)\n",
      "Requirement already satisfied: rsa<=4.9 in /opt/conda/lib/python3.10/site-packages (from dsub) (4.9)\n",
      "Requirement already satisfied: tabulate<=0.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.9.0)\n",
      "Requirement already satisfied: tenacity<=8.2.3 in /opt/conda/lib/python3.10/site-packages (from dsub) (8.2.3)\n",
      "Requirement already satisfied: uritemplate<=4.1.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (3.0.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.dev0,>=1.56.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (1.66.0)\n",
      "Requirement already satisfied: proto-plus<2.0.0dev,>=1.22.3 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (1.25.0)\n",
      "Requirement already satisfied: requests<3.0.0.dev0,>=2.18.0 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (2.32.3)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (4.2.4)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.33.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-batch<=0.17.20->dsub) (1.68.1)\n",
      "Requirement already satisfied: grpcio-status<2.0.dev0,>=1.33.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-batch<=0.17.20->dsub) (1.48.2)\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in /opt/conda/lib/python3.10/site-packages (from httplib2<=0.22.0->dsub) (3.2.0)\n",
      "Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.10/site-packages (from python-dateutil<=2.9.0->dsub) (1.17.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (3.4.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (1.26.20)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (2024.12.14)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install --upgrade dsub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /home/jupyter/aou_dsub.bash\n"
     ]
    }
   ],
   "source": [
    "%%writefile ~/aou_dsub.bash\n",
    "#!/bin/bash\n",
    "function aou_dsub () {\n",
    "\n",
    "  # Get a shorter username to leave more characters for the job name.\n",
    "  local DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "  # For AoU RWB projects network name is \"network\".\n",
    "  local AOU_NETWORK=network\n",
    "  local AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "  dsub \\\n",
    "      --provider google-cls-v2 \\\n",
    "      --user-project \"${GOOGLE_PROJECT}\"\\\n",
    "      --project \"${GOOGLE_PROJECT}\"\\\n",
    "      --image 'marketplace.gcr.io/google/ubuntu1804:latest' \\\n",
    "      --network \"${AOU_NETWORK}\" \\\n",
    "      --subnetwork \"${AOU_SUBNETWORK}\" \\\n",
    "      --service-account \"$(gcloud config get-value account)\" \\\n",
    "      --user \"${DSUB_USER_NAME}\" \\\n",
    "      --regions us-central1 \\\n",
    "      --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "      \"$@\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo source ~/aou_dsub.bash >> ~/.bashrc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Variants.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Variants.R\n",
    "rm(list = ls())\n",
    "\n",
    "chr <- as.numeric(commandArgs(TRUE)[1])\n",
    "print(chr)\n",
    "\n",
    "BED_file <- commandArgs(TRUE)[2]\n",
    "BED_file <- gsub(\"\\\\.bed\",\"\",BED_file)\n",
    "print(BED_file)\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[5]\n",
    "print(Ancestry_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "system(paste0(\"rm \",Ancestry_File))\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --geno 0.10 --make-bed --out tmp1\"),intern = TRUE)\n",
    "system(paste0(\"rm \",BED_file,\".bed\"));system(paste0(\"rm \",BED_file,\".bim\"));system(paste0(\"rm \",BED_file,\".fam\"))\n",
    "system(paste0(\"plink --bfile tmp1 --mind 0.05 --make-bed --out tmp2\"),intern = TRUE)\n",
    "system(\"rm tmp1.bed\");system(\"rm tmp1.bim\");system(\"rm tmp1.fam\")\n",
    "\n",
    "\n",
    "AFR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AFR\"])\n",
    "AMR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AMR\"])\n",
    "EUR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"EUR\"])\n",
    "\n",
    "write.table(AFR_Keep,\"AFR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(AMR_Keep,\"AMR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(EUR_Keep,\"EUR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile tmp2 --keep AFR_Keep.txt --freq --out chr\",chr,\"_freq_AFR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile tmp2 --keep AMR_Keep.txt --freq --out chr\",chr,\"_freq_AMR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile tmp2 --keep EUR_Keep.txt --freq --out chr\",chr,\"_freq_EUR\"),intern = TRUE)\n",
    "\n",
    "a <- read.csv(paste0(\"chr\",chr,\"_freq_AFR.frq\"),sep = \"\")\n",
    "b <- read.csv(paste0(\"chr\",chr,\"_freq_AMR.frq\"),sep = \"\")\n",
    "c <- read.csv(paste0(\"chr\",chr,\"_freq_EUR.frq\"),sep = \"\")\n",
    "\n",
    "exclude_list <- a$SNP[is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF) | a$MAF <= 0.01 | b$MAF <= 0.01 | c$MAF <= 0.01]\n",
    "write.table(exclude_list,\"exclude_list.txt\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "system(paste0(\"plink --bfile tmp2 --exclude exclude_list.txt --make-bed --out \",OUTPUT_PATH,\"/chr\",chr,\"_corrected\"),intern = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for (chrom in 1:12) {\n",
    "  tasks <- rbind(tasks, data.frame(\n",
    "    '--env CHR'=chrom,\n",
    "    '--input BED_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/acaf_threshold_v7.1/plink_bed/acaf_threshold.chr\",chrom,\".bed\"),\n",
    "    '--input BIM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/acaf_threshold_v7.1/plink_bed/acaf_threshold.chr\",chrom,\".bim\"),\n",
    "    '--input FAM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/acaf_threshold_v7.1/plink_bed/acaf_threshold.chr\",chrom,\".fam\"),\n",
    "    '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Common_Variants.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "}\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Variants.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Variants.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${CHR} ${BED_File} ${BIM_File} ${FAM_File} ${Ancestry_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Common_Variants.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.1 KiB/  2.1 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.1 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Common_Variants.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: common-var--williamsjacr--250410-125128-66\n",
      "  job-name: common-variants\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/609494964870255421\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12779355517925240545\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5962273838059799743\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12207914947044021913\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9136467727547312618\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4436065283973798346\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1749630105134682131\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17896130386990975270\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9012679296694992357\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7210797182379846626\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7944522520059069733\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3180217167494848690\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14797301264773578671\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4984858478917309752\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9888901832816483618\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7968910471927449901\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11289955813765974662\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/77098744562983382\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13835991048086782229\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8777991212067187520\n",
      "Launched job-id: common-var--williamsjacr--250410-125128-66\n",
      "20 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--250410-125128-66' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--250410-125128-66' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 1000 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Common_Variants.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Variants_BED_Sum.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Variants_BED_Sum.R\n",
    "rm(list = ls())\n",
    "\n",
    "chr <- as.numeric(commandArgs(TRUE)[1])\n",
    "print(chr)\n",
    "\n",
    "BED_file <- commandArgs(TRUE)[2]\n",
    "BED_file <- gsub(\".bed\",\"\",BED_file,fixed = TRUE)\n",
    "print(BED_file)\n",
    "\n",
    "CT_BIM_FILE <- commandArgs(TRUE)[3]\n",
    "print(CT_BIM_FILE)\n",
    "\n",
    "PROSPER_BIM_FILE <- commandArgs(TRUE)[4]\n",
    "print(PROSPER_BIM_FILE)\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[5]\n",
    "print(Ancestry_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "\n",
    "AFR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AFR\"])\n",
    "AMR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AMR\"])\n",
    "EUR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"EUR\"])\n",
    "\n",
    "\n",
    "write.table(AFR_Keep,\"AFR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(AMR_Keep,\"AMR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(EUR_Keep,\"EUR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AFR_Keep.txt --freq --out chr\",chr,\"_freq_AFR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep AMR_Keep.txt --freq --out chr\",chr,\"_freq_AMR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_file,\" --keep EUR_Keep.txt --freq --out chr\",chr,\"_freq_EUR\"),intern = TRUE)\n",
    "\n",
    "a <- read.csv(paste0(\"chr\",chr,\"_freq_AFR.frq\"),sep = \"\")\n",
    "b <- read.csv(paste0(\"chr\",chr,\"_freq_AMR.frq\"),sep = \"\")\n",
    "c <- read.csv(paste0(\"chr\",chr,\"_freq_EUR.frq\"),sep = \"\")\n",
    "\n",
    "exclude_list <- a$SNP[is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF) | a$MAF <= 0.01 | b$MAF <= 0.01 | c$MAF <= 0.01]\n",
    "rarevariant_list <- a$SNP[is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF)]\n",
    "\n",
    "Number_Common_Variants <- nrow(a) - length(exclude_list)\n",
    "Number_Rare_Variants <- nrow(a) - length(rarevariant_list) - Number_Common_Variants\n",
    "\n",
    "CT_BIM <- read.delim(CT_BIM_FILE,header = FALSE)\n",
    "CT_SNPs <- nrow(CT_BIM)\n",
    "\n",
    "PROSPER_BIM <- read.delim(PROSPER_BIM_FILE,header = FALSE)\n",
    "PROSPER_SNPs <- nrow(PROSPER_BIM)\n",
    "\n",
    "tmp <- data.frame(Chr = chr,\n",
    "                  Number_Common_Variants = Number_Common_Variants,\n",
    "                  Number_Rare_Variants = Number_Rare_Variants,\n",
    "                 CT_SNPs = CT_SNPs,\n",
    "                 PROSPER_SNPs = PROSPER_SNPs)\n",
    "write.csv(tmp,row.names = FALSE,file = paste0(OUTPUT_PATH,\"/\",chr,\"_NumSNPs.csv\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for (chrom in 1:22) {\n",
    "  tasks <- rbind(tasks, data.frame(\n",
    "    '--env CHR'=chrom,\n",
    "    '--input BED_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/plink_bed/exome.chr\",chrom,\".bed\"),\n",
    "    '--input BIM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/plink_bed/exome.chr\",chrom,\".bim\"),\n",
    "    '--input FAM_File'=paste0(\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/plink_bed/exome.chr\",chrom,\".fam\"),\n",
    "    '--input CT_BIM_FILE'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input PROSPER_BIM_FILE'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",  \n",
    "    '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Common_Variants_BED_Sum.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Number_SNPs\",\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "}\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Common_Variants_BED_Sum.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Variants_BED_Sum.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${CHR} ${BED_File} ${CT_BIM_FILE} ${PROSPER_BIM_FILE} ${Ancestry_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Common_Variants_BED_Sum.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.3 KiB/  2.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Common_Variants_BED_Sum.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: common-var--williamsjacr--240901-131943-66\n",
      "  job-name: common-variants-bed-sum\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11196432129391680803\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7824575572013587726\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9758527684499917372\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15865405136015907616\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13738081057261705017\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1215447884564362733\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12706738842639517649\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9128748637006943225\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17641149042402703339\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9298657386333898969\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8368203038466850967\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6101676909241046982\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10799503408380772836\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9032221348798239212\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7251358705368745639\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9939768856006987805\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13772409251482507036\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17914230607144014707\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13585512460409370635\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10923239924929804856\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13427428528047602329\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13504958931632819405\n",
      "Launched job-id: common-var--williamsjacr--240901-131943-66\n",
      "22 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--240901-131943-66' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-var--williamsjacr--240901-131943-66' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 400 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 32 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Common_Variants_BED_Sum.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting merge_bed.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile merge_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[1]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "merge_list_train <- paste0(INPUT_PATH,\"/chr\",2:22,\"_corrected\")\n",
    "write.table(merge_list_train,file = \"merge_list.txt\",col.names = F,row.names = F,quote=F)\n",
    "system(paste0(\"plink --bfile \",INPUT_PATH,\"/chr\",1,\"_corrected --merge-list merge_list.txt --make-bed --out \",OUTPUT_PATH,\"/all_chr\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/merge_bed.R\",\n",
    "  '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "  # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting merge_bed.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile merge_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://merge_bed.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  417.0 B/  417.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/417.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp merge_bed.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: merge-bed--williamsjacr--250414-112849-13\n",
      "  job-name: merge-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1734854414342758511\n",
      "Launched job-id: merge-bed--williamsjacr--250414-112849-13\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'merge-bed--williamsjacr--250414-112849-13' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'merge-bed--williamsjacr--250414-112849-13' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 2000 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 72 \\\n",
    "  --timeout \"72h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script merge_bed.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting all_phenotypes.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile all_phenotypes.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Related_File <- commandArgs(TRUE)[1]\n",
    "print(Related_File)\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[2]\n",
    "print(Ancestry_File)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "flagged_relatedness <- read.delim(Related_File)\n",
    "colnames(flagged_relatedness) <- \"IID\"\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "aou_BMI_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_BMI_pheno.tsv\"))\n",
    "aou_LDL_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_LDL_pheno.tsv\"))\n",
    "aou_HDL_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_HDL_pheno.tsv\"))\n",
    "aou_TG_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_TG_pheno.tsv\"))\n",
    "colnames(aou_TG_pheno)[3] <- \"logTG\"\n",
    "aou_TC_pheno <- read.delim(paste0(INPUT_PATH,\"/aou_TC_pheno.tsv\"))\n",
    "aou_Height_pheno <- read.csv(paste0(INPUT_PATH,\"/aou_Height_pheno.tsv\"),sep =\"\")\n",
    "\n",
    "aou_BMI_cov <- read.delim(paste0(INPUT_PATH,\"/aou_BMI_cov.tsv\"))\n",
    "aou_LDL_cov <- read.csv(paste0(INPUT_PATH,\"/aou_LDL_cov.tsv\"),sep =\"\")\n",
    "aou_HDL_cov <- read.csv(paste0(INPUT_PATH,\"/aou_HDL_cov.tsv\"),sep =\"\")\n",
    "aou_TG_cov <- read.csv(paste0(INPUT_PATH,\"/aou_TG_cov.tsv\"),sep =\"\")\n",
    "aou_TC_cov <- read.csv(paste0(INPUT_PATH,\"/aou_TC_cov.tsv\"),sep =\"\")\n",
    "aou_Height_cov <- read.csv(paste0(INPUT_PATH,\"/aou_Height_cov.tsv\"),sep =\"\")\n",
    "\n",
    "library(dplyr)\n",
    "unique_ids_pheno <- unique(c(aou_BMI_pheno$IID,aou_LDL_pheno$IID,aou_HDL_pheno$IID,aou_TG_pheno$IID,aou_TC_pheno$IID,aou_Height_pheno$IID))\n",
    "unique_ids_cov <- unique(c(aou_BMI_cov$IID,aou_LDL_cov$IID,aou_HDL_cov$IID,aou_TG_cov$IID,aou_TC_cov$IID,aou_Height_cov$IID))\n",
    "\n",
    "\n",
    "all_phenotypes <- data.frame(IID = unique(c(unique_ids_pheno,unique_ids_cov)))\n",
    "all_phenotypes <- all_phenotypes[!(all_phenotypes$IID %in% flagged_relatedness$IID),,drop = FALSE]\n",
    "\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_BMI_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_LDL_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_HDL_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TG_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TC_pheno)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_Height_pheno)\n",
    "\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_BMI_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_LDL_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_HDL_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TG_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_TC_cov)\n",
    "all_phenotypes <- left_join(all_phenotypes,aou_Height_cov)\n",
    "\n",
    "all_phenotypes <- left_join(all_phenotypes,ancestries)\n",
    "\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$age),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$female),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC1),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC2),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC3),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC4),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC5),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC6),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC7),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC8),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC9),]\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes$PC10),]\n",
    "\n",
    "write.csv(all_phenotypes,file = paste0(OUTPUT_PATH,\"/all_phenotypes.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/all_phenotypes.R\",\n",
    "  '--input Related_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/relatedness/relatedness_flagged_samples.tsv\",\n",
    "  '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "  '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/Tony_Files\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/\",\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting all_phenotypes.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile all_phenotypes.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Related_File} ${Ancestry_File} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://all_phenotypes.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  3.4 KiB/  3.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/3.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp all_phenotypes.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: all-phenot--williamsjacr--240627-152247-11\n",
      "  job-name: all-phenotypes\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3165015638967736981\n",
      "Launched job-id: all-phenot--williamsjacr--240627-152247-11\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'all-phenot--williamsjacr--240627-152247-11' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'all-phenot--williamsjacr--240627-152247-11' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script all_phenotypes.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Train_Tune_Validation.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Train_Tune_Validation.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "all_phenotype_File <- commandArgs(TRUE)[1]\n",
    "print(all_phenotype_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotype_File)\n",
    "\n",
    "i <- (1:length(all_phenotypes$IID))[all_phenotypes$ancestry %in% c(\"EUR\",\"AFR\",\"AMR\")]\n",
    "train_number <- round(length(all_phenotypes$IID)*0.7) + 1\n",
    "train <- sample(i, train_number)\n",
    "\n",
    "i <- (1:length(all_phenotypes$IID))[!((1:length(all_phenotypes$IID)) %in% train)]\n",
    "i_EUR <- i[all_phenotypes$ancestry[i] == \"EUR\"]\n",
    "i_AFR <- i[all_phenotypes$ancestry[i] == \"AFR\"]\n",
    "i_SAS <- i[all_phenotypes$ancestry[i] == \"SAS\"]\n",
    "i_EAS <- i[all_phenotypes$ancestry[i] == \"EAS\"]\n",
    "i_MID <- i[all_phenotypes$ancestry[i] == \"MID\"]\n",
    "i_AMR <- i[all_phenotypes$ancestry[i] == \"AMR\"]\n",
    "\n",
    "\n",
    "tune <- c(sample(i_EUR,round(length(i_EUR)/2)),\n",
    "          sample(i_AFR,round(length(i_AFR)/2)),\n",
    "          sample(i_SAS,round(length(i_SAS)/2)),\n",
    "          sample(i_EAS,round(length(i_EAS)/2)),\n",
    "          sample(i_MID,round(length(i_MID)/2)),\n",
    "          sample(i_AMR,round(length(i_AMR)/2)))\n",
    "\n",
    "validation <- i[!(i %in% tune)]\n",
    "\n",
    "\n",
    "train <- all_phenotypes$IID[train]\n",
    "tune <- all_phenotypes$IID[tune]\n",
    "validation <- all_phenotypes$IID[validation]\n",
    "\n",
    "\n",
    "write.table(train,paste0(OUTPUT_PATH,\"/train.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "write.table(tune,paste0(OUTPUT_PATH,\"/tune.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "write.table(validation,paste0(OUTPUT_PATH,\"/validation.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "reference <- train[sample(1:length(train),3000,replace = FALSE)]\n",
    "\n",
    "write.table(reference,paste0(OUTPUT_PATH,\"/reference.txt\"),row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "colnames(all_phenotypes)[10] <- \"sex\"\n",
    "\n",
    "phenotype_train <- all_phenotypes[all_phenotypes$IID %in% train,]\n",
    "phenotype_tune <- all_phenotypes[all_phenotypes$IID %in% tune,]\n",
    "phenotype_validation <- all_phenotypes[all_phenotypes$IID %in% validation,]\n",
    "\n",
    "phenotype_train$FID <- 0\n",
    "phenotype_tune$FID <- 0\n",
    "phenotype_validation$FID <- 0\n",
    "\n",
    "phenotype_train <- phenotype_train[,c(\"IID\",\"FID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "phenotype_train$age2 <- phenotype_train$age^2\n",
    "phenotype_tune <- phenotype_tune[,c(\"IID\",\"FID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "phenotype_tune$age2 <- phenotype_tune$age^2\n",
    "phenotype_validation <- phenotype_validation[,c(\"IID\",\"FID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "phenotype_validation$age2 <- phenotype_validation$age^2\n",
    "\n",
    "colnames(phenotype_train) <- c(\"IID\",colnames(phenotype_train)[-1])\n",
    "colnames(phenotype_tune) <- c(\"IID\",colnames(phenotype_tune)[-1])\n",
    "colnames(phenotype_validation) <- c(\"IID\",colnames(phenotype_validation)[-1])\n",
    "\n",
    "phenotype_train[c(9,11:21)] <- lapply(phenotype_train[c(9,11:21)], function(x){c(scale(x))})\n",
    "phenotype_tune[c(9,11:21)] <- lapply(phenotype_tune[c(9,11:21)], function(x){c(scale(x))})\n",
    "phenotype_validation[c(9,11:21)] <- lapply(phenotype_validation[c(9,11:21)], function(x){c(scale(x))})\n",
    "\n",
    "write.table(phenotype_train,file = paste0(OUTPUT_PATH,\"/All_Train.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "write.table(phenotype_tune,file = paste0(OUTPUT_PATH,\"/All_Tune.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "write.table(phenotype_validation,file = paste0(OUTPUT_PATH,\"/All_Validation.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Train_Tune_Validation.R\",\n",
    "  '--input all_phenotype_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/\",\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Train_Tune_Validation.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Train_Tune_Validation.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${all_phenotype_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Train_Tune_Validation.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  3.4 KiB/  3.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/3.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Train_Tune_Validation.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: train-tune--williamsjacr--240627-153410-54\n",
      "  job-name: train-tune-validation\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9958687970653510792\n",
      "Launched job-id: train-tune--williamsjacr--240627-153410-54\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'train-tune--williamsjacr--240627-153410-54' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'train-tune--williamsjacr--240627-153410-54' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 60 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Train_Tune_Validation.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting REGENIE_Prep_TraitSpecific.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE_Prep_TraitSpecific.R\n",
    "rm(list = ls())\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[1]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "all_train <- read.delim(all_train_file)\n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    for(trait in c(\"BMI\",\"TC\",\"logTG\",\"LDL\",\"HDL\",\"Height\")){\n",
    "        tmp <- all_train[all_train$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == anc],]\n",
    "        tmp <- tmp[!is.na(tmp[,trait]),]\n",
    "        tmp <- tmp[,c(\"FID\",\"IID\",\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\",\"age\",\"age2\",\"sex\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")]\n",
    "        write.table(tmp,file = paste0(OUTPUT_PATH,\"/\",\"All_Train_\",trait,\"_\",anc,\"_REGENIE.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/REGENIE_Prep_TraitSpecific.R\",\n",
    "  '--input all_phenotype_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "  '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/\",\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting REGENIE_Prep_TraitSpecific.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE_Prep_TraitSpecific.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${all_phenotype_File} ${all_train_file} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://REGENIE_Prep_TraitSpecific.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  840.0 B/  840.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/840.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp REGENIE_Prep_TraitSpecific.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: regenie-pr--williamsjacr--250426-143127-77\n",
      "  job-name: regenie-prep-traitspecific\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5063303981653871125\n",
      "Launched job-id: regenie-pr--williamsjacr--250426-143127-77\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-pr--williamsjacr--250426-143127-77' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-pr--williamsjacr--250426-143127-77' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 10 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 10 \\\n",
    "  --timeout \"4h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script REGENIE_Prep_TraitSpecific.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting REGENIE_Prep_BED.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE_Prep_BED.R\n",
    "rm(list = ls())\n",
    "\n",
    "Ancestry_File <- commandArgs(TRUE)[1]\n",
    "print(Ancestry_File)\n",
    "\n",
    "BED_File <- gsub(\".bed\",\"\",commandArgs(TRUE)[2])\n",
    "print(BED_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "ancestries <- read.delim(Ancestry_File)\n",
    "ancestries <- data.frame(IID = as.numeric(ancestries[,c(\"research_id\")]),ancestry = toupper(ancestries[,c(\"ancestry_pred\")]))\n",
    "\n",
    "AFR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AFR\"])\n",
    "AMR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"AMR\"])\n",
    "EUR_Keep <- data.frame(FID = 0,IID = ancestries$IID[ancestries$ancestry == \"EUR\"])\n",
    "\n",
    "write.table(AFR_Keep,\"AFR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(AMR_Keep,\"AMR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "write.table(EUR_Keep,\"EUR_Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_File,\" --keep AFR_Keep.txt --freq --out freq_AFR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_File,\" --keep AMR_Keep.txt --freq --out freq_AMR\"),intern = TRUE)\n",
    "system(paste0(\"plink --bfile \",BED_File,\" --keep EUR_Keep.txt --freq --out freq_EUR\"),intern = TRUE)\n",
    "\n",
    "a <- read.csv(paste0(\"freq_AFR.frq\"),sep = \"\")\n",
    "b <- read.csv(paste0(\"freq_AMR.frq\"),sep = \"\")\n",
    "c <- read.csv(paste0(\"freq_EUR.frq\"),sep = \"\")\n",
    "\n",
    "print(summary(a))\n",
    "print(summary(b))\n",
    "print(summary(c))\n",
    "\n",
    "exclude_list <- which(is.na(a$MAF) | is.na(b$MAF) | is.na(c$MAF) | a$MAF <= 0.05 | b$MAF <= 0.05 | c$MAF <= 0.05)\n",
    "\n",
    "print(length(exclude_list)/nrow(a))\n",
    "\n",
    "all_chr <- read.table(paste0(BED_File,\".bim\"), header=FALSE)\n",
    "samps <- sample((1:nrow(all_chr))[-exclude_list],900000,replace = FALSE)\n",
    "samps <- samps[order(samps)]\n",
    "all_chr <- all_chr[samps,2]\n",
    "write.table(all_chr,file = \"extract_snps\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "system(paste0(\"plink2 --bfile \",BED_File,\" --extract extract_snps --make-bed --out \",OUTPUT_PATH,\"/\",\"REGENIE_Step1\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "      '--input BED_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "      '--input BIM_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "      '--input FAM_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "      '--input Ancestry_File'=\"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv\",\n",
    "      '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/REGENIE_Prep_BED.R\",\n",
    "      '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "      check.names = FALSE\n",
    "    ))\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting REGENIE_Prep_BED.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE_Prep_BED.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Ancestry_File} ${BED_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://REGENIE_Prep_BED.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.8 KiB/  1.8 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.8 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp REGENIE_Prep_BED.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: regenie-pr--williamsjacr--250415-112618-55\n",
      "  job-name: regenie-prep-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12235351109116825389\n",
      "Launched job-id: regenie-pr--williamsjacr--250415-112618-55\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-pr--williamsjacr--250415-112618-55' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-pr--williamsjacr--250415-112618-55' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 1000 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script REGENIE_Prep_BED.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing REGENIE_Test.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE_Test.R\n",
    "rm(list = ls())\n",
    "\n",
    "print(1 + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "      '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/REGENIE_Test.R\",\n",
    "      '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "      check.names = FALSE\n",
    "    ))\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing REGENIE_Test.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE_Test.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://REGENIE_Test.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][   30.0 B/   30.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/30.0 B.                                       \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp REGENIE_Test.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: regenie-te--williamsjacr--250523-160030-74\n",
      "  job-name: regenie-test\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11227930139315377219\n",
      "Launched job-id: regenie-te--williamsjacr--250523-160030-74\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-te--williamsjacr--250523-160030-74' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-te--williamsjacr--250523-160030-74' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 100 \\\n",
    "  --min-ram 100 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script REGENIE_Test.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name        Task  Status                                      Last Update\r\n",
      "------------  ------  ------------------------------------------  --------------\r\n",
      "regenie-test       1  pulling image: docker pull: retry budge...  05-23 16:06:49\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-te--williamsjacr--250523-154923-89' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name        Task  Status                                   Last Update\r\n",
      "------------  ------  ---------------------------------------  --------------\r\n",
      "regenie-test       1  Pulling \"willja16/r_with_plink_regenie\"  05-23 16:01:19\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-te--williamsjacr--250523-160014-73' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name        Task  Status                           Last Update\r\n",
      "------------  ------  -------------------------------  --------------\r\n",
      "regenie-test       1  Pulling \"willja16/r_with_plink\"  05-23 16:01:19\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie-te--williamsjacr--250523-160030-74' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"logTG\",\"HDL\",\"TC\",\"Height\")){\n",
    "     tasks <- rbind(tasks, data.frame(\n",
    "  '--input Regenie_Step1_Bed'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/REGENIE_Step1.bed\",\n",
    "  '--input Regenie_Step1_Bim'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/REGENIE_Step1.bim\",\n",
    "  '--input Regenie_Step1_Fam'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/REGENIE_Step1.fam\",\n",
    "  '--input Regenie_Step2_Bed'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "  '--input Regenie_Step2_Bim'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "  '--input Regenie_Step2_Fam'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "  '--env anc'=anc,\n",
    "  '--env trait'=trait,\n",
    "  '--input Regenie_Train_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train_\",trait,\"_\",anc,\"_REGENIE.txt\"),\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "  check.names = FALSE\n",
    "))   \n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting REGENIE.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile REGENIE.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "regenie --step 1 --bed /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/REGENIE_Step1 --phenoFile ${Regenie_Train_File} --phenoColList ${trait} --covarFile ${Regenie_Train_File} --covarColList age,age2,sex,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 --qt --bsize 1000 --lowmem --lowmem-prefix regenie_step1_tmp --out regenie_step1_withsex &\n",
    "wait\n",
    "regenie --step 2 --bed /mnt/data/input/gs/fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr --phenoFile ${Regenie_Train_File} --phenoColList ${trait} --covarFile ${Regenie_Train_File} --covarColList age,age2,sex,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 --pred regenie_step1_withsex_pred.list --qt --bsize 400 --out ${OUTPUT_PATH}/regenie_step2_${anc}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: regenie--williamsjacr--250427-234426-19\n",
      "  job-name: regenie\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5130628365722022552\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12685230392961720833\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1553640902003377733\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10114656875896946847\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2348449403932750315\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9354960312708151806\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3715688646150788466\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3795155697146286595\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16281371663921910182\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16192095260364623036\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14661004121178686769\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17485031717009553070\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1967623286063097859\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/327760482951804624\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11869808876151920710\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11558810775159043775\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17902267762882538100\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1217365069406410441\n",
      "Launched job-id: regenie--williamsjacr--250427-234426-19\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie--williamsjacr--250427-234426-19' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'regenie--williamsjacr--250427-234426-19' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink_regenie \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script REGENIE.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing GWAS_SumStats.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats.R\n",
    "rm(list = ls())\n",
    "\n",
    "BED_file <- commandArgs(TRUE)[1]\n",
    "BED_file <- gsub(\".bed\",\"\",BED_file)\n",
    "print(BED_file)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[3]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "anc <- commandArgs(TRUE)[4]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[5]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "tmp <- read.delim(all_train_file)\n",
    "write.table(tmp[tmp$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == anc],],file = paste0(\"All_Train_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE)\n",
    "\n",
    "system(paste0(\"plink2 --bfile \",BED_file,\" --pheno All_Train_\",anc,\".txt --pheno-name \",trait,\" --linear --covar All_Train_\",anc,\".txt --covar-name age, age2, sex, PC1, PC2, PC3, PC4, PC5, PC6, PC7, PC8, PC9, PC10 --vif 999 --out \",OUTPUT_PATH,\"/\",trait,\"_sumstats_\",anc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "  for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "      '--input BED_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "      '--input BIM_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "      '--input FAM_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "      '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "      '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "      '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats.R\",\n",
    "      '--env anc'=anc,\n",
    "      '--env trait'=trait,\n",
    "      '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "      check.names = FALSE\n",
    "    ))\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing GWAS_SumStats.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${BED_File} ${all_phenotypes_file} ${all_train_file} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  905.0 B/  905.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/905.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--250424-173116-13\n",
      "  job-name: gwas-sumstats\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3214562358622228061\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17511041392178526707\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1214001968959801695\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4518281125617921996\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7051887284840326656\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10142665077198811106\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6126429204444064864\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6321005950925884043\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14889961924423328504\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10946661215398721144\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14979466789088587599\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6472251751245820480\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16013292871016644477\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9621027801065218490\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8768921485236746079\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14883738892659652685\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12889058732588816708\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12129646829671858356\n",
      "Launched job-id: gwas-sumst--williamsjacr--250424-173116-13\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250424-173116-13' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250424-173116-13' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean_Plink.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean_Plink.R\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "\n",
    "anc <- commandArgs(TRUE)[2]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[3]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(data.table)\n",
    "\n",
    "sumstats <- fread(sumstats)\n",
    "print(str(sumstats))\n",
    "sumstats <- as.data.frame(sumstats)\n",
    "colnames(sumstats) <- c(\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"PROVISIONAL_REF\",\"A1\",\"OMITTED\",\"A1_FREQ\",\"TEST\",\"OBS_CT\",\"BETA\",\"SE\",\"T_STAT\",\"P\",\"ERRCODE\")\n",
    "sumstats <- sumstats[sumstats$TEST == \"ADD\",]\n",
    "sumstats <- sumstats[,c(\"CHROM\",\"ID\",\"POS\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"A1\",\"OBS_CT\",\"BETA\",\"SE\",\"P\")\n",
    "write.table(sumstats, file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_Plink_GWAS_SumStats_Cleaned\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean_Plink.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean_Plink.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"EUR\"\n",
    "trait <- \"HDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AMR\",\"AFR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"Height\",\"TC\",\"logTG\")){\n",
    "     tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",trait,\"_sumstats_\",anc,\".\",trait,\".glm.linear\"),\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats_Clean_Plink.R\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats_Clean_Plink.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  793.0 B/  793.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/793.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats_Clean_Plink.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--250425-122607-63\n",
      "  job-name: gwas-sumstats-clean-plink\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14674126123918469510\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12452681274097805714\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14256211189955613424\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14878270159713398819\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7764722351286179164\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10541295791420624510\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4394802813209458800\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4845252920812637302\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1469963412370931263\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5354820749031751041\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9732277646725809268\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1632418665649549808\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6185408955732946971\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13731778908855475143\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5468435370668041078\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10461296564359068687\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11051572426884747324\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1533208424754907236\n",
      "Launched job-id: gwas-sumst--williamsjacr--250425-122607-63\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250425-122607-63' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250425-122607-63' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 40 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 60 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats_Clean_Plink.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_anc_bed.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_anc_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "all_File <- commandArgs(TRUE)[3]\n",
    "print(all_File)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[4]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "anc <- commandArgs(TRUE)[5]\n",
    "print(anc)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Tune$IID[Tune$IID %in% All$IID[All$ancestry == anc]])\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune_\",anc),intern = TRUE)\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Valid$IID[Valid$IID %in% All$IID[All$ancestry == anc]])\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_valid_\",anc),intern = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(anc in c(\"AFR\",\"EUR\",\"AMR\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--env anc'=anc,\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input all_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/tune_validation_anc_bed.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_anc_bed.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_anc_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${all_File} ${BED_Full_File} ${anc} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://tune_validation_anc_bed.R [Content-Type=application/octet-stream]...\n",
      "- [1/1 files][  1.0 KiB/  1.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.0 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp tune_validation_anc_bed.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: tune-valid--williamsjacr--250417-112318-03\n",
      "  job-name: tune-validation-anc-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7194155009525131398\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/984999540801392388\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12867490442755241193\n",
      "Launched job-id: tune-valid--williamsjacr--250417-112318-03\n",
      "3 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--250417-112318-03' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--250417-112318-03' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 700 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script tune_validation_anc_bed.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting JointPRS_MUSSEL_PROSPER_BED_Files.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_MUSSEL_PROSPER_BED_Files.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "library(dplyr)\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "all_File <- commandArgs(TRUE)[3]\n",
    "print(all_File)\n",
    "\n",
    "snp_info_rsids <- commandArgs(TRUE)[4]\n",
    "print(snp_info_rsids)\n",
    "\n",
    "snp_info_extractlist <- commandArgs(TRUE)[5]\n",
    "print(snp_info_extractlist)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[6]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "all_chr_bim <- read.delim(paste0(BED_Full_File,\".bim\"), header=FALSE)\n",
    "all_chr_bim <- as.data.frame(all_chr_bim)\n",
    "colnames(all_chr_bim) <- c(\"CHR\",\"SNP\",\"IDK\",\"BP\",\"A1\",\"A2\")\n",
    "\n",
    "all_chr_bim$snp <- all_chr_bim$SNP\n",
    "all_chr_bim$rs_ID <- all_chr_bim$SNP\n",
    "\n",
    "all_chr_bim$SNP <- toupper(all_chr_bim$SNP)\n",
    "all_chr_bim$SNP <- gsub(\" \",\"\",all_chr_bim$SNP)\n",
    "\n",
    "SNP_GRCh37_38_match_update <- readRDS(snp_info_rsids)\n",
    "\n",
    "system(paste0(\"rm \",snp_info_rsids))\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele1_38,\":\",SNP_GRCh37_38_match_update$allele2_38)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele2_38,\":\",SNP_GRCh37_38_match_update$allele1_38)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- toupper(SNP_GRCh37_38_match_update$unique_id1)\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id1)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- toupper(SNP_GRCh37_38_match_update$unique_id2)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id2)\n",
    "\n",
    "all_chr_bim <- left_join(all_chr_bim,SNP_GRCh37_38_match_update[,c(\"unique_id1\",\"rsid\")],by = c(\"SNP\"=\"unique_id1\"))\n",
    "all_chr_bim$rs_ID[!is.na(all_chr_bim$rsid)] <- all_chr_bim$rsid[!is.na(all_chr_bim$rsid)]\n",
    "    \n",
    "all_chr_bim <- subset(all_chr_bim,select = -c(rsid))\n",
    "    \n",
    "all_chr_bim <- left_join(all_chr_bim,SNP_GRCh37_38_match_update[,c(\"unique_id2\",\"rsid\")],by = c(\"SNP\"=\"unique_id2\"))\n",
    "all_chr_bim$rs_ID[!is.na(all_chr_bim$rsid)] <- all_chr_bim$rsid[!is.na(all_chr_bim$rsid)]\n",
    "\n",
    "snp_info_extractlist <- read.csv(snp_info_extractlist)\n",
    "extract_list <- all_chr_bim$snp[all_chr_bim$rs_ID %in% snp_info_extractlist$SNP]\n",
    "write.table(extract_list,\"extract_list.txt\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "# Extract Needed SNPs in Full File\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --extract extract_list.txt --make-bed --out \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER\"),intern = TRUE)\n",
    "\n",
    "system(paste0(\"rm \",BED_Full_File,\".bed\"))\n",
    "system(paste0(\"rm \",BED_Full_File,\".bim\"))\n",
    "system(paste0(\"rm \",BED_Full_File,\".fam\"))\n",
    "\n",
    "write.table(all_chr_bim[,c(\"snp\",\"rs_ID\")],file = \"snp_rename.txt\",row.names = F,col.names = F,quote=F)\n",
    "\n",
    "# Rename SNPs to rsids\n",
    "system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --update-name snp_rename.txt 2 1 --make-bed --out \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER\"))\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Tune$IID)\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "## Full Tune\n",
    "system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune_JointPRS_MUSSEL_PROSPER\"),intern = TRUE)\n",
    "\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "# Anc Specific Tune\n",
    "for(anc in c(\"AFR\",\"EUR\",\"AMR\")){\n",
    "    Keep <- data.frame(FID = 0,IID = Tune$IID[Tune$IID %in% All$IID[All$ancestry == anc]])\n",
    "    write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "    system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune_JointPRS_MUSSEL_PROSPER_\",anc),intern = TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Valid$IID)\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "## Full Valid\n",
    "system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_valid_JointPRS_MUSSEL_PROSPER\"),intern = TRUE)\n",
    "\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "# Anc Specific Tune\n",
    "for(anc in c(\"AFR\",\"EUR\",\"AMR\")){\n",
    "    Keep <- data.frame(FID = 0,IID = Valid$IID[Valid$IID %in% All$IID[All$ancestry == anc]])\n",
    "    write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "    system(paste0(\"plink --bfile \",OUTPUT_PATH,\"/all_JointPRS_MUSSEL_PROSPER --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_Valid_JointPRS_MUSSEL_PROSPER_\",anc),intern = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input all_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input snp_info_rsids'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\",\n",
    "    '--input snp_info_extractlist'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_Info_JointPRS_PROSPER_MUSSEL.csv\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/JointPRS_MUSSEL_PROSPER_BED_Files.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED\",\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting JointPRS_MUSSEL_PROSPER_BED_Files.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_MUSSEL_PROSPER_BED_Files.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${all_File} ${snp_info_rsids} ${snp_info_extractlist} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://JointPRS_MUSSEL_PROSPER_BED_Files.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  4.5 KiB/  4.5 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/4.5 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp JointPRS_MUSSEL_PROSPER_BED_Files.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: jointprs-m--williamsjacr--250417-151205-37\n",
      "  job-name: jointprs-mussel-prosper-bed-files\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2830465926782299950\n",
      "Launched job-id: jointprs-m--williamsjacr--250417-151205-37\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-m--williamsjacr--250417-151205-37' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-m--williamsjacr--250417-151205-37' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 700 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 80 \\\n",
    "  --timeout \"168h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script JointPRS_MUSSEL_PROSPER_BED_Files.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_bed.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[3]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Tune$IID)\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_tune\"),intern = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "\n",
    "Keep <- data.frame(FID = 0,IID = Valid$IID)\n",
    "\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_valid\"),intern = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/tune_validation_bed.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_bed.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://tune_validation_bed.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  823.0 B/  823.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/823.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp tune_validation_bed.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: tune-valid--williamsjacr--250417-112549-30\n",
      "  job-name: tune-validation-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5390226490247056300\n",
      "Launched job-id: tune-valid--williamsjacr--250417-112549-30\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--250417-112549-30' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--250417-112549-30' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 700 \\\n",
    "  --boot-disk-size 40 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script tune_validation_bed.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_phenotypes.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_phenotypes.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "Tune_File <- commandArgs(TRUE)[1]\n",
    "print(Tune_File)\n",
    "\n",
    "Valid_File <- commandArgs(TRUE)[2]\n",
    "print(Valid_File)\n",
    "\n",
    "all_File <- commandArgs(TRUE)[3]\n",
    "print(all_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Tune <- read.delim(Tune_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "covar_tmp <- data.frame(FID = Tune$FID,IID = Tune$IID,age = Tune$age,age2 = Tune$age2,sex = Tune$sex,\n",
    "                        PC1 = Tune$PC1,PC2 = Tune$PC2,PC3 = Tune$PC3,PC4 = Tune$PC4,PC5 = Tune$PC5,\n",
    "                        PC6 = Tune$PC6,PC7 = Tune$PC7,PC8 = Tune$PC8,PC9 = Tune$PC9,PC10 = Tune$PC10)\n",
    "\n",
    "write.table(covar_tmp,file = paste0(OUTPUT_PATH,\"/full_tune_covar.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    covar_tmp_anc <- covar_tmp[covar_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "    write.table(covar_tmp_anc,file = paste0(OUTPUT_PATH,\"/full_tune_covar_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "}\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    fam_tmp <- data.frame(FID = Tune$FID,IID = Tune$IID,Y = Tune[,trait])\n",
    "    write.table(fam_tmp,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_tune_y.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "    for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "        fam_tmp_anc <- fam_tmp[fam_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "        write.table(fam_tmp_anc,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_tune_y_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Valid <- read.delim(Valid_File)\n",
    "All <- read.csv(all_File)\n",
    "\n",
    "covar_tmp <- data.frame(FID = Valid$FID,IID = Valid$IID,age = Valid$age,age2 = Valid$age2,sex = Valid$sex,\n",
    "                        PC1 = Valid$PC1,PC2 = Valid$PC2,PC3 = Valid$PC3,PC4 = Valid$PC4,PC5 = Valid$PC5,\n",
    "                        PC6 = Valid$PC6,PC7 = Valid$PC7,PC8 = Valid$PC8,PC9 = Valid$PC9,PC10 = Valid$PC10)\n",
    "\n",
    "write.table(covar_tmp,file = paste0(OUTPUT_PATH,\"/full_valid_covar.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    covar_tmp_anc <- covar_tmp[covar_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "    write.table(covar_tmp_anc,file = paste0(OUTPUT_PATH,\"/full_valid_covar_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "}\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    fam_tmp <- data.frame(FID = Valid$FID,IID = Valid$IID,Y = Valid[,trait])\n",
    "    write.table(fam_tmp,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_valid_y.txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    \n",
    "    for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "        fam_tmp_anc <- fam_tmp[fam_tmp$IID %in% All$IID[All$ancestry == anc],]\n",
    "        write.table(fam_tmp_anc,file = paste0(OUTPUT_PATH,\"/\",trait,\"_full_valid_y_\",anc,\".txt\"),sep = '\\t',row.names = FALSE,quote = FALSE,col.names = FALSE)\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Tune_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Valid_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input all_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/tune_validation_phenotypes.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes\",\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing tune_validation_phenotypes.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile tune_validation_phenotypes.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Tune_File} ${Valid_File} ${all_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://tune_validation_phenotypes.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.8 KiB/  2.8 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.8 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp tune_validation_phenotypes.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: tune-valid--williamsjacr--240722-175637-83\n",
      "  job-name: tune-validation-phenotypes\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16970955615058126312\n",
      "Launched job-id: tune-valid--williamsjacr--240722-175637-83\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240722-175637-83' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'tune-valid--williamsjacr--240722-175637-83' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 40 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 8 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script tune_validation_phenotypes.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing reference_bed.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile reference_bed.R\n",
    "\n",
    "rm(list = ls())\n",
    "\n",
    "pheno_train <- commandArgs(TRUE)[1]\n",
    "print(pheno_train)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[3]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "pheno_train <- read.delim(pheno_train)\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\",\"EUR\")){\n",
    "    tmp <- pheno_train[pheno_train$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == anc],]\n",
    "    Keep <- data.frame(FID = 0,IID = sample(tmp$IID,3000))\n",
    "    write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "    system(paste0(\"plink --bfile \",BED_Full_File,\" --keep Keep.txt --make-bed --out \",OUTPUT_PATH,\"/all_chr_reference_\",anc),intern = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input pheno_train'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/reference_bed.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants\",\n",
    "    # '--output out_file'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/test\",chrom,\".Rdata\"),\n",
    "    check.names = FALSE\n",
    "  ))\n",
    "\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing reference_bed.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile reference_bed.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${pheno_train} ${all_phenotypes_file} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://reference_bed.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  799.0 B/  799.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/799.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp reference_bed.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: reference---williamsjacr--250418-123556-81\n",
      "  job-name: reference-bed\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16501531450550937699\n",
      "Launched job-id: reference---williamsjacr--250418-123556-81\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'reference---williamsjacr--250418-123556-81' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'reference---williamsjacr--250418-123556-81' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 700 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script reference_bed.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean.R\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "\n",
    "anc <- commandArgs(TRUE)[2]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[3]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "sumstats <- read.csv(sumstats, sep=\"\")\n",
    "colnames(sumstats) <- c(\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"A1_FREQ\",\"N\",\"TEST\",\"BETA\",\"SE\",\"CHISQ\",\"LOG10P\",\"EXTRA\")\n",
    "sumstats$P <- 10^(-1*sumstats$LOG10P)\n",
    "sumstats <- sumstats[,c(\"CHROM\",\"ID\",\"POS\",\"REF\",\"ALT\",\"N\",\"BETA\",\"SE\",\"P\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\")\n",
    "write.table(sumstats, file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"EUR\"\n",
    "trait <- \"HDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AMR\",\"AFR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"Height\",\"TC\",\"logTG\")){\n",
    "     tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/regenie_step2_\",anc,\"_\",trait,\".regenie\"),\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats_Clean.R\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats_Clean.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  662.0 B/  662.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/662.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats_Clean.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--250428-173207-76\n",
      "  job-name: gwas-sumstats-clean\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7034685886252727257\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4082315975050440483\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7938481039455991176\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15274961768330524042\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11793126664326088779\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3982309362940487567\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10594059716017372268\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3991644517047102351\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1008738729791118479\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18411546459672876864\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13243142025854682127\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6320383088505898924\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6745302760374208265\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5828754584733337655\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14171595120965413852\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15631755643604011817\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7771514375847259507\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15875377060250257645\n",
      "Launched job-id: gwas-sumst--williamsjacr--250428-173207-76\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250428-173207-76' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250428-173207-76' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 10 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 20 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats_Clean.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean_QQPlots.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean_QQPlots.R\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "\n",
    "anc <- commandArgs(TRUE)[2]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[3]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "sumstats <- read.csv(sumstats, sep=\"\")\n",
    "colnames(sumstats) <- c(\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"A1_FREQ\",\"N\",\"TEST\",\"BETA\",\"SE\",\"CHISQ\",\"LOG10P\",\"EXTRA\")\n",
    "sumstats$P <- 10^(-1*sumstats$LOG10P)\n",
    "sumstats <- sumstats[,c(\"CHROM\",\"ID\",\"POS\",\"A1_FREQ\",\"P\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"A1_FREQ\",\"P\")\n",
    "write.table(sumstats, file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned_QQPlots\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting GWAS_SumStats_Clean_QQPlots.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile GWAS_SumStats_Clean_QQPlots.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"EUR\"\n",
    "trait <- \"HDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AMR\",\"AFR\")){\n",
    "    for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"Height\",\"TC\",\"logTG\")){\n",
    "     tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/regenie_step2_\",anc,\"_\",trait,\".regenie\"),\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/GWAS_SumStats_Clean_QQPlots.R\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://GWAS_SumStats_Clean_QQPlots.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  628.0 B/  628.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/628.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp GWAS_SumStats_Clean_QQPlots.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: gwas-sumst--williamsjacr--250428-173226-64\n",
      "  job-name: gwas-sumstats-clean-qqplots\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2403754769293709418\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13781852294473506066\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12577162130862398103\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10308808330836817463\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8410554366338355886\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4842582575030511619\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2512501573097638363\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5789696100044841828\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10582451307309159188\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6482660427374822750\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5868419785274217146\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18074007484370393753\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1966499168238485339\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8165884351610526811\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/125768957691495756\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17627801425712592314\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14107775627323434360\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11046213553488011657\n",
      "Launched job-id: gwas-sumst--williamsjacr--250428-173226-64\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250428-173226-64' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'gwas-sumst--williamsjacr--250428-173226-64' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 10 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 10 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script GWAS_SumStats_Clean_QQPlots.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Add_RSIDs_Sumstats.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Add_RSIDs_Sumstats.R\n",
    "rm(list = ls())\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "print(sumstats)\n",
    "\n",
    "snp_info <- commandArgs(TRUE)[2]\n",
    "print(snp_info)\n",
    "\n",
    "anc <- commandArgs(TRUE)[3]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[4]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "SNP_GRCh37_38_match_update <- readRDS(snp_info)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele1_38,\":\",SNP_GRCh37_38_match_update$allele2_38)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele2_38,\":\",SNP_GRCh37_38_match_update$allele1_38)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- toupper(SNP_GRCh37_38_match_update$unique_id1)\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id1)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- toupper(SNP_GRCh37_38_match_update$unique_id2)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id2)\n",
    "\n",
    "\n",
    "sumstats <- fread(sumstats)\n",
    "    \n",
    "sumstats <- as.data.frame(sumstats)\n",
    "sumstats <- sumstats[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "    \n",
    "sumstats$SNP <- toupper(sumstats$SNP)\n",
    "sumstats$SNP <- gsub(\" \",\"\",sumstats$SNP)\n",
    "    \n",
    "sumstats <- left_join(sumstats,SNP_GRCh37_38_match_update[,c(\"unique_id1\",\"rsid\")],by = c(\"SNP\"=\"unique_id1\"))\n",
    "sumstats$rs_ID[!is.na(sumstats$rsid)] <- sumstats$rsid[!is.na(sumstats$rsid)]\n",
    "    \n",
    "sumstats <- subset(sumstats,select = -c(rsid))\n",
    "    \n",
    "sumstats <- left_join(sumstats,SNP_GRCh37_38_match_update[,c(\"unique_id2\",\"rsid\")],by = c(\"SNP\"=\"unique_id2\"))\n",
    "sumstats$rs_ID[!is.na(sumstats$rsid)] <- sumstats$rsid[!is.na(sumstats$rsid)]\n",
    "    \n",
    "sumstats <- sumstats[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "    \n",
    "sumstats$SNP <- sumstats$rs_ID\n",
    "    \n",
    "doubles <- names(which(table(sumstats$rs_ID) > 1))\n",
    "    \n",
    "sumstats <- sumstats[!(sumstats$SNP %in% doubles),]\n",
    "    \n",
    "write.csv(sumstats,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_RSIDs.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Add_RSIDs_Sumstats.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Add_RSIDs_Sumstats.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${snp_info} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input snp_info'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Add_RSIDs_Sumstats.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Add_RSIDs_Sumstats.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.3 KiB/  2.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Add_RSIDs_Sumstats.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: add-rsids---williamsjacr--250428-174430-30\n",
      "  job-name: add-rsids-sumstats\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6998726100006922535\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4293973618065607811\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3001410113428019042\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15389202591089947051\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14367914056569811633\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15698518150691807952\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11655765565357011582\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3650844991244160939\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/641746769860945503\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2546523258750150982\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10837516147411403051\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17331868544448187863\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3881060599115021434\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4253637643572861887\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8954228243063715673\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11663461168700474763\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17810486694260462307\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9230077776326058325\n",
      "Launched job-id: add-rsids---williamsjacr--250428-174430-30\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'add-rsids---williamsjacr--250428-174430-30' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'add-rsids---williamsjacr--250428-174430-30' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 10 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 80 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Add_RSIDs_Sumstats.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_SumStats_Format.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_SumStats_Format.R\n",
    "rm(list = ls())\n",
    "\n",
    "sumstats <- commandArgs(TRUE)[1]\n",
    "print(sumstats)\n",
    "\n",
    "ref_bim <- commandArgs(TRUE)[2]\n",
    "print(ref_bim)\n",
    "\n",
    "anc <- commandArgs(TRUE)[3]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[4]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "sumstats <- fread(sumstats)\n",
    "    \n",
    "sumstats <- as.data.frame(sumstats)\n",
    "colnames(sumstats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "\n",
    "sumstats <- sumstats[,c(\"rs_ID\",\"CHR\",\"ALT\",\"REF\",\"BETA\",\"SE\",\"OBS_CT\")]\n",
    "colnames(sumstats) <- c(\"rsid\",\"chr\",\"a1\",\"a0\",\"beta\",\"beta_se\",\"n_eff\")\n",
    "\n",
    "sumstats$n_eff <- as.numeric(sumstats$n_eff)\n",
    "\n",
    "ref_bim <- fread(ref_bim)\n",
    "ref_bim <- as.data.frame(ref_bim)\n",
    "\n",
    "sumstats <- sumstats[sumstats$rsid %in% ref_bim[,2],]\n",
    "\n",
    "ref_bim <- ref_bim[ref_bim[,2] %in% sumstats$rsid,]\n",
    "\n",
    "order_vec <- match(sumstats$rsid,ref_bim[,2])\n",
    "\n",
    "ref_bim <- ref_bim[order_vec,]\n",
    "\n",
    "a <- which((sumstats$a0 != ref_bim[,6]) & (sumstats$a1 != ref_bim[,6]))\n",
    "\n",
    "b <- which((sumstats$a0 != ref_bim[,5]) & (sumstats$a1 != ref_bim[,5]))\n",
    "\n",
    "sumstats$a0[a] <- ref_bim$V5[a]\n",
    "sumstats$a1[a] <- ref_bim$V6[a]\n",
    "\n",
    "b <- which((sumstats$a0 != ref_bim[,5]) & (sumstats$a1 != ref_bim[,5]))\n",
    "\n",
    "if(length(b) > 0){\n",
    "  sumstats$a0[b] <- ref_bim$V5[b]\n",
    "  sumstats$a1[b] <- ref_bim$V6[b]\n",
    "}\n",
    "\n",
    "print(summary(sumstats$n_eff))\n",
    "\n",
    "sumstats <- sumstats[sumstats$n_eff > (median(sumstats$n_eff)*.9),]\n",
    "\n",
    "write.csv(sumstats,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_SumStats_Format.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_SumStats_Format.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${sumstats} ${ref_bim} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"EUR\",\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_RSIDs.csv\"),\n",
    "    '--input ref_bim'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER/ref_bim.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_SumStats_Format.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_SumStats_Format.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.4 KiB/  1.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_SumStats_Format.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-su--williamsjacr--250428-184627-04\n",
      "  job-name: prosper-sumstats-format\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11674139616768273866\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13484182636830392265\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3962884142189636599\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13392857554919715171\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12029767872049378365\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/895516038504869640\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/93042629627218191\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2397665020924857596\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11471171650775769630\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5819963965045770968\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6013071152884751414\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7074770186856007126\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15998534996542236713\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9274313101758974667\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15119132157433336279\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10410743032992554110\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1348849601247607995\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10468480096631514927\n",
      "Launched job-id: prosper-su--williamsjacr--250428-184627-04\n",
      "18 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-su--williamsjacr--250428-184627-04' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-su--williamsjacr--250428-184627-04' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 10 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 8 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_SumStats_Format.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting CT_SLEB_dimCT_prseb.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_dimCT_prseb.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(devtools)\n",
    "library(CTSLEB)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "BED_Tune_File <- commandArgs(TRUE)[1]\n",
    "BED_Tune_File <- gsub(\".bed\",\"\",BED_Tune_File)\n",
    "print(BED_Tune_File)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[2]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "#EUR\n",
    "BED_Ref_pop1_File <- commandArgs(TRUE)[3]\n",
    "BED_Ref_pop1_File <- gsub(\".bed\",\"\",BED_Ref_pop1_File)\n",
    "print(BED_Ref_pop1_File)\n",
    "\n",
    "#ANC\n",
    "BED_Ref_pop2_File <- commandArgs(TRUE)[4]\n",
    "BED_Ref_pop2_File <- gsub(\".bed\",\"\",BED_Ref_pop2_File)\n",
    "print(BED_Ref_pop2_File)\n",
    "\n",
    "pop1_sumstats <- commandArgs(TRUE)[5]\n",
    "print(pop1_sumstats)\n",
    "\n",
    "pop2_sumstats <- commandArgs(TRUE)[6]\n",
    "print(pop2_sumstats)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[7]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[8]\n",
    "print(all_tune_file)\n",
    "\n",
    "anc <- commandArgs(TRUE)[9]\n",
    "print(anc)\n",
    "\n",
    "trait <- commandArgs(TRUE)[10]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[11]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "system(\"mkdir CTSLEB_TestCode\")\n",
    "\n",
    "sum_pop1 <- fread(pop1_sumstats)\n",
    "\n",
    "sum_pop1 <- as.data.frame(sum_pop1)\n",
    "sum_pop1 <- sum_pop1[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(sum_pop1) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "sum_pop1 <- sum_pop1[,c(\"CHR\",\"SNP\",\"BP\",\"ALT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "colnames(sum_pop1) <- c(\"CHR\",\"SNP\",\"BP\",\"A1\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "\n",
    "sum_pop1$P <- as.numeric(sum_pop1$P)\n",
    "sum_pop1 <- sum_pop1[which(sum_pop1$P < 0.1),]\n",
    "\n",
    "sum_pop2 <- fread(pop2_sumstats)\n",
    "\n",
    "sum_pop2 <- as.data.frame(sum_pop2)\n",
    "sum_pop2 <- sum_pop2[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(sum_pop2) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "sum_pop2 <- sum_pop2[,c(\"CHR\",\"SNP\",\"BP\",\"ALT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "colnames(sum_pop2) <- c(\"CHR\",\"SNP\",\"BP\",\"A1\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "\n",
    "sum_pop2$P <- as.numeric(sum_pop2$P)\n",
    "sum_pop2 <- sum_pop2[which(sum_pop2$P < 0.1),]\n",
    "\n",
    "print(nrow(sum_pop1));print(nrow(sum_pop2))\n",
    "\n",
    "PRS_farm <- SetParamsFarm(plink19_exec = \"plink\",plink2_exec = \"plink2\",mem = 80490,pthres = c(5E-08,5E-07,5E-06,5E-05,5E-04,5E-03,0.05,0.1))\n",
    "\n",
    "prs_mat <- dimCT(results_dir = \"CTSLEB_TestCode/\",sum_target = sum_pop2,sum_ref = sum_pop1,\n",
    "                 ref_plink = BED_Ref_pop1_File,\n",
    "                 target_plink = BED_Ref_pop2_File,\n",
    "                 test_target_plink = BED_Tune_File,\n",
    "                 out_prefix = \"dimCT\",\n",
    "                 params_farm = PRS_farm)\n",
    "\n",
    "# save(prs_mat,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_prs_mat_dimCT.RData\"))\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,prs_mat,by = \"IID\")\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "\n",
    "n.total.prs <- length(pthres)^2*length(r2_vec)*length(wc_base_vec)\n",
    "prs_r2_vec_test <- rep(0,n.total.prs)\n",
    "indices <- (ncol(pheno_tuning) - n.total.prs + 1):ncol(pheno_tuning)\n",
    "\n",
    "for(p_ind in 1:n.total.prs){\n",
    "  prs <- pheno_tuning[!is.na(pheno_tuning[,trait]),indices[p_ind]]\n",
    "  model <- lm(model.null$residual~prs)\n",
    "  prs_r2_vec_test[p_ind] <- summary(model)$r.square\n",
    "}\n",
    "\n",
    "max_ind <- which.max(prs_r2_vec_test)\n",
    "\n",
    "print(colnames(prs_mat)[indices[max_ind]])\n",
    "\n",
    "best_snps <- colnames(prs_mat)[indices[max_ind]]\n",
    "#calculate eb effect using EB coefficients\n",
    "prs_mat_eb <- CalculateEBEffectSize(bfile = BED_Full_File,\n",
    "                                    snp_ind = best_snps,\n",
    "                                    plink_list = plink_list,\n",
    "                                    out_prefix = \"EB\",\n",
    "                                    results_dir = \"CTSLEB_TestCode/\",\n",
    "                                    params_farm = PRS_farm)\n",
    "\n",
    "\n",
    "\n",
    "score_full <- matrix(NA,ncol = ncol(prs_mat_eb) - 2,nrow = nrow(unique_infor_post))\n",
    "names_score_full <- NULL\n",
    "count <- 1\n",
    "for(i in 1:length(pthres)){\n",
    "  for(j in 1:length(pthres)){\n",
    "    for(k in 1:24){\n",
    "      tmp <- matrix(0,nrow = nrow(unique_infor_post),ncol = 1)\n",
    "      tmp[which(unique_infor_post$P < pthres[j] | unique_infor_post$P_ref < pthres[i]),1] <- plink_list_eb$scores_eb[which(unique_infor_post$P < pthres[j] | unique_infor_post$P_ref < pthres[i]),k + 2]\n",
    "      names_score_full <- c(names_score_full,paste0(colnames(plink_list_eb$scores_eb)[k + 2],\"_p_other_\",pthres[i],\"_p_tar_\",pthres[j]))\n",
    "      score_full[,count] <- tmp[,1]\n",
    "      count <- count + 1\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "final <- data.frame(SNP = unique_infor_post$SNP,A1 = unique_infor_post$A1,score_full)\n",
    "colnames(final) <- c(\"SNP\",\"A1\",names_score_full)\n",
    "\n",
    "\n",
    "save(final,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_betas_eb.RData\"))\n",
    "save(prs_mat_eb,file = paste0(OUTPUT_PATH,\"/\",anc,\"_\",trait,\"_prs_mat_eb.RData\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting CT_SLEB_dimCT_prseb.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_dimCT_prseb.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${BED_Tune_File} ${BED_Full_File} ${BED_Ref_pop1_File} ${BED_Ref_pop2_File} ${pop1_sumstats} ${pop2_sumstats} ${all_phenotypes_file} ${all_tune_file} ${anc} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "anc <- \"AFR\"\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(anc in c(\"AFR\",\"AMR\")){\n",
    "   for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input BED_Tune_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles/all_tune_\",anc,\".bed\"),\n",
    "    '--input BIM_Tune_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles/all_tune_\",anc,\".bim\"),\n",
    "    '--input FAM_Tune_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_BedFiles/all_tune_\",anc,\".fam\"),\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input BED_Ref_pop1_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",\"EUR\",\".bed\"),\n",
    "    '--input BIM_Ref_pop1_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",\"EUR\",\".bim\"),\n",
    "    '--input FAM_Ref_pop1_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",\"EUR\",\".fam\"),\n",
    "    '--input BED_Ref_pop2_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",anc,\".bed\"),\n",
    "    '--input BIM_Ref_pop2_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",anc,\".bim\"),\n",
    "    '--input FAM_Ref_pop2_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr_reference_\",anc,\".fam\"),\n",
    "    '--input pop1_sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/EUR_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input pop2_sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",anc,\"_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/CT_SLEB_dimCT_prseb.R\",\n",
    "    '--env anc'=anc,\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://CT_SLEB_dimCT_prseb.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  4.5 KiB/  4.5 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/4.5 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp CT_SLEB_dimCT_prseb.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: ct-sleb-di--williamsjacr--250428-174505-06\n",
      "  job-name: ct-sleb-dimct-prseb\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12444147214727260830\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8856746294784154544\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15219274510332855362\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3476450934576849204\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1798407164403620421\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/752621326411423424\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12262684282836058048\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17686097527311174742\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12747996702934129425\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3710844933542487776\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14540941010227904456\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2238258436553127586\n",
      "Launched job-id: ct-sleb-di--williamsjacr--250428-174505-06\n",
      "12 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-di--williamsjacr--250428-174505-06' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-di--williamsjacr--250428-174505-06' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 800 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 30 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script CT_SLEB_dimCT_prseb.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_LASSOSUM.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_LASSOSUM.sh\n",
    "#!/bin/bash\n",
    "\n",
    "path_plink='plink2'\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "tuning_bed_EUR=${tuning_bed_EUR%??}\n",
    "tuning_bed_AFR=${tuning_bed_AFR%??}\n",
    "tuning_bed_AMR=${tuning_bed_AMR%??}\n",
    "\n",
    "valid_bed_EUR=${valid_bed_EUR%??}\n",
    "valid_bed_AFR=${valid_bed_AFR%??}\n",
    "valid_bed_AMR=${valid_bed_AMR%??}\n",
    "\n",
    "\n",
    "Rscript ${PROSPER_Scripts}/scripts/lassosum2.R \\\n",
    "--PATH_package ${PROSPER_Scripts} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--PATH_plink ${path_plink} \\\n",
    "--FILE_sst ${EUR_Sumstats},${AFR_Sumstats},${AMR_Sumstats} \\\n",
    "--pop EUR,AFR,AMR \\\n",
    "--chrom 1-22 \\\n",
    "--bfile_tuning ${tuning_bed_EUR},${tuning_bed_AFR},${tuning_bed_AMR} \\\n",
    "--pheno_tuning ${tuning_y_EUR},${tuning_y_AFR},${tuning_y_AMR} \\\n",
    "--covar_tuning ${tuning_covar_EUR},${tuning_covar_AFR},${tuning_covar_AMR} \\\n",
    "--bfile_testing ${valid_bed_EUR},${valid_bed_AFR},${valid_bed_AMR} \\\n",
    "--pheno_testing ${valid_y_EUR},${valid_y_AFR},${valid_y_AMR} \\\n",
    "--covar_testing ${valid_covar_EUR},${valid_covar_AFR},${valid_covar_AMR} \\\n",
    "--testing TRUE \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input EUR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/EUR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AFR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AFR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AMR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AMR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input-recursive PROSPER_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER\",\n",
    "        '--input tuning_bed_EUR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_EUR.*\",\n",
    "        '--input tuning_bed_AMR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AMR.*\",\n",
    "        '--input tuning_bed_AFR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AFR.*\",\n",
    "        '--input valid_bed_EUR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_Valid_JointPRS_MUSSEL_PROSPER_EUR.*\",\n",
    "        '--input valid_bed_AMR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_Valid_JointPRS_MUSSEL_PROSPER_AMR.*\",\n",
    "        '--input valid_bed_AFR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_Valid_JointPRS_MUSSEL_PROSPER_AFR.*\",\n",
    "        '--input tuning_y_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y_EUR.txt\"),\n",
    "        '--input tuning_y_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y_AMR.txt\"),\n",
    "        '--input tuning_y_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y_AFR.txt\"),\n",
    "        '--input valid_y_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y_EUR.txt\"),\n",
    "        '--input valid_y_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y_AMR.txt\"),\n",
    "        '--input valid_y_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y_AFR.txt\"),\n",
    "        '--input tuning_covar_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar_EUR.txt\"),\n",
    "        '--input tuning_covar_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar_AMR.txt\"),\n",
    "        '--input tuning_covar_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar_AFR.txt\"),\n",
    "        '--input valid_covar_EUR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar_EUR.txt\"),\n",
    "        '--input valid_covar_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar_AMR.txt\"),\n",
    "        '--input valid_covar_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar_AFR.txt\"),\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-la--williamsjacr--250428-190015-72\n",
      "  job-name: prosper-lassosum\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13383691184684028660\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14622073818099952684\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/96705514830176920\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9905529682641647141\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/340593287056854147\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15735860106761377480\n",
      "Launched job-id: prosper-la--williamsjacr--250428-190015-72\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-la--williamsjacr--250428-190015-72' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-la--williamsjacr--250428-190015-72' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 60 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_LASSOSUM.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting MUSSEL_LDPred2.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile MUSSEL_LDPred2.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "tuning_bed_EUR=${tuning_bed_EUR%??}\n",
    "tuning_bed_AFR=${tuning_bed_AFR%??}\n",
    "tuning_bed_AMR=${tuning_bed_AMR%??}\n",
    "\n",
    "Rscript ${MUSSEL_Scripts}/R/LDpred2.R \\\n",
    "--PATH_package ${MUSSEL_Scripts} \\\n",
    "--PATH_ref ${MUSSEL_LD} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--FILE_sst ${EUR_Sumstats},${AFR_Sumstats},${AMR_Sumstats} \\\n",
    "--pop EUR,AFR,AMR \\\n",
    "--bfile_tuning ${tuning_bed_EUR},${tuning_bed_AFR},${tuning_bed_AMR} \\\n",
    "--verbose 2 \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input EUR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/EUR_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),\n",
    "        '--input AFR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/AFR_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),\n",
    "        '--input AMR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/AMR_\",trait,\"_GWAS_SumStats_MUSSEL.csv\"),\n",
    "        '--input-recursive MUSSEL_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/MUSSEL\",\n",
    "        '--input-recursive MUSSEL_LD'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/MUSSEL/MUSSEL_LD\",\n",
    "        '--input tuning_bed_EUR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_EUR.*\",\n",
    "        '--input tuning_bed_AMR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AMR.*\",\n",
    "        '--input tuning_bed_AFR'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER_AFR.*\",\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/MUSSEL/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: mussel-ldp--williamsjacr--240722-180603-76\n",
      "  job-name: mussel-ldpred2\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1876996133311691452\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2094648451624967140\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13690489926848506048\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5508878527126587761\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16970031399968614553\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5105032732639826203\n",
      "Launched job-id: mussel-ldp--williamsjacr--240722-180603-76\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-ldp--williamsjacr--240722-180603-76' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'mussel-ldp--williamsjacr--240722-180603-76' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script MUSSEL_LDPred2.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Prosper.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Prosper.sh\n",
    "\n",
    "path_plink='plink2'\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${PROSPER_Scripts}/scripts/PROSPER.R \\\n",
    "--PATH_package ${PROSPER_Scripts} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--FILE_sst ${EUR_Sumstats},${AFR_Sumstats},${AMR_Sumstats} \\\n",
    "--pop EUR,AFR,AMR \\\n",
    "--lassosum_param ${EUR_Optim_Param},${AFR_Optim_Param},${AMR_Optim_Param} \\\n",
    "--chrom 1-22 \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input EUR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/EUR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AFR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AFR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input AMR_Sumstats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/AMR_\",trait,\"_GWAS_SumStats_PROSPER.csv\"),\n",
    "        '--input EUR_Optim_Param'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/EUR/optimal_param.txt\"),\n",
    "        '--input AFR_Optim_Param'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/AFR/optimal_param.txt\"),\n",
    "        '--input AMR_Optim_Param'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/AMR/optimal_param.txt\"),\n",
    "        '--input-recursive PROSPER_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER\",\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-pr--williamsjacr--250428-212815-36\n",
      "  job-name: prosper-prosper\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7388948163095099351\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15375682377900572725\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3784459766502440264\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17282830596162308910\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3601401621939499964\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7321139786719999179\n",
      "Launched job-id: prosper-pr--williamsjacr--250428-212815-36\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-pr--williamsjacr--250428-212815-36' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-pr--williamsjacr--250428-212815-36' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 60 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Prosper.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_SL.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_SL.sh\n",
    "\n",
    "path_plink='plink2'\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "tuning_bed=${tuning_bed%??}\n",
    "\n",
    "valid_bed=${valid_bed%??}\n",
    "\n",
    "Rscript ${PROSPER_Scripts}/scripts/tuning_testing.R \\\n",
    "--PATH_plink ${path_plink} \\\n",
    "--PATH_out ${OUTPUT_PATH} \\\n",
    "--PATH_input ${INPUT_PATH} \\\n",
    "--prefix 'full_results' \\\n",
    "--SL_library 'SL.glmnet,SL.ridge,SL.lm' \\\n",
    "--testing TRUE \\\n",
    "--linear_score TRUE \\\n",
    "--bfile_tuning ${tuning_bed} \\\n",
    "--pheno_tuning ${tuning_y} \\\n",
    "--covar_tuning ${tuning_covar} \\\n",
    "--bfile_testing ${valid_bed} \\\n",
    "--pheno_testing ${valid_y} \\\n",
    "--covar_testing ${valid_covar} \\\n",
    "--cleanup F \\\n",
    "--verbose 2 \\\n",
    "--NCORES 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "trait <- \"LDL\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "        '--input-recursive PROSPER_Scripts'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER\",\n",
    "        '--input-recursive INPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "        '--input tuning_bed'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_tune_JointPRS_MUSSEL_PROSPER.*\",\n",
    "        '--input valid_bed'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_valid_JointPRS_MUSSEL_PROSPER.*\",\n",
    "        '--input tuning_y'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_tune_y.txt\"),\n",
    "        '--input valid_y'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/\",trait,\"_full_valid_y.txt\"),\n",
    "        '--input tuning_covar'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_tune_covar.txt\"),\n",
    "        '--input valid_covar'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/Tune_Validation_Phenotypes/full_valid_covar.txt\"),\n",
    "        '--output-recursive OUTPUT_PATH'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait),\n",
    "    check.names = FALSE))   \n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-sl--williamsjacr--250429-000731-50\n",
      "  job-name: prosper-sl\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7692107175929691071\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3463749443434059570\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16794735172943099674\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8470576111325403060\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14809784528643477355\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/744517568757117769\n",
      "Launched job-id: prosper-sl--williamsjacr--250429-000731-50\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sl--williamsjacr--250429-000731-50' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sl--williamsjacr--250429-000731-50' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 60 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --min-cores 1 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_SL.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing JointPRS_Score.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_Score.R\n",
    "rm(list = ls())\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[1]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[2]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    print(trait)\n",
    "    a <- read.csv(paste0(INPUT_PATH,\"/JointPRS_META_Score_\",trait,\".csv\"))\n",
    "    colnames(a) <- c(\"CHR\", \"rsID\", \"BP\", \"A1\", \"A2\", \"Post_EffectSize\")\n",
    "    a <- data.frame(SNP = a$rsID,A1 = a$A1,BETA = a$Post_EffectSize)\n",
    "    write.table(a,file = \"Final_PRS_Coefficients\",col.names = T,row.names = F,quote=F)\n",
    "    \n",
    "    system(paste0(\"plink2 --threads 2 --score Final_PRS_Coefficients cols=+scoresums,-scoreavgs header no-mean-imputation  --bfile \",BED_Full_File,\" --out \",OUTPUT_PATH,\"/PRS_META_\",trait))\n",
    "} \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing JointPRS_Score.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile JointPRS_Score.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${INPUT_PATH} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/Scores\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/JointPRS_Score.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://JointPRS_Score.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  812.0 B/  812.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/812.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp JointPRS_Score.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: jointprs-s--williamsjacr--250501-113726-10\n",
      "  job-name: jointprs-score\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2359695272220548302\n",
      "Launched job-id: jointprs-s--williamsjacr--250501-113726-10\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-s--williamsjacr--250501-113726-10' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'jointprs-s--williamsjacr--250501-113726-10' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script JointPRS_Score.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Score.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score.R\n",
    "rm(list = ls())\n",
    "\n",
    "Score_BMI <- commandArgs(TRUE)[1]\n",
    "print(Score_BMI)\n",
    "\n",
    "Score_LDL <- commandArgs(TRUE)[2]\n",
    "print(Score_LDL)\n",
    "\n",
    "Score_HDL <- commandArgs(TRUE)[3]\n",
    "print(Score_HDL)\n",
    "\n",
    "Score_logTG <- commandArgs(TRUE)[4]\n",
    "print(Score_logTG)\n",
    "\n",
    "Score_TC <- commandArgs(TRUE)[5]\n",
    "print(Score_TC)\n",
    "\n",
    "Score_Height <- commandArgs(TRUE)[6]\n",
    "print(Score_Height)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[7]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[8]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "    \n",
    "    if(trait == \"BMI\"){\n",
    "        Score_File <- Score_BMI\n",
    "    }else if(trait == \"LDL\"){\n",
    "        Score_File <- Score_LDL\n",
    "    }else if(trait == \"HDL\"){\n",
    "        Score_File <- Score_HDL\n",
    "    }else if(trait == \"logTG\"){\n",
    "        Score_File <- Score_logTG\n",
    "    }else if(trait == \"TC\"){\n",
    "        Score_File <- Score_TC\n",
    "    }else {\n",
    "        Score_File <- Score_Height\n",
    "    }\n",
    "    \n",
    "    b <- read.delim(Score_File)\n",
    "    b <- data.frame(SNP = b$rsid,A1 = b$a1,BETA = b$weight)\n",
    "    write.table(b,file = \"Final_PRS_Coefficients\",col.names = T,row.names = F,quote=F)\n",
    "    \n",
    "    system(paste0(\"plink2 --threads 2 --score Final_PRS_Coefficients cols=+scoresums,-scoreavgs header no-mean-imputation  --bfile \",BED_Full_File,\" --out \",OUTPUT_PATH,\"/PRS_\",trait))\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Score.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Score_BMI} ${Score_LDL} ${Score_HDL} ${Score_logTG} ${Score_TC} ${Score_Height} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input Score_BMI'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/BMI/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_LDL'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/LDL/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_HDL'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/HDL/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_logTG'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/logTG/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_TC'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/TC/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input Score_Height'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Height/after_ensemble_full_results/PROSPER_prs_file.txt\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Score.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_Score.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.3 KiB/  1.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_Score.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-sc--williamsjacr--250429-112419-13\n",
      "  job-name: prosper-score\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7246999809422609194\n",
      "Launched job-id: prosper-sc--williamsjacr--250429-112419-13\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--250429-112419-13' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--250429-112419-13' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Score.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Score_All.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score_All.R\n",
    "rm(list = ls())\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Score_Trait <- commandArgs(TRUE)[2]\n",
    "print(Score_Trait)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[3]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Score_File <- Score_Trait\n",
    "    \n",
    "b <- read.delim(Score_File)\n",
    "\n",
    "system(paste0(\"plink2 --threads 2 --score \",Score_Trait,\" cols=+scoresums,-scoreavgs header no-mean-imputation --score-col-nums 4-\",ncol(b),\" --bfile \",BED_Full_File,\" --out \",OUTPUT_PATH,\"/PRS_All_\",trait))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Score_All.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Score_All.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Score_Trait} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "    '--env trait'=trait,\n",
    "    '--input Score_Trait'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/before_ensemble/score_file.txt\"),\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/JointPRS_MUSSEL_PROSPER_BED/all_JointPRS_MUSSEL_PROSPER.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Score_All.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   \n",
    "}\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_Score_All.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  548.0 B/  548.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/548.0 B.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_Score_All.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-sc--williamsjacr--250429-112439-24\n",
      "  job-name: prosper-score-all\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10280476335283298601\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11984450039516978224\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/157940060024560932\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16892770664284640810\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12521981021396864440\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/465680316409281147\n",
      "Launched job-id: prosper-sc--williamsjacr--250429-112439-24\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--250429-112439-24' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-sc--williamsjacr--250429-112439-24' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 200 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"144h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Score_All.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "code_folding": [
     145,
     161
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing CT_SLEB_SL.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_SL.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "AFR_PRS <- commandArgs(TRUE)[1]\n",
    "print(AFR_PRS)\n",
    "\n",
    "AMR_PRS <- commandArgs(TRUE)[2]\n",
    "print(AMR_PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[3]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[4]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[5]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[6]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[7]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[8]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "\n",
    "load(AFR_PRS)\n",
    "colnames(prs_mat_eb)[3:ncol(prs_mat_eb)] <- paste0(\"AFR_\",colnames(prs_mat_eb)[3:ncol(prs_mat_eb)])\n",
    "\n",
    "PRS_Mat_AFR <- prs_mat_eb\n",
    "rm(prs_mat_eb)\n",
    "\n",
    "load(AMR_PRS)\n",
    "colnames(prs_mat_eb)[3:ncol(prs_mat_eb)] <- paste0(\"AMR_\",colnames(prs_mat_eb)[3:ncol(prs_mat_eb)])\n",
    "\n",
    "PRS_Mat_AMR <- prs_mat_eb\n",
    "rm(prs_mat_eb)\n",
    "\n",
    "PRS_Mat_All <- inner_join(PRS_Mat_AFR,PRS_Mat_AMR)\n",
    "colnames(PRS_Mat_All) <- gsub(\"-\",\"_\",colnames(PRS_Mat_All))\n",
    "\n",
    "rm(PRS_Mat_AFR);rm(PRS_Mat_AMR)\n",
    "\n",
    "\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "ndrop <- ncol(pheno_training)\n",
    "\n",
    "PRS_Mat_Train <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "prs_train_all <- pheno_training[,-c(1:ndrop)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "prs_tune_all <- pheno_tuning[,-c(1:ndrop)]\n",
    "print(dim(prs_tune_all))\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "prs_validation_all <- pheno_valid[,-c(1:ndrop)]\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "mtx <- cor(prs_tune_all)\n",
    "drop <- names(prs_tune_all)[apply(mtx,2,function(x){sum(is.na(x))}) == (nrow(mtx) - 1)]\n",
    "\n",
    "prs_train_all <- dplyr::select(prs_train_all, -c(drop))\n",
    "prs_tune_all <- dplyr::select(prs_tune_all, -c(drop))\n",
    "prs_validation_all <- dplyr::select(prs_validation_all, -c(drop))\n",
    "\n",
    "drop <- caret::findLinearCombos(prs_tune_all)$remove\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))\n",
    "\n",
    "mtx <- cor(prs_tune_all)\n",
    "drop <- findCorrelation(mtx,cutoff=0.98)\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))    \n",
    "\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "print(dim(prs_tune_all))\n",
    "print(sum(is.na(prs_tune_all)))\n",
    "\n",
    "print(sum(is.na(prs_tune_all[!is.na(pheno_tuning[,\"y_tune\"]),])))\n",
    "\n",
    "Ensemble_Function_Continuous <- function(x,y){\n",
    "  x <- as.matrix(x[!is.na(y),])\n",
    "  y <- y[!is.na(y)]\n",
    "  \n",
    "  lasso_mod <- cv.glmnet(x,y,family = \"gaussian\",alpha = 1,type.measure = \"mse\",nfold = 10)\n",
    "  ridge_mod <- cv.glmnet(x,y,family = \"gaussian\",alpha = 0,type.measure = \"mse\",nfold = 10)\n",
    "  \n",
    "  lasso_prediction_x <- predict(lasso_mod, x,s = \"lambda.min\")\n",
    "  ridge_prediction_x <- predict(ridge_mod, x,s = \"lambda.min\")\n",
    "  \n",
    "  ensemble_mod <- lm(y~.,data = data.frame(lasso_prediction_x,ridge_prediction_x))\n",
    "  \n",
    "  ensemble_prediction_x <- ensemble_mod$fitted\n",
    "  \n",
    "  coefficients_x <- coef(lm(y~.,data.frame(y = ensemble_prediction_x,x)))\n",
    "  return(list(Coefficients = coefficients_x))\n",
    "}\n",
    "Ensemble_Function_Binary <- function(x,y){\n",
    "  x <- as.matrix(x[!is.na(y),])\n",
    "  y <- y[!is.na(y)]\n",
    "  \n",
    "  lasso_mod <- cv.glmnet(x,y,family = \"binomial\",alpha = 1,type.measure = \"auc\")\n",
    "  ridge_mod <- cv.glmnet(x,y,family = \"binomial\",alpha = 0,type.measure = \"auc\")\n",
    "  \n",
    "  lasso_prediction_x <- predict(lasso_mod, x,type = \"link\")\n",
    "  ridge_prediction_x <- predict(ridge_mod, x,type = \"link\")\n",
    "  \n",
    "  ensemble_mod <- glm(y~.,data = data.frame(lasso_prediction_x,ridge_prediction_x),family = binomial())\n",
    "  ensemble_prediction_x <- predict(ensemble_mod,data.frame(lasso_prediction_x,ridge_prediction_x),type = \"link\")\n",
    "  \n",
    "  coefficients_x <- coef(lm(y~.,data.frame(y = ensemble_prediction_x,x)))\n",
    "  return(list(Coefficients = coefficients_x))\n",
    "}\n",
    "Ensemble_Function <- function(x,y,family = c(\"continuous\",\"binary\")){\n",
    "  if(family == \"continuous\"){\n",
    "    return(Ensemble_Function_Continuous(x,y))\n",
    "  }else{\n",
    "    return(Ensemble_Function_Binary(x,y))\n",
    "  }\n",
    "}\n",
    "\n",
    "Results <- Ensemble_Function(x = prs_tune_all,y = pheno_tuning[,\"y_tune\"],family = \"continuous\")\n",
    "Results$Coefficients[is.na(Results$Coefficients)] <- 0\n",
    "save(Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Final_Coefficients.csv\"))\n",
    "print(names(Results$Coefficients))\n",
    "print(colnames(pheno_training))\n",
    "print(colnames(pheno_tuning))\n",
    "print(colnames(pheno_valid))\n",
    "PRS_Train <- as.matrix(pheno_training[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "PRS_Tune <- as.matrix(pheno_tuning[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "PRS_Validation <- as.matrix(pheno_valid[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "\n",
    "PRS_Train <- data.frame(IID = pheno_training$IID,PRS = PRS_Train)\n",
    "PRS_Tune <- data.frame(IID = pheno_tuning$IID,PRS = PRS_Tune)\n",
    "PRS_Validation <- data.frame(IID = pheno_valid$IID,PRS = PRS_Validation)\n",
    "\n",
    "write.table(PRS_Train,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Train_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(PRS_Tune,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Tune_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(PRS_Validation,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Validation_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- inner_join(pheno_valid[,c(\"IID\",\"age\",\"age2\",\"sex\",paste0(\"PC\",1:10),\"y_validation\")],PRS_Validation)\n",
    "pheno_validation_adjusted <- pheno_validation_raw\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"PRS\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"PRS\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"PRS\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR[,\"PRS\"] <- scale(pheno_validation_raw_EUR[,\"PRS\"])\n",
    "pheno_validation_raw_SAS[,\"PRS\"] <- scale(pheno_validation_raw_SAS[,\"PRS\"])\n",
    "pheno_validation_raw_AMR[,\"PRS\"] <- scale(pheno_validation_raw_AMR[,\"PRS\"])\n",
    "pheno_validation_raw_AFR[,\"PRS\"] <- scale(pheno_validation_raw_AFR[,\"PRS\"])\n",
    "pheno_validation_raw_EAS[,\"PRS\"] <- scale(pheno_validation_raw_EAS[,\"PRS\"])\n",
    "pheno_validation_raw_MID[,\"PRS\"] <- scale(pheno_validation_raw_MID[,\"PRS\"])\n",
    "\n",
    "Beta_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = boot_data))[2]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = boot_data))$r.squared\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_MID_boot <- boot_beta$t\n",
    "beta_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_raw_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_MID_boot <- boot_R2$t\n",
    "R2_se_validation_raw_MID <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_MID_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = pheno_validation_adjusted_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_MID_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_MID <- sd(boot_R2$t)\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         beta_se_raw = c(beta_se_validation_raw_EUR,beta_se_validation_raw_SAS,beta_se_validation_raw_AMR,beta_se_validation_raw_AFR,beta_se_validation_raw_EAS,beta_se_validation_raw_MID), \n",
    "                         R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                         R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         beta_se_adjusted = c(beta_se_validation_adjusted_EUR,beta_se_validation_adjusted_SAS,beta_se_validation_adjusted_AMR,beta_se_validation_adjusted_AFR,beta_se_validation_adjusted_EAS,beta_se_validation_adjusted_MID), \n",
    "                         R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                         R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "CV_Boot_Results <- data.frame(trait = trait,beta_raw_EUR_boot,R2_raw_EUR_boot,beta_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_raw_AMR_boot,R2_raw_AMR_boot,beta_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_raw_EAS_boot,R2_raw_EAS_boot,beta_raw_MID_boot,R2_raw_MID_boot,\n",
    "                              beta_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_adjusted_EAS_boot,R2_adjusted_EAS_boot,\n",
    "                              beta_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(CV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Bootstraps.csv\"),row.names = FALSE)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing CT_SLEB_SL.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile CT_SLEB_SL.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${AFR_PRS} ${AMR_PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input AFR_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AFR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input AMR_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AMR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/CT_SLEB_SL.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/Results/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://CT_SLEB_SL.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 19.5 KiB/ 19.5 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/19.5 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp CT_SLEB_SL.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: ct-sleb-sl--williamsjacr--250516-185556-27\n",
      "  job-name: ct-sleb-sl\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/135171374334310759\n",
      "Launched job-id: ct-sleb-sl--williamsjacr--250516-185556-27\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-sl--williamsjacr--250516-185556-27' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-sl--williamsjacr--250516-185556-27' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script CT_SLEB_SL.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name      Task  Status    Last Update\r\n",
      "----------  ------  --------  --------------\r\n",
      "ct-sleb-sl       1  Success   05-16 19:48:45\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ct-sleb-sl--williamsjacr--250516-185556-27' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Joint_PRS_Results.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Joint_PRS_Results.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "\n",
    "PRS <- commandArgs(TRUE)[1]\n",
    "print(PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[3]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[4]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[5]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[6]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "PRS_Mat_All <- read.delim(PRS)\n",
    "colnames(PRS_Mat_All) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"prs\")\n",
    "PRS_Mat_All <- PRS_Mat_All[,c(\"IID\",\"prs\")]\n",
    "\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- pheno_valid\n",
    "pheno_validation_adjusted <- pheno_valid\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"prs\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(y_hat < 0) > 0){\n",
    "  mod <- lm(y~1,data = tmp)\n",
    "  y_hat <- predict(mod,tmp)\n",
    "}\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"prs\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"prs\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR[,\"prs\"] <- scale(pheno_validation_raw_EUR[,\"prs\"])\n",
    "pheno_validation_raw_SAS[,\"prs\"] <- scale(pheno_validation_raw_SAS[,\"prs\"])\n",
    "pheno_validation_raw_AMR[,\"prs\"] <- scale(pheno_validation_raw_AMR[,\"prs\"])\n",
    "pheno_validation_raw_AFR[,\"prs\"] <- scale(pheno_validation_raw_AFR[,\"prs\"])\n",
    "pheno_validation_raw_EAS[,\"prs\"] <- scale(pheno_validation_raw_EAS[,\"prs\"])\n",
    "pheno_validation_raw_MID[,\"prs\"] <- scale(pheno_validation_raw_MID[,\"prs\"])\n",
    "\n",
    "\n",
    "Beta_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = boot_data))[2]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = boot_data))$r.squared\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_MID_boot <- boot_beta$t\n",
    "beta_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_MID_boot <- boot_R2$t\n",
    "R2_se_validation_raw_MID <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_MID_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_MID_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_MID <- sd(boot_R2$t)\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         beta_se_raw = c(beta_se_validation_raw_EUR,beta_se_validation_raw_SAS,beta_se_validation_raw_AMR,beta_se_validation_raw_AFR,beta_se_validation_raw_EAS,beta_se_validation_raw_MID), \n",
    "                         R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                         R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         beta_se_adjusted = c(beta_se_validation_adjusted_EUR,beta_se_validation_adjusted_SAS,beta_se_validation_adjusted_AMR,beta_se_validation_adjusted_AFR,beta_se_validation_adjusted_EAS,beta_se_validation_adjusted_MID), \n",
    "                         R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                         R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "CV_Boot_Results <- data.frame(trait = trait,beta_raw_EUR_boot,R2_raw_EUR_boot,beta_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_raw_AMR_boot,R2_raw_AMR_boot,beta_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_raw_EAS_boot,R2_raw_EAS_boot,beta_raw_MID_boot,R2_raw_MID_boot,\n",
    "                              beta_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_adjusted_EAS_boot,R2_adjusted_EAS_boot,\n",
    "                              beta_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(CV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Bootstraps.csv\"),row.names = FALSE) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Joint_PRS_Results.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Joint_PRS_Results.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs/PRS_META_\",trait,\".sscore\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Joint_PRS_Results.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/Results/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Joint_PRS_Results.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 15.2 KiB/ 15.2 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/15.2 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Joint_PRS_Results.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: joint-prs---williamsjacr--250514-130145-77\n",
      "  job-name: joint-prs-results\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4362912663453334680\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8026149672767830864\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3362018703464373505\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5321072471225326166\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14179213402507147645\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17956511654413766800\n",
      "Launched job-id: joint-prs---williamsjacr--250514-130145-77\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'joint-prs---williamsjacr--250514-130145-77' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'joint-prs---williamsjacr--250514-130145-77' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 30 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 8 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Joint_PRS_Results.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Results.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Results.R\n",
    "rm(list = ls())\n",
    "\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "\n",
    "PRS <- commandArgs(TRUE)[1]\n",
    "print(PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[2]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[3]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[4]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[5]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[6]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "\n",
    "PRS_Mat_All <- read.delim(PRS)\n",
    "colnames(PRS_Mat_All) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"prs\")\n",
    "PRS_Mat_All <- PRS_Mat_All[,c(\"IID\",\"prs\")]\n",
    "\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_Mat_All[PRS_Mat_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- pheno_valid\n",
    "pheno_validation_adjusted <- pheno_valid\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"prs\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"prs\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"prs\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR[,\"prs\"] <- scale(pheno_validation_raw_EUR[,\"prs\"])\n",
    "pheno_validation_raw_SAS[,\"prs\"] <- scale(pheno_validation_raw_SAS[,\"prs\"])\n",
    "pheno_validation_raw_AMR[,\"prs\"] <- scale(pheno_validation_raw_AMR[,\"prs\"])\n",
    "pheno_validation_raw_AFR[,\"prs\"] <- scale(pheno_validation_raw_AFR[,\"prs\"])\n",
    "pheno_validation_raw_EAS[,\"prs\"] <- scale(pheno_validation_raw_EAS[,\"prs\"])\n",
    "pheno_validation_raw_MID[,\"prs\"] <- scale(pheno_validation_raw_MID[,\"prs\"])\n",
    "\n",
    "\n",
    "Beta_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = boot_data))[2]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = boot_data))$r.squared\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_MID_boot <- boot_beta$t\n",
    "beta_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_MID_boot <- boot_R2$t\n",
    "R2_se_validation_raw_MID <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_MID_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_MID_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_MID <- sd(boot_R2$t)\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         beta_se_raw = c(beta_se_validation_raw_EUR,beta_se_validation_raw_SAS,beta_se_validation_raw_AMR,beta_se_validation_raw_AFR,beta_se_validation_raw_EAS,beta_se_validation_raw_MID), \n",
    "                         R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                         R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         beta_se_adjusted = c(beta_se_validation_adjusted_EUR,beta_se_validation_adjusted_SAS,beta_se_validation_adjusted_AMR,beta_se_validation_adjusted_AFR,beta_se_validation_adjusted_EAS,beta_se_validation_adjusted_MID), \n",
    "                         R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                         R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "CV_Boot_Results <- data.frame(trait = trait,beta_raw_EUR_boot,R2_raw_EUR_boot,beta_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_raw_AMR_boot,R2_raw_AMR_boot,beta_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_raw_EAS_boot,R2_raw_EAS_boot,beta_raw_MID_boot,R2_raw_MID_boot,\n",
    "                              beta_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_adjusted_EAS_boot,R2_adjusted_EAS_boot,\n",
    "                              beta_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(CV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Bootstraps.csv\"),row.names = FALSE) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing PROSPER_Results.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile PROSPER_Results.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${PRS} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/PRS_\",trait,\".sscore\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/PROSPER_Results.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/Results/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://PROSPER_Results.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 15.1 KiB/ 15.1 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/15.1 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp PROSPER_Results.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: prosper-re--williamsjacr--250514-130216-09\n",
      "  job-name: prosper-results\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16373633924048282295\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/503809240254683343\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2765906364711580764\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2079727502755999668\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9050702195468526515\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13293838478095224273\n",
      "Launched job-id: prosper-re--williamsjacr--250514-130216-09\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-re--williamsjacr--250514-130216-09' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'prosper-re--williamsjacr--250514-130216-09' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script PROSPER_Results.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting OneCommonPRS.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile OneCommonPRS.R\n",
    "rm(list = ls())\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "PROSPER_PRSs <- commandArgs(TRUE)[1]\n",
    "print(PROSPER_PRSs)\n",
    "\n",
    "JointPRS_PRS <- commandArgs(TRUE)[2]\n",
    "print(JointPRS_PRS)\n",
    "\n",
    "CTSLEB_AFR <- commandArgs(TRUE)[3]\n",
    "print(CTSLEB_AFR)\n",
    "\n",
    "CTSLEB_AMR <- commandArgs(TRUE)[4]\n",
    "print(CTSLEB_AMR)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[5]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[6]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[7]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[8]\n",
    "print(all_validation_file)\n",
    "\n",
    "trait <- commandArgs(TRUE)[9]\n",
    "print(trait)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[10]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "PROSPER_PRSs <- read.delim(PROSPER_PRSs)\n",
    "colnames(PROSPER_PRSs) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_PRSs) - 4)))\n",
    "PROSPER_PRSs <- PROSPER_PRSs[,c(\"IID\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_PRSs) - 4)))]\n",
    "\n",
    "JointPRS_PRS <- read.delim(JointPRS_PRS)\n",
    "colnames(JointPRS_PRS) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"JointPRS_PRS\")\n",
    "JointPRS_PRS <- JointPRS_PRS[,c(\"IID\",\"JointPRS_PRS\")]\n",
    "\n",
    "load(CTSLEB_AFR)\n",
    "CTSLEB_PRS_AFR <- prs_mat_eb;rm(prs_mat_eb)\n",
    "colnames(CTSLEB_PRS_AFR) <- c(\"FID\",\"IID\",paste0(\"AFR_\",colnames(CTSLEB_PRS_AFR)[3:ncol(CTSLEB_PRS_AFR)]))\n",
    "colnames(CTSLEB_PRS_AFR) <- gsub(\"-\",\"_\",colnames(CTSLEB_PRS_AFR))\n",
    "CTSLEB_PRS_AFR <- CTSLEB_PRS_AFR[,c(\"IID\",colnames(CTSLEB_PRS_AFR)[3:ncol(CTSLEB_PRS_AFR)])]\n",
    "\n",
    "load(CTSLEB_AMR)\n",
    "CTSLEB_PRS_AMR <- prs_mat_eb;rm(prs_mat_eb)\n",
    "colnames(CTSLEB_PRS_AMR) <- c(\"FID\",\"IID\",paste0(\"AMR_\",colnames(CTSLEB_PRS_AMR)[3:ncol(CTSLEB_PRS_AMR)]))\n",
    "colnames(CTSLEB_PRS_AMR) <- gsub(\"-\",\"_\",colnames(CTSLEB_PRS_AMR))\n",
    "CTSLEB_PRS_AMR <- CTSLEB_PRS_AMR[,c(\"IID\",colnames(CTSLEB_PRS_AMR)[3:ncol(CTSLEB_PRS_AMR)])]\n",
    "\n",
    "PRS_All <- inner_join(PROSPER_PRSs,JointPRS_PRS)\n",
    "PRS_All <- inner_join(PRS_All,CTSLEB_PRS_AFR)\n",
    "PRS_All <- inner_join(PRS_All,CTSLEB_PRS_AMR)\n",
    "\n",
    "colnames(PRS_All) <- gsub(\"-\",\"_\",colnames(PRS_All))\n",
    "\n",
    "pheno_training <- read.delim(all_train_file)\n",
    "\n",
    "PRS_Mat_Train <- PRS_All[PRS_All$IID %in% pheno_training$IID,]\n",
    "\n",
    "pheno_training <- left_join(pheno_training,PRS_Mat_Train,by = \"IID\")\n",
    "\n",
    "prs_train_all <- pheno_training[,-c(1:22)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_tuning <- read.delim(all_tune_file)\n",
    "\n",
    "PRS_Mat_Tune <- PRS_All[PRS_All$IID %in% pheno_tuning$IID,]\n",
    "\n",
    "pheno_tuning <- left_join(pheno_tuning,PRS_Mat_Tune,by = \"IID\")\n",
    "\n",
    "prs_tune_all <- pheno_tuning[,-c(1:22)]\n",
    "\n",
    "\n",
    "\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "\n",
    "PRS_Mat_Validation <- PRS_All[PRS_All$IID %in% pheno_valid$IID,]\n",
    "\n",
    "pheno_valid <- left_join(pheno_valid,PRS_Mat_Validation,by = \"IID\")\n",
    "\n",
    "prs_validation_all <- pheno_valid[,-c(1:22)]\n",
    "\n",
    "\n",
    "## Null Models\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_training)\n",
    "y_train <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tuning)\n",
    "y_tune <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_validation <- model.null$residual\n",
    "\n",
    "\n",
    "drop <- caret::findLinearCombos(prs_tune_all)$remove\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))\n",
    "\n",
    "\n",
    "mtx <- cor(prs_tune_all)\n",
    "drop <- findCorrelation(mtx,cutoff=0.98)\n",
    "drop <- names(prs_tune_all)[drop]\n",
    "\n",
    "prs_train_all = prs_train_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_tune_all = prs_tune_all %>% \n",
    "  select(-all_of(drop))\n",
    "prs_validation_all = prs_validation_all %>% \n",
    "  select(-all_of(drop))    \n",
    "\n",
    "\n",
    "pheno_tuning$y_tune <- NA\n",
    "pheno_tuning$y_tune[!is.na(pheno_tuning[,trait])] <- y_tune\n",
    "\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_validation\n",
    "\n",
    "Ensemble_Function_Continuous <- function(x,y){\n",
    "  x <- as.matrix(x[!is.na(y),])\n",
    "  y <- y[!is.na(y)]\n",
    "  \n",
    "  lasso_mod <- cv.glmnet(x,y,family = \"gaussian\",alpha = 1,type.measure = \"mse\",nfold = 10)\n",
    "  ridge_mod <- cv.glmnet(x,y,family = \"gaussian\",alpha = 0,type.measure = \"mse\",nfold = 10)\n",
    "  \n",
    "  lasso_prediction_x <- predict(lasso_mod, x,s = \"lambda.min\")\n",
    "  ridge_prediction_x <- predict(ridge_mod, x,s = \"lambda.min\")\n",
    "  \n",
    "  ensemble_mod <- lm(y~.,data = data.frame(lasso_prediction_x,ridge_prediction_x))\n",
    "  \n",
    "  ensemble_prediction_x <- ensemble_mod$fitted\n",
    "  \n",
    "  coefficients_x <- coef(lm(y~.,data.frame(y = ensemble_prediction_x,x)))\n",
    "  return(list(Coefficients = coefficients_x))\n",
    "}\n",
    "Ensemble_Function_Binary <- function(x,y){\n",
    "  x <- as.matrix(x[!is.na(y),])\n",
    "  y <- y[!is.na(y)]\n",
    "  \n",
    "  lasso_mod <- cv.glmnet(x,y,family = \"binomial\",alpha = 1,type.measure = \"auc\")\n",
    "  ridge_mod <- cv.glmnet(x,y,family = \"binomial\",alpha = 0,type.measure = \"auc\")\n",
    "  \n",
    "  lasso_prediction_x <- predict(lasso_mod, x,type = \"link\")\n",
    "  ridge_prediction_x <- predict(ridge_mod, x,type = \"link\")\n",
    "  \n",
    "  ensemble_mod <- glm(y~.,data = data.frame(lasso_prediction_x,ridge_prediction_x),family = binomial())\n",
    "  ensemble_prediction_x <- predict(ensemble_mod,data.frame(lasso_prediction_x,ridge_prediction_x),type = \"link\")\n",
    "  \n",
    "  coefficients_x <- coef(lm(y~.,data.frame(y = ensemble_prediction_x,x)))\n",
    "  return(list(Coefficients = coefficients_x))\n",
    "}\n",
    "Ensemble_Function <- function(x,y,family = c(\"continuous\",\"binary\")){\n",
    "  if(family == \"continuous\"){\n",
    "    return(Ensemble_Function_Continuous(x,y))\n",
    "  }else{\n",
    "    return(Ensemble_Function_Binary(x,y))\n",
    "  }\n",
    "}\n",
    "\n",
    "Results <- Ensemble_Function(x = prs_tune_all,y = pheno_tuning[,\"y_tune\"],family = \"continuous\")\n",
    "Results$Coefficients[is.na(Results$Coefficients)] <- 0\n",
    "write.csv(Results$Coefficients,file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"))\n",
    "\n",
    "PRS_Train <- as.matrix(pheno_training[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "PRS_Tune <- as.matrix(pheno_tuning[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "PRS_Validation <- as.matrix(pheno_valid[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "\n",
    "prs_best_train <- data.frame(IID = pheno_training$IID,prs = PRS_Train)\n",
    "prs_best_tune <- data.frame(IID = pheno_tuning$IID,prs = PRS_Tune)\n",
    "prs_best_validation <- data.frame(IID = pheno_valid$IID,prs = PRS_Validation)\n",
    "\n",
    "write.table(prs_best_train,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Train_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(prs_best_tune,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Tune_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "write.table(prs_best_validation,file=paste0(OUTPUT_PATH,\"/\",trait,\"_Best_Validation_All.txt\"),sep = \"\\t\",row.names = FALSE)\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "pheno_validation_raw <- inner_join(pheno_valid[,c(\"IID\",\"age\",\"age2\",\"sex\",paste0(\"PC\",1:10),\"y_validation\")],prs_best_validation)\n",
    "pheno_validation_adjusted <- pheno_validation_raw\n",
    "\n",
    "tmp <- data.frame(y = pheno_validation_adjusted[,\"prs\"],pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,pheno_validation_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  pheno_validation_adjusted[,\"prs\"] <- 0\n",
    "}else{\n",
    "  pheno_validation_adjusted[,\"prs\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "pheno_validation_raw_EUR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_raw_SAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_raw_AMR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_raw_AFR <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_raw_EAS <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_raw_MID <- pheno_validation_raw[pheno_validation_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "pheno_validation_adjusted_EUR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "pheno_validation_adjusted_SAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "pheno_validation_adjusted_AMR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "pheno_validation_adjusted_AFR <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "pheno_validation_adjusted_EAS <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "pheno_validation_adjusted_MID <- pheno_validation_adjusted[pheno_validation_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "pheno_validation_raw_EUR$y_validation <- scale(pheno_validation_raw_EUR$y_validation)\n",
    "pheno_validation_raw_SAS$y_validation <- scale(pheno_validation_raw_SAS$y_validation)\n",
    "pheno_validation_raw_AMR$y_validation <- scale(pheno_validation_raw_AMR$y_validation)\n",
    "pheno_validation_raw_AFR$y_validation <- scale(pheno_validation_raw_AFR$y_validation)\n",
    "pheno_validation_raw_EAS$y_validation <- scale(pheno_validation_raw_EAS$y_validation)\n",
    "pheno_validation_raw_MID$y_validation <- scale(pheno_validation_raw_MID$y_validation)\n",
    "\n",
    "pheno_validation_adjusted_EUR$y_validation <- scale(pheno_validation_adjusted_EUR$y_validation)\n",
    "pheno_validation_adjusted_SAS$y_validation <- scale(pheno_validation_adjusted_SAS$y_validation)\n",
    "pheno_validation_adjusted_AMR$y_validation <- scale(pheno_validation_adjusted_AMR$y_validation)\n",
    "pheno_validation_adjusted_AFR$y_validation <- scale(pheno_validation_adjusted_AFR$y_validation)\n",
    "pheno_validation_adjusted_EAS$y_validation <- scale(pheno_validation_adjusted_EAS$y_validation)\n",
    "pheno_validation_adjusted_MID$y_validation <- scale(pheno_validation_adjusted_MID$y_validation)\n",
    "\n",
    "pheno_validation_raw_EUR[,\"prs\"] <- scale(pheno_validation_raw_EUR[,\"prs\"])\n",
    "pheno_validation_raw_SAS[,\"prs\"] <- scale(pheno_validation_raw_SAS[,\"prs\"])\n",
    "pheno_validation_raw_AMR[,\"prs\"] <- scale(pheno_validation_raw_AMR[,\"prs\"])\n",
    "pheno_validation_raw_AFR[,\"prs\"] <- scale(pheno_validation_raw_AFR[,\"prs\"])\n",
    "pheno_validation_raw_EAS[,\"prs\"] <- scale(pheno_validation_raw_EAS[,\"prs\"])\n",
    "pheno_validation_raw_MID[,\"prs\"] <- scale(pheno_validation_raw_MID[,\"prs\"])\n",
    "\n",
    "Beta_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = boot_data))[2]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = boot_data))$r.squared\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_raw_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_MID_boot <- boot_beta$t\n",
    "beta_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_raw_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_raw_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_MID_boot <- boot_R2$t\n",
    "R2_se_validation_raw_MID <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EUR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_SAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AMR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_AFR))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_EAS))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))[2]\n",
    "boot_beta <- boot(data = pheno_validation_adjusted_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_MID_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"prs\")),data = pheno_validation_adjusted_MID))$r.squared\n",
    "boot_R2 <- boot(data = pheno_validation_adjusted_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_MID_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_MID <- sd(boot_R2$t)\n",
    "\n",
    "CV_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                         beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                         beta_se_raw = c(beta_se_validation_raw_EUR,beta_se_validation_raw_SAS,beta_se_validation_raw_AMR,beta_se_validation_raw_AFR,beta_se_validation_raw_EAS,beta_se_validation_raw_MID), \n",
    "                         R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                         R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                         beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                         beta_se_adjusted = c(beta_se_validation_adjusted_EUR,beta_se_validation_adjusted_SAS,beta_se_validation_adjusted_AMR,beta_se_validation_adjusted_AFR,beta_se_validation_adjusted_EAS,beta_se_validation_adjusted_MID), \n",
    "                         R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                         R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "CV_Boot_Results <- data.frame(trait = trait,beta_raw_EUR_boot,R2_raw_EUR_boot,beta_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_raw_AMR_boot,R2_raw_AMR_boot,beta_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_raw_EAS_boot,R2_raw_EAS_boot,beta_raw_MID_boot,R2_raw_MID_boot,\n",
    "                              beta_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_adjusted_EAS_boot,R2_adjusted_EAS_boot,\n",
    "                              beta_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "write.csv(CV_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(CV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Bootstraps.csv\"),row.names = FALSE) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting OneCommonPRS.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile OneCommonPRS.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${PROSPER_PRSs} ${JointPRS_PRS} ${CTSLEB_AFR} ${CTSLEB_AMR} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${trait} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"logTG\")){\n",
    "    tasks <- rbind(tasks, data.frame(\n",
    "    '--input PROSPER_PRSs'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/PRS_All_\",trait,\".sscore\"),\n",
    "    '--input JointPRS_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs/PRS_META_\",trait,\".sscore\"),\n",
    "    '--input CTSLEB_AFR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AFR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input CTSLEB_AMR'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AMR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/OneCommonPRS.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   }\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://OneCommonPRS.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 20.1 KiB/ 20.1 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/20.1 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp OneCommonPRS.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: onecommonp--williamsjacr--250516-214332-35\n",
      "  job-name: onecommonprs\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7513765309207800356\n",
      "Launched job-id: onecommonp--williamsjacr--250516-214332-35\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'onecommonp--williamsjacr--250516-214332-35' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'onecommonp--williamsjacr--250516-214332-35' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script OneCommonPRS.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Extract_Betas_All.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Extract_Betas_All.R\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "library(data.table)\n",
    "\n",
    "GWAS_SumStats <- commandArgs(TRUE)[1]\n",
    "print(GWAS_SumStats)\n",
    "\n",
    "snp_info <- commandArgs(TRUE)[2]\n",
    "print(snp_info)\n",
    "\n",
    "PROSPER_PRSs <- commandArgs(TRUE)[3]\n",
    "print(PROSPER_PRSs)\n",
    "\n",
    "JointPRS_PRS <- commandArgs(TRUE)[4]\n",
    "print(JointPRS_PRS)\n",
    "\n",
    "CTSLEB_AFR_PRSs <- commandArgs(TRUE)[5]\n",
    "print(CTSLEB_AFR_PRSs)\n",
    "\n",
    "CTSLEB_AMR_PRSs <- commandArgs(TRUE)[6]\n",
    "print(CTSLEB_AMR_PRSs)\n",
    "\n",
    "PROSPER_Scores <- commandArgs(TRUE)[7]\n",
    "print(PROSPER_Scores)\n",
    "\n",
    "JointPRS_Score <- commandArgs(TRUE)[8]\n",
    "print(JointPRS_Score)\n",
    "\n",
    "CTSLEB_AFR_Scores <- commandArgs(TRUE)[9]\n",
    "print(CTSLEB_AFR_Scores)\n",
    "\n",
    "CTSLEB_AMR_Scores <- commandArgs(TRUE)[10]\n",
    "print(CTSLEB_AMR_Scores)\n",
    "\n",
    "RICE_CV_PRS_Tune <- commandArgs(TRUE)[11]\n",
    "print(RICE_CV_PRS_Tune)\n",
    "\n",
    "RICE_CV_PRS_Validation <- commandArgs(TRUE)[12]\n",
    "print(RICE_CV_PRS_Validation)\n",
    "\n",
    "Validation <- commandArgs(TRUE)[13]\n",
    "print(Validation)\n",
    "\n",
    "trait <- commandArgs(TRUE)[14]\n",
    "print(trait)\n",
    "\n",
    "BED_Full_File <- commandArgs(TRUE)[15]\n",
    "BED_Full_File <- gsub(\".bed\",\"\",BED_Full_File)\n",
    "print(BED_Full_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[16]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "PROSPER_PRSs <- read.delim(PROSPER_PRSs)\n",
    "colnames(PROSPER_PRSs) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_PRSs) - 4)))\n",
    "PROSPER_PRSs <- PROSPER_PRSs[,c(\"IID\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_PRSs) - 4)))]\n",
    "\n",
    "JointPRS_PRS <- read.delim(JointPRS_PRS)\n",
    "colnames(JointPRS_PRS) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"JointPRS_PRS\")\n",
    "JointPRS_PRS <- JointPRS_PRS[,c(\"IID\",\"JointPRS_PRS\")]\n",
    "\n",
    "load(CTSLEB_AFR_PRSs)\n",
    "CTSLEB_PRS_AFR <- prs_mat_eb;rm(prs_mat_eb)\n",
    "colnames(CTSLEB_PRS_AFR) <- c(\"FID\",\"IID\",paste0(\"AFR_\",colnames(CTSLEB_PRS_AFR)[3:ncol(CTSLEB_PRS_AFR)]))\n",
    "colnames(CTSLEB_PRS_AFR) <- gsub(\"-\",\"_\",colnames(CTSLEB_PRS_AFR))\n",
    "CTSLEB_PRS_AFR <- CTSLEB_PRS_AFR[,c(\"IID\",colnames(CTSLEB_PRS_AFR)[3:ncol(CTSLEB_PRS_AFR)])]\n",
    "\n",
    "load(CTSLEB_AMR_PRSs)\n",
    "CTSLEB_PRS_AMR <- prs_mat_eb;rm(prs_mat_eb)\n",
    "colnames(CTSLEB_PRS_AMR) <- c(\"FID\",\"IID\",paste0(\"AMR_\",colnames(CTSLEB_PRS_AMR)[3:ncol(CTSLEB_PRS_AMR)]))\n",
    "colnames(CTSLEB_PRS_AMR) <- gsub(\"-\",\"_\",colnames(CTSLEB_PRS_AMR))\n",
    "CTSLEB_PRS_AMR <- CTSLEB_PRS_AMR[,c(\"IID\",colnames(CTSLEB_PRS_AMR)[3:ncol(CTSLEB_PRS_AMR)])]\n",
    "\n",
    "PRS_All <- inner_join(PROSPER_PRSs,JointPRS_PRS)\n",
    "PRS_All <- inner_join(PRS_All,CTSLEB_PRS_AFR)\n",
    "PRS_All <- inner_join(PRS_All,CTSLEB_PRS_AMR)\n",
    "\n",
    "rm(CTSLEB_PRS_AMR);rm(CTSLEB_PRS_AFR);rm(JointPRS_PRS);rm(PROSPER_PRSs)\n",
    "gc()\n",
    "\n",
    "colnames(PRS_All) <- gsub(\"-\",\"_\",colnames(PRS_All))\n",
    "\n",
    "\n",
    "GWAS_SumStats <- fread(GWAS_SumStats)\n",
    "\n",
    "GWAS_SumStats <- as.data.frame(GWAS_SumStats)\n",
    "GWAS_SumStats <- GWAS_SumStats[,c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"SNP\")]\n",
    "colnames(GWAS_SumStats) <- c(\"CHR\",\"SNP\",\"BP\",\"REF\",\"ALT\",\"OBS_CT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "GWAS_SumStats <- GWAS_SumStats[,c(\"CHR\",\"SNP\",\"BP\",\"ALT\",\"BETA\",\"SE\",\"P\",\"rs_ID\")]\n",
    "colnames(GWAS_SumStats) <- c(\"CHR\",\"SNP\",\"BP\",\"A1\",\"BETA\",\"SE\",\"P\",\"rs_ID\")\n",
    "\n",
    "GWAS_SumStats$SNP_Merge <- toupper(GWAS_SumStats$SNP)\n",
    "GWAS_SumStats$SNP_Merge <- gsub(\" \",\"\",GWAS_SumStats$SNP_Merge)\n",
    "\n",
    "SNP_GRCh37_38_match_update <- readRDS(snp_info)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele1_38,\":\",SNP_GRCh37_38_match_update$allele2_38)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- paste0(\"chr\",SNP_GRCh37_38_match_update$chr,\":\",SNP_GRCh37_38_match_update$pos38,\":\",\n",
    "                                                SNP_GRCh37_38_match_update$allele2_38,\":\",SNP_GRCh37_38_match_update$allele1_38)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- toupper(SNP_GRCh37_38_match_update$unique_id1)\n",
    "SNP_GRCh37_38_match_update$unique_id1 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id1)\n",
    "\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- toupper(SNP_GRCh37_38_match_update$unique_id2)\n",
    "SNP_GRCh37_38_match_update$unique_id2 <- gsub(\" \",\"\",SNP_GRCh37_38_match_update$unique_id2)\n",
    "\n",
    "GWAS_SumStats <- left_join(GWAS_SumStats,SNP_GRCh37_38_match_update[,c(\"unique_id1\",\"rsid\")],by = c(\"SNP_Merge\"=\"unique_id1\"))\n",
    "GWAS_SumStats$rs_ID[!is.na(GWAS_SumStats$rsid)] <- GWAS_SumStats$rsid[!is.na(GWAS_SumStats$rsid)]\n",
    "\n",
    "GWAS_SumStats <- subset(GWAS_SumStats,select = -c(rsid))\n",
    "\n",
    "GWAS_SumStats <- left_join(GWAS_SumStats,SNP_GRCh37_38_match_update[,c(\"unique_id2\",\"rsid\")],by = c(\"SNP_Merge\"=\"unique_id2\"))\n",
    "GWAS_SumStats$rs_ID[!is.na(GWAS_SumStats$rsid)] <- GWAS_SumStats$rsid[!is.na(GWAS_SumStats$rsid)]\n",
    "\n",
    "GWAS_SumStats <- GWAS_SumStats[,c(\"SNP\",\"A1\",\"rs_ID\")]\n",
    "\n",
    "rm(SNP_GRCh37_38_match_update);gc()\n",
    "\n",
    "\n",
    "PROSPER_Scores <- read.delim(PROSPER_Scores)\n",
    "colnames(PROSPER_Scores) <- c(\"rs_ID\",\"A1\",\"A0\",paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_Scores) - 3)))\n",
    "\n",
    "JointPRS_Score <- read.csv(JointPRS_Score,header = TRUE)\n",
    "colnames(JointPRS_Score) <- c(\"CHR\",\"rs_ID\",\"BP\",\"A1\",\"A0\",\"JointPRS_PRS\")\n",
    "\n",
    "CTSLEB_AFR_Scores <- get(load(CTSLEB_AFR_Scores))\n",
    "CTSLEB_AFR_Scores <- CTSLEB_AFR_Scores[,-c(ncol(CTSLEB_AFR_Scores) - 1,ncol(CTSLEB_AFR_Scores))]\n",
    "colnames(CTSLEB_AFR_Scores) <- c(\"SNP\",\"A1\",paste0(\"AFR_\",colnames(CTSLEB_AFR_Scores)[3:ncol(CTSLEB_AFR_Scores)]))\n",
    "colnames(CTSLEB_AFR_Scores) <- gsub(\"-\",\"_\",colnames(CTSLEB_AFR_Scores))\n",
    "\n",
    "CTSLEB_AMR_Scores <- get(load(CTSLEB_AMR_Scores))\n",
    "CTSLEB_AMR_Scores <- CTSLEB_AMR_Scores[,-c(ncol(CTSLEB_AMR_Scores) - 1,ncol(CTSLEB_AMR_Scores))]\n",
    "colnames(CTSLEB_AMR_Scores) <- c(\"SNP\",\"A1\",paste0(\"AMR_\",colnames(CTSLEB_AMR_Scores)[3:ncol(CTSLEB_AMR_Scores)]))\n",
    "colnames(CTSLEB_AMR_Scores) <- gsub(\"-\",\"_\",colnames(CTSLEB_AMR_Scores))\n",
    "\n",
    "PROSPER_Scores[!(paste0(PROSPER_Scores$rs_ID,\"_\",PROSPER_Scores$A1) %in% paste0(GWAS_SumStats$rs_ID,\"_\",GWAS_SumStats$A1)),paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_Scores) - 3))] <- (-1)*PROSPER_Scores[!(paste0(PROSPER_Scores$rs_ID,\"_\",PROSPER_Scores$A1) %in% paste0(GWAS_SumStats$rs_ID,\"_\",GWAS_SumStats$A1)),paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_Scores) - 3))]\n",
    "JointPRS_Score[!(paste0(JointPRS_Score$rs_ID,\"_\",JointPRS_Score$A1) %in% paste0(GWAS_SumStats$rs_ID,\"_\",GWAS_SumStats$A1)),\"JointPRS_PRS\"] <- (-1)*JointPRS_Score[!(paste0(JointPRS_Score$rs_ID,\"_\",JointPRS_Score$A1) %in% paste0(GWAS_SumStats$rs_ID,\"_\",GWAS_SumStats$A1)),\"JointPRS_PRS\"]\n",
    "CTSLEB_AFR_Scores[!(paste0(CTSLEB_AFR_Scores$SNP,\"_\",CTSLEB_AFR_Scores$A1) %in% paste0(GWAS_SumStats$SNP,\"_\",GWAS_SumStats$A1)),3:ncol(CTSLEB_AFR_Scores)] <- (-1)*CTSLEB_AFR_Scores[!(paste0(CTSLEB_AFR_Scores$SNP,\"_\",CTSLEB_AFR_Scores$A1) %in% paste0(GWAS_SumStats$SNP,\"_\",GWAS_SumStats$A1)),3:ncol(CTSLEB_AFR_Scores)]\n",
    "CTSLEB_AMR_Scores[!(paste0(CTSLEB_AMR_Scores$SNP,\"_\",CTSLEB_AMR_Scores$A1) %in% paste0(GWAS_SumStats$SNP,\"_\",GWAS_SumStats$A1)),3:ncol(CTSLEB_AMR_Scores)] <- (-1)*CTSLEB_AMR_Scores[!(paste0(CTSLEB_AMR_Scores$SNP,\"_\",CTSLEB_AMR_Scores$A1) %in% paste0(GWAS_SumStats$SNP,\"_\",GWAS_SumStats$A1)),3:ncol(CTSLEB_AMR_Scores)]\n",
    "\n",
    "\n",
    "## Merge covariates and y for tuning with the prs_mat\n",
    "Best_Tune_All <- read.delim(RICE_CV_PRS_Tune)\n",
    "Best_Tune_All <- inner_join(Best_Tune_All,PRS_All)\n",
    "Best_Tune_All <- subset(Best_Tune_All,select = -IID)\n",
    "\n",
    "tmp1 <- lm(prs ~.,data = Best_Tune_All) \n",
    "tmp <- coef(tmp1)\n",
    "Beta_Star <- matrix(unname(tmp),ncol = 1)\n",
    "Beta_Star <- Beta_Star[-1,,drop = FALSE]\n",
    "Beta_Star[is.na(Beta_Star)] <- 0\n",
    "\n",
    "PROSPER_Scores$Final_Prosper <- as.numeric(as.matrix(PROSPER_Scores[,paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_Scores) - 3))])%*% Beta_Star[names(tmp)[-1] %in% paste0(\"PROSPER_PRS_\",1:(ncol(PROSPER_Scores) - 3)),1,drop = FALSE])\n",
    "JointPRS_Score$Final_JointPRS <- as.numeric(as.matrix(JointPRS_Score[,\"JointPRS_PRS\"])%*% Beta_Star[names(tmp)[-1] %in% \"JointPRS_PRS\",1,drop = FALSE])\n",
    "CTSLEB_AMR_Scores$Final_CTSLEB_AMR <- as.numeric(as.matrix(CTSLEB_AMR_Scores[,colnames(CTSLEB_AMR_Scores)[3:ncol(CTSLEB_AMR_Scores)]])%*% Beta_Star[names(tmp)[-1] %in% colnames(CTSLEB_AMR_Scores)[3:ncol(CTSLEB_AMR_Scores)],1,drop = FALSE])\n",
    "CTSLEB_AFR_Scores$Final_CTSLEB_AFR <- as.numeric(as.matrix(CTSLEB_AFR_Scores[,colnames(CTSLEB_AFR_Scores)[3:ncol(CTSLEB_AFR_Scores)]])%*% Beta_Star[names(tmp)[-1] %in% colnames(CTSLEB_AFR_Scores)[3:ncol(CTSLEB_AFR_Scores)],1,drop = FALSE])\n",
    "\n",
    "GWAS_SumStats <- left_join(GWAS_SumStats,PROSPER_Scores[,c(\"rs_ID\",\"Final_Prosper\")])  \n",
    "GWAS_SumStats <- left_join(GWAS_SumStats,JointPRS_Score[,c(\"rs_ID\",\"Final_JointPRS\")]) \n",
    "GWAS_SumStats <- left_join(GWAS_SumStats,CTSLEB_AMR_Scores[,c(\"SNP\",\"Final_CTSLEB_AMR\")])  \n",
    "GWAS_SumStats <- left_join(GWAS_SumStats,CTSLEB_AFR_Scores[,c(\"SNP\",\"Final_CTSLEB_AFR\")]) \n",
    "\n",
    "GWAS_SumStats[is.na(GWAS_SumStats)] <- 0 \n",
    "GWAS_SumStats$BETA <- GWAS_SumStats$Final_Prosper +GWAS_SumStats$Final_JointPRS +GWAS_SumStats$Final_CTSLEB_AMR +GWAS_SumStats$Final_CTSLEB_AFR\n",
    "\n",
    "GWAS_SumStats <- GWAS_SumStats[,c(\"SNP\",\"A1\",\"BETA\")]\n",
    "GWAS_SumStats <- GWAS_SumStats[GWAS_SumStats$BETA != 0,]\n",
    "\n",
    "write.table(GWAS_SumStats,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Final_Score\"),col.names = T,row.names = F,quote=F)\n",
    "\n",
    "Validation <- read.delim(Validation)\n",
    "Keep <- data.frame(FID = 0,IID = Validation$IID)\n",
    "write.table(Keep,\"Keep.txt\",row.names = FALSE,col.names = FALSE)\n",
    "\n",
    "system(paste0(\"plink2 --threads 2 --score \",OUTPUT_PATH,\"/\",trait,\"_Final_Score cols=+scoresums,-scoreavgs header no-mean-imputation --keep Keep.txt --bfile \",BED_Full_File,\" --out test_validation\"),intern = TRUE)\n",
    "\n",
    "Best_Validation_All <- read.csv(RICE_CV_PRS_Validation,sep = \"\\t\")\n",
    "test_validation <- read.delim(\"test_validation.sscore\", header=FALSE, comment.char=\"#\")\n",
    "\n",
    "all.equal(Best_Validation_All$IID,test_validation[,2])\n",
    "all.equal(Best_Validation_All$prs,test_validation[,5])\n",
    "cor(Best_Validation_All$prs,test_validation[,5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Extract_Betas_All.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Extract_Betas_All.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${GWAS_SumStats} ${snp_info} ${PROSPER_PRSs} ${JointPRS_PRS} ${CTSLEB_AFR_PRSs} ${CTSLEB_AMR_PRSs} ${PROSPER_Scores} ${JointPRS_Score} ${CTSLEB_AFR_Scores} ${CTSLEB_AMR_Scores} ${RICE_CV_PRS_Tune} ${RICE_CV_PRS_Validation} ${Validation} ${trait} ${BED_Full_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"TC\",\"logTG\",\"Height\")){\n",
    "  tasks <- rbind(tasks, data.frame(\n",
    "    '--input GWAS_SumStats'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GWAS_SummaryStats/\",\"EUR\",\"_\",trait,\"_GWAS_SumStats_Cleaned\"),\n",
    "    '--input snp_info'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/SNP_GRCh37_38_match_update.rds\",\n",
    "    '--input PROSPER_PRSs'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/PRS_All_\",trait,\".sscore\"),\n",
    "    '--input JointPRS_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs/PRS_META_\",trait,\".sscore\"),\n",
    "    '--input CTSLEB_AFR_PRSs'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AFR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input CTSLEB_AMR_PRSs'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AMR_\",trait,\"_prs_mat_eb.RData\"),\n",
    "    '--input PROSPER_Scores'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/\",trait,\"/before_ensemble/score_file.txt\"),\n",
    "    '--input JointPRS_Score'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/Scores/JointPRS_META_Score_\",trait,\".csv\"),\n",
    "    '--input CTSLEB_AFR_Scores'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AFR_\",trait,\"_betas_eb.RData\"),\n",
    "    '--input CTSLEB_AMR_Scores'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/AMR_\",trait,\"_betas_eb.RData\"),\n",
    "    '--input RICE_CV_PRS_Tune'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",trait,\"_Best_Tune_All.txt\"),\n",
    "    '--input RICE_CV_PRS_Validation'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",trait,\"_Best_Validation_All.txt\"),\n",
    "    '--input Validation'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input BED_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bed\",\n",
    "    '--input BIM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.bim\",\n",
    "    '--input FAM_Full_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Data/CommonVariants/all_chr.fam\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Extract_Betas_All.R\",\n",
    "    '--env trait'=trait,\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",\n",
    "    check.names = FALSE\n",
    "  ))   \n",
    "}\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Extract_Betas_All.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  9.3 KiB/  9.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/9.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Extract_Betas_All.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: extract-be--williamsjacr--250516-234309-35\n",
      "  job-name: extract-betas-all\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3885708353785787379\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17699662607382907460\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/5596598642924517590\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6714268682014752827\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4772606334659318813\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6504039311456878092\n",
      "Launched job-id: extract-be--williamsjacr--250516-234309-35\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'extract-be--williamsjacr--250516-234309-35' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'extract-be--williamsjacr--250516-234309-35' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 700 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 72 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Extract_Betas_All.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting NullModel.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile NullModel.R\n",
    "rm(list = ls())\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[1]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[2]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[3]\n",
    "print(all_validation_file)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(dplyr)\n",
    "library(STAAR)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(Matrix)\n",
    "library(SCANG)\n",
    "library(STAARpipeline)\n",
    "\n",
    "trait <- \"BMI\"\n",
    "\n",
    "for(trait in c(\"BMI\",\"TC\",\"LDL\",\"HDL\",\"logTG\",\"Height\")){\n",
    "  pheno_train <- read.delim(all_train_file)\n",
    "  \n",
    "  common_prs <- read.delim(paste0(INPUT_PATH,\"/\",trait,\"_Best_Train_All.txt\"))\n",
    "  \n",
    "  pheno_train <- inner_join(pheno_train,common_prs,by = \"IID\")\n",
    "  \n",
    "  obj.STAAR.UKB <- fit_nullmodel(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10 + prs\")), data = pheno_train,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "  save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Train_Null_Model.RData\"))\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  pheno_tune <- read.delim(all_tune_file)\n",
    "  \n",
    "  common_prs <- read.delim(paste0(INPUT_PATH,\"/\",trait,\"_Best_Tune_All.txt\"))\n",
    "  \n",
    "  pheno_tune <- inner_join(pheno_tune,common_prs,by = \"IID\")\n",
    "  \n",
    "  obj.STAAR.UKB <- fit_nullmodel(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10 + prs\")), data = pheno_tune,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "  save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Tune_Null_Model.RData\"))\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  pheno_validation <- read.delim(all_validation_file)\n",
    "  \n",
    "  common_prs <- read.delim(paste0(INPUT_PATH,\"/\",trait,\"_Best_Validation_All.txt\"))\n",
    "  \n",
    "  pheno_validation <- inner_join(pheno_validation,common_prs,by = \"IID\")\n",
    "  \n",
    "  obj.STAAR.UKB <- fit_nullmodel(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10 + prs\")), data = pheno_validation,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "  save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Validation_Null_Model.RData\"))\n",
    "  \n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "    '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "    '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS\",\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/NullModel.R\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels\",\n",
    "    check.names = FALSE\n",
    "))   \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting NullModel.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile NullModel.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${all_train_file} ${all_tune_file} ${all_validation_file} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://NullModel.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.0 KiB/  2.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.0 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp NullModel.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: nullmodel--williamsjacr--240801-214711-99\n",
      "  job-name: nullmodel\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4654998437295274832\n",
      "Launched job-id: nullmodel--williamsjacr--240801-214711-99\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240801-214711-99' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240801-214711-99' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script NullModel.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /home/jupyter/aou_dsub.bash\n"
     ]
    }
   ],
   "source": [
    "%%writefile ~/aou_dsub.bash\n",
    "#!/bin/bash\n",
    "function aou_dsub () {\n",
    "\n",
    "  # Get a shorter username to leave more characters for the job name.\n",
    "  local DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "  # For AoU RWB projects network name is \"network\".\n",
    "  local AOU_NETWORK=network\n",
    "  local AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "  dsub \\\n",
    "      --provider google-cls-v2 \\\n",
    "      --user-project \"${GOOGLE_PROJECT}\"\\\n",
    "      --project \"${GOOGLE_PROJECT}\"\\\n",
    "      --image 'marketplace.gcr.io/google/ubuntu1804:latest' \\\n",
    "      --network \"${AOU_NETWORK}\" \\\n",
    "      --subnetwork \"${AOU_SUBNETWORK}\" \\\n",
    "      --service-account \"$(gcloud config get-value account)\" \\\n",
    "      --user \"${DSUB_USER_NAME}\" \\\n",
    "      --regions us-central1 \\\n",
    "      --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "      \"$@\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo source ~/aou_dsub.bash >> ~/.bashrc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "code_folding": [
     81,
     271
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ncRNA_Analysis.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile ncRNA_Analysis.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "trait <- commandArgs(TRUE)[2]\n",
    "print(trait)\n",
    "\n",
    "Train_Null_Model <- commandArgs(TRUE)[3]\n",
    "print(Train_Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[4]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[5]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "\n",
    "obj_nullmodel <- get(load(Train_Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output file name\n",
    "output_file_name <- paste0(OUTPUT_PATH,\"/\",trait,\"_AoU_WGS_ncRNA_Train\")\n",
    "\n",
    "## gene number in job\n",
    "gene_num_in_array <- 100 \n",
    "group.num.allchr <- ceiling(table(ncRNA_gene[,1])/gene_num_in_array)\n",
    "\n",
    "chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if (chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "ncRNA_gene_chr <- ncRNA_gene[ncRNA_gene[,1]==chr,]\n",
    "ncRNA_longmasks <- c(\"KCNQ1OT1\",\"AC006548.28\",\"RP3-394A18.1\",\"RP3-323A16.1\",\"RP5-1039K5.19\",\"SNHG14\",\"LL22NC03-86G7.1\")\n",
    "ncRNA_gene_chr <- ncRNA_gene_chr[!ncRNA_gene_chr[,2] %in% ncRNA_longmasks,]\n",
    "sub_seq_num <- dim(ncRNA_gene_chr)[1]\n",
    "\n",
    "if(groupid < group.num){\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "###########################################################\n",
    "#           Main Function \n",
    "###########################################################\n",
    "\n",
    "ncRNA_nolongmask <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                             rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                             QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                             Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                             Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                             SPA_p_filter=TRUE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ## ncRNA SNVs\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  is.in <- ((GENCODE.Category==\"ncRNA_exonic\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\"))&(SNVlist)\n",
    "  \n",
    "  variant.id.ncRNA <- variant.id[is.in]\n",
    "  \n",
    "  rm(GENCODE.Category)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.ncRNA,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.ncRNA)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[;]\")\n",
    "  Gene <- as.character(sapply(GENCODE.Info.split,function(z) gsub(\"\\\\(.*\\\\)\",\"\",z[1])))\n",
    "  \n",
    "  Gene_list_1 <- as.character(sapply(strsplit(Gene,','),'[',1))\n",
    "  Gene_list_2 <- as.character(sapply(strsplit(Gene,','),'[',2))\n",
    "  Gene_list_3 <- as.character(sapply(strsplit(Gene,','),'[',3))\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  variant.id.ncRNA <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- union(which(Gene_list_1==gene_name),which(Gene_list_2==gene_name))\n",
    "  is.in <- union(is.in,which(Gene_list_3==gene_name))\n",
    "  \n",
    "  variant.is.in <- variant.id.ncRNA[is.in]\n",
    "  \n",
    "  if(length(variant.is.in) > 10000){\n",
    "    return(NULL)\n",
    "  }\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"ncRNA\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "agds.path <- GDS_File\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "gene_centric_ncRNA_dnanexus <- function(ncRNA_gene_chr,gene_name,chr,genofile,obj_nullmodel,rare_maf_cutoff,\n",
    "                                        QC_label,variant_type,geno_missing_imputation,\n",
    "                                        Annotation_dir,Annotation_name_catalog,\n",
    "                                        Use_annotation_weights,Annotation_name,silent){\n",
    "  results <- try(ncRNA_nolongmask(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                  QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                  Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                  Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=silent),silent=TRUE)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "results_ncRNA <- c()\n",
    "for(kk in sub_seq_id){\n",
    "  print(kk)\n",
    "  gene_name <- ncRNA_gene_chr[kk,2]\n",
    "  results <- gene_centric_ncRNA_dnanexus(ncRNA_gene_chr=ncRNA_gene_chr,gene_name = gene_name,chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=max.maf,\n",
    "                                         QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                         Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=TRUE)\n",
    "  \n",
    "  results_ncRNA <- append(results_ncRNA,results)\n",
    "}\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "rm(list=setdiff(ls(), c(\"results_ncRNA\", \"output_file_name\",\"arrayid\",\"agds.path\",\"trait\"))); gc()\n",
    "save(results_ncRNA, file = paste0(output_file_name,\"_\",arrayid,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ncRNA_Analysis.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile ncRNA_Analysis.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${trait} ${Train_Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"Height\",\"TC\")){\n",
    "    for(arrayid in 222){\n",
    "        if(arrayid < 19){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 35){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 46){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 56){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 69){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 79){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 89){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 100){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 108){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 117){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 128){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 147){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 156){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 166){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 178){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 191){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 198){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 208){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 214){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 218){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--env trait'=trait,\n",
    "            '--input Train_Null_Model'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/acaf_threshold_v7/acaf_threshold.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/ncRNA_Analysis.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentric_ncRNA\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "}  \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://ncRNA_Analysis.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 10.5 KiB/ 10.5 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/10.5 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp ncRNA_Analysis.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: ncrna-anal--williamsjacr--240802-141249-95\n",
      "  job-name: ncrna-analysis\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15945330792918160714\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12474725655731960867\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10526454129034710544\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2471146011125752468\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16532595984810075093\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6231998838750485626\n",
      "Launched job-id: ncrna-anal--williamsjacr--240802-141249-95\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ncrna-anal--williamsjacr--240802-141249-95' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'ncrna-anal--williamsjacr--240802-141249-95' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 32 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script ncRNA_Analysis.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "code_folding": [
     85,
     184,
     1082,
     1213,
     1344,
     1474,
     1604
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Noncoding_Analysis.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Noncoding_Analysis.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "trait <- commandArgs(TRUE)[2]\n",
    "print(trait)\n",
    "\n",
    "Train_Null_Model <- commandArgs(TRUE)[3]\n",
    "print(Train_Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[4]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[5]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "\n",
    "obj_nullmodel <- get(load(Train_Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\t\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output file name\n",
    "output_file_name <- paste0(OUTPUT_PATH,\"/\",trait,\"_AoU_WGS_Noncoding_Train\")\n",
    "\n",
    "###########################################################\n",
    "#           Main Function \n",
    "###########################################################\n",
    "\n",
    "\n",
    "## gene number in job\n",
    "gene_num_in_array <- 50 \n",
    "group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "\n",
    "chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if(chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "noncoding_longmasks <- c(\"ARID4B\",\"RERE\",\"CRIM1\",\"EML4\",\"EML6\",\"GLI2\",\"ITGB6\",\"FNDC3B\",\"AFAP1\",\"TRIO\",\n",
    "                         \"NR3C1\",\"SGK1\",\"PRDM1\",\"TNS3\",\"GNA12\",\"HOXA2\",\"PLEC\",\"ASAP1\",\"CSGALNACT1\",\"RAPGEF1\",\n",
    "                         \"KLF6\",\"DNAJB12\",\"ABLIM1\",\"ETV6\",\"IGF1R\",\"RMI2\",\"CMIP\",\"HOXB2\",\"SEPT9\",\"SPECC1\",\n",
    "                         \"CIRBP\",\"ADNP\",\"CDC42EP3\",\"EPAS1\",\"MGAT5\")\n",
    "genes_info_chr <- genes_info_chr[!genes_info_chr[,1] %in% noncoding_longmasks,]\n",
    "sub_seq_num <- dim(genes_info_chr)[1]\n",
    "\n",
    "if(groupid < group.num){\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "Gene_Centric_Noncoding_preload <- function(chr,gene_name,category=c(\"all_categories\",\"downstream\",\"upstream\",\"UTR\",\"promoter_CAGE\",\"promoter_DHS\",\"enhancer_CAGE\",\"enhancer_DHS\"),\n",
    "                                           genofile,obj_nullmodel,\n",
    "                                           dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                                           dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                                           dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                                           dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                                           rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                           QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                           Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                           Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                           SPA_p_filter=TRUE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  if(category==\"all_categories\"){\n",
    "    results <- noncoding_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                 dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                 dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                 dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                 dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                 rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                 Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                 Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                 SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"downstream\"){\n",
    "    results <- downstream(chr,gene_name,genofile,obj_nullmodel,\n",
    "                          rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                          QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                          Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                          Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                          SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"upstream\"){\n",
    "    results <- upstream(chr,gene_name,genofile,obj_nullmodel,\n",
    "                        rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                        Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                        SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"UTR\"){\n",
    "    results <- UTR(chr,gene_name,genofile,obj_nullmodel,\n",
    "                   rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                   QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                   Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                   Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                   SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"promoter_CAGE\"){\n",
    "    results <- promoter_CAGE_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                     dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                     rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                     QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                     Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                     Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                     SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"promoter_DHS\"){\n",
    "    results <- promoter_DHS_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                    dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                    rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                    QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                    Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                    Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                    SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"enhancer_CAGE\"){\n",
    "    results <- enhancer_CAGE_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                     dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                     rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                     QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                     Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                     Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                     SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"enhancer_DHS\"){\n",
    "    results <- enhancer_DHS_preload(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                    dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                    rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                    QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                    Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                    Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                    SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "noncoding_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                              dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                              dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                              dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                              dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                              rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                              QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                              Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                              Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                              SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  #####################################\n",
    "  #   Gene Info\n",
    "  ## get SNV id\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ########################################\n",
    "  #   Downstream\n",
    "  ## downstream SNVs\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  is.in <- (GENCODE.Category==\"downstream\")&(SNVlist)\n",
    "  variant.id.downstream <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.downstream,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.downstream)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[,]\")\n",
    "  variant_gene_num <- sapply(GENCODE.Info.split,function(z) length(z))\n",
    "  \n",
    "  variant.id.SNV <- seqGetData(genofile, \"variant.id\")\n",
    "  variant.id.SNV <- rep(variant.id.SNV,variant_gene_num)\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  rm(variant_gene_num)\n",
    "  gc()\n",
    "  \n",
    "  Gene <- as.character(unlist(GENCODE.Info.split))\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(Gene==gene_name)\n",
    "  variant.is.in <- variant.id.SNV[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_downstream <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"downstream\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_downstream <- rbind(results_downstream,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_downstream)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_downstream) <- colnames(results_downstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_downstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_downstream)[(dim(results_downstream)[2]-1):dim(results_downstream)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_downstream) <- colnames(results_downstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_downstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_downstream)[dim(results_downstream)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ########################################\n",
    "  #   Upstream\n",
    "  \n",
    "  is.in <- (GENCODE.Category==\"upstream\")&(SNVlist)\n",
    "  variant.id.upstream <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.upstream,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.upstream)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[,]\")\n",
    "  variant_gene_num <- sapply(GENCODE.Info.split,function(z) length(z))\n",
    "  \n",
    "  variant.id.SNV <- seqGetData(genofile, \"variant.id\")\n",
    "  variant.id.SNV <- rep(variant.id.SNV,variant_gene_num)\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  rm(variant_gene_num)\n",
    "  gc()\n",
    "  \n",
    "  Gene <- as.character(unlist(GENCODE.Info.split))\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(Gene==gene_name)\n",
    "  variant.is.in <- variant.id.SNV[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_upstream <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"upstream\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_upstream <- rbind(results_upstream,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_upstream)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_upstream) <- colnames(results_upstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_upstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_upstream)[(dim(results_upstream)[2]-1):dim(results_upstream)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_upstream) <- colnames(results_upstream, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_upstream)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_upstream)[dim(results_upstream)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ########################################################\n",
    "  #                UTR\n",
    "  \n",
    "  is.in <- ((GENCODE.Category==\"UTR3\")|(GENCODE.Category==\"UTR5\")|(GENCODE.Category==\"UTR5;UTR3\"))&(SNVlist)\n",
    "  variant.id.UTR <- variant.id[is.in]\n",
    "  \n",
    "  rm(GENCODE.Category)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.UTR,sample.id=phenotype.id)\n",
    "  \n",
    "  rm(variant.id.UTR)\n",
    "  gc()\n",
    "  \n",
    "  GENCODE.Info <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "  GENCODE.Info.split <- strsplit(GENCODE.Info, split = \"[(]\")\n",
    "  \n",
    "  rm(GENCODE.Info)\n",
    "  gc()\n",
    "  \n",
    "  # Gene <- as.character(sapply(GENCODE.Info.split,function(z) z[seq(1,length(z),2)]))\n",
    "  Gene <- as.character(sapply(GENCODE.Info.split,function(z) z[1]))\n",
    "  \n",
    "  rm(GENCODE.Info.split)\n",
    "  gc()\n",
    "  \n",
    "  variant.id.SNV <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(Gene==gene_name)\n",
    "  variant.is.in <- variant.id.SNV[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_UTR <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- rep(NA,4)\n",
    "    results_temp[3] <- \"UTR\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_UTR <- rbind(results_UTR,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_UTR)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_UTR) <- colnames(results_UTR, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_UTR)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_UTR)[(dim(results_UTR)[2]-1):dim(results_UTR)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_UTR) <- colnames(results_UTR, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_UTR)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_UTR)[dim(results_UTR)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  #############################################\n",
    "  #   Promoter-CAGE\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_promoter_CAGE <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_promoter_CAGE <- rbind(results_promoter_CAGE,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_promoter_CAGE)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_promoter_CAGE) <- colnames(results_promoter_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_CAGE)[(dim(results_promoter_CAGE)[2]-1):dim(results_promoter_CAGE)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_promoter_CAGE) <- colnames(results_promoter_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_CAGE)[dim(results_promoter_CAGE)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ##################################################\n",
    "  #       Promoter-DHS\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_promoter_DHS <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_promoter_DHS <- rbind(results_promoter_DHS ,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_promoter_DHS)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_promoter_DHS) <- colnames(results_promoter_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_DHS)[(dim(results_promoter_DHS)[2]-1):dim(results_promoter_DHS)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_promoter_DHS) <- colnames(results_promoter_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_promoter_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_promoter_DHS)[dim(results_promoter_DHS)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ###########################################\n",
    "  #        Enhancer-CAGE\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_enhancer_CAGE <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_enhancer_CAGE <- rbind(results_enhancer_CAGE,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_enhancer_CAGE)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_enhancer_CAGE) <- colnames(results_enhancer_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_CAGE)[(dim(results_enhancer_CAGE)[2]-1):dim(results_enhancer_CAGE)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_enhancer_CAGE) <- colnames(results_enhancer_CAGE, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_CAGE)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_CAGE)[dim(results_enhancer_CAGE)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ##################################################\n",
    "  #       Enhancer-DHS\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_enhancer_DHS <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_enhancer_DHS <- rbind(results_enhancer_DHS,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_enhancer_DHS)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results_enhancer_DHS) <- colnames(results_enhancer_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_DHS)[(dim(results_enhancer_DHS)[2]-1):dim(results_enhancer_DHS)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results_enhancer_DHS) <- colnames(results_enhancer_DHS, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_enhancer_DHS)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_enhancer_DHS)[dim(results_enhancer_DHS)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  ############################################\n",
    "  #           results\n",
    "  \n",
    "  results_noncoding <- list(upstream=results_upstream,downstream=results_downstream,UTR=results_UTR,\n",
    "                            promoter_CAGE=results_promoter_CAGE,promoter_DHS=results_promoter_DHS,\n",
    "                            enhancer_CAGE=results_enhancer_CAGE,enhancer_DHS=results_enhancer_DHS)\n",
    "  \n",
    "  return(results_noncoding)\n",
    "  \n",
    "}\n",
    "\n",
    "promoter_CAGE_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                  dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                                  rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                  QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                  Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                  Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                  SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "promoter_DHS_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                 dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                                 rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                 QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                 Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                 SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfPromrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.PromrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfPromrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"promoter_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "enhancer_CAGE_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                  dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                                  rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                  QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                  Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                  Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                  SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerCAGEVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerCAGE[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerCAGEVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_CAGE\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "enhancer_DHS_preload <- function(chr,gene_name,genofile,obj_nullmodel,\n",
    "                                 dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                                 rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                 QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                 Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                 SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA)){\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else{\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ### Gene\n",
    "  is.in <- which(dfHancerrOCRsVarGene.SNV[,5]==gene_name)\n",
    "  variant.is.in <- variant.id.SNV.HancerrOCRs[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.is.in,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno))){\n",
    "    if(dim(Geno)[2]>0){\n",
    "      if(geno_missing_imputation==\"mean\"){\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\"){\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    if(Use_annotation_weights){\n",
    "      for(k in 1:length(Annotation_name)){\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name){\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\"){\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\"){\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1){\n",
    "    if(!use_SPA){\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else{\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\")){\n",
    "    results_temp <- dfHancerrOCRsVarGene.SNV[1,1:4]\n",
    "    results_temp[3] <- \"enhancer_DHS\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(gene_name)\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA){\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else{\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results)){\n",
    "    if(!use_SPA){\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else{\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "gene_centric_noncoding_dnanexus <- function(genes_info_chr,gene_name,chr,genofile,obj_nullmodel,\n",
    "                                            dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE,\n",
    "                                            dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs,\n",
    "                                            dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE,\n",
    "                                            dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs,\n",
    "                                            rare_maf_cutoff,QC_label,variant_type,geno_missing_imputation,\n",
    "                                            Annotation_dir,Annotation_name_catalog,\n",
    "                                            Use_annotation_weights,Annotation_name,silent){\n",
    "  results <- try(Gene_Centric_Noncoding_preload(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,\n",
    "                                                dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                                dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                                dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                                dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                                rare_maf_cutoff=rare_maf_cutoff,QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                                Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=silent),silent=TRUE)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "#########################################################\n",
    "#             Promoter_CAGE\n",
    "#########################################################\n",
    "\n",
    "agds.path <- GDS_File\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "promGobj <- promoters(genes(txdb), upstream = 3000, downstream = 3000)\n",
    "\n",
    "#Subsetting Promoters that within +/-3kb of TSS and have CAGE signals\n",
    "CAGEAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"CAGE\")]))\n",
    "CAGEBvt <- CAGEAnno!=\"\"\n",
    "CAGEidx <- which(CAGEBvt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[CAGEidx])\n",
    "seqSetFilter(genofile,promGobj,intersect=TRUE)\n",
    "CAGEpromgene <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "CAGEGene <- unlist(lapply(strsplit(CAGEpromgene,\"\\\\(|\\\\,|;|-\"),`[[`,1))\n",
    "##obtain variants info\n",
    "CAGEvchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "CAGEvpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "CAGEvref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "CAGEvalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfPromCAGEVarGene <- data.frame(CAGEvchr,CAGEvpos,CAGEvref,CAGEvalt,CAGEGene)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.PromCAGE <- variant.id[SNVlist]\n",
    "\n",
    "dfPromCAGEVarGene.SNV <- dfPromCAGEVarGene[SNVlist,]\n",
    "dfPromCAGEVarGene.SNV$CAGEvpos <- as.character(dfPromCAGEVarGene.SNV$CAGEvpos)\n",
    "dfPromCAGEVarGene.SNV$CAGEvref <- as.character(dfPromCAGEVarGene.SNV$CAGEvref)\n",
    "dfPromCAGEVarGene.SNV$CAGEvalt <- as.character(dfPromCAGEVarGene.SNV$CAGEvalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfPromCAGEVarGene)\n",
    "gc()\n",
    "\n",
    "#########################################################\n",
    "#             Promoter_DHS\n",
    "#########################################################\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "promGobj <- promoters(genes(txdb), upstream = 3000, downstream = 3000)\n",
    "\n",
    "# Subsetting Promoters that within +/-3kb of TSS and have rOCRs signals\n",
    "rOCRsAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"DHS\")]))\n",
    "rOCRsBvt <- rOCRsAnno!=\"\"\n",
    "rOCRsidx <- which(rOCRsBvt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[rOCRsidx])\n",
    "\n",
    "seqSetFilter(genofile,promGobj,intersect=TRUE)\n",
    "rOCRspromgene <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Info\")]))\n",
    "rOCRsGene <- unlist(lapply(strsplit(rOCRspromgene,\"\\\\(|\\\\,|;|-\"),`[[`,1))\n",
    "## obtain variants info\n",
    "rOCRsvchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "rOCRsvpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "rOCRsvref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "rOCRsvalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfPromrOCRsVarGene <- data.frame(rOCRsvchr,rOCRsvpos,rOCRsvref,rOCRsvalt,rOCRsGene)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.PromrOCRs <- variant.id[SNVlist]\n",
    "\n",
    "dfPromrOCRsVarGene.SNV <- dfPromrOCRsVarGene[SNVlist,]\n",
    "dfPromrOCRsVarGene.SNV$rOCRsvpos <- as.character(dfPromrOCRsVarGene.SNV$rOCRsvpos)\n",
    "dfPromrOCRsVarGene.SNV$rOCRsvref <- as.character(dfPromrOCRsVarGene.SNV$rOCRsvref)\n",
    "dfPromrOCRsVarGene.SNV$rOCRsvalt <- as.character(dfPromrOCRsVarGene.SNV$rOCRsvalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfPromrOCRsVarGene)\n",
    "gc()\n",
    "\n",
    "#########################################################\n",
    "#             Enhancer_CAGE\n",
    "#########################################################\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "\n",
    "#Now extract the GeneHancer with CAGE Signal Overlay\n",
    "genehancerAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "genehancer <- genehancerAnno!=\"\"\n",
    "\n",
    "CAGEAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"CAGE\")]))\n",
    "CAGE <- CAGEAnno!=\"\"\n",
    "CAGEGeneHancervt <- CAGEAnno!=\"\"&genehancerAnno!=\"\"\n",
    "CAGEGeneHanceridx <- which(CAGEGeneHancervt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[CAGEGeneHanceridx])\n",
    "\n",
    "# variants that covered by whole GeneHancer without CAGE overlap.\n",
    "genehancerSet <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "enhancerGene <- unlist(lapply(strsplit(genehancerSet,\"=\"),`[[`,4))\n",
    "enhancer2GENE <- unlist(lapply(strsplit(enhancerGene,\";\"),`[[`,1))\n",
    "enhancervchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "enhancervpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "enhancervref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "enhancervalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfHancerCAGEVarGene <- data.frame(enhancervchr,enhancervpos,enhancervref,enhancervalt,enhancer2GENE)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.HancerCAGE <- variant.id[SNVlist]\n",
    "\n",
    "dfHancerCAGEVarGene.SNV <- dfHancerCAGEVarGene[SNVlist,]\n",
    "dfHancerCAGEVarGene.SNV$enhancervpos <- as.character(dfHancerCAGEVarGene.SNV$enhancervpos)\n",
    "dfHancerCAGEVarGene.SNV$enhancervref <- as.character(dfHancerCAGEVarGene.SNV$enhancervref)\n",
    "dfHancerCAGEVarGene.SNV$enhancervalt <- as.character(dfHancerCAGEVarGene.SNV$enhancervalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfHancerCAGEVarGene)\n",
    "gc()\n",
    "\n",
    "#########################################################\n",
    "#             Enhancer_DHS\n",
    "#########################################################\n",
    "\n",
    "varid <- seqGetData(genofile, \"variant.id\")\n",
    "\n",
    "#Now extract the GeneHancer with rOCRs Signal Overlay\n",
    "genehancerAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "genehancer <- genehancerAnno!=\"\"\n",
    "\n",
    "rOCRsAnno <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"DHS\")]))\n",
    "rOCRs <- rOCRsAnno!=\"\"\n",
    "rOCRsGeneHancervt <- rOCRsAnno!=\"\"&genehancerAnno!=\"\"\n",
    "rOCRsGeneHanceridx <- which(rOCRsGeneHancervt,useNames=TRUE)\n",
    "seqSetFilter(genofile,variant.id=varid[rOCRsGeneHanceridx])\n",
    "# variants that covered by whole GeneHancer without rOCRs overlap.\n",
    "\n",
    "genehancerSet <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GeneHancer\")]))\n",
    "enhancerGene <- unlist(lapply(strsplit(genehancerSet,\"=\"),`[[`,4))\n",
    "enhancer2GENE <- unlist(lapply(strsplit(enhancerGene,\";\"),`[[`,1))\n",
    "enhancervchr <- as.numeric(seqGetData(genofile,\"chromosome\"))\n",
    "enhancervpos <- as.numeric(seqGetData(genofile,\"position\"))\n",
    "enhancervref <- as.character(seqGetData(genofile,\"$ref\"))\n",
    "enhancervalt <- as.character(seqGetData(genofile,\"$alt\"))\n",
    "dfHancerrOCRsVarGene <- data.frame(enhancervchr,enhancervpos,enhancervref,enhancervalt,enhancer2GENE)\n",
    "\n",
    "rm(varid)\n",
    "gc()\n",
    "\n",
    "## get SNV id\n",
    "filter <- seqGetData(genofile, QC_label)\n",
    "if(variant_type==\"variant\"){\n",
    "  SNVlist <- filter == \"PASS\"\n",
    "}\n",
    "\n",
    "if(variant_type==\"SNV\"){\n",
    "  SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "}\n",
    "\n",
    "if(variant_type==\"Indel\"){\n",
    "  SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "}\n",
    "\n",
    "variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "variant.id.SNV.HancerrOCRs <- variant.id[SNVlist]\n",
    "\n",
    "dfHancerrOCRsVarGene.SNV <- dfHancerrOCRsVarGene[SNVlist,]\n",
    "dfHancerrOCRsVarGene.SNV$enhancervpos <- as.character(dfHancerrOCRsVarGene.SNV$enhancervpos)\n",
    "dfHancerrOCRsVarGene.SNV$enhancervref <- as.character(dfHancerrOCRsVarGene.SNV$enhancervref)\n",
    "dfHancerrOCRsVarGene.SNV$enhancervalt <- as.character(dfHancerrOCRsVarGene.SNV$enhancervalt)\n",
    "\n",
    "seqResetFilter(genofile)\n",
    "\n",
    "rm(dfHancerrOCRsVarGene)\n",
    "gc()\n",
    "\n",
    "results_noncoding <- list()\n",
    "count <- 1\n",
    "for(kk in sub_seq_id){\n",
    "  print(kk)\n",
    "  gene_name <- genes_info_chr[kk,1]\n",
    "  results <- gene_centric_noncoding_dnanexus(genes_info_chr=genes_info_chr,gene_name = gene_name,chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,\n",
    "                                             dfPromCAGEVarGene.SNV=dfPromCAGEVarGene.SNV,variant.id.SNV.PromCAGE=variant.id.SNV.PromCAGE,\n",
    "                                             dfPromrOCRsVarGene.SNV=dfPromrOCRsVarGene.SNV,variant.id.SNV.PromrOCRs=variant.id.SNV.PromrOCRs,\n",
    "                                             dfHancerCAGEVarGene.SNV=dfHancerCAGEVarGene.SNV,variant.id.SNV.HancerCAGE=variant.id.SNV.HancerCAGE,\n",
    "                                             dfHancerrOCRsVarGene.SNV=dfHancerrOCRsVarGene.SNV,variant.id.SNV.HancerrOCRs=variant.id.SNV.HancerrOCRs,\n",
    "                                             rare_maf_cutoff=0.01,QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                             Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                             Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=TRUE)\n",
    "  \n",
    "  results_noncoding[[count]] <-results\n",
    "  count <- count + 1\n",
    "}\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "rm(list=setdiff(ls(), c(\"results_noncoding\", \"output_file_name\",\"arrayid\",\"agds.path\",\"trait\"))); gc()\n",
    "save(results_noncoding, file = paste0(output_file_name,\"_\",arrayid,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Noncoding_Analysis.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Noncoding_Analysis.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${trait} ${Train_Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"Height\",\"TC\")){\n",
    "    for(arrayid in 379){\n",
    "        if(arrayid < 41){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 66){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 87){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 102){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 120){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 159){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 173){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 189){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 204){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 230){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 250){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 257){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 269){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 281){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 298){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 321){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 327){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 355){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 366){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 371){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--env trait'=trait,\n",
    "            '--input Train_Null_Model'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/acaf_threshold_v7/acaf_threshold.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Noncoding_Analysis.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricNoncoding\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "}  \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "code_folding": [
     33,
     785,
     975,
     1168,
     1644,
     1834,
     2022,
     2229,
     2439,
     3408
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "trait <- commandArgs(TRUE)[2]\n",
    "print(trait)\n",
    "\n",
    "Train_Null_Model <- commandArgs(TRUE)[3]\n",
    "print(Train_Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[4]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[5]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "\n",
    "coding <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                   rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                   QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                   Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                   Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                   SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  rm(position)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ################################################\n",
    "  #           Coding\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.coding <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.EXONIC.Category==\"nonsynonymous SNV\")|(GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.coding]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ################################################\n",
    "  #                  plof_ds\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof_ds <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof_ds <- rbind(results_plof_ds,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof_ds))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[(dim(results_plof_ds)[2]-1):dim(results_plof_ds)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[dim(results_plof_ds)[2]] <- c(\"STAAR-B\")\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #####################################################\n",
    "  #                      plof\n",
    "  #####################################################\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof <- rbind(results_plof,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[(dim(results_plof)[2]-1):dim(results_plof)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[dim(results_plof)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #############################################\n",
    "  #             synonymous\n",
    "  #############################################\n",
    "  lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.synonymous]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.synonymous,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_synonymous <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"synonymous\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_synonymous <- rbind(results_synonymous,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_synonymous))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[(dim(results_synonymous)[2]-1):dim(results_synonymous)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[dim(results_synonymous)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #        missense\n",
    "  #################################################\n",
    "  lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.missense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.missense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #         disruptive missense\n",
    "  #################################################\n",
    "  lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.dmissense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.dmissense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results <- cbind(results,matrix(1,1,6))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }else{\n",
    "          results <- cbind(results,matrix(1,1,2))\n",
    "          colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }\n",
    "      }else\n",
    "      {\n",
    "        results_missense <- c()\n",
    "        results_ds <- results\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,6))\n",
    "          names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT(1,25)\",\"SKAT(1,1)\",\"Burden(1,25)\",\"Burden(1,1)\",\"ACAT-V(1,25)\",\"ACAT-V(1,1)\")]\n",
    "          apc_num <- (length(results_m)-19)/6\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "          results_m[\"STAAR-O\"] <- CCT(as.numeric(results_m[6:length(results_m)][p_seq]))\n",
    "          results_m[\"STAAR-S(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "          results_m[\"STAAR-S(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "          results_m[\"STAAR-B(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "          results_m[\"STAAR-B(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "          results_m[\"STAAR-A(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "          results_m[\"STAAR-A(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }else\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,2))\n",
    "          names(results_m)[(length(results_m)-1):length(results_m)] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-1):length(results_m)] <- results[2,c(\"Burden(1,25)\",\"Burden(1,1)\")]\n",
    "          \n",
    "          ## check whether the p-values is NA. If so, set NA equals 1.\n",
    "          if(is.na(results_m[(length(results_m)-1)]))\n",
    "          {\n",
    "            results_m[(length(results_m)-1)] <- 1\n",
    "          }\n",
    "          \n",
    "          if(is.na(results_m[length(results_m)]))\n",
    "          {\n",
    "            results_m[length(results_m)] <- 1\n",
    "          }\n",
    "          \n",
    "          apc_num <- (length(results_m)-10)/2\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),(length(results_m)-6):(length(results_m)-5))\n",
    "          \n",
    "          ## calculate STAAR-B\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][p_seq])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,25)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num,(length(results_m)-6))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,1)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),(length(results_m)-5))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "    results_ds <- c()\n",
    "  }\n",
    "  \n",
    "  results_coding <- list(plof=results_plof,plof_ds=results_plof_ds,missense=results,disruptive_missense=results_ds,synonymous=results_synonymous)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results_coding)\n",
    "}\n",
    "\n",
    "plof <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                 rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                 QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                 Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                 SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "plof_ds <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                    rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                    QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                    Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                    Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                    SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof_ds\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else\n",
    "    {\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "missense <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                     rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                     QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                     Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                     Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                     SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  ############################################################\n",
    "  #                      missense\n",
    "  ############################################################\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.missense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  ############################################################\n",
    "  #                      disruptive missense\n",
    "  ############################################################\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.dmissense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results <- cbind(results,matrix(1,1,6))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "        }else{\n",
    "          results <- cbind(results,matrix(1,1,2))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "        }\n",
    "      }else\n",
    "      {\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,6))\n",
    "          names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT(1,25)\",\"SKAT(1,1)\",\"Burden(1,25)\",\"Burden(1,1)\",\"ACAT-V(1,25)\",\"ACAT-V(1,1)\")]\n",
    "          apc_num <- (length(results_m)-19)/6\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "          results_m[\"STAAR-O\"] <- CCT(as.numeric(results_m[6:length(results_m)][p_seq]))\n",
    "          results_m[\"STAAR-S(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "          results_m[\"STAAR-S(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "          results_m[\"STAAR-B(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "          results_m[\"STAAR-B(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "          results_m[\"STAAR-A(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "          results_m[\"STAAR-A(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }else\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,2))\n",
    "          names(results_m)[(length(results_m)-1):length(results_m)] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-1):length(results_m)] <- results[2,c(\"Burden(1,25)\",\"Burden(1,1)\")]\n",
    "          \n",
    "          ## check whether the p-values is NA. If so, set NA equals 1.\n",
    "          if(is.na(results_m[(length(results_m)-1)]))\n",
    "          {\n",
    "            results_m[(length(results_m)-1)] <- 1\n",
    "          }\n",
    "          \n",
    "          if(is.na(results_m[length(results_m)]))\n",
    "          {\n",
    "            results_m[length(results_m)] <- 1\n",
    "          }\n",
    "          \n",
    "          apc_num <- (length(results_m)-10)/2\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),(length(results_m)-6):(length(results_m)-5))\n",
    "          \n",
    "          ## calculate STAAR-B\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][p_seq])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,25)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num,(length(results_m)-6))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,1)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),(length(results_m)-5))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "          \n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "disruptive_missense <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                                rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## disruptive_missense\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.ds <- ((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene <- variant.id.gene[lof.in.ds]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "synonymous <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                       rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                       QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                       Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                       Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                       SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ### synonymous\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.synonymous]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"synonymous\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "ptv <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel\n",
    "  }\n",
    "  \n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "ptv_ds <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                   rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                   QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                   Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                   Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                   SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof_ds\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else\n",
    "    {\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "coding_incl_ptv <- function(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                            rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                            QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                            Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                            Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                            SPA_p_filter=FALSE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  n_pheno <- obj_nullmodel$n.pheno\n",
    "  \n",
    "  ## SPA status\n",
    "  if(!is.null(obj_nullmodel$use_SPA))\n",
    "  {\n",
    "    use_SPA <- obj_nullmodel$use_SPA\n",
    "  }else\n",
    "  {\n",
    "    use_SPA <- FALSE\n",
    "  }\n",
    "  \n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  rm(position)\n",
    "  gc()\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ################################################\n",
    "  #           Coding\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.coding <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.EXONIC.Category==\"nonsynonymous SNV\")|(GENCODE.EXONIC.Category==\"synonymous SNV\")|(GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "  variant.id.gene <- variant.id.gene[lof.in.coding]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "        {\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "          Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "          \n",
    "          if(Annotation_name[k]==\"CADD\")\n",
    "          {\n",
    "            Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "          }\n",
    "          \n",
    "          if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "          {\n",
    "            Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "            Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "          }\n",
    "          Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ################################################\n",
    "  #                  plof_ds\n",
    "  ################################################\n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof_ds <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof_ds <- rbind(results_plof_ds,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof_ds))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[(dim(results_plof_ds)[2]-1):dim(results_plof_ds)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof_ds)[dim(results_plof_ds)[2]] <- c(\"STAAR-B\")\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ################################################\n",
    "  #                  ptv_ds\n",
    "  ################################################\n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "  }\n",
    "  \n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_ptv_ds <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv_ds\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_ptv_ds <- rbind(results_ptv_ds,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_ptv_ds))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_ptv_ds) <- colnames(results_ptv_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv_ds)[(dim(results_ptv_ds)[2]-1):dim(results_ptv_ds)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_ptv_ds) <- colnames(results_ptv_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv_ds)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv_ds)[dim(results_ptv_ds)[2]] <- c(\"STAAR-B\")\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #####################################################\n",
    "  #                      plof\n",
    "  #####################################################\n",
    "  lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_plof <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"plof\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_plof <- rbind(results_plof,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_plof))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[(dim(results_plof)[2]-1):dim(results_plof)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_plof)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_plof)[dim(results_plof)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #####################################################\n",
    "  #                      ptv\n",
    "  #####################################################\n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\")\n",
    "  {\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"variant\")\n",
    "  {\n",
    "    lof.in.plof.snv <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")\n",
    "    lof.in.plof.indel <- (GENCODE.EXONIC.Category==\"frameshift deletion\")|(GENCODE.EXONIC.Category==\"frameshift insertion\")\n",
    "    lof.in.plof <- lof.in.plof.snv|lof.in.plof.indel\n",
    "  }\n",
    "  \n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_ptv <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"ptv\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_ptv <- rbind(results_ptv,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_ptv))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_ptv) <- colnames(results_ptv, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv)[(dim(results_ptv)[2]-1):dim(results_ptv)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_ptv) <- colnames(results_ptv, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_ptv)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_ptv)[dim(results_ptv)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #############################################\n",
    "  #             synonymous\n",
    "  #############################################\n",
    "  lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.synonymous]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.synonymous,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results_synonymous <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"synonymous\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results_synonymous <- rbind(results_synonymous,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results_synonymous))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[(dim(results_synonymous)[2]-1):dim(results_synonymous)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results_synonymous)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results_synonymous)[dim(results_synonymous)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #        missense\n",
    "  #################################################\n",
    "  lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.missense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.missense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  results <- c()\n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  #################################################\n",
    "  #         disruptive missense\n",
    "  #################################################\n",
    "  lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "  variant.id.gene.category <- variant.id.gene[lof.in.dmissense]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.dmissense,]\n",
    "  \n",
    "  pvalues <- 0\n",
    "  if(n_pheno == 1)\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      try(pvalues <- STAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "    }else{\n",
    "      try(pvalues <- STAAR_Binary_SPA(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff),silent=silent)\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    try(pvalues <- MultiSTAAR(Geno,obj_nullmodel,Anno.Int.PHRED.sub.category,rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff),silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(inherits(pvalues, \"list\"))\n",
    "  {\n",
    "    results_temp <- as.vector(genes[kk,])\n",
    "    results_temp[3] <- \"disruptive_missense\"\n",
    "    results_temp[2] <- chr\n",
    "    results_temp[1] <- as.character(genes[kk,1])\n",
    "    results_temp[4] <- pvalues$num_variant\n",
    "    \n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,pvalues$results_STAAR_S_1_25,pvalues$results_STAAR_S_1_1,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_A_1_25,\n",
    "                        pvalues$results_STAAR_A_1_1,pvalues$results_ACAT_O,pvalues$results_STAAR_O)\n",
    "    }else\n",
    "    {\n",
    "      results_temp <- c(results_temp,pvalues$cMAC,\n",
    "                        pvalues$results_STAAR_B_1_25,pvalues$results_STAAR_B_1_1,pvalues$results_STAAR_B)\n",
    "    }\n",
    "    \n",
    "    results <- rbind(results,results_temp)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    if(!use_SPA)\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O\",\"STAAR-O\")\n",
    "    }else\n",
    "    {\n",
    "      colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "      colnames(results)[1:5] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"cMAC\")\n",
    "      colnames(results)[dim(results)[2]] <- c(\"STAAR-B\")\n",
    "    }\n",
    "    \n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results <- cbind(results,matrix(1,1,6))\n",
    "          colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }else{\n",
    "          results <- cbind(results,matrix(1,1,2))\n",
    "          colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_missense <- results\n",
    "          results_ds <- c()\n",
    "        }\n",
    "      }else\n",
    "      {\n",
    "        results_missense <- c()\n",
    "        results_ds <- results\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        if(!use_SPA)\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,6))\n",
    "          names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT(1,25)-Disruptive\",\"SKAT(1,1)-Disruptive\",\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\",\"ACAT-V(1,25)-Disruptive\",\"ACAT-V(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT(1,25)\",\"SKAT(1,1)\",\"Burden(1,25)\",\"Burden(1,1)\",\"ACAT-V(1,25)\",\"ACAT-V(1,1)\")]\n",
    "          apc_num <- (length(results_m)-19)/6\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "          results_m[\"STAAR-O\"] <- CCT(as.numeric(results_m[6:length(results_m)][p_seq]))\n",
    "          results_m[\"STAAR-S(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "          results_m[\"STAAR-S(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "          results_m[\"STAAR-B(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "          results_m[\"STAAR-B(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "          results_m[\"STAAR-A(1,25)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "          results_m[\"STAAR-A(1,1)\"] <- CCT(as.numeric(results_m[6:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }else\n",
    "        {\n",
    "          results_m <- c(results[1,],rep(0,2))\n",
    "          names(results_m)[(length(results_m)-1):length(results_m)] <- c(\"Burden(1,25)-Disruptive\",\"Burden(1,1)-Disruptive\")\n",
    "          results_m[(length(results_m)-1):length(results_m)] <- results[2,c(\"Burden(1,25)\",\"Burden(1,1)\")]\n",
    "          \n",
    "          ## check whether the p-values is NA. If so, set NA equals 1.\n",
    "          if(is.na(results_m[(length(results_m)-1)]))\n",
    "          {\n",
    "            results_m[(length(results_m)-1)] <- 1\n",
    "          }\n",
    "          \n",
    "          if(is.na(results_m[length(results_m)]))\n",
    "          {\n",
    "            results_m[length(results_m)] <- 1\n",
    "          }\n",
    "          \n",
    "          apc_num <- (length(results_m)-10)/2\n",
    "          p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),(length(results_m)-6):(length(results_m)-5))\n",
    "          \n",
    "          ## calculate STAAR-B\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][p_seq])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,25)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num,(length(results_m)-6))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,25)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          ## calculate STAAR-B(1,1)\n",
    "          pvalues_sub <- as.numeric(results_m[6:length(results_m)][c(1:apc_num+(apc_num+1),(length(results_m)-5))])\n",
    "          if(sum(is.na(pvalues_sub))>0)\n",
    "          {\n",
    "            if(sum(is.na(pvalues_sub))==length(pvalues_sub))\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }else\n",
    "            {\n",
    "              ## not all NAs\n",
    "              pvalues_sub <- pvalues_sub[!is.na(pvalues_sub)]\n",
    "              if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "              {\n",
    "                ## not all ones\n",
    "                results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "                \n",
    "              }else\n",
    "              {\n",
    "                results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "                \n",
    "              }\n",
    "            }\n",
    "          }else\n",
    "          {\n",
    "            if(sum(pvalues_sub[pvalues_sub<1])>0)\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- CCT(pvalues_sub[pvalues_sub<1])\n",
    "            }else\n",
    "            {\n",
    "              results_m[\"STAAR-B(1,1)\"] <- 1\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          results_ds <- c()\n",
    "          results_ds <- rbind(results_ds,results[2,])\n",
    "          \n",
    "          results <- c()\n",
    "          results <- rbind(results,results_m)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "    results_ds <- c()\n",
    "  }\n",
    "  \n",
    "  results_coding <- list(plof=results_plof,plof_ds=results_plof_ds,missense=results,disruptive_missense=results_ds,synonymous=results_synonymous,ptv=results_ptv,ptv_ds=results_ptv_ds)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(results_coding)\n",
    "}\n",
    "\n",
    "Gene_Centric_Coding <- function(chr,gene_name,category=c(\"all_categories\",\"plof\",\"plof_ds\",\"missense\",\"disruptive_missense\",\"synonymous\",\"ptv\",\"ptv_ds\",\"all_categories_incl_ptv\"),\n",
    "                                genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                SPA_p_filter=TRUE,p_filter_cutoff=0.05,silent=FALSE){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  genes <- genes_info[genes_info[,2]==chr,]\n",
    "  \n",
    "  if(category==\"all_categories\")\n",
    "  {\n",
    "    results <- coding(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                      rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                      QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                      Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                      Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                      SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"plof\")\n",
    "  {\n",
    "    results <- plof(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                    rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                    QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                    Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                    Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                    SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"plof_ds\")\n",
    "  {\n",
    "    results <- plof_ds(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                       rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                       QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                       Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                       Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                       SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"missense\")\n",
    "  {\n",
    "    results <- missense(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                        rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                        QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                        Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                        Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                        SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"disruptive_missense\")\n",
    "  {\n",
    "    results <- disruptive_missense(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                                   rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                                   QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                   Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                   Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                                   SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  if(category==\"synonymous\")\n",
    "  {\n",
    "    results <- synonymous(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                          rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                          QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                          Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                          Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                          SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(category==\"ptv\")\n",
    "  {\n",
    "    results <- ptv(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                   rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                   QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                   Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                   Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                   SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(category==\"ptv_ds\")\n",
    "  {\n",
    "    results <- ptv_ds(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                      rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                      QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                      Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                      Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                      SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  if(category==\"all_categories_incl_ptv\")\n",
    "  {\n",
    "    results <- coding_incl_ptv(chr,gene_name,genofile,obj_nullmodel,genes,\n",
    "                               rare_maf_cutoff=rare_maf_cutoff,rv_num_cutoff=rv_num_cutoff,\n",
    "                               QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                               Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                               Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,\n",
    "                               SPA_p_filter=SPA_p_filter,p_filter_cutoff=p_filter_cutoff,silent=silent)\n",
    "  }\n",
    "  \n",
    "  return(results)\n",
    "}\n",
    "\n",
    "obj_nullmodel <- get(load(Train_Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output file name\n",
    "output_file_name <- paste0(OUTPUT_PATH,\"/\",trait,\"_AoU_WGS_Coding_Train\")\n",
    "\n",
    "## Chr\n",
    "gene_num_in_array <- 50 \n",
    "group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "sum(group.num.allchr)\n",
    "\n",
    "chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if (chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "coding_longmasks <- c(\"TTN\",\"PCDHA2\",\"PCDHA3\",\"PCDHGA1\",\"PCDHGA2\",\"PCDHGA3\",\"PCDHGB1\",\"PCDHGA4\",\"PCDHGB2\",\"PCDHGB3\")\n",
    "genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "genes_info_chr <- genes_info_chr[!genes_info_chr[,1] %in% coding_longmasks,]\n",
    "sub_seq_num <- dim(genes_info_chr)[1]\n",
    "\n",
    "if(groupid < group.num){ \n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "agds.path <- GDS_File\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "gene_centric_coding_dnanexus <- function(genes_info_chr,gene_name,chr,genofile,obj_nullmodel,rare_maf_cutoff,\n",
    "                                         QC_label,variant_type,geno_missing_imputation,\n",
    "                                         Annotation_dir,Annotation_name_catalog,\n",
    "                                         Use_annotation_weights,Annotation_name,silent){\n",
    "  results <- try(Gene_Centric_Coding(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                     QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                     Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                     Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=silent),silent=TRUE)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "results_coding <- list()\n",
    "count <- 1\n",
    "for(kk in sub_seq_id){\n",
    "  print(kk)\n",
    "  gene_name <- genes_info_chr[kk,1]\n",
    "  results <- gene_centric_coding_dnanexus(genes_info_chr=genes_info_chr,gene_name = gene_name,chr=chr,genofile=genofile,obj_nullmodel=obj_nullmodel,rare_maf_cutoff=0.01,\n",
    "                                          QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                          Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                          Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name,silent=TRUE)\n",
    "  \n",
    "  results_coding[[count]] <- results\n",
    "  count <- count + 1\n",
    "}\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "rm(list=setdiff(ls(), c(\"results_coding\", \"output_file_name\",\"arrayid\",\"agds.path\",\"trait\"))); gc()\n",
    "save(results_coding, file = paste0(output_file_name,\"_\",arrayid,\".Rdata\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${trait} ${Train_Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"Height\")){\n",
    "    for(arrayid in c(86,112,113,332)){\n",
    "        if(arrayid < 41){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 66){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 87){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 102){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 120){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 159){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 173){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 189){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 204){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 230){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 250){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 257){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 269){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 281){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 298){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 321){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 327){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 355){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 366){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 371){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--env trait'=trait,\n",
    "            '--input Train_Null_Model'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/exome.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Coding_Analysis.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "}  \n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Coding_Analysis.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][129.1 KiB/129.1 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/129.1 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Coding_Analysis.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: coding-ana--williamsjacr--240805-152344-37\n",
      "  job-name: coding-analysis\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/17061318534041850195\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/840692259032739568\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9318378643191275524\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15924022795533132610\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3860031643969366395\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18348738718336721892\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10678717588606309768\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9692262135783927507\n",
      "Launched job-id: coding-ana--williamsjacr--240805-152344-37\n",
      "8 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-152344-37' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-152344-37' --users 'williamsjacr'\n",
      "Waiting for job to complete...\n",
      "Monitoring for failed tasks to retry...\n",
      "*** This dsub process must continue running to retry failed tasks.\n",
      "  coding-ana--williamsjacr--240805-152344-37.6 (attempt 1) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/2964234674434847443\n",
      "  coding-ana--williamsjacr--240805-152344-37.3 (attempt 1) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11529758471240069288\n",
      "  coding-ana--williamsjacr--240805-152344-37.2 (attempt 1) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/8349154917546125030\n",
      "  coding-ana--williamsjacr--240805-152344-37.3 (attempt 2) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7701904607760559155\n",
      "  coding-ana--williamsjacr--240805-152344-37.3 (attempt 3) failed. Retrying.\n",
      "  Failure message: worker was terminated\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7912129358242913412\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Coding_Analysis.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis_Summary.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis_Summary.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "library(data.table)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[1]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "gene_centric_coding_jobs_num <- 379\n",
    "\n",
    "trait <- \"BMI\"\n",
    "i <- 1\n",
    "\n",
    "for(trait in c(\"BMI\",\"TC\",\"HDL\",\"LDL\",\"logTG\",\"Height\")){\n",
    "  gene_centric_results_name <- paste0(trait,\"_AoU_WGS_Coding_Train\")\n",
    "  \n",
    "  results_coding_genome <- NULL\n",
    "  for (i in 1:gene_centric_coding_jobs_num){\n",
    "    results <- get(load(paste0(INPUT_PATH,\"/\",gene_centric_results_name,\"_\",i,\".Rdata\")))\n",
    "    results_coding_genome <- c(results_coding_genome, results)\n",
    "  }\n",
    "  \n",
    "  combine_jake <- function(x){\n",
    "    a <- data.frame(Gene = c(unlist(x$plof[,c(\"Gene name\")]),unlist(x$plof_ds[,c(\"Gene name\")]),unlist(x$missense[,c(\"Gene name\")]),unlist(x$disruptive_missense[,c(\"Gene name\")]),unlist(x$synonymous[,c(\"Gene name\")])),\n",
    "                    Chr = c(unlist(x$plof[,c(\"Chr\")]),unlist(x$plof_ds[,c(\"Chr\")]),unlist(x$missense[,c(\"Chr\")]),unlist(x$disruptive_missense[,c(\"Chr\")]),unlist(x$synonymous[,c(\"Chr\")])),\n",
    "                    Category = c(unlist(x$plof[,c(\"Category\")]),unlist(x$plof_ds[,c(\"Category\")]),unlist(x$missense[,c(\"Category\")]),unlist(x$disruptive_missense[,c(\"Category\")]),unlist(x$synonymous[,c(\"Category\")])),\n",
    "                    Number_SNV = c(unlist(x$plof[,c(\"#SNV\")]),unlist(x$plof_ds[,c(\"#SNV\")]),unlist(x$missense[,c(\"#SNV\")]),unlist(x$disruptive_missense[,c(\"#SNV\")]),unlist(x$synonymous[,c(\"#SNV\")])),\n",
    "                    Burden_1_1 = c(unlist(x$plof[,c(\"Burden(1,1)\")]),unlist(x$plof_ds[,c(\"Burden(1,1)\")]),unlist(x$missense[,c(\"Burden(1,1)\")]),unlist(x$disruptive_missense[,c(\"Burden(1,1)\")]),unlist(x$synonymous[,c(\"Burden(1,1)\")])),\n",
    "                    STAARB = c(unlist(x$plof[,c(\"STAAR-B(1,1)\")]),unlist(x$plof_ds[,c(\"STAAR-B(1,1)\")]),unlist(x$missense[,c(\"STAAR-B(1,1)\")]),unlist(x$disruptive_missense[,c(\"STAAR-B(1,1)\")]),unlist(x$synonymous[,c(\"STAAR-B(1,1)\")])))\n",
    "    return(a)\n",
    "  }\n",
    "  \n",
    "  results_coding_genome <- lapply(results_coding_genome, combine_jake)\n",
    "  results_coding_genome <- rbindlist(results_coding_genome)\n",
    "  results_coding_genome$Number_SNV <- as.numeric(results_coding_genome$Number_SNV)\n",
    "  results_coding_genome <- results_coding_genome[results_coding_genome$Number_SNV < 2000,]\n",
    "  \n",
    "  write.csv(results_coding_genome,file = paste0(OUTPUT_PATH,\"/\",trait,\"_coding_sig.csv\"),row.names = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Coding_Analysis_Summary.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Coding_Analysis_Summary.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${INPUT_PATH} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "            '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Coding_Analysis_Summary.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding\",\n",
    "            check.names = FALSE\n",
    "        ))\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Coding_Analysis_Summary.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  2.3 KiB/  2.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.3 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Coding_Analysis_Summary.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: coding-ana--williamsjacr--240805-203638-20\n",
      "  job-name: coding-analysis-summary\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/18328105492588917735\n",
      "Launched job-id: coding-ana--williamsjacr--240805-203638-20\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-203638-20' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'coding-ana--williamsjacr--240805-203638-20' --users 'williamsjacr'\n",
      "Waiting for job to complete...\n",
      "Monitoring for failed tasks to retry...\n",
      "*** This dsub process must continue running to retry failed tasks.\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Coding_Analysis_Summary.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     19
    ]
   },
   "outputs": [],
   "source": [
    "%%writefile Single_RareVariant_PRS.R\n",
    "rm(list = ls())\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "Gene_Centric_Coding_G_Star <- function(chr,gene_name,category=c(\"plof\",\"plof_ds\",\"missense\",\"disruptive_missense\",\"synonymous\"),\n",
    "                                       genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                       QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                       Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,silent=FALSE){\n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  genes <- genes_info[genes_info[,2]==chr,]  \n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category  <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")  \n",
    "  \n",
    "  if(category == \"plof\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"plof_ds\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"missense\"){\n",
    "    lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.missense]\n",
    "  }else if(category == \"disruptive_missense\"){\n",
    "    lof.in.ds <- ((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.ds]\n",
    "  }else{\n",
    "    lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.synonymous]\n",
    "  }\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  genotype <- Geno\n",
    "  \n",
    "  if(dim(genotype)[2] == 1){\n",
    "    return(matrix(0,nrow = dim(genotype)[1],ncol = 1))\n",
    "  }\n",
    "  \n",
    "  if(!is.null(attr(class(genotype), \"package\")) && attr(class(genotype), \"package\") == \"Matrix\"){\n",
    "    genotype <- as.matrix(genotype)\n",
    "  }\n",
    "  genotype <- matrix_flip(genotype)\n",
    "  MAF <- genotype$MAF\n",
    "  RV_label <- as.vector((MAF<rare_maf_cutoff)&(MAF>0))\n",
    "  Geno_rare <- genotype$Geno[,RV_label]\n",
    "  G <- Geno_rare\n",
    "  rm(Geno_rare)\n",
    "  gc()\n",
    "  \n",
    "  if(is.null(dim(G))){\n",
    "    G <- matrix(G,ncol = 1)\n",
    "  }\n",
    "  \n",
    "  C <- G%*%matrix(1,nrow=ncol(G),ncol = 1)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(C)\n",
    "}\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Train_Pvals <- commandArgs(TRUE)[2]\n",
    "print(Train_Pvals)\n",
    "\n",
    "Train_NullModel <- commandArgs(TRUE)[3]\n",
    "print(Train_NullModel)\n",
    "\n",
    "Tune_NullModel <- commandArgs(TRUE)[4]\n",
    "print(Tune_NullModel)\n",
    "\n",
    "Validation_NullModel <- commandArgs(TRUE)[5]\n",
    "print(Validation_NullModel)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[6]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[8]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "all_train_file <- commandArgs(TRUE)[9]\n",
    "print(all_train_file)\n",
    "\n",
    "all_tune_file <- commandArgs(TRUE)[10]\n",
    "print(all_tune_file)\n",
    "\n",
    "all_validation_file <- commandArgs(TRUE)[11]\n",
    "print(all_validation_file)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[12]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Train_PVals_All <- read.csv(Train_Pvals)\n",
    "Train_PVals_All <- Train_PVals_All[Train_PVals_All$STAARB <= 1e-03,]\n",
    "\n",
    "## Null Model\n",
    "obj_nullmodel_train <- get(load(Train_NullModel))\n",
    "obj_nullmodel_tune <- get(load(Tune_NullModel))\n",
    "obj_nullmodel_validation <- get(load(Validation_NullModel))\n",
    "\n",
    "obj_nullmodel <- obj_nullmodel_train\n",
    "obj_nullmodel$id_include <- c(obj_nullmodel_train$id_include,obj_nullmodel_tune$id_include,obj_nullmodel_validation$id_include)\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "\n",
    "G_star_gene_centric_coding <- list()\n",
    "\n",
    "for(i in 1:nrow(Train_PVals_All)){\n",
    "  ## Chr\n",
    "  chr <- Train_PVals_All$Chr[i]\n",
    "  ## Gene name\n",
    "  gene_name <- Train_PVals_All$Gene[i]\n",
    "  ## Coding mask\n",
    "  category <- Train_PVals_All$Category[i]\n",
    "  \n",
    "  ### gds file\n",
    "  gds.path <- paste0(INPUT_PATH,\"/exome.chr\",chr,\".gds\")\n",
    "  genofile <- seqOpen(gds.path)\n",
    "  \n",
    "  G_star_gene_centric_coding[[i]] <- Gene_Centric_Coding_G_Star(chr=chr,gene_name=gene_name,category=category ,\n",
    "                                                                genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                                                QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,silent=FALSE) \n",
    "  seqClose(genofile) \n",
    "} \n",
    "\n",
    "G_star_gene_centric_coding <- do.call(cbind,G_star_gene_centric_coding)\n",
    "\n",
    "col_remove <- apply(G_star_gene_centric_coding,2,function(x){sum(x != 0)}) > 10 & colSums(G_star_gene_centric_coding) > 10 \n",
    "G_star_gene_centric_coding <- G_star_gene_centric_coding[,col_remove,drop = FALSE]\n",
    "\n",
    "Train_PVals_All <- Train_PVals_All[col_remove,]\n",
    "\n",
    "ids_gstar <- obj_nullmodel$id_include\n",
    "\n",
    "G_star_gene_centric_coding_train <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_train$id_include,]\n",
    "G_star_gene_centric_coding_tune <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_tune$id_include,]\n",
    "G_star_gene_centric_coding_vad <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_validation$id_include,]\n",
    "\n",
    "rm(G_star_gene_centric_coding)\n",
    "\n",
    "X_train <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_train$id_include],G_star_gene_centric_coding_train)\n",
    "pheno_train <- read.delim(all_train_file)\n",
    "pheno_train <- inner_join(pheno_train,X_train)\n",
    "print(\"Made it\")\n",
    "X_train <- as.matrix(pheno_train[,22:ncol(pheno_train),drop = FALSE])\n",
    "print(\"Made it\")\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_train)\n",
    "y_train <- model.null$residual\n",
    "pheno_train$y_train <- NA\n",
    "pheno_train$y_train[!is.na(pheno_train[,trait])] <- y_train\n",
    "\n",
    "X_tune <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_tune$id_include],G_star_gene_centric_coding_tune)\n",
    "pheno_tune <- read.delim(all_tune_file)\n",
    "pheno_tune <- inner_join(pheno_tune,X_tune)\n",
    "X_tune <- as.matrix(pheno_tune[,22:ncol(pheno_tune),drop = FALSE])\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tune)\n",
    "y_tune <- model.null$residual\n",
    "pheno_tune$y_tune <- NA\n",
    "pheno_tune$y_tune[!is.na(pheno_tune[,trait])] <- y_tune\n",
    "\n",
    "X_valid <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_validation$id_include],G_star_gene_centric_coding_vad)\n",
    "pheno_valid <- read.delim(all_validation_file)\n",
    "pheno_valid <- inner_join(pheno_valid,X_valid)\n",
    "X_valid <- as.matrix(pheno_valid[,22:ncol(pheno_valid),drop = FALSE])\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_valid)\n",
    "y_valid <- model.null$residual\n",
    "pheno_valid$y_validation <- NA\n",
    "pheno_valid$y_validation[!is.na(pheno_valid[,trait])] <- y_valid\n",
    "\n",
    "print(\"Made it\")\n",
    "\n",
    "lasso_train <- glmnet(X_train,pheno_train[,\"y_train\"],family = \"gaussian\",alpha = 1)\n",
    "ridge_train <- glmnet(X_train,pheno_train[,\"y_train\"],family = \"gaussian\",alpha = 0)\n",
    "lm_train <- lm.fit(cbind(1,X_train),pheno_train[,\"y_train\"])\n",
    "lm_train$coefficients[is.na(lm_train$coefficients)] <- 0\n",
    "\n",
    "beta_matrix <- cbind(as.matrix(lasso_train$beta),as.matrix(ridge_train$beta),lm_train$coefficients[-1])\n",
    "beta_matrix <- as.data.frame(beta_matrix)\n",
    "colnames(beta_matrix) <- c(paste0(\"lasso_prs\",1:ncol(as.matrix(lasso_train$beta))),paste0(\"ridge_prs\",1:ncol(as.matrix(ridge_train$beta))),paste0(\"lm_prs\",1))\n",
    "\n",
    "beta_matrix <- cbind(Train_PVals_All[,c(1:4)],beta_matrix)\n",
    "\n",
    "print(\"Made it\")\n",
    "\n",
    "lasso_prs_tune <- predict(lasso_train,X_tune)\n",
    "ridge_prs_tune <- predict(ridge_train,X_tune)\n",
    "lm_prs_tune <- as.numeric(cbind(1,X_tune)%*%matrix(lm_train$coefficients,ncol = 1))\n",
    "\n",
    "lasso_prs_vad <- predict(lasso_train,X_valid)\n",
    "ridge_prs_vad <- predict(ridge_train,X_valid)\n",
    "lm_prs_vad <- as.numeric(cbind(1,X_valid)%*%matrix(lm_train$coefficients,ncol = 1))\n",
    "\n",
    "\n",
    "lasso_tune_dat <- data.frame(y = pheno_tune[,\"y_tune\"],lasso_prs_tune)\n",
    "colnames(lasso_tune_dat) <- c(\"y\",paste0(\"lasso_prs\",1:(ncol(lasso_tune_dat) - 1)))\n",
    "lasso_valid_dat <- data.frame(y = pheno_valid[,\"y_validation\"],lasso_prs_vad)\n",
    "colnames(lasso_valid_dat) <- c(\"y\",paste0(\"lasso_prs\",1:(ncol(lasso_valid_dat) - 1)))\n",
    "\n",
    "\n",
    "ridge_tune_dat <- data.frame(y = pheno_tune[,\"y_tune\"],ridge_prs_tune)\n",
    "colnames(ridge_tune_dat) <- c(\"y\",paste0(\"ridge_prs\",1:(ncol(ridge_tune_dat) - 1)))\n",
    "ridge_valid_dat <- data.frame(y = pheno_valid[,\"y_validation\"],ridge_prs_vad)\n",
    "colnames(ridge_valid_dat) <- c(\"y\",paste0(\"ridge_prs\",1:(ncol(ridge_valid_dat) - 1)))\n",
    "\n",
    "lm_tune_dat <- data.frame(y = pheno_tune[,\"y_tune\"],lm_prs_tune)\n",
    "colnames(lm_tune_dat) <- c(\"y\",paste0(\"lm_prs\",1:(ncol(lm_tune_dat) - 1)))\n",
    "lm_valid_dat <- data.frame(y = pheno_valid[,\"y_validation\"],lm_prs_vad)\n",
    "colnames(lm_valid_dat) <- c(\"y\",paste0(\"lm_prs\",1:(ncol(lm_valid_dat) - 1)))\n",
    "\n",
    "\n",
    "all_prs_tune <- cbind(lasso_prs_tune,ridge_prs_tune,lm_prs_tune)\n",
    "colnames(all_prs_tune) <- c(paste0(\"lasso_prs\",1:ncol(lasso_prs_tune)),paste0(\"ridge_prs\",1:ncol(ridge_prs_tune)),\"lm_prs1\")\n",
    "all_prs_valid <- cbind(lasso_prs_vad,ridge_prs_vad,lm_prs_vad)\n",
    "colnames(all_prs_valid) <- c(paste0(\"lasso_prs\",1:ncol(lasso_prs_vad)),paste0(\"ridge_prs\",1:ncol(ridge_prs_vad)),\"lm_prs1\")\n",
    "\n",
    "all_prs_tune <- as.data.frame(all_prs_tune)\n",
    "all_prs_valid <- as.data.frame(all_prs_valid)\n",
    "\n",
    "mtx <- cor(all_prs_tune)\n",
    "drop <- names(all_prs_tune)[apply(mtx,2,function(x){sum(is.na(x))}) == (nrow(mtx) - 1)]\n",
    "\n",
    "all_prs_tune <- dplyr::select(all_prs_tune, -c(drop))\n",
    "all_prs_valid <- dplyr::select(all_prs_valid, -c(drop))\n",
    "\n",
    "mtx <- cor(all_prs_tune)\n",
    "drop <- findCorrelation(mtx,cutoff=0.98)\n",
    "drop <- names(all_prs_tune)[drop]\n",
    "\n",
    "all_prs_tune <- dplyr::select(all_prs_tune, -c(drop))\n",
    "all_prs_valid <- dplyr::select(all_prs_valid, -c(drop))\n",
    "\n",
    "drop <- findLinearCombos(all_prs_tune)$remove\n",
    "drop <- names(data.frame(all_prs_tune))[drop]\n",
    "\n",
    "all_prs_tune <- dplyr::select(all_prs_tune, -c(drop))\n",
    "all_prs_valid <- dplyr::select(all_prs_valid, -c(drop))\n",
    "\n",
    "print(\"Made it\")\n",
    "\n",
    "\n",
    "  Ensemble_Function_Continuous <- function(x,y){\n",
    "    x <- as.matrix(x[!is.na(y),])\n",
    "    y <- y[!is.na(y)]\n",
    "    \n",
    "    lasso_train <- glmnet(x,y,family = \"gaussian\",alpha = 1)\n",
    "    ridge_train <- glmnet(x,y,family = \"gaussian\",alpha = 0)\n",
    "    \n",
    "    lasso_prs_tune <- predict(lasso_train,x)\n",
    "    ridge_prs_tune <- predict(ridge_train,x)\n",
    "    \n",
    "    all <- cbind(lasso_prs_tune,ridge_prs_tune)\n",
    "    \n",
    "    R2_Vector <- vector()\n",
    "    for(i in 1:ncol(all)){\n",
    "      tmp <- data.frame(y = y, x_try = all[,i])\n",
    "      R2_Vector[i] <- summary(lm(y~x_try,data = tmp))$r.square\n",
    "    }\n",
    "    \n",
    "    coefficients_x <- coef(lm(y~.,data.frame(y = all[,which.max(R2_Vector)],x)))\n",
    "    return(list(Coefficients = coefficients_x))\n",
    "  }\n",
    "  Ensemble_Function_Binary <- function(x,y){\n",
    "    x <- as.matrix(x[!is.na(y),])\n",
    "    y <- y[!is.na(y)]\n",
    "    \n",
    "    lasso_train <- glmnet(x,y,family = \"binomial\",alpha = 1)\n",
    "    ridge_train <- glmnet(x,y,family = \"binomial\",alpha = 0)\n",
    "    \n",
    "    lasso_prs_tune <- predict(lasso_train,x)\n",
    "    ridge_prs_tune <- predict(ridge_train,x)\n",
    "    \n",
    "    all <- cbind(lasso_prs_tune,ridge_prs_tune)\n",
    "    \n",
    "    AUC_Vector <- vector()\n",
    "    for(i in 1:ncol(all)){\n",
    "      tmp <- data.frame(y = y, x_try = all[,i])\n",
    "      roc_obj <- roc.binary(status = \"y\",\n",
    "                            variable = \"x_try\",\n",
    "                            confounders = \"~1\",\n",
    "                            data = tmp,\n",
    "                            precision=seq(0.05,0.95, by=0.05))\n",
    "      AUC_Vector[i] <- roc_obj$auc\n",
    "    }\n",
    "    \n",
    "    coefficients_x <- coef(lm(y~.,data.frame(y = all[,which.max(AUC_Vector)],x)))\n",
    "    return(list(Coefficients = coefficients_x))\n",
    "  }\n",
    "  Ensemble_Function <- function(x,y,family = c(\"continuous\",\"binary\")){\n",
    "    if(family == \"continuous\"){\n",
    "      return(Ensemble_Function_Continuous(x,y))\n",
    "    }else{\n",
    "      return(Ensemble_Function_Binary(x,y))\n",
    "    }\n",
    "  }\n",
    "\n",
    "print(str(all_prs_tune))\n",
    "\n",
    "Results <- Ensemble_Function(x = all_prs_tune,y = pheno_tune[,\"y_tune\"],family = \"continuous\")\n",
    "Results$Coefficients[is.na(Results$Coefficients)] <- 0\n",
    "\n",
    "print(\"Made it\")\n",
    "\n",
    "Final_Coefficients <- data.frame(beta_matrix[,1:4],BETA = as.matrix(beta_matrix[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1))\n",
    "print(\"Made it\")\n",
    "write.csv(Final_Coefficients,file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "PRS_Tune <- as.matrix(all_prs_tune[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "PRS_Validation <- as.matrix(all_prs_valid[,names(Results$Coefficients)[-1]]) %*% matrix(Results$Coefficients[-1],ncol = 1)\n",
    "\n",
    "print(\"Made it\")\n",
    "\n",
    "PRS_Tune <- data.frame(IID = pheno_tune$IID,PRS = PRS_Tune)\n",
    "write.csv(PRS_Tune,paste0(OUTPUT_PATH,\"/\",trait,\"_PRS_Tune.csv\"),row.names = FALSE)\n",
    "PRS_Validation <- data.frame(IID = pheno_valid$IID,PRS = PRS_Validation)\n",
    "write.csv(PRS_Validation,paste0(OUTPUT_PATH,\"/\",trait,\"_PRS_Validation.csv\"),row.names = FALSE)\n",
    "\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "RV_PRS_raw <- inner_join(pheno_valid[,c(\"IID\",\"age\",\"age2\",\"sex\",paste0(\"PC\",1:10),\"y_validation\")],PRS_Validation)\n",
    "RV_PRS_adjusted <- RV_PRS_raw\n",
    "\n",
    "\n",
    "tmp <- data.frame(y = RV_PRS_adjusted[,\"PRS\"],RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "R <- mod$residuals\n",
    "tmp <- data.frame(y = R^2,RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "mod <- lm(y~.,data = tmp)\n",
    "y_hat <- predict(mod,tmp)\n",
    "if(sum(y_hat < 0) > 0){\n",
    "  mod <- lm(y~1,data = tmp)\n",
    "  y_hat <- predict(mod,tmp)\n",
    "}\n",
    "if(sum(sqrt(y_hat)) == 0){\n",
    "  RV_PRS_adjusted[,\"PRS\"] <- 0\n",
    "}else{\n",
    "  RV_PRS_adjusted[,\"PRS\"] <- R/sqrt(y_hat)\n",
    "}\n",
    "\n",
    "\n",
    "RV_PRS_raw_EUR <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "RV_PRS_raw_SAS <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "RV_PRS_raw_AMR <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "RV_PRS_raw_AFR <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "RV_PRS_raw_EAS <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "RV_PRS_raw_MID <- RV_PRS_raw[RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "RV_PRS_adjusted_EUR <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "RV_PRS_adjusted_SAS <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "RV_PRS_adjusted_AMR <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "RV_PRS_adjusted_AFR <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "RV_PRS_adjusted_EAS <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "RV_PRS_adjusted_MID <- RV_PRS_adjusted[RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "RV_PRS_raw_EUR$y_validation <- scale(RV_PRS_raw_EUR$y_validation)\n",
    "RV_PRS_raw_SAS$y_validation <- scale(RV_PRS_raw_SAS$y_validation)\n",
    "RV_PRS_raw_AMR$y_validation <- scale(RV_PRS_raw_AMR$y_validation)\n",
    "RV_PRS_raw_AFR$y_validation <- scale(RV_PRS_raw_AFR$y_validation)\n",
    "RV_PRS_raw_EAS$y_validation <- scale(RV_PRS_raw_EAS$y_validation)\n",
    "RV_PRS_raw_MID$y_validation <- scale(RV_PRS_raw_MID$y_validation)\n",
    "\n",
    "RV_PRS_adjusted_EUR$y_validation <- scale(RV_PRS_adjusted_EUR$y_validation)\n",
    "RV_PRS_adjusted_SAS$y_validation <- scale(RV_PRS_adjusted_SAS$y_validation)\n",
    "RV_PRS_adjusted_AMR$y_validation <- scale(RV_PRS_adjusted_AMR$y_validation)\n",
    "RV_PRS_adjusted_AFR$y_validation <- scale(RV_PRS_adjusted_AFR$y_validation)\n",
    "RV_PRS_adjusted_EAS$y_validation <- scale(RV_PRS_adjusted_EAS$y_validation)\n",
    "RV_PRS_adjusted_MID$y_validation <- scale(RV_PRS_adjusted_MID$y_validation)\n",
    "\n",
    "RV_PRS_raw_EUR$PRS <- scale(RV_PRS_raw_EUR$PRS)\n",
    "RV_PRS_raw_SAS$PRS <- scale(RV_PRS_raw_SAS$PRS)\n",
    "RV_PRS_raw_AMR$PRS <- scale(RV_PRS_raw_AMR$PRS)\n",
    "RV_PRS_raw_AFR$PRS <- scale(RV_PRS_raw_AFR$PRS)\n",
    "RV_PRS_raw_EAS$PRS <- scale(RV_PRS_raw_EAS$PRS)\n",
    "RV_PRS_raw_MID$PRS <- scale(RV_PRS_raw_MID$PRS)\n",
    "\n",
    "Beta_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = boot_data))[2]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = boot_data))$r.squared\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "beta_validation_raw_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_EUR))[2]\n",
    "boot_beta <- boot(data = RV_PRS_raw_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_EUR))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_raw_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_SAS))[2]\n",
    "boot_beta <- boot(data = RV_PRS_raw_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_SAS))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_raw_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_AMR))[2]\n",
    "boot_beta <- boot(data = RV_PRS_raw_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_AMR))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_raw_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_AFR))[2]\n",
    "boot_beta <- boot(data = RV_PRS_raw_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_AFR))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_raw_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_EAS))[2]\n",
    "boot_beta <- boot(data = RV_PRS_raw_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_EAS))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_raw_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_raw_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_MID))[2]\n",
    "boot_beta <- boot(data = RV_PRS_raw_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_raw_MID_boot <- boot_beta$t\n",
    "beta_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_raw_MID))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_raw_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_MID_boot <- boot_R2$t\n",
    "R2_se_validation_raw_MID <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EUR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_EUR))[2]\n",
    "boot_beta <- boot(data = RV_PRS_adjusted_EUR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EUR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_EUR))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_adjusted_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EUR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_SAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_SAS))[2]\n",
    "boot_beta <- boot(data = RV_PRS_adjusted_SAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_SAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_SAS))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_adjusted_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_SAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AMR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_AMR))[2]\n",
    "boot_beta <- boot(data = RV_PRS_adjusted_AMR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AMR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_AMR))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_adjusted_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AMR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_AFR <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_AFR))[2]\n",
    "boot_beta <- boot(data = RV_PRS_adjusted_AFR, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AFR <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_AFR))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_adjusted_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AFR <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_EAS <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_EAS))[2]\n",
    "boot_beta <- boot(data = RV_PRS_adjusted_EAS, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EAS <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_EAS))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_adjusted_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EAS <- sd(boot_R2$t)\n",
    "\n",
    "beta_validation_adjusted_MID <- coef(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_MID))[2]\n",
    "boot_beta <- boot(data = RV_PRS_adjusted_MID, statistic = Beta_Boot, R = 10000)\n",
    "beta_adjusted_MID_boot <- boot_beta$t\n",
    "beta_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_MID <- summary(lm(as.formula(paste0(\"y_validation~\",\"PRS\")),data = RV_PRS_adjusted_MID))$r.squared\n",
    "boot_R2 <- boot(data = RV_PRS_adjusted_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_MID_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_MID <- sd(boot_R2$t)\n",
    "\n",
    "RV_PRS_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                             beta_raw = c(beta_validation_raw_EUR,beta_validation_raw_SAS,beta_validation_raw_AMR,beta_validation_raw_AFR,beta_validation_raw_EAS,beta_validation_raw_MID), \n",
    "                             beta_se_raw = c(beta_se_validation_raw_EUR,beta_se_validation_raw_SAS,beta_se_validation_raw_AMR,beta_se_validation_raw_AFR,beta_se_validation_raw_EAS,beta_se_validation_raw_MID), \n",
    "                             R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                             R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                             beta_adjusted = c(beta_validation_adjusted_EUR,beta_validation_adjusted_SAS,beta_validation_adjusted_AMR,beta_validation_adjusted_AFR,beta_validation_adjusted_EAS,beta_validation_adjusted_MID), \n",
    "                             beta_se_adjusted = c(beta_se_validation_adjusted_EUR,beta_se_validation_adjusted_SAS,beta_se_validation_adjusted_AMR,beta_se_validation_adjusted_AFR,beta_se_validation_adjusted_EAS,beta_se_validation_adjusted_MID), \n",
    "                             R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                             R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "RV_Boot_Results <- data.frame(trait = trait,beta_raw_EUR_boot,R2_raw_EUR_boot,beta_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_raw_AMR_boot,R2_raw_AMR_boot,beta_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_raw_EAS_boot,R2_raw_EAS_boot,beta_raw_MID_boot,R2_raw_MID_boot,beta_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_adjusted_EAS_boot,R2_adjusted_EAS_boot,beta_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "write.csv(RV_PRS_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(RV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Bootstraps.csv\"),row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Single_RareVariant_PRS.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Single_RareVariant_PRS.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Train_Pvals} ${Train_NullModel} ${Tune_NullModel} ${Validation_NullModel} ${Annotation_name_catalog} ${INPUT_PATH} ${all_phenotypes_file} ${all_train_file} ${all_tune_file} ${all_validation_file} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "            '--env trait'=trait,\n",
    "            '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Single_RareVariant_PRS.R\",\n",
    "            '--input Train_Pvals'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding/\",trait,\"_coding_sig.csv\"),\n",
    "            '--input Train_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Tune_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Tune_Null_Model.RData\"),\n",
    "            '--input Validation_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Validation_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "            '--input all_train_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Train.txt\",\n",
    "            '--input all_tune_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "            '--input all_validation_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS\",\n",
    "            check.names = FALSE\n",
    "        ))   \n",
    "}\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Single_RareVariant_PRS.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 27.8 KiB/ 27.8 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/27.8 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Single_RareVariant_PRS.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: single-rar--williamsjacr--250511-005419-70\n",
      "  job-name: single-rarevariant-prs\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4588539315175791449\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10993787666976377948\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4946105226847872983\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10491198882476459573\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4444183596653775443\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/716398368260328813\n",
      "Launched job-id: single-rar--williamsjacr--250511-005419-70\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'single-rar--williamsjacr--250511-005419-70' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'single-rar--williamsjacr--250511-005419-70' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Single_RareVariant_PRS.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "code_folding": [
     19
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing RareVariantCoefficients.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile RareVariantCoefficients.R\n",
    "rm(list = ls())\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "Gene_Centric_Coding_G_Star <- function(chr,gene_name,category=c(\"plof\",\"plof_ds\",\"missense\",\"disruptive_missense\",\"synonymous\"),\n",
    "                                       genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                       QC_label=\"annotation/filter\",variant_type=c(\"SNV\",\"Indel\",\"variant\"),geno_missing_imputation=c(\"mean\",\"minor\"),\n",
    "                                       Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,silent=FALSE){\n",
    "  ## evaluate choices\n",
    "  category <- match.arg(category)\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  geno_missing_imputation <- match.arg(geno_missing_imputation)\n",
    "  \n",
    "  genes <- genes_info[genes_info[,2]==chr,]  \n",
    "  \n",
    "  phenotype.id <- obj_nullmodel$id_include\n",
    "  ## get SNV id, position, REF, ALT (whole genome)\n",
    "  filter <- seqGetData(genofile, QC_label)\n",
    "  if(variant_type==\"variant\"){\n",
    "    SNVlist <- filter == \"PASS\"\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"SNV\"){\n",
    "    SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "  }\n",
    "  \n",
    "  if(variant_type==\"Indel\"){\n",
    "    SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "  }\n",
    "  \n",
    "  position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "  variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "  \n",
    "  rm(filter)\n",
    "  gc()\n",
    "  \n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "  \n",
    "  sub_start_loc <- genes[kk,3]\n",
    "  sub_end_loc <- genes[kk,4]\n",
    "  \n",
    "  is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "  variant.id.gene <- variant.id[is.in]\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## plof\n",
    "  ## Gencode_Exonic\n",
    "  GENCODE.EXONIC.Category  <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "  ## Gencode\n",
    "  GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "  ## Meta.SVM.Pred\n",
    "  MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "  \n",
    "  variant.id.gene <- seqGetData(genofile, \"variant.id\")  \n",
    "  \n",
    "  if(category == \"plof\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"plof_ds\"){\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.plof]\n",
    "  }else if(category == \"missense\"){\n",
    "    lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.missense]\n",
    "  }else if(category == \"disruptive_missense\"){\n",
    "    lof.in.ds <- ((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene <- variant.id.gene[lof.in.ds]\n",
    "  }else{\n",
    "    lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.synonymous]\n",
    "  }\n",
    "  \n",
    "  seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "  \n",
    "  ## genotype id\n",
    "  id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "  # id.genotype.match <- rep(0,length(id.genotype))\n",
    "  \n",
    "  id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "  phenotype.id.merge <- data.frame(phenotype.id)\n",
    "  phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "  id.genotype.match <- phenotype.id.merge$index\n",
    "  \n",
    "  ## Genotype\n",
    "  Geno <- seqGetData(genofile, \"$dosage\")\n",
    "  Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "  \n",
    "  ## impute missing\n",
    "  if(!is.null(dim(Geno)))\n",
    "  {\n",
    "    if(dim(Geno)[2]>0)\n",
    "    {\n",
    "      if(geno_missing_imputation==\"mean\")\n",
    "      {\n",
    "        Geno <- matrix_flip_mean(Geno)$Geno\n",
    "      }\n",
    "      if(geno_missing_imputation==\"minor\")\n",
    "      {\n",
    "        Geno <- matrix_flip_minor(Geno)$Geno\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  genotype <- Geno\n",
    "  \n",
    "  if(dim(genotype)[2] == 1){\n",
    "    return(matrix(0,nrow = dim(genotype)[1],ncol = 1))\n",
    "  }\n",
    "  \n",
    "  if(!is.null(attr(class(genotype), \"package\")) && attr(class(genotype), \"package\") == \"Matrix\"){\n",
    "    genotype <- as.matrix(genotype)\n",
    "  }\n",
    "  genotype <- matrix_flip(genotype)\n",
    "  MAF <- genotype$MAF\n",
    "  RV_label <- as.vector((MAF<rare_maf_cutoff)&(MAF>0))\n",
    "  Geno_rare <- genotype$Geno[,RV_label]\n",
    "  G <- Geno_rare\n",
    "  rm(Geno_rare)\n",
    "  gc()\n",
    "  \n",
    "  if(is.null(dim(G))){\n",
    "    G <- matrix(G,ncol = 1)\n",
    "  }\n",
    "  \n",
    "  C <- G%*%matrix(1,nrow=ncol(G),ncol = 1)\n",
    "  \n",
    "  seqResetFilter(genofile)\n",
    "  \n",
    "  return(C)\n",
    "}\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Train_Pvals <- commandArgs(TRUE)[2]\n",
    "print(Train_Pvals)\n",
    "\n",
    "Train_NullModel <- commandArgs(TRUE)[3]\n",
    "print(Train_NullModel)\n",
    "\n",
    "Tune_NullModel <- commandArgs(TRUE)[4]\n",
    "print(Tune_NullModel)\n",
    "\n",
    "Validation_NullModel <- commandArgs(TRUE)[5]\n",
    "print(Validation_NullModel)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[6]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "INPUT_PATH <- commandArgs(TRUE)[7]\n",
    "print(INPUT_PATH)\n",
    "\n",
    "RV_PRS <- commandArgs(TRUE)[8]\n",
    "print(RV_PRS)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[9]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "Train_PVals_All <- read.csv(Train_Pvals)\n",
    "Train_PVals_All <- Train_PVals_All[Train_PVals_All$STAARB <= 1e-03,]\n",
    "\n",
    "## Null Model\n",
    "obj_nullmodel_train <- get(load(Train_NullModel))\n",
    "obj_nullmodel_tune <- get(load(Tune_NullModel))\n",
    "obj_nullmodel_validation <- get(load(Validation_NullModel))\n",
    "\n",
    "obj_nullmodel <- obj_nullmodel_train\n",
    "obj_nullmodel$id_include <- c(obj_nullmodel_train$id_include,obj_nullmodel_tune$id_include,obj_nullmodel_validation$id_include)\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "\n",
    "G_star_gene_centric_coding <- list()\n",
    "\n",
    "for(i in 1:nrow(Train_PVals_All)){\n",
    "  ## Chr\n",
    "  chr <- Train_PVals_All$Chr[i]\n",
    "  ## Gene name\n",
    "  gene_name <- Train_PVals_All$Gene[i]\n",
    "  ## Coding mask\n",
    "  category <- Train_PVals_All$Category[i]\n",
    "  \n",
    "  ### gds file\n",
    "  gds.path <- paste0(INPUT_PATH,\"/exome.chr\",chr,\".gds\")\n",
    "  genofile <- seqOpen(gds.path)\n",
    "  \n",
    "  G_star_gene_centric_coding[[i]] <- Gene_Centric_Coding_G_Star(chr=chr,gene_name=gene_name,category=category ,\n",
    "                                                                genofile,obj_nullmodel,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                                                QC_label=QC_label,variant_type=variant_type,geno_missing_imputation=geno_missing_imputation,\n",
    "                                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,silent=FALSE) \n",
    "  seqClose(genofile) \n",
    "} \n",
    "\n",
    "G_star_gene_centric_coding <- do.call(cbind,G_star_gene_centric_coding)\n",
    "\n",
    "col_remove <- apply(G_star_gene_centric_coding,2,function(x){sum(x != 0)}) > 10 & colSums(G_star_gene_centric_coding) > 10 \n",
    "G_star_gene_centric_coding <- G_star_gene_centric_coding[,col_remove,drop = FALSE]\n",
    "\n",
    "Train_PVals_All <- Train_PVals_All[col_remove,]\n",
    "\n",
    "print(str(G_star_gene_centric_coding))\n",
    "\n",
    "ids_gstar <- obj_nullmodel$id_include\n",
    "\n",
    "G_star_gene_centric_coding_train <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_train$id_include,]\n",
    "G_star_gene_centric_coding_tune <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_tune$id_include,]\n",
    "G_star_gene_centric_coding_vad <- G_star_gene_centric_coding[ids_gstar %in% obj_nullmodel_validation$id_include,]\n",
    "\n",
    "rm(G_star_gene_centric_coding)\n",
    "\n",
    "\n",
    "\n",
    "X_valid <- data.frame(IID = ids_gstar[ids_gstar %in% obj_nullmodel_validation$id_include],G_star_gene_centric_coding_vad)\n",
    "RV_PRS <- read.csv(RV_PRS)\n",
    "tmp <- inner_join(RV_PRS[,c(\"IID\",\"PRS\")],X_valid)\n",
    "tmp <- subset(tmp,select = -c(IID))\n",
    "Train_PVals_All$Beta <- coef(lm(PRS~.,tmp))[-1]\n",
    "Train_PVals_All$Beta[is.na(Train_PVals_All$Beta)] <- 0\n",
    "Train_PVals_All$Beta[abs(Train_PVals_All$Beta) < 1e-10] <- 0\n",
    "\n",
    "write.csv(Train_PVals_All,file = paste0(OUTPUT_PATH,\"/\",trait,\"_final_coef.csv\"),row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing RareVariantCoefficients.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile RareVariantCoefficients.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Train_Pvals} ${Train_NullModel} ${Tune_NullModel} ${Validation_NullModel} ${Annotation_name_catalog} ${INPUT_PATH} ${RV_PRS} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    " tasks <- rbind(tasks, data.frame(\n",
    "            '--env trait'=trait,\n",
    "            '--input-recursive INPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/RareVariantCoefficients.R\",\n",
    "            '--input Train_Pvals'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/GeneCentricCoding/\",trait,\"_coding_sig.csv\"),\n",
    "            '--input Train_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Train_Null_Model.RData\"),\n",
    "            '--input Tune_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Tune_Null_Model.RData\"),\n",
    "            '--input Validation_NullModel'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/NullModels/\",trait,\"_Validation_Null_Model.RData\"),\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input RV_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\",trait,\"_PRS_Validation.csv\"),\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS\",\n",
    "            check.names = FALSE\n",
    "        ))   \n",
    "}\n",
    "\n",
    "   \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://RareVariantCoefficients.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  8.4 KiB/  8.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/8.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp RareVariantCoefficients.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: rarevarian--williamsjacr--250511-125417-59\n",
      "  job-name: rarevariantcoefficients\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/10641838240056700816\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/16025562658320565953\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12772089216612072498\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7742665512195031476\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/11176103684187754673\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7247589663043056178\n",
      "Launched job-id: rarevarian--williamsjacr--250511-125417-59\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'rarevarian--williamsjacr--250511-125417-59' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'rarevarian--williamsjacr--250511-125417-59' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script RareVariantCoefficients.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Common_Plus_Rare.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Plus_Rare.R\n",
    "rm(list = ls())\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(caret)\n",
    "library(ranger)\n",
    "library(SuperLearner)\n",
    "library(dplyr)\n",
    "library(boot)\n",
    "library(stringr)\n",
    "library(glmnet)\n",
    "\n",
    "trait <- commandArgs(TRUE)[1]\n",
    "print(trait)\n",
    "\n",
    "Tune_Data <- commandArgs(TRUE)[2]\n",
    "print(Tune_Data)\n",
    "\n",
    "Validation_Data <- commandArgs(TRUE)[3]\n",
    "print(Validation_Data)\n",
    "\n",
    "CommonVariantPRS_Tune <- commandArgs(TRUE)[4]\n",
    "print(CommonVariantPRS_Tune)\n",
    "\n",
    "RareVariantPRS_Tune <- commandArgs(TRUE)[5]\n",
    "print(RareVariantPRS_Tune)\n",
    "\n",
    "CommonVariantPRS_Validation <- commandArgs(TRUE)[6]\n",
    "print(CommonVariantPRS_Validation)\n",
    "\n",
    "RareVariantPRS_Validation <- commandArgs(TRUE)[7]\n",
    "print(RareVariantPRS_Validation)\n",
    "\n",
    "JointPRS <- commandArgs(TRUE)[8]\n",
    "print(JointPRS)\n",
    "\n",
    "PROSPER_PRS <- commandArgs(TRUE)[9]\n",
    "print(PROSPER_PRS)\n",
    "\n",
    "CTSLEB_PRS <- commandArgs(TRUE)[10]\n",
    "print(CTSLEB_PRS)\n",
    "\n",
    "all_phenotypes_file <- commandArgs(TRUE)[11]\n",
    "print(all_phenotypes_file)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[12]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "JointPRS <- read.delim(JointPRS)\n",
    "colnames(JointPRS) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"JointPRS_PRS\")\n",
    "JointPRS <- JointPRS[,c(\"IID\",\"JointPRS_PRS\")]\n",
    "\n",
    "PROSPER_PRS <- read.delim(PROSPER_PRS)\n",
    "colnames(PROSPER_PRS) <- c(\"FID\",\"IID\",\"ALLELE_CT\",\"NAMED_ALLELE_DOSAGE_SUM\",\"PROSPER_PRS\")\n",
    "PROSPER_PRS <- PROSPER_PRS[,c(\"IID\",\"PROSPER_PRS\")]\n",
    "\n",
    "CTSLEB_PRS <- read.delim(CTSLEB_PRS,sep = \"\\t\",header = TRUE)\n",
    "colnames(CTSLEB_PRS) <- c(\"IID\",\"CTSLEB_PRS\")\n",
    "\n",
    "pheno_tune <- read.delim(Tune_Data)\n",
    "CV_PRS_Tune <- read.delim(CommonVariantPRS_Tune)\n",
    "colnames(CV_PRS_Tune) <- c(\"IID\",\"CV_PRS\")\n",
    "pheno_tune <- inner_join(pheno_tune,CV_PRS_Tune)\n",
    "RV_PRS_Tune <- read.csv(RareVariantPRS_Tune)\n",
    "colnames(RV_PRS_Tune) <- c(\"IID\",\"RV_PRS\")\n",
    "pheno_tune <- inner_join(pheno_tune,RV_PRS_Tune)\n",
    "\n",
    "\n",
    "pheno_validation <- read.delim(Validation_Data)\n",
    "CV_PRS_Validation <- read.delim(CommonVariantPRS_Validation)\n",
    "colnames(CV_PRS_Validation) <- c(\"IID\",\"CV_PRS\")\n",
    "pheno_validation <- inner_join(pheno_validation,CV_PRS_Validation)\n",
    "RV_PRS_Validation <- read.csv(RareVariantPRS_Validation)\n",
    "colnames(RV_PRS_Validation) <- c(\"IID\",\"RV_PRS\")\n",
    "pheno_validation <- inner_join(pheno_validation,RV_PRS_Validation)\n",
    "pheno_validation <- inner_join(pheno_validation,JointPRS)\n",
    "pheno_validation <- inner_join(pheno_validation,PROSPER_PRS)\n",
    "pheno_validation <- inner_join(pheno_validation,CTSLEB_PRS)\n",
    "\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_tune)\n",
    "pheno_tune$y_tune <- NA\n",
    "pheno_tune$y_tune[!is.na(pheno_tune[,trait])] <- model.null$residual\n",
    "\n",
    "model.null <- lm(as.formula(paste0(trait,\"~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")),data=pheno_validation)\n",
    "pheno_validation$y_validation <- NA\n",
    "pheno_validation$y_validation[!is.na(pheno_validation[,trait])] <- model.null$residual\n",
    "\n",
    "RICE_Model <- lm(y_tune ~ CV_PRS + RV_PRS,data = pheno_tune)\n",
    "pheno_validation$PRS <- predict(RICE_Model,pheno_validation)\n",
    "\n",
    "CV_RV_PRS_raw <- pheno_validation\n",
    "CV_RV_PRS_adjusted <- pheno_validation\n",
    "\n",
    "for(i in c(\"CV_PRS\",\"RV_PRS\",\"PRS\",\"JointPRS_PRS\",\"PROSPER_PRS\",\"CTSLEB_PRS\")){\n",
    "  tmp <- data.frame(y = CV_RV_PRS_adjusted[,i],CV_RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "  mod <- lm(y~.,data = tmp)\n",
    "  R <- mod$residuals\n",
    "  tmp <- data.frame(y = R^2,CV_RV_PRS_adjusted[,c(\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\")])\n",
    "  mod <- lm(y~.,data = tmp)\n",
    "  y_hat <- predict(mod,tmp)\n",
    "  if(sum(y_hat < 0) > 0){\n",
    "    mod <- lm(y~1,data = tmp)\n",
    "    y_hat <- predict(mod,tmp)\n",
    "  }\n",
    "  if(sum(sqrt(y_hat)) == 0){\n",
    "    CV_RV_PRS_adjusted[,i] <- 0\n",
    "  }else{\n",
    "    CV_RV_PRS_adjusted[,i] <- R/sqrt(y_hat)\n",
    "  }\n",
    "}\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotypes_file)\n",
    "\n",
    "CV_RV_PRS_raw_EUR <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "CV_RV_PRS_raw_SAS <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "CV_RV_PRS_raw_AMR <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "CV_RV_PRS_raw_AFR <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "CV_RV_PRS_raw_EAS <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "CV_RV_PRS_raw_MID <- CV_RV_PRS_raw[CV_RV_PRS_raw$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "CV_RV_PRS_adjusted_EUR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EUR\"],]\n",
    "CV_RV_PRS_adjusted_SAS <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"SAS\"],]\n",
    "CV_RV_PRS_adjusted_AMR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AMR\"],]\n",
    "CV_RV_PRS_adjusted_AFR <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"AFR\"],]\n",
    "CV_RV_PRS_adjusted_EAS <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"EAS\"],]\n",
    "CV_RV_PRS_adjusted_MID <- CV_RV_PRS_adjusted[CV_RV_PRS_adjusted$IID %in% all_phenotypes$IID[all_phenotypes$ancestry == \"MID\"],]\n",
    "\n",
    "CV_RV_PRS_raw_EUR$y_validation <- scale(CV_RV_PRS_raw_EUR$y_validation)\n",
    "CV_RV_PRS_raw_SAS$y_validation <- scale(CV_RV_PRS_raw_SAS$y_validation)\n",
    "CV_RV_PRS_raw_AMR$y_validation <- scale(CV_RV_PRS_raw_AMR$y_validation)\n",
    "CV_RV_PRS_raw_AFR$y_validation <- scale(CV_RV_PRS_raw_AFR$y_validation)\n",
    "CV_RV_PRS_raw_EAS$y_validation <- scale(CV_RV_PRS_raw_EAS$y_validation)\n",
    "CV_RV_PRS_raw_MID$y_validation <- scale(CV_RV_PRS_raw_MID$y_validation)\n",
    "\n",
    "CV_RV_PRS_adjusted_EUR$y_validation <- scale(CV_RV_PRS_adjusted_EUR$y_validation)\n",
    "CV_RV_PRS_adjusted_SAS$y_validation <- scale(CV_RV_PRS_adjusted_SAS$y_validation)\n",
    "CV_RV_PRS_adjusted_AMR$y_validation <- scale(CV_RV_PRS_adjusted_AMR$y_validation)\n",
    "CV_RV_PRS_adjusted_AFR$y_validation <- scale(CV_RV_PRS_adjusted_AFR$y_validation)\n",
    "CV_RV_PRS_adjusted_EAS$y_validation <- scale(CV_RV_PRS_adjusted_EAS$y_validation)\n",
    "CV_RV_PRS_adjusted_MID$y_validation <- scale(CV_RV_PRS_adjusted_MID$y_validation)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$RV_PRS <- scale(CV_RV_PRS_raw_EUR$RV_PRS)\n",
    "CV_RV_PRS_raw_SAS$RV_PRS <- scale(CV_RV_PRS_raw_SAS$RV_PRS)\n",
    "CV_RV_PRS_raw_AMR$RV_PRS <- scale(CV_RV_PRS_raw_AMR$RV_PRS)\n",
    "CV_RV_PRS_raw_AFR$RV_PRS <- scale(CV_RV_PRS_raw_AFR$RV_PRS)\n",
    "CV_RV_PRS_raw_EAS$RV_PRS <- scale(CV_RV_PRS_raw_EAS$RV_PRS)\n",
    "CV_RV_PRS_raw_MID$RV_PRS <- scale(CV_RV_PRS_raw_MID$RV_PRS)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$CV_PRS <- scale(CV_RV_PRS_raw_EUR$CV_PRS)\n",
    "CV_RV_PRS_raw_SAS$CV_PRS <- scale(CV_RV_PRS_raw_SAS$CV_PRS)\n",
    "CV_RV_PRS_raw_AMR$CV_PRS <- scale(CV_RV_PRS_raw_AMR$CV_PRS)\n",
    "CV_RV_PRS_raw_AFR$CV_PRS <- scale(CV_RV_PRS_raw_AFR$CV_PRS)\n",
    "CV_RV_PRS_raw_EAS$CV_PRS <- scale(CV_RV_PRS_raw_EAS$CV_PRS)\n",
    "CV_RV_PRS_raw_MID$CV_PRS <- scale(CV_RV_PRS_raw_MID$CV_PRS)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$PRS <- scale(CV_RV_PRS_raw_EUR$PRS)\n",
    "CV_RV_PRS_raw_SAS$PRS <- scale(CV_RV_PRS_raw_SAS$PRS)\n",
    "CV_RV_PRS_raw_AMR$PRS <- scale(CV_RV_PRS_raw_AMR$PRS)\n",
    "CV_RV_PRS_raw_AFR$PRS <- scale(CV_RV_PRS_raw_AFR$PRS)\n",
    "CV_RV_PRS_raw_EAS$PRS <- scale(CV_RV_PRS_raw_EAS$PRS)\n",
    "CV_RV_PRS_raw_MID$PRS <- scale(CV_RV_PRS_raw_MID$PRS)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$JointPRS_PRS <- scale(CV_RV_PRS_raw_EUR$JointPRS_PRS)\n",
    "CV_RV_PRS_raw_SAS$JointPRS_PRS <- scale(CV_RV_PRS_raw_SAS$JointPRS_PRS)\n",
    "CV_RV_PRS_raw_AMR$JointPRS_PRS <- scale(CV_RV_PRS_raw_AMR$JointPRS_PRS)\n",
    "CV_RV_PRS_raw_AFR$JointPRS_PRS <- scale(CV_RV_PRS_raw_AFR$JointPRS_PRS)\n",
    "CV_RV_PRS_raw_EAS$JointPRS_PRS <- scale(CV_RV_PRS_raw_EAS$JointPRS_PRS)\n",
    "CV_RV_PRS_raw_MID$JointPRS_PRS <- scale(CV_RV_PRS_raw_MID$JointPRS_PRS)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$PROSPER_PRS <- scale(CV_RV_PRS_raw_EUR$PROSPER_PRS)\n",
    "CV_RV_PRS_raw_SAS$PROSPER_PRS <- scale(CV_RV_PRS_raw_SAS$PROSPER_PRS)\n",
    "CV_RV_PRS_raw_AMR$PROSPER_PRS <- scale(CV_RV_PRS_raw_AMR$PROSPER_PRS)\n",
    "CV_RV_PRS_raw_AFR$PROSPER_PRS <- scale(CV_RV_PRS_raw_AFR$PROSPER_PRS)\n",
    "CV_RV_PRS_raw_EAS$PROSPER_PRS <- scale(CV_RV_PRS_raw_EAS$PROSPER_PRS)\n",
    "CV_RV_PRS_raw_MID$PROSPER_PRS <- scale(CV_RV_PRS_raw_MID$PROSPER_PRS)\n",
    "\n",
    "CV_RV_PRS_raw_EUR$CTSLEB_PRS <- scale(CV_RV_PRS_raw_EUR$CTSLEB_PRS)\n",
    "CV_RV_PRS_raw_SAS$CTSLEB_PRS <- scale(CV_RV_PRS_raw_SAS$CTSLEB_PRS)\n",
    "CV_RV_PRS_raw_AMR$CTSLEB_PRS <- scale(CV_RV_PRS_raw_AMR$CTSLEB_PRS)\n",
    "CV_RV_PRS_raw_AFR$CTSLEB_PRS <- scale(CV_RV_PRS_raw_AFR$CTSLEB_PRS)\n",
    "CV_RV_PRS_raw_EAS$CTSLEB_PRS <- scale(CV_RV_PRS_raw_EAS$CTSLEB_PRS)\n",
    "CV_RV_PRS_raw_MID$CTSLEB_PRS <- scale(CV_RV_PRS_raw_MID$CTSLEB_PRS)\n",
    "\n",
    "\n",
    "Beta_CV_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(y_validation~CV_PRS + RV_PRS,data = boot_data))[2]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "Beta_RV_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- coef(lm(y_validation~CV_PRS + RV_PRS,data = boot_data))[3]\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  result <- summary(lm(y_validation~PRS,data = boot_data))$r.squared\n",
    "  return(c(result))\n",
    "}\n",
    "\n",
    "R2_Comparison_Boot <- function(data,indices){\n",
    "  boot_data <- data[indices, ]\n",
    "  RICE_R2 <- summary(lm(y_validation~PRS,data = boot_data))$r.squared\n",
    "  CTSLEB_R2 <- summary(lm(y_validation~CTSLEB_PRS,data = boot_data))$r.squared\n",
    "  PROSPER_R2 <- summary(lm(y_validation~PROSPER_PRS,data = boot_data))$r.squared\n",
    "  JointPRS_R2 <- summary(lm(y_validation~JointPRS_PRS,data = boot_data))$r.squared\n",
    "  return(c(RICE_R2 - CTSLEB_R2,RICE_R2 - PROSPER_R2,RICE_R2 - JointPRS_R2))\n",
    "}\n",
    "\n",
    "beta_CV_validation_raw_EUR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_EUR))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_EUR, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_raw_EUR_boot <- boot_beta$t\n",
    "beta_CV_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_raw_EUR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_EUR))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_EUR, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_raw_EUR_boot <- boot_beta$t\n",
    "beta_RV_se_validation_raw_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EUR <- summary(lm(y_validation~PRS,data = CV_RV_PRS_raw_EUR))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EUR <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_EUR, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_raw_EUR_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_raw_EUR_boot) <- c(\"R2_raw_EUR_RICE_vs_CTSLEB\",\"R2_raw_EUR_RICE_vs_PROSPER\",\"R2_raw_EUR_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_raw_SAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_SAS))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_SAS, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_raw_SAS_boot <- boot_beta$t\n",
    "beta_CV_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_raw_SAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_SAS))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_SAS, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_raw_SAS_boot <- boot_beta$t\n",
    "beta_RV_se_validation_raw_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_SAS <- summary(lm(y_validation~PRS,data = CV_RV_PRS_raw_SAS))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_SAS <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_SAS, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_raw_SAS_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_raw_SAS_boot) <- c(\"R2_raw_SAS_RICE_vs_CTSLEB\",\"R2_raw_SAS_RICE_vs_PROSPER\",\"R2_raw_SAS_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_raw_AMR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_AMR))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_AMR, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_raw_AMR_boot <- boot_beta$t\n",
    "beta_CV_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_raw_AMR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_AMR))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_AMR, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_raw_AMR_boot <- boot_beta$t\n",
    "beta_RV_se_validation_raw_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AMR <- summary(lm(y_validation~PRS,data = CV_RV_PRS_raw_AMR))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AMR <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_AMR, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_raw_AMR_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_raw_AMR_boot) <- c(\"R2_raw_AMR_RICE_vs_CTSLEB\",\"R2_raw_AMR_RICE_vs_PROSPER\",\"R2_raw_AMR_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_raw_AFR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_AFR))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_AFR, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_raw_AFR_boot <- boot_beta$t\n",
    "beta_CV_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_raw_AFR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_AFR))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_AFR, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_raw_AFR_boot <- boot_beta$t\n",
    "beta_RV_se_validation_raw_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_AFR <- summary(lm(y_validation~PRS,data = CV_RV_PRS_raw_AFR))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_raw_AFR <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_AFR, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_raw_AFR_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_raw_AFR_boot) <- c(\"R2_raw_AFR_RICE_vs_CTSLEB\",\"R2_raw_AFR_RICE_vs_PROSPER\",\"R2_raw_AFR_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_raw_EAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_EAS))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_EAS, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_raw_EAS_boot <- boot_beta$t\n",
    "beta_CV_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_raw_EAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_EAS))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_EAS, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_raw_EAS_boot <- boot_beta$t\n",
    "beta_RV_se_validation_raw_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_EAS <- summary(lm(y_validation~PRS,data = CV_RV_PRS_raw_EAS))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_raw_EAS <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_EAS, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_raw_EAS_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_raw_EAS_boot) <- c(\"R2_raw_EAS_RICE_vs_CTSLEB\",\"R2_raw_EAS_RICE_vs_PROSPER\",\"R2_raw_EAS_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_raw_MID <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_MID))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_MID, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_raw_MID_boot <- boot_beta$t\n",
    "beta_CV_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_raw_MID <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_raw_MID))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_raw_MID, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_raw_MID_boot <- boot_beta$t\n",
    "beta_RV_se_validation_raw_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_raw_MID <- summary(lm(y_validation~PRS,data = CV_RV_PRS_raw_MID))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_raw_MID_boot <- boot_R2$t\n",
    "R2_se_validation_raw_MID <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_raw_MID, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_raw_MID_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_raw_MID_boot) <- c(\"R2_raw_MID_RICE_vs_CTSLEB\",\"R2_raw_MID_RICE_vs_PROSPER\",\"R2_raw_MID_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_adjusted_EUR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_EUR))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_EUR, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_CV_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_adjusted_EUR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_EUR))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_EUR, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_adjusted_EUR_boot <- boot_beta$t\n",
    "beta_RV_se_validation_adjusted_EUR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EUR <- summary(lm(y_validation~PRS,data = CV_RV_PRS_adjusted_EUR))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_EUR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EUR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EUR <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_EUR, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_adjusted_EUR_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_adjusted_EUR_boot) <- c(\"R2_adjusted_EUR_RICE_vs_CTSLEB\",\"R2_adjusted_EUR_RICE_vs_PROSPER\",\"R2_adjusted_EUR_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_adjusted_SAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_SAS))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_SAS, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_CV_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_adjusted_SAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_SAS))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_SAS, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_adjusted_SAS_boot <- boot_beta$t\n",
    "beta_RV_se_validation_adjusted_SAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_SAS <- summary(lm(y_validation~PRS,data = CV_RV_PRS_adjusted_SAS))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_SAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_SAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_SAS <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_SAS, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_adjusted_SAS_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_adjusted_SAS_boot) <- c(\"R2_adjusted_SAS_RICE_vs_CTSLEB\",\"R2_adjusted_SAS_RICE_vs_PROSPER\",\"R2_adjusted_SAS_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_adjusted_AMR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_AMR))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_AMR, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_CV_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_adjusted_AMR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_AMR))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_AMR, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_adjusted_AMR_boot <- boot_beta$t\n",
    "beta_RV_se_validation_adjusted_AMR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AMR <- summary(lm(y_validation~PRS,data = CV_RV_PRS_adjusted_AMR))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_AMR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AMR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AMR <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_AMR, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_adjusted_AMR_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_adjusted_AMR_boot) <- c(\"R2_adjusted_AMR_RICE_vs_CTSLEB\",\"R2_adjusted_AMR_RICE_vs_PROSPER\",\"R2_adjusted_AMR_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_adjusted_AFR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_AFR))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_AFR, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_CV_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_adjusted_AFR <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_AFR))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_AFR, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_adjusted_AFR_boot <- boot_beta$t\n",
    "beta_RV_se_validation_adjusted_AFR <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_AFR <- summary(lm(y_validation~PRS,data = CV_RV_PRS_adjusted_AFR))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_AFR, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_AFR_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_AFR <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_AFR, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_adjusted_AFR_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_adjusted_AFR_boot) <- c(\"R2_adjusted_AFR_RICE_vs_CTSLEB\",\"R2_adjusted_AFR_RICE_vs_PROSPER\",\"R2_adjusted_AFR_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_adjusted_EAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_EAS))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_EAS, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_CV_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_adjusted_EAS <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_EAS))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_EAS, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_adjusted_EAS_boot <- boot_beta$t\n",
    "beta_RV_se_validation_adjusted_EAS <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_EAS <- summary(lm(y_validation~PRS,data = CV_RV_PRS_adjusted_EAS))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_EAS, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_EAS_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_EAS <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_EAS, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_adjusted_EAS_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_adjusted_EAS_boot) <- c(\"R2_adjusted_EAS_RICE_vs_CTSLEB\",\"R2_adjusted_EAS_RICE_vs_PROSPER\",\"R2_adjusted_EAS_RICE_vs_JointPRS\")\n",
    "\n",
    "beta_CV_validation_adjusted_MID <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_MID))[2]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_MID, statistic = Beta_CV_Boot, R = 10000)\n",
    "beta_CV_adjusted_MID_boot <- boot_beta$t\n",
    "beta_CV_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "beta_RV_validation_adjusted_MID <- coef(lm(y_validation~CV_PRS + RV_PRS,data = CV_RV_PRS_adjusted_MID))[3]\n",
    "boot_beta <- boot(data = CV_RV_PRS_adjusted_MID, statistic = Beta_RV_Boot, R = 10000)\n",
    "beta_RV_adjusted_MID_boot <- boot_beta$t\n",
    "beta_RV_se_validation_adjusted_MID <- sd(boot_beta$t)\n",
    "\n",
    "R2_validation_adjusted_MID <- summary(lm(y_validation~PRS,data = CV_RV_PRS_adjusted_MID))$r.squared\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_MID, statistic = R2_Boot, R = 10000)\n",
    "R2_adjusted_MID_boot <- boot_R2$t\n",
    "R2_se_validation_adjusted_MID <- sd(boot_R2$t)\n",
    "\n",
    "boot_R2 <- boot(data = CV_RV_PRS_adjusted_MID, statistic = R2_Comparison_Boot, R = 10000)\n",
    "R2_comparison_adjusted_MID_boot <- as.data.frame(boot_R2$t)\n",
    "colnames(R2_comparison_adjusted_MID_boot) <- c(\"R2_adjusted_MID_RICE_vs_CTSLEB\",\"R2_adjusted_MID_RICE_vs_PROSPER\",\"R2_adjusted_MID_RICE_vs_JointPRS\")\n",
    "\n",
    "CV_PRS_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                             beta_raw = c(beta_CV_validation_raw_EUR,beta_CV_validation_raw_SAS,beta_CV_validation_raw_AMR,beta_CV_validation_raw_AFR,beta_CV_validation_raw_EAS,beta_CV_validation_raw_MID), \n",
    "                             beta_se_raw = c(beta_CV_se_validation_raw_EUR,beta_CV_se_validation_raw_SAS,beta_CV_se_validation_raw_AMR,beta_CV_se_validation_raw_AFR,beta_CV_se_validation_raw_EAS,beta_CV_se_validation_raw_MID), \n",
    "                             R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                             R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                             beta_adjusted = c(beta_CV_validation_adjusted_EUR,beta_CV_validation_adjusted_SAS,beta_CV_validation_adjusted_AMR,beta_CV_validation_adjusted_AFR,beta_CV_validation_adjusted_EAS,beta_CV_validation_adjusted_MID), \n",
    "                             beta_se_adjusted = c(beta_CV_se_validation_adjusted_EUR,beta_CV_se_validation_adjusted_SAS,beta_CV_se_validation_adjusted_AMR,beta_CV_se_validation_adjusted_AFR,beta_CV_se_validation_adjusted_EAS,beta_CV_se_validation_adjusted_MID), \n",
    "                             R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                             R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "CV_Boot_Results <- data.frame(trait = trait,beta_CV_raw_EUR_boot,R2_raw_EUR_boot,beta_CV_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_CV_raw_AMR_boot,R2_raw_AMR_boot,beta_CV_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_CV_raw_EAS_boot,R2_raw_EAS_boot,beta_CV_raw_MID_boot,R2_raw_MID_boot,beta_CV_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_CV_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_CV_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_CV_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_CV_adjusted_EAS_boot,R2_adjusted_EAS_boot,beta_CV_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "Comparison_Boot_Results <- data.frame(trait = trait,R2_comparison_raw_EUR_boot,R2_comparison_raw_SAS_boot,R2_comparison_raw_AMR_boot,R2_comparison_raw_AFR_boot,R2_comparison_raw_EAS_boot,R2_comparison_raw_MID_boot,\n",
    "                                      R2_comparison_adjusted_EUR_boot,R2_comparison_adjusted_SAS_boot,R2_comparison_adjusted_AMR_boot,R2_comparison_adjusted_AFR_boot,R2_comparison_adjusted_EAS_boot,R2_comparison_adjusted_MID_boot)\n",
    "\n",
    "RV_PRS_Results <- data.frame(trait = trait,ancestry = c(\"EUR\",\"SAS\",\"AMR\",\"AFR\",\"EAS\",\"MID\"), \n",
    "                             beta_raw = c(beta_RV_validation_raw_EUR,beta_RV_validation_raw_SAS,beta_RV_validation_raw_AMR,beta_RV_validation_raw_AFR,beta_RV_validation_raw_EAS,beta_RV_validation_raw_MID), \n",
    "                             beta_se_raw = c(beta_RV_se_validation_raw_EUR,beta_RV_se_validation_raw_SAS,beta_RV_se_validation_raw_AMR,beta_RV_se_validation_raw_AFR,beta_RV_se_validation_raw_EAS,beta_RV_se_validation_raw_MID), \n",
    "                             R2_raw = c(R2_validation_raw_EUR,R2_validation_raw_SAS,R2_validation_raw_AMR,R2_validation_raw_AFR,R2_validation_raw_EAS,R2_validation_raw_MID),\n",
    "                             R2_se_raw = c(R2_se_validation_raw_EUR,R2_se_validation_raw_SAS,R2_se_validation_raw_AMR,R2_se_validation_raw_AFR,R2_se_validation_raw_EAS,R2_se_validation_raw_MID),\n",
    "                             beta_adjusted = c(beta_RV_validation_adjusted_EUR,beta_RV_validation_adjusted_SAS,beta_RV_validation_adjusted_AMR,beta_RV_validation_adjusted_AFR,beta_RV_validation_adjusted_EAS,beta_RV_validation_adjusted_MID), \n",
    "                             beta_se_adjusted = c(beta_RV_se_validation_adjusted_EUR,beta_RV_se_validation_adjusted_SAS,beta_RV_se_validation_adjusted_AMR,beta_RV_se_validation_adjusted_AFR,beta_RV_se_validation_adjusted_EAS,beta_RV_se_validation_adjusted_MID), \n",
    "                             R2_adjusted = c(R2_validation_adjusted_EUR,R2_validation_adjusted_SAS,R2_validation_adjusted_AMR,R2_validation_adjusted_AFR,R2_validation_adjusted_EAS,R2_validation_adjusted_MID),\n",
    "                             R2_se_adjusted = c(R2_se_validation_adjusted_EUR,R2_se_validation_adjusted_SAS,R2_se_validation_adjusted_AMR,R2_se_validation_adjusted_AFR,R2_se_validation_adjusted_EAS,R2_se_validation_adjusted_MID))\n",
    "\n",
    "RV_Boot_Results <- data.frame(trait = trait,beta_RV_raw_EUR_boot,R2_raw_EUR_boot,beta_RV_raw_SAS_boot,R2_raw_SAS_boot,\n",
    "                              beta_RV_raw_AMR_boot,R2_raw_AMR_boot,beta_RV_raw_AFR_boot,R2_raw_AFR_boot,\n",
    "                              beta_RV_raw_EAS_boot,R2_raw_EAS_boot,beta_RV_raw_MID_boot,R2_raw_MID_boot,beta_RV_adjusted_EUR_boot,R2_adjusted_EUR_boot,\n",
    "                              beta_RV_adjusted_SAS_boot,R2_adjusted_SAS_boot,beta_RV_adjusted_AMR_boot,R2_adjusted_AMR_boot,\n",
    "                              beta_RV_adjusted_AFR_boot,R2_adjusted_AFR_boot,beta_RV_adjusted_EAS_boot,R2_adjusted_EAS_boot,beta_RV_adjusted_MID_boot,R2_adjusted_MID_boot)\n",
    "\n",
    "write.csv(Comparison_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",trait,\"_Comparison_Bootstraps.csv\"),row.names = FALSE)\n",
    "write.csv(CV_PRS_Results,file = paste0(OUTPUT_PATH,\"/\",\"CV_\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(CV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",\"CV_\",trait,\"_Bootstraps.csv\"),row.names = FALSE)\n",
    "write.csv(RV_PRS_Results,file = paste0(OUTPUT_PATH,\"/\",\"RV_\",trait,\"Best_Betas.csv\"),row.names = FALSE)\n",
    "write.csv(RV_Boot_Results,file = paste0(OUTPUT_PATH,\"/\",\"RV_\",trait,\"_Bootstraps.csv\"),row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Common_Plus_Rare.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Common_Plus_Rare.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${trait} ${Tune_Data} ${Validation_Data} ${CommonVariantPRS_Tune} ${RareVariantPRS_Tune} ${CommonVariantPRS_Validation} ${RareVariantPRS_Validation}  ${JointPRS}  ${PROSPER_PRS}  ${CTSLEB_PRS} ${all_phenotypes_file} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(trait in c(\"BMI\",\"LDL\",\"HDL\",\"logTG\",\"TC\",\"Height\")){\n",
    "  tasks <- rbind(tasks, data.frame(\n",
    "    '--env trait'=trait,\n",
    "    '--input Tune_Data'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Tune.txt\",\n",
    "    '--input Validation_Data'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/All_Validation.txt\",\n",
    "    '--input CommonVariantPRS_Tune'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",trait,\"_Best_Tune_All.txt\"),\n",
    "    '--input RareVariantPRS_Tune'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\",trait,\"_PRS_Tune.csv\"),\n",
    "    '--input CommonVariantPRS_Validation'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/OneCommonPRS/\",trait,\"_Best_Validation_All.txt\"),\n",
    "    '--input RareVariantPRS_Validation'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/RareVariant_PRS/\",trait,\"_PRS_Validation.csv\"),\n",
    "    '--input JointPRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/JointPRS/PRSs/PRS_META_\",trait,\".sscore\"),\n",
    "    '--input PROSPER_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/PROSPER/PRSs/PRS_\",trait,\".sscore\"),\n",
    "    '--input CTSLEB_PRS'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/CTSLEB/Results/\",trait,\"_Best_Validation_All.txt\"),\n",
    "    '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/Common_Plus_Rare.R\",\n",
    "    '--input all_phenotypes_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "    '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/Continuous/Common_Plus_Rare/\",\n",
    "    check.names = FALSE\n",
    "  ))   \n",
    "}\n",
    "\n",
    "\n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Common_Plus_Rare.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 27.6 KiB/ 27.6 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/27.6 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Common_Plus_Rare.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: common-plu--williamsjacr--250516-234338-22\n",
      "  job-name: common-plus-rare\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/7209186891319781713\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13861044095870257966\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/13852388308955964713\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/4893768840273172404\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1490482200950451986\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/12978605675624031067\n",
      "Launched job-id: common-plu--williamsjacr--250516-234338-22\n",
      "6 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-plu--williamsjacr--250516-234338-22' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'common-plu--williamsjacr--250516-234338-22' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Common_Plus_Rare.sh \\\n",
    "  --tasks score_task.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
